{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "load('recon3d_qflux.mat')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "15\tAlanine and aspartate metabolism\n",
      "5\tAlkaloid synthesis\n",
      "39\tAminoacyl-tRNA biosynthesis\n",
      "31\tAminosugar metabolism\n",
      "51\tAndrogen and estrogen synthesis and metabolism\n",
      "84\tArachidonic acid metabolism\n",
      "37\tArginine and proline metabolism\n",
      "22\tBeta-Alanine metabolism\n",
      "58\tBile acid synthesis\n",
      "2\tBiotin metabolism\n",
      "13\tBlood group synthesis\n",
      "11\tButanoate metabolism\n",
      "3\tC5-branched dibasic acid metabolism\n",
      "37\tCholesterol metabolism\n",
      "10\tChondroitin sulfate degradation\n",
      "41\tChondroitin synthesis\n",
      "28\tCitric acid cycle\n",
      "11\tCoA catabolism\n",
      "11\tCoA synthesis\n",
      "16\tCytochrome metabolism\n",
      "3\tD-alanine metabolism\n",
      "0\tDietary fiber binding\n",
      "55\tDrug metabolism\n",
      "97\tEicosanoid metabolism\n",
      "40\tExchange/demand reaction\n",
      "94\tFatty acid oxidation\n",
      "70\tFatty acid synthesis\n",
      "18\tFolate metabolism\n",
      "36\tFructose and mannose metabolism\n",
      "19\tGalactose metabolism\n",
      "21\tGlutamate metabolism\n",
      "25\tGlutathione metabolism\n",
      "163\tGlycerophospholipid metabolism\n",
      "30\tGlycine, serine, alanine, and threonine metabolism\n",
      "81\tGlycolysis/gluconeogenesis\n",
      "58\tGlycosphingolipid metabolism\n",
      "30\tGlyoxylate and dicarboxylate metabolism\n",
      "5\tHeme degradation\n",
      "13\tHeme synthesis\n",
      "6\tHeparan sulfate degradation\n",
      "0\tHippurate metabolism\n",
      "37\tHistidine metabolism\n",
      "7\tHyaluronan metabolism\n",
      "105\tInositol phosphate metabolism\n",
      "11\tKeratan sulfate degradation\n",
      "21\tKeratan sulfate synthesis\n",
      "64\tLeukotriene metabolism\n",
      "15\tLimonene and pinene degradation\n",
      "58\tLinoleate metabolism\n",
      "4\tLipoate metabolism\n",
      "52\tLysine metabolism\n",
      "41\tMethionine and cysteine metabolism\n",
      "1043\tMiscellaneous\n",
      "15\tN-glycan degradation\n",
      "17\tN-glycan metabolism\n",
      "36\tN-glycan synthesis\n",
      "35\tNAD metabolism\n",
      "115\tNucleotide interconversion\n",
      "69\tNucleotide metabolism\n",
      "2\tNucleotide salvage pathway\n",
      "2\tNucleotide sugar metabolism\n",
      "24\tO-glycan metabolism\n",
      "99\tOxidative phosphorylation\n",
      "26\tPentose phosphate pathway\n",
      "0\tPeptide metabolism\n",
      "36\tPhenylalanine metabolism\n",
      "40\tPhosphatidylinositol phosphate metabolism\n",
      "50\tPropanoate metabolism\n",
      "1\tProtein assembly\n",
      "18\tProtein degradation\n",
      "0\tProtein formation\n",
      "197\tProtein modification\n",
      "33\tPurine catabolism\n",
      "9\tPurine synthesis\n",
      "29\tPyrimidine catabolism\n",
      "22\tPyrimidine synthesis\n",
      "42\tPyruvate metabolism\n",
      "0\tR group synthesis\n",
      "9\tROS detoxification\n",
      "27\tSelenoamino acid metabolism\n",
      "125\tSphingolipid metabolism\n",
      "7\tSqualene and cholesterol synthesis\n",
      "31\tStarch and sucrose metabolism\n",
      "61\tSteroid metabolism\n",
      "2\tStilbene, coumarine and lignin synthesis\n",
      "6\tTaurine and hypotaurine metabolism\n",
      "19\tTetrahydrobiopterin metabolism\n",
      "5\tThiamine metabolism\n",
      "29\tTransport, endoplasmic reticular\n",
      "317\tTransport, extracellular\n",
      "28\tTransport, golgi apparatus\n",
      "36\tTransport, lysosomal\n",
      "61\tTransport, mitochondrial\n",
      "54\tTransport, nuclear\n",
      "25\tTransport, peroxisomal\n",
      "29\tTriacylglycerol synthesis\n",
      "0\tTriglycerides formation\n",
      "77\tTryptophan metabolism\n",
      "72\tTyrosine metabolism\n",
      "10\tUbiquinone synthesis\n",
      "26\tUrea cycle\n",
      "43\tValine, leucine, and isoleucine metabolism\n",
      "75\tVitamin A metabolism\n",
      "1\tVitamin B12 metabolism\n",
      "10\tVitamin B2 metabolism\n",
      "4\tVitamin B6 metabolism\n",
      "13\tVitamin C metabolism\n",
      "6\tVitamin D metabolism\n",
      "33\tVitamin E metabolism\n",
      "1\tVitamin K metabolism\n",
      "91\tXenobiotics metabolism\n"
     ]
    }
   ],
   "source": [
    "% subsystems\n",
    "unique_subsystems = unique(model.subSystems);\n",
    "for a = 1:length(unique_subsystems)\n",
    "    genes = {};\n",
    "    for i = 1:length(model.rxns)\n",
    "        if strcmp(model.subSystems{i},unique_subsystems{a})\n",
    "            for geneid = find(model.rxnGeneMat(i,:))\n",
    "                if ~any(strcmp(genes,model.geneSymbols{geneid}))\n",
    "                    genes{end+1} = model.geneSymbols{geneid};\n",
    "                end\n",
    "            end\n",
    "        end\n",
    "    end\n",
    "    fprintf('%d\\t%s\\n',length(genes),unique_subsystems{a})\n",
    "end"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "   134\n",
      "\n"
     ]
    }
   ],
   "source": [
    "% nadp+ to nadph (c or m)\n",
    "reactions = [];\n",
    "for i = 1:length(model.rxns)\n",
    "    if (full(model.S(strcmp(model.mets,'nadp[c]'),i)) < 0 && full(model.S(strcmp(model.mets,'nadph[c]'),i)) > 0 && model.ub(i) > 0) || (full(model.S(strcmp(model.mets,'nadp[c]'),i)) > 0 && full(model.S(strcmp(model.mets,'nadph[c]'),i)) < 0 && model.lb(i) < 0) || (full(model.S(strcmp(model.mets,'nadp[m]'),i)) < 0 && full(model.S(strcmp(model.mets,'nadph[m]'),i)) > 0 && model.ub(i) > 0) || (full(model.S(strcmp(model.mets,'nadp[m]'),i)) > 0 && full(model.S(strcmp(model.mets,'nadph[m]'),i)) < 0 && model.lb(i) < 0) \n",
    "        reactions(end+1) = i;\n",
    "    end\n",
    "end\n",
    "genes = {};\n",
    "for i = reactions\n",
    "    for geneid = find(model.rxnGeneMat(i,:))\n",
    "        if ~any(strcmp(genes,model.geneSymbols{geneid}))\n",
    "            genes{end+1} = model.geneSymbols{geneid};\n",
    "        end\n",
    "    end\n",
    "end\n",
    "disp(length(genes))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "   216\n",
      "\n"
     ]
    }
   ],
   "source": [
    "% nadph to nadp+ (c or m)\n",
    "reactions = [];\n",
    "for i = 1:length(model.rxns)\n",
    "    if (full(model.S(strcmp(model.mets,'nadp[c]'),i)) > 0 && full(model.S(strcmp(model.mets,'nadph[c]'),i)) < 0 && model.ub(i) > 0) || (full(model.S(strcmp(model.mets,'nadp[c]'),i)) < 0 && full(model.S(strcmp(model.mets,'nadph[c]'),i)) > 0 && model.lb(i) < 0) || (full(model.S(strcmp(model.mets,'nadp[m]'),i)) > 0 && full(model.S(strcmp(model.mets,'nadph[m]'),i)) < 0 && model.ub(i) > 0) || (full(model.S(strcmp(model.mets,'nadp[m]'),i)) < 0 && full(model.S(strcmp(model.mets,'nadph[m]'),i)) > 0 && model.lb(i) < 0) \n",
    "        reactions(end+1) = i;\n",
    "    end\n",
    "end\n",
    "genes = {};\n",
    "for i = reactions\n",
    "    for geneid = find(model.rxnGeneMat(i,:))\n",
    "        if ~any(strcmp(genes,model.geneSymbols{geneid}))\n",
    "            genes{end+1} = model.geneSymbols{geneid};\n",
    "        end\n",
    "    end\n",
    "end\n",
    "disp(length(genes))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "   134\n",
      "\n"
     ]
    }
   ],
   "source": [
    "% nad+ to nadh (c or m)\n",
    "reactions = [];\n",
    "for i = 1:length(model.rxns)\n",
    "    if (full(model.S(strcmp(model.mets,'nad[c]'),i)) < 0 && full(model.S(strcmp(model.mets,'nadh[c]'),i)) > 0 && model.ub(i) > 0) || (full(model.S(strcmp(model.mets,'nad[c]'),i)) > 0 && full(model.S(strcmp(model.mets,'nadh[c]'),i)) < 0 && model.lb(i) < 0) || (full(model.S(strcmp(model.mets,'nad[m]'),i)) < 0 && full(model.S(strcmp(model.mets,'nadh[m]'),i)) > 0 && model.ub(i) > 0) || (full(model.S(strcmp(model.mets,'nad[m]'),i)) > 0 && full(model.S(strcmp(model.mets,'nadh[m]'),i)) < 0 && model.lb(i) < 0) \n",
    "        reactions(end+1) = i;\n",
    "    end\n",
    "end\n",
    "genes = {};\n",
    "for i = reactions\n",
    "    for geneid = find(model.rxnGeneMat(i,:))\n",
    "        if ~any(strcmp(genes,model.geneSymbols{geneid}))\n",
    "            genes{end+1} = model.geneSymbols{geneid};\n",
    "        end\n",
    "    end\n",
    "end\n",
    "disp(length(genes))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "   150\n",
      "\n"
     ]
    }
   ],
   "source": [
    "% nadh to nad+ (c or m)\n",
    "reactions = [];\n",
    "for i = 1:length(model.rxns)\n",
    "    if (full(model.S(strcmp(model.mets,'nad[c]'),i)) > 0 && full(model.S(strcmp(model.mets,'nadh[c]'),i)) < 0 && model.ub(i) > 0) || (full(model.S(strcmp(model.mets,'nad[c]'),i)) < 0 && full(model.S(strcmp(model.mets,'nadh[c]'),i)) > 0 && model.lb(i) < 0) || (full(model.S(strcmp(model.mets,'nad[m]'),i)) > 0 && full(model.S(strcmp(model.mets,'nadh[m]'),i)) < 0 && model.ub(i) > 0) || (full(model.S(strcmp(model.mets,'nad[m]'),i)) < 0 && full(model.S(strcmp(model.mets,'nadh[m]'),i)) > 0 && model.lb(i) < 0) \n",
    "        reactions(end+1) = i;\n",
    "    end\n",
    "end\n",
    "genes = {};\n",
    "for i = reactions\n",
    "    for geneid = find(model.rxnGeneMat(i,:))\n",
    "        if ~any(strcmp(genes,model.geneSymbols{geneid}))\n",
    "            genes{end+1} = model.geneSymbols{geneid};\n",
    "        end\n",
    "    end\n",
    "end\n",
    "disp(length(genes))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "genes =\n",
      "\n",
      "  1x134 cell array\n",
      "\n",
      "  Columns 1 through 6\n",
      "\n",
      "    {'LDHAL6A'}    {'LDHA'}    {'LDHB'}    {'LDHC'}    {'UEVLD'}    {'LDHAL6B'}\n",
      "\n",
      "  Columns 7 through 12\n",
      "\n",
      "    {'DLD'}    {'DLST'}    {'OGDH'}    {'PDHX'}    {'ADH1A'}    {'ADH1B'}\n",
      "\n",
      "  Columns 13 through 18\n",
      "\n",
      "    {'ADH1C'}    {'ADH4'}    {'ADH5'}    {'ADH6'}    {'ADH7'}    {'ADHFE1'}\n",
      "\n",
      "  Columns 19 through 23\n",
      "\n",
      "    {'ALDH3A1'}    {'ALDH1A3'}    {'ALDH3B1'}    {'ALDH3B2'}    {'ALDH4A1'}\n",
      "\n",
      "  Columns 24 through 28\n",
      "\n",
      "    {'ALDH2'}    {'ALDH1B1'}    {'ZADH2'}    {'ALDH1A1'}    {'ALDH9A1'}\n",
      "\n",
      "  Columns 29 through 33\n",
      "\n",
      "    {'ALDH3A2'}    {'ALDH7A1'}    {'ALDH1A2'}    {'BDH1'}    {'ACADM'}\n",
      "\n",
      "  Columns 34 through 38\n",
      "\n",
      "    {'ACADS'}    {'ALDH1L1'}    {'ALDH6A1'}    {'ALDH18A1'}    {'ALDH5A1'}\n",
      "\n",
      "  Columns 39 through 44\n",
      "\n",
      "    {'GCSH'}    {'GLDC'}    {'AMT'}    {'GLUD1'}    {'GLUD2'}    {'HSD17B10'}\n",
      "\n",
      "  Columns 45 through 49\n",
      "\n",
      "    {'HADHA'}    {'HADHB'}    {'HADH'}    {'EHHADH'}    {'HIBADH'}\n",
      "\n",
      "  Columns 50 through 55\n",
      "\n",
      "    {'HSD3B1'}    {'IDH3A'}    {'IDH3B'}    {'IDH3G'}    {'LDHD'}    {'MAOB'}\n",
      "\n",
      "  Columns 56 through 61\n",
      "\n",
      "    {'MDH1'}    {'MDH2'}    {'ME2'}    {'MTHFD2'}    {'DBT'}    {'BCKDHA'}\n",
      "\n",
      "  Columns 62 through 67\n",
      "\n",
      "    {'BCKDHB'}    {'DLAT'}    {'PDHA1'}    {'PDHA2'}    {'PDHB'}    {'PHGDH'}\n",
      "\n",
      "  Columns 68 through 73\n",
      "\n",
      "    {'PRODH'}    {'RDH5'}    {'RDH16'}    {'SORD'}    {'MDH1B'}    {'UGDH'}\n",
      "\n",
      "  Columns 74 through 79\n",
      "\n",
      "    {'GSR'}    {'XDH'}    {'CMAHP'}    {'IDH1'}    {'IDH2'}    {'DHFR'}\n",
      "\n",
      "  Columns 80 through 84\n",
      "\n",
      "    {'AASS'}    {'PYCR2'}    {'PYCR1'}    {'HSD17B4'}    {'HSD3B2'}\n",
      "\n",
      "  Columns 85 through 90\n",
      "\n",
      "    {'CYP27A1'}    {'CBR1'}    {'CBR3'}    {'DHRS4'}    {'ACAA2'}    {'ECHS1'}\n",
      "\n",
      "  Columns 91 through 96\n",
      "\n",
      "    {'ACADVL'}    {'HPGD'}    {'GPD1'}    {'GAPDH'}    {'GAPDHS'}    {'IMPDH1'}\n",
      "\n",
      "  Columns 97 through 101\n",
      "\n",
      "    {'IMPDH2'}    {'PRODH2'}    {'MTHFR'}    {'MTHFD2P1'}    {'MTHFD2L'}\n",
      "\n",
      "  Columns 102 through 106\n",
      "\n",
      "    {'ACOT8'}    {'ACOT2'}    {'SLC27A4'}    {'SLC27A3'}    {'SLC27A2'}\n",
      "\n",
      "  Columns 107 through 112\n",
      "\n",
      "    {'ACOT7'}    {'ACOT4'}    {'AASDH'}    {'ACSF3'}    {'AMACR'}    {'ACAA1'}\n",
      "\n",
      "  Columns 113 through 117\n",
      "\n",
      "    {'LOC344967'}    {'SLC27A1'}    {'ACOX1'}    {'BAAT'}    {'ACOT1'}\n",
      "\n",
      "  Columns 118 through 122\n",
      "\n",
      "    {'ACSF2'}    {'ACOX3'}    {'PTGR1'}    {'TECRL'}    {'HSD17B12'}\n",
      "\n",
      "  Columns 123 through 128\n",
      "\n",
      "    {'AKR1C3'}    {'NSDHL'}    {'HSD3B7'}    {'ACAD9'}    {'ACADL'}    {'CHDH'}\n",
      "\n",
      "  Columns 129 through 134\n",
      "\n",
      "    {'ECI2'}    {'ECI1'}    {'DECR1'}    {'DECR2'}    {'NQO1'}    {'NQO2'}\n",
      "\n"
     ]
    }
   ],
   "source": [
    "genes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Matlab",
   "language": "matlab",
   "name": "matlab"
  },
  "language_info": {
   "codemirror_mode": "octave",
   "file_extension": ".m",
   "help_links": [
    {
     "text": "MetaKernel Magics",
     "url": "https://github.com/calysto/metakernel/blob/master/metakernel/magics/README.md"
    }
   ],
   "mimetype": "text/x-octave",
   "name": "matlab",
   "version": "0.15.1"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
