{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Flux Balance Analysis"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Necessary User Input"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Name of input folder (output folder will have same name)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "% input folder name\n",
    "input_folder = 'Test';\n",
    "\n",
    "% signify if gene knockdown screen\n",
    "% if True, will take all settings (besides gene knockdown) from first file in input folder, and use these for all subsequent files\n",
    "gene_knockdown_screen = false;\n",
    "\n",
    "% signify if objective function screen\n",
    "% if True, will take all settings (besides objective function) from first file in input folder, and use these for all subsequent files\n",
    "objective_function_screen = false;\n",
    "\n",
    "% number of parallel cores\n",
    "parallel_cores = 1;\n",
    "\n",
    "% look for already-completed input files\n",
    "look_for_completed = false;\n",
    "completed_output_folder = 'RAD_knockout_nadh';"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Get all input files"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Look for already-completed files"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "% don't do already-completed files\n",
    "if look_for_completed\n",
    "    ds = {sprintf('results/%s',completed_output_folder)};\n",
    "    overlap = {};\n",
    "    for i = 1:length(ds)\n",
    "        directory = dir(ds{i});\n",
    "        directory = directory(3:end);\n",
    "        for j = 1:length(directory)\n",
    "            overlap{end+1} = directory(j).name;\n",
    "        end\n",
    "    end\n",
    "else\n",
    "    overlap = {};\n",
    "end"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Get input files"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "directory = dir(sprintf('input/%s/*.xlsx',input_folder));\n",
    "input_files = {};\n",
    "for i = 1:length(directory)\n",
    "    fn = sprintf('input/%s/%s',input_folder,directory(i).name);\n",
    "    test = strsplit(fn,'/');\n",
    "    if ~strcmp(extractBetween(test{end},1,2),'~$')\n",
    "        temp = strsplit(test{end},'.xlsx');\n",
    "        if ~any(strcmp(overlap,temp{1}))\n",
    "            input_files{end+1} = fn;\n",
    "        end\n",
    "    end\n",
    "end"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Initialize output folder"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "% name output folder\n",
    "if ~(exist(sprintf('results/%s',input_folder)) == 7)\n",
    "    output_folder = input_folder;\n",
    "else\n",
    "    num = 1;\n",
    "    while exist(sprintf('results/%s',sprintf('%s_%d',input_folder,num))) == 7\n",
    "        num = num + 1;\n",
    "    end\n",
    "    output_folder = sprintf('%s_%d',input_folder,num);\n",
    "end\n",
    "\n",
    "% create output folder\n",
    "mkdir(sprintf('results/%s',output_folder))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Parse Input File"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### General"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%file parse_general.m\n",
    "\n",
    "function [type_of_analysis,pfba_fraction,fva_all_fraction,fva_select_fraction,fva_select_groups,fva_select_reactions,sampler_fraction,sampler_samples,sampler_steps,media_sensitivity_critical_fraction,media_sensitivity_critical_metabolite1,media_sensitivity_spectrum_fraction,media_sensitivity_spectrum_metabolite1,media_sensitivity_spectrum_metabolite2,media_sensitivity_spectrum_points] = parse_general(input_file)\n",
    "\n",
    "    % load data\n",
    "    [~,~,raw] = xlsread(input_file,'General');\n",
    "\n",
    "    % type of analysis\n",
    "    for row = 2:8\n",
    "        value = raw(row,1);\n",
    "        value = value{1};\n",
    "        if ~isnan(value)\n",
    "            if ischar(value)\n",
    "                if strcmpi(value,'X')\n",
    "                    if exist('type_of_analysis','var')\n",
    "                        error('ERROR - General - Type of Analysis - Can only choose one type of analysis')\n",
    "                    else\n",
    "                        type_of_analysis = row - 1;\n",
    "                    end\n",
    "                else\n",
    "                    error('ERROR - General - Type of Analysis - Values must either be X or blank')\n",
    "                end\n",
    "            else\n",
    "                error('ERROR - General - Type of Analysis - Values must either be X or blank')\n",
    "            end\n",
    "        end\n",
    "    end\n",
    "    if ~exist('type_of_analysis','var')\n",
    "        error('ERROR - General - Type of Analysis - Must select one type of analysis')\n",
    "    end\n",
    "    \n",
    "    % options - pfba\n",
    "    if type_of_analysis == 2\n",
    "    \n",
    "        % fraction of maximum objective value\n",
    "        value = raw(11,1);\n",
    "        value = value{1};\n",
    "        if ~isnan(value)\n",
    "            if isfloat(value)\n",
    "                if (value >= 0) && (value <= 1)\n",
    "                    pfba_fraction = value;\n",
    "                else\n",
    "                    error('ERROR - General - Options - pFBA - Fraction of maximum objective value must either be >= 0 and <= 1')\n",
    "                end\n",
    "            else\n",
    "                error('ERROR - General - Options - pFBA - Fraction of maximum objective value must either be >= 0 and <= 1')\n",
    "            end\n",
    "        else\n",
    "            error('ERROR - General - Options - pFBA - Must enter fraction of maximum objective value')\n",
    "        end\n",
    "        if ~exist('pfba_fraction','var')\n",
    "            error('ERROR - General - Options - pFBA - Must enter fraction of maximum objective value')\n",
    "        end \n",
    "    \n",
    "    else\n",
    "        pfba_fraction = NaN; \n",
    "    end\n",
    "    \n",
    "    % options - fva - all reactions\n",
    "    if type_of_analysis == 3\n",
    "    \n",
    "        % fraction of maximum objective value\n",
    "        value = raw(14,1);\n",
    "        value = value{1};\n",
    "        if ~isnan(value)\n",
    "            if isfloat(value)\n",
    "                if (value >= 0) && (value <= 1)\n",
    "                    fva_all_fraction = value;\n",
    "                else\n",
    "                    error('ERROR - General - Options - FVA - All reactions - Fraction of maximum objective value must either be >= 0 and <= 1')\n",
    "                end\n",
    "            else\n",
    "                error('ERROR - General - Options - FVA - All reactions - Fraction of maximum objective value must either be >= 0 and <= 1')\n",
    "            end\n",
    "        else\n",
    "            error('ERROR - General - Options - FVA - All reactions - Must enter fraction of maximum objective value')\n",
    "        end\n",
    "        if ~exist('fva_all_fraction','var')\n",
    "            error('ERROR - General - Options - FVA - All reactions - Must enter fraction of maximum objective value')\n",
    "        end \n",
    "        \n",
    "    else    \n",
    "        fva_all_fraction = NaN;      \n",
    "    end\n",
    "    \n",
    "    % options - fva - select reactions\n",
    "    if type_of_analysis == 4\n",
    "        \n",
    "        % fraction of maximum objective value\n",
    "        value = raw(17,1);\n",
    "        value = value{1};\n",
    "        if ~isnan(value)\n",
    "            if isfloat(value)\n",
    "                if (value >= 0) && (value <= 1)\n",
    "                    fva_select_fraction = value;\n",
    "                else\n",
    "                    error('ERROR - General - Options - FVA - Select reactions - Fraction of maximum objective value must either be >= 0 and <= 1')\n",
    "                end\n",
    "            else\n",
    "                error('ERROR - General - Options - FVA - Select reactions - Fraction of maximum objective value must either be >= 0 and <= 1')\n",
    "            end\n",
    "        else\n",
    "            error('ERROR - General - Options - FVA - Select reactions - Must enter fraction of maximum objective value')\n",
    "        end\n",
    "        if ~exist('fva_select_fraction','var')\n",
    "            error('ERROR - General - Options - FVA - Select reactions - Must enter fraction of maximum objective value')\n",
    "        end\n",
    "        \n",
    "        % reaction groups\n",
    "        fva_select_groups = [];\n",
    "        for row = 19:23\n",
    "            value = raw(row,1);\n",
    "            value = value{1};\n",
    "            if ~isnan(value)\n",
    "                if ischar(value)\n",
    "                    if strcmpi(value,'X')\n",
    "                        fva_select_groups(end+1) = row - 18;\n",
    "                    else\n",
    "                        error('ERROR - General - Options - FVA - Select groups - Values must either be X or blank')\n",
    "                    end\n",
    "                else\n",
    "                    error('ERROR - General - Options - FVA - Select groups - Values must either be X or blank')\n",
    "                end\n",
    "            end\n",
    "        end\n",
    "\n",
    "        % reaction list\n",
    "        fva_select_reactions = {};\n",
    "        for row = 25:size(raw,1)\n",
    "            value = raw(row,2);\n",
    "            value = value{1};\n",
    "            if ~isnan(value)\n",
    "                if ischar(value)\n",
    "                    fva_select_reactions{end+1} = value;\n",
    "                else\n",
    "                    error('ERROR - General - Options - FVA - Select reactions - Reaction name must be a string')\n",
    "                end\n",
    "            end\n",
    "        end\n",
    "        if (length(fva_select_groups) == 0) && (length(fva_select_reactions) == 0)\n",
    "            error('ERROR - General - Options - FVA - Select reactions - Must provide at least one reaction group or name')\n",
    "        end\n",
    "        \n",
    "    else\n",
    "        fva_select_fraction = NaN;\n",
    "        fva_select_groups = [];\n",
    "        fva_select_reactions = {};\n",
    "    end\n",
    "    \n",
    "    % options - uniform sampling\n",
    "    if type_of_analysis == 5\n",
    "        \n",
    "        % fraction of maximum objective value\n",
    "        value = raw(11,5);\n",
    "        value = value{1};\n",
    "        if ~isnan(value)\n",
    "            if isfloat(value)\n",
    "                if (value >= 0) && (value <= 1)\n",
    "                    sampler_fraction = value;\n",
    "                else\n",
    "                    error('ERROR - General - Options - Uniform sampling - Fraction of maximum objective value must either be >= 0 and <= 1')\n",
    "                end\n",
    "            else\n",
    "                error('ERROR - General - Options - Uniform sampling - Fraction of maximum objective value must either be >= 0 and <= 1')\n",
    "            end\n",
    "        else\n",
    "            error('ERROR - General - Options - Uniform sampling - Must enter fraction of maximum objective value')\n",
    "        end\n",
    "        if ~exist('sampler_fraction','var')\n",
    "            error('ERROR - General - Options - Uniform sampling - Must enter fraction of maximum objective value')\n",
    "        end\n",
    "        \n",
    "        % number of samples\n",
    "        value = raw(12,5);\n",
    "        value = value{1};\n",
    "        if ~isnan(value)\n",
    "            if isfloat(value)\n",
    "                if mod(value,1) == 0\n",
    "                    if value >= 1\n",
    "                        sampler_samples = value;\n",
    "                    else\n",
    "                        error('ERROR - General - Options - Uniform sampling - Number of Samples - Value must be >= 1')\n",
    "                    end\n",
    "                else\n",
    "                    error('ERROR - General - Options - Uniform sampling - Number of Samples - Value must be an integer')\n",
    "                end\n",
    "            else\n",
    "                error('ERROR - General - Options - Uniform sampling - Number of Samples - Value must be a number')\n",
    "            end\n",
    "        end\n",
    "        if ~exist('sampler_samples','var')\n",
    "            error('ERROR - General - Options - Uniform sampling - Number of Samples - Must enter value')\n",
    "        end\n",
    "\n",
    "        % number of steps per sample\n",
    "        value = raw(13,5);\n",
    "        value = value{1};\n",
    "        if ~isnan(value)\n",
    "            if isfloat(value)\n",
    "                if mod(value,1) == 0\n",
    "                    if value >= 1\n",
    "                        sampler_steps = value;\n",
    "                    else\n",
    "                        error('ERROR - General - Options - Uniform sampling - Number of Steps per Sample - Value must be >= 1')\n",
    "                    end\n",
    "                else\n",
    "                    error('ERROR - General - Options - Uniform sampling - Number of Steps per Sample - Value must be an integer')\n",
    "                end\n",
    "            else\n",
    "                error('ERROR - General - Options - Uniform sampling - Number of Steps per Sample - Value must be a number')\n",
    "            end\n",
    "        end\n",
    "        if ~exist('sampler_steps','var')\n",
    "            error('ERROR - General - Options - Uniform sampling - Number of Steps per Sample - Must enter value')\n",
    "        end\n",
    "\n",
    "    else\n",
    "        sampler_fraction = NaN;\n",
    "        sampler_samples = NaN;\n",
    "        sampler_steps = NaN;\n",
    "    end\n",
    "    \n",
    "    % options - media sensitivity - critical value\n",
    "    if type_of_analysis == 6\n",
    "        \n",
    "        % fraction of maximum objective value\n",
    "        value = raw(16,5);\n",
    "        value = value{1};\n",
    "        if ~isnan(value)\n",
    "            if isfloat(value)\n",
    "                if (value >= 0) && (value <= 1)\n",
    "                    media_sensitivity_critical_fraction = value;\n",
    "                else\n",
    "                    error('ERROR - General - Options - Media sensitivity - Critical value - Fraction of maximum objective value must either be >= 0 and <= 1')\n",
    "                end\n",
    "            else\n",
    "                error('ERROR - General - Options - Media sensitivity - Critical value - Fraction of maximum objective value must either be >= 0 and <= 1')\n",
    "            end\n",
    "        else\n",
    "            error('ERROR - General - Options - Media sensitivity - Critical values - Must enter fraction of maximum objective value')\n",
    "        end\n",
    "        if ~exist('media_sensitivity_critical_fraction','var')\n",
    "            error('ERROR - General - Options - Media sensitivity - Critical value - Must enter fraction of maximum objective value')\n",
    "        end\n",
    "        \n",
    "        % metabolite 1\n",
    "        value = raw(17,5);\n",
    "        value = value{1};\n",
    "        if ~isnan(value)\n",
    "            if ischar(value)\n",
    "                media_sensitivity_critical_metabolite1 = value;\n",
    "            else\n",
    "                error('ERROR - General - Options - Media sensitivity - Critical value - Metabolite - Metabolite 1 name must be a string')\n",
    "            end \n",
    "        end\n",
    "        if ~exist('media_sensitivity_critical_metabolite1','var')\n",
    "            error('ERROR - General - Options - Media sensitivity - Critical value - Metabolite - Must enter metabolite 1 name')\n",
    "        end\n",
    "    \n",
    "    else\n",
    "        media_sensitivity_critical_fraction = NaN;\n",
    "        media_sensitivity_critical_metabolite1 = NaN;    \n",
    "    end\n",
    "    \n",
    "    % options - media sensitivity - spectrum\n",
    "    if type_of_analysis == 7\n",
    "        \n",
    "        % fraction of maximum objective value\n",
    "        value = raw(20,5);\n",
    "        value = value{1};\n",
    "        if ~isnan(value)\n",
    "            if isfloat(value)\n",
    "                if (value >= 0) && (value <= 1)\n",
    "                    media_sensitivity_spectrum_fraction = value;\n",
    "                else\n",
    "                    error('ERROR - General - Options - Media sensitivity - Spectrum - Fraction of maximum objective value must either be >= 0 and <= 1')\n",
    "                end\n",
    "            else\n",
    "                error('ERROR - General - Options - Media sensitivity - Spectrum - Fraction of maximum objective value must either be >= 0 and <= 1')\n",
    "            end\n",
    "        else\n",
    "            error('ERROR - General - Options - Media sensitivity - Spectrum - Must enter fraction of maximum objective value')\n",
    "        end\n",
    "        if ~exist('media_sensitivity_spectrum_fraction','var')\n",
    "            error('ERROR - General - Options - Media sensitivity - Spectrum - Must enter fraction of maximum objective value')\n",
    "        end\n",
    "        \n",
    "        % metabolite 1\n",
    "        value = raw(21,5);\n",
    "        value = value{1};\n",
    "        if ~isnan(value)\n",
    "            if ischar(value)\n",
    "                media_sensitivity_spectrum_metabolite1 = value;\n",
    "            else\n",
    "                error('ERROR - General - Options - Media sensitivity - Spectrum - Metabolite - Metabolite 1 name must be a string')\n",
    "            end \n",
    "        end\n",
    "        if ~exist('media_sensitivity_spectrum_metabolite1','var')\n",
    "            error('ERROR - General - Options - Media sensitivity - Spectrum - Metabolite - Must enter metabolite 1 name')\n",
    "        end\n",
    "        \n",
    "        % metabolite 2\n",
    "        value = raw(22,5);\n",
    "        value = value{1};\n",
    "        if ~isnan(value)\n",
    "            if ischar(value)\n",
    "                media_sensitivity_spectrum_metabolite2 = value;\n",
    "            else\n",
    "                error('ERROR - General - Options - Media sensitivity - Spectrum - Metabolite - Metabolite 2 name must be a string')\n",
    "            end \n",
    "        end\n",
    "        if ~exist('media_sensitivity_spectrum_metabolite2','var')\n",
    "            media_sensitivity_spectrum_metabolite2 = NaN;\n",
    "        end\n",
    "        \n",
    "        % number of concentration points\n",
    "        value = raw(23,5);\n",
    "        value = value{1};\n",
    "        if ~isnan(value)\n",
    "            if isfloat(value)\n",
    "                if mod(value,1) == 0\n",
    "                    if value >= 1\n",
    "                        media_sensitivity_spectrum_points = value;\n",
    "                    else\n",
    "                        error('ERROR - General - Options - Media sensitivity - Spectrum - Number of Concentration Points - Value must be >= 1')\n",
    "                    end\n",
    "                else\n",
    "                    error('ERROR - General - Options - Media sensitivity - Spectrum - Number of Concentration Points - Value must be an integer')\n",
    "                end\n",
    "            else\n",
    "                error('ERROR - General - Options - Media sensitivity - Spectrum - Number of Concentration Points - Value must be a number')\n",
    "            end\n",
    "        end\n",
    "        if ~exist('media_sensitivity_spectrum_points','var')\n",
    "            error('ERROR - General - Options - Media sensitivity - Spectrum - Number of Concentration Points - Must enter value')\n",
    "        end\n",
    "    \n",
    "    else\n",
    "        media_sensitivity_spectrum_fraction = NaN;\n",
    "        media_sensitivity_spectrum_metabolite1 = NaN;\n",
    "        media_sensitivity_spectrum_metabolite2 = NaN;\n",
    "        media_sensitivity_spectrum_points = NaN;\n",
    "    end  \n",
    "end "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "###  Constraints - Proliferation"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%file parse_constraints_proliferation.m\n",
    "\n",
    "function [constraints_proliferation,constraints_proliferation_type,constraints_proliferation_value] = parse_constraints_proliferation(input_file)\n",
    "   \n",
    "    % load data\n",
    "    [~,~,raw] = xlsread(input_file,'Constraints - Proliferation');\n",
    "\n",
    "    % proliferation constraint\n",
    "    row = 3;\n",
    "    \n",
    "    % constraint type - equal\n",
    "    value = raw(row,1);\n",
    "    value = value{1};\n",
    "    if ~isnan(value)\n",
    "        if isfloat(value)           \n",
    "            if value >= 0\n",
    "                constraints_proliferation = 1;\n",
    "                constraints_proliferation_type = 'equal';\n",
    "                constraints_proliferation_value = value;\n",
    "            else\n",
    "                error('ERROR - Constraints - Proliferation - Constraint fraction must be >= 0')\n",
    "            end\n",
    "        else\n",
    "            error('ERROR - Constraints - Proliferation - Constraint fraction must be a number')\n",
    "        end\n",
    "    end\n",
    "\n",
    "    % constraint type - less than\n",
    "    value = raw(row,2);\n",
    "    value = value{1};\n",
    "    if ~isnan(value)\n",
    "        if ~exist('constraints_proliferation_type','var')\n",
    "            if isfloat(value)           \n",
    "                if value >= 0\n",
    "                    constraints_proliferation = 1;\n",
    "                    constraints_proliferation_type = 'less';\n",
    "                    constraints_proliferation_value = value;\n",
    "                else\n",
    "                    error('ERROR - Constraints - Proliferation - Constraint fraction must be >= 0')\n",
    "                end\n",
    "            else\n",
    "                error('ERROR - Constraints - Proliferation - Constraint fraction must be a number')\n",
    "            end\n",
    "        else\n",
    "            error('ERROR - Constraints - Proliferation - Can only choose one constraint type')\n",
    "        end\n",
    "    end\n",
    "\n",
    "    % constraint type - greater than\n",
    "    value = raw(row,3);\n",
    "    value = value{1};\n",
    "    if ~isnan(value)\n",
    "        if ~exist('constraints_proliferation_type','var')\n",
    "            if isfloat(value)           \n",
    "                if value >= 0\n",
    "                    constraints_proliferation = 1;\n",
    "                    constraints_proliferation_type = 'greater';\n",
    "                    constraints_proliferation_value = value;\n",
    "                else\n",
    "                    error('ERROR - Constraints - Proliferation - Constraint fraction must be >= 0')\n",
    "                end\n",
    "            else\n",
    "                error('ERROR - Constraints - Proliferation - Constraint fraction must be a number')\n",
    "            end\n",
    "        else\n",
    "            error('ERROR - Constraints - Proliferation - Can only choose one constraint type')\n",
    "        end\n",
    "    end\n",
    "\n",
    "    % constraint type - range around - fraction\n",
    "    value1 = raw(row,4);\n",
    "    value1 = value1{1};\n",
    "    value2 = raw(row,5);\n",
    "    value2 = value2{1};\n",
    "    if ~isnan(value1) && ~isnan(value2)\n",
    "        if ~exist('constraints_proliferation_type','var')\n",
    "            if isfloat(value1) && isfloat(value2)       \n",
    "                if (value1 >= 0) && (value2 >= 0)\n",
    "                    constraints_proliferation = 1;\n",
    "                    constraints_proliferation_type = 'range|fraction';\n",
    "                    constraints_proliferation_value = [value1,value2];\n",
    "                else\n",
    "                    error('ERROR - Constraints - Proliferation - Constraint median and range must both be >= 0')\n",
    "                end\n",
    "            else\n",
    "                error('ERROR - Constraints - Proliferation - Constraint median and range must both be a number')\n",
    "            end\n",
    "        else\n",
    "            error('ERROR - Constraints - Proliferation - Can only choose one constraint type')\n",
    "        end\n",
    "    elseif ~isnan(value1) || ~isnan(value2)\n",
    "        error('ERROR - Constraints - Proliferation - If choosing \"Range Around\" constraint, must enter both median and range/distance values');\n",
    "    end\n",
    "\n",
    "    % constraint type - range around - distance\n",
    "    value1 = raw(row,6);\n",
    "    value1 = value1{1};\n",
    "    value2 = raw(row,7);\n",
    "    value2 = value2{1};\n",
    "    if ~isnan(value1) && ~isnan(value2)\n",
    "        if ~exist('constraints_proliferation_type','var')\n",
    "            if isfloat(value1) && isfloat(value2)       \n",
    "                if (value1 >= 0) && (value2 >= 0)\n",
    "                    constraints_proliferation = 1;\n",
    "                    constraints_proliferation_type = 'range|distance';\n",
    "                    constraints_proliferation_value = [value1,value2];\n",
    "                else\n",
    "                    error('ERROR - Constraints - Proliferation - Constraint median and range must both be >= 0')\n",
    "                end\n",
    "            else\n",
    "                error('ERROR - Constraints - Proliferation - Constraint median and range must both be a number')\n",
    "            end\n",
    "        else\n",
    "            error('ERROR - Constraints - Proliferation - Can only choose one constraint type')\n",
    "        end\n",
    "    elseif ~isnan(value1) || ~isnan(value2)\n",
    "        error('ERROR - Constraints - Proliferation - If choosing \"Range Around\" constraint, must enter both median and range/distance values');\n",
    "    end\n",
    "\n",
    "    % none\n",
    "    if ~exist('constraints_proliferation_type','var')\n",
    "        constraints_proliferation = 0;\n",
    "        constraints_proliferation_type = NaN;\n",
    "        constraints_proliferation_value = NaN;\n",
    "    end\n",
    "end"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Constraints - Value"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%file parse_constraints_value.m\n",
    "\n",
    "function [constraints_value_type,constraints_value_value,constraints_value_reaction,constraints_value_count] = parse_constraints_value(input_file)\n",
    "  \n",
    "    % load data\n",
    "    [~,~,raw] = xlsread(input_file,'Constraints - Value');\n",
    "\n",
    "    % value constraints\n",
    "    constraints_value_type = {};\n",
    "    constraints_value_value = {};\n",
    "    constraints_value_reaction = {};\n",
    "    constraints_value_count = 0;\n",
    "\n",
    "    for row = 3:size(raw,1)\n",
    "        added = false;\n",
    "        \n",
    "        % constraint type - equal\n",
    "        value = raw(row,1);\n",
    "        value = value{1};\n",
    "        if ~isnan(value)\n",
    "            if isfloat(value)           \n",
    "                if value >= 0\n",
    "                    added = true;\n",
    "                    constraints_value_type{end+1} = 'equal';\n",
    "                    constraints_value_value{end+1} = value;\n",
    "                else\n",
    "                    error('ERROR - Constraints - Value - Constraint value must be >= 0')\n",
    "                end\n",
    "            else\n",
    "                error('ERROR - Constraints - Value - Constraint value must be a number')\n",
    "            end\n",
    "        end\n",
    "\n",
    "        % constraint type - less than\n",
    "        value = raw(row,2);\n",
    "        value = value{1};\n",
    "        if ~isnan(value)\n",
    "            if ~added;\n",
    "                if isfloat(value)           \n",
    "                    if value >= 0\n",
    "                        added = true;\n",
    "                        constraints_value_type{end+1} = 'less';\n",
    "                        constraints_value_value{end+1} = value;\n",
    "                    else\n",
    "                        error('ERROR - Constraints - Value - Constraint value must be >= 0')\n",
    "                    end\n",
    "                else\n",
    "                    error('ERROR - Constraints - Value - Constraint value must be a number')\n",
    "                end\n",
    "            else\n",
    "                error('ERROR - Constraints - Value - Can only choose one constraint type at a time')\n",
    "            end\n",
    "        end\n",
    "\n",
    "        % constraint type - greater than\n",
    "        value = raw(row,3);\n",
    "        value = value{1};\n",
    "        if ~isnan(value)\n",
    "            if ~added;\n",
    "                if isfloat(value)           \n",
    "                    if value >= 0\n",
    "                        added = true;\n",
    "                        constraints_value_type{end+1} = 'greater';\n",
    "                        constraints_value_value{end+1} = value;\n",
    "                    else\n",
    "                        error('ERROR - Constraints - Value - Constraint value must be >= 0')\n",
    "                    end\n",
    "                else\n",
    "                    error('ERROR - Constraints - Value - Constraint value must be a number')\n",
    "                end\n",
    "            else\n",
    "                error('ERROR - Constraints - Value - Can only choose one constraint type at a time')\n",
    "            end\n",
    "        end\n",
    "\n",
    "        % constraint type - range\n",
    "        value1 = raw(row,4);\n",
    "        value1 = value1{1};\n",
    "        value2 = raw(row,5);\n",
    "        value2 = value2{1};\n",
    "        if ~isnan(value1) && ~isnan(value2)\n",
    "            if ~added;\n",
    "                if isfloat(value1) && isfloat(value2)       \n",
    "                    if (value1 >= 0) && (value2 >= 0)\n",
    "                        if value1 <= value2\n",
    "                            added = true;\n",
    "                            constraints_value_type{end+1} = 'range';\n",
    "                            constraints_value_value{end+1} = [value1,value2];\n",
    "                        else\n",
    "                            error('ERROR - Constraints - Value - Minimum value must be <= maximum value')\n",
    "                        end\n",
    "                    else\n",
    "                        error('ERROR - Constraints - Value - Constraint values must be >= 0')\n",
    "                    end\n",
    "                else\n",
    "                    error('ERROR - Constraints - Value - Constraint values must be a number')\n",
    "                end\n",
    "            else\n",
    "                error('ERROR - Constraints - Value - Can only choose one constraint type at a time')\n",
    "            end\n",
    "        elseif ~isnan(value1) || ~isnan(value2)\n",
    "            error('ERROR - Constraints - Value - If choosing \"Range\" constraint, must enter both minimum and maximum values');\n",
    "        end\n",
    "\n",
    "        % reaction formula\n",
    "        if added\n",
    "            value = raw(row,6);\n",
    "            value = value{1};\n",
    "            if ~isnan(value)\n",
    "                if ischar(value)\n",
    "                    constraints_value_reaction{end+1} = value;\n",
    "                else\n",
    "                    error('ERROR - Constraints - Value - Reaction formula must be valid string')\n",
    "                end\n",
    "            else\n",
    "                error('ERROR - Constraints - Value - Must give reaction formula');\n",
    "            end\n",
    "        end\n",
    "        \n",
    "        % constraints counter\n",
    "        if added\n",
    "            constraints_value_count = constraints_value_count + 1;\n",
    "        end\n",
    "    end\n",
    "end"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Constraints - Fraction"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%file parse_constraints_fraction.m\n",
    "\n",
    "function [constraints_fraction_type,constraints_fraction_value,constraints_fraction_reaction,constraints_fraction_count] = parse_constraints_fraction(input_file)\n",
    "  \n",
    "    % load data\n",
    "    [~,~,raw] = xlsread(input_file,'Constraints - Fraction');\n",
    "\n",
    "    % fraction of maximum constraints\n",
    "    constraints_fraction_type = {};\n",
    "    constraints_fraction_value = {};\n",
    "    constraints_fraction_reaction = {};\n",
    "    constraints_fraction_count = 0;\n",
    "\n",
    "    for row = 3:size(raw,1)\n",
    "        added = false;\n",
    "\n",
    "        % constraint type - equal\n",
    "        value = raw(row,1);\n",
    "        value = value{1};\n",
    "        if ~isnan(value)\n",
    "            if isfloat(value)           \n",
    "                if (value >= 0) && (value <= 1)\n",
    "                    added = true;\n",
    "                    constraints_fraction_type{end+1} = 'equal';\n",
    "                    constraints_fraction_value{end+1} = value;\n",
    "                else\n",
    "                    error('ERROR - Constraints - Fraction - Constraint fraction must be >= 0 and <= 1')\n",
    "                end\n",
    "            else\n",
    "                error('ERROR - Constraints - Fraction - Constraint fraction must be a number')\n",
    "            end\n",
    "        end\n",
    "\n",
    "        % constraint type - less than\n",
    "        value = raw(row,2);\n",
    "        value = value{1};\n",
    "        if ~isnan(value)\n",
    "            if ~added\n",
    "                if isfloat(value)           \n",
    "                    if (value >= 0) && (value <= 1)\n",
    "                        added = true;\n",
    "                        constraints_fraction_type{end+1} = 'less';\n",
    "                        constraints_fraction_value{end+1} = value;\n",
    "                    else\n",
    "                        error('ERROR - Constraints - Fraction - Constraint fraction must be >= 0 and <= 1')\n",
    "                    end\n",
    "                else\n",
    "                    error('ERROR - Constraints - Fraction - Constraint fraction must be a number')\n",
    "                end\n",
    "            else\n",
    "                error('ERROR - Constraints - Fraction - Can only choose one constraint type at a time')\n",
    "            end\n",
    "        end\n",
    "\n",
    "        % constraint type - greater than\n",
    "        value = raw(row,3);\n",
    "        value = value{1};\n",
    "        if ~isnan(value)\n",
    "            if ~added\n",
    "                if isfloat(value)           \n",
    "                    if (value >= 0) && (value <= 1)\n",
    "                        added = true;\n",
    "                        constraints_fraction_type{end+1} = 'greater';\n",
    "                        constraints_fraction_value{end+1} = value;\n",
    "                    else\n",
    "                        error('ERROR - Constraints - Fraction - Constraint fraction must be >= 0 and <= 1')\n",
    "                    end\n",
    "                else\n",
    "                    error('ERROR - Constraints - Fraction - Constraint fraction must be a number')\n",
    "                end\n",
    "            else\n",
    "                error('ERROR - Constraints - Fraction - Can only choose one constraint type at a time')\n",
    "            end\n",
    "        end\n",
    "\n",
    "        % constraint type - range\n",
    "        value1 = raw(row,4);\n",
    "        value1 = value1{1};\n",
    "        value2 = raw(row,5);\n",
    "        value2 = value2{1};\n",
    "        if ~isnan(value1) && ~isnan(value2)\n",
    "            if ~added\n",
    "                if isfloat(value1) && isfloat(value2)       \n",
    "                    if (value1 >= 0) && (value1 <= 1) && (value2 >= 0) && (value2 <= 1)\n",
    "                        if value1 <= value2\n",
    "                            added = true;\n",
    "                            constraints_fraction_type{end+1} = 'range';\n",
    "                            constraints_fraction_value{end+1} = [value1,value2];\n",
    "                        else\n",
    "                            error('ERROR - Constraints - Fraction - Minimum fraction must be <= maximum fraction')\n",
    "                        end\n",
    "                    else\n",
    "                        error('ERROR - Constraints - Fraction - Constraint fractions must be >= 0 and <= 1')\n",
    "                    end\n",
    "                else\n",
    "                    error('ERROR - Constraints - Fraction - Constraint values must be a number')\n",
    "                end\n",
    "            else\n",
    "                error('ERROR - Constraints - Fraction - Can only choose one constraint type at a time')\n",
    "            end\n",
    "        elseif ~isnan(value1) || ~isnan(value2)\n",
    "            error('ERROR - Constraints - Fraction - If choosing \"Range\" constraint, must enter both minimum and maximum fractions');\n",
    "        end\n",
    "\n",
    "        % reaction formula\n",
    "        if added\n",
    "            value = raw(row,6);\n",
    "            value = value{1};\n",
    "            if ~isnan(value)\n",
    "                if ischar(value)\n",
    "                    constraints_fraction_reaction{end+1} = value;\n",
    "                else\n",
    "                    error('ERROR - Constraints - Fraction - Reaction formula must be valid string')\n",
    "                end\n",
    "            else\n",
    "                error('ERROR - Constraints - Fraction - Must give reaction formula');\n",
    "            end\n",
    "        end\n",
    "        \n",
    "        % constraints counter\n",
    "        if added\n",
    "            constraints_fraction_count = constraints_fraction_count + 1;\n",
    "        end\n",
    "    end\n",
    "end"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Constraints - Max Multi"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%file parse_constraints_maxmulti.m\n",
    "\n",
    "function [constraints_maxmulti_weight,constraints_maxmulti_reaction,constraints_maxmulti_fraction,constraints_maxmulti_count] = parse_constraints_maxmulti(input_file)\n",
    "  \n",
    "    % load data\n",
    "    [~,~,raw] = xlsread(input_file,'Constraints - Max Multi');\n",
    "\n",
    "    % initialize max multi constraints\n",
    "    constraints_maxmulti_weight = {};\n",
    "    constraints_maxmulti_reaction = {};\n",
    "    constraints_maxmulti_count = 0;\n",
    "    constraints_maxmulti_fraction = NaN;\n",
    "\n",
    "    for row = 2:size(raw,1)\n",
    "\n",
    "        % weight\n",
    "        value = raw(row,1);\n",
    "        value = value{1};\n",
    "        if ~isnan(value)\n",
    "            if isfloat(value)\n",
    "                if value>0\n",
    "\n",
    "                    % set value\n",
    "                    constraints_maxmulti_weight{end+1} = value;\n",
    "\n",
    "                    % reaction\n",
    "                    value = raw(row,2);\n",
    "                    value = value{1};\n",
    "                    if ~isnan(value)\n",
    "                        if ischar(value)\n",
    "                            constraints_maxmulti_reaction{end+1} = value;\n",
    "                            constraints_maxmulti_count = constraints_maxmulti_count + 1;\n",
    "                        else\n",
    "                            error('ERROR - Constraints - Max Multi - Reaction formula must be valid string')\n",
    "                        end\n",
    "                    else\n",
    "                        error('ERROR - Constraints - Max Multi - Must give reaction formula');\n",
    "                    end\n",
    "\n",
    "                    % fraction of maximum\n",
    "                    value = raw(1,4);\n",
    "                    value = value{1};\n",
    "                    if ~isnan(value)\n",
    "                        if isfloat(value)\n",
    "                            if (value >= 0) && (value <= 1)\n",
    "                                constraints_maxmulti_fraction = value;\n",
    "                            else\n",
    "                                error('ERROR - Constraints - Max Multi - Fraction of maximum objective value must either be >= 0 and <= 1')\n",
    "                            end\n",
    "                        else\n",
    "                            error('ERROR - Constraints - Max Multi - Fraction of maximum objective value must either be >= 0 and <= 1')\n",
    "                        end\n",
    "                    else\n",
    "                        error('ERROR - Constraints - Max Multi - Must enter fraction of maximum objective value')\n",
    "                    end\n",
    "\n",
    "                else\n",
    "                    error('ERROR - Constraints - Max Multi - Weight must be > 0')\n",
    "                end\n",
    "            elseif isfloat(str2num(value))\n",
    "                if str2num(value)>0\n",
    "\n",
    "                    % set value\n",
    "                    constraints_maxmulti_weight{end+1} = str2num(value);\n",
    "\n",
    "                    % reaction\n",
    "                    value = raw(row,2);\n",
    "                    value = value{1};\n",
    "                    if ~isnan(value)\n",
    "                        if ischar(value)\n",
    "                            constraints_maxmulti_reaction{end+1} = value;\n",
    "                            constraints_maxmulti_count = constraints_maxmulti_count + 1;\n",
    "                        else\n",
    "                            error('ERROR - Constraints - Max Multi - Reaction formula must be valid string')\n",
    "                        end\n",
    "                    else\n",
    "                        error('ERROR - Constraints - Max Multi - Must give reaction formula');\n",
    "                    end\n",
    "\n",
    "                    % fraction of maximum\n",
    "                    value = raw(1,4);\n",
    "                    value = value{1};\n",
    "                    if ~isnan(value)\n",
    "                        if isfloat(value)\n",
    "                            if (value >= 0) && (value <= 1)\n",
    "                                constraints_maxmulti_fraction = value;\n",
    "                            else\n",
    "                                error('ERROR - Constraints - Max Multi - Fraction of maximum objective value must either be >= 0 and <= 1')\n",
    "                            end\n",
    "                        else\n",
    "                            error('ERROR - Constraints - Max Multi - Fraction of maximum objective value must either be >= 0 and <= 1')\n",
    "                        end\n",
    "                    else\n",
    "                        error('ERROR - Constraints - Max Multi - Must enter fraction of maximum objective value')\n",
    "                    end\n",
    "\n",
    "                else\n",
    "                    error('ERROR - Constraints - Max Multi - Weight must be > 0')\n",
    "                end\n",
    "            else\n",
    "                error('ERROR - Constraints - Max Multi - Weight must be a number')\n",
    "            end\n",
    "        end\n",
    "    end\n",
    "end"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Objective Function"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%file parse_objective.m\n",
    "\n",
    "function [objective_function_direction,objective_function_weight,objective_function_reaction] = parse_objective(input_file)\n",
    "  \n",
    "% load data\n",
    "[~,~,raw] = xlsread(input_file,'Objective Function');\n",
    "\n",
    "% initialize objective functions\n",
    "objective_function_direction = {};\n",
    "objective_function_weight = {};\n",
    "objective_function_reaction = {};\n",
    "\n",
    "for row = 2:size(raw,1)\n",
    "\n",
    "    % objective function\n",
    "    value = raw(row,1);\n",
    "    value = value{1};\n",
    "    if ~isnan(value)\n",
    "        if ischar(value)\n",
    "            if strcmpi(value,'X')\n",
    "\n",
    "                % max or min\n",
    "                value = raw(row,2);\n",
    "                value = value{1};\n",
    "                if ~isnan(value)\n",
    "                    if ischar(value)\n",
    "                        if any(strcmp({'MAX','MIN'},value))\n",
    "                            objective_function_direction{end+1} = value;\n",
    "                        else\n",
    "                            error('ERROR - Objective Function - Must state MAX or MIN')\n",
    "                        end\n",
    "                    else\n",
    "                        error('ERROR - Objective Function - Must state MAX or MIN')\n",
    "                    end\n",
    "                else\n",
    "                    error('ERROR - Objective Function - Must state MAX or MIN');\n",
    "                end\n",
    "\n",
    "                % weight\n",
    "                value = raw(row,3);\n",
    "                value = value{1};\n",
    "                if ~isnan(value)\n",
    "                    if isfloat(value)\n",
    "                        if value>0\n",
    "                            objective_function_weight{end+1} = value;\n",
    "                        else\n",
    "                            error('ERROR - Objective Function - Weight must be > 0')\n",
    "                        end\n",
    "                    elseif isfloat(str2num(value))\n",
    "                        if str2num(value)>0\n",
    "                            objective_function_weight{end+1} = str2num(value);\n",
    "                        else\n",
    "                            error('ERROR - Objective Function - Weight must be > 0')\n",
    "                        end\n",
    "                    else\n",
    "                        error('ERROR - Objective Function - Weight must be a number')\n",
    "                    end\n",
    "                else\n",
    "                    error('ERROR - Objective Function - Weight must be a number');\n",
    "                end\n",
    "\n",
    "                % reaction formula\n",
    "                value = raw(row,4);\n",
    "                value = value{1};\n",
    "                if ~isnan(value)\n",
    "                    if ischar(value)\n",
    "                        objective_function_reaction{end+1} = value;\n",
    "                    else\n",
    "                        error('ERROR - Objective Function - Reaction formula must be valid string')\n",
    "                    end\n",
    "                else\n",
    "                    error('ERROR - Objective Function - Must give reaction formula');\n",
    "                end\n",
    "\n",
    "           else\n",
    "                error('ERROR - Objective Function - Entry in SELECT column must either be X or blank')\n",
    "            end\n",
    "        else\n",
    "            error('ERROR - Objective Function - Entry in SELECT column must either be X or blank')\n",
    "        end\n",
    "    end\n",
    "end"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Drugs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%file parse_drugs.m\n",
    "\n",
    "function [required_drugs] = parse_drugs(input_file)\n",
    "    \n",
    "    % load data\n",
    "    [~,~,raw] = xlsread(input_file,'Drugs');\n",
    "\n",
    "    % samples\n",
    "    required_drugs = {};\n",
    "    for row = 3:390\n",
    "        value = raw(row,1);\n",
    "        value = value{1};\n",
    "\n",
    "        if ~isnan(value)\n",
    "            if ischar(value)\n",
    "                if strcmpi(value,'X')\n",
    "                    switch row\n",
    "                        case 3, required_drugs{end+1} = '681640';\n",
    "                        case 4, required_drugs{end+1} = '(5Z)-7-Oxozeaenol';\n",
    "                        case 5, required_drugs{end+1} = '81C6';\n",
    "                        case 6, required_drugs{end+1} = 'A-443654';\n",
    "                        case 7, required_drugs{end+1} = 'A-770041';\n",
    "                        case 8, required_drugs{end+1} = 'Afatinib';\n",
    "                        case 9, required_drugs{end+1} = 'AICA Ribonucleotide';\n",
    "                        case 10, required_drugs{end+1} = 'AKT inhibitor VIII';\n",
    "                        case 11, required_drugs{end+1} = 'Aldesleukin';\n",
    "                        case 12, required_drugs{end+1} = 'Alectinib';\n",
    "                        case 13, required_drugs{end+1} = 'Alisertib';\n",
    "                        case 14, required_drugs{end+1} = 'Alvesin';\n",
    "                        case 15, required_drugs{end+1} = 'Amino Acids';\n",
    "                        case 16, required_drugs{end+1} = 'Amuvatinib';\n",
    "                        case 17, required_drugs{end+1} = 'Anastrozole';\n",
    "                        case 18, required_drugs{end+1} = 'Angiocal';\n",
    "                        case 19, required_drugs{end+1} = 'AR-42';\n",
    "                        case 20, required_drugs{end+1} = 'AS601245';\n",
    "                        case 21, required_drugs{end+1} = 'AS605240';\n",
    "                        case 22, required_drugs{end+1} = 'AT-7519';\n",
    "                        case 23, required_drugs{end+1} = 'ATCV';\n",
    "                        case 24, required_drugs{end+1} = 'Avagacestat';\n",
    "                        case 25, required_drugs{end+1} = 'Axitinib';\n",
    "                        case 26, required_drugs{end+1} = 'AZ628';\n",
    "                        case 27, required_drugs{end+1} = 'AZD6482';\n",
    "                        case 28, required_drugs{end+1} = 'AZD7762';\n",
    "                        case 29, required_drugs{end+1} = 'AZD8055';\n",
    "                        case 30, required_drugs{end+1} = 'BAY-61-3606';\n",
    "                        case 31, required_drugs{end+1} = 'BCG';\n",
    "                        case 32, required_drugs{end+1} = 'Belinostat';\n",
    "                        case 33, required_drugs{end+1} = 'Bevacizumab';\n",
    "                        case 34, required_drugs{end+1} = 'Bexarotene';\n",
    "                        case 35, required_drugs{end+1} = 'BI-2536';\n",
    "                        case 36, required_drugs{end+1} = 'Bicalutamide';\n",
    "                        case 37, required_drugs{end+1} = 'BIX02189';\n",
    "                        case 38, required_drugs{end+1} = 'Bleomycin';\n",
    "                        case 39, required_drugs{end+1} = 'BMS-345541';\n",
    "                        case 40, required_drugs{end+1} = 'BMS-509744';\n",
    "                        case 41, required_drugs{end+1} = 'BMS-536924';\n",
    "                        case 42, required_drugs{end+1} = 'BMS-754807';\n",
    "                        case 43, required_drugs{end+1} = 'Bortezomib';\n",
    "                        case 44, required_drugs{end+1} = 'Bosutinib';\n",
    "                        case 45, required_drugs{end+1} = 'Bryostatin 1';\n",
    "                        case 46, required_drugs{end+1} = 'BX796';\n",
    "                        case 47, required_drugs{end+1} = 'BX-912';\n",
    "                        case 48, required_drugs{end+1} = 'Cabazitaxel';\n",
    "                        case 49, required_drugs{end+1} = 'Cabozantinib';\n",
    "                        case 50, required_drugs{end+1} = 'Camptothecin';\n",
    "                        case 51, required_drugs{end+1} = 'Capecitabine';\n",
    "                        case 52, required_drugs{end+1} = 'Capecitabine+Oxaliplatin';\n",
    "                        case 53, required_drugs{end+1} = 'Carboplatin';\n",
    "                        case 54, required_drugs{end+1} = 'Carboplatin+Paclitaxel';\n",
    "                        case 55, required_drugs{end+1} = 'Carmustine';\n",
    "                        case 56, required_drugs{end+1} = 'CAY10603';\n",
    "                        case 57, required_drugs{end+1} = 'CCT007093';\n",
    "                        case 58, required_drugs{end+1} = 'CCT-018159';\n",
    "                        case 59, required_drugs{end+1} = 'Cediranib';\n",
    "                        case 60, required_drugs{end+1} = 'Celecoxib';\n",
    "                        case 61, required_drugs{end+1} = 'Cetuximab';\n",
    "                        case 62, required_drugs{end+1} = 'CGP-082996';\n",
    "                        case 63, required_drugs{end+1} = 'CGP-60474';\n",
    "                        case 64, required_drugs{end+1} = 'CHIR-99021';\n",
    "                        case 65, required_drugs{end+1} = 'CI-1040';\n",
    "                        case 66, required_drugs{end+1} = 'Cilengitide';\n",
    "                        case 67, required_drugs{end+1} = 'Cisplatin';\n",
    "                        case 68, required_drugs{end+1} = 'Cisplatin+Gemcitabine';\n",
    "                        case 69, required_drugs{end+1} = 'CMK';\n",
    "                        case 70, required_drugs{end+1} = 'CP466722';\n",
    "                        case 71, required_drugs{end+1} = 'CP724714';\n",
    "                        case 72, required_drugs{end+1} = 'Crizotinib';\n",
    "                        case 73, required_drugs{end+1} = 'CUDC-101';\n",
    "                        case 74, required_drugs{end+1} = 'CX-5461';\n",
    "                        case 75, required_drugs{end+1} = 'Cyanocobalamin';\n",
    "                        case 76, required_drugs{end+1} = 'Cyclopamine';\n",
    "                        case 77, required_drugs{end+1} = 'Cyclophosphamide';\n",
    "                        case 78, required_drugs{end+1} = 'Cyclophosphamide+Polysaccharide-K';\n",
    "                        case 79, required_drugs{end+1} = 'Cyclophosphamide+Vincristine+Dacarbazine';\n",
    "                        case 80, required_drugs{end+1} = 'Cyclosporine';\n",
    "                        case 81, required_drugs{end+1} = 'Cytarabine';\n",
    "                        case 82, required_drugs{end+1} = 'Dabrafenib';\n",
    "                        case 83, required_drugs{end+1} = 'Dacarbazine';\n",
    "                        case 84, required_drugs{end+1} = 'Dacinostat';\n",
    "                        case 85, required_drugs{end+1} = 'Dactinomycin';\n",
    "                        case 86, required_drugs{end+1} = 'Dactolisib';\n",
    "                        case 87, required_drugs{end+1} = 'Daporinad';\n",
    "                        case 88, required_drugs{end+1} = 'Dasatinib';\n",
    "                        case 89, required_drugs{end+1} = 'Degarelix';\n",
    "                        case 90, required_drugs{end+1} = 'Denosumab';\n",
    "                        case 91, required_drugs{end+1} = 'Dexamethasone';\n",
    "                        case 92, required_drugs{end+1} = 'Didox';\n",
    "                        case 93, required_drugs{end+1} = 'DMOG';\n",
    "                        case 94, required_drugs{end+1} = 'Docetaxel';\n",
    "                        case 95, required_drugs{end+1} = 'Docetaxel+Vandetanib';\n",
    "                        case 96, required_drugs{end+1} = 'Doramapimod';\n",
    "                        case 97, required_drugs{end+1} = 'Doxifluridine';\n",
    "                        case 98, required_drugs{end+1} = 'Doxorubicin';\n",
    "                        case 99, required_drugs{end+1} = 'Doxorubicin+Cyclophosphamide';\n",
    "                        case 100, required_drugs{end+1} = 'DS-8273a';\n",
    "                        case 101, required_drugs{end+1} = 'EHT-1864';\n",
    "                        case 102, required_drugs{end+1} = 'Elesclomol';\n",
    "                        case 103, required_drugs{end+1} = 'Embelin';\n",
    "                        case 104, required_drugs{end+1} = 'Entinostat';\n",
    "                        case 105, required_drugs{end+1} = 'Enzastaurin';\n",
    "                        case 106, required_drugs{end+1} = 'Epirubicin';\n",
    "                        case 107, required_drugs{end+1} = 'Epoetin';\n",
    "                        case 108, required_drugs{end+1} = 'Eribulin';\n",
    "                        case 109, required_drugs{end+1} = 'Erlotinib';\n",
    "                        case 110, required_drugs{end+1} = 'Etoposide';\n",
    "                        case 111, required_drugs{end+1} = 'Everolimus';\n",
    "                        case 112, required_drugs{end+1} = 'Everolimus+Gemcitabine+Cisplatin';\n",
    "                        case 113, required_drugs{end+1} = 'Evofosfamide';\n",
    "                        case 114, required_drugs{end+1} = 'Exemestane';\n",
    "                        case 115, required_drugs{end+1} = 'EZN-2968';\n",
    "                        case 116, required_drugs{end+1} = 'Fedratinib';\n",
    "                        case 117, required_drugs{end+1} = 'FH535';\n",
    "                        case 118, required_drugs{end+1} = 'Fluorouracil';\n",
    "                        case 119, required_drugs{end+1} = 'FMK';\n",
    "                        case 120, required_drugs{end+1} = 'Fotemustine';\n",
    "                        case 121, required_drugs{end+1} = 'FR-180204';\n",
    "                        case 122, required_drugs{end+1} = 'FTI-277';\n",
    "                        case 123, required_drugs{end+1} = 'Fulvestrant';\n",
    "                        case 124, required_drugs{end+1} = 'GDC-0980';\n",
    "                        case 125, required_drugs{end+1} = 'Gefitinib';\n",
    "                        case 126, required_drugs{end+1} = 'Gemcitabine';\n",
    "                        case 127, required_drugs{end+1} = 'Gemcitabine+Oxaliplatin';\n",
    "                        case 128, required_drugs{end+1} = 'Genentech Cpd 10';\n",
    "                        case 129, required_drugs{end+1} = 'GL-ONC1';\n",
    "                        case 130, required_drugs{end+1} = 'GNF-2';\n",
    "                        case 131, required_drugs{end+1} = 'Goserelin';\n",
    "                        case 132, required_drugs{end+1} = 'GP100';\n",
    "                        case 133, required_drugs{end+1} = 'GSK1070916';\n",
    "                        case 134, required_drugs{end+1} = 'GSK1904529A';\n",
    "                        case 135, required_drugs{end+1} = 'GSK2302025A';\n",
    "                        case 136, required_drugs{end+1} = 'GSK269962A';\n",
    "                        case 137, required_drugs{end+1} = 'GSK319347A';\n",
    "                        case 138, required_drugs{end+1} = 'GSK429286A';\n",
    "                        case 139, required_drugs{end+1} = 'GSK650394';\n",
    "                        case 140, required_drugs{end+1} = 'GSK690693';\n",
    "                        case 141, required_drugs{end+1} = 'GW-2580';\n",
    "                        case 142, required_drugs{end+1} = 'GW441756';\n",
    "                        case 143, required_drugs{end+1} = 'GW843682X';\n",
    "                        case 144, required_drugs{end+1} = 'Halichondrin B';\n",
    "                        case 145, required_drugs{end+1} = 'HG-5-113-01';\n",
    "                        case 146, required_drugs{end+1} = 'HG-5-88-01';\n",
    "                        case 147, required_drugs{end+1} = 'HG6-64-1';\n",
    "                        case 148, required_drugs{end+1} = 'Hydrocortisone';\n",
    "                        case 149, required_drugs{end+1} = 'Hydroxyurea';\n",
    "                        case 150, required_drugs{end+1} = 'I-BET-762';\n",
    "                        case 151, required_drugs{end+1} = 'Idelalisib';\n",
    "                        case 152, required_drugs{end+1} = 'Ifosfamide';\n",
    "                        case 153, required_drugs{end+1} = 'Imatinib';\n",
    "                        case 154, required_drugs{end+1} = 'Infliximab';\n",
    "                        case 155, required_drugs{end+1} = 'Interferon A';\n",
    "                        case 156, required_drugs{end+1} = 'IOX2';\n",
    "                        case 157, required_drugs{end+1} = 'IPA-3';\n",
    "                        case 158, required_drugs{end+1} = 'Ipilimumab';\n",
    "                        case 159, required_drugs{end+1} = 'Irinotecan';\n",
    "                        case 160, required_drugs{end+1} = 'Irinotecan+Cetuximab';\n",
    "                        case 161, required_drugs{end+1} = 'Isotretinoin';\n",
    "                        case 162, required_drugs{end+1} = 'Ispinesib';\n",
    "                        case 163, required_drugs{end+1} = 'Ixabepilone';\n",
    "                        case 164, required_drugs{end+1} = 'JNK Inhibitor VIII';\n",
    "                        case 165, required_drugs{end+1} = 'JNK-9L';\n",
    "                        case 166, required_drugs{end+1} = 'JQ1';\n",
    "                        case 167, required_drugs{end+1} = 'JQ12';\n",
    "                        case 168, required_drugs{end+1} = 'JW-7-24-1';\n",
    "                        case 169, required_drugs{end+1} = 'JW-7-52-1';\n",
    "                        case 170, required_drugs{end+1} = 'Karenitecin';\n",
    "                        case 171, required_drugs{end+1} = 'KIN001-236';\n",
    "                        case 172, required_drugs{end+1} = 'KIN001-244';\n",
    "                        case 173, required_drugs{end+1} = 'KIN001-260';\n",
    "                        case 174, required_drugs{end+1} = 'KIN001-266';\n",
    "                        case 175, required_drugs{end+1} = 'KIN001-270';\n",
    "                        case 176, required_drugs{end+1} = 'KU-55933';\n",
    "                        case 177, required_drugs{end+1} = 'Lapatinib';\n",
    "                        case 178, required_drugs{end+1} = 'Leflunomide';\n",
    "                        case 179, required_drugs{end+1} = 'Lenalidomide';\n",
    "                        case 180, required_drugs{end+1} = 'Lestaurtinib';\n",
    "                        case 181, required_drugs{end+1} = 'Letrozole';\n",
    "                        case 182, required_drugs{end+1} = 'Leucovorin';\n",
    "                        case 183, required_drugs{end+1} = 'Leucovorin+Fluorouracil+Irinotecan';\n",
    "                        case 184, required_drugs{end+1} = 'Leucovorin+Fluorouracil+Irinotecan+Aflibercept';\n",
    "                        case 185, required_drugs{end+1} = 'Leucovorin+Fluorouracil+Irinotecan+Bevacizumab';\n",
    "                        case 186, required_drugs{end+1} = 'Leucovorin+Fluorouracil+Oxaliplatin';\n",
    "                        case 187, required_drugs{end+1} = 'Leuprolide';\n",
    "                        case 188, required_drugs{end+1} = 'Levothyroxine';\n",
    "                        case 189, required_drugs{end+1} = 'LFM-A13';\n",
    "                        case 190, required_drugs{end+1} = 'LHRH';\n",
    "                        case 191, required_drugs{end+1} = 'Linifanib';\n",
    "                        case 192, required_drugs{end+1} = 'Linsitinib';\n",
    "                        case 193, required_drugs{end+1} = 'Liothyronine';\n",
    "                        case 194, required_drugs{end+1} = 'Lomustine';\n",
    "                        case 195, required_drugs{end+1} = 'Luminespib';\n",
    "                        case 196, required_drugs{end+1} = 'LY228820';\n",
    "                        case 197, required_drugs{end+1} = 'MAGE-3-AS-AS15-ASCI';\n",
    "                        case 198, required_drugs{end+1} = 'MAGE-A3';\n",
    "                        case 199, required_drugs{end+1} = 'Masitinib';\n",
    "                        case 200, required_drugs{end+1} = 'Megestrol';\n",
    "                        case 201, required_drugs{end+1} = 'Melphalan';\n",
    "                        case 202, required_drugs{end+1} = 'Mesna';\n",
    "                        case 203, required_drugs{end+1} = 'Methotrexate';\n",
    "                        case 204, required_drugs{end+1} = 'Methylprednisolone';\n",
    "                        case 205, required_drugs{end+1} = 'MG-132';\n",
    "                        case 206, required_drugs{end+1} = 'Mibefradil';\n",
    "                        case 207, required_drugs{end+1} = 'Midostaurin';\n",
    "                        case 208, required_drugs{end+1} = 'Mitomycin C';\n",
    "                        case 209, required_drugs{end+1} = 'Mitoxantrone';\n",
    "                        case 210, required_drugs{end+1} = 'MK2206';\n",
    "                        case 211, required_drugs{end+1} = 'MK-2206';\n",
    "                        case 212, required_drugs{end+1} = 'MORAb-004';\n",
    "                        case 213, required_drugs{end+1} = 'Motesanib';\n",
    "                        case 214, required_drugs{end+1} = 'MPS-1-IN-1';\n",
    "                        case 215, required_drugs{end+1} = 'MR1-1';\n",
    "                        case 216, required_drugs{end+1} = 'Navitoclax';\n",
    "                        case 217, required_drugs{end+1} = 'NG-25';\n",
    "                        case 218, required_drugs{end+1} = 'Nilotinib';\n",
    "                        case 219, required_drugs{end+1} = 'Nilutamide';\n",
    "                        case 220, required_drugs{end+1} = 'Nimustine';\n",
    "                        case 221, required_drugs{end+1} = 'Nivolumab';\n",
    "                        case 222, required_drugs{end+1} = 'NPK76-II-72-1';\n",
    "                        case 223, required_drugs{end+1} = 'NSC-207895';\n",
    "                        case 224, required_drugs{end+1} = 'NSC-87877';\n",
    "                        case 225, required_drugs{end+1} = 'NU7441';\n",
    "                        case 226, required_drugs{end+1} = 'Nutlin-3a';\n",
    "                        case 227, required_drugs{end+1} = 'NVP-BHG712';\n",
    "                        case 228, required_drugs{end+1} = 'NVP-TAE684';\n",
    "                        case 229, required_drugs{end+1} = 'O6-Benzylguanine';\n",
    "                        case 230, required_drugs{end+1} = 'Obatoclax';\n",
    "                        case 231, required_drugs{end+1} = 'Olaparib';\n",
    "                        case 232, required_drugs{end+1} = 'Omipalisib';\n",
    "                        case 233, required_drugs{end+1} = 'Ondansetron';\n",
    "                        case 234, required_drugs{end+1} = 'ONYX-015';\n",
    "                        case 235, required_drugs{end+1} = 'OSI-027';\n",
    "                        case 236, required_drugs{end+1} = 'OSI-930';\n",
    "                        case 237, required_drugs{end+1} = 'OSU-03012';\n",
    "                        case 238, required_drugs{end+1} = 'Oxaliplatin';\n",
    "                        case 239, required_drugs{end+1} = 'PAC-1';\n",
    "                        case 240, required_drugs{end+1} = 'Paclitaxel';\n",
    "                        case 241, required_drugs{end+1} = 'Palbociclib';\n",
    "                        case 242, required_drugs{end+1} = 'Palonosetron';\n",
    "                        case 243, required_drugs{end+1} = 'Pamidronate';\n",
    "                        case 244, required_drugs{end+1} = 'Panitumumab';\n",
    "                        case 245, required_drugs{end+1} = 'Parthenolide';\n",
    "                        case 246, required_drugs{end+1} = 'Pasireotide';\n",
    "                        case 247, required_drugs{end+1} = 'Patupilone';\n",
    "                        case 248, required_drugs{end+1} = 'Pazopanib';\n",
    "                        case 249, required_drugs{end+1} = 'PD0325901';\n",
    "                        case 250, required_drugs{end+1} = 'PD173074';\n",
    "                        case 251, required_drugs{end+1} = 'Pegfilgrastim';\n",
    "                        case 252, required_drugs{end+1} = 'Pelitinib';\n",
    "                        case 253, required_drugs{end+1} = 'Pembrolizumab';\n",
    "                        case 254, required_drugs{end+1} = 'Pemetrexed';\n",
    "                        case 255, required_drugs{end+1} = 'Pevonedistat';\n",
    "                        case 256, required_drugs{end+1} = 'PF-4708671';\n",
    "                        case 257, required_drugs{end+1} = 'PF-562271';\n",
    "                        case 258, required_drugs{end+1} = 'PFI-1';\n",
    "                        case 259, required_drugs{end+1} = 'PHA-665752';\n",
    "                        case 260, required_drugs{end+1} = 'PHA-793887';\n",
    "                        case 261, required_drugs{end+1} = 'Phenformin';\n",
    "                        case 262, required_drugs{end+1} = 'PI-103';\n",
    "                        case 263, required_drugs{end+1} = 'PI-88';\n",
    "                        case 264, required_drugs{end+1} = 'Pictilisib';\n",
    "                        case 265, required_drugs{end+1} = 'PIK-93';\n",
    "                        case 266, required_drugs{end+1} = 'Piperlongumine';\n",
    "                        case 267, required_drugs{end+1} = 'Platinum';\n",
    "                        case 268, required_drugs{end+1} = 'PLFE';\n",
    "                        case 269, required_drugs{end+1} = 'PLX3397';\n",
    "                        case 270, required_drugs{end+1} = 'PLX-4720';\n",
    "                        case 271, required_drugs{end+1} = 'PNU-159548';\n",
    "                        case 272, required_drugs{end+1} = 'Poly E';\n",
    "                        case 273, required_drugs{end+1} = 'Polyvalent Melanoma Vaccine';\n",
    "                        case 274, required_drugs{end+1} = 'Ponatinib';\n",
    "                        case 275, required_drugs{end+1} = 'Prednisone';\n",
    "                        case 276, required_drugs{end+1} = 'Procarbazine';\n",
    "                        case 277, required_drugs{end+1} = 'PV-10';\n",
    "                        case 278, required_drugs{end+1} = 'px-866';\n",
    "                        case 279, required_drugs{end+1} = 'Pyrimethamine';\n",
    "                        case 280, required_drugs{end+1} = 'QL-VIII-58';\n",
    "                        case 281, required_drugs{end+1} = 'QL-X-138';\n",
    "                        case 282, required_drugs{end+1} = 'QL-XI-92';\n",
    "                        case 283, required_drugs{end+1} = 'QL-XII-47';\n",
    "                        case 284, required_drugs{end+1} = 'QL-XII-61';\n",
    "                        case 285, required_drugs{end+1} = 'QS11';\n",
    "                        case 286, required_drugs{end+1} = 'Quizartinib';\n",
    "                        case 287, required_drugs{end+1} = 'Raloxifene';\n",
    "                        case 288, required_drugs{end+1} = 'Rapamycin';\n",
    "                        case 289, required_drugs{end+1} = 'Refametinib';\n",
    "                        case 290, required_drugs{end+1} = 'REGN421';\n",
    "                        case 291, required_drugs{end+1} = 'Regorafenib';\n",
    "                        case 292, required_drugs{end+1} = 'Reolysin';\n",
    "                        case 293, required_drugs{end+1} = 'Resiquimod';\n",
    "                        case 294, required_drugs{end+1} = 'Ridaforolimus';\n",
    "                        case 295, required_drugs{end+1} = 'Rigosertib';\n",
    "                        case 296, required_drugs{end+1} = 'Rituximab';\n",
    "                        case 297, required_drugs{end+1} = 'RO-3306';\n",
    "                        case 298, required_drugs{end+1} = 'rTRAIL';\n",
    "                        case 299, required_drugs{end+1} = 'Rucaparib';\n",
    "                        case 300, required_drugs{end+1} = 'Ruxolitinib';\n",
    "                        case 301, required_drugs{end+1} = 'Salubrinal';\n",
    "                        case 302, required_drugs{end+1} = 'Saracatinib';\n",
    "                        case 303, required_drugs{end+1} = 'Sargramostim';\n",
    "                        case 304, required_drugs{end+1} = 'SB216763';\n",
    "                        case 305, required_drugs{end+1} = 'SB-505124';\n",
    "                        case 306, required_drugs{end+1} = 'SB52334';\n",
    "                        case 307, required_drugs{end+1} = 'SB590885';\n",
    "                        case 308, required_drugs{end+1} = 'Seliciclib';\n",
    "                        case 309, required_drugs{end+1} = 'Selisistat';\n",
    "                        case 310, required_drugs{end+1} = 'Selumetinib';\n",
    "                        case 311, required_drugs{end+1} = 'Sepantronium bromide';\n",
    "                        case 312, required_drugs{end+1} = 'Serdemetan';\n",
    "                        case 313, required_drugs{end+1} = 'SGC0946';\n",
    "                        case 314, required_drugs{end+1} = 'Shikonin';\n",
    "                        case 315, required_drugs{end+1} = 'SL0101';\n",
    "                        case 316, required_drugs{end+1} = 'SNX-2112';\n",
    "                        case 317, required_drugs{end+1} = 'Sorafenib';\n",
    "                        case 318, required_drugs{end+1} = 'STF-62247';\n",
    "                        case 319, required_drugs{end+1} = 'S-Trityl-L-cysteine';\n",
    "                        case 320, required_drugs{end+1} = 'Sulindac';\n",
    "                        case 321, required_drugs{end+1} = 'Sunitinib';\n",
    "                        case 322, required_drugs{end+1} = 'T0901317';\n",
    "                        case 323, required_drugs{end+1} = 'TAK-715';\n",
    "                        case 324, required_drugs{end+1} = 'Talazoparib';\n",
    "                        case 325, required_drugs{end+1} = 'Talimogene';\n",
    "                        case 326, required_drugs{end+1} = 'Tamoxifen';\n",
    "                        case 327, required_drugs{end+1} = 'Tanespimycin';\n",
    "                        case 328, required_drugs{end+1} = 'Tegafur+Gimeracil+Oteracil';\n",
    "                        case 329, required_drugs{end+1} = 'Temozolomide';\n",
    "                        case 330, required_drugs{end+1} = 'Temsirolimus';\n",
    "                        case 331, required_drugs{end+1} = 'Teniposide';\n",
    "                        case 332, required_drugs{end+1} = 'Teprotumumab';\n",
    "                        case 333, required_drugs{end+1} = 'TGX221';\n",
    "                        case 334, required_drugs{end+1} = 'Thalidomide';\n",
    "                        case 335, required_drugs{end+1} = 'Thapsigargin';\n",
    "                        case 336, required_drugs{end+1} = 'THZ-2-102-1';\n",
    "                        case 337, required_drugs{end+1} = 'THZ-2-49';\n",
    "                        case 338, required_drugs{end+1} = 'Tipifarnib';\n",
    "                        case 339, required_drugs{end+1} = 'Tivantinib';\n",
    "                        case 340, required_drugs{end+1} = 'Tivozanib';\n",
    "                        case 341, required_drugs{end+1} = 'TL-1-85';\n",
    "                        case 342, required_drugs{end+1} = 'TL-2-105';\n",
    "                        case 343, required_drugs{end+1} = 'Topotecan';\n",
    "                        case 344, required_drugs{end+1} = 'Toremifene';\n",
    "                        case 345, required_drugs{end+1} = 'Tozasertib';\n",
    "                        case 346, required_drugs{end+1} = 'TPCA-1';\n",
    "                        case 347, required_drugs{end+1} = 'Trabectedin';\n",
    "                        case 348, required_drugs{end+1} = 'Trametinib';\n",
    "                        case 349, required_drugs{end+1} = 'Trastuzumab';\n",
    "                        case 350, required_drugs{end+1} = 'Treosulfan';\n",
    "                        case 351, required_drugs{end+1} = 'Tretinoin';\n",
    "                        case 352, required_drugs{end+1} = 'Triptorelin';\n",
    "                        case 353, required_drugs{end+1} = 'Tubastatin A';\n",
    "                        case 354, required_drugs{end+1} = 'TW 37';\n",
    "                        case 355, required_drugs{end+1} = 'UNC0638';\n",
    "                        case 356, required_drugs{end+1} = 'UNC1215';\n",
    "                        case 357, required_drugs{end+1} = 'Valproic Acid';\n",
    "                        case 358, required_drugs{end+1} = 'Veliparib';\n",
    "                        case 359, required_drugs{end+1} = 'Veliparib+Temozolomide';\n",
    "                        case 360, required_drugs{end+1} = 'Vemurafenib';\n",
    "                        case 361, required_drugs{end+1} = 'Vinblastine';\n",
    "                        case 362, required_drugs{end+1} = 'Vincristine';\n",
    "                        case 363, required_drugs{end+1} = 'Vinorelbine';\n",
    "                        case 364, required_drugs{end+1} = 'Vismodegib';\n",
    "                        case 365, required_drugs{end+1} = 'VMCL';\n",
    "                        case 366, required_drugs{end+1} = 'VNLG/124';\n",
    "                        case 367, required_drugs{end+1} = 'Vorinostat';\n",
    "                        case 368, required_drugs{end+1} = 'VX-11e';\n",
    "                        case 369, required_drugs{end+1} = 'VX-702';\n",
    "                        case 370, required_drugs{end+1} = 'WH-4-023';\n",
    "                        case 371, required_drugs{end+1} = 'WHI-P97';\n",
    "                        case 372, required_drugs{end+1} = 'WIKI4';\n",
    "                        case 373, required_drugs{end+1} = 'WZ-1-84';\n",
    "                        case 374, required_drugs{end+1} = 'WZ3105';\n",
    "                        case 375, required_drugs{end+1} = 'XAV939';\n",
    "                        case 376, required_drugs{end+1} = 'XMD11-85h';\n",
    "                        case 377, required_drugs{end+1} = 'XMD13-2';\n",
    "                        case 378, required_drugs{end+1} = 'XMD14-99';\n",
    "                        case 379, required_drugs{end+1} = 'XMD15-27';\n",
    "                        case 380, required_drugs{end+1} = 'XMD8-85';\n",
    "                        case 381, required_drugs{end+1} = 'XMD8-92';\n",
    "                        case 382, required_drugs{end+1} = 'Y-39983';\n",
    "                        case 383, required_drugs{end+1} = 'YK-4-279';\n",
    "                        case 384, required_drugs{end+1} = 'YM201636';\n",
    "                        case 385, required_drugs{end+1} = 'ZG-10';\n",
    "                        case 386, required_drugs{end+1} = 'Zibotentan';\n",
    "                        case 387, required_drugs{end+1} = 'Z-LLNle-CHO';\n",
    "                        case 388, required_drugs{end+1} = 'ZM447439';\n",
    "                        case 389, required_drugs{end+1} = 'Zoledronic acid';\n",
    "                        case 390, required_drugs{end+1} = 'ZSTK474';\n",
    "                    end\n",
    "                else\n",
    "                    error('ERROR - Drugs - Values must either be X or blank')\n",
    "                end\n",
    "            else\n",
    "                error('ERROR - Drugs - Values must either be X or blank')\n",
    "            end\n",
    "        end\n",
    "    end\n",
    "end"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Mutations"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%file parse_mutations.m\n",
    "\n",
    "function [use_mutation] = parse_mutations(input_file)\n",
    "\n",
    "    % load data\n",
    "    [~,~,raw] = xlsread(input_file,'Mutations');\n",
    "    \n",
    "    % use mutation data\n",
    "    row = 3;\n",
    "    value = raw(row,1);\n",
    "    value = value{1};\n",
    "    if ~isnan(value)\n",
    "        if ischar(value)\n",
    "            if strcmpi(value,'X')\n",
    "                use_mutation = 1;\n",
    "            else\n",
    "                error('ERROR - Mutations - Value must either be X or blank')\n",
    "            end\n",
    "        else\n",
    "            error('ERROR - Mutations - Value must either be X or blank')\n",
    "        end\n",
    "    end\n",
    "    \n",
    "    row = 4;\n",
    "    value = raw(row,1);\n",
    "    value = value{1};\n",
    "    if ~isnan(value)\n",
    "        if ischar(value)\n",
    "            if strcmpi(value,'X')\n",
    "                if ~exist('use_mutation','var')\n",
    "                    use_mutation = 2;\n",
    "                else\n",
    "                    error('ERROR - Mutations - Cannot select more than one option') \n",
    "                end\n",
    "            else\n",
    "                error('ERROR - Mutations - Value must either be X or blank')\n",
    "            end\n",
    "        else\n",
    "            error('ERROR - Mutations - Value must either be X or blank')\n",
    "        end\n",
    "    end\n",
    "    \n",
    "    row = 5;\n",
    "    value = raw(row,1);\n",
    "    value = value{1};\n",
    "    if ~isnan(value)\n",
    "        if ischar(value)\n",
    "            if strcmpi(value,'X')\n",
    "                if ~exist('use_mutation','var')\n",
    "                    use_mutation = 0;\n",
    "                else\n",
    "                    error('ERROR - Mutations - Cannot select more than one option') \n",
    "                end\n",
    "            else\n",
    "                error('ERROR - Mutations - Value must either be X or blank')\n",
    "            end\n",
    "        else\n",
    "            error('ERROR - Mutations - Value must either be X or blank')\n",
    "        end\n",
    "    end\n",
    "    \n",
    "    if ~exist('use_mutation','var')\n",
    "        error('ERROR - Mutations - Must select an option') \n",
    "    end\n",
    "end"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Samples - TCGA"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%file parse_samples_tcga.m\n",
    "\n",
    "function [samples_all,samples_all_source,samples_all_source_protein] = parse_samples_tcga(input_file,constraints_proliferation,required_drugs,use_mutation)\n",
    "    \n",
    "    % load data - TCGA\n",
    "    [~,~,raw] = xlsread(input_file,'Samples - TCGA');\n",
    "\n",
    "    % samples - TCGA\n",
    "    tcga_types = {};\n",
    "    tcga_needs_proliferation = 0;\n",
    "    tcga_needs_radiation = 0;\n",
    "    for row = 3:35\n",
    "        value = raw(row,1);\n",
    "        value = value{1};\n",
    "\n",
    "        if ~isnan(value)\n",
    "            if ischar(value)\n",
    "                if strcmpi(value,'X')\n",
    "                    switch row\n",
    "                        case 3, tcga_types{end+1} = 'ACC';\n",
    "                        case 4, tcga_types{end+1} = 'BLCA';\n",
    "                        case 5, tcga_types{end+1} = 'BRCA';\n",
    "                        case 6, tcga_types{end+1} = 'CESC';\n",
    "                        case 7, tcga_types{end+1} = 'CHOL';\n",
    "                        case 8, tcga_types{end+1} = 'COAD';\n",
    "                        case 9, tcga_types{end+1} = 'DLBC';\n",
    "                        case 10, tcga_types{end+1} = 'ESCA';\n",
    "                        case 11, tcga_types{end+1} = 'GBM';\n",
    "                        case 12, tcga_types{end+1} = 'HNSC';\n",
    "                        case 13, tcga_types{end+1} = 'KICH';\n",
    "                        case 14, tcga_types{end+1} = 'KIRC';\n",
    "                        case 15, tcga_types{end+1} = 'KIRP';\n",
    "                        case 16, tcga_types{end+1} = 'LAML';\n",
    "                        case 17, tcga_types{end+1} = 'LGG';\n",
    "                        case 18, tcga_types{end+1} = 'LIHC';\n",
    "                        case 19, tcga_types{end+1} = 'LUAD';\n",
    "                        case 20, tcga_types{end+1} = 'LUSC';\n",
    "                        case 21, tcga_types{end+1} = 'MESO';\n",
    "                        case 22, tcga_types{end+1} = 'OV';\n",
    "                        case 23, tcga_types{end+1} = 'PAAD';\n",
    "                        case 24, tcga_types{end+1} = 'PCPG';\n",
    "                        case 25, tcga_types{end+1} = 'PRAD';\n",
    "                        case 26, tcga_types{end+1} = 'READ';\n",
    "                        case 27, tcga_types{end+1} = 'SARC';\n",
    "                        case 28, tcga_types{end+1} = 'SKCM';\n",
    "                        case 29, tcga_types{end+1} = 'STAD';\n",
    "                        case 30, tcga_types{end+1} = 'TGCT';\n",
    "                        case 31, tcga_types{end+1} = 'THCA';\n",
    "                        case 32, tcga_types{end+1} = 'THYM';\n",
    "                        case 33, tcga_types{end+1} = 'UCEC';\n",
    "                        case 34, tcga_types{end+1} = 'UCS';\n",
    "                        case 35, tcga_types{end+1} = 'UVM';\n",
    "                    end\n",
    "                else\n",
    "                    error('ERROR - Samples - TCGA - Values must either be X or blank')\n",
    "                end\n",
    "            else\n",
    "                error('ERROR - Samples - TCGA - Values must either be X or blank')\n",
    "            end\n",
    "        end\n",
    "    end\n",
    "    \n",
    "    % old or new protein abundance\n",
    "    if length(tcga_types) > 0\n",
    "        for row = 3:4\n",
    "            value = raw(row,5);\n",
    "            value = value{1};\n",
    "            if ~isnan(value)\n",
    "                if ischar(value)\n",
    "                    if strcmpi(value,'X')\n",
    "                        if exist('tcga_protein','var')\n",
    "                            error('ERROR - Samples - TCGA - Old or new protein abundance - Can only choose one option')\n",
    "                        else\n",
    "                            if row == 3\n",
    "                                tcga_protein = 'TCGA_old';\n",
    "                            elseif row == 4\n",
    "                                tcga_protein = 'TCGA';\n",
    "                            end\n",
    "                        end\n",
    "                    else\n",
    "                        error('ERROR - Samples - TCGA - Old or new protein abundance - Values must either be X or blank')\n",
    "                    end\n",
    "                else\n",
    "                    error('ERROR - Samples - TCGA - Old or new protein abundance - Values must either be X or blank')\n",
    "                end\n",
    "            end\n",
    "        end\n",
    "        if ~exist('tcga_protein','var')\n",
    "            error('ERROR - Samples - TCGA - Old or new protein abundance - Must select one option')\n",
    "        end\n",
    "    else\n",
    "        tcga_protein = NaN;\n",
    "    end\n",
    "    \n",
    "    % include primary tumor samples\n",
    "    row = 8;\n",
    "    value = raw(row,5);\n",
    "    value = value{1};\n",
    "\n",
    "    if ~isnan(value)\n",
    "        if ischar(value)\n",
    "            if strcmpi(value,'X')\n",
    "                tcga_include_tumor = 1;\n",
    "            else\n",
    "                error('ERROR - Samples - TCGA - Include primary tumor samples - Value must either be X or blank')\n",
    "            end\n",
    "        else\n",
    "            error('ERROR - Samples - TCGA - Include primary tumor samples - Value must either be X or blank')\n",
    "        end\n",
    "    else\n",
    "        tcga_include_tumor = 0;\n",
    "    end\n",
    "\n",
    "    % include normal tissue samples\n",
    "    row = 9;\n",
    "    value = raw(row,5);\n",
    "    value = value{1};\n",
    "\n",
    "    if ~isnan(value)\n",
    "        if ischar(value)\n",
    "            if strcmpi(value,'X')\n",
    "                tcga_include_normal = 1;\n",
    "            else\n",
    "                error('ERROR - Samples - TCGA - Include normal tissue samples - Value must either be X or blank')\n",
    "            end\n",
    "        else\n",
    "            error('ERROR - Samples - TCGA - Include normal tissue samples - Value must either be X or blank')\n",
    "        end\n",
    "    else\n",
    "        tcga_include_normal = 0;\n",
    "    end\n",
    "\n",
    "    % needs doubling time\n",
    "    row = 13;\n",
    "    value = raw(row,5);\n",
    "    value = value{1};\n",
    "\n",
    "    if ~isnan(value)\n",
    "        if ischar(value)\n",
    "            if strcmpi(value,'X')\n",
    "                tcga_needs_proliferation = 1;\n",
    "            else\n",
    "                error('ERROR - Samples - TCGA - Restrict samples to those with doubling time - Value must either be X or blank')\n",
    "            end\n",
    "        else\n",
    "            error('ERROR - Samples - TCGA - Restrict samples to those with doubling time - Value must either be X or blank')\n",
    "        end\n",
    "    elseif (constraints_proliferation == 1) && (~isempty(samples_tcga))\n",
    "        error('ERROR - Samples - TCGA - Restrict samples to those with doubling time - Must choose this option if imposing proliferation constraint')\n",
    "    else\n",
    "        tcga_needs_proliferation = 0;\n",
    "    end\n",
    "\n",
    "    % needs mutation data\n",
    "    row = 14;\n",
    "    value = raw(row,5);\n",
    "    value = value{1};\n",
    "\n",
    "    if ~isnan(value)\n",
    "        if ischar(value)\n",
    "            if strcmpi(value,'X')\n",
    "                tcga_needs_mutation = 1;\n",
    "            else\n",
    "                error('ERROR - Samples - TCGA - Restrict samples to those with mutation data - Value must either be X or blank')\n",
    "            end\n",
    "        else\n",
    "            error('ERROR - Samples - TCGA - Restrict samples to those with mutation data - Value must either be X or blank')\n",
    "        end\n",
    "    elseif (use_mutation == 1) && (~isempty(samples_tcga))\n",
    "        error('ERROR - Samples - TCGA - Restrict samples to those with mutation data - Must choose this option if using mutation data')\n",
    "    else\n",
    "        tcga_needs_mutation = 0;\n",
    "    end\n",
    "\n",
    "    % needs radiation response\n",
    "    row = 15;\n",
    "    value = raw(row,5);\n",
    "    value = value{1};\n",
    "\n",
    "    if ~isnan(value)\n",
    "        if ischar(value)\n",
    "            if strcmpi(value,'X')\n",
    "                tcga_needs_radiation = 1;\n",
    "            else\n",
    "                error('ERROR - Samples - TCGA - Restrict samples to those with radiation response - Value must either be X or blank')\n",
    "            end\n",
    "        else\n",
    "            error('ERROR - Samples - TCGA - Restrict samples to those with radiation response - Value must either be X or blank')\n",
    "        end\n",
    "    else\n",
    "        tcga_needs_radiation = 0;\n",
    "    end\n",
    "\n",
    "    % needs any drug response\n",
    "    row = 16;\n",
    "    value = raw(row,5);\n",
    "    value = value{1};\n",
    "\n",
    "    if ~isnan(value)\n",
    "        if ischar(value)\n",
    "            if strcmpi(value,'X')\n",
    "                tcga_needs_any_drug = 1;\n",
    "            else\n",
    "                error('ERROR - Samples - TCGA - Restrict samples to those with any drug response - Value must either be X or blank')\n",
    "            end\n",
    "        else\n",
    "            error('ERROR - Samples - TCGA - Restrict samples to those with any drug response - Value must either be X or blank')\n",
    "        end\n",
    "    elseif (length(required_drugs) > 0) && (~isempty(samples_tcga))\n",
    "        error('ERROR - Samples - TCGA - Restrict samples to those with mutation data - Must choose this option if using mutation data')\n",
    "    else\n",
    "        tcga_needs_any_drug = 0;\n",
    "    end\n",
    "\n",
    "    % get TCGA samples\n",
    "    samples_all = {};\n",
    "    samples_all_source = {};\n",
    "    samples_all_source_protein = {};\n",
    "    if ~isempty(tcga_types)\n",
    "\n",
    "        % get list of all TCGA samples\n",
    "        files = dir('../data/clinical/TCGA/');\n",
    "        fn = {};\n",
    "        for i = 3:length(files)\n",
    "            temp = strsplit(files(i).name,'.csv');\n",
    "            fn{end+1} = temp{1};\n",
    "        end\n",
    "        \n",
    "        % get list of all TCGA vmax data\n",
    "        files = dir(sprintf('../data/vmax/%s/no_mutation/',tcga_protein));\n",
    "        fn_protein = {};\n",
    "        for i = 3:length(files)\n",
    "            temp = strsplit(files(i).name,'.csv');\n",
    "            fn_protein{end+1} = temp{1};\n",
    "        end\n",
    "        \n",
    "        % get all samples with mutation data\n",
    "        files = dir('../data/mutation/TCGA/');\n",
    "        fn_mutation = {};\n",
    "        for i = 3:length(files)\n",
    "            temp = strsplit(files(i).name,'.csv');\n",
    "            fn_mutation{end+1} = temp{1};\n",
    "        end\n",
    "        idh1_patients_H133Q = importdata('../data/idh1/H133Q.txt')';\n",
    "        idh1_patients_A134D = importdata('../data/idh1/A134D.txt')';\n",
    "        idh1_patients_R100Q = importdata('../data/idh1/R100Q.txt')';\n",
    "        idh1_patients_R132H = importdata('../data/idh1/R132H.txt')';\n",
    "        idh1_patients_R132C = importdata('../data/idh1/R132C.txt')';\n",
    "        idh1_patients_R132G = importdata('../data/idh1/R132G.txt')';\n",
    "        idh1_patients_R132W = importdata('../data/idh1/R132W.txt')';\n",
    "        idh1_patients_R132A = importdata('../data/idh1/R132A.txt')';\n",
    "        idh1_patients_R132Q = importdata('../data/idh1/R132Q.txt')';\n",
    "        idh1_patients_R132K = importdata('../data/idh1/R132K.txt')';\n",
    "        idh1_patients_R132N = importdata('../data/idh1/R132N.txt')';\n",
    "        fn_mutation = [fn_mutation,idh1_patients_H133Q,idh1_patients_A134D,idh1_patients_R100Q,idh1_patients_R132H,idh1_patients_R132C,idh1_patients_R132G,idh1_patients_R132W,idh1_patients_R132A,idh1_patients_R132Q,idh1_patients_R132K,idh1_patients_R132N];\n",
    "        \n",
    "        % iterate over samples\n",
    "        for i = 1:length(fn)\n",
    "            keep = 1;\n",
    "\n",
    "            % load clinical file\n",
    "            f = fopen(sprintf('../data/clinical/TCGA/%s.csv',fn{i}),'r');\n",
    "            data = textscan(f,'%s %s','Delimiter',',');\n",
    "            fclose(f);\n",
    "            \n",
    "            % determine if has vmax data\n",
    "            if ~any(strcmp(fn_protein,fn{i}))\n",
    "                keep = 0;\n",
    "\n",
    "            % determine if in selected cancer type\n",
    "            elseif ~any(strcmp(tcga_types,data{2}{find(strcmp(data{1},'COHORT'))}))\n",
    "                keep = 0;\n",
    "\n",
    "            % if primary tumor\n",
    "            elseif strcmp(data{2}{find(strcmp(data{1},'TYPE'))},'TUMOR') && (tcga_include_tumor == 0)\n",
    "                keep = 0;\n",
    "\n",
    "            % if normal tissue\n",
    "            elseif strcmp(data{2}{find(strcmp(data{1},'TYPE'))},'NORMAL') && (tcga_include_normal == 0)\n",
    "                keep = 0;\n",
    "\n",
    "            % if needs doubling time\n",
    "            else\n",
    "                [~, status] = str2num(data{2}{find(strcmp(data{1},'PROLIFERATION [1/hr]'))});\n",
    "                if (status == 0) && (tcga_needs_proliferation == 1)\n",
    "                    keep = 0;\n",
    "\n",
    "                % if needs radiation response\n",
    "                elseif strcmp(data{2}{find(strcmp(data{1},'RESPONSE RADIATION'))},'') && (tcga_needs_radiation == 1)\n",
    "                    keep = 0;\n",
    "\n",
    "                % if needs mutation data\n",
    "                elseif (tcga_needs_mutation == 1) && ~any(strcmp(fn_mutation,fn{i}))\n",
    "                    keep = 0;\n",
    "\n",
    "                % if needs any drug response \n",
    "                elseif tcga_needs_any_drug == 1     \n",
    "                    response = 0;\n",
    "                    for j = 1:length(data{1})\n",
    "                        temp = strsplit(data{1}{j},' ');\n",
    "                        if length(temp) > 1\n",
    "                            if strcmp(temp{2},'DRUG') && (~strcmp(data{2}{j},''))\n",
    "                                response = 1;\n",
    "                            end\n",
    "                        end\n",
    "                    end\n",
    "                    if response == 0\n",
    "                        keep = 0;\n",
    "                    end\n",
    "                end\n",
    "            end\n",
    "\n",
    "            % required drugs\n",
    "            if keep == 1\n",
    "                for j = 1:length(required_drugs)\n",
    "\n",
    "                    % is drug response available\n",
    "                    if ~any(strcmp(data{1},sprintf('RESPONSE DRUG %s',required_drugs{j})))\n",
    "                        keep = 0;\n",
    "                    elseif strcmp(data{2}{find(strcmp(data{1},sprintf('RESPONSE DRUG %s',required_drugs{j})))},'')\n",
    "                        keep = 0;\n",
    "                    end\n",
    "                end\n",
    "            end\n",
    "\n",
    "            % add to samples list\n",
    "            if keep == 1\n",
    "                samples_all{end+1} = fn{i};\n",
    "                samples_all_source{end+1} = 'TCGA';\n",
    "                samples_all_source_protein{end+1} = tcga_protein;\n",
    "            end\n",
    "        end\n",
    "    end\n",
    "end"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Samples - CCLE"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%file parse_samples_ccle.m\n",
    "\n",
    "function [samples_all,samples_all_source,samples_all_source_protein] = parse_samples_ccle(input_file,constraints_proliferation,required_drugs,use_mutation,samples_all,samples_all_source,samples_all_source_protein)\n",
    "    \n",
    "    % load data - CCLE\n",
    "    [~,~,raw] = xlsread(input_file,'Samples - CCLE');\n",
    "\n",
    "    % samples - CCLE\n",
    "    ccle_types = {};\n",
    "    ccle_needs_radiation = 0;\n",
    "    for row = 3:26\n",
    "        value = raw(row,1);\n",
    "        value = value{1};\n",
    "\n",
    "        if ~isnan(value)\n",
    "            if ischar(value)\n",
    "                if strcmpi(value,'X')\n",
    "                    switch row\n",
    "                        case 3, ccle_types{end+1} = 'Autonomic ganglia';\n",
    "                        case 4, ccle_types{end+1} = 'Biliary tract';\n",
    "                        case 5, ccle_types{end+1} = 'Bone';\n",
    "                        case 6, ccle_types{end+1} = 'Breast';\n",
    "                        case 7, ccle_types{end+1} = 'Central nervous system';\n",
    "                        case 8, ccle_types{end+1} = 'Endometrium';\n",
    "                        case 9, ccle_types{end+1} = 'Oesophagus';\n",
    "                        case 10, ccle_types{end+1} = 'Haematopoietic and lymphoid tissue';\n",
    "                        case 11, ccle_types{end+1} = 'Kidney';\n",
    "                        case 12, ccle_types{end+1} = 'Large intestine';\n",
    "                        case 13, ccle_types{end+1} = 'Liver';\n",
    "                        case 14, ccle_types{end+1} = 'Lung';\n",
    "                        case 15, ccle_types{end+1} = 'Ovary';\n",
    "                        case 16, ccle_types{end+1} = 'Pancreas';\n",
    "                        case 17, ccle_types{end+1} = 'Pleura';\n",
    "                        case 18, ccle_types{end+1} = 'Prostate';\n",
    "                        case 19, ccle_types{end+1} = 'Salivary gland';\n",
    "                        case 20, ccle_types{end+1} = 'Skin';\n",
    "                        case 21, ccle_types{end+1} = 'Small intestine';\n",
    "                        case 22, ccle_types{end+1} = 'Soft tissue';\n",
    "                        case 23, ccle_types{end+1} = 'Stomach';\n",
    "                        case 24, ccle_types{end+1} = 'Thyroid';\n",
    "                        case 25, ccle_types{end+1} = 'Upper aerodigestive tract';\n",
    "                        case 26, ccle_types{end+1} = 'Urinary tract';\n",
    "                    end\n",
    "                else\n",
    "                    error('ERROR - Samples - CCLE - Values must either be X or blank')\n",
    "                end\n",
    "            else\n",
    "                error('ERROR - Samples - CCLE - Values must either be X or blank')\n",
    "            end\n",
    "        end\n",
    "    end\n",
    " \n",
    "    % needs mutation data\n",
    "    row = 3;\n",
    "    value = raw(row,4);\n",
    "    value = value{1};\n",
    "\n",
    "    if ~isnan(value)\n",
    "        if ischar(value)\n",
    "            if strcmpi(value,'X')\n",
    "                ccle_needs_mutation = 1;\n",
    "            else\n",
    "                error('ERROR - Samples - CCLE - Restrict samples to those with mutation data - Value must either be X or blank')\n",
    "            end\n",
    "        else\n",
    "            error('ERROR - Samples - CCLE - Restrict samples to those with mutation data - Value must either be X or blank')\n",
    "        end\n",
    "    elseif (use_mutation == 1) && (~isempty(samples_ccle))\n",
    "        error('ERROR - Samples - CCLE - Restrict samples to those with mutation data - Must choose this option if using mutation data')\n",
    "    else\n",
    "        ccle_needs_mutation = 0;\n",
    "    end\n",
    "\n",
    "    % needs radiation response\n",
    "    row = 4;\n",
    "    value = raw(row,4);\n",
    "    value = value{1};\n",
    "\n",
    "    if ~isnan(value)\n",
    "        if ischar(value)\n",
    "            if strcmpi(value,'X')\n",
    "                ccle_needs_radiation = 1;\n",
    "            else\n",
    "                error('ERROR - Samples - CCLE - Restrict samples to those with radiation response - Value must either be X or blank')\n",
    "            end\n",
    "        else\n",
    "            error('ERROR - Samples - CCLE - Restrict samples to those with radiation response - Value must either be X or blank')\n",
    "        end\n",
    "    else\n",
    "        ccle_needs_radiation = 0;\n",
    "    end\n",
    "\n",
    "    % needs any drug response\n",
    "    row = 5;\n",
    "    value = raw(row,4);\n",
    "    value = value{1};\n",
    "\n",
    "    if ~isnan(value)\n",
    "        if ischar(value)\n",
    "            if strcmpi(value,'X')\n",
    "                ccle_needs_any_drug = 1;\n",
    "            else\n",
    "                error('ERROR - Samples - CCLE - Restrict samples to those with any drug response - Value must either be X or blank')\n",
    "            end\n",
    "        else\n",
    "            error('ERROR - Samples - CCLE - Restrict samples to those with any drug response - Value must either be X or blank')\n",
    "        end\n",
    "    elseif (length(required_drugs) > 0) && (~isempty(samples_ccle))\n",
    "        error('ERROR - Samples - CCLE - Restrict samples to those with mutation data - Must choose this option if using mutation data')\n",
    "    else\n",
    "        ccle_needs_any_drug = 0;\n",
    "    end\n",
    "\n",
    "    % get CCLE samples\n",
    "    if ~isempty(ccle_types)\n",
    "\n",
    "        % get list of all CCLE samples\n",
    "        files = dir('../data/clinical/CCLE/');\n",
    "        fn = {};\n",
    "        for i = 3:length(files)\n",
    "            temp = strsplit(files(i).name,'.csv');\n",
    "            fn{end+1} = temp{1};\n",
    "        end\n",
    "        \n",
    "        % get list of all CCLE vmax data\n",
    "        files = dir('../data/vmax/CCLE/no_mutation/');\n",
    "        fn_protein = {};\n",
    "        for i = 3:length(files)\n",
    "            temp = strsplit(files(i).name,'.csv');\n",
    "            fn_protein{end+1} = temp{1};\n",
    "        end\n",
    "        \n",
    "        % get all samples with mutation data\n",
    "        files = dir('../data/mutation/CCLE/');\n",
    "        fn_mutation = {};\n",
    "        for i = 3:length(files)\n",
    "            temp = strsplit(files(i).name,'.csv');\n",
    "            fn_mutation{end+1} = temp{1};\n",
    "        end\n",
    "        idh1_patients_H133Q = importdata('../data/idh1/H133Q.txt')';\n",
    "        idh1_patients_A134D = importdata('../data/idh1/A134D.txt')';\n",
    "        idh1_patients_R100Q = importdata('../data/idh1/R100Q.txt')';\n",
    "        idh1_patients_R132H = importdata('../data/idh1/R132H.txt')';\n",
    "        idh1_patients_R132C = importdata('../data/idh1/R132C.txt')';\n",
    "        idh1_patients_R132G = importdata('../data/idh1/R132G.txt')';\n",
    "        idh1_patients_R132W = importdata('../data/idh1/R132W.txt')';\n",
    "        idh1_patients_R132A = importdata('../data/idh1/R132A.txt')';\n",
    "        idh1_patients_R132Q = importdata('../data/idh1/R132Q.txt')';\n",
    "        idh1_patients_R132K = importdata('../data/idh1/R132K.txt')';\n",
    "        idh1_patients_R132N = importdata('../data/idh1/R132N.txt')';\n",
    "        fn_mutation = [fn_mutation,idh1_patients_H133Q,idh1_patients_A134D,idh1_patients_R100Q,idh1_patients_R132H,idh1_patients_R132C,idh1_patients_R132G,idh1_patients_R132W,idh1_patients_R132A,idh1_patients_R132Q,idh1_patients_R132K,idh1_patients_R132N];\n",
    "        \n",
    "        % iterate over samples\n",
    "        for i = 1:length(fn)\n",
    "            keep = 1;\n",
    "\n",
    "            % load clinical file\n",
    "            f = fopen(sprintf('../data/clinical/CCLE/%s.csv',fn{i}),'r');\n",
    "            data = textscan(f,'%s %s','Delimiter',',');\n",
    "            fclose(f);\n",
    "            \n",
    "            % determine if has vmax data\n",
    "            if ~any(strcmp(fn_protein,fn{i}))\n",
    "                keep = 0;\n",
    "\n",
    "            % determine if in selected cancer type\n",
    "            elseif ~any(strcmp(ccle_types,data{2}{find(strcmp(data{1},'PRIMARY SITE'))}))\n",
    "                keep = 0;\n",
    "\n",
    "            % if needs radiation response\n",
    "            elseif strcmp(data{2}{find(strcmp(data{1},'RADIATION AUC'))},'') && (ccle_needs_radiation == 1)\n",
    "                keep = 0;\n",
    "\n",
    "            % if needs mutation data\n",
    "            elseif (ccle_needs_mutation == 1) && ~any(strcmp(fn_mutation,fn{i}))\n",
    "                keep = 0;\n",
    "\n",
    "            % if needs any drug response \n",
    "            elseif ccle_needs_any_drug == 1     \n",
    "                response = 0;\n",
    "                for j = 1:length(data{1})\n",
    "                    temp = strsplit(data{1}{j},' ');\n",
    "                    if length(temp) > 1\n",
    "                        if strcmp(temp{2},'DRUG') && (~strcmp(data{2}{j},''))\n",
    "                            response = 1;\n",
    "                        end\n",
    "                    end\n",
    "                end\n",
    "                if response == 0\n",
    "                    keep = 0;\n",
    "                end\n",
    "            end\n",
    "\n",
    "            % required drugs\n",
    "            if keep == 1\n",
    "                for j = 1:length(required_drugs)\n",
    "\n",
    "                    % is drug response available\n",
    "                    if ~any(strcmp(data{1},sprintf('RESPONSE DRUG %s',required_drugs{j})))\n",
    "                        keep = 0;\n",
    "                    elseif strcmp(data{2}{find(strcmp(data{1},sprintf('RESPONSE DRUG %s',required_drugs{j})))},'')\n",
    "                        keep = 0;\n",
    "                    end\n",
    "                end\n",
    "            end\n",
    "\n",
    "            % add to samples list\n",
    "            if keep == 1\n",
    "                samples_all{end+1} = fn{i};\n",
    "                samples_all_source{end+1} = 'CCLE';\n",
    "                samples_all_source_protein{end+1} = 'CCLE';\n",
    "            end\n",
    "        end\n",
    "    end\n",
    "end"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Samples - GTEx"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%file parse_samples_gtex.m\n",
    "\n",
    "function [samples_all,samples_all_source,samples_all_source_protein] = parse_samples_gtex(input_file,constraints_proliferation,required_drugs,use_mutation,samples_all,samples_all_source,samples_all_source_protein)\n",
    "    \n",
    "    % load data - GTEx\n",
    "    [~,~,raw] = xlsread(input_file,'Samples - GTEx');\n",
    "\n",
    "    % samples - GTEx\n",
    "    gtex_types = {};\n",
    "    for row = 3:56\n",
    "        value = raw(row,1);\n",
    "        value = value{1};\n",
    "\n",
    "        if ~isnan(value)\n",
    "            if ischar(value)\n",
    "                if strcmpi(value,'X')\n",
    "                    switch row\n",
    "                        case 3, gtex_types{end+1} = 'Adipose - Subcutaneous';\n",
    "                        case 4, gtex_types{end+1} = 'Adipose - Subcutaneous';\n",
    "                        case 5, gtex_types{end+1} = 'Adrenal Gland';\n",
    "                        case 6, gtex_types{end+1} = 'Artery - Aorta';\n",
    "                        case 7, gtex_types{end+1} = 'Artery - Coronary';\n",
    "                        case 8, gtex_types{end+1} = 'Artery - Tibial';\n",
    "                        case 9, gtex_types{end+1} = 'Bladder';\n",
    "                        case 10, gtex_types{end+1} = 'Brain - Amygdala';\n",
    "                        case 11, gtex_types{end+1} = 'Brain - Anterior cingulate cortex (BA24)';\n",
    "                        case 12, gtex_types{end+1} = 'Brain - Caudate (basal ganglia)';\n",
    "                        case 13, gtex_types{end+1} = 'Brain - Cerebellar Hemisphere';\n",
    "                        case 14, gtex_types{end+1} = 'Brain - Cerebellum';\n",
    "                        case 15, gtex_types{end+1} = 'Brain - Cortex';\n",
    "                        case 16, gtex_types{end+1} = 'Brain - Frontal Cortex (BA9)';\n",
    "                        case 17, gtex_types{end+1} = 'Brain - Hippocampus';\n",
    "                        case 18, gtex_types{end+1} = 'Brain - Hypothalamus';\n",
    "                        case 19, gtex_types{end+1} = 'Brain - Nucleus accumbens (basal ganglia)';\n",
    "                        case 20, gtex_types{end+1} = 'Brain - Putamen (basal ganglia)';\n",
    "                        case 21, gtex_types{end+1} = 'Brain - Spinal cord (cervical c-1)';\n",
    "                        case 22, gtex_types{end+1} = 'Brain - Substantia nigra';\n",
    "                        case 23, gtex_types{end+1} = 'Breast - Mammary Tissue';\n",
    "                        case 24, gtex_types{end+1} = 'Cells - EBV-transformed lymphocytes';\n",
    "                        case 25, gtex_types{end+1} = 'Cells - Leukemia cell line (CML)';\n",
    "                        case 26, gtex_types{end+1} = 'Cells - Transformed fibroblasts';\n",
    "                        case 27, gtex_types{end+1} = 'Cervix - Ectocervix';\n",
    "                        case 28, gtex_types{end+1} = 'Cervix - Endocervix';\n",
    "                        case 29, gtex_types{end+1} = 'Colon - Sigmoid';\n",
    "                        case 30, gtex_types{end+1} = 'Colon - Transverse';\n",
    "                        case 31, gtex_types{end+1} = 'Esophagus - Gastroesophageal Junction';\n",
    "                        case 32, gtex_types{end+1} = 'Esophagus - Mucosa';\n",
    "                        case 33, gtex_types{end+1} = 'Esophagus - Muscularis';\n",
    "                        case 34, gtex_types{end+1} = 'Fallopian Tube';\n",
    "                        case 35, gtex_types{end+1} = 'Heart - Atrial Appendage';\n",
    "                        case 36, gtex_types{end+1} = 'Heart - Left Ventricle';\n",
    "                        case 37, gtex_types{end+1} = 'Kidney - Cortex';\n",
    "                        case 38, gtex_types{end+1} = 'Liver';\n",
    "                        case 39, gtex_types{end+1} = 'Lung';\n",
    "                        case 40, gtex_types{end+1} = 'Minor Salivary Gland';\n",
    "                        case 41, gtex_types{end+1} = 'Muscle - Skeletal';\n",
    "                        case 42, gtex_types{end+1} = 'Nerve - Tibial';\n",
    "                        case 43, gtex_types{end+1} = 'Ovary';\n",
    "                        case 44, gtex_types{end+1} = 'Pancreas';\n",
    "                        case 45, gtex_types{end+1} = 'Pituitary';\n",
    "                        case 46, gtex_types{end+1} = 'Prostate';\n",
    "                        case 47, gtex_types{end+1} = 'Skin - Not Sun Exposed (Suprapubic)';\n",
    "                        case 48, gtex_types{end+1} = 'Skin - Sun Exposed (Lower leg)';\n",
    "                        case 49, gtex_types{end+1} = 'Small Intestine - Terminal Ileum';\n",
    "                        case 50, gtex_types{end+1} = 'Spleen';\n",
    "                        case 51, gtex_types{end+1} = 'Stomach';\n",
    "                        case 52, gtex_types{end+1} = 'Testis';\n",
    "                        case 53, gtex_types{end+1} = 'Thyroid';\n",
    "                        case 54, gtex_types{end+1} = 'Uterus';\n",
    "                        case 55, gtex_types{end+1} = 'Vagina';\n",
    "                        case 56, gtex_types{end+1} = 'Whole Blood';\n",
    "                    end\n",
    "                else\n",
    "                    error('ERROR - Samples - GTEx - Values must either be X or blank')\n",
    "                end\n",
    "            else\n",
    "                error('ERROR - Samples - GTEx - Values must either be X or blank')\n",
    "            end\n",
    "        end\n",
    "    end\n",
    " \n",
    "    % get GTEx samples\n",
    "    if ~isempty(gtex_types)\n",
    "\n",
    "        % get list of all GTEx samples\n",
    "        files = dir('../data/clinical/GTEx/');\n",
    "        fn = {};\n",
    "        for i = 3:length(files)\n",
    "            temp = strsplit(files(i).name,'.csv');\n",
    "            fn{end+1} = temp{1};\n",
    "        end\n",
    "        \n",
    "        % get list of all GTEx vmax data\n",
    "        files = dir('../data/vmax/GTEx/no_mutation/');\n",
    "        fn_protein = {};\n",
    "        for i = 3:length(files)\n",
    "            temp = strsplit(files(i).name,'.csv');\n",
    "            fn_protein{end+1} = temp{1};\n",
    "        end\n",
    "        \n",
    "        % iterate over samples\n",
    "        for i = 1:length(fn)\n",
    "            keep = 1;\n",
    "\n",
    "            % load clinical file\n",
    "            f = fopen(sprintf('../data/clinical/GTEx/%s.csv',fn{i}),'r');\n",
    "            data = textscan(f,'%s %s','Delimiter',',');\n",
    "            fclose(f);\n",
    "            \n",
    "            % determine if has vmax data\n",
    "            if ~any(strcmp(fn_protein,fn{i}))\n",
    "                keep = 0;\n",
    "\n",
    "            % determine if in selected cancer type\n",
    "            elseif ~any(strcmp(gtex_types,data{2}{find(strcmp(data{1},'TISSUE TYPE'))}))\n",
    "                keep = 0;\n",
    "\n",
    "            % required drugs\n",
    "            elseif length(required_drugs) > 0\n",
    "                keep = 0;\n",
    "            end\n",
    "\n",
    "            % add to samples list\n",
    "            if keep == 1\n",
    "                samples_all{end+1} = fn{i};\n",
    "                samples_all_source{end+1} = 'GTEx';\n",
    "                samples_all_source_protein{end+1} = 'GTEx';\n",
    "            end\n",
    "        end\n",
    "    end\n",
    "end"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Samples - Other"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%file parse_samples_other.m\n",
    "\n",
    "function [samples_all,samples_all_source,samples_all_source_protein] = parse_samples_other(input_file,constraints_proliferation,required_drugs,use_mutation,samples_all,samples_all_source,samples_all_source_protein)\n",
    "    \n",
    "    % load data - Other samples\n",
    "    [~,~,raw] = xlsread(input_file,'Samples - Other');\n",
    "    \n",
    "    % other samples\n",
    "    other_samples_folder = {};\n",
    "    other_samples = {};\n",
    "\n",
    "    for row = 2:size(raw,1)\n",
    "        added = false;\n",
    "\n",
    "        % folder name\n",
    "        value = raw(row,1);\n",
    "        value = value{1};\n",
    "        if ~isnan(value)\n",
    "            if ischar(value)\n",
    "                \n",
    "                % sample folder\n",
    "                other_samples_folder{end+1} = value;\n",
    "                \n",
    "                % sample name\n",
    "                value = raw(row,2);\n",
    "                value = value{1};\n",
    "                if ~isnan(value)\n",
    "                    if ischar(value)\n",
    "                        other_samples{end+1} = value;\n",
    "                    elseif ischar(num2str(value))\n",
    "                        other_samples{end+1} = num2str(value);\n",
    "                    else\n",
    "                        error('ERROR - Samples - Other - Sample name must be a valid string')\n",
    "                    end\n",
    "                else\n",
    "                    error('ERROR - Samples - Other - Must enter sample name if provide sample folder')\n",
    "                end\n",
    "            end\n",
    "        end\n",
    "    end\n",
    "    \n",
    "    % needs doubling time\n",
    "    row = 6;\n",
    "    value = raw(row,4);\n",
    "    value = value{1};\n",
    "\n",
    "    if ~isnan(value)\n",
    "        if ischar(value)\n",
    "            if strcmpi(value,'X')\n",
    "                other_needs_proliferation = 1;\n",
    "            else\n",
    "                error('ERROR - Samples - Other - Restrict samples to those with doubling time - Value must either be X or blank')\n",
    "            end\n",
    "        else\n",
    "            error('ERROR - Samples - Other - Restrict samples to those with doubling time - Value must either be X or blank')\n",
    "        end\n",
    "    elseif (constraints_proliferation == 1) && (~isempty(other_samples))\n",
    "        error('ERROR - Samples - Other - Restrict samples to those with doubling time - Must choose this option if imposing proliferation constraint')\n",
    "    else\n",
    "        other_needs_proliferation = 0;\n",
    "    end\n",
    "\n",
    "    % needs mutation data\n",
    "    row = 7;\n",
    "    value = raw(row,4);\n",
    "    value = value{1};\n",
    "\n",
    "    if ~isnan(value)\n",
    "        if ischar(value)\n",
    "            if strcmpi(value,'X')\n",
    "                other_needs_mutation = 1;\n",
    "            else\n",
    "                error('ERROR - Samples - Other - Restrict samples to those with mutation data - Value must either be X or blank')\n",
    "            end\n",
    "        else\n",
    "            error('ERROR - Samples - Other - Restrict samples to those with mutation data - Value must either be X or blank')\n",
    "        end\n",
    "    elseif (use_mutation == 1) && (~isempty(other_samples))\n",
    "        error('ERROR - Samples - Other - Restrict samples to those with mutation data - Must choose this option if using mutation data')\n",
    "    else\n",
    "        other_needs_mutation = 0;\n",
    "    end\n",
    "\n",
    "    % needs any drug response\n",
    "    row = 8;\n",
    "    value = raw(row,4);\n",
    "    value = value{1};\n",
    "\n",
    "    if ~isnan(value)\n",
    "        if ischar(value)\n",
    "            if strcmpi(value,'X')\n",
    "                other_needs_any_drug = 1;\n",
    "            else\n",
    "                error('ERROR - Samples - Other - Restrict samples to those with any drug response - Value must either be X or blank')\n",
    "            end\n",
    "        else\n",
    "            error('ERROR - Samples - Other - Restrict samples to those with any drug response - Value must either be X or blank')\n",
    "        end\n",
    "    elseif (length(required_drugs) > 0) && (~isempty(other_samples))\n",
    "        error('ERROR - Samples - Other - Restrict samples to those with mutation data - Must choose this option if using mutation data')\n",
    "    else\n",
    "        other_needs_any_drug = 0;\n",
    "    end\n",
    "    \n",
    "    % extract samples from _ALL_ folders\n",
    "    remove = [];\n",
    "    for i = 1:length(other_samples_folder)\n",
    "        if strcmp(other_samples{i},'_ALL_')\n",
    "            \n",
    "            % remove index\n",
    "            remove(end+1) = i;\n",
    "            \n",
    "            % get samples in folder\n",
    "            files = dir(sprintf('../data/vmax/%s/no_mutation/',other_samples_folder{i}));\n",
    "            fn_protein = {};\n",
    "            for j = 3:length(files)\n",
    "                temp = strsplit(files(j).name,'.csv');\n",
    "                fn_protein{end+1} = temp{1};\n",
    "            end\n",
    "            \n",
    "            % add samples\n",
    "            for j = 1:length(fn_protein)\n",
    "                other_samples_folder{end+1} = other_samples_folder{i};\n",
    "                other_samples{end+1} = fn_protein{j};\n",
    "            end \n",
    "        end\n",
    "    end\n",
    "    for i = fliplr(sort(remove))\n",
    "        other_samples_folder(i) = [];\n",
    "        other_samples(i) = [];\n",
    "    end\n",
    "    \n",
    "    % patients with idh1 mutations\n",
    "    idh1_patients_H133Q = importdata('../data/idh1/H133Q.txt')';\n",
    "    idh1_patients_A134D = importdata('../data/idh1/A134D.txt')';\n",
    "    idh1_patients_R100Q = importdata('../data/idh1/R100Q.txt')';\n",
    "    idh1_patients_R132H = importdata('../data/idh1/R132H.txt')';\n",
    "    idh1_patients_R132C = importdata('../data/idh1/R132C.txt')';\n",
    "    idh1_patients_R132G = importdata('../data/idh1/R132G.txt')';\n",
    "    idh1_patients_R132W = importdata('../data/idh1/R132W.txt')';\n",
    "    idh1_patients_R132A = importdata('../data/idh1/R132A.txt')';\n",
    "    idh1_patients_R132Q = importdata('../data/idh1/R132Q.txt')';\n",
    "    idh1_patients_R132K = importdata('../data/idh1/R132K.txt')';\n",
    "    idh1_patients_R132N = importdata('../data/idh1/R132N.txt')';\n",
    "    \n",
    "    % iterate over sample folders\n",
    "    for i = 1:length(other_samples_folder)\n",
    "        keep = 1;\n",
    "        \n",
    "        % folder\n",
    "        temp = strsplit(other_samples_folder{i},'_');\n",
    "        prefix = temp{1};\n",
    "        switch prefix\n",
    "            case 'TCGA'\n",
    "                nonprotein_folder = 'TCGA';\n",
    "            case 'NCI60'\n",
    "                nonprotein_folder = 'NCI60';\n",
    "            otherwise\n",
    "                nonprotein_folder = other_samples_folder{i};\n",
    "        end\n",
    "\n",
    "        % check vmax data\n",
    "        if ~(exist(sprintf('../data/vmax/%s',other_samples_folder{i})) == 7)\n",
    "            error('ERROR - Samples - Other - Folder \"%s\" does not exist in vmax data',other_samples_folder{i})\n",
    "        else\n",
    "            files = dir(sprintf('../data/vmax/%s/no_mutation/',other_samples_folder{i}));\n",
    "            fn_protein = {};\n",
    "            for j = 3:length(files)\n",
    "                temp = strsplit(files(j).name,'.csv');\n",
    "                fn_protein{end+1} = temp{1};\n",
    "            end\n",
    "            if ~any(strcmp(fn_protein,other_samples{i}))\n",
    "                error('ERROR - Samples - Other - Sample \"%s\" does not exist in folder \"%s\" of vmax data',other_samples{i},other_samples_folder{i})\n",
    "            else\n",
    "                \n",
    "                % check mutation data\n",
    "                if (other_needs_mutation == 1)\n",
    "                    if ~(exist(sprintf('../data/mutation/%s',nonprotein_folder)) == 7)\n",
    "                        keep = 0;\n",
    "                    else\n",
    "                        files = dir(sprintf('../data/mutation/%s/',nonprotein_folder));\n",
    "                        fn_mutation = {};\n",
    "                        for j = 3:length(files)\n",
    "                            temp = strsplit(files(j).name,'.csv');\n",
    "                            fn_mutation{end+1} = temp{1};\n",
    "                        end\n",
    "                        if strcmp(prefix,'TCGA')\n",
    "                            fn_mutation = [fn_mutation,idh1_patients_H133Q,idh1_patients_A134D,idh1_patients_R100Q,idh1_patients_R132H,idh1_patients_R132C,idh1_patients_R132G,idh1_patients_R132W,idh1_patients_R132A,idh1_patients_R132Q,idh1_patients_R132K,idh1_patients_R132N];\n",
    "                        end\n",
    "                        if ~any(strcmp(fn_mutation,other_samples{i}))\n",
    "                            keep = 0;\n",
    "                        else\n",
    "                \n",
    "                            % check clinical data\n",
    "                            if (other_needs_proliferation == 1) || (other_nees_any_drug == 1)\n",
    "                                if ~(exist(sprintf('../data/clinical/%s',nonprotein_folder)) == 7)\n",
    "                                    keep = 0;\n",
    "                                else\n",
    "                                    files = dir(sprintf('../data/clinical/%s/',nonprotein_folder));\n",
    "                                    fn_protein = {};\n",
    "                                    for j = 3:length(files)\n",
    "                                        temp = strsplit(files(j).name,'.csv');\n",
    "                                        fn_protein{end+1} = temp{1};\n",
    "                                    end\n",
    "                                    if ~any(strcmp(fn_protein,other_samples{i}))\n",
    "                                        keep = 0;\n",
    "                                    else\n",
    "                                    \n",
    "                                        % load clinical data\n",
    "                                        f = fopen(sprintf('../data/clinical/%s/%s.csv',nonprotein_folder,other_samples{i}),'r');\n",
    "                                        data = textscan(f,'%s %s','Delimiter',',');\n",
    "                                        fclose(f);\n",
    "\n",
    "                                        % if needs doubling time\n",
    "                                        [~, status] = str2num(data{2}{find(strcmp(data{1},'PROLIFERATION [1/hr]'))});\n",
    "                                        if (status == 0) && (other_needs_proliferation == 1)\n",
    "                                            keep = 0;\n",
    "\n",
    "                                        % if needs any drug response \n",
    "                                        elseif other_needs_any_drug == 1     \n",
    "                                            response = 0;\n",
    "                                            for j = 1:length(data{1})\n",
    "                                                temp = strsplit(data{1}{j},' ');\n",
    "                                                if length(temp) > 1\n",
    "                                                    if strcmp(temp{2},'DRUG') && (~strcmp(data{2}{j},''))\n",
    "                                                        response = 1;\n",
    "                                                    end\n",
    "                                                end\n",
    "                                            end\n",
    "                                            if response == 0\n",
    "                                                keep = 0;\n",
    "                                            end\n",
    "                                            \n",
    "                                            % required drugs\n",
    "                                            if keep == 1\n",
    "                                                for j = 1:length(required_drugs)\n",
    "\n",
    "                                                    % is drug response available\n",
    "                                                    if ~any(strcmp(data{1},sprintf('RESPONSE DRUG %s',required_drugs{j})))\n",
    "                                                        keep = 0;\n",
    "                                                    elseif strcmp(data{2}{find(strcmp(data{1},sprintf('RESPONSE DRUG %s',required_drugs{j})))},'')\n",
    "                                                        keep = 0;\n",
    "                                                    end\n",
    "                                                end\n",
    "                                            end\n",
    "                                        end\n",
    "                                    end\n",
    "                                end\n",
    "                            end\n",
    "                        end\n",
    "                    end\n",
    "                end\n",
    "            end\n",
    "        end\n",
    "\n",
    "        % add to samples list\n",
    "        if keep == 1\n",
    "            samples_all{end+1} = other_samples{i};\n",
    "            samples_all_source{end+1} = nonprotein_folder;\n",
    "            samples_all_source_protein{end+1} = other_samples_folder{i};\n",
    "        end\n",
    "    end\n",
    "end"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Media"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%file parse_media.m\n",
    "\n",
    "function [media_choice,media_constraint_metabolite,media_constraint_uptake] = parse_media(input_file)\n",
    "\n",
    "    % load data\n",
    "    [~,~,raw] = xlsread(input_file,'Media');\n",
    "    \n",
    "    % media to add\n",
    "    media_choice = {};\n",
    "    for row = 9:size(raw,1)\n",
    "        value = raw(row,1);\n",
    "        value = value{1};\n",
    "\n",
    "        if ~isnan(value)\n",
    "            if ischar(value)\n",
    "                media_choice{end+1} = value;\n",
    "            else\n",
    "                error('ERROR - Media - Medium Makeup - Value must either be X or blank')\n",
    "            end\n",
    "        end\n",
    "    end\n",
    "    \n",
    "    % specific media constraints\n",
    "    media_constraint_metabolite = {};\n",
    "    media_constraint_uptake = [];\n",
    "\n",
    "    for row = 3:size(raw,1)\n",
    "        value = raw(row,5);\n",
    "        value = value{1};\n",
    "        \n",
    "        % metabolite name\n",
    "        if ~isnan(value)\n",
    "            if ischar(value)\n",
    "                media_constraint_metabolite{end+1} = value;\n",
    "\n",
    "                % uptake rate\n",
    "                value = raw(row,6);\n",
    "                value = value{1};\n",
    "                if ~isnan(value)\n",
    "                    if isfloat(value)           \n",
    "                        if value >= 0\n",
    "                            media_constraint_uptake(end+1) = value;\n",
    "                        else\n",
    "                            error('ERROR - Media - Specific Media Constraints - Maximum uptake rate must be >= 0')\n",
    "                        end\n",
    "                    else\n",
    "                        error('ERROR - Media - Specific Media Constraints - Maximum uptake rate must be a number')\n",
    "                    end\n",
    "                else\n",
    "                    error('ERROR - Media - Specific Media Constraints - Must give maximum uptake rate')\n",
    "                end\n",
    "            else\n",
    "                error('ERROR - Media - Specific Media Constraints - Metabolite name must be a string')\n",
    "            end\n",
    "        end\n",
    "    end\n",
    "end"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Concentrations"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%file parse_concentrations.m\n",
    "\n",
    "function [recalculate_thermodynamics,default_concentration_lb,default_concentration_ub,concentration_ranges_all,concentration_ranges_sample_folder,concentration_ranges_sample] = parse_concentrations(input_file)\n",
    "\n",
    "    % load data\n",
    "    [~,~,raw] = xlsread(input_file,'Concentrations');\n",
    "\n",
    "    % recalculate thermodynamics\n",
    "    row = 3;\n",
    "    value = raw(row,1);\n",
    "    value = value{1};\n",
    "    if ~isnan(value)\n",
    "        if ischar(value)\n",
    "            if strcmpi(value,'X')\n",
    "                recalculate_thermodynamics = 1;\n",
    "            else\n",
    "                error('ERROR - Concentrations - Recalculate Thermodynamics - Value must either be X or blank')\n",
    "            end\n",
    "        else\n",
    "            error('ERROR - Concentrations - Recalculate Thermodynamics - Value must either be X or blank')\n",
    "        end\n",
    "    else\n",
    "        recalculate_thermodynamics = 0;\n",
    "    end\n",
    "\n",
    "    row = 4;\n",
    "    value = raw(row,1);\n",
    "    value = value{1};\n",
    "    if ~isnan(value)\n",
    "        if ischar(value)\n",
    "            if strcmpi(value,'X')\n",
    "                if recalculate_thermodynamics == 1;\n",
    "                   error('ERROR - Concentrations - Recalculate Thermodynamics - Cannot select both Yes and No') \n",
    "                end\n",
    "            else\n",
    "                error('ERROR - Concentrations - Recalculate Thermodynamics - Value must either be X or blank')\n",
    "            end\n",
    "        else\n",
    "            error('ERROR - Concentrations - Recalculate Thermodynamics - Value must either be X or blank')\n",
    "        end\n",
    "    elseif recalculate_thermodynamics == 0;\n",
    "        error('ERROR - Concentrations - Recalculate Thermodynamics - Must select either Yes and No')\n",
    "    end\n",
    "    \n",
    "    % default concentration ranges - lower bound\n",
    "    value = raw(8,1);\n",
    "    value = value{1};\n",
    "    if recalculate_thermodynamics == 1\n",
    "        if ~isnan(value)\n",
    "            if isfloat(value)\n",
    "                if value >= 0\n",
    "                    default_concentration_lb = value;\n",
    "                else\n",
    "                    error('ERROR - Concentrations - Default Lower Bound - Value must be >= 0')\n",
    "                end\n",
    "            else\n",
    "                error('ERROR - Concentrations - Default Lower Bound - Value must be a number')\n",
    "            end\n",
    "        else\n",
    "            error('ERROR - Concentrations - Default Lower Bound - Must provide value if recalculating thermodynamics')\n",
    "        end\n",
    "    else\n",
    "        default_concentration_lb = NaN;\n",
    "    end\n",
    "    \n",
    "    % default concentration ranges - upper bound\n",
    "    value = raw(9,1);\n",
    "    value = value{1};\n",
    "    if recalculate_thermodynamics == 1\n",
    "        if ~isnan(value)\n",
    "            if isfloat(value)\n",
    "                if value >= 0\n",
    "                    if value >= default_concentration_lb\n",
    "                        default_concentration_ub = value;\n",
    "                    else\n",
    "                        error('ERROR - Concentrations - Default Upper Bound - Value must be >= default lower bound')\n",
    "                    end\n",
    "                else\n",
    "                    error('ERROR - Concentrations - Default Upper Bound - Value must be >= 0')\n",
    "                end\n",
    "            else\n",
    "                error('ERROR - Concentrations - Default Upper Bound - Value must be a number')\n",
    "            end\n",
    "        else\n",
    "            error('ERROR - Concentrations - Default Upper Bound - Must provide value if recalculating thermodynamics')\n",
    "        end\n",
    "    else\n",
    "        default_concentration_ub = NaN;\n",
    "    end\n",
    "    \n",
    "    % concentration values/ranges applied to all samples\n",
    "    concentration_ranges_all = {};\n",
    "    for row = 13:size(raw,1)\n",
    "        value = raw(row,2);\n",
    "        value = value{1};\n",
    "\n",
    "        if ~isnan(value)\n",
    "            if ischar(value)\n",
    "                concentration_ranges_all{end+1} = value;\n",
    "            else\n",
    "                error('ERROR - Concentrations - Ranges Applied to All Samples - File name must be valid string')\n",
    "            end\n",
    "        end\n",
    "    end\n",
    "    \n",
    "    % concentration values/ranges applied to individual samples\n",
    "    concentration_ranges_sample_folder = {};\n",
    "    concentration_ranges_sample = {};\n",
    "    for row = 13:size(raw,1)\n",
    "\n",
    "        value = raw(row,4);\n",
    "        value = value{1};\n",
    "        if ~isnan(value)\n",
    "            if ischar(value)\n",
    "            \n",
    "                % folder name\n",
    "                concentration_ranges_sample_folder{end+1} = value;\n",
    "                \n",
    "                % sample name\n",
    "                value = raw(row,5);\n",
    "                value = value{1};\n",
    "                if ~isnan(value)\n",
    "                    if ischar(value)\n",
    "                        concentration_ranges_sample{end+1} = value;\n",
    "                    else\n",
    "                        error('ERROR - Concentrations - Ranges Applied to Individual Samples - Sample name must be valid string')\n",
    "                    end\n",
    "                else\n",
    "                    error('ERROR - Concentrations - Ranges Applied to Individual Samples - Must provide sample name')\n",
    "                end       \n",
    "            else\n",
    "                error('ERROR - Concentrations - Ranges Applied to Individual Samples - Folder name must be valid string')\n",
    "            end\n",
    "        end\n",
    "     end\n",
    "end"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Modules"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%file parse_modules.m\n",
    "\n",
    "function [catalase_version, module_5fu, module_cis, module_cpa, module_dox] = parse_modules(input_file)\n",
    "\n",
    "    % load data\n",
    "    [~,~,raw] = xlsread(input_file,'Modules');\n",
    "    \n",
    "    % catalase version\n",
    "    value = raw(3,1);\n",
    "    value = value{1};\n",
    "    if ~isnan(value)\n",
    "        if ischar(value)\n",
    "            if strcmpi(value,'X')\n",
    "                catalase_version = 1;\n",
    "            else\n",
    "                error('ERROR - Modules - Catalase Version - Value must either be X or blank')\n",
    "            end\n",
    "        else\n",
    "            error('ERROR - Modules - Catalase Version - Value must either be X or blank')\n",
    "        end\n",
    "    else\n",
    "        catalase_version = 0;\n",
    "    end\n",
    "    \n",
    "    value = raw(4,1);\n",
    "    value = value{1};\n",
    "    if ~isnan(value)\n",
    "        if ischar(value)\n",
    "            if strcmpi(value,'X')\n",
    "                if catalase_version == 1;\n",
    "                   error('ERROR - Modules - Catalase Version - Cannot select both options') \n",
    "                end\n",
    "            else\n",
    "                error('ERROR - Modules - Catalase Version - Value must either be X or blank')\n",
    "            end\n",
    "        else\n",
    "            error('ERROR - Modules - Catalase Version - Value must either be X or blank')\n",
    "        end\n",
    "    elseif catalase_version == 0;\n",
    "        error('ERROR - Modules - Catalase Version - Must select an option')\n",
    "    end\n",
    "    \n",
    "    % 5-fluorouracil module\n",
    "    value = raw(8,1);\n",
    "    value = value{1};\n",
    "    if ~isnan(value)\n",
    "        if ischar(value)\n",
    "            if strcmpi(value,'X')\n",
    "                module_5fu = 1;\n",
    "            else\n",
    "                error('ERROR - Modules - 5-Fluorouracil - Value must either be X or blank')\n",
    "            end\n",
    "        else\n",
    "            error('ERROR - Modules - 5-Fluorouracil - Value must either be X or blank')\n",
    "        end\n",
    "    else\n",
    "        module_5fu = 0;\n",
    "    end\n",
    "    \n",
    "    % cisplatin module\n",
    "    value = raw(9,1);\n",
    "    value = value{1};\n",
    "    if ~isnan(value)\n",
    "        if ischar(value)\n",
    "            if strcmpi(value,'X')\n",
    "                module_cis = 1;\n",
    "            else\n",
    "                error('ERROR - Modules - Cisplatin - Value must either be X or blank')\n",
    "            end\n",
    "        else\n",
    "            error('ERROR - Modules - Cisplatin - Value must either be X or blank')\n",
    "        end\n",
    "    else\n",
    "        module_cis = 0;\n",
    "    end\n",
    "    \n",
    "    % cyclophosphamide module\n",
    "    value = raw(10,1);\n",
    "    value = value{1};\n",
    "    if ~isnan(value)\n",
    "        if ischar(value)\n",
    "            if strcmpi(value,'X')\n",
    "                module_cpa = 1;\n",
    "            else\n",
    "                error('ERROR - Modules - Cyclophosphamide - Value must either be X or blank')\n",
    "            end\n",
    "        else\n",
    "            error('ERROR - Modules - Cyclophosphamide - Value must either be X or blank')\n",
    "        end\n",
    "    else\n",
    "        module_cpa = 0;\n",
    "    end\n",
    "    \n",
    "    % doxorubicin module\n",
    "    value = raw(11,1);\n",
    "    value = value{1};\n",
    "    if ~isnan(value)\n",
    "        if ischar(value)\n",
    "            if strcmpi(value,'X')\n",
    "                module_dox = 1;\n",
    "            else\n",
    "                error('ERROR - Modules - Doxorubicin - Value must either be X or blank')\n",
    "            end\n",
    "        else\n",
    "            error('ERROR - Modules - Doxorubicin - Value must either be X or blank')\n",
    "        end\n",
    "    else\n",
    "        module_dox = 0;\n",
    "    end\n",
    " \n",
    "end"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Gene Knockdown"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%file parse_knockdown.m\n",
    "\n",
    "function [knockdown_genes,knockdown_fractions] = parse_knockdown(input_file)\n",
    "\n",
    "    % load data\n",
    "    [~,~,raw] = xlsread(input_file,'Gene Knockdown');\n",
    "\n",
    "    % genes and fractions to knockdown\n",
    "    knockdown_genes = {};\n",
    "    knockdown_fractions = [];\n",
    "    for row = 3:3270\n",
    "        value = raw(row,1);\n",
    "        value = value{1};\n",
    "        if ~isnan(value)\n",
    "            if ischar(value)\n",
    "                if strcmpi(value,'X')\n",
    "\n",
    "                    % knockdown gene\n",
    "                    switch row\n",
    "                        case 3, knockdown_genes{end+1} = '10';\n",
    "                        case 4, knockdown_genes{end+1} = '100';\n",
    "                        case 5, knockdown_genes{end+1} = '10000';\n",
    "                        case 6, knockdown_genes{end+1} = '10005';\n",
    "                        case 7, knockdown_genes{end+1} = '10007';\n",
    "                        case 8, knockdown_genes{end+1} = '100137049';\n",
    "                        case 9, knockdown_genes{end+1} = '10020';\n",
    "                        case 10, knockdown_genes{end+1} = '10026';\n",
    "                        case 11, knockdown_genes{end+1} = '100287513';\n",
    "                        case 12, knockdown_genes{end+1} = '100287639';\n",
    "                        case 13, knockdown_genes{end+1} = '100288072';\n",
    "                        case 14, knockdown_genes{end+1} = '10038';\n",
    "                        case 15, knockdown_genes{end+1} = '10039';\n",
    "                        case 16, knockdown_genes{end+1} = '10050';\n",
    "                        case 17, knockdown_genes{end+1} = '100526794';\n",
    "                        case 18, knockdown_genes{end+1} = '100529261';\n",
    "                        case 19, knockdown_genes{end+1} = '10054';\n",
    "                        case 20, knockdown_genes{end+1} = '10055';\n",
    "                        case 21, knockdown_genes{end+1} = '10056';\n",
    "                        case 22, knockdown_genes{end+1} = '10057';\n",
    "                        case 23, knockdown_genes{end+1} = '10058';\n",
    "                        case 24, knockdown_genes{end+1} = '10060';\n",
    "                        case 25, knockdown_genes{end+1} = '10075';\n",
    "                        case 26, knockdown_genes{end+1} = '10076';\n",
    "                        case 27, knockdown_genes{end+1} = '10087';\n",
    "                        case 28, knockdown_genes{end+1} = '10090';\n",
    "                        case 29, knockdown_genes{end+1} = '100996939';\n",
    "                        case 30, knockdown_genes{end+1} = '10105';\n",
    "                        case 31, knockdown_genes{end+1} = '10106';\n",
    "                        case 32, knockdown_genes{end+1} = '10110';\n",
    "                        case 33, knockdown_genes{end+1} = '10114';\n",
    "                        case 34, knockdown_genes{end+1} = '10130';\n",
    "                        case 35, knockdown_genes{end+1} = '10134';\n",
    "                        case 36, knockdown_genes{end+1} = '10135';\n",
    "                        case 37, knockdown_genes{end+1} = '10157';\n",
    "                        case 38, knockdown_genes{end+1} = '10162';\n",
    "                        case 39, knockdown_genes{end+1} = '10164';\n",
    "                        case 40, knockdown_genes{end+1} = '10165';\n",
    "                        case 41, knockdown_genes{end+1} = '10166';\n",
    "                        case 42, knockdown_genes{end+1} = '1017';\n",
    "                        case 43, knockdown_genes{end+1} = '10170';\n",
    "                        case 44, knockdown_genes{end+1} = '10171';\n",
    "                        case 45, knockdown_genes{end+1} = '1018';\n",
    "                        case 46, knockdown_genes{end+1} = '10188';\n",
    "                        case 47, knockdown_genes{end+1} = '1019';\n",
    "                        case 48, knockdown_genes{end+1} = '10193';\n",
    "                        case 49, knockdown_genes{end+1} = '10195';\n",
    "                        case 50, knockdown_genes{end+1} = '1020';\n",
    "                        case 51, knockdown_genes{end+1} = '10201';\n",
    "                        case 52, knockdown_genes{end+1} = '10202';\n",
    "                        case 53, knockdown_genes{end+1} = '1021';\n",
    "                        case 54, knockdown_genes{end+1} = '10210';\n",
    "                        case 55, knockdown_genes{end+1} = '10217';\n",
    "                        case 56, knockdown_genes{end+1} = '1022';\n",
    "                        case 57, knockdown_genes{end+1} = '10226';\n",
    "                        case 58, knockdown_genes{end+1} = '10229';\n",
    "                        case 59, knockdown_genes{end+1} = '1024';\n",
    "                        case 60, knockdown_genes{end+1} = '10243';\n",
    "                        case 61, knockdown_genes{end+1} = '10246';\n",
    "                        case 62, knockdown_genes{end+1} = '10249';\n",
    "                        case 63, knockdown_genes{end+1} = '1025';\n",
    "                        case 64, knockdown_genes{end+1} = '10257';\n",
    "                        case 65, knockdown_genes{end+1} = '102724197';\n",
    "                        case 66, knockdown_genes{end+1} = '102724560';\n",
    "                        case 67, knockdown_genes{end+1} = '10273';\n",
    "                        case 68, knockdown_genes{end+1} = '10277';\n",
    "                        case 69, knockdown_genes{end+1} = '10283';\n",
    "                        case 70, knockdown_genes{end+1} = '10290';\n",
    "                        case 71, knockdown_genes{end+1} = '10295';\n",
    "                        case 72, knockdown_genes{end+1} = '10298';\n",
    "                        case 73, knockdown_genes{end+1} = '10299';\n",
    "                        case 74, knockdown_genes{end+1} = '10312';\n",
    "                        case 75, knockdown_genes{end+1} = '10317';\n",
    "                        case 76, knockdown_genes{end+1} = '10327';\n",
    "                        case 77, knockdown_genes{end+1} = '1033';\n",
    "                        case 78, knockdown_genes{end+1} = '10331';\n",
    "                        case 79, knockdown_genes{end+1} = '10351';\n",
    "                        case 80, knockdown_genes{end+1} = '10352';\n",
    "                        case 81, knockdown_genes{end+1} = '1036';\n",
    "                        case 82, knockdown_genes{end+1} = '10380';\n",
    "                        case 83, knockdown_genes{end+1} = '10390';\n",
    "                        case 84, knockdown_genes{end+1} = '10396';\n",
    "                        case 85, knockdown_genes{end+1} = '1040';\n",
    "                        case 86, knockdown_genes{end+1} = '10400';\n",
    "                        case 87, knockdown_genes{end+1} = '10402';\n",
    "                        case 88, knockdown_genes{end+1} = '10420';\n",
    "                        case 89, knockdown_genes{end+1} = '10423';\n",
    "                        case 90, knockdown_genes{end+1} = '10434';\n",
    "                        case 91, knockdown_genes{end+1} = '10449';\n",
    "                        case 92, knockdown_genes{end+1} = '10450';\n",
    "                        case 93, knockdown_genes{end+1} = '10455';\n",
    "                        case 94, knockdown_genes{end+1} = '10461';\n",
    "                        case 95, knockdown_genes{end+1} = '10465';\n",
    "                        case 96, knockdown_genes{end+1} = '10476';\n",
    "                        case 97, knockdown_genes{end+1} = '10477';\n",
    "                        case 98, knockdown_genes{end+1} = '10478';\n",
    "                        case 99, knockdown_genes{end+1} = '10494';\n",
    "                        case 100, knockdown_genes{end+1} = '10498';\n",
    "                        case 101, knockdown_genes{end+1} = '10499';\n",
    "                        case 102, knockdown_genes{end+1} = '10514';\n",
    "                        case 103, knockdown_genes{end+1} = '10524';\n",
    "                        case 104, knockdown_genes{end+1} = '10553';\n",
    "                        case 105, knockdown_genes{end+1} = '10554';\n",
    "                        case 106, knockdown_genes{end+1} = '10555';\n",
    "                        case 107, knockdown_genes{end+1} = '10558';\n",
    "                        case 108, knockdown_genes{end+1} = '10559';\n",
    "                        case 109, knockdown_genes{end+1} = '1056';\n",
    "                        case 110, knockdown_genes{end+1} = '10560';\n",
    "                        case 111, knockdown_genes{end+1} = '10568';\n",
    "                        case 112, knockdown_genes{end+1} = '10585';\n",
    "                        case 113, knockdown_genes{end+1} = '10587';\n",
    "                        case 114, knockdown_genes{end+1} = '10588';\n",
    "                        case 115, knockdown_genes{end+1} = '10595';\n",
    "                        case 116, knockdown_genes{end+1} = '10599';\n",
    "                        case 117, knockdown_genes{end+1} = '10600';\n",
    "                        case 118, knockdown_genes{end+1} = '10606';\n",
    "                        case 119, knockdown_genes{end+1} = '10610';\n",
    "                        case 120, knockdown_genes{end+1} = '10617';\n",
    "                        case 121, knockdown_genes{end+1} = '10621';\n",
    "                        case 122, knockdown_genes{end+1} = '10622';\n",
    "                        case 123, knockdown_genes{end+1} = '10623';\n",
    "                        case 124, knockdown_genes{end+1} = '10632';\n",
    "                        case 125, knockdown_genes{end+1} = '10645';\n",
    "                        case 126, knockdown_genes{end+1} = '10654';\n",
    "                        case 127, knockdown_genes{end+1} = '1066';\n",
    "                        case 128, knockdown_genes{end+1} = '10667';\n",
    "                        case 129, knockdown_genes{end+1} = '10678';\n",
    "                        case 130, knockdown_genes{end+1} = '10682';\n",
    "                        case 131, knockdown_genes{end+1} = '10690';\n",
    "                        case 132, knockdown_genes{end+1} = '107';\n",
    "                        case 133, knockdown_genes{end+1} = '10714';\n",
    "                        case 134, knockdown_genes{end+1} = '10715';\n",
    "                        case 135, knockdown_genes{end+1} = '10720';\n",
    "                        case 136, knockdown_genes{end+1} = '10721';\n",
    "                        case 137, knockdown_genes{end+1} = '10723';\n",
    "                        case 138, knockdown_genes{end+1} = '10724';\n",
    "                        case 139, knockdown_genes{end+1} = '10728';\n",
    "                        case 140, knockdown_genes{end+1} = '10733';\n",
    "                        case 141, knockdown_genes{end+1} = '10746';\n",
    "                        case 142, knockdown_genes{end+1} = '1075';\n",
    "                        case 143, knockdown_genes{end+1} = '10768';\n",
    "                        case 144, knockdown_genes{end+1} = '10769';\n",
    "                        case 145, knockdown_genes{end+1} = '10783';\n",
    "                        case 146, knockdown_genes{end+1} = '10786';\n",
    "                        case 147, knockdown_genes{end+1} = '10797';\n",
    "                        case 148, knockdown_genes{end+1} = '108';\n",
    "                        case 149, knockdown_genes{end+1} = '1080';\n",
    "                        case 150, knockdown_genes{end+1} = '10825';\n",
    "                        case 151, knockdown_genes{end+1} = '10826';\n",
    "                        case 152, knockdown_genes{end+1} = '10840';\n",
    "                        case 153, knockdown_genes{end+1} = '10841';\n",
    "                        case 154, knockdown_genes{end+1} = '10846';\n",
    "                        case 155, knockdown_genes{end+1} = '10858';\n",
    "                        case 156, knockdown_genes{end+1} = '10861';\n",
    "                        case 157, knockdown_genes{end+1} = '10864';\n",
    "                        case 158, knockdown_genes{end+1} = '10868';\n",
    "                        case 159, knockdown_genes{end+1} = '10869';\n",
    "                        case 160, knockdown_genes{end+1} = '10870';\n",
    "                        case 161, knockdown_genes{end+1} = '10873';\n",
    "                        case 162, knockdown_genes{end+1} = '109';\n",
    "                        case 163, knockdown_genes{end+1} = '10901';\n",
    "                        case 164, knockdown_genes{end+1} = '10905';\n",
    "                        case 165, knockdown_genes{end+1} = '10908';\n",
    "                        case 166, knockdown_genes{end+1} = '10914';\n",
    "                        case 167, knockdown_genes{end+1} = '10919';\n",
    "                        case 168, knockdown_genes{end+1} = '10922';\n",
    "                        case 169, knockdown_genes{end+1} = '10928';\n",
    "                        case 170, knockdown_genes{end+1} = '10935';\n",
    "                        case 171, knockdown_genes{end+1} = '10941';\n",
    "                        case 172, knockdown_genes{end+1} = '10948';\n",
    "                        case 173, knockdown_genes{end+1} = '10951';\n",
    "                        case 174, knockdown_genes{end+1} = '10954';\n",
    "                        case 175, knockdown_genes{end+1} = '10965';\n",
    "                        case 176, knockdown_genes{end+1} = '10975';\n",
    "                        case 177, knockdown_genes{end+1} = '10991';\n",
    "                        case 178, knockdown_genes{end+1} = '10993';\n",
    "                        case 179, knockdown_genes{end+1} = '10998';\n",
    "                        case 180, knockdown_genes{end+1} = '10999';\n",
    "                        case 181, knockdown_genes{end+1} = '11000';\n",
    "                        case 182, knockdown_genes{end+1} = '11001';\n",
    "                        case 183, knockdown_genes{end+1} = '11010';\n",
    "                        case 184, knockdown_genes{end+1} = '11011';\n",
    "                        case 185, knockdown_genes{end+1} = '11019';\n",
    "                        case 186, knockdown_genes{end+1} = '1103';\n",
    "                        case 187, knockdown_genes{end+1} = '11035';\n",
    "                        case 188, knockdown_genes{end+1} = '11040';\n",
    "                        case 189, knockdown_genes{end+1} = '11041';\n",
    "                        case 190, knockdown_genes{end+1} = '11044';\n",
    "                        case 191, knockdown_genes{end+1} = '11046';\n",
    "                        case 192, knockdown_genes{end+1} = '11059';\n",
    "                        case 193, knockdown_genes{end+1} = '11060';\n",
    "                        case 194, knockdown_genes{end+1} = '11065';\n",
    "                        case 195, knockdown_genes{end+1} = '11072';\n",
    "                        case 196, knockdown_genes{end+1} = '1109';\n",
    "                        case 197, knockdown_genes{end+1} = '11099';\n",
    "                        case 198, knockdown_genes{end+1} = '111';\n",
    "                        case 199, knockdown_genes{end+1} = '11101';\n",
    "                        case 200, knockdown_genes{end+1} = '11102';\n",
    "                        case 201, knockdown_genes{end+1} = '1111';\n",
    "                        case 202, knockdown_genes{end+1} = '11112';\n",
    "                        case 203, knockdown_genes{end+1} = '11113';\n",
    "                        case 204, knockdown_genes{end+1} = '11122';\n",
    "                        case 205, knockdown_genes{end+1} = '11128';\n",
    "                        case 206, knockdown_genes{end+1} = '11136';\n",
    "                        case 207, knockdown_genes{end+1} = '11143';\n",
    "                        case 208, knockdown_genes{end+1} = '11145';\n",
    "                        case 209, knockdown_genes{end+1} = '11156';\n",
    "                        case 210, knockdown_genes{end+1} = '1116';\n",
    "                        case 211, knockdown_genes{end+1} = '11163';\n",
    "                        case 212, knockdown_genes{end+1} = '11164';\n",
    "                        case 213, knockdown_genes{end+1} = '11165';\n",
    "                        case 214, knockdown_genes{end+1} = '1118';\n",
    "                        case 215, knockdown_genes{end+1} = '11181';\n",
    "                        case 216, knockdown_genes{end+1} = '11182';\n",
    "                        case 217, knockdown_genes{end+1} = '11183';\n",
    "                        case 218, knockdown_genes{end+1} = '11184';\n",
    "                        case 219, knockdown_genes{end+1} = '11185';\n",
    "                        case 220, knockdown_genes{end+1} = '1119';\n",
    "                        case 221, knockdown_genes{end+1} = '112';\n",
    "                        case 222, knockdown_genes{end+1} = '1120';\n",
    "                        case 223, knockdown_genes{end+1} = '11200';\n",
    "                        case 224, knockdown_genes{end+1} = '11201';\n",
    "                        case 225, knockdown_genes{end+1} = '11212';\n",
    "                        case 226, knockdown_genes{end+1} = '11213';\n",
    "                        case 227, knockdown_genes{end+1} = '11219';\n",
    "                        case 228, knockdown_genes{end+1} = '11221';\n",
    "                        case 229, knockdown_genes{end+1} = '11226';\n",
    "                        case 230, knockdown_genes{end+1} = '11227';\n",
    "                        case 231, knockdown_genes{end+1} = '11232';\n",
    "                        case 232, knockdown_genes{end+1} = '11236';\n",
    "                        case 233, knockdown_genes{end+1} = '11238';\n",
    "                        case 234, knockdown_genes{end+1} = '11240';\n",
    "                        case 235, knockdown_genes{end+1} = '112476';\n",
    "                        case 236, knockdown_genes{end+1} = '112483';\n",
    "                        case 237, knockdown_genes{end+1} = '11253';\n",
    "                        case 238, knockdown_genes{end+1} = '11254';\n",
    "                        case 239, knockdown_genes{end+1} = '112611';\n",
    "                        case 240, knockdown_genes{end+1} = '11266';\n",
    "                        case 241, knockdown_genes{end+1} = '112724';\n",
    "                        case 242, knockdown_genes{end+1} = '11274';\n",
    "                        case 243, knockdown_genes{end+1} = '11277';\n",
    "                        case 244, knockdown_genes{end+1} = '11282';\n",
    "                        case 245, knockdown_genes{end+1} = '11283';\n",
    "                        case 246, knockdown_genes{end+1} = '11284';\n",
    "                        case 247, knockdown_genes{end+1} = '11285';\n",
    "                        case 248, knockdown_genes{end+1} = '112858';\n",
    "                        case 249, knockdown_genes{end+1} = '113';\n",
    "                        case 250, knockdown_genes{end+1} = '113026';\n",
    "                        case 251, knockdown_genes{end+1} = '11309';\n",
    "                        case 252, knockdown_genes{end+1} = '11313';\n",
    "                        case 253, knockdown_genes{end+1} = '11316';\n",
    "                        case 254, knockdown_genes{end+1} = '113189';\n",
    "                        case 255, knockdown_genes{end+1} = '11320';\n",
    "                        case 256, knockdown_genes{end+1} = '113235';\n",
    "                        case 257, knockdown_genes{end+1} = '113278';\n",
    "                        case 258, knockdown_genes{end+1} = '11328';\n",
    "                        case 259, knockdown_genes{end+1} = '11329';\n",
    "                        case 260, knockdown_genes{end+1} = '11332';\n",
    "                        case 261, knockdown_genes{end+1} = '11342';\n",
    "                        case 262, knockdown_genes{end+1} = '11343';\n",
    "                        case 263, knockdown_genes{end+1} = '113451';\n",
    "                        case 264, knockdown_genes{end+1} = '113612';\n",
    "                        case 265, knockdown_genes{end+1} = '113675';\n",
    "                        case 266, knockdown_genes{end+1} = '114';\n",
    "                        case 267, knockdown_genes{end+1} = '114112';\n",
    "                        case 268, knockdown_genes{end+1} = '114134';\n",
    "                        case 269, knockdown_genes{end+1} = '114571';\n",
    "                        case 270, knockdown_genes{end+1} = '1147';\n",
    "                        case 271, knockdown_genes{end+1} = '114770';\n",
    "                        case 272, knockdown_genes{end+1} = '114771';\n",
    "                        case 273, knockdown_genes{end+1} = '114783';\n",
    "                        case 274, knockdown_genes{end+1} = '114805';\n",
    "                        case 275, knockdown_genes{end+1} = '1149';\n",
    "                        case 276, knockdown_genes{end+1} = '114971';\n",
    "                        case 277, knockdown_genes{end+1} = '115';\n",
    "                        case 278, knockdown_genes{end+1} = '115019';\n",
    "                        case 279, knockdown_genes{end+1} = '115111';\n",
    "                        case 280, knockdown_genes{end+1} = '115123';\n",
    "                        case 281, knockdown_genes{end+1} = '1152';\n",
    "                        case 282, knockdown_genes{end+1} = '115286';\n",
    "                        case 283, knockdown_genes{end+1} = '115426';\n",
    "                        case 284, knockdown_genes{end+1} = '1155';\n",
    "                        case 285, knockdown_genes{end+1} = '115584';\n",
    "                        case 286, knockdown_genes{end+1} = '1158';\n",
    "                        case 287, knockdown_genes{end+1} = '1159';\n",
    "                        case 288, knockdown_genes{end+1} = '1160';\n",
    "                        case 289, knockdown_genes{end+1} = '116085';\n",
    "                        case 290, knockdown_genes{end+1} = '116179';\n",
    "                        case 291, knockdown_genes{end+1} = '116255';\n",
    "                        case 292, knockdown_genes{end+1} = '116285';\n",
    "                        case 293, knockdown_genes{end+1} = '116369';\n",
    "                        case 294, knockdown_genes{end+1} = '116519';\n",
    "                        case 295, knockdown_genes{end+1} = '116969';\n",
    "                        case 296, knockdown_genes{end+1} = '117247';\n",
    "                        case 297, knockdown_genes{end+1} = '117248';\n",
    "                        case 298, knockdown_genes{end+1} = '117283';\n",
    "                        case 299, knockdown_genes{end+1} = '1178';\n",
    "                        case 300, knockdown_genes{end+1} = '118424';\n",
    "                        case 301, knockdown_genes{end+1} = '118881';\n",
    "                        case 302, knockdown_genes{end+1} = '119391';\n",
    "                        case 303, knockdown_genes{end+1} = '1195';\n",
    "                        case 304, knockdown_genes{end+1} = '119548';\n",
    "                        case 305, knockdown_genes{end+1} = '1196';\n",
    "                        case 306, knockdown_genes{end+1} = '1198';\n",
    "                        case 307, knockdown_genes{end+1} = '12';\n",
    "                        case 308, knockdown_genes{end+1} = '1200';\n",
    "                        case 309, knockdown_genes{end+1} = '120103';\n",
    "                        case 310, knockdown_genes{end+1} = '1208';\n",
    "                        case 311, knockdown_genes{end+1} = '120892';\n",
    "                        case 312, knockdown_genes{end+1} = '121214';\n",
    "                        case 313, knockdown_genes{end+1} = '121260';\n",
    "                        case 314, knockdown_genes{end+1} = '121278';\n",
    "                        case 315, knockdown_genes{end+1} = '1215';\n",
    "                        case 316, knockdown_genes{end+1} = '122011';\n",
    "                        case 317, knockdown_genes{end+1} = '122481';\n",
    "                        case 318, knockdown_genes{end+1} = '122618';\n",
    "                        case 319, knockdown_genes{end+1} = '122622';\n",
    "                        case 320, knockdown_genes{end+1} = '122970';\n",
    "                        case 321, knockdown_genes{end+1} = '123';\n",
    "                        case 322, knockdown_genes{end+1} = '123041';\n",
    "                        case 323, knockdown_genes{end+1} = '123096';\n",
    "                        case 324, knockdown_genes{end+1} = '123099';\n",
    "                        case 325, knockdown_genes{end+1} = '123263';\n",
    "                        case 326, knockdown_genes{end+1} = '123264';\n",
    "                        case 327, knockdown_genes{end+1} = '123283';\n",
    "                        case 328, knockdown_genes{end+1} = '123688';\n",
    "                        case 329, knockdown_genes{end+1} = '123745';\n",
    "                        case 330, knockdown_genes{end+1} = '123803';\n",
    "                        case 331, knockdown_genes{end+1} = '123876';\n",
    "                        case 332, knockdown_genes{end+1} = '124';\n",
    "                        case 333, knockdown_genes{end+1} = '1244';\n",
    "                        case 334, knockdown_genes{end+1} = '124454';\n",
    "                        case 335, knockdown_genes{end+1} = '124583';\n",
    "                        case 336, knockdown_genes{end+1} = '124637';\n",
    "                        case 337, knockdown_genes{end+1} = '124739';\n",
    "                        case 338, knockdown_genes{end+1} = '124872';\n",
    "                        case 339, knockdown_genes{end+1} = '124935';\n",
    "                        case 340, knockdown_genes{end+1} = '124975';\n",
    "                        case 341, knockdown_genes{end+1} = '125';\n",
    "                        case 342, knockdown_genes{end+1} = '125061';\n",
    "                        case 343, knockdown_genes{end+1} = '125206';\n",
    "                        case 344, knockdown_genes{end+1} = '125965';\n",
    "                        case 345, knockdown_genes{end+1} = '125981';\n",
    "                        case 346, knockdown_genes{end+1} = '126';\n",
    "                        case 347, knockdown_genes{end+1} = '126129';\n",
    "                        case 348, knockdown_genes{end+1} = '1263';\n",
    "                        case 349, knockdown_genes{end+1} = '126328';\n",
    "                        case 350, knockdown_genes{end+1} = '126410';\n",
    "                        case 351, knockdown_genes{end+1} = '1267';\n",
    "                        case 352, knockdown_genes{end+1} = '126792';\n",
    "                        case 353, knockdown_genes{end+1} = '127';\n",
    "                        case 354, knockdown_genes{end+1} = '127124';\n",
    "                        case 355, knockdown_genes{end+1} = '127544';\n",
    "                        case 356, knockdown_genes{end+1} = '127933';\n",
    "                        case 357, knockdown_genes{end+1} = '128';\n",
    "                        case 358, knockdown_genes{end+1} = '128486';\n",
    "                        case 359, knockdown_genes{end+1} = '128853';\n",
    "                        case 360, knockdown_genes{end+1} = '128869';\n",
    "                        case 361, knockdown_genes{end+1} = '129607';\n",
    "                        case 362, knockdown_genes{end+1} = '129642';\n",
    "                        case 363, knockdown_genes{end+1} = '129807';\n",
    "                        case 364, knockdown_genes{end+1} = '13';\n",
    "                        case 365, knockdown_genes{end+1} = '130';\n",
    "                        case 366, knockdown_genes{end+1} = '130013';\n",
    "                        case 367, knockdown_genes{end+1} = '130367';\n",
    "                        case 368, knockdown_genes{end+1} = '130399';\n",
    "                        case 369, knockdown_genes{end+1} = '130507';\n",
    "                        case 370, knockdown_genes{end+1} = '130589';\n",
    "                        case 371, knockdown_genes{end+1} = '130752';\n",
    "                        case 372, knockdown_genes{end+1} = '131';\n",
    "                        case 373, knockdown_genes{end+1} = '1312';\n",
    "                        case 374, knockdown_genes{end+1} = '1314';\n",
    "                        case 375, knockdown_genes{end+1} = '1315';\n",
    "                        case 376, knockdown_genes{end+1} = '131669';\n",
    "                        case 377, knockdown_genes{end+1} = '1317';\n",
    "                        case 378, knockdown_genes{end+1} = '131890';\n",
    "                        case 379, knockdown_genes{end+1} = '132';\n",
    "                        case 380, knockdown_genes{end+1} = '132158';\n",
    "                        case 381, knockdown_genes{end+1} = '132160';\n",
    "                        case 382, knockdown_genes{end+1} = '1326';\n",
    "                        case 383, knockdown_genes{end+1} = '1327';\n",
    "                        case 384, knockdown_genes{end+1} = '132789';\n",
    "                        case 385, knockdown_genes{end+1} = '1329';\n",
    "                        case 386, knockdown_genes{end+1} = '132949';\n",
    "                        case 387, knockdown_genes{end+1} = '133121';\n",
    "                        case 388, knockdown_genes{end+1} = '133688';\n",
    "                        case 389, knockdown_genes{end+1} = '1337';\n",
    "                        case 390, knockdown_genes{end+1} = '1339';\n",
    "                        case 391, knockdown_genes{end+1} = '1340';\n",
    "                        case 392, knockdown_genes{end+1} = '134111';\n",
    "                        case 393, knockdown_genes{end+1} = '1345';\n",
    "                        case 394, knockdown_genes{end+1} = '134510';\n",
    "                        case 395, knockdown_genes{end+1} = '134526';\n",
    "                        case 396, knockdown_genes{end+1} = '1346';\n",
    "                        case 397, knockdown_genes{end+1} = '1347';\n",
    "                        case 398, knockdown_genes{end+1} = '1349';\n",
    "                        case 399, knockdown_genes{end+1} = '1350';\n",
    "                        case 400, knockdown_genes{end+1} = '1351';\n",
    "                        case 401, knockdown_genes{end+1} = '135152';\n",
    "                        case 402, knockdown_genes{end+1} = '1356';\n",
    "                        case 403, knockdown_genes{end+1} = '1357';\n",
    "                        case 404, knockdown_genes{end+1} = '1358';\n",
    "                        case 405, knockdown_genes{end+1} = '1371';\n",
    "                        case 406, knockdown_genes{end+1} = '1373';\n",
    "                        case 407, knockdown_genes{end+1} = '1374';\n",
    "                        case 408, knockdown_genes{end+1} = '1375';\n",
    "                        case 409, knockdown_genes{end+1} = '1376';\n",
    "                        case 410, knockdown_genes{end+1} = '137872';\n",
    "                        case 411, knockdown_genes{end+1} = '137902';\n",
    "                        case 412, knockdown_genes{end+1} = '137964';\n",
    "                        case 413, knockdown_genes{end+1} = '1384';\n",
    "                        case 414, knockdown_genes{end+1} = '138428';\n",
    "                        case 415, knockdown_genes{end+1} = '138429';\n",
    "                        case 416, knockdown_genes{end+1} = '138639';\n",
    "                        case 417, knockdown_genes{end+1} = '1387';\n",
    "                        case 418, knockdown_genes{end+1} = '139596';\n",
    "                        case 419, knockdown_genes{end+1} = '139728';\n",
    "                        case 420, knockdown_genes{end+1} = '140469';\n",
    "                        case 421, knockdown_genes{end+1} = '140609';\n",
    "                        case 422, knockdown_genes{end+1} = '140679';\n",
    "                        case 423, knockdown_genes{end+1} = '140687';\n",
    "                        case 424, knockdown_genes{end+1} = '140733';\n",
    "                        case 425, knockdown_genes{end+1} = '140739';\n",
    "                        case 426, knockdown_genes{end+1} = '140803';\n",
    "                        case 427, knockdown_genes{end+1} = '140838';\n",
    "                        case 428, knockdown_genes{end+1} = '140901';\n",
    "                        case 429, knockdown_genes{end+1} = '141';\n",
    "                        case 430, knockdown_genes{end+1} = '142';\n",
    "                        case 431, knockdown_genes{end+1} = '142678';\n",
    "                        case 432, knockdown_genes{end+1} = '142679';\n",
    "                        case 433, knockdown_genes{end+1} = '142680';\n",
    "                        case 434, knockdown_genes{end+1} = '143';\n",
    "                        case 435, knockdown_genes{end+1} = '1431';\n",
    "                        case 436, knockdown_genes{end+1} = '143162';\n",
    "                        case 437, knockdown_genes{end+1} = '1432';\n",
    "                        case 438, knockdown_genes{end+1} = '143279';\n",
    "                        case 439, knockdown_genes{end+1} = '1436';\n",
    "                        case 440, knockdown_genes{end+1} = '144193';\n",
    "                        case 441, knockdown_genes{end+1} = '144195';\n",
    "                        case 442, knockdown_genes{end+1} = '1445';\n",
    "                        case 443, knockdown_genes{end+1} = '145173';\n",
    "                        case 444, knockdown_genes{end+1} = '1452';\n",
    "                        case 445, knockdown_genes{end+1} = '145226';\n",
    "                        case 446, knockdown_genes{end+1} = '1453';\n",
    "                        case 447, knockdown_genes{end+1} = '1454';\n",
    "                        case 448, knockdown_genes{end+1} = '1455';\n",
    "                        case 449, knockdown_genes{end+1} = '145553';\n",
    "                        case 450, knockdown_genes{end+1} = '1456';\n",
    "                        case 451, knockdown_genes{end+1} = '1457';\n",
    "                        case 452, knockdown_genes{end+1} = '1459';\n",
    "                        case 453, knockdown_genes{end+1} = '1460';\n",
    "                        case 454, knockdown_genes{end+1} = '146664';\n",
    "                        case 455, knockdown_genes{end+1} = '146712';\n",
    "                        case 456, knockdown_genes{end+1} = '1468';\n",
    "                        case 457, knockdown_genes{end+1} = '146802';\n",
    "                        case 458, knockdown_genes{end+1} = '148581';\n",
    "                        case 459, knockdown_genes{end+1} = '1487';\n",
    "                        case 460, knockdown_genes{end+1} = '148789';\n",
    "                        case 461, knockdown_genes{end+1} = '148867';\n",
    "                        case 462, knockdown_genes{end+1} = '1490';\n",
    "                        case 463, knockdown_genes{end+1} = '1491';\n",
    "                        case 464, knockdown_genes{end+1} = '149603';\n",
    "                        case 465, knockdown_genes{end+1} = '15';\n",
    "                        case 466, knockdown_genes{end+1} = '150290';\n",
    "                        case 467, knockdown_genes{end+1} = '1503';\n",
    "                        case 468, knockdown_genes{end+1} = '150465';\n",
    "                        case 469, knockdown_genes{end+1} = '150763';\n",
    "                        case 470, knockdown_genes{end+1} = '1508';\n",
    "                        case 471, knockdown_genes{end+1} = '1509';\n",
    "                        case 472, knockdown_genes{end+1} = '1510';\n",
    "                        case 473, knockdown_genes{end+1} = '151056';\n",
    "                        case 474, knockdown_genes{end+1} = '1511';\n",
    "                        case 475, knockdown_genes{end+1} = '151112';\n",
    "                        case 476, knockdown_genes{end+1} = '1512';\n",
    "                        case 477, knockdown_genes{end+1} = '1513';\n",
    "                        case 478, knockdown_genes{end+1} = '1514';\n",
    "                        case 479, knockdown_genes{end+1} = '1515';\n",
    "                        case 480, knockdown_genes{end+1} = '151531';\n",
    "                        case 481, knockdown_genes{end+1} = '1519';\n",
    "                        case 482, knockdown_genes{end+1} = '1520';\n",
    "                        case 483, knockdown_genes{end+1} = '1521';\n",
    "                        case 484, knockdown_genes{end+1} = '1522';\n",
    "                        case 485, knockdown_genes{end+1} = '152586';\n",
    "                        case 486, knockdown_genes{end+1} = '152926';\n",
    "                        case 487, knockdown_genes{end+1} = '153201';\n",
    "                        case 488, knockdown_genes{end+1} = '1536';\n",
    "                        case 489, knockdown_genes{end+1} = '1537';\n",
    "                        case 490, knockdown_genes{end+1} = '153769';\n",
    "                        case 491, knockdown_genes{end+1} = '1540';\n",
    "                        case 492, knockdown_genes{end+1} = '154091';\n",
    "                        case 493, knockdown_genes{end+1} = '154141';\n",
    "                        case 494, knockdown_genes{end+1} = '154214';\n",
    "                        case 495, knockdown_genes{end+1} = '1543';\n",
    "                        case 496, knockdown_genes{end+1} = '1544';\n",
    "                        case 497, knockdown_genes{end+1} = '1545';\n",
    "                        case 498, knockdown_genes{end+1} = '1548';\n",
    "                        case 499, knockdown_genes{end+1} = '1549';\n",
    "                        case 500, knockdown_genes{end+1} = '1551';\n",
    "                        case 501, knockdown_genes{end+1} = '155184';\n",
    "                        case 502, knockdown_genes{end+1} = '1553';\n",
    "                        case 503, knockdown_genes{end+1} = '1555';\n",
    "                        case 504, knockdown_genes{end+1} = '1557';\n",
    "                        case 505, knockdown_genes{end+1} = '1558';\n",
    "                        case 506, knockdown_genes{end+1} = '1559';\n",
    "                        case 507, knockdown_genes{end+1} = '156';\n",
    "                        case 508, knockdown_genes{end+1} = '1562';\n",
    "                        case 509, knockdown_genes{end+1} = '1565';\n",
    "                        case 510, knockdown_genes{end+1} = '157';\n",
    "                        case 511, knockdown_genes{end+1} = '1571';\n",
    "                        case 512, knockdown_genes{end+1} = '1572';\n",
    "                        case 513, knockdown_genes{end+1} = '157285';\n",
    "                        case 514, knockdown_genes{end+1} = '1573';\n",
    "                        case 515, knockdown_genes{end+1} = '157506';\n",
    "                        case 516, knockdown_genes{end+1} = '1576';\n",
    "                        case 517, knockdown_genes{end+1} = '1577';\n",
    "                        case 518, knockdown_genes{end+1} = '1579';\n",
    "                        case 519, knockdown_genes{end+1} = '158';\n",
    "                        case 520, knockdown_genes{end+1} = '1580';\n",
    "                        case 521, knockdown_genes{end+1} = '1581';\n",
    "                        case 522, knockdown_genes{end+1} = '158160';\n",
    "                        case 523, knockdown_genes{end+1} = '1582';\n",
    "                        case 524, knockdown_genes{end+1} = '1583';\n",
    "                        case 525, knockdown_genes{end+1} = '1584';\n",
    "                        case 526, knockdown_genes{end+1} = '1585';\n",
    "                        case 527, knockdown_genes{end+1} = '158506';\n",
    "                        case 528, knockdown_genes{end+1} = '158584';\n",
    "                        case 529, knockdown_genes{end+1} = '1586';\n",
    "                        case 530, knockdown_genes{end+1} = '1588';\n",
    "                        case 531, knockdown_genes{end+1} = '158833';\n",
    "                        case 532, knockdown_genes{end+1} = '158835';\n",
    "                        case 533, knockdown_genes{end+1} = '158880';\n",
    "                        case 534, knockdown_genes{end+1} = '1589';\n",
    "                        case 535, knockdown_genes{end+1} = '159';\n",
    "                        case 536, knockdown_genes{end+1} = '1591';\n",
    "                        case 537, knockdown_genes{end+1} = '1592';\n",
    "                        case 538, knockdown_genes{end+1} = '1593';\n",
    "                        case 539, knockdown_genes{end+1} = '1594';\n",
    "                        case 540, knockdown_genes{end+1} = '1595';\n",
    "                        case 541, knockdown_genes{end+1} = '159963';\n",
    "                        case 542, knockdown_genes{end+1} = '16';\n",
    "                        case 543, knockdown_genes{end+1} = '160287';\n",
    "                        case 544, knockdown_genes{end+1} = '1603';\n",
    "                        case 545, knockdown_genes{end+1} = '160428';\n",
    "                        case 546, knockdown_genes{end+1} = '1606';\n",
    "                        case 547, knockdown_genes{end+1} = '1607';\n",
    "                        case 548, knockdown_genes{end+1} = '160728';\n",
    "                        case 549, knockdown_genes{end+1} = '1608';\n",
    "                        case 550, knockdown_genes{end+1} = '160851';\n",
    "                        case 551, knockdown_genes{end+1} = '1609';\n",
    "                        case 552, knockdown_genes{end+1} = '1610';\n",
    "                        case 553, knockdown_genes{end+1} = '1612';\n",
    "                        case 554, knockdown_genes{end+1} = '161247';\n",
    "                        case 555, knockdown_genes{end+1} = '1613';\n",
    "                        case 556, knockdown_genes{end+1} = '1615';\n",
    "                        case 557, knockdown_genes{end+1} = '1621';\n",
    "                        case 558, knockdown_genes{end+1} = '1622';\n",
    "                        case 559, knockdown_genes{end+1} = '162333';\n",
    "                        case 560, knockdown_genes{end+1} = '162417';\n",
    "                        case 561, knockdown_genes{end+1} = '162466';\n",
    "                        case 562, knockdown_genes{end+1} = '1629';\n",
    "                        case 563, knockdown_genes{end+1} = '1632';\n",
    "                        case 564, knockdown_genes{end+1} = '1633';\n",
    "                        case 565, knockdown_genes{end+1} = '163404';\n",
    "                        case 566, knockdown_genes{end+1} = '1635';\n",
    "                        case 567, knockdown_genes{end+1} = '1636';\n",
    "                        case 568, knockdown_genes{end+1} = '1638';\n",
    "                        case 569, knockdown_genes{end+1} = '1644';\n",
    "                        case 570, knockdown_genes{end+1} = '1645';\n",
    "                        case 571, knockdown_genes{end+1} = '1646';\n",
    "                        case 572, knockdown_genes{end+1} = '1650';\n",
    "                        case 573, knockdown_genes{end+1} = '165631';\n",
    "                        case 574, knockdown_genes{end+1} = '166012';\n",
    "                        case 575, knockdown_genes{end+1} = '1666';\n",
    "                        case 576, knockdown_genes{end+1} = '166614';\n",
    "                        case 577, knockdown_genes{end+1} = '166785';\n",
    "                        case 578, knockdown_genes{end+1} = '166929';\n",
    "                        case 579, knockdown_genes{end+1} = '167127';\n",
    "                        case 580, knockdown_genes{end+1} = '167153';\n",
    "                        case 581, knockdown_genes{end+1} = '168391';\n",
    "                        case 582, knockdown_genes{end+1} = '168433';\n",
    "                        case 583, knockdown_genes{end+1} = '169355';\n",
    "                        case 584, knockdown_genes{end+1} = '170384';\n",
    "                        case 585, knockdown_genes{end+1} = '170685';\n",
    "                        case 586, knockdown_genes{end+1} = '170712';\n",
    "                        case 587, knockdown_genes{end+1} = '171425';\n",
    "                        case 588, knockdown_genes{end+1} = '1716';\n",
    "                        case 589, knockdown_genes{end+1} = '1717';\n",
    "                        case 590, knockdown_genes{end+1} = '1718';\n",
    "                        case 591, knockdown_genes{end+1} = '1719';\n",
    "                        case 592, knockdown_genes{end+1} = '1723';\n",
    "                        case 593, knockdown_genes{end+1} = '1727';\n",
    "                        case 594, knockdown_genes{end+1} = '1728';\n",
    "                        case 595, knockdown_genes{end+1} = '1733';\n",
    "                        case 596, knockdown_genes{end+1} = '1734';\n",
    "                        case 597, knockdown_genes{end+1} = '1735';\n",
    "                        case 598, knockdown_genes{end+1} = '1737';\n",
    "                        case 599, knockdown_genes{end+1} = '1738';\n",
    "                        case 600, knockdown_genes{end+1} = '1743';\n",
    "                        case 601, knockdown_genes{end+1} = '175';\n",
    "                        case 602, knockdown_genes{end+1} = '1757';\n",
    "                        case 603, knockdown_genes{end+1} = '1760';\n",
    "                        case 604, knockdown_genes{end+1} = '178';\n",
    "                        case 605, knockdown_genes{end+1} = '1786';\n",
    "                        case 606, knockdown_genes{end+1} = '1787';\n",
    "                        case 607, knockdown_genes{end+1} = '1788';\n",
    "                        case 608, knockdown_genes{end+1} = '1789';\n",
    "                        case 609, knockdown_genes{end+1} = '1798';\n",
    "                        case 610, knockdown_genes{end+1} = '18';\n",
    "                        case 611, knockdown_genes{end+1} = '1800';\n",
    "                        case 612, knockdown_genes{end+1} = '1806';\n",
    "                        case 613, knockdown_genes{end+1} = '1807';\n",
    "                        case 614, knockdown_genes{end+1} = '1808';\n",
    "                        case 615, knockdown_genes{end+1} = '1809';\n",
    "                        case 616, knockdown_genes{end+1} = '1811';\n",
    "                        case 617, knockdown_genes{end+1} = '183';\n",
    "                        case 618, knockdown_genes{end+1} = '1836';\n",
    "                        case 619, knockdown_genes{end+1} = '1841';\n",
    "                        case 620, knockdown_genes{end+1} = '1843';\n",
    "                        case 621, knockdown_genes{end+1} = '1844';\n",
    "                        case 622, knockdown_genes{end+1} = '1845';\n",
    "                        case 623, knockdown_genes{end+1} = '1846';\n",
    "                        case 624, knockdown_genes{end+1} = '1847';\n",
    "                        case 625, knockdown_genes{end+1} = '1848';\n",
    "                        case 626, knockdown_genes{end+1} = '1849';\n",
    "                        case 627, knockdown_genes{end+1} = '1850';\n",
    "                        case 628, knockdown_genes{end+1} = '1852';\n",
    "                        case 629, knockdown_genes{end+1} = '1854';\n",
    "                        case 630, knockdown_genes{end+1} = '1859';\n",
    "                        case 631, knockdown_genes{end+1} = '189';\n",
    "                        case 632, knockdown_genes{end+1} = '1890';\n",
    "                        case 633, knockdown_genes{end+1} = '1891';\n",
    "                        case 634, knockdown_genes{end+1} = '1892';\n",
    "                        case 635, knockdown_genes{end+1} = '19';\n",
    "                        case 636, knockdown_genes{end+1} = '191';\n",
    "                        case 637, knockdown_genes{end+1} = '192111';\n",
    "                        case 638, knockdown_genes{end+1} = '192134';\n",
    "                        case 639, knockdown_genes{end+1} = '192286';\n",
    "                        case 640, knockdown_genes{end+1} = '1956';\n",
    "                        case 641, knockdown_genes{end+1} = '195814';\n",
    "                        case 642, knockdown_genes{end+1} = '196051';\n",
    "                        case 643, knockdown_genes{end+1} = '1962';\n",
    "                        case 644, knockdown_genes{end+1} = '196743';\n",
    "                        case 645, knockdown_genes{end+1} = '196883';\n",
    "                        case 646, knockdown_genes{end+1} = '1969';\n",
    "                        case 647, knockdown_genes{end+1} = '197131';\n",
    "                        case 648, knockdown_genes{end+1} = '197257';\n",
    "                        case 649, knockdown_genes{end+1} = '197258';\n",
    "                        case 650, knockdown_genes{end+1} = '197259';\n",
    "                        case 651, knockdown_genes{end+1} = '197322';\n",
    "                        case 652, knockdown_genes{end+1} = '199857';\n",
    "                        case 653, knockdown_genes{end+1} = '199974';\n",
    "                        case 654, knockdown_genes{end+1} = '200010';\n",
    "                        case 655, knockdown_genes{end+1} = '200576';\n",
    "                        case 656, knockdown_genes{end+1} = '200810';\n",
    "                        case 657, knockdown_genes{end+1} = '200931';\n",
    "                        case 658, knockdown_genes{end+1} = '2011';\n",
    "                        case 659, knockdown_genes{end+1} = '201164';\n",
    "                        case 660, knockdown_genes{end+1} = '201288';\n",
    "                        case 661, knockdown_genes{end+1} = '201562';\n",
    "                        case 662, knockdown_genes{end+1} = '201595';\n",
    "                        case 663, knockdown_genes{end+1} = '2023';\n",
    "                        case 664, knockdown_genes{end+1} = '202374';\n",
    "                        case 665, knockdown_genes{end+1} = '2026';\n",
    "                        case 666, knockdown_genes{end+1} = '2027';\n",
    "                        case 667, knockdown_genes{end+1} = '203';\n",
    "                        case 668, knockdown_genes{end+1} = '2030';\n",
    "                        case 669, knockdown_genes{end+1} = '2033';\n",
    "                        case 670, knockdown_genes{end+1} = '203611';\n",
    "                        case 671, knockdown_genes{end+1} = '204';\n",
    "                        case 672, knockdown_genes{end+1} = '2041';\n",
    "                        case 673, knockdown_genes{end+1} = '2042';\n",
    "                        case 674, knockdown_genes{end+1} = '204219';\n",
    "                        case 675, knockdown_genes{end+1} = '2043';\n",
    "                        case 676, knockdown_genes{end+1} = '2044';\n",
    "                        case 677, knockdown_genes{end+1} = '2045';\n",
    "                        case 678, knockdown_genes{end+1} = '2046';\n",
    "                        case 679, knockdown_genes{end+1} = '2047';\n",
    "                        case 680, knockdown_genes{end+1} = '2048';\n",
    "                        case 681, knockdown_genes{end+1} = '204851';\n",
    "                        case 682, knockdown_genes{end+1} = '2049';\n",
    "                        case 683, knockdown_genes{end+1} = '205';\n",
    "                        case 684, knockdown_genes{end+1} = '2050';\n",
    "                        case 685, knockdown_genes{end+1} = '2051';\n",
    "                        case 686, knockdown_genes{end+1} = '2052';\n",
    "                        case 687, knockdown_genes{end+1} = '2053';\n",
    "                        case 688, knockdown_genes{end+1} = '2058';\n",
    "                        case 689, knockdown_genes{end+1} = '206358';\n",
    "                        case 690, knockdown_genes{end+1} = '2064';\n",
    "                        case 691, knockdown_genes{end+1} = '2065';\n",
    "                        case 692, knockdown_genes{end+1} = '2066';\n",
    "                        case 693, knockdown_genes{end+1} = '207';\n",
    "                        case 694, knockdown_genes{end+1} = '2070';\n",
    "                        case 695, knockdown_genes{end+1} = '208';\n",
    "                        case 696, knockdown_genes{end+1} = '2081';\n",
    "                        case 697, knockdown_genes{end+1} = '2098';\n",
    "                        case 698, knockdown_genes{end+1} = '21';\n",
    "                        case 699, knockdown_genes{end+1} = '210';\n",
    "                        case 700, knockdown_genes{end+1} = '2108';\n",
    "                        case 701, knockdown_genes{end+1} = '2109';\n",
    "                        case 702, knockdown_genes{end+1} = '211';\n",
    "                        case 703, knockdown_genes{end+1} = '2110';\n",
    "                        case 704, knockdown_genes{end+1} = '212';\n",
    "                        case 705, knockdown_genes{end+1} = '213';\n",
    "                        case 706, knockdown_genes{end+1} = '2131';\n",
    "                        case 707, knockdown_genes{end+1} = '2132';\n",
    "                        case 708, knockdown_genes{end+1} = '2134';\n",
    "                        case 709, knockdown_genes{end+1} = '2135';\n",
    "                        case 710, knockdown_genes{end+1} = '2137';\n",
    "                        case 711, knockdown_genes{end+1} = '2138';\n",
    "                        case 712, knockdown_genes{end+1} = '2139';\n",
    "                        case 713, knockdown_genes{end+1} = '2140';\n",
    "                        case 714, knockdown_genes{end+1} = '2145';\n",
    "                        case 715, knockdown_genes{end+1} = '2146';\n",
    "                        case 716, knockdown_genes{end+1} = '2147';\n",
    "                        case 717, knockdown_genes{end+1} = '215';\n",
    "                        case 718, knockdown_genes{end+1} = '216';\n",
    "                        case 719, knockdown_genes{end+1} = '2162';\n",
    "                        case 720, knockdown_genes{end+1} = '2165';\n",
    "                        case 721, knockdown_genes{end+1} = '2166';\n",
    "                        case 722, knockdown_genes{end+1} = '2167';\n",
    "                        case 723, knockdown_genes{end+1} = '2168';\n",
    "                        case 724, knockdown_genes{end+1} = '2169';\n",
    "                        case 725, knockdown_genes{end+1} = '217';\n",
    "                        case 726, knockdown_genes{end+1} = '2170';\n",
    "                        case 727, knockdown_genes{end+1} = '2171';\n",
    "                        case 728, knockdown_genes{end+1} = '2172';\n",
    "                        case 729, knockdown_genes{end+1} = '2173';\n",
    "                        case 730, knockdown_genes{end+1} = '218';\n",
    "                        case 731, knockdown_genes{end+1} = '2180';\n",
    "                        case 732, knockdown_genes{end+1} = '2181';\n",
    "                        case 733, knockdown_genes{end+1} = '2182';\n",
    "                        case 734, knockdown_genes{end+1} = '2184';\n",
    "                        case 735, knockdown_genes{end+1} = '2185';\n",
    "                        case 736, knockdown_genes{end+1} = '219';\n",
    "                        case 737, knockdown_genes{end+1} = '2193';\n",
    "                        case 738, knockdown_genes{end+1} = '219333';\n",
    "                        case 739, knockdown_genes{end+1} = '2194';\n",
    "                        case 740, knockdown_genes{end+1} = '220';\n",
    "                        case 741, knockdown_genes{end+1} = '220074';\n",
    "                        case 742, knockdown_genes{end+1} = '2203';\n",
    "                        case 743, knockdown_genes{end+1} = '220441';\n",
    "                        case 744, knockdown_genes{end+1} = '220972';\n",
    "                        case 745, knockdown_genes{end+1} = '221';\n",
    "                        case 746, knockdown_genes{end+1} = '221223';\n",
    "                        case 747, knockdown_genes{end+1} = '221357';\n",
    "                        case 748, knockdown_genes{end+1} = '221443';\n",
    "                        case 749, knockdown_genes{end+1} = '221687';\n",
    "                        case 750, knockdown_genes{end+1} = '221823';\n",
    "                        case 751, knockdown_genes{end+1} = '221955';\n",
    "                        case 752, knockdown_genes{end+1} = '222';\n",
    "                        case 753, knockdown_genes{end+1} = '2222';\n",
    "                        case 754, knockdown_genes{end+1} = '2224';\n",
    "                        case 755, knockdown_genes{end+1} = '222537';\n",
    "                        case 756, knockdown_genes{end+1} = '222962';\n",
    "                        case 757, knockdown_genes{end+1} = '223';\n",
    "                        case 758, knockdown_genes{end+1} = '223082';\n",
    "                        case 759, knockdown_genes{end+1} = '2232';\n",
    "                        case 760, knockdown_genes{end+1} = '2235';\n",
    "                        case 761, knockdown_genes{end+1} = '224';\n",
    "                        case 762, knockdown_genes{end+1} = '2241';\n",
    "                        case 763, knockdown_genes{end+1} = '2242';\n",
    "                        case 764, knockdown_genes{end+1} = '2243';\n",
    "                        case 765, knockdown_genes{end+1} = '225';\n",
    "                        case 766, knockdown_genes{end+1} = '225689';\n",
    "                        case 767, knockdown_genes{end+1} = '226';\n",
    "                        case 768, knockdown_genes{end+1} = '2260';\n",
    "                        case 769, knockdown_genes{end+1} = '2261';\n",
    "                        case 770, knockdown_genes{end+1} = '2263';\n",
    "                        case 771, knockdown_genes{end+1} = '2264';\n",
    "                        case 772, knockdown_genes{end+1} = '2268';\n",
    "                        case 773, knockdown_genes{end+1} = '2271';\n",
    "                        case 774, knockdown_genes{end+1} = '2272';\n",
    "                        case 775, knockdown_genes{end+1} = '2274';\n",
    "                        case 776, knockdown_genes{end+1} = '2280';\n",
    "                        case 777, knockdown_genes{end+1} = '2281';\n",
    "                        case 778, knockdown_genes{end+1} = '22820';\n",
    "                        case 779, knockdown_genes{end+1} = '22843';\n",
    "                        case 780, knockdown_genes{end+1} = '22848';\n",
    "                        case 781, knockdown_genes{end+1} = '22853';\n",
    "                        case 782, knockdown_genes{end+1} = '22856';\n",
    "                        case 783, knockdown_genes{end+1} = '22858';\n",
    "                        case 784, knockdown_genes{end+1} = '2286';\n",
    "                        case 785, knockdown_genes{end+1} = '2287';\n",
    "                        case 786, knockdown_genes{end+1} = '22876';\n",
    "                        case 787, knockdown_genes{end+1} = '2288';\n",
    "                        case 788, knockdown_genes{end+1} = '2289';\n",
    "                        case 789, knockdown_genes{end+1} = '229';\n",
    "                        case 790, knockdown_genes{end+1} = '22908';\n",
    "                        case 791, knockdown_genes{end+1} = '22928';\n",
    "                        case 792, knockdown_genes{end+1} = '22929';\n",
    "                        case 793, knockdown_genes{end+1} = '22933';\n",
    "                        case 794, knockdown_genes{end+1} = '22934';\n",
    "                        case 795, knockdown_genes{end+1} = '22949';\n",
    "                        case 796, knockdown_genes{end+1} = '22954';\n",
    "                        case 797, knockdown_genes{end+1} = '22977';\n",
    "                        case 798, knockdown_genes{end+1} = '22978';\n",
    "                        case 799, knockdown_genes{end+1} = '22983';\n",
    "                        case 800, knockdown_genes{end+1} = '230';\n",
    "                        case 801, knockdown_genes{end+1} = '23007';\n",
    "                        case 802, knockdown_genes{end+1} = '23012';\n",
    "                        case 803, knockdown_genes{end+1} = '23031';\n",
    "                        case 804, knockdown_genes{end+1} = '23032';\n",
    "                        case 805, knockdown_genes{end+1} = '23035';\n",
    "                        case 806, knockdown_genes{end+1} = '23039';\n",
    "                        case 807, knockdown_genes{end+1} = '23043';\n",
    "                        case 808, knockdown_genes{end+1} = '23049';\n",
    "                        case 809, knockdown_genes{end+1} = '23054';\n",
    "                        case 810, knockdown_genes{end+1} = '23057';\n",
    "                        case 811, knockdown_genes{end+1} = '23067';\n",
    "                        case 812, knockdown_genes{end+1} = '23070';\n",
    "                        case 813, knockdown_genes{end+1} = '23072';\n",
    "                        case 814, knockdown_genes{end+1} = '23077';\n",
    "                        case 815, knockdown_genes{end+1} = '23097';\n",
    "                        case 816, knockdown_genes{end+1} = '231';\n",
    "                        case 817, knockdown_genes{end+1} = '23127';\n",
    "                        case 818, knockdown_genes{end+1} = '23139';\n",
    "                        case 819, knockdown_genes{end+1} = '23169';\n",
    "                        case 820, knockdown_genes{end+1} = '23175';\n",
    "                        case 821, knockdown_genes{end+1} = '23178';\n",
    "                        case 822, knockdown_genes{end+1} = '23179';\n",
    "                        case 823, knockdown_genes{end+1} = '23203';\n",
    "                        case 824, knockdown_genes{end+1} = '23205';\n",
    "                        case 825, knockdown_genes{end+1} = '2321';\n",
    "                        case 826, knockdown_genes{end+1} = '23214';\n",
    "                        case 827, knockdown_genes{end+1} = '2322';\n",
    "                        case 828, knockdown_genes{end+1} = '23228';\n",
    "                        case 829, knockdown_genes{end+1} = '23235';\n",
    "                        case 830, knockdown_genes{end+1} = '23236';\n",
    "                        case 831, knockdown_genes{end+1} = '23239';\n",
    "                        case 832, knockdown_genes{end+1} = '2324';\n",
    "                        case 833, knockdown_genes{end+1} = '23250';\n",
    "                        case 834, knockdown_genes{end+1} = '2326';\n",
    "                        case 835, knockdown_genes{end+1} = '23262';\n",
    "                        case 836, knockdown_genes{end+1} = '2327';\n",
    "                        case 837, knockdown_genes{end+1} = '2328';\n",
    "                        case 838, knockdown_genes{end+1} = '2329';\n",
    "                        case 839, knockdown_genes{end+1} = '23295';\n",
    "                        case 840, knockdown_genes{end+1} = '2330';\n",
    "                        case 841, knockdown_genes{end+1} = '23304';\n",
    "                        case 842, knockdown_genes{end+1} = '23305';\n",
    "                        case 843, knockdown_genes{end+1} = '23309';\n",
    "                        case 844, knockdown_genes{end+1} = '23315';\n",
    "                        case 845, knockdown_genes{end+1} = '23318';\n",
    "                        case 846, knockdown_genes{end+1} = '23326';\n",
    "                        case 847, knockdown_genes{end+1} = '23327';\n",
    "                        case 848, knockdown_genes{end+1} = '23352';\n",
    "                        case 849, knockdown_genes{end+1} = '23358';\n",
    "                        case 850, knockdown_genes{end+1} = '23363';\n",
    "                        case 851, knockdown_genes{end+1} = '23371';\n",
    "                        case 852, knockdown_genes{end+1} = '23382';\n",
    "                        case 853, knockdown_genes{end+1} = '23387';\n",
    "                        case 854, knockdown_genes{end+1} = '2339';\n",
    "                        case 855, knockdown_genes{end+1} = '23395';\n",
    "                        case 856, knockdown_genes{end+1} = '23396';\n",
    "                        case 857, knockdown_genes{end+1} = '23398';\n",
    "                        case 858, knockdown_genes{end+1} = '23408';\n",
    "                        case 859, knockdown_genes{end+1} = '23410';\n",
    "                        case 860, knockdown_genes{end+1} = '23411';\n",
    "                        case 861, knockdown_genes{end+1} = '23417';\n",
    "                        case 862, knockdown_genes{end+1} = '2342';\n",
    "                        case 863, knockdown_genes{end+1} = '23428';\n",
    "                        case 864, knockdown_genes{end+1} = '23430';\n",
    "                        case 865, knockdown_genes{end+1} = '23438';\n",
    "                        case 866, knockdown_genes{end+1} = '23439';\n",
    "                        case 867, knockdown_genes{end+1} = '23443';\n",
    "                        case 868, knockdown_genes{end+1} = '23463';\n",
    "                        case 869, knockdown_genes{end+1} = '23464';\n",
    "                        case 870, knockdown_genes{end+1} = '23475';\n",
    "                        case 871, knockdown_genes{end+1} = '23478';\n",
    "                        case 872, knockdown_genes{end+1} = '2348';\n",
    "                        case 873, knockdown_genes{end+1} = '23483';\n",
    "                        case 874, knockdown_genes{end+1} = '23491';\n",
    "                        case 875, knockdown_genes{end+1} = '23498';\n",
    "                        case 876, knockdown_genes{end+1} = '2350';\n",
    "                        case 877, knockdown_genes{end+1} = '2352';\n",
    "                        case 878, knockdown_genes{end+1} = '23522';\n",
    "                        case 879, knockdown_genes{end+1} = '23529';\n",
    "                        case 880, knockdown_genes{end+1} = '23530';\n",
    "                        case 881, knockdown_genes{end+1} = '23533';\n",
    "                        case 882, knockdown_genes{end+1} = '23539';\n",
    "                        case 883, knockdown_genes{end+1} = '23545';\n",
    "                        case 884, knockdown_genes{end+1} = '23552';\n",
    "                        case 885, knockdown_genes{end+1} = '23556';\n",
    "                        case 886, knockdown_genes{end+1} = '2356';\n",
    "                        case 887, knockdown_genes{end+1} = '23563';\n",
    "                        case 888, knockdown_genes{end+1} = '23564';\n",
    "                        case 889, knockdown_genes{end+1} = '23569';\n",
    "                        case 890, knockdown_genes{end+1} = '23576';\n",
    "                        case 891, knockdown_genes{end+1} = '23600';\n",
    "                        case 892, knockdown_genes{end+1} = '23604';\n",
    "                        case 893, knockdown_genes{end+1} = '23608';\n",
    "                        case 894, knockdown_genes{end+1} = '23609';\n",
    "                        case 895, knockdown_genes{end+1} = '23617';\n",
    "                        case 896, knockdown_genes{end+1} = '23624';\n",
    "                        case 897, knockdown_genes{end+1} = '23632';\n",
    "                        case 898, knockdown_genes{end+1} = '23633';\n",
    "                        case 899, knockdown_genes{end+1} = '23646';\n",
    "                        case 900, knockdown_genes{end+1} = '23649';\n",
    "                        case 901, knockdown_genes{end+1} = '23657';\n",
    "                        case 902, knockdown_genes{end+1} = '23659';\n",
    "                        case 903, knockdown_genes{end+1} = '23678';\n",
    "                        case 904, knockdown_genes{end+1} = '23683';\n",
    "                        case 905, knockdown_genes{end+1} = '23729';\n",
    "                        case 906, knockdown_genes{end+1} = '23743';\n",
    "                        case 907, knockdown_genes{end+1} = '23759';\n",
    "                        case 908, knockdown_genes{end+1} = '23761';\n",
    "                        case 909, knockdown_genes{end+1} = '23770';\n",
    "                        case 910, knockdown_genes{end+1} = '238';\n",
    "                        case 911, knockdown_genes{end+1} = '239';\n",
    "                        case 912, knockdown_genes{end+1} = '24';\n",
    "                        case 913, knockdown_genes{end+1} = '240';\n",
    "                        case 914, knockdown_genes{end+1} = '242';\n",
    "                        case 915, knockdown_genes{end+1} = '2444';\n",
    "                        case 916, knockdown_genes{end+1} = '245972';\n",
    "                        case 917, knockdown_genes{end+1} = '245973';\n",
    "                        case 918, knockdown_genes{end+1} = '246';\n",
    "                        case 919, knockdown_genes{end+1} = '246213';\n",
    "                        case 920, knockdown_genes{end+1} = '247';\n",
    "                        case 921, knockdown_genes{end+1} = '248';\n",
    "                        case 922, knockdown_genes{end+1} = '249';\n",
    "                        case 923, knockdown_genes{end+1} = '25';\n",
    "                        case 924, knockdown_genes{end+1} = '250';\n",
    "                        case 925, knockdown_genes{end+1} = '251';\n",
    "                        case 926, knockdown_genes{end+1} = '2517';\n",
    "                        case 927, knockdown_genes{end+1} = '2519';\n",
    "                        case 928, knockdown_genes{end+1} = '2523';\n",
    "                        case 929, knockdown_genes{end+1} = '2524';\n",
    "                        case 930, knockdown_genes{end+1} = '2525';\n",
    "                        case 931, knockdown_genes{end+1} = '2526';\n",
    "                        case 932, knockdown_genes{end+1} = '2527';\n",
    "                        case 933, knockdown_genes{end+1} = '2528';\n",
    "                        case 934, knockdown_genes{end+1} = '2529';\n",
    "                        case 935, knockdown_genes{end+1} = '2530';\n",
    "                        case 936, knockdown_genes{end+1} = '253017';\n",
    "                        case 937, knockdown_genes{end+1} = '2531';\n",
    "                        case 938, knockdown_genes{end+1} = '253175';\n",
    "                        case 939, knockdown_genes{end+1} = '2534';\n",
    "                        case 940, knockdown_genes{end+1} = '253430';\n",
    "                        case 941, knockdown_genes{end+1} = '253558';\n",
    "                        case 942, knockdown_genes{end+1} = '253782';\n",
    "                        case 943, knockdown_genes{end+1} = '2538';\n",
    "                        case 944, knockdown_genes{end+1} = '2539';\n",
    "                        case 945, knockdown_genes{end+1} = '254173';\n",
    "                        case 946, knockdown_genes{end+1} = '2542';\n",
    "                        case 947, knockdown_genes{end+1} = '254428';\n",
    "                        case 948, knockdown_genes{end+1} = '254531';\n",
    "                        case 949, knockdown_genes{end+1} = '2548';\n",
    "                        case 950, knockdown_genes{end+1} = '255189';\n",
    "                        case 951, knockdown_genes{end+1} = '255488';\n",
    "                        case 952, knockdown_genes{end+1} = '256435';\n",
    "                        case 953, knockdown_genes{end+1} = '257068';\n",
    "                        case 954, knockdown_genes{end+1} = '2571';\n",
    "                        case 955, knockdown_genes{end+1} = '2572';\n",
    "                        case 956, knockdown_genes{end+1} = '257202';\n",
    "                        case 957, knockdown_genes{end+1} = '257218';\n",
    "                        case 958, knockdown_genes{end+1} = '25769';\n",
    "                        case 959, knockdown_genes{end+1} = '25778';\n",
    "                        case 960, knockdown_genes{end+1} = '25796';\n",
    "                        case 961, knockdown_genes{end+1} = '25797';\n",
    "                        case 962, knockdown_genes{end+1} = '2580';\n",
    "                        case 963, knockdown_genes{end+1} = '2581';\n",
    "                        case 964, knockdown_genes{end+1} = '2582';\n",
    "                        case 965, knockdown_genes{end+1} = '25828';\n",
    "                        case 966, knockdown_genes{end+1} = '2583';\n",
    "                        case 967, knockdown_genes{end+1} = '25830';\n",
    "                        case 968, knockdown_genes{end+1} = '25831';\n",
    "                        case 969, knockdown_genes{end+1} = '25834';\n",
    "                        case 970, knockdown_genes{end+1} = '2584';\n",
    "                        case 971, knockdown_genes{end+1} = '2585';\n",
    "                        case 972, knockdown_genes{end+1} = '25862';\n",
    "                        case 973, knockdown_genes{end+1} = '25865';\n",
    "                        case 974, knockdown_genes{end+1} = '25874';\n",
    "                        case 975, knockdown_genes{end+1} = '2588';\n",
    "                        case 976, knockdown_genes{end+1} = '25885';\n",
    "                        case 977, knockdown_genes{end+1} = '2589';\n",
    "                        case 978, knockdown_genes{end+1} = '25897';\n",
    "                        case 979, knockdown_genes{end+1} = '25898';\n",
    "                        case 980, knockdown_genes{end+1} = '2590';\n",
    "                        case 981, knockdown_genes{end+1} = '25902';\n",
    "                        case 982, knockdown_genes{end+1} = '2591';\n",
    "                        case 983, knockdown_genes{end+1} = '2592';\n",
    "                        case 984, knockdown_genes{end+1} = '259230';\n",
    "                        case 985, knockdown_genes{end+1} = '2593';\n",
    "                        case 986, knockdown_genes{end+1} = '25930';\n",
    "                        case 987, knockdown_genes{end+1} = '259307';\n",
    "                        case 988, knockdown_genes{end+1} = '25942';\n",
    "                        case 989, knockdown_genes{end+1} = '2595';\n",
    "                        case 990, knockdown_genes{end+1} = '2597';\n",
    "                        case 991, knockdown_genes{end+1} = '25973';\n",
    "                        case 992, knockdown_genes{end+1} = '25976';\n",
    "                        case 993, knockdown_genes{end+1} = '25989';\n",
    "                        case 994, knockdown_genes{end+1} = '26';\n",
    "                        case 995, knockdown_genes{end+1} = '26001';\n",
    "                        case 996, knockdown_genes{end+1} = '26002';\n",
    "                        case 997, knockdown_genes{end+1} = '26007';\n",
    "                        case 998, knockdown_genes{end+1} = '26027';\n",
    "                        case 999, knockdown_genes{end+1} = '260293';\n",
    "                        case 1000, knockdown_genes{end+1} = '26035';\n",
    "                        case 1001, knockdown_genes{end+1} = '26061';\n",
    "                        case 1002, knockdown_genes{end+1} = '26062';\n",
    "                        case 1003, knockdown_genes{end+1} = '26063';\n",
    "                        case 1004, knockdown_genes{end+1} = '26090';\n",
    "                        case 1005, knockdown_genes{end+1} = '26091';\n",
    "                        case 1006, knockdown_genes{end+1} = '2617';\n",
    "                        case 1007, knockdown_genes{end+1} = '2618';\n",
    "                        case 1008, knockdown_genes{end+1} = '26191';\n",
    "                        case 1009, knockdown_genes{end+1} = '262';\n",
    "                        case 1010, knockdown_genes{end+1} = '26227';\n",
    "                        case 1011, knockdown_genes{end+1} = '26229';\n",
    "                        case 1012, knockdown_genes{end+1} = '26230';\n",
    "                        case 1013, knockdown_genes{end+1} = '26266';\n",
    "                        case 1014, knockdown_genes{end+1} = '26275';\n",
    "                        case 1015, knockdown_genes{end+1} = '26279';\n",
    "                        case 1016, knockdown_genes{end+1} = '2628';\n",
    "                        case 1017, knockdown_genes{end+1} = '26289';\n",
    "                        case 1018, knockdown_genes{end+1} = '2629';\n",
    "                        case 1019, knockdown_genes{end+1} = '26290';\n",
    "                        case 1020, knockdown_genes{end+1} = '26301';\n",
    "                        case 1021, knockdown_genes{end+1} = '2632';\n",
    "                        case 1022, knockdown_genes{end+1} = '26330';\n",
    "                        case 1023, knockdown_genes{end+1} = '2638';\n",
    "                        case 1024, knockdown_genes{end+1} = '2639';\n",
    "                        case 1025, knockdown_genes{end+1} = '2643';\n",
    "                        case 1026, knockdown_genes{end+1} = '2645';\n",
    "                        case 1027, knockdown_genes{end+1} = '26469';\n",
    "                        case 1028, knockdown_genes{end+1} = '2648';\n",
    "                        case 1029, knockdown_genes{end+1} = '2650';\n",
    "                        case 1030, knockdown_genes{end+1} = '26503';\n",
    "                        case 1031, knockdown_genes{end+1} = '2651';\n",
    "                        case 1032, knockdown_genes{end+1} = '26524';\n",
    "                        case 1033, knockdown_genes{end+1} = '2653';\n",
    "                        case 1034, knockdown_genes{end+1} = '26576';\n",
    "                        case 1035, knockdown_genes{end+1} = '26580';\n",
    "                        case 1036, knockdown_genes{end+1} = '266722';\n",
    "                        case 1037, knockdown_genes{end+1} = '266740';\n",
    "                        case 1038, knockdown_genes{end+1} = '267';\n",
    "                        case 1039, knockdown_genes{end+1} = '267020';\n",
    "                        case 1040, knockdown_genes{end+1} = '2673';\n",
    "                        case 1041, knockdown_genes{end+1} = '26750';\n",
    "                        case 1042, knockdown_genes{end+1} = '2678';\n",
    "                        case 1043, knockdown_genes{end+1} = '2679';\n",
    "                        case 1044, knockdown_genes{end+1} = '2683';\n",
    "                        case 1045, knockdown_genes{end+1} = '2686';\n",
    "                        case 1046, knockdown_genes{end+1} = '2687';\n",
    "                        case 1047, knockdown_genes{end+1} = '26873';\n",
    "                        case 1048, knockdown_genes{end+1} = '269';\n",
    "                        case 1049, knockdown_genes{end+1} = '2694';\n",
    "                        case 1050, knockdown_genes{end+1} = '26958';\n",
    "                        case 1051, knockdown_genes{end+1} = '27';\n",
    "                        case 1052, knockdown_genes{end+1} = '270';\n",
    "                        case 1053, knockdown_genes{end+1} = '27005';\n",
    "                        case 1054, knockdown_genes{end+1} = '27010';\n",
    "                        case 1055, knockdown_genes{end+1} = '27034';\n",
    "                        case 1056, knockdown_genes{end+1} = '27035';\n",
    "                        case 1057, knockdown_genes{end+1} = '27063';\n",
    "                        case 1058, knockdown_genes{end+1} = '27068';\n",
    "                        case 1059, knockdown_genes{end+1} = '27087';\n",
    "                        case 1060, knockdown_genes{end+1} = '27089';\n",
    "                        case 1061, knockdown_genes{end+1} = '27090';\n",
    "                        case 1062, knockdown_genes{end+1} = '271';\n",
    "                        case 1063, knockdown_genes{end+1} = '2710';\n",
    "                        case 1064, knockdown_genes{end+1} = '27102';\n",
    "                        case 1065, knockdown_genes{end+1} = '27109';\n",
    "                        case 1066, knockdown_genes{end+1} = '27115';\n",
    "                        case 1067, knockdown_genes{end+1} = '2712';\n",
    "                        case 1068, knockdown_genes{end+1} = '27124';\n",
    "                        case 1069, knockdown_genes{end+1} = '2713';\n",
    "                        case 1070, knockdown_genes{end+1} = '27148';\n",
    "                        case 1071, knockdown_genes{end+1} = '27159';\n",
    "                        case 1072, knockdown_genes{end+1} = '27163';\n",
    "                        case 1073, knockdown_genes{end+1} = '27165';\n",
    "                        case 1074, knockdown_genes{end+1} = '2717';\n",
    "                        case 1075, knockdown_genes{end+1} = '272';\n",
    "                        case 1076, knockdown_genes{end+1} = '2720';\n",
    "                        case 1077, knockdown_genes{end+1} = '27232';\n",
    "                        case 1078, knockdown_genes{end+1} = '27233';\n",
    "                        case 1079, knockdown_genes{end+1} = '27235';\n",
    "                        case 1080, knockdown_genes{end+1} = '27242';\n",
    "                        case 1081, knockdown_genes{end+1} = '27246';\n",
    "                        case 1082, knockdown_genes{end+1} = '27284';\n",
    "                        case 1083, knockdown_genes{end+1} = '2729';\n",
    "                        case 1084, knockdown_genes{end+1} = '27294';\n",
    "                        case 1085, knockdown_genes{end+1} = '2730';\n",
    "                        case 1086, knockdown_genes{end+1} = '27306';\n",
    "                        case 1087, knockdown_genes{end+1} = '2731';\n",
    "                        case 1088, knockdown_genes{end+1} = '27330';\n",
    "                        case 1089, knockdown_genes{end+1} = '27338';\n",
    "                        case 1090, knockdown_genes{end+1} = '27339';\n",
    "                        case 1091, knockdown_genes{end+1} = '27343';\n",
    "                        case 1092, knockdown_genes{end+1} = '27347';\n",
    "                        case 1093, knockdown_genes{end+1} = '27349';\n",
    "                        case 1094, knockdown_genes{end+1} = '2739';\n",
    "                        case 1095, knockdown_genes{end+1} = '27430';\n",
    "                        case 1096, knockdown_genes{end+1} = '27434';\n",
    "                        case 1097, knockdown_genes{end+1} = '2744';\n",
    "                        case 1098, knockdown_genes{end+1} = '2745';\n",
    "                        case 1099, knockdown_genes{end+1} = '2746';\n",
    "                        case 1100, knockdown_genes{end+1} = '2747';\n",
    "                        case 1101, knockdown_genes{end+1} = '275';\n",
    "                        case 1102, knockdown_genes{end+1} = '2752';\n",
    "                        case 1103, knockdown_genes{end+1} = '276';\n",
    "                        case 1104, knockdown_genes{end+1} = '2760';\n",
    "                        case 1105, knockdown_genes{end+1} = '2762';\n",
    "                        case 1106, knockdown_genes{end+1} = '2766';\n",
    "                        case 1107, knockdown_genes{end+1} = '277';\n",
    "                        case 1108, knockdown_genes{end+1} = '278';\n",
    "                        case 1109, knockdown_genes{end+1} = '279';\n",
    "                        case 1110, knockdown_genes{end+1} = '2799';\n",
    "                        case 1111, knockdown_genes{end+1} = '28';\n",
    "                        case 1112, knockdown_genes{end+1} = '280';\n",
    "                        case 1113, knockdown_genes{end+1} = '2805';\n",
    "                        case 1114, knockdown_genes{end+1} = '2806';\n",
    "                        case 1115, knockdown_genes{end+1} = '2819';\n",
    "                        case 1116, knockdown_genes{end+1} = '2820';\n",
    "                        case 1117, knockdown_genes{end+1} = '2821';\n",
    "                        case 1118, knockdown_genes{end+1} = '2822';\n",
    "                        case 1119, knockdown_genes{end+1} = '28227';\n",
    "                        case 1120, knockdown_genes{end+1} = '28231';\n",
    "                        case 1121, knockdown_genes{end+1} = '28232';\n",
    "                        case 1122, knockdown_genes{end+1} = '28234';\n",
    "                        case 1123, knockdown_genes{end+1} = '282974';\n",
    "                        case 1124, knockdown_genes{end+1} = '283208';\n",
    "                        case 1125, knockdown_genes{end+1} = '283209';\n",
    "                        case 1126, knockdown_genes{end+1} = '283358';\n",
    "                        case 1127, knockdown_genes{end+1} = '283629';\n",
    "                        case 1128, knockdown_genes{end+1} = '283748';\n",
    "                        case 1129, knockdown_genes{end+1} = '283871';\n",
    "                        case 1130, knockdown_genes{end+1} = '283985';\n",
    "                        case 1131, knockdown_genes{end+1} = '284004';\n",
    "                        case 1132, knockdown_genes{end+1} = '284086';\n",
    "                        case 1133, knockdown_genes{end+1} = '284098';\n",
    "                        case 1134, knockdown_genes{end+1} = '284111';\n",
    "                        case 1135, knockdown_genes{end+1} = '284129';\n",
    "                        case 1136, knockdown_genes{end+1} = '284273';\n",
    "                        case 1137, knockdown_genes{end+1} = '284656';\n",
    "                        case 1138, knockdown_genes{end+1} = '284996';\n",
    "                        case 1139, knockdown_genes{end+1} = '285193';\n",
    "                        case 1140, knockdown_genes{end+1} = '285220';\n",
    "                        case 1141, knockdown_genes{end+1} = '285440';\n",
    "                        case 1142, knockdown_genes{end+1} = '285671';\n",
    "                        case 1143, knockdown_genes{end+1} = '285755';\n",
    "                        case 1144, knockdown_genes{end+1} = '286016';\n",
    "                        case 1145, knockdown_genes{end+1} = '286297';\n",
    "                        case 1146, knockdown_genes{end+1} = '286410';\n",
    "                        case 1147, knockdown_genes{end+1} = '2872';\n",
    "                        case 1148, knockdown_genes{end+1} = '2875';\n",
    "                        case 1149, knockdown_genes{end+1} = '2876';\n",
    "                        case 1150, knockdown_genes{end+1} = '2877';\n",
    "                        case 1151, knockdown_genes{end+1} = '2878';\n",
    "                        case 1152, knockdown_genes{end+1} = '2879';\n",
    "                        case 1153, knockdown_genes{end+1} = '2880';\n",
    "                        case 1154, knockdown_genes{end+1} = '2882';\n",
    "                        case 1155, knockdown_genes{end+1} = '28965';\n",
    "                        case 1156, knockdown_genes{end+1} = '28972';\n",
    "                        case 1157, knockdown_genes{end+1} = '28976';\n",
    "                        case 1158, knockdown_genes{end+1} = '28982';\n",
    "                        case 1159, knockdown_genes{end+1} = '28992';\n",
    "                        case 1160, knockdown_genes{end+1} = '28996';\n",
    "                        case 1161, knockdown_genes{end+1} = '290';\n",
    "                        case 1162, knockdown_genes{end+1} = '29072';\n",
    "                        case 1163, knockdown_genes{end+1} = '29082';\n",
    "                        case 1164, knockdown_genes{end+1} = '29085';\n",
    "                        case 1165, knockdown_genes{end+1} = '29089';\n",
    "                        case 1166, knockdown_genes{end+1} = '291';\n",
    "                        case 1167, knockdown_genes{end+1} = '29101';\n",
    "                        case 1168, knockdown_genes{end+1} = '29110';\n",
    "                        case 1169, knockdown_genes{end+1} = '29116';\n",
    "                        case 1170, knockdown_genes{end+1} = '29124';\n",
    "                        case 1171, knockdown_genes{end+1} = '29128';\n",
    "                        case 1172, knockdown_genes{end+1} = '292';\n",
    "                        case 1173, knockdown_genes{end+1} = '2923';\n",
    "                        case 1174, knockdown_genes{end+1} = '293';\n",
    "                        case 1175, knockdown_genes{end+1} = '2936';\n",
    "                        case 1176, knockdown_genes{end+1} = '2937';\n",
    "                        case 1177, knockdown_genes{end+1} = '2938';\n",
    "                        case 1178, knockdown_genes{end+1} = '2939';\n",
    "                        case 1179, knockdown_genes{end+1} = '2940';\n",
    "                        case 1180, knockdown_genes{end+1} = '2941';\n",
    "                        case 1181, knockdown_genes{end+1} = '2944';\n",
    "                        case 1182, knockdown_genes{end+1} = '2946';\n",
    "                        case 1183, knockdown_genes{end+1} = '2947';\n",
    "                        case 1184, knockdown_genes{end+1} = '2948';\n",
    "                        case 1185, knockdown_genes{end+1} = '2949';\n",
    "                        case 1186, knockdown_genes{end+1} = '2950';\n",
    "                        case 1187, knockdown_genes{end+1} = '2952';\n",
    "                        case 1188, knockdown_genes{end+1} = '2953';\n",
    "                        case 1189, knockdown_genes{end+1} = '2954';\n",
    "                        case 1190, knockdown_genes{end+1} = '2962';\n",
    "                        case 1191, knockdown_genes{end+1} = '2974';\n",
    "                        case 1192, knockdown_genes{end+1} = '29761';\n",
    "                        case 1193, knockdown_genes{end+1} = '2977';\n",
    "                        case 1194, knockdown_genes{end+1} = '2978';\n",
    "                        case 1195, knockdown_genes{end+1} = '29785';\n",
    "                        case 1196, knockdown_genes{end+1} = '29796';\n",
    "                        case 1197, knockdown_genes{end+1} = '2982';\n",
    "                        case 1198, knockdown_genes{end+1} = '2983';\n",
    "                        case 1199, knockdown_genes{end+1} = '2984';\n",
    "                        case 1200, knockdown_genes{end+1} = '29841';\n",
    "                        case 1201, knockdown_genes{end+1} = '2986';\n",
    "                        case 1202, knockdown_genes{end+1} = '2987';\n",
    "                        case 1203, knockdown_genes{end+1} = '29880';\n",
    "                        case 1204, knockdown_genes{end+1} = '29881';\n",
    "                        case 1205, knockdown_genes{end+1} = '2990';\n",
    "                        case 1206, knockdown_genes{end+1} = '29906';\n",
    "                        case 1207, knockdown_genes{end+1} = '2992';\n",
    "                        case 1208, knockdown_genes{end+1} = '29920';\n",
    "                        case 1209, knockdown_genes{end+1} = '29922';\n",
    "                        case 1210, knockdown_genes{end+1} = '29925';\n",
    "                        case 1211, knockdown_genes{end+1} = '29926';\n",
    "                        case 1212, knockdown_genes{end+1} = '29929';\n",
    "                        case 1213, knockdown_genes{end+1} = '29941';\n",
    "                        case 1214, knockdown_genes{end+1} = '29943';\n",
    "                        case 1215, knockdown_genes{end+1} = '29947';\n",
    "                        case 1216, knockdown_genes{end+1} = '29953';\n",
    "                        case 1217, knockdown_genes{end+1} = '29954';\n",
    "                        case 1218, knockdown_genes{end+1} = '29956';\n",
    "                        case 1219, knockdown_genes{end+1} = '29958';\n",
    "                        case 1220, knockdown_genes{end+1} = '29968';\n",
    "                        case 1221, knockdown_genes{end+1} = '2997';\n",
    "                        case 1222, knockdown_genes{end+1} = '2998';\n",
    "                        case 1223, knockdown_genes{end+1} = '29988';\n",
    "                        case 1224, knockdown_genes{end+1} = '30';\n",
    "                        case 1225, knockdown_genes{end+1} = '3000';\n",
    "                        case 1226, knockdown_genes{end+1} = '30061';\n",
    "                        case 1227, knockdown_genes{end+1} = '3028';\n",
    "                        case 1228, knockdown_genes{end+1} = '3029';\n",
    "                        case 1229, knockdown_genes{end+1} = '3030';\n",
    "                        case 1230, knockdown_genes{end+1} = '3032';\n",
    "                        case 1231, knockdown_genes{end+1} = '3033';\n",
    "                        case 1232, knockdown_genes{end+1} = '3034';\n",
    "                        case 1233, knockdown_genes{end+1} = '3035';\n",
    "                        case 1234, knockdown_genes{end+1} = '3036';\n",
    "                        case 1235, knockdown_genes{end+1} = '3037';\n",
    "                        case 1236, knockdown_genes{end+1} = '3038';\n",
    "                        case 1237, knockdown_genes{end+1} = '3052';\n",
    "                        case 1238, knockdown_genes{end+1} = '3055';\n",
    "                        case 1239, knockdown_genes{end+1} = '306';\n",
    "                        case 1240, knockdown_genes{end+1} = '3067';\n",
    "                        case 1241, knockdown_genes{end+1} = '3069';\n",
    "                        case 1242, knockdown_genes{end+1} = '3073';\n",
    "                        case 1243, knockdown_genes{end+1} = '3074';\n",
    "                        case 1244, knockdown_genes{end+1} = '3081';\n",
    "                        case 1245, knockdown_genes{end+1} = '30811';\n",
    "                        case 1246, knockdown_genes{end+1} = '30814';\n",
    "                        case 1247, knockdown_genes{end+1} = '30815';\n",
    "                        case 1248, knockdown_genes{end+1} = '30833';\n",
    "                        case 1249, knockdown_genes{end+1} = '30834';\n",
    "                        case 1250, knockdown_genes{end+1} = '30849';\n",
    "                        case 1251, knockdown_genes{end+1} = '3093';\n",
    "                        case 1252, knockdown_genes{end+1} = '3098';\n",
    "                        case 1253, knockdown_genes{end+1} = '3099';\n",
    "                        case 1254, knockdown_genes{end+1} = '31';\n",
    "                        case 1255, knockdown_genes{end+1} = '3101';\n",
    "                        case 1256, knockdown_genes{end+1} = '313';\n",
    "                        case 1257, knockdown_genes{end+1} = '314';\n",
    "                        case 1258, knockdown_genes{end+1} = '3141';\n",
    "                        case 1259, knockdown_genes{end+1} = '3145';\n",
    "                        case 1260, knockdown_genes{end+1} = '3155';\n",
    "                        case 1261, knockdown_genes{end+1} = '3156';\n",
    "                        case 1262, knockdown_genes{end+1} = '3157';\n",
    "                        case 1263, knockdown_genes{end+1} = '3158';\n",
    "                        case 1264, knockdown_genes{end+1} = '316';\n",
    "                        case 1265, knockdown_genes{end+1} = '3162';\n",
    "                        case 1266, knockdown_genes{end+1} = '3163';\n",
    "                        case 1267, knockdown_genes{end+1} = '3176';\n",
    "                        case 1268, knockdown_genes{end+1} = '3177';\n",
    "                        case 1269, knockdown_genes{end+1} = '318';\n",
    "                        case 1270, knockdown_genes{end+1} = '32';\n",
    "                        case 1271, knockdown_genes{end+1} = '3242';\n",
    "                        case 1272, knockdown_genes{end+1} = '3248';\n",
    "                        case 1273, knockdown_genes{end+1} = '3251';\n",
    "                        case 1274, knockdown_genes{end+1} = '3283';\n",
    "                        case 1275, knockdown_genes{end+1} = '3284';\n",
    "                        case 1276, knockdown_genes{end+1} = '3290';\n",
    "                        case 1277, knockdown_genes{end+1} = '3291';\n",
    "                        case 1278, knockdown_genes{end+1} = '3292';\n",
    "                        case 1279, knockdown_genes{end+1} = '3293';\n",
    "                        case 1280, knockdown_genes{end+1} = '3294';\n",
    "                        case 1281, knockdown_genes{end+1} = '3295';\n",
    "                        case 1282, knockdown_genes{end+1} = '33';\n",
    "                        case 1283, knockdown_genes{end+1} = '3339';\n",
    "                        case 1284, knockdown_genes{end+1} = '3340';\n",
    "                        case 1285, knockdown_genes{end+1} = '335';\n",
    "                        case 1286, knockdown_genes{end+1} = '336';\n",
    "                        case 1287, knockdown_genes{end+1} = '3376';\n",
    "                        case 1288, knockdown_genes{end+1} = '337876';\n",
    "                        case 1289, knockdown_genes{end+1} = '338';\n",
    "                        case 1290, knockdown_genes{end+1} = '338328';\n",
    "                        case 1291, knockdown_genes{end+1} = '338599';\n",
    "                        case 1292, knockdown_genes{end+1} = '338707';\n",
    "                        case 1293, knockdown_genes{end+1} = '339221';\n",
    "                        case 1294, knockdown_genes{end+1} = '3396';\n",
    "                        case 1295, knockdown_genes{end+1} = '34';\n",
    "                        case 1296, knockdown_genes{end+1} = '340024';\n",
    "                        case 1297, knockdown_genes{end+1} = '340156';\n",
    "                        case 1298, knockdown_genes{end+1} = '340485';\n",
    "                        case 1299, knockdown_genes{end+1} = '340811';\n",
    "                        case 1300, knockdown_genes{end+1} = '341';\n",
    "                        case 1301, knockdown_genes{end+1} = '341392';\n",
    "                        case 1302, knockdown_genes{end+1} = '341676';\n",
    "                        case 1303, knockdown_genes{end+1} = '3417';\n",
    "                        case 1304, knockdown_genes{end+1} = '3418';\n",
    "                        case 1305, knockdown_genes{end+1} = '3419';\n",
    "                        case 1306, knockdown_genes{end+1} = '341947';\n",
    "                        case 1307, knockdown_genes{end+1} = '3420';\n",
    "                        case 1308, knockdown_genes{end+1} = '3421';\n",
    "                        case 1309, knockdown_genes{end+1} = '3422';\n",
    "                        case 1310, knockdown_genes{end+1} = '3423';\n",
    "                        case 1311, knockdown_genes{end+1} = '3425';\n",
    "                        case 1312, knockdown_genes{end+1} = '343';\n",
    "                        case 1313, knockdown_genes{end+1} = '343641';\n",
    "                        case 1314, knockdown_genes{end+1} = '344';\n",
    "                        case 1315, knockdown_genes{end+1} = '344387';\n",
    "                        case 1316, knockdown_genes{end+1} = '344558';\n",
    "                        case 1317, knockdown_genes{end+1} = '344967';\n",
    "                        case 1318, knockdown_genes{end+1} = '345';\n",
    "                        case 1319, knockdown_genes{end+1} = '345274';\n",
    "                        case 1320, knockdown_genes{end+1} = '346606';\n",
    "                        case 1321, knockdown_genes{end+1} = '347734';\n",
    "                        case 1322, knockdown_genes{end+1} = '348';\n",
    "                        case 1323, knockdown_genes{end+1} = '3480';\n",
    "                        case 1324, knockdown_genes{end+1} = '348158';\n",
    "                        case 1325, knockdown_genes{end+1} = '348932';\n",
    "                        case 1326, knockdown_genes{end+1} = '349565';\n",
    "                        case 1327, knockdown_genes{end+1} = '35';\n",
    "                        case 1328, knockdown_genes{end+1} = '353';\n",
    "                        case 1329, knockdown_genes{end+1} = '353238';\n",
    "                        case 1330, knockdown_genes{end+1} = '353497';\n",
    "                        case 1331, knockdown_genes{end+1} = '355';\n",
    "                        case 1332, knockdown_genes{end+1} = '3551';\n",
    "                        case 1333, knockdown_genes{end+1} = '358';\n",
    "                        case 1334, knockdown_genes{end+1} = '359';\n",
    "                        case 1335, knockdown_genes{end+1} = '36';\n",
    "                        case 1336, knockdown_genes{end+1} = '360';\n",
    "                        case 1337, knockdown_genes{end+1} = '360132';\n",
    "                        case 1338, knockdown_genes{end+1} = '360203';\n",
    "                        case 1339, knockdown_genes{end+1} = '361';\n",
    "                        case 1340, knockdown_genes{end+1} = '3611';\n",
    "                        case 1341, knockdown_genes{end+1} = '3612';\n",
    "                        case 1342, knockdown_genes{end+1} = '3613';\n",
    "                        case 1343, knockdown_genes{end+1} = '3614';\n",
    "                        case 1344, knockdown_genes{end+1} = '3615';\n",
    "                        case 1345, knockdown_genes{end+1} = '362';\n",
    "                        case 1346, knockdown_genes{end+1} = '3620';\n",
    "                        case 1347, knockdown_genes{end+1} = '3628';\n",
    "                        case 1348, knockdown_genes{end+1} = '3631';\n",
    "                        case 1349, knockdown_genes{end+1} = '3632';\n",
    "                        case 1350, knockdown_genes{end+1} = '3633';\n",
    "                        case 1351, knockdown_genes{end+1} = '3635';\n",
    "                        case 1352, knockdown_genes{end+1} = '3636';\n",
    "                        case 1353, knockdown_genes{end+1} = '364';\n",
    "                        case 1354, knockdown_genes{end+1} = '3640';\n",
    "                        case 1355, knockdown_genes{end+1} = '3643';\n",
    "                        case 1356, knockdown_genes{end+1} = '3645';\n",
    "                        case 1357, knockdown_genes{end+1} = '3654';\n",
    "                        case 1358, knockdown_genes{end+1} = '3656';\n",
    "                        case 1359, knockdown_genes{end+1} = '3658';\n",
    "                        case 1360, knockdown_genes{end+1} = '366';\n",
    "                        case 1361, knockdown_genes{end+1} = '369';\n",
    "                        case 1362, knockdown_genes{end+1} = '37';\n",
    "                        case 1363, knockdown_genes{end+1} = '3702';\n",
    "                        case 1364, knockdown_genes{end+1} = '3703';\n",
    "                        case 1365, knockdown_genes{end+1} = '3704';\n",
    "                        case 1366, knockdown_genes{end+1} = '3705';\n",
    "                        case 1367, knockdown_genes{end+1} = '3706';\n",
    "                        case 1368, knockdown_genes{end+1} = '3707';\n",
    "                        case 1369, knockdown_genes{end+1} = '3712';\n",
    "                        case 1370, knockdown_genes{end+1} = '3716';\n",
    "                        case 1371, knockdown_genes{end+1} = '3717';\n",
    "                        case 1372, knockdown_genes{end+1} = '3718';\n",
    "                        case 1373, knockdown_genes{end+1} = '372';\n",
    "                        case 1374, knockdown_genes{end+1} = '373156';\n",
    "                        case 1375, knockdown_genes{end+1} = '3735';\n",
    "                        case 1376, knockdown_genes{end+1} = '373856';\n",
    "                        case 1377, knockdown_genes{end+1} = '374291';\n",
    "                        case 1378, knockdown_genes{end+1} = '374462';\n",
    "                        case 1379, knockdown_genes{end+1} = '374569';\n",
    "                        case 1380, knockdown_genes{end+1} = '374875';\n",
    "                        case 1381, knockdown_genes{end+1} = '374907';\n",
    "                        case 1382, knockdown_genes{end+1} = '375056';\n",
    "                        case 1383, knockdown_genes{end+1} = '375449';\n",
    "                        case 1384, knockdown_genes{end+1} = '375775';\n",
    "                        case 1385, knockdown_genes{end+1} = '3764';\n",
    "                        case 1386, knockdown_genes{end+1} = '376497';\n",
    "                        case 1387, knockdown_genes{end+1} = '3767';\n",
    "                        case 1388, knockdown_genes{end+1} = '377630';\n",
    "                        case 1389, knockdown_genes{end+1} = '377677';\n",
    "                        case 1390, knockdown_genes{end+1} = '377841';\n",
    "                        case 1391, knockdown_genes{end+1} = '378884';\n",
    "                        case 1392, knockdown_genes{end+1} = '3791';\n",
    "                        case 1393, knockdown_genes{end+1} = '3795';\n",
    "                        case 1394, knockdown_genes{end+1} = '38';\n",
    "                        case 1395, knockdown_genes{end+1} = '3815';\n",
    "                        case 1396, knockdown_genes{end+1} = '383';\n",
    "                        case 1397, knockdown_genes{end+1} = '384';\n",
    "                        case 1398, knockdown_genes{end+1} = '386757';\n",
    "                        case 1399, knockdown_genes{end+1} = '387775';\n",
    "                        case 1400, knockdown_genes{end+1} = '387787';\n",
    "                        case 1401, knockdown_genes{end+1} = '387893';\n",
    "                        case 1402, knockdown_genes{end+1} = '388228';\n",
    "                        case 1403, knockdown_genes{end+1} = '388714';\n",
    "                        case 1404, knockdown_genes{end+1} = '389015';\n",
    "                        case 1405, knockdown_genes{end+1} = '3897';\n",
    "                        case 1406, knockdown_genes{end+1} = '389840';\n",
    "                        case 1407, knockdown_genes{end+1} = '389898';\n",
    "                        case 1408, knockdown_genes{end+1} = '39';\n",
    "                        case 1409, knockdown_genes{end+1} = '3906';\n",
    "                        case 1410, knockdown_genes{end+1} = '390928';\n",
    "                        case 1411, knockdown_genes{end+1} = '391013';\n",
    "                        case 1412, knockdown_genes{end+1} = '391627';\n",
    "                        case 1413, knockdown_genes{end+1} = '3930';\n",
    "                        case 1414, knockdown_genes{end+1} = '3931';\n",
    "                        case 1415, knockdown_genes{end+1} = '3932';\n",
    "                        case 1416, knockdown_genes{end+1} = '3938';\n",
    "                        case 1417, knockdown_genes{end+1} = '3939';\n",
    "                        case 1418, knockdown_genes{end+1} = '3945';\n",
    "                        case 1419, knockdown_genes{end+1} = '3948';\n",
    "                        case 1420, knockdown_genes{end+1} = '3949';\n",
    "                        case 1421, knockdown_genes{end+1} = '3984';\n",
    "                        case 1422, knockdown_genes{end+1} = '3985';\n",
    "                        case 1423, knockdown_genes{end+1} = '3988';\n",
    "                        case 1424, knockdown_genes{end+1} = '3990';\n",
    "                        case 1425, knockdown_genes{end+1} = '3991';\n",
    "                        case 1426, knockdown_genes{end+1} = '3992';\n",
    "                        case 1427, knockdown_genes{end+1} = '3995';\n",
    "                        case 1428, knockdown_genes{end+1} = '401494';\n",
    "                        case 1429, knockdown_genes{end+1} = '4015';\n",
    "                        case 1430, knockdown_genes{end+1} = '4017';\n",
    "                        case 1431, knockdown_genes{end+1} = '4023';\n",
    "                        case 1432, knockdown_genes{end+1} = '4025';\n",
    "                        case 1433, knockdown_genes{end+1} = '403313';\n",
    "                        case 1434, knockdown_genes{end+1} = '4035';\n",
    "                        case 1435, knockdown_genes{end+1} = '4036';\n",
    "                        case 1436, knockdown_genes{end+1} = '4047';\n",
    "                        case 1437, knockdown_genes{end+1} = '4048';\n",
    "                        case 1438, knockdown_genes{end+1} = '4051';\n",
    "                        case 1439, knockdown_genes{end+1} = '4056';\n",
    "                        case 1440, knockdown_genes{end+1} = '4058';\n",
    "                        case 1441, knockdown_genes{end+1} = '4067';\n",
    "                        case 1442, knockdown_genes{end+1} = '410';\n",
    "                        case 1443, knockdown_genes{end+1} = '411';\n",
    "                        case 1444, knockdown_genes{end+1} = '4117';\n",
    "                        case 1445, knockdown_genes{end+1} = '412';\n",
    "                        case 1446, knockdown_genes{end+1} = '4121';\n",
    "                        case 1447, knockdown_genes{end+1} = '4122';\n",
    "                        case 1448, knockdown_genes{end+1} = '4123';\n",
    "                        case 1449, knockdown_genes{end+1} = '4124';\n",
    "                        case 1450, knockdown_genes{end+1} = '4125';\n",
    "                        case 1451, knockdown_genes{end+1} = '4126';\n",
    "                        case 1452, knockdown_genes{end+1} = '4128';\n",
    "                        case 1453, knockdown_genes{end+1} = '4129';\n",
    "                        case 1454, knockdown_genes{end+1} = '4139';\n",
    "                        case 1455, knockdown_genes{end+1} = '4140';\n",
    "                        case 1456, knockdown_genes{end+1} = '4141';\n",
    "                        case 1457, knockdown_genes{end+1} = '4143';\n",
    "                        case 1458, knockdown_genes{end+1} = '4144';\n",
    "                        case 1459, knockdown_genes{end+1} = '4145';\n",
    "                        case 1460, knockdown_genes{end+1} = '415116';\n",
    "                        case 1461, knockdown_genes{end+1} = '4166';\n",
    "                        case 1462, knockdown_genes{end+1} = '417';\n",
    "                        case 1463, knockdown_genes{end+1} = '419';\n",
    "                        case 1464, knockdown_genes{end+1} = '4190';\n",
    "                        case 1465, knockdown_genes{end+1} = '4191';\n",
    "                        case 1466, knockdown_genes{end+1} = '4193';\n",
    "                        case 1467, knockdown_genes{end+1} = '4199';\n",
    "                        case 1468, knockdown_genes{end+1} = '420';\n",
    "                        case 1469, knockdown_genes{end+1} = '4200';\n",
    "                        case 1470, knockdown_genes{end+1} = '4214';\n",
    "                        case 1471, knockdown_genes{end+1} = '4215';\n",
    "                        case 1472, knockdown_genes{end+1} = '4216';\n",
    "                        case 1473, knockdown_genes{end+1} = '4217';\n",
    "                        case 1474, knockdown_genes{end+1} = '4233';\n",
    "                        case 1475, knockdown_genes{end+1} = '4234';\n",
    "                        case 1476, knockdown_genes{end+1} = '4245';\n",
    "                        case 1477, knockdown_genes{end+1} = '4247';\n",
    "                        case 1478, knockdown_genes{end+1} = '4248';\n",
    "                        case 1479, knockdown_genes{end+1} = '4249';\n",
    "                        case 1480, knockdown_genes{end+1} = '4255';\n",
    "                        case 1481, knockdown_genes{end+1} = '4257';\n",
    "                        case 1482, knockdown_genes{end+1} = '4258';\n",
    "                        case 1483, knockdown_genes{end+1} = '4259';\n",
    "                        case 1484, knockdown_genes{end+1} = '427';\n",
    "                        case 1485, knockdown_genes{end+1} = '4281';\n",
    "                        case 1486, knockdown_genes{end+1} = '4282';\n",
    "                        case 1487, knockdown_genes{end+1} = '4284';\n",
    "                        case 1488, knockdown_genes{end+1} = '4293';\n",
    "                        case 1489, knockdown_genes{end+1} = '4294';\n",
    "                        case 1490, knockdown_genes{end+1} = '4296';\n",
    "                        case 1491, knockdown_genes{end+1} = '4297';\n",
    "                        case 1492, knockdown_genes{end+1} = '43';\n",
    "                        case 1493, knockdown_genes{end+1} = '4329';\n",
    "                        case 1494, knockdown_genes{end+1} = '4337';\n",
    "                        case 1495, knockdown_genes{end+1} = '4338';\n",
    "                        case 1496, knockdown_genes{end+1} = '4342';\n",
    "                        case 1497, knockdown_genes{end+1} = '435';\n",
    "                        case 1498, knockdown_genes{end+1} = '4351';\n",
    "                        case 1499, knockdown_genes{end+1} = '4353';\n",
    "                        case 1500, knockdown_genes{end+1} = '4357';\n",
    "                        case 1501, knockdown_genes{end+1} = '4363';\n",
    "                        case 1502, knockdown_genes{end+1} = '438';\n",
    "                        case 1503, knockdown_genes{end+1} = '440';\n",
    "                        case 1504, knockdown_genes{end+1} = '440138';\n",
    "                        case 1505, knockdown_genes{end+1} = '440275';\n",
    "                        case 1506, knockdown_genes{end+1} = '441024';\n",
    "                        case 1507, knockdown_genes{end+1} = '441061';\n",
    "                        case 1508, knockdown_genes{end+1} = '441282';\n",
    "                        case 1509, knockdown_genes{end+1} = '443';\n",
    "                        case 1510, knockdown_genes{end+1} = '444';\n",
    "                        case 1511, knockdown_genes{end+1} = '445';\n",
    "                        case 1512, knockdown_genes{end+1} = '4482';\n",
    "                        case 1513, knockdown_genes{end+1} = '4486';\n",
    "                        case 1514, knockdown_genes{end+1} = '4507';\n",
    "                        case 1515, knockdown_genes{end+1} = '4512';\n",
    "                        case 1516, knockdown_genes{end+1} = '4513';\n",
    "                        case 1517, knockdown_genes{end+1} = '4514';\n",
    "                        case 1518, knockdown_genes{end+1} = '4519';\n",
    "                        case 1519, knockdown_genes{end+1} = '4522';\n",
    "                        case 1520, knockdown_genes{end+1} = '4524';\n",
    "                        case 1521, knockdown_genes{end+1} = '4535';\n",
    "                        case 1522, knockdown_genes{end+1} = '4536';\n",
    "                        case 1523, knockdown_genes{end+1} = '4537';\n",
    "                        case 1524, knockdown_genes{end+1} = '4538';\n",
    "                        case 1525, knockdown_genes{end+1} = '4539';\n",
    "                        case 1526, knockdown_genes{end+1} = '4540';\n",
    "                        case 1527, knockdown_genes{end+1} = '4541';\n",
    "                        case 1528, knockdown_genes{end+1} = '4547';\n",
    "                        case 1529, knockdown_genes{end+1} = '4548';\n",
    "                        case 1530, knockdown_genes{end+1} = '4552';\n",
    "                        case 1531, knockdown_genes{end+1} = '4593';\n",
    "                        case 1532, knockdown_genes{end+1} = '4594';\n",
    "                        case 1533, knockdown_genes{end+1} = '4597';\n",
    "                        case 1534, knockdown_genes{end+1} = '4598';\n",
    "                        case 1535, knockdown_genes{end+1} = '4638';\n",
    "                        case 1536, knockdown_genes{end+1} = '4645';\n",
    "                        case 1537, knockdown_genes{end+1} = '4668';\n",
    "                        case 1538, knockdown_genes{end+1} = '4669';\n",
    "                        case 1539, knockdown_genes{end+1} = '4677';\n",
    "                        case 1540, knockdown_genes{end+1} = '4694';\n",
    "                        case 1541, knockdown_genes{end+1} = '4695';\n",
    "                        case 1542, knockdown_genes{end+1} = '4696';\n",
    "                        case 1543, knockdown_genes{end+1} = '4697';\n",
    "                        case 1544, knockdown_genes{end+1} = '4698';\n",
    "                        case 1545, knockdown_genes{end+1} = '47';\n",
    "                        case 1546, knockdown_genes{end+1} = '4700';\n",
    "                        case 1547, knockdown_genes{end+1} = '4701';\n",
    "                        case 1548, knockdown_genes{end+1} = '4702';\n",
    "                        case 1549, knockdown_genes{end+1} = '4704';\n",
    "                        case 1550, knockdown_genes{end+1} = '4705';\n",
    "                        case 1551, knockdown_genes{end+1} = '4706';\n",
    "                        case 1552, knockdown_genes{end+1} = '4707';\n",
    "                        case 1553, knockdown_genes{end+1} = '4708';\n",
    "                        case 1554, knockdown_genes{end+1} = '4709';\n",
    "                        case 1555, knockdown_genes{end+1} = '471';\n",
    "                        case 1556, knockdown_genes{end+1} = '4710';\n",
    "                        case 1557, knockdown_genes{end+1} = '4711';\n",
    "                        case 1558, knockdown_genes{end+1} = '4712';\n",
    "                        case 1559, knockdown_genes{end+1} = '4713';\n",
    "                        case 1560, knockdown_genes{end+1} = '4714';\n",
    "                        case 1561, knockdown_genes{end+1} = '4715';\n",
    "                        case 1562, knockdown_genes{end+1} = '4716';\n",
    "                        case 1563, knockdown_genes{end+1} = '4717';\n",
    "                        case 1564, knockdown_genes{end+1} = '4718';\n",
    "                        case 1565, knockdown_genes{end+1} = '4719';\n",
    "                        case 1566, knockdown_genes{end+1} = '472';\n",
    "                        case 1567, knockdown_genes{end+1} = '4720';\n",
    "                        case 1568, knockdown_genes{end+1} = '4722';\n",
    "                        case 1569, knockdown_genes{end+1} = '4723';\n",
    "                        case 1570, knockdown_genes{end+1} = '4724';\n",
    "                        case 1571, knockdown_genes{end+1} = '4725';\n",
    "                        case 1572, knockdown_genes{end+1} = '4726';\n",
    "                        case 1573, knockdown_genes{end+1} = '4728';\n",
    "                        case 1574, knockdown_genes{end+1} = '4729';\n",
    "                        case 1575, knockdown_genes{end+1} = '4731';\n",
    "                        case 1576, knockdown_genes{end+1} = '4734';\n",
    "                        case 1577, knockdown_genes{end+1} = '4750';\n",
    "                        case 1578, knockdown_genes{end+1} = '4751';\n",
    "                        case 1579, knockdown_genes{end+1} = '4752';\n",
    "                        case 1580, knockdown_genes{end+1} = '4758';\n",
    "                        case 1581, knockdown_genes{end+1} = '4759';\n",
    "                        case 1582, knockdown_genes{end+1} = '476';\n",
    "                        case 1583, knockdown_genes{end+1} = '477';\n",
    "                        case 1584, knockdown_genes{end+1} = '478';\n",
    "                        case 1585, knockdown_genes{end+1} = '48';\n",
    "                        case 1586, knockdown_genes{end+1} = '480';\n",
    "                        case 1587, knockdown_genes{end+1} = '481';\n",
    "                        case 1588, knockdown_genes{end+1} = '482';\n",
    "                        case 1589, knockdown_genes{end+1} = '4820';\n",
    "                        case 1590, knockdown_genes{end+1} = '483';\n",
    "                        case 1591, knockdown_genes{end+1} = '4830';\n",
    "                        case 1592, knockdown_genes{end+1} = '4831';\n",
    "                        case 1593, knockdown_genes{end+1} = '4832';\n",
    "                        case 1594, knockdown_genes{end+1} = '4833';\n",
    "                        case 1595, knockdown_genes{end+1} = '4835';\n",
    "                        case 1596, knockdown_genes{end+1} = '4836';\n",
    "                        case 1597, knockdown_genes{end+1} = '4837';\n",
    "                        case 1598, knockdown_genes{end+1} = '4842';\n",
    "                        case 1599, knockdown_genes{end+1} = '4843';\n",
    "                        case 1600, knockdown_genes{end+1} = '4846';\n",
    "                        case 1601, knockdown_genes{end+1} = '4850';\n",
    "                        case 1602, knockdown_genes{end+1} = '4860';\n",
    "                        case 1603, knockdown_genes{end+1} = '4862';\n",
    "                        case 1604, knockdown_genes{end+1} = '4864';\n",
    "                        case 1605, knockdown_genes{end+1} = '4881';\n",
    "                        case 1606, knockdown_genes{end+1} = '4882';\n",
    "                        case 1607, knockdown_genes{end+1} = '4891';\n",
    "                        case 1608, knockdown_genes{end+1} = '4899';\n",
    "                        case 1609, knockdown_genes{end+1} = '490';\n",
    "                        case 1610, knockdown_genes{end+1} = '4905';\n",
    "                        case 1611, knockdown_genes{end+1} = '4907';\n",
    "                        case 1612, knockdown_genes{end+1} = '491';\n",
    "                        case 1613, knockdown_genes{end+1} = '4914';\n",
    "                        case 1614, knockdown_genes{end+1} = '4915';\n",
    "                        case 1615, knockdown_genes{end+1} = '4916';\n",
    "                        case 1616, knockdown_genes{end+1} = '4919';\n",
    "                        case 1617, knockdown_genes{end+1} = '492';\n",
    "                        case 1618, knockdown_genes{end+1} = '4920';\n",
    "                        case 1619, knockdown_genes{end+1} = '4921';\n",
    "                        case 1620, knockdown_genes{end+1} = '493';\n",
    "                        case 1621, knockdown_genes{end+1} = '493869';\n",
    "                        case 1622, knockdown_genes{end+1} = '4942';\n",
    "                        case 1623, knockdown_genes{end+1} = '494551';\n",
    "                        case 1624, knockdown_genes{end+1} = '495';\n",
    "                        case 1625, knockdown_genes{end+1} = '4952';\n",
    "                        case 1626, knockdown_genes{end+1} = '4953';\n",
    "                        case 1627, knockdown_genes{end+1} = '496';\n",
    "                        case 1628, knockdown_genes{end+1} = '4967';\n",
    "                        case 1629, knockdown_genes{end+1} = '498';\n",
    "                        case 1630, knockdown_genes{end+1} = '50';\n",
    "                        case 1631, knockdown_genes{end+1} = '5002';\n",
    "                        case 1632, knockdown_genes{end+1} = '5009';\n",
    "                        case 1633, knockdown_genes{end+1} = '501';\n",
    "                        case 1634, knockdown_genes{end+1} = '5019';\n",
    "                        case 1635, knockdown_genes{end+1} = '5033';\n",
    "                        case 1636, knockdown_genes{end+1} = '5034';\n",
    "                        case 1637, knockdown_genes{end+1} = '50484';\n",
    "                        case 1638, knockdown_genes{end+1} = '50486';\n",
    "                        case 1639, knockdown_genes{end+1} = '50487';\n",
    "                        case 1640, knockdown_genes{end+1} = '50488';\n",
    "                        case 1641, knockdown_genes{end+1} = '5049';\n",
    "                        case 1642, knockdown_genes{end+1} = '5050';\n",
    "                        case 1643, knockdown_genes{end+1} = '50506';\n",
    "                        case 1644, knockdown_genes{end+1} = '50507';\n",
    "                        case 1645, knockdown_genes{end+1} = '50508';\n",
    "                        case 1646, knockdown_genes{end+1} = '5051';\n",
    "                        case 1647, knockdown_genes{end+1} = '50515';\n",
    "                        case 1648, knockdown_genes{end+1} = '5052';\n",
    "                        case 1649, knockdown_genes{end+1} = '5053';\n",
    "                        case 1650, knockdown_genes{end+1} = '5058';\n",
    "                        case 1651, knockdown_genes{end+1} = '506';\n",
    "                        case 1652, knockdown_genes{end+1} = '50614';\n",
    "                        case 1653, knockdown_genes{end+1} = '50617';\n",
    "                        case 1654, knockdown_genes{end+1} = '5062';\n",
    "                        case 1655, knockdown_genes{end+1} = '5063';\n",
    "                        case 1656, knockdown_genes{end+1} = '50640';\n",
    "                        case 1657, knockdown_genes{end+1} = '5066';\n",
    "                        case 1658, knockdown_genes{end+1} = '5068';\n",
    "                        case 1659, knockdown_genes{end+1} = '50700';\n",
    "                        case 1660, knockdown_genes{end+1} = '5071';\n",
    "                        case 1661, knockdown_genes{end+1} = '50808';\n",
    "                        case 1662, knockdown_genes{end+1} = '50814';\n",
    "                        case 1663, knockdown_genes{end+1} = '509';\n",
    "                        case 1664, knockdown_genes{end+1} = '5091';\n",
    "                        case 1665, knockdown_genes{end+1} = '5092';\n",
    "                        case 1666, knockdown_genes{end+1} = '50940';\n",
    "                        case 1667, knockdown_genes{end+1} = '5095';\n",
    "                        case 1668, knockdown_genes{end+1} = '5096';\n",
    "                        case 1669, knockdown_genes{end+1} = '51';\n",
    "                        case 1670, knockdown_genes{end+1} = '51004';\n",
    "                        case 1671, knockdown_genes{end+1} = '51005';\n",
    "                        case 1672, knockdown_genes{end+1} = '51022';\n",
    "                        case 1673, knockdown_genes{end+1} = '5105';\n",
    "                        case 1674, knockdown_genes{end+1} = '51056';\n",
    "                        case 1675, knockdown_genes{end+1} = '5106';\n",
    "                        case 1676, knockdown_genes{end+1} = '51067';\n",
    "                        case 1677, knockdown_genes{end+1} = '51071';\n",
    "                        case 1678, knockdown_genes{end+1} = '51079';\n",
    "                        case 1679, knockdown_genes{end+1} = '51082';\n",
    "                        case 1680, knockdown_genes{end+1} = '51086';\n",
    "                        case 1681, knockdown_genes{end+1} = '51099';\n",
    "                        case 1682, knockdown_genes{end+1} = '5110';\n",
    "                        case 1683, knockdown_genes{end+1} = '51102';\n",
    "                        case 1684, knockdown_genes{end+1} = '51109';\n",
    "                        case 1685, knockdown_genes{end+1} = '51111';\n",
    "                        case 1686, knockdown_genes{end+1} = '51129';\n",
    "                        case 1687, knockdown_genes{end+1} = '51135';\n",
    "                        case 1688, knockdown_genes{end+1} = '51144';\n",
    "                        case 1689, knockdown_genes{end+1} = '51146';\n",
    "                        case 1690, knockdown_genes{end+1} = '51148';\n",
    "                        case 1691, knockdown_genes{end+1} = '51166';\n",
    "                        case 1692, knockdown_genes{end+1} = '51170';\n",
    "                        case 1693, knockdown_genes{end+1} = '51171';\n",
    "                        case 1694, knockdown_genes{end+1} = '51172';\n",
    "                        case 1695, knockdown_genes{end+1} = '51179';\n",
    "                        case 1696, knockdown_genes{end+1} = '51181';\n",
    "                        case 1697, knockdown_genes{end+1} = '51196';\n",
    "                        case 1698, knockdown_genes{end+1} = '51205';\n",
    "                        case 1699, knockdown_genes{end+1} = '51207';\n",
    "                        case 1700, knockdown_genes{end+1} = '51227';\n",
    "                        case 1701, knockdown_genes{end+1} = '51228';\n",
    "                        case 1702, knockdown_genes{end+1} = '51231';\n",
    "                        case 1703, knockdown_genes{end+1} = '51251';\n",
    "                        case 1704, knockdown_genes{end+1} = '51257';\n",
    "                        case 1705, knockdown_genes{end+1} = '51265';\n",
    "                        case 1706, knockdown_genes{end+1} = '51268';\n",
    "                        case 1707, knockdown_genes{end+1} = '5127';\n",
    "                        case 1708, knockdown_genes{end+1} = '5128';\n",
    "                        case 1709, knockdown_genes{end+1} = '5129';\n",
    "                        case 1710, knockdown_genes{end+1} = '51292';\n",
    "                        case 1711, knockdown_genes{end+1} = '51296';\n",
    "                        case 1712, knockdown_genes{end+1} = '513';\n",
    "                        case 1713, knockdown_genes{end+1} = '5130';\n",
    "                        case 1714, knockdown_genes{end+1} = '51301';\n",
    "                        case 1715, knockdown_genes{end+1} = '51302';\n",
    "                        case 1716, knockdown_genes{end+1} = '51303';\n",
    "                        case 1717, knockdown_genes{end+1} = '51312';\n",
    "                        case 1718, knockdown_genes{end+1} = '51347';\n",
    "                        case 1719, knockdown_genes{end+1} = '5136';\n",
    "                        case 1720, knockdown_genes{end+1} = '51366';\n",
    "                        case 1721, knockdown_genes{end+1} = '5137';\n",
    "                        case 1722, knockdown_genes{end+1} = '5138';\n",
    "                        case 1723, knockdown_genes{end+1} = '51380';\n",
    "                        case 1724, knockdown_genes{end+1} = '51382';\n",
    "                        case 1725, knockdown_genes{end+1} = '5139';\n",
    "                        case 1726, knockdown_genes{end+1} = '514';\n",
    "                        case 1727, knockdown_genes{end+1} = '5140';\n",
    "                        case 1728, knockdown_genes{end+1} = '5141';\n",
    "                        case 1729, knockdown_genes{end+1} = '5142';\n",
    "                        case 1730, knockdown_genes{end+1} = '51422';\n",
    "                        case 1731, knockdown_genes{end+1} = '51426';\n",
    "                        case 1732, knockdown_genes{end+1} = '5143';\n",
    "                        case 1733, knockdown_genes{end+1} = '5144';\n",
    "                        case 1734, knockdown_genes{end+1} = '51444';\n",
    "                        case 1735, knockdown_genes{end+1} = '51447';\n",
    "                        case 1736, knockdown_genes{end+1} = '51449';\n",
    "                        case 1737, knockdown_genes{end+1} = '5145';\n",
    "                        case 1738, knockdown_genes{end+1} = '51458';\n",
    "                        case 1739, knockdown_genes{end+1} = '5146';\n",
    "                        case 1740, knockdown_genes{end+1} = '51465';\n",
    "                        case 1741, knockdown_genes{end+1} = '51477';\n",
    "                        case 1742, knockdown_genes{end+1} = '51478';\n",
    "                        case 1743, knockdown_genes{end+1} = '5149';\n",
    "                        case 1744, knockdown_genes{end+1} = '51495';\n",
    "                        case 1745, knockdown_genes{end+1} = '51496';\n",
    "                        case 1746, knockdown_genes{end+1} = '515';\n",
    "                        case 1747, knockdown_genes{end+1} = '5150';\n",
    "                        case 1748, knockdown_genes{end+1} = '5151';\n",
    "                        case 1749, knockdown_genes{end+1} = '5152';\n",
    "                        case 1750, knockdown_genes{end+1} = '51520';\n",
    "                        case 1751, knockdown_genes{end+1} = '51529';\n",
    "                        case 1752, knockdown_genes{end+1} = '5153';\n",
    "                        case 1753, knockdown_genes{end+1} = '51540';\n",
    "                        case 1754, knockdown_genes{end+1} = '51547';\n",
    "                        case 1755, knockdown_genes{end+1} = '51548';\n",
    "                        case 1756, knockdown_genes{end+1} = '51550';\n",
    "                        case 1757, knockdown_genes{end+1} = '51557';\n",
    "                        case 1758, knockdown_genes{end+1} = '51559';\n",
    "                        case 1759, knockdown_genes{end+1} = '5156';\n",
    "                        case 1760, knockdown_genes{end+1} = '51573';\n",
    "                        case 1761, knockdown_genes{end+1} = '5158';\n",
    "                        case 1762, knockdown_genes{end+1} = '5159';\n",
    "                        case 1763, knockdown_genes{end+1} = '51592';\n",
    "                        case 1764, knockdown_genes{end+1} = '516';\n",
    "                        case 1765, knockdown_genes{end+1} = '5160';\n",
    "                        case 1766, knockdown_genes{end+1} = '51601';\n",
    "                        case 1767, knockdown_genes{end+1} = '51604';\n",
    "                        case 1768, knockdown_genes{end+1} = '51606';\n",
    "                        case 1769, knockdown_genes{end+1} = '5161';\n",
    "                        case 1770, knockdown_genes{end+1} = '51611';\n",
    "                        case 1771, knockdown_genes{end+1} = '51619';\n",
    "                        case 1772, knockdown_genes{end+1} = '5162';\n",
    "                        case 1773, knockdown_genes{end+1} = '5163';\n",
    "                        case 1774, knockdown_genes{end+1} = '5164';\n",
    "                        case 1775, knockdown_genes{end+1} = '51645';\n",
    "                        case 1776, knockdown_genes{end+1} = '5165';\n",
    "                        case 1777, knockdown_genes{end+1} = '51651';\n",
    "                        case 1778, knockdown_genes{end+1} = '51657';\n",
    "                        case 1779, knockdown_genes{end+1} = '5166';\n",
    "                        case 1780, knockdown_genes{end+1} = '51660';\n",
    "                        case 1781, knockdown_genes{end+1} = '51661';\n",
    "                        case 1782, knockdown_genes{end+1} = '5167';\n",
    "                        case 1783, knockdown_genes{end+1} = '5168';\n",
    "                        case 1784, knockdown_genes{end+1} = '5169';\n",
    "                        case 1785, knockdown_genes{end+1} = '517';\n",
    "                        case 1786, knockdown_genes{end+1} = '5170';\n",
    "                        case 1787, knockdown_genes{end+1} = '51700';\n",
    "                        case 1788, knockdown_genes{end+1} = '51701';\n",
    "                        case 1789, knockdown_genes{end+1} = '51702';\n",
    "                        case 1790, knockdown_genes{end+1} = '51703';\n",
    "                        case 1791, knockdown_genes{end+1} = '51706';\n",
    "                        case 1792, knockdown_genes{end+1} = '5172';\n",
    "                        case 1793, knockdown_genes{end+1} = '51727';\n",
    "                        case 1794, knockdown_genes{end+1} = '51728';\n",
    "                        case 1795, knockdown_genes{end+1} = '51733';\n",
    "                        case 1796, knockdown_genes{end+1} = '51741';\n",
    "                        case 1797, knockdown_genes{end+1} = '51755';\n",
    "                        case 1798, knockdown_genes{end+1} = '51765';\n",
    "                        case 1799, knockdown_genes{end+1} = '51776';\n",
    "                        case 1800, knockdown_genes{end+1} = '518';\n",
    "                        case 1801, knockdown_genes{end+1} = '51805';\n",
    "                        case 1802, knockdown_genes{end+1} = '51809';\n",
    "                        case 1803, knockdown_genes{end+1} = '5184';\n",
    "                        case 1804, knockdown_genes{end+1} = '5198';\n",
    "                        case 1805, knockdown_genes{end+1} = '52';\n",
    "                        case 1806, knockdown_genes{end+1} = '5205';\n",
    "                        case 1807, knockdown_genes{end+1} = '5207';\n",
    "                        case 1808, knockdown_genes{end+1} = '5208';\n",
    "                        case 1809, knockdown_genes{end+1} = '5209';\n",
    "                        case 1810, knockdown_genes{end+1} = '521';\n",
    "                        case 1811, knockdown_genes{end+1} = '5210';\n",
    "                        case 1812, knockdown_genes{end+1} = '5211';\n",
    "                        case 1813, knockdown_genes{end+1} = '5213';\n",
    "                        case 1814, knockdown_genes{end+1} = '5214';\n",
    "                        case 1815, knockdown_genes{end+1} = '5218';\n",
    "                        case 1816, knockdown_genes{end+1} = '522';\n",
    "                        case 1817, knockdown_genes{end+1} = '5223';\n",
    "                        case 1818, knockdown_genes{end+1} = '5224';\n",
    "                        case 1819, knockdown_genes{end+1} = '5226';\n",
    "                        case 1820, knockdown_genes{end+1} = '523';\n",
    "                        case 1821, knockdown_genes{end+1} = '5230';\n",
    "                        case 1822, knockdown_genes{end+1} = '5232';\n",
    "                        case 1823, knockdown_genes{end+1} = '5236';\n",
    "                        case 1824, knockdown_genes{end+1} = '5238';\n",
    "                        case 1825, knockdown_genes{end+1} = '5243';\n",
    "                        case 1826, knockdown_genes{end+1} = '5244';\n",
    "                        case 1827, knockdown_genes{end+1} = '525';\n",
    "                        case 1828, knockdown_genes{end+1} = '5250';\n",
    "                        case 1829, knockdown_genes{end+1} = '5251';\n",
    "                        case 1830, knockdown_genes{end+1} = '5255';\n",
    "                        case 1831, knockdown_genes{end+1} = '5256';\n",
    "                        case 1832, knockdown_genes{end+1} = '5257';\n",
    "                        case 1833, knockdown_genes{end+1} = '526';\n",
    "                        case 1834, knockdown_genes{end+1} = '5260';\n",
    "                        case 1835, knockdown_genes{end+1} = '5261';\n",
    "                        case 1836, knockdown_genes{end+1} = '5264';\n",
    "                        case 1837, knockdown_genes{end+1} = '5265';\n",
    "                        case 1838, knockdown_genes{end+1} = '527';\n",
    "                        case 1839, knockdown_genes{end+1} = '5277';\n",
    "                        case 1840, knockdown_genes{end+1} = '5279';\n",
    "                        case 1841, knockdown_genes{end+1} = '528';\n",
    "                        case 1842, knockdown_genes{end+1} = '5281';\n",
    "                        case 1843, knockdown_genes{end+1} = '5283';\n",
    "                        case 1844, knockdown_genes{end+1} = '5286';\n",
    "                        case 1845, knockdown_genes{end+1} = '5287';\n",
    "                        case 1846, knockdown_genes{end+1} = '5288';\n",
    "                        case 1847, knockdown_genes{end+1} = '5289';\n",
    "                        case 1848, knockdown_genes{end+1} = '529';\n",
    "                        case 1849, knockdown_genes{end+1} = '5290';\n",
    "                        case 1850, knockdown_genes{end+1} = '5291';\n",
    "                        case 1851, knockdown_genes{end+1} = '5292';\n",
    "                        case 1852, knockdown_genes{end+1} = '5293';\n",
    "                        case 1853, knockdown_genes{end+1} = '5294';\n",
    "                        case 1854, knockdown_genes{end+1} = '5295';\n",
    "                        case 1855, knockdown_genes{end+1} = '5296';\n",
    "                        case 1856, knockdown_genes{end+1} = '5297';\n",
    "                        case 1857, knockdown_genes{end+1} = '5298';\n",
    "                        case 1858, knockdown_genes{end+1} = '53';\n",
    "                        case 1859, knockdown_genes{end+1} = '5300';\n",
    "                        case 1860, knockdown_genes{end+1} = '5303';\n",
    "                        case 1861, knockdown_genes{end+1} = '5305';\n",
    "                        case 1862, knockdown_genes{end+1} = '5313';\n",
    "                        case 1863, knockdown_genes{end+1} = '5315';\n",
    "                        case 1864, knockdown_genes{end+1} = '5319';\n",
    "                        case 1865, knockdown_genes{end+1} = '5320';\n",
    "                        case 1866, knockdown_genes{end+1} = '5321';\n",
    "                        case 1867, knockdown_genes{end+1} = '5322';\n",
    "                        case 1868, knockdown_genes{end+1} = '533';\n",
    "                        case 1869, knockdown_genes{end+1} = '5330';\n",
    "                        case 1870, knockdown_genes{end+1} = '5331';\n",
    "                        case 1871, knockdown_genes{end+1} = '5332';\n",
    "                        case 1872, knockdown_genes{end+1} = '5333';\n",
    "                        case 1873, knockdown_genes{end+1} = '5334';\n",
    "                        case 1874, knockdown_genes{end+1} = '53343';\n",
    "                        case 1875, knockdown_genes{end+1} = '5335';\n",
    "                        case 1876, knockdown_genes{end+1} = '53354';\n",
    "                        case 1877, knockdown_genes{end+1} = '5336';\n",
    "                        case 1878, knockdown_genes{end+1} = '5337';\n",
    "                        case 1879, knockdown_genes{end+1} = '5338';\n",
    "                        case 1880, knockdown_genes{end+1} = '534';\n",
    "                        case 1881, knockdown_genes{end+1} = '5340';\n",
    "                        case 1882, knockdown_genes{end+1} = '5346';\n",
    "                        case 1883, knockdown_genes{end+1} = '5347';\n",
    "                        case 1884, knockdown_genes{end+1} = '535';\n",
    "                        case 1885, knockdown_genes{end+1} = '5351';\n",
    "                        case 1886, knockdown_genes{end+1} = '5352';\n",
    "                        case 1887, knockdown_genes{end+1} = '5360';\n",
    "                        case 1888, knockdown_genes{end+1} = '53630';\n",
    "                        case 1889, knockdown_genes{end+1} = '5365';\n",
    "                        case 1890, knockdown_genes{end+1} = '5372';\n",
    "                        case 1891, knockdown_genes{end+1} = '5373';\n",
    "                        case 1892, knockdown_genes{end+1} = '539';\n",
    "                        case 1893, knockdown_genes{end+1} = '53904';\n",
    "                        case 1894, knockdown_genes{end+1} = '53905';\n",
    "                        case 1895, knockdown_genes{end+1} = '53919';\n",
    "                        case 1896, knockdown_genes{end+1} = '53938';\n",
    "                        case 1897, knockdown_genes{end+1} = '53944';\n",
    "                        case 1898, knockdown_genes{end+1} = '53947';\n",
    "                        case 1899, knockdown_genes{end+1} = '54';\n",
    "                        case 1900, knockdown_genes{end+1} = '54020';\n",
    "                        case 1901, knockdown_genes{end+1} = '5406';\n",
    "                        case 1902, knockdown_genes{end+1} = '5407';\n",
    "                        case 1903, knockdown_genes{end+1} = '5408';\n",
    "                        case 1904, knockdown_genes{end+1} = '5409';\n",
    "                        case 1905, knockdown_genes{end+1} = '54101';\n",
    "                        case 1906, knockdown_genes{end+1} = '54107';\n",
    "                        case 1907, knockdown_genes{end+1} = '54187';\n",
    "                        case 1908, knockdown_genes{end+1} = '5422';\n",
    "                        case 1909, knockdown_genes{end+1} = '5423';\n",
    "                        case 1910, knockdown_genes{end+1} = '5424';\n",
    "                        case 1911, knockdown_genes{end+1} = '5425';\n",
    "                        case 1912, knockdown_genes{end+1} = '5426';\n",
    "                        case 1913, knockdown_genes{end+1} = '5427';\n",
    "                        case 1914, knockdown_genes{end+1} = '5428';\n",
    "                        case 1915, knockdown_genes{end+1} = '5429';\n",
    "                        case 1916, knockdown_genes{end+1} = '5430';\n",
    "                        case 1917, knockdown_genes{end+1} = '5431';\n",
    "                        case 1918, knockdown_genes{end+1} = '5432';\n",
    "                        case 1919, knockdown_genes{end+1} = '5433';\n",
    "                        case 1920, knockdown_genes{end+1} = '5434';\n",
    "                        case 1921, knockdown_genes{end+1} = '54344';\n",
    "                        case 1922, knockdown_genes{end+1} = '5435';\n",
    "                        case 1923, knockdown_genes{end+1} = '5436';\n",
    "                        case 1924, knockdown_genes{end+1} = '54363';\n",
    "                        case 1925, knockdown_genes{end+1} = '5437';\n",
    "                        case 1926, knockdown_genes{end+1} = '5438';\n",
    "                        case 1927, knockdown_genes{end+1} = '5439';\n",
    "                        case 1928, knockdown_genes{end+1} = '5440';\n",
    "                        case 1929, knockdown_genes{end+1} = '54407';\n",
    "                        case 1930, knockdown_genes{end+1} = '5441';\n",
    "                        case 1931, knockdown_genes{end+1} = '54414';\n",
    "                        case 1932, knockdown_genes{end+1} = '5442';\n",
    "                        case 1933, knockdown_genes{end+1} = '54434';\n",
    "                        case 1934, knockdown_genes{end+1} = '5446';\n",
    "                        case 1935, knockdown_genes{end+1} = '5447';\n",
    "                        case 1936, knockdown_genes{end+1} = '54476';\n",
    "                        case 1937, knockdown_genes{end+1} = '54480';\n",
    "                        case 1938, knockdown_genes{end+1} = '54490';\n",
    "                        case 1939, knockdown_genes{end+1} = '54492';\n",
    "                        case 1940, knockdown_genes{end+1} = '54495';\n",
    "                        case 1941, knockdown_genes{end+1} = '545';\n",
    "                        case 1942, knockdown_genes{end+1} = '54511';\n",
    "                        case 1943, knockdown_genes{end+1} = '54575';\n",
    "                        case 1944, knockdown_genes{end+1} = '54576';\n",
    "                        case 1945, knockdown_genes{end+1} = '54577';\n",
    "                        case 1946, knockdown_genes{end+1} = '54578';\n",
    "                        case 1947, knockdown_genes{end+1} = '54579';\n",
    "                        case 1948, knockdown_genes{end+1} = '54600';\n",
    "                        case 1949, knockdown_genes{end+1} = '54625';\n",
    "                        case 1950, knockdown_genes{end+1} = '5464';\n",
    "                        case 1951, knockdown_genes{end+1} = '5465';\n",
    "                        case 1952, knockdown_genes{end+1} = '54657';\n",
    "                        case 1953, knockdown_genes{end+1} = '54658';\n",
    "                        case 1954, knockdown_genes{end+1} = '54659';\n",
    "                        case 1955, knockdown_genes{end+1} = '54675';\n",
    "                        case 1956, knockdown_genes{end+1} = '54677';\n",
    "                        case 1957, knockdown_genes{end+1} = '54682';\n",
    "                        case 1958, knockdown_genes{end+1} = '5469';\n",
    "                        case 1959, knockdown_genes{end+1} = '5470';\n",
    "                        case 1960, knockdown_genes{end+1} = '54704';\n",
    "                        case 1961, knockdown_genes{end+1} = '54708';\n",
    "                        case 1962, knockdown_genes{end+1} = '5471';\n",
    "                        case 1963, knockdown_genes{end+1} = '54716';\n",
    "                        case 1964, knockdown_genes{end+1} = '5475';\n",
    "                        case 1965, knockdown_genes{end+1} = '5476';\n",
    "                        case 1966, knockdown_genes{end+1} = '5478';\n",
    "                        case 1967, knockdown_genes{end+1} = '5479';\n",
    "                        case 1968, knockdown_genes{end+1} = '5480';\n",
    "                        case 1969, knockdown_genes{end+1} = '54802';\n",
    "                        case 1970, knockdown_genes{end+1} = '5481';\n",
    "                        case 1971, knockdown_genes{end+1} = '54822';\n",
    "                        case 1972, knockdown_genes{end+1} = '54857';\n",
    "                        case 1973, knockdown_genes{end+1} = '548596';\n",
    "                        case 1974, knockdown_genes{end+1} = '54861';\n",
    "                        case 1975, knockdown_genes{end+1} = '54872';\n",
    "                        case 1976, knockdown_genes{end+1} = '54898';\n",
    "                        case 1977, knockdown_genes{end+1} = '549';\n",
    "                        case 1978, knockdown_genes{end+1} = '54904';\n",
    "                        case 1979, knockdown_genes{end+1} = '54926';\n",
    "                        case 1980, knockdown_genes{end+1} = '54935';\n",
    "                        case 1981, knockdown_genes{end+1} = '54938';\n",
    "                        case 1982, knockdown_genes{end+1} = '5494';\n",
    "                        case 1983, knockdown_genes{end+1} = '54941';\n",
    "                        case 1984, knockdown_genes{end+1} = '54947';\n",
    "                        case 1985, knockdown_genes{end+1} = '5495';\n",
    "                        case 1986, knockdown_genes{end+1} = '54956';\n",
    "                        case 1987, knockdown_genes{end+1} = '5496';\n",
    "                        case 1988, knockdown_genes{end+1} = '54961';\n",
    "                        case 1989, knockdown_genes{end+1} = '54963';\n",
    "                        case 1990, knockdown_genes{end+1} = '54965';\n",
    "                        case 1991, knockdown_genes{end+1} = '5498';\n",
    "                        case 1992, knockdown_genes{end+1} = '54981';\n",
    "                        case 1993, knockdown_genes{end+1} = '54986';\n",
    "                        case 1994, knockdown_genes{end+1} = '54988';\n",
    "                        case 1995, knockdown_genes{end+1} = '5499';\n",
    "                        case 1996, knockdown_genes{end+1} = '54995';\n",
    "                        case 1997, knockdown_genes{end+1} = '55';\n",
    "                        case 1998, knockdown_genes{end+1} = '5500';\n",
    "                        case 1999, knockdown_genes{end+1} = '5501';\n",
    "                        case 2000, knockdown_genes{end+1} = '55016';\n",
    "                        case 2001, knockdown_genes{end+1} = '5502';\n",
    "                        case 2002, knockdown_genes{end+1} = '55031';\n",
    "                        case 2003, knockdown_genes{end+1} = '55033';\n",
    "                        case 2004, knockdown_genes{end+1} = '5504';\n",
    "                        case 2005, knockdown_genes{end+1} = '5506';\n",
    "                        case 2006, knockdown_genes{end+1} = '55065';\n",
    "                        case 2007, knockdown_genes{end+1} = '55066';\n",
    "                        case 2008, knockdown_genes{end+1} = '5507';\n",
    "                        case 2009, knockdown_genes{end+1} = '55072';\n",
    "                        case 2010, knockdown_genes{end+1} = '55089';\n",
    "                        case 2011, knockdown_genes{end+1} = '55093';\n",
    "                        case 2012, knockdown_genes{end+1} = '55120';\n",
    "                        case 2013, knockdown_genes{end+1} = '55140';\n",
    "                        case 2014, knockdown_genes{end+1} = '55148';\n",
    "                        case 2015, knockdown_genes{end+1} = '55149';\n",
    "                        case 2016, knockdown_genes{end+1} = '5515';\n",
    "                        case 2017, knockdown_genes{end+1} = '55157';\n",
    "                        case 2018, knockdown_genes{end+1} = '55159';\n",
    "                        case 2019, knockdown_genes{end+1} = '5516';\n",
    "                        case 2020, knockdown_genes{end+1} = '55163';\n",
    "                        case 2021, knockdown_genes{end+1} = '5518';\n",
    "                        case 2022, knockdown_genes{end+1} = '5519';\n",
    "                        case 2023, knockdown_genes{end+1} = '55190';\n",
    "                        case 2024, knockdown_genes{end+1} = '55191';\n",
    "                        case 2025, knockdown_genes{end+1} = '5520';\n",
    "                        case 2026, knockdown_genes{end+1} = '55209';\n",
    "                        case 2027, knockdown_genes{end+1} = '5521';\n",
    "                        case 2028, knockdown_genes{end+1} = '55217';\n",
    "                        case 2029, knockdown_genes{end+1} = '5522';\n",
    "                        case 2030, knockdown_genes{end+1} = '55224';\n",
    "                        case 2031, knockdown_genes{end+1} = '55229';\n",
    "                        case 2032, knockdown_genes{end+1} = '5523';\n",
    "                        case 2033, knockdown_genes{end+1} = '55230';\n",
    "                        case 2034, knockdown_genes{end+1} = '55236';\n",
    "                        case 2035, knockdown_genes{end+1} = '55244';\n",
    "                        case 2036, knockdown_genes{end+1} = '55256';\n",
    "                        case 2037, knockdown_genes{end+1} = '55268';\n",
    "                        case 2038, knockdown_genes{end+1} = '55276';\n",
    "                        case 2039, knockdown_genes{end+1} = '55277';\n",
    "                        case 2040, knockdown_genes{end+1} = '55284';\n",
    "                        case 2041, knockdown_genes{end+1} = '55293';\n",
    "                        case 2042, knockdown_genes{end+1} = '5530';\n",
    "                        case 2043, knockdown_genes{end+1} = '55301';\n",
    "                        case 2044, knockdown_genes{end+1} = '55304';\n",
    "                        case 2045, knockdown_genes{end+1} = '5531';\n",
    "                        case 2046, knockdown_genes{end+1} = '55312';\n",
    "                        case 2047, knockdown_genes{end+1} = '55315';\n",
    "                        case 2048, knockdown_genes{end+1} = '5532';\n",
    "                        case 2049, knockdown_genes{end+1} = '55326';\n",
    "                        case 2050, knockdown_genes{end+1} = '5533';\n",
    "                        case 2051, knockdown_genes{end+1} = '55331';\n",
    "                        case 2052, knockdown_genes{end+1} = '5534';\n",
    "                        case 2053, knockdown_genes{end+1} = '55343';\n",
    "                        case 2054, knockdown_genes{end+1} = '55349';\n",
    "                        case 2055, knockdown_genes{end+1} = '5535';\n",
    "                        case 2056, knockdown_genes{end+1} = '55350';\n",
    "                        case 2057, knockdown_genes{end+1} = '55351';\n",
    "                        case 2058, knockdown_genes{end+1} = '55359';\n",
    "                        case 2059, knockdown_genes{end+1} = '5536';\n",
    "                        case 2060, knockdown_genes{end+1} = '55361';\n",
    "                        case 2061, knockdown_genes{end+1} = '5537';\n",
    "                        case 2062, knockdown_genes{end+1} = '5538';\n",
    "                        case 2063, knockdown_genes{end+1} = '55454';\n",
    "                        case 2064, knockdown_genes{end+1} = '5550';\n",
    "                        case 2065, knockdown_genes{end+1} = '55500';\n",
    "                        case 2066, knockdown_genes{end+1} = '55501';\n",
    "                        case 2067, knockdown_genes{end+1} = '55512';\n",
    "                        case 2068, knockdown_genes{end+1} = '55561';\n",
    "                        case 2069, knockdown_genes{end+1} = '55568';\n",
    "                        case 2070, knockdown_genes{end+1} = '55577';\n",
    "                        case 2071, knockdown_genes{end+1} = '55585';\n",
    "                        case 2072, knockdown_genes{end+1} = '55586';\n",
    "                        case 2073, knockdown_genes{end+1} = '55589';\n",
    "                        case 2074, knockdown_genes{end+1} = '55593';\n",
    "                        case 2075, knockdown_genes{end+1} = '55613';\n",
    "                        case 2076, knockdown_genes{end+1} = '5562';\n",
    "                        case 2077, knockdown_genes{end+1} = '55621';\n",
    "                        case 2078, knockdown_genes{end+1} = '55624';\n",
    "                        case 2079, knockdown_genes{end+1} = '55627';\n",
    "                        case 2080, knockdown_genes{end+1} = '5563';\n",
    "                        case 2081, knockdown_genes{end+1} = '5565';\n",
    "                        case 2082, knockdown_genes{end+1} = '55650';\n",
    "                        case 2083, knockdown_genes{end+1} = '55658';\n",
    "                        case 2084, knockdown_genes{end+1} = '5566';\n",
    "                        case 2085, knockdown_genes{end+1} = '55662';\n",
    "                        case 2086, knockdown_genes{end+1} = '5567';\n",
    "                        case 2087, knockdown_genes{end+1} = '55676';\n",
    "                        case 2088, knockdown_genes{end+1} = '5568';\n",
    "                        case 2089, knockdown_genes{end+1} = '55683';\n",
    "                        case 2090, knockdown_genes{end+1} = '55687';\n",
    "                        case 2091, knockdown_genes{end+1} = '55699';\n",
    "                        case 2092, knockdown_genes{end+1} = '55703';\n",
    "                        case 2093, knockdown_genes{end+1} = '55711';\n",
    "                        case 2094, knockdown_genes{end+1} = '55718';\n",
    "                        case 2095, knockdown_genes{end+1} = '5573';\n",
    "                        case 2096, knockdown_genes{end+1} = '55743';\n",
    "                        case 2097, knockdown_genes{end+1} = '55748';\n",
    "                        case 2098, knockdown_genes{end+1} = '5575';\n",
    "                        case 2099, knockdown_genes{end+1} = '55750';\n",
    "                        case 2100, knockdown_genes{end+1} = '55757';\n",
    "                        case 2101, knockdown_genes{end+1} = '5576';\n",
    "                        case 2102, knockdown_genes{end+1} = '55768';\n",
    "                        case 2103, knockdown_genes{end+1} = '5577';\n",
    "                        case 2104, knockdown_genes{end+1} = '5578';\n",
    "                        case 2105, knockdown_genes{end+1} = '55781';\n",
    "                        case 2106, knockdown_genes{end+1} = '5579';\n",
    "                        case 2107, knockdown_genes{end+1} = '55790';\n",
    "                        case 2108, knockdown_genes{end+1} = '558';\n",
    "                        case 2109, knockdown_genes{end+1} = '5580';\n",
    "                        case 2110, knockdown_genes{end+1} = '55808';\n",
    "                        case 2111, knockdown_genes{end+1} = '5581';\n",
    "                        case 2112, knockdown_genes{end+1} = '55811';\n",
    "                        case 2113, knockdown_genes{end+1} = '55819';\n",
    "                        case 2114, knockdown_genes{end+1} = '5582';\n",
    "                        case 2115, knockdown_genes{end+1} = '55821';\n",
    "                        case 2116, knockdown_genes{end+1} = '55825';\n",
    "                        case 2117, knockdown_genes{end+1} = '5583';\n",
    "                        case 2118, knockdown_genes{end+1} = '55830';\n",
    "                        case 2119, knockdown_genes{end+1} = '5584';\n",
    "                        case 2120, knockdown_genes{end+1} = '5585';\n",
    "                        case 2121, knockdown_genes{end+1} = '55856';\n",
    "                        case 2122, knockdown_genes{end+1} = '5586';\n",
    "                        case 2123, knockdown_genes{end+1} = '55867';\n",
    "                        case 2124, knockdown_genes{end+1} = '5587';\n",
    "                        case 2125, knockdown_genes{end+1} = '55870';\n",
    "                        case 2126, knockdown_genes{end+1} = '55872';\n",
    "                        case 2127, knockdown_genes{end+1} = '5588';\n",
    "                        case 2128, knockdown_genes{end+1} = '5589';\n",
    "                        case 2129, knockdown_genes{end+1} = '5590';\n",
    "                        case 2130, knockdown_genes{end+1} = '55902';\n",
    "                        case 2131, knockdown_genes{end+1} = '55904';\n",
    "                        case 2132, knockdown_genes{end+1} = '55907';\n",
    "                        case 2133, knockdown_genes{end+1} = '5591';\n",
    "                        case 2134, knockdown_genes{end+1} = '5592';\n",
    "                        case 2135, knockdown_genes{end+1} = '5593';\n",
    "                        case 2136, knockdown_genes{end+1} = '5594';\n",
    "                        case 2137, knockdown_genes{end+1} = '5595';\n",
    "                        case 2138, knockdown_genes{end+1} = '5596';\n",
    "                        case 2139, knockdown_genes{end+1} = '55967';\n",
    "                        case 2140, knockdown_genes{end+1} = '5597';\n",
    "                        case 2141, knockdown_genes{end+1} = '5598';\n",
    "                        case 2142, knockdown_genes{end+1} = '5599';\n",
    "                        case 2143, knockdown_genes{end+1} = '5600';\n",
    "                        case 2144, knockdown_genes{end+1} = '5601';\n",
    "                        case 2145, knockdown_genes{end+1} = '5602';\n",
    "                        case 2146, knockdown_genes{end+1} = '5603';\n",
    "                        case 2147, knockdown_genes{end+1} = '5604';\n",
    "                        case 2148, knockdown_genes{end+1} = '5605';\n",
    "                        case 2149, knockdown_genes{end+1} = '56052';\n",
    "                        case 2150, knockdown_genes{end+1} = '5606';\n",
    "                        case 2151, knockdown_genes{end+1} = '5607';\n",
    "                        case 2152, knockdown_genes{end+1} = '5608';\n",
    "                        case 2153, knockdown_genes{end+1} = '5609';\n",
    "                        case 2154, knockdown_genes{end+1} = '5610';\n",
    "                        case 2155, knockdown_genes{end+1} = '5613';\n",
    "                        case 2156, knockdown_genes{end+1} = '5616';\n",
    "                        case 2157, knockdown_genes{end+1} = '56164';\n",
    "                        case 2158, knockdown_genes{end+1} = '5625';\n",
    "                        case 2159, knockdown_genes{end+1} = '56254';\n",
    "                        case 2160, knockdown_genes{end+1} = '56261';\n",
    "                        case 2161, knockdown_genes{end+1} = '56267';\n",
    "                        case 2162, knockdown_genes{end+1} = '56301';\n",
    "                        case 2163, knockdown_genes{end+1} = '5631';\n",
    "                        case 2164, knockdown_genes{end+1} = '56339';\n",
    "                        case 2165, knockdown_genes{end+1} = '5634';\n",
    "                        case 2166, knockdown_genes{end+1} = '5641';\n",
    "                        case 2167, knockdown_genes{end+1} = '56474';\n",
    "                        case 2168, knockdown_genes{end+1} = '56548';\n",
    "                        case 2169, knockdown_genes{end+1} = '5660';\n",
    "                        case 2170, knockdown_genes{end+1} = '56603';\n",
    "                        case 2171, knockdown_genes{end+1} = '56606';\n",
    "                        case 2172, knockdown_genes{end+1} = '56623';\n",
    "                        case 2173, knockdown_genes{end+1} = '56624';\n",
    "                        case 2174, knockdown_genes{end+1} = '56655';\n",
    "                        case 2175, knockdown_genes{end+1} = '5681';\n",
    "                        case 2176, knockdown_genes{end+1} = '56829';\n",
    "                        case 2177, knockdown_genes{end+1} = '56848';\n",
    "                        case 2178, knockdown_genes{end+1} = '56852';\n",
    "                        case 2179, knockdown_genes{end+1} = '56894';\n",
    "                        case 2180, knockdown_genes{end+1} = '56895';\n",
    "                        case 2181, knockdown_genes{end+1} = '56898';\n",
    "                        case 2182, knockdown_genes{end+1} = '56903';\n",
    "                        case 2183, knockdown_genes{end+1} = '56913';\n",
    "                        case 2184, knockdown_genes{end+1} = '56922';\n",
    "                        case 2185, knockdown_genes{end+1} = '56924';\n",
    "                        case 2186, knockdown_genes{end+1} = '56940';\n",
    "                        case 2187, knockdown_genes{end+1} = '56953';\n",
    "                        case 2188, knockdown_genes{end+1} = '56954';\n",
    "                        case 2189, knockdown_genes{end+1} = '56965';\n",
    "                        case 2190, knockdown_genes{end+1} = '56994';\n",
    "                        case 2191, knockdown_genes{end+1} = '570';\n",
    "                        case 2192, knockdown_genes{end+1} = '57016';\n",
    "                        case 2193, knockdown_genes{end+1} = '57026';\n",
    "                        case 2194, knockdown_genes{end+1} = '57030';\n",
    "                        case 2195, knockdown_genes{end+1} = '57038';\n",
    "                        case 2196, knockdown_genes{end+1} = '57084';\n",
    "                        case 2197, knockdown_genes{end+1} = '57094';\n",
    "                        case 2198, knockdown_genes{end+1} = '57097';\n",
    "                        case 2199, knockdown_genes{end+1} = '57104';\n",
    "                        case 2200, knockdown_genes{end+1} = '57115';\n",
    "                        case 2201, knockdown_genes{end+1} = '57118';\n",
    "                        case 2202, knockdown_genes{end+1} = '57127';\n",
    "                        case 2203, knockdown_genes{end+1} = '57134';\n",
    "                        case 2204, knockdown_genes{end+1} = '57144';\n",
    "                        case 2205, knockdown_genes{end+1} = '57154';\n",
    "                        case 2206, knockdown_genes{end+1} = '57171';\n",
    "                        case 2207, knockdown_genes{end+1} = '57172';\n",
    "                        case 2208, knockdown_genes{end+1} = '57176';\n",
    "                        case 2209, knockdown_genes{end+1} = '57194';\n",
    "                        case 2210, knockdown_genes{end+1} = '57198';\n",
    "                        case 2211, knockdown_genes{end+1} = '5723';\n",
    "                        case 2212, knockdown_genes{end+1} = '5728';\n",
    "                        case 2213, knockdown_genes{end+1} = '57282';\n",
    "                        case 2214, knockdown_genes{end+1} = '5730';\n",
    "                        case 2215, knockdown_genes{end+1} = '57379';\n",
    "                        case 2216, knockdown_genes{end+1} = '57393';\n",
    "                        case 2217, knockdown_genes{end+1} = '57396';\n",
    "                        case 2218, knockdown_genes{end+1} = '5740';\n",
    "                        case 2219, knockdown_genes{end+1} = '57406';\n",
    "                        case 2220, knockdown_genes{end+1} = '57412';\n",
    "                        case 2221, knockdown_genes{end+1} = '57419';\n",
    "                        case 2222, knockdown_genes{end+1} = '5742';\n",
    "                        case 2223, knockdown_genes{end+1} = '5743';\n",
    "                        case 2224, knockdown_genes{end+1} = '57448';\n",
    "                        case 2225, knockdown_genes{end+1} = '57468';\n",
    "                        case 2226, knockdown_genes{end+1} = '5747';\n",
    "                        case 2227, knockdown_genes{end+1} = '57478';\n",
    "                        case 2228, knockdown_genes{end+1} = '57486';\n",
    "                        case 2229, knockdown_genes{end+1} = '57505';\n",
    "                        case 2230, knockdown_genes{end+1} = '57520';\n",
    "                        case 2231, knockdown_genes{end+1} = '5753';\n",
    "                        case 2232, knockdown_genes{end+1} = '57531';\n",
    "                        case 2233, knockdown_genes{end+1} = '57534';\n",
    "                        case 2234, knockdown_genes{end+1} = '5754';\n",
    "                        case 2235, knockdown_genes{end+1} = '57546';\n",
    "                        case 2236, knockdown_genes{end+1} = '57551';\n",
    "                        case 2237, knockdown_genes{end+1} = '57558';\n",
    "                        case 2238, knockdown_genes{end+1} = '57559';\n",
    "                        case 2239, knockdown_genes{end+1} = '57571';\n",
    "                        case 2240, knockdown_genes{end+1} = '57574';\n",
    "                        case 2241, knockdown_genes{end+1} = '57595';\n",
    "                        case 2242, knockdown_genes{end+1} = '57602';\n",
    "                        case 2243, knockdown_genes{end+1} = '57630';\n",
    "                        case 2244, knockdown_genes{end+1} = '57646';\n",
    "                        case 2245, knockdown_genes{end+1} = '57663';\n",
    "                        case 2246, knockdown_genes{end+1} = '57665';\n",
    "                        case 2247, knockdown_genes{end+1} = '57678';\n",
    "                        case 2248, knockdown_genes{end+1} = '57695';\n",
    "                        case 2249, knockdown_genes{end+1} = '5770';\n",
    "                        case 2250, knockdown_genes{end+1} = '57704';\n",
    "                        case 2251, knockdown_genes{end+1} = '5771';\n",
    "                        case 2252, knockdown_genes{end+1} = '57733';\n",
    "                        case 2253, knockdown_genes{end+1} = '5774';\n",
    "                        case 2254, knockdown_genes{end+1} = '5775';\n",
    "                        case 2255, knockdown_genes{end+1} = '57761';\n",
    "                        case 2256, knockdown_genes{end+1} = '5777';\n",
    "                        case 2257, knockdown_genes{end+1} = '5778';\n",
    "                        case 2258, knockdown_genes{end+1} = '57787';\n",
    "                        case 2259, knockdown_genes{end+1} = '5780';\n",
    "                        case 2260, knockdown_genes{end+1} = '57804';\n",
    "                        case 2261, knockdown_genes{end+1} = '5781';\n",
    "                        case 2262, knockdown_genes{end+1} = '57818';\n",
    "                        case 2263, knockdown_genes{end+1} = '5782';\n",
    "                        case 2264, knockdown_genes{end+1} = '57820';\n",
    "                        case 2265, knockdown_genes{end+1} = '5783';\n",
    "                        case 2266, knockdown_genes{end+1} = '57834';\n",
    "                        case 2267, knockdown_genes{end+1} = '57835';\n",
    "                        case 2268, knockdown_genes{end+1} = '5784';\n",
    "                        case 2269, knockdown_genes{end+1} = '5786';\n",
    "                        case 2270, knockdown_genes{end+1} = '5787';\n",
    "                        case 2271, knockdown_genes{end+1} = '5788';\n",
    "                        case 2272, knockdown_genes{end+1} = '5789';\n",
    "                        case 2273, knockdown_genes{end+1} = '5791';\n",
    "                        case 2274, knockdown_genes{end+1} = '5792';\n",
    "                        case 2275, knockdown_genes{end+1} = '5793';\n",
    "                        case 2276, knockdown_genes{end+1} = '5794';\n",
    "                        case 2277, knockdown_genes{end+1} = '5795';\n",
    "                        case 2278, knockdown_genes{end+1} = '5796';\n",
    "                        case 2279, knockdown_genes{end+1} = '5797';\n",
    "                        case 2280, knockdown_genes{end+1} = '5798';\n",
    "                        case 2281, knockdown_genes{end+1} = '5799';\n",
    "                        case 2282, knockdown_genes{end+1} = '580';\n",
    "                        case 2283, knockdown_genes{end+1} = '5800';\n",
    "                        case 2284, knockdown_genes{end+1} = '5801';\n",
    "                        case 2285, knockdown_genes{end+1} = '5802';\n",
    "                        case 2286, knockdown_genes{end+1} = '5803';\n",
    "                        case 2287, knockdown_genes{end+1} = '5805';\n",
    "                        case 2288, knockdown_genes{end+1} = '5810';\n",
    "                        case 2289, knockdown_genes{end+1} = '58190';\n",
    "                        case 2290, knockdown_genes{end+1} = '5825';\n",
    "                        case 2291, knockdown_genes{end+1} = '5831';\n",
    "                        case 2292, knockdown_genes{end+1} = '5832';\n",
    "                        case 2293, knockdown_genes{end+1} = '5833';\n",
    "                        case 2294, knockdown_genes{end+1} = '5834';\n",
    "                        case 2295, knockdown_genes{end+1} = '5836';\n",
    "                        case 2296, knockdown_genes{end+1} = '5837';\n",
    "                        case 2297, knockdown_genes{end+1} = '58478';\n",
    "                        case 2298, knockdown_genes{end+1} = '58508';\n",
    "                        case 2299, knockdown_genes{end+1} = '58510';\n",
    "                        case 2300, knockdown_genes{end+1} = '5859';\n",
    "                        case 2301, knockdown_genes{end+1} = '586';\n",
    "                        case 2302, knockdown_genes{end+1} = '5860';\n",
    "                        case 2303, knockdown_genes{end+1} = '587';\n",
    "                        case 2304, knockdown_genes{end+1} = '5871';\n",
    "                        case 2305, knockdown_genes{end+1} = '5883';\n",
    "                        case 2306, knockdown_genes{end+1} = '5891';\n",
    "                        case 2307, knockdown_genes{end+1} = '5894';\n",
    "                        case 2308, knockdown_genes{end+1} = '5896';\n",
    "                        case 2309, knockdown_genes{end+1} = '590';\n",
    "                        case 2310, knockdown_genes{end+1} = '5917';\n",
    "                        case 2311, knockdown_genes{end+1} = '59272';\n",
    "                        case 2312, knockdown_genes{end+1} = '593';\n",
    "                        case 2313, knockdown_genes{end+1} = '5930';\n",
    "                        case 2314, knockdown_genes{end+1} = '594';\n",
    "                        case 2315, knockdown_genes{end+1} = '5947';\n",
    "                        case 2316, knockdown_genes{end+1} = '5948';\n",
    "                        case 2317, knockdown_genes{end+1} = '5950';\n",
    "                        case 2318, knockdown_genes{end+1} = '5959';\n",
    "                        case 2319, knockdown_genes{end+1} = '5973';\n",
    "                        case 2320, knockdown_genes{end+1} = '5979';\n",
    "                        case 2321, knockdown_genes{end+1} = '5980';\n",
    "                        case 2322, knockdown_genes{end+1} = '5985';\n",
    "                        case 2323, knockdown_genes{end+1} = '6005';\n",
    "                        case 2324, knockdown_genes{end+1} = '6011';\n",
    "                        case 2325, knockdown_genes{end+1} = '6015';\n",
    "                        case 2326, knockdown_genes{end+1} = '60386';\n",
    "                        case 2327, knockdown_genes{end+1} = '6045';\n",
    "                        case 2328, knockdown_genes{end+1} = '6048';\n",
    "                        case 2329, knockdown_genes{end+1} = '60481';\n",
    "                        case 2330, knockdown_genes{end+1} = '60482';\n",
    "                        case 2331, knockdown_genes{end+1} = '60487';\n",
    "                        case 2332, knockdown_genes{end+1} = '60490';\n",
    "                        case 2333, knockdown_genes{end+1} = '60496';\n",
    "                        case 2334, knockdown_genes{end+1} = '60559';\n",
    "                        case 2335, knockdown_genes{end+1} = '60681';\n",
    "                        case 2336, knockdown_genes{end+1} = '6093';\n",
    "                        case 2337, knockdown_genes{end+1} = '6095';\n",
    "                        case 2338, knockdown_genes{end+1} = '6098';\n",
    "                        case 2339, knockdown_genes{end+1} = '6120';\n",
    "                        case 2340, knockdown_genes{end+1} = '6121';\n",
    "                        case 2341, knockdown_genes{end+1} = '613';\n",
    "                        case 2342, knockdown_genes{end+1} = '6184';\n",
    "                        case 2343, knockdown_genes{end+1} = '6185';\n",
    "                        case 2344, knockdown_genes{end+1} = '6195';\n",
    "                        case 2345, knockdown_genes{end+1} = '6196';\n",
    "                        case 2346, knockdown_genes{end+1} = '6197';\n",
    "                        case 2347, knockdown_genes{end+1} = '6198';\n",
    "                        case 2348, knockdown_genes{end+1} = '6199';\n",
    "                        case 2349, knockdown_genes{end+1} = '622';\n",
    "                        case 2350, knockdown_genes{end+1} = '6240';\n",
    "                        case 2351, knockdown_genes{end+1} = '6241';\n",
    "                        case 2352, knockdown_genes{end+1} = '6256';\n",
    "                        case 2353, knockdown_genes{end+1} = '6259';\n",
    "                        case 2354, knockdown_genes{end+1} = '6296';\n",
    "                        case 2355, knockdown_genes{end+1} = '6300';\n",
    "                        case 2356, knockdown_genes{end+1} = '6301';\n",
    "                        case 2357, knockdown_genes{end+1} = '6303';\n",
    "                        case 2358, knockdown_genes{end+1} = '6307';\n",
    "                        case 2359, knockdown_genes{end+1} = '6309';\n",
    "                        case 2360, knockdown_genes{end+1} = '6319';\n",
    "                        case 2361, knockdown_genes{end+1} = '6342';\n",
    "                        case 2362, knockdown_genes{end+1} = '635';\n",
    "                        case 2363, knockdown_genes{end+1} = '6382';\n",
    "                        case 2364, knockdown_genes{end+1} = '6389';\n",
    "                        case 2365, knockdown_genes{end+1} = '63891';\n",
    "                        case 2366, knockdown_genes{end+1} = '63893';\n",
    "                        case 2367, knockdown_genes{end+1} = '6390';\n",
    "                        case 2368, knockdown_genes{end+1} = '63904';\n",
    "                        case 2369, knockdown_genes{end+1} = '6391';\n",
    "                        case 2370, knockdown_genes{end+1} = '63917';\n",
    "                        case 2371, knockdown_genes{end+1} = '6392';\n",
    "                        case 2372, knockdown_genes{end+1} = '640';\n",
    "                        case 2373, knockdown_genes{end+1} = '64064';\n",
    "                        case 2374, knockdown_genes{end+1} = '64077';\n",
    "                        case 2375, knockdown_genes{end+1} = '64078';\n",
    "                        case 2376, knockdown_genes{end+1} = '64080';\n",
    "                        case 2377, knockdown_genes{end+1} = '64087';\n",
    "                        case 2378, knockdown_genes{end+1} = '64090';\n",
    "                        case 2379, knockdown_genes{end+1} = '64122';\n",
    "                        case 2380, knockdown_genes{end+1} = '64131';\n",
    "                        case 2381, knockdown_genes{end+1} = '64132';\n",
    "                        case 2382, knockdown_genes{end+1} = '641371';\n",
    "                        case 2383, knockdown_genes{end+1} = '641372';\n",
    "                        case 2384, knockdown_genes{end+1} = '64146';\n",
    "                        case 2385, knockdown_genes{end+1} = '6416';\n",
    "                        case 2386, knockdown_genes{end+1} = '641649';\n",
    "                        case 2387, knockdown_genes{end+1} = '64174';\n",
    "                        case 2388, knockdown_genes{end+1} = '64180';\n",
    "                        case 2389, knockdown_genes{end+1} = '6419';\n",
    "                        case 2390, knockdown_genes{end+1} = '64219';\n",
    "                        case 2391, knockdown_genes{end+1} = '64220';\n",
    "                        case 2392, knockdown_genes{end+1} = '64240';\n",
    "                        case 2393, knockdown_genes{end+1} = '64241';\n",
    "                        case 2394, knockdown_genes{end+1} = '64320';\n",
    "                        case 2395, knockdown_genes{end+1} = '64324';\n",
    "                        case 2396, knockdown_genes{end+1} = '64326';\n",
    "                        case 2397, knockdown_genes{end+1} = '64377';\n",
    "                        case 2398, knockdown_genes{end+1} = '644';\n",
    "                        case 2399, knockdown_genes{end+1} = '64419';\n",
    "                        case 2400, knockdown_genes{end+1} = '64425';\n",
    "                        case 2401, knockdown_genes{end+1} = '644378';\n",
    "                        case 2402, knockdown_genes{end+1} = '644591';\n",
    "                        case 2403, knockdown_genes{end+1} = '6446';\n",
    "                        case 2404, knockdown_genes{end+1} = '6448';\n",
    "                        case 2405, knockdown_genes{end+1} = '644974';\n",
    "                        case 2406, knockdown_genes{end+1} = '64499';\n",
    "                        case 2407, knockdown_genes{end+1} = '645';\n",
    "                        case 2408, knockdown_genes{end+1} = '645740';\n",
    "                        case 2409, knockdown_genes{end+1} = '64579';\n",
    "                        case 2410, knockdown_genes{end+1} = '64600';\n",
    "                        case 2411, knockdown_genes{end+1} = '646480';\n",
    "                        case 2412, knockdown_genes{end+1} = '646486';\n",
    "                        case 2413, knockdown_genes{end+1} = '6470';\n",
    "                        case 2414, knockdown_genes{end+1} = '64711';\n",
    "                        case 2415, knockdown_genes{end+1} = '64714';\n",
    "                        case 2416, knockdown_genes{end+1} = '6472';\n",
    "                        case 2417, knockdown_genes{end+1} = '64748';\n",
    "                        case 2418, knockdown_genes{end+1} = '64750';\n",
    "                        case 2419, knockdown_genes{end+1} = '64754';\n",
    "                        case 2420, knockdown_genes{end+1} = '6476';\n",
    "                        case 2421, knockdown_genes{end+1} = '64761';\n",
    "                        case 2422, knockdown_genes{end+1} = '6477';\n",
    "                        case 2423, knockdown_genes{end+1} = '64772';\n",
    "                        case 2424, knockdown_genes{end+1} = '6478';\n",
    "                        case 2425, knockdown_genes{end+1} = '64781';\n",
    "                        case 2426, knockdown_genes{end+1} = '6480';\n",
    "                        case 2427, knockdown_genes{end+1} = '64802';\n",
    "                        case 2428, knockdown_genes{end+1} = '64816';\n",
    "                        case 2429, knockdown_genes{end+1} = '6482';\n",
    "                        case 2430, knockdown_genes{end+1} = '6483';\n",
    "                        case 2431, knockdown_genes{end+1} = '64834';\n",
    "                        case 2432, knockdown_genes{end+1} = '6484';\n",
    "                        case 2433, knockdown_genes{end+1} = '64841';\n",
    "                        case 2434, knockdown_genes{end+1} = '64844';\n",
    "                        case 2435, knockdown_genes{end+1} = '64849';\n",
    "                        case 2436, knockdown_genes{end+1} = '64850';\n",
    "                        case 2437, knockdown_genes{end+1} = '64852';\n",
    "                        case 2438, knockdown_genes{end+1} = '64854';\n",
    "                        case 2439, knockdown_genes{end+1} = '6487';\n",
    "                        case 2440, knockdown_genes{end+1} = '6489';\n",
    "                        case 2441, knockdown_genes{end+1} = '64895';\n",
    "                        case 2442, knockdown_genes{end+1} = '649';\n",
    "                        case 2443, knockdown_genes{end+1} = '64900';\n",
    "                        case 2444, knockdown_genes{end+1} = '64902';\n",
    "                        case 2445, knockdown_genes{end+1} = '65010';\n",
    "                        case 2446, knockdown_genes{end+1} = '65018';\n",
    "                        case 2447, knockdown_genes{end+1} = '6505';\n",
    "                        case 2448, knockdown_genes{end+1} = '6506';\n",
    "                        case 2449, knockdown_genes{end+1} = '65061';\n",
    "                        case 2450, knockdown_genes{end+1} = '6507';\n",
    "                        case 2451, knockdown_genes{end+1} = '6508';\n",
    "                        case 2452, knockdown_genes{end+1} = '6509';\n",
    "                        case 2453, knockdown_genes{end+1} = '6510';\n",
    "                        case 2454, knockdown_genes{end+1} = '6511';\n",
    "                        case 2455, knockdown_genes{end+1} = '6512';\n",
    "                        case 2456, knockdown_genes{end+1} = '65125';\n",
    "                        case 2457, knockdown_genes{end+1} = '6513';\n",
    "                        case 2458, knockdown_genes{end+1} = '6514';\n",
    "                        case 2459, knockdown_genes{end+1} = '6515';\n",
    "                        case 2460, knockdown_genes{end+1} = '6517';\n",
    "                        case 2461, knockdown_genes{end+1} = '6518';\n",
    "                        case 2462, knockdown_genes{end+1} = '6519';\n",
    "                        case 2463, knockdown_genes{end+1} = '6520';\n",
    "                        case 2464, knockdown_genes{end+1} = '6521';\n",
    "                        case 2465, knockdown_genes{end+1} = '6522';\n",
    "                        case 2466, knockdown_genes{end+1} = '65220';\n",
    "                        case 2467, knockdown_genes{end+1} = '6523';\n",
    "                        case 2468, knockdown_genes{end+1} = '6524';\n",
    "                        case 2469, knockdown_genes{end+1} = '6526';\n",
    "                        case 2470, knockdown_genes{end+1} = '65263';\n",
    "                        case 2471, knockdown_genes{end+1} = '65264';\n",
    "                        case 2472, knockdown_genes{end+1} = '65266';\n",
    "                        case 2473, knockdown_genes{end+1} = '65267';\n",
    "                        case 2474, knockdown_genes{end+1} = '65268';\n",
    "                        case 2475, knockdown_genes{end+1} = '6528';\n",
    "                        case 2476, knockdown_genes{end+1} = '6529';\n",
    "                        case 2477, knockdown_genes{end+1} = '6530';\n",
    "                        case 2478, knockdown_genes{end+1} = '6531';\n",
    "                        case 2479, knockdown_genes{end+1} = '6532';\n",
    "                        case 2480, knockdown_genes{end+1} = '6533';\n",
    "                        case 2481, knockdown_genes{end+1} = '6534';\n",
    "                        case 2482, knockdown_genes{end+1} = '6535';\n",
    "                        case 2483, knockdown_genes{end+1} = '653505';\n",
    "                        case 2484, knockdown_genes{end+1} = '653598';\n",
    "                        case 2485, knockdown_genes{end+1} = '6536';\n",
    "                        case 2486, knockdown_genes{end+1} = '6538';\n",
    "                        case 2487, knockdown_genes{end+1} = '6539';\n",
    "                        case 2488, knockdown_genes{end+1} = '6540';\n",
    "                        case 2489, knockdown_genes{end+1} = '6541';\n",
    "                        case 2490, knockdown_genes{end+1} = '6542';\n",
    "                        case 2491, knockdown_genes{end+1} = '6543';\n",
    "                        case 2492, knockdown_genes{end+1} = '6546';\n",
    "                        case 2493, knockdown_genes{end+1} = '6547';\n",
    "                        case 2494, knockdown_genes{end+1} = '6548';\n",
    "                        case 2495, knockdown_genes{end+1} = '6549';\n",
    "                        case 2496, knockdown_genes{end+1} = '6550';\n",
    "                        case 2497, knockdown_genes{end+1} = '6553';\n",
    "                        case 2498, knockdown_genes{end+1} = '6554';\n",
    "                        case 2499, knockdown_genes{end+1} = '6555';\n",
    "                        case 2500, knockdown_genes{end+1} = '6556';\n",
    "                        case 2501, knockdown_genes{end+1} = '6557';\n",
    "                        case 2502, knockdown_genes{end+1} = '6558';\n",
    "                        case 2503, knockdown_genes{end+1} = '6559';\n",
    "                        case 2504, knockdown_genes{end+1} = '6560';\n",
    "                        case 2505, knockdown_genes{end+1} = '6561';\n",
    "                        case 2506, knockdown_genes{end+1} = '6563';\n",
    "                        case 2507, knockdown_genes{end+1} = '6564';\n",
    "                        case 2508, knockdown_genes{end+1} = '6565';\n",
    "                        case 2509, knockdown_genes{end+1} = '6566';\n",
    "                        case 2510, knockdown_genes{end+1} = '6567';\n",
    "                        case 2511, knockdown_genes{end+1} = '6568';\n",
    "                        case 2512, knockdown_genes{end+1} = '6569';\n",
    "                        case 2513, knockdown_genes{end+1} = '657';\n",
    "                        case 2514, knockdown_genes{end+1} = '6570';\n",
    "                        case 2515, knockdown_genes{end+1} = '6571';\n",
    "                        case 2516, knockdown_genes{end+1} = '6572';\n",
    "                        case 2517, knockdown_genes{end+1} = '6573';\n",
    "                        case 2518, knockdown_genes{end+1} = '6574';\n",
    "                        case 2519, knockdown_genes{end+1} = '6575';\n",
    "                        case 2520, knockdown_genes{end+1} = '6576';\n",
    "                        case 2521, knockdown_genes{end+1} = '6578';\n",
    "                        case 2522, knockdown_genes{end+1} = '6579';\n",
    "                        case 2523, knockdown_genes{end+1} = '658';\n",
    "                        case 2524, knockdown_genes{end+1} = '6580';\n",
    "                        case 2525, knockdown_genes{end+1} = '6581';\n",
    "                        case 2526, knockdown_genes{end+1} = '6582';\n",
    "                        case 2527, knockdown_genes{end+1} = '6583';\n",
    "                        case 2528, knockdown_genes{end+1} = '6584';\n",
    "                        case 2529, knockdown_genes{end+1} = '659';\n",
    "                        case 2530, knockdown_genes{end+1} = '6596';\n",
    "                        case 2531, knockdown_genes{end+1} = '65975';\n",
    "                        case 2532, knockdown_genes{end+1} = '65985';\n",
    "                        case 2533, knockdown_genes{end+1} = '660';\n",
    "                        case 2534, knockdown_genes{end+1} = '66002';\n",
    "                        case 2535, knockdown_genes{end+1} = '66035';\n",
    "                        case 2536, knockdown_genes{end+1} = '6604';\n",
    "                        case 2537, knockdown_genes{end+1} = '6609';\n",
    "                        case 2538, knockdown_genes{end+1} = '661';\n",
    "                        case 2539, knockdown_genes{end+1} = '6610';\n",
    "                        case 2540, knockdown_genes{end+1} = '6611';\n",
    "                        case 2541, knockdown_genes{end+1} = '6646';\n",
    "                        case 2542, knockdown_genes{end+1} = '6647';\n",
    "                        case 2543, knockdown_genes{end+1} = '6648';\n",
    "                        case 2544, knockdown_genes{end+1} = '6649';\n",
    "                        case 2545, knockdown_genes{end+1} = '6652';\n",
    "                        case 2546, knockdown_genes{end+1} = '6675';\n",
    "                        case 2547, knockdown_genes{end+1} = '669';\n",
    "                        case 2548, knockdown_genes{end+1} = '6697';\n",
    "                        case 2549, knockdown_genes{end+1} = '6713';\n",
    "                        case 2550, knockdown_genes{end+1} = '6714';\n",
    "                        case 2551, knockdown_genes{end+1} = '6715';\n",
    "                        case 2552, knockdown_genes{end+1} = '6716';\n",
    "                        case 2553, knockdown_genes{end+1} = '6718';\n",
    "                        case 2554, knockdown_genes{end+1} = '6720';\n",
    "                        case 2555, knockdown_genes{end+1} = '6723';\n",
    "                        case 2556, knockdown_genes{end+1} = '6725';\n",
    "                        case 2557, knockdown_genes{end+1} = '673';\n",
    "                        case 2558, knockdown_genes{end+1} = '6732';\n",
    "                        case 2559, knockdown_genes{end+1} = '6733';\n",
    "                        case 2560, knockdown_genes{end+1} = '6748';\n",
    "                        case 2561, knockdown_genes{end+1} = '6783';\n",
    "                        case 2562, knockdown_genes{end+1} = '6785';\n",
    "                        case 2563, knockdown_genes{end+1} = '6787';\n",
    "                        case 2564, knockdown_genes{end+1} = '6788';\n",
    "                        case 2565, knockdown_genes{end+1} = '6789';\n",
    "                        case 2566, knockdown_genes{end+1} = '6790';\n",
    "                        case 2567, knockdown_genes{end+1} = '6792';\n",
    "                        case 2568, knockdown_genes{end+1} = '6793';\n",
    "                        case 2569, knockdown_genes{end+1} = '6794';\n",
    "                        case 2570, knockdown_genes{end+1} = '6795';\n",
    "                        case 2571, knockdown_genes{end+1} = '6799';\n",
    "                        case 2572, knockdown_genes{end+1} = '6817';\n",
    "                        case 2573, knockdown_genes{end+1} = '6818';\n",
    "                        case 2574, knockdown_genes{end+1} = '6819';\n",
    "                        case 2575, knockdown_genes{end+1} = '682';\n",
    "                        case 2576, knockdown_genes{end+1} = '6820';\n",
    "                        case 2577, knockdown_genes{end+1} = '6821';\n",
    "                        case 2578, knockdown_genes{end+1} = '6822';\n",
    "                        case 2579, knockdown_genes{end+1} = '6833';\n",
    "                        case 2580, knockdown_genes{end+1} = '6839';\n",
    "                        case 2581, knockdown_genes{end+1} = '6850';\n",
    "                        case 2582, knockdown_genes{end+1} = '686';\n",
    "                        case 2583, knockdown_genes{end+1} = '6872';\n",
    "                        case 2584, knockdown_genes{end+1} = '6885';\n",
    "                        case 2585, knockdown_genes{end+1} = '6888';\n",
    "                        case 2586, knockdown_genes{end+1} = '6897';\n",
    "                        case 2587, knockdown_genes{end+1} = '6898';\n",
    "                        case 2588, knockdown_genes{end+1} = '6907';\n",
    "                        case 2589, knockdown_genes{end+1} = '6916';\n",
    "                        case 2590, knockdown_genes{end+1} = '699';\n",
    "                        case 2591, knockdown_genes{end+1} = '6999';\n",
    "                        case 2592, knockdown_genes{end+1} = '7001';\n",
    "                        case 2593, knockdown_genes{end+1} = '701';\n",
    "                        case 2594, knockdown_genes{end+1} = '7016';\n",
    "                        case 2595, knockdown_genes{end+1} = '7037';\n",
    "                        case 2596, knockdown_genes{end+1} = '7046';\n",
    "                        case 2597, knockdown_genes{end+1} = '7047';\n",
    "                        case 2598, knockdown_genes{end+1} = '7048';\n",
    "                        case 2599, knockdown_genes{end+1} = '7051';\n",
    "                        case 2600, knockdown_genes{end+1} = '7052';\n",
    "                        case 2601, knockdown_genes{end+1} = '7053';\n",
    "                        case 2602, knockdown_genes{end+1} = '7054';\n",
    "                        case 2603, knockdown_genes{end+1} = '7083';\n",
    "                        case 2604, knockdown_genes{end+1} = '7084';\n",
    "                        case 2605, knockdown_genes{end+1} = '7086';\n",
    "                        case 2606, knockdown_genes{end+1} = '7108';\n",
    "                        case 2607, knockdown_genes{end+1} = '7166';\n",
    "                        case 2608, knockdown_genes{end+1} = '7167';\n",
    "                        case 2609, knockdown_genes{end+1} = '7172';\n",
    "                        case 2610, knockdown_genes{end+1} = '7173';\n",
    "                        case 2611, knockdown_genes{end+1} = '7177';\n",
    "                        case 2612, knockdown_genes{end+1} = '7179';\n",
    "                        case 2613, knockdown_genes{end+1} = '7204';\n",
    "                        case 2614, knockdown_genes{end+1} = '7263';\n",
    "                        case 2615, knockdown_genes{end+1} = '7264';\n",
    "                        case 2616, knockdown_genes{end+1} = '7267';\n",
    "                        case 2617, knockdown_genes{end+1} = '7272';\n",
    "                        case 2618, knockdown_genes{end+1} = '7273';\n",
    "                        case 2619, knockdown_genes{end+1} = '728386';\n",
    "                        case 2620, knockdown_genes{end+1} = '728419';\n",
    "                        case 2621, knockdown_genes{end+1} = '728441';\n",
    "                        case 2622, knockdown_genes{end+1} = '728642';\n",
    "                        case 2623, knockdown_genes{end+1} = '729359';\n",
    "                        case 2624, knockdown_genes{end+1} = '7295';\n",
    "                        case 2625, knockdown_genes{end+1} = '7296';\n",
    "                        case 2626, knockdown_genes{end+1} = '7298';\n",
    "                        case 2627, knockdown_genes{end+1} = '7299';\n",
    "                        case 2628, knockdown_genes{end+1} = '7301';\n",
    "                        case 2629, knockdown_genes{end+1} = '730262';\n",
    "                        case 2630, knockdown_genes{end+1} = '7306';\n",
    "                        case 2631, knockdown_genes{end+1} = '7317';\n",
    "                        case 2632, knockdown_genes{end+1} = '7318';\n",
    "                        case 2633, knockdown_genes{end+1} = '7319';\n",
    "                        case 2634, knockdown_genes{end+1} = '7320';\n",
    "                        case 2635, knockdown_genes{end+1} = '7321';\n",
    "                        case 2636, knockdown_genes{end+1} = '7322';\n",
    "                        case 2637, knockdown_genes{end+1} = '7323';\n",
    "                        case 2638, knockdown_genes{end+1} = '7324';\n",
    "                        case 2639, knockdown_genes{end+1} = '7325';\n",
    "                        case 2640, knockdown_genes{end+1} = '7326';\n",
    "                        case 2641, knockdown_genes{end+1} = '7327';\n",
    "                        case 2642, knockdown_genes{end+1} = '7328';\n",
    "                        case 2643, knockdown_genes{end+1} = '7329';\n",
    "                        case 2644, knockdown_genes{end+1} = '7332';\n",
    "                        case 2645, knockdown_genes{end+1} = '7334';\n",
    "                        case 2646, knockdown_genes{end+1} = '7337';\n",
    "                        case 2647, knockdown_genes{end+1} = '7350';\n",
    "                        case 2648, knockdown_genes{end+1} = '7351';\n",
    "                        case 2649, knockdown_genes{end+1} = '7352';\n",
    "                        case 2650, knockdown_genes{end+1} = '7355';\n",
    "                        case 2651, knockdown_genes{end+1} = '7357';\n",
    "                        case 2652, knockdown_genes{end+1} = '7358';\n",
    "                        case 2653, knockdown_genes{end+1} = '7360';\n",
    "                        case 2654, knockdown_genes{end+1} = '7363';\n",
    "                        case 2655, knockdown_genes{end+1} = '7364';\n",
    "                        case 2656, knockdown_genes{end+1} = '7365';\n",
    "                        case 2657, knockdown_genes{end+1} = '7366';\n",
    "                        case 2658, knockdown_genes{end+1} = '7367';\n",
    "                        case 2659, knockdown_genes{end+1} = '7368';\n",
    "                        case 2660, knockdown_genes{end+1} = '7371';\n",
    "                        case 2661, knockdown_genes{end+1} = '7372';\n",
    "                        case 2662, knockdown_genes{end+1} = '7375';\n",
    "                        case 2663, knockdown_genes{end+1} = '7378';\n",
    "                        case 2664, knockdown_genes{end+1} = '7381';\n",
    "                        case 2665, knockdown_genes{end+1} = '7384';\n",
    "                        case 2666, knockdown_genes{end+1} = '7385';\n",
    "                        case 2667, knockdown_genes{end+1} = '7386';\n",
    "                        case 2668, knockdown_genes{end+1} = '7388';\n",
    "                        case 2669, knockdown_genes{end+1} = '7389';\n",
    "                        case 2670, knockdown_genes{end+1} = '7390';\n",
    "                        case 2671, knockdown_genes{end+1} = '7398';\n",
    "                        case 2672, knockdown_genes{end+1} = '7407';\n",
    "                        case 2673, knockdown_genes{end+1} = '7443';\n",
    "                        case 2674, knockdown_genes{end+1} = '7444';\n",
    "                        case 2675, knockdown_genes{end+1} = '7453';\n",
    "                        case 2676, knockdown_genes{end+1} = '7465';\n",
    "                        case 2677, knockdown_genes{end+1} = '7468';\n",
    "                        case 2678, knockdown_genes{end+1} = '747';\n",
    "                        case 2679, knockdown_genes{end+1} = '7498';\n",
    "                        case 2680, knockdown_genes{end+1} = '7525';\n",
    "                        case 2681, knockdown_genes{end+1} = '7535';\n",
    "                        case 2682, knockdown_genes{end+1} = '759';\n",
    "                        case 2683, knockdown_genes{end+1} = '760';\n",
    "                        case 2684, knockdown_genes{end+1} = '761';\n",
    "                        case 2685, knockdown_genes{end+1} = '762';\n",
    "                        case 2686, knockdown_genes{end+1} = '763';\n",
    "                        case 2687, knockdown_genes{end+1} = '765';\n",
    "                        case 2688, knockdown_genes{end+1} = '766';\n",
    "                        case 2689, knockdown_genes{end+1} = '767';\n",
    "                        case 2690, knockdown_genes{end+1} = '768';\n",
    "                        case 2691, knockdown_genes{end+1} = '7681';\n",
    "                        case 2692, knockdown_genes{end+1} = '771';\n",
    "                        case 2693, knockdown_genes{end+1} = '7779';\n",
    "                        case 2694, knockdown_genes{end+1} = '7786';\n",
    "                        case 2695, knockdown_genes{end+1} = '780';\n",
    "                        case 2696, knockdown_genes{end+1} = '7837';\n",
    "                        case 2697, knockdown_genes{end+1} = '7841';\n",
    "                        case 2698, knockdown_genes{end+1} = '7844';\n",
    "                        case 2699, knockdown_genes{end+1} = '7862';\n",
    "                        case 2700, knockdown_genes{end+1} = '7867';\n",
    "                        case 2701, knockdown_genes{end+1} = '7874';\n",
    "                        case 2702, knockdown_genes{end+1} = '788';\n",
    "                        case 2703, knockdown_genes{end+1} = '790';\n",
    "                        case 2704, knockdown_genes{end+1} = '79001';\n",
    "                        case 2705, knockdown_genes{end+1} = '79017';\n",
    "                        case 2706, knockdown_genes{end+1} = '7903';\n",
    "                        case 2707, knockdown_genes{end+1} = '79053';\n",
    "                        case 2708, knockdown_genes{end+1} = '79071';\n",
    "                        case 2709, knockdown_genes{end+1} = '79077';\n",
    "                        case 2710, knockdown_genes{end+1} = '79087';\n",
    "                        case 2711, knockdown_genes{end+1} = '7915';\n",
    "                        case 2712, knockdown_genes{end+1} = '79158';\n",
    "                        case 2713, knockdown_genes{end+1} = '79178';\n",
    "                        case 2714, knockdown_genes{end+1} = '7922';\n",
    "                        case 2715, knockdown_genes{end+1} = '7923';\n",
    "                        case 2716, knockdown_genes{end+1} = '79369';\n",
    "                        case 2717, knockdown_genes{end+1} = '79400';\n",
    "                        case 2718, knockdown_genes{end+1} = '7941';\n",
    "                        case 2719, knockdown_genes{end+1} = '79411';\n",
    "                        case 2720, knockdown_genes{end+1} = '79567';\n",
    "                        case 2721, knockdown_genes{end+1} = '7957';\n",
    "                        case 2722, knockdown_genes{end+1} = '79581';\n",
    "                        case 2723, knockdown_genes{end+1} = '79586';\n",
    "                        case 2724, knockdown_genes{end+1} = '79587';\n",
    "                        case 2725, knockdown_genes{end+1} = '79589';\n",
    "                        case 2726, knockdown_genes{end+1} = '79594';\n",
    "                        case 2727, knockdown_genes{end+1} = '79595';\n",
    "                        case 2728, knockdown_genes{end+1} = '79603';\n",
    "                        case 2729, knockdown_genes{end+1} = '79611';\n",
    "                        case 2730, knockdown_genes{end+1} = '79623';\n",
    "                        case 2731, knockdown_genes{end+1} = '79644';\n",
    "                        case 2732, knockdown_genes{end+1} = '79646';\n",
    "                        case 2733, knockdown_genes{end+1} = '79654';\n",
    "                        case 2734, knockdown_genes{end+1} = '79660';\n",
    "                        case 2735, knockdown_genes{end+1} = '79668';\n",
    "                        case 2736, knockdown_genes{end+1} = '79670';\n",
    "                        case 2737, knockdown_genes{end+1} = '79672';\n",
    "                        case 2738, knockdown_genes{end+1} = '79690';\n",
    "                        case 2739, knockdown_genes{end+1} = '79691';\n",
    "                        case 2740, knockdown_genes{end+1} = '79695';\n",
    "                        case 2741, knockdown_genes{end+1} = '79705';\n",
    "                        case 2742, knockdown_genes{end+1} = '79707';\n",
    "                        case 2743, knockdown_genes{end+1} = '79709';\n",
    "                        case 2744, knockdown_genes{end+1} = '79717';\n",
    "                        case 2745, knockdown_genes{end+1} = '79718';\n",
    "                        case 2746, knockdown_genes{end+1} = '79723';\n",
    "                        case 2747, knockdown_genes{end+1} = '79731';\n",
    "                        case 2748, knockdown_genes{end+1} = '79751';\n",
    "                        case 2749, knockdown_genes{end+1} = '79796';\n",
    "                        case 2750, knockdown_genes{end+1} = '79799';\n",
    "                        case 2751, knockdown_genes{end+1} = '79813';\n",
    "                        case 2752, knockdown_genes{end+1} = '79814';\n",
    "                        case 2753, knockdown_genes{end+1} = '79834';\n",
    "                        case 2754, knockdown_genes{end+1} = '79837';\n",
    "                        case 2755, knockdown_genes{end+1} = '79858';\n",
    "                        case 2756, knockdown_genes{end+1} = '79868';\n",
    "                        case 2757, knockdown_genes{end+1} = '79871';\n",
    "                        case 2758, knockdown_genes{end+1} = '79872';\n",
    "                        case 2759, knockdown_genes{end+1} = '79888';\n",
    "                        case 2760, knockdown_genes{end+1} = '79896';\n",
    "                        case 2761, knockdown_genes{end+1} = '79901';\n",
    "                        case 2762, knockdown_genes{end+1} = '7991';\n",
    "                        case 2763, knockdown_genes{end+1} = '7994';\n",
    "                        case 2764, knockdown_genes{end+1} = '79944';\n",
    "                        case 2765, knockdown_genes{end+1} = '79948';\n",
    "                        case 2766, knockdown_genes{end+1} = '79966';\n",
    "                        case 2767, knockdown_genes{end+1} = '79993';\n",
    "                        case 2768, knockdown_genes{end+1} = '80024';\n",
    "                        case 2769, knockdown_genes{end+1} = '80025';\n",
    "                        case 2770, knockdown_genes{end+1} = '80055';\n",
    "                        case 2771, knockdown_genes{end+1} = '801';\n",
    "                        case 2772, knockdown_genes{end+1} = '80142';\n",
    "                        case 2773, knockdown_genes{end+1} = '80146';\n",
    "                        case 2774, knockdown_genes{end+1} = '80150';\n",
    "                        case 2775, knockdown_genes{end+1} = '80162';\n",
    "                        case 2776, knockdown_genes{end+1} = '80168';\n",
    "                        case 2777, knockdown_genes{end+1} = '80201';\n",
    "                        case 2778, knockdown_genes{end+1} = '80205';\n",
    "                        case 2779, knockdown_genes{end+1} = '80221';\n",
    "                        case 2780, knockdown_genes{end+1} = '80222';\n",
    "                        case 2781, knockdown_genes{end+1} = '80235';\n",
    "                        case 2782, knockdown_genes{end+1} = '80270';\n",
    "                        case 2783, knockdown_genes{end+1} = '80271';\n",
    "                        case 2784, knockdown_genes{end+1} = '8029';\n",
    "                        case 2785, knockdown_genes{end+1} = '80308';\n",
    "                        case 2786, knockdown_genes{end+1} = '80324';\n",
    "                        case 2787, knockdown_genes{end+1} = '80339';\n",
    "                        case 2788, knockdown_genes{end+1} = '8034';\n",
    "                        case 2789, knockdown_genes{end+1} = '80347';\n",
    "                        case 2790, knockdown_genes{end+1} = '80351';\n",
    "                        case 2791, knockdown_genes{end+1} = '8050';\n",
    "                        case 2792, knockdown_genes{end+1} = '8061';\n",
    "                        case 2793, knockdown_genes{end+1} = '80704';\n",
    "                        case 2794, knockdown_genes{end+1} = '80724';\n",
    "                        case 2795, knockdown_genes{end+1} = '8073';\n",
    "                        case 2796, knockdown_genes{end+1} = '80824';\n",
    "                        case 2797, knockdown_genes{end+1} = '8085';\n",
    "                        case 2798, knockdown_genes{end+1} = '80854';\n",
    "                        case 2799, knockdown_genes{end+1} = '80895';\n",
    "                        case 2800, knockdown_genes{end+1} = '81031';\n",
    "                        case 2801, knockdown_genes{end+1} = '814';\n",
    "                        case 2802, knockdown_genes{end+1} = '8140';\n",
    "                        case 2803, knockdown_genes{end+1} = '81490';\n",
    "                        case 2804, knockdown_genes{end+1} = '815';\n",
    "                        case 2805, knockdown_genes{end+1} = '81537';\n",
    "                        case 2806, knockdown_genes{end+1} = '81539';\n",
    "                        case 2807, knockdown_genes{end+1} = '81559';\n",
    "                        case 2808, knockdown_genes{end+1} = '81579';\n",
    "                        case 2809, knockdown_genes{end+1} = '816';\n",
    "                        case 2810, knockdown_genes{end+1} = '81616';\n",
    "                        case 2811, knockdown_genes{end+1} = '81627';\n",
    "                        case 2812, knockdown_genes{end+1} = '81629';\n",
    "                        case 2813, knockdown_genes{end+1} = '81693';\n",
    "                        case 2814, knockdown_genes{end+1} = '817';\n",
    "                        case 2815, knockdown_genes{end+1} = '8170';\n",
    "                        case 2816, knockdown_genes{end+1} = '81788';\n",
    "                        case 2817, knockdown_genes{end+1} = '818';\n",
    "                        case 2818, knockdown_genes{end+1} = '81847';\n",
    "                        case 2819, knockdown_genes{end+1} = '81849';\n",
    "                        case 2820, knockdown_genes{end+1} = '81855';\n",
    "                        case 2821, knockdown_genes{end+1} = '81888';\n",
    "                        case 2822, knockdown_genes{end+1} = '81890';\n",
    "                        case 2823, knockdown_genes{end+1} = '81894';\n",
    "                        case 2824, knockdown_genes{end+1} = '8202';\n",
    "                        case 2825, knockdown_genes{end+1} = '8228';\n",
    "                        case 2826, knockdown_genes{end+1} = '8237';\n",
    "                        case 2827, knockdown_genes{end+1} = '8239';\n",
    "                        case 2828, knockdown_genes{end+1} = '8260';\n",
    "                        case 2829, knockdown_genes{end+1} = '8277';\n",
    "                        case 2830, knockdown_genes{end+1} = '8287';\n",
    "                        case 2831, knockdown_genes{end+1} = '8288';\n",
    "                        case 2832, knockdown_genes{end+1} = '8309';\n",
    "                        case 2833, knockdown_genes{end+1} = '8310';\n",
    "                        case 2834, knockdown_genes{end+1} = '8315';\n",
    "                        case 2835, knockdown_genes{end+1} = '8317';\n",
    "                        case 2836, knockdown_genes{end+1} = '833';\n",
    "                        case 2837, knockdown_genes{end+1} = '83401';\n",
    "                        case 2838, knockdown_genes{end+1} = '83440';\n",
    "                        case 2839, knockdown_genes{end+1} = '83468';\n",
    "                        case 2840, knockdown_genes{end+1} = '83539';\n",
    "                        case 2841, knockdown_genes{end+1} = '83549';\n",
    "                        case 2842, knockdown_genes{end+1} = '83594';\n",
    "                        case 2843, knockdown_genes{end+1} = '83666';\n",
    "                        case 2844, knockdown_genes{end+1} = '83697';\n",
    "                        case 2845, knockdown_genes{end+1} = '8372';\n",
    "                        case 2846, knockdown_genes{end+1} = '83732';\n",
    "                        case 2847, knockdown_genes{end+1} = '83733';\n",
    "                        case 2848, knockdown_genes{end+1} = '83844';\n",
    "                        case 2849, knockdown_genes{end+1} = '83852';\n",
    "                        case 2850, knockdown_genes{end+1} = '83884';\n",
    "                        case 2851, knockdown_genes{end+1} = '8394';\n",
    "                        case 2852, knockdown_genes{end+1} = '83942';\n",
    "                        case 2853, knockdown_genes{end+1} = '8395';\n",
    "                        case 2854, knockdown_genes{end+1} = '8396';\n",
    "                        case 2855, knockdown_genes{end+1} = '8398';\n",
    "                        case 2856, knockdown_genes{end+1} = '83983';\n",
    "                        case 2857, knockdown_genes{end+1} = '8399';\n",
    "                        case 2858, knockdown_genes{end+1} = '84002';\n",
    "                        case 2859, knockdown_genes{end+1} = '8402';\n",
    "                        case 2860, knockdown_genes{end+1} = '84076';\n",
    "                        case 2861, knockdown_genes{end+1} = '8408';\n",
    "                        case 2862, knockdown_genes{end+1} = '84101';\n",
    "                        case 2863, knockdown_genes{end+1} = '84102';\n",
    "                        case 2864, knockdown_genes{end+1} = '84105';\n",
    "                        case 2865, knockdown_genes{end+1} = '84129';\n",
    "                        case 2866, knockdown_genes{end+1} = '84132';\n",
    "                        case 2867, knockdown_genes{end+1} = '84148';\n",
    "                        case 2868, knockdown_genes{end+1} = '84152';\n",
    "                        case 2869, knockdown_genes{end+1} = '84172';\n",
    "                        case 2870, knockdown_genes{end+1} = '8418';\n",
    "                        case 2871, knockdown_genes{end+1} = '84188';\n",
    "                        case 2872, knockdown_genes{end+1} = '84196';\n",
    "                        case 2873, knockdown_genes{end+1} = '84231';\n",
    "                        case 2874, knockdown_genes{end+1} = '8424';\n",
    "                        case 2875, knockdown_genes{end+1} = '84254';\n",
    "                        case 2876, knockdown_genes{end+1} = '84264';\n",
    "                        case 2877, knockdown_genes{end+1} = '84265';\n",
    "                        case 2878, knockdown_genes{end+1} = '84274';\n",
    "                        case 2879, knockdown_genes{end+1} = '8428';\n",
    "                        case 2880, knockdown_genes{end+1} = '8435';\n",
    "                        case 2881, knockdown_genes{end+1} = '8443';\n",
    "                        case 2882, knockdown_genes{end+1} = '8444';\n",
    "                        case 2883, knockdown_genes{end+1} = '84444';\n",
    "                        case 2884, knockdown_genes{end+1} = '84446';\n",
    "                        case 2885, knockdown_genes{end+1} = '84447';\n",
    "                        case 2886, knockdown_genes{end+1} = '8445';\n",
    "                        case 2887, knockdown_genes{end+1} = '84451';\n",
    "                        case 2888, knockdown_genes{end+1} = '8446';\n",
    "                        case 2889, knockdown_genes{end+1} = '84513';\n",
    "                        case 2890, knockdown_genes{end+1} = '84532';\n",
    "                        case 2891, knockdown_genes{end+1} = '84569';\n",
    "                        case 2892, knockdown_genes{end+1} = '8459';\n",
    "                        case 2893, knockdown_genes{end+1} = '8460';\n",
    "                        case 2894, knockdown_genes{end+1} = '84618';\n",
    "                        case 2895, knockdown_genes{end+1} = '84640';\n",
    "                        case 2896, knockdown_genes{end+1} = '84647';\n",
    "                        case 2897, knockdown_genes{end+1} = '84649';\n",
    "                        case 2898, knockdown_genes{end+1} = '84669';\n",
    "                        case 2899, knockdown_genes{end+1} = '84676';\n",
    "                        case 2900, knockdown_genes{end+1} = '84679';\n",
    "                        case 2901, knockdown_genes{end+1} = '8468';\n",
    "                        case 2902, knockdown_genes{end+1} = '84693';\n",
    "                        case 2903, knockdown_genes{end+1} = '847';\n",
    "                        case 2904, knockdown_genes{end+1} = '84701';\n",
    "                        case 2905, knockdown_genes{end+1} = '84706';\n",
    "                        case 2906, knockdown_genes{end+1} = '84720';\n",
    "                        case 2907, knockdown_genes{end+1} = '8473';\n",
    "                        case 2908, knockdown_genes{end+1} = '84735';\n",
    "                        case 2909, knockdown_genes{end+1} = '84749';\n",
    "                        case 2910, knockdown_genes{end+1} = '84750';\n",
    "                        case 2911, knockdown_genes{end+1} = '84752';\n",
    "                        case 2912, knockdown_genes{end+1} = '8476';\n",
    "                        case 2913, knockdown_genes{end+1} = '84787';\n",
    "                        case 2914, knockdown_genes{end+1} = '8479';\n",
    "                        case 2915, knockdown_genes{end+1} = '84803';\n",
    "                        case 2916, knockdown_genes{end+1} = '84812';\n",
    "                        case 2917, knockdown_genes{end+1} = '84814';\n",
    "                        case 2918, knockdown_genes{end+1} = '84867';\n",
    "                        case 2919, knockdown_genes{end+1} = '84869';\n",
    "                        case 2920, knockdown_genes{end+1} = '84875';\n",
    "                        case 2921, knockdown_genes{end+1} = '84881';\n",
    "                        case 2922, knockdown_genes{end+1} = '84889';\n",
    "                        case 2923, knockdown_genes{end+1} = '84890';\n",
    "                        case 2924, knockdown_genes{end+1} = '8491';\n",
    "                        case 2925, knockdown_genes{end+1} = '84912';\n",
    "                        case 2926, knockdown_genes{end+1} = '84920';\n",
    "                        case 2927, knockdown_genes{end+1} = '84926';\n",
    "                        case 2928, knockdown_genes{end+1} = '8493';\n",
    "                        case 2929, knockdown_genes{end+1} = '84930';\n",
    "                        case 2930, knockdown_genes{end+1} = '84937';\n",
    "                        case 2931, knockdown_genes{end+1} = '84992';\n",
    "                        case 2932, knockdown_genes{end+1} = '85007';\n",
    "                        case 2933, knockdown_genes{end+1} = '8501';\n",
    "                        case 2934, knockdown_genes{end+1} = '85015';\n",
    "                        case 2935, knockdown_genes{end+1} = '8503';\n",
    "                        case 2936, knockdown_genes{end+1} = '8509';\n",
    "                        case 2937, knockdown_genes{end+1} = '8513';\n",
    "                        case 2938, knockdown_genes{end+1} = '8520';\n",
    "                        case 2939, knockdown_genes{end+1} = '8525';\n",
    "                        case 2940, knockdown_genes{end+1} = '8526';\n",
    "                        case 2941, knockdown_genes{end+1} = '8527';\n",
    "                        case 2942, knockdown_genes{end+1} = '8528';\n",
    "                        case 2943, knockdown_genes{end+1} = '8529';\n",
    "                        case 2944, knockdown_genes{end+1} = '85313';\n",
    "                        case 2945, knockdown_genes{end+1} = '85320';\n",
    "                        case 2946, knockdown_genes{end+1} = '8534';\n",
    "                        case 2947, knockdown_genes{end+1} = '8536';\n",
    "                        case 2948, knockdown_genes{end+1} = '85365';\n",
    "                        case 2949, knockdown_genes{end+1} = '85366';\n",
    "                        case 2950, knockdown_genes{end+1} = '8540';\n",
    "                        case 2951, knockdown_genes{end+1} = '85413';\n",
    "                        case 2952, knockdown_genes{end+1} = '85441';\n",
    "                        case 2953, knockdown_genes{end+1} = '85464';\n",
    "                        case 2954, knockdown_genes{end+1} = '85465';\n",
    "                        case 2955, knockdown_genes{end+1} = '85481';\n",
    "                        case 2956, knockdown_genes{end+1} = '8550';\n",
    "                        case 2957, knockdown_genes{end+1} = '8555';\n",
    "                        case 2958, knockdown_genes{end+1} = '8556';\n",
    "                        case 2959, knockdown_genes{end+1} = '8558';\n",
    "                        case 2960, knockdown_genes{end+1} = '8560';\n",
    "                        case 2961, knockdown_genes{end+1} = '8564';\n",
    "                        case 2962, knockdown_genes{end+1} = '8565';\n",
    "                        case 2963, knockdown_genes{end+1} = '8566';\n",
    "                        case 2964, knockdown_genes{end+1} = '8569';\n",
    "                        case 2965, knockdown_genes{end+1} = '8573';\n",
    "                        case 2966, knockdown_genes{end+1} = '8574';\n",
    "                        case 2967, knockdown_genes{end+1} = '8576';\n",
    "                        case 2968, knockdown_genes{end+1} = '8604';\n",
    "                        case 2969, knockdown_genes{end+1} = '8605';\n",
    "                        case 2970, knockdown_genes{end+1} = '8608';\n",
    "                        case 2971, knockdown_genes{end+1} = '8611';\n",
    "                        case 2972, knockdown_genes{end+1} = '8612';\n",
    "                        case 2973, knockdown_genes{end+1} = '8613';\n",
    "                        case 2974, knockdown_genes{end+1} = '8621';\n",
    "                        case 2975, knockdown_genes{end+1} = '8622';\n",
    "                        case 2976, knockdown_genes{end+1} = '8630';\n",
    "                        case 2977, knockdown_genes{end+1} = '8634';\n",
    "                        case 2978, knockdown_genes{end+1} = '8637';\n",
    "                        case 2979, knockdown_genes{end+1} = '8639';\n",
    "                        case 2980, knockdown_genes{end+1} = '8644';\n",
    "                        case 2981, knockdown_genes{end+1} = '8647';\n",
    "                        case 2982, knockdown_genes{end+1} = '8648';\n",
    "                        case 2983, knockdown_genes{end+1} = '8654';\n",
    "                        case 2984, knockdown_genes{end+1} = '8658';\n",
    "                        case 2985, knockdown_genes{end+1} = '8659';\n",
    "                        case 2986, knockdown_genes{end+1} = '8671';\n",
    "                        case 2987, knockdown_genes{end+1} = '8681';\n",
    "                        case 2988, knockdown_genes{end+1} = '8693';\n",
    "                        case 2989, knockdown_genes{end+1} = '8694';\n",
    "                        case 2990, knockdown_genes{end+1} = '8702';\n",
    "                        case 2991, knockdown_genes{end+1} = '8703';\n",
    "                        case 2992, knockdown_genes{end+1} = '8704';\n",
    "                        case 2993, knockdown_genes{end+1} = '8705';\n",
    "                        case 2994, knockdown_genes{end+1} = '8706';\n",
    "                        case 2995, knockdown_genes{end+1} = '8707';\n",
    "                        case 2996, knockdown_genes{end+1} = '8708';\n",
    "                        case 2997, knockdown_genes{end+1} = '8711';\n",
    "                        case 2998, knockdown_genes{end+1} = '8714';\n",
    "                        case 2999, knockdown_genes{end+1} = '87178';\n",
    "                        case 3000, knockdown_genes{end+1} = '8722';\n",
    "                        case 3001, knockdown_genes{end+1} = '873';\n",
    "                        case 3002, knockdown_genes{end+1} = '8731';\n",
    "                        case 3003, knockdown_genes{end+1} = '8732';\n",
    "                        case 3004, knockdown_genes{end+1} = '8733';\n",
    "                        case 3005, knockdown_genes{end+1} = '8737';\n",
    "                        case 3006, knockdown_genes{end+1} = '874';\n",
    "                        case 3007, knockdown_genes{end+1} = '875';\n",
    "                        case 3008, knockdown_genes{end+1} = '8760';\n",
    "                        case 3009, knockdown_genes{end+1} = '8767';\n",
    "                        case 3010, knockdown_genes{end+1} = '8776';\n",
    "                        case 3011, knockdown_genes{end+1} = '8780';\n",
    "                        case 3012, knockdown_genes{end+1} = '8781';\n",
    "                        case 3013, knockdown_genes{end+1} = '8789';\n",
    "                        case 3014, knockdown_genes{end+1} = '8790';\n",
    "                        case 3015, knockdown_genes{end+1} = '8798';\n",
    "                        case 3016, knockdown_genes{end+1} = '8800';\n",
    "                        case 3017, knockdown_genes{end+1} = '8801';\n",
    "                        case 3018, knockdown_genes{end+1} = '8802';\n",
    "                        case 3019, knockdown_genes{end+1} = '8803';\n",
    "                        case 3020, knockdown_genes{end+1} = '8813';\n",
    "                        case 3021, knockdown_genes{end+1} = '8814';\n",
    "                        case 3022, knockdown_genes{end+1} = '8818';\n",
    "                        case 3023, knockdown_genes{end+1} = '8821';\n",
    "                        case 3024, knockdown_genes{end+1} = '8824';\n",
    "                        case 3025, knockdown_genes{end+1} = '883';\n",
    "                        case 3026, knockdown_genes{end+1} = '8833';\n",
    "                        case 3027, knockdown_genes{end+1} = '8836';\n",
    "                        case 3028, knockdown_genes{end+1} = '8841';\n",
    "                        case 3029, knockdown_genes{end+1} = '8850';\n",
    "                        case 3030, knockdown_genes{end+1} = '8854';\n",
    "                        case 3031, knockdown_genes{end+1} = '8859';\n",
    "                        case 3032, knockdown_genes{end+1} = '8867';\n",
    "                        case 3033, knockdown_genes{end+1} = '8869';\n",
    "                        case 3034, knockdown_genes{end+1} = '8871';\n",
    "                        case 3035, knockdown_genes{end+1} = '8875';\n",
    "                        case 3036, knockdown_genes{end+1} = '8876';\n",
    "                        case 3037, knockdown_genes{end+1} = '8877';\n",
    "                        case 3038, knockdown_genes{end+1} = '8879';\n",
    "                        case 3039, knockdown_genes{end+1} = '8884';\n",
    "                        case 3040, knockdown_genes{end+1} = '8897';\n",
    "                        case 3041, knockdown_genes{end+1} = '8898';\n",
    "                        case 3042, knockdown_genes{end+1} = '8899';\n",
    "                        case 3043, knockdown_genes{end+1} = '8908';\n",
    "                        case 3044, knockdown_genes{end+1} = '8916';\n",
    "                        case 3045, knockdown_genes{end+1} = '8924';\n",
    "                        case 3046, knockdown_genes{end+1} = '8925';\n",
    "                        case 3047, knockdown_genes{end+1} = '8942';\n",
    "                        case 3048, knockdown_genes{end+1} = '8972';\n",
    "                        case 3049, knockdown_genes{end+1} = '8974';\n",
    "                        case 3050, knockdown_genes{end+1} = '8975';\n",
    "                        case 3051, knockdown_genes{end+1} = '89792';\n",
    "                        case 3052, knockdown_genes{end+1} = '8985';\n",
    "                        case 3053, knockdown_genes{end+1} = '8986';\n",
    "                        case 3054, knockdown_genes{end+1} = '89869';\n",
    "                        case 3055, knockdown_genes{end+1} = '89872';\n",
    "                        case 3056, knockdown_genes{end+1} = '89874';\n",
    "                        case 3057, knockdown_genes{end+1} = '89910';\n",
    "                        case 3058, knockdown_genes{end+1} = '8992';\n",
    "                        case 3059, knockdown_genes{end+1} = '8993';\n",
    "                        case 3060, knockdown_genes{end+1} = '8997';\n",
    "                        case 3061, knockdown_genes{end+1} = '8999';\n",
    "                        case 3062, knockdown_genes{end+1} = '9';\n",
    "                        case 3063, knockdown_genes{end+1} = '90';\n",
    "                        case 3064, knockdown_genes{end+1} = '9016';\n",
    "                        case 3065, knockdown_genes{end+1} = '90161';\n",
    "                        case 3066, knockdown_genes{end+1} = '9020';\n",
    "                        case 3067, knockdown_genes{end+1} = '9023';\n",
    "                        case 3068, knockdown_genes{end+1} = '9024';\n",
    "                        case 3069, knockdown_genes{end+1} = '9025';\n",
    "                        case 3070, knockdown_genes{end+1} = '90381';\n",
    "                        case 3071, knockdown_genes{end+1} = '9039';\n",
    "                        case 3072, knockdown_genes{end+1} = '9040';\n",
    "                        case 3073, knockdown_genes{end+1} = '90423';\n",
    "                        case 3074, knockdown_genes{end+1} = '9045';\n",
    "                        case 3075, knockdown_genes{end+1} = '9056';\n",
    "                        case 3076, knockdown_genes{end+1} = '9057';\n",
    "                        case 3077, knockdown_genes{end+1} = '9058';\n",
    "                        case 3078, knockdown_genes{end+1} = '9060';\n",
    "                        case 3079, knockdown_genes{end+1} = '9061';\n",
    "                        case 3080, knockdown_genes{end+1} = '9064';\n",
    "                        case 3081, knockdown_genes{end+1} = '90678';\n",
    "                        case 3082, knockdown_genes{end+1} = '90701';\n",
    "                        case 3083, knockdown_genes{end+1} = '9085';\n",
    "                        case 3084, knockdown_genes{end+1} = '9088';\n",
    "                        case 3085, knockdown_genes{end+1} = '9091';\n",
    "                        case 3086, knockdown_genes{end+1} = '9097';\n",
    "                        case 3087, knockdown_genes{end+1} = '9098';\n",
    "                        case 3088, knockdown_genes{end+1} = '9099';\n",
    "                        case 3089, knockdown_genes{end+1} = '91';\n",
    "                        case 3090, knockdown_genes{end+1} = '9100';\n",
    "                        case 3091, knockdown_genes{end+1} = '9101';\n",
    "                        case 3092, knockdown_genes{end+1} = '91012';\n",
    "                        case 3093, knockdown_genes{end+1} = '9107';\n",
    "                        case 3094, knockdown_genes{end+1} = '9108';\n",
    "                        case 3095, knockdown_genes{end+1} = '9113';\n",
    "                        case 3096, knockdown_genes{end+1} = '9114';\n",
    "                        case 3097, knockdown_genes{end+1} = '91179';\n",
    "                        case 3098, knockdown_genes{end+1} = '91227';\n",
    "                        case 3099, knockdown_genes{end+1} = '9123';\n",
    "                        case 3100, knockdown_genes{end+1} = '9128';\n",
    "                        case 3101, knockdown_genes{end+1} = '91373';\n",
    "                        case 3102, knockdown_genes{end+1} = '91445';\n",
    "                        case 3103, knockdown_genes{end+1} = '9148';\n",
    "                        case 3104, knockdown_genes{end+1} = '9149';\n",
    "                        case 3105, knockdown_genes{end+1} = '9150';\n",
    "                        case 3106, knockdown_genes{end+1} = '9152';\n",
    "                        case 3107, knockdown_genes{end+1} = '9153';\n",
    "                        case 3108, knockdown_genes{end+1} = '9154';\n",
    "                        case 3109, knockdown_genes{end+1} = '9162';\n",
    "                        case 3110, knockdown_genes{end+1} = '9167';\n",
    "                        case 3111, knockdown_genes{end+1} = '91703';\n",
    "                        case 3112, knockdown_genes{end+1} = '91734';\n",
    "                        case 3113, knockdown_genes{end+1} = '9175';\n",
    "                        case 3114, knockdown_genes{end+1} = '91754';\n",
    "                        case 3115, knockdown_genes{end+1} = '91807';\n",
    "                        case 3116, knockdown_genes{end+1} = '9187';\n",
    "                        case 3117, knockdown_genes{end+1} = '9194';\n",
    "                        case 3118, knockdown_genes{end+1} = '9197';\n",
    "                        case 3119, knockdown_genes{end+1} = '92';\n",
    "                        case 3120, knockdown_genes{end+1} = '9200';\n",
    "                        case 3121, knockdown_genes{end+1} = '9201';\n",
    "                        case 3122, knockdown_genes{end+1} = '92086';\n",
    "                        case 3123, knockdown_genes{end+1} = '9212';\n",
    "                        case 3124, knockdown_genes{end+1} = '92235';\n",
    "                        case 3125, knockdown_genes{end+1} = '9227';\n",
    "                        case 3126, knockdown_genes{end+1} = '92370';\n",
    "                        case 3127, knockdown_genes{end+1} = '9245';\n",
    "                        case 3128, knockdown_genes{end+1} = '9246';\n",
    "                        case 3129, knockdown_genes{end+1} = '92483';\n",
    "                        case 3130, knockdown_genes{end+1} = '9249';\n",
    "                        case 3131, knockdown_genes{end+1} = '9252';\n",
    "                        case 3132, knockdown_genes{end+1} = '92579';\n",
    "                        case 3133, knockdown_genes{end+1} = '9261';\n",
    "                        case 3134, knockdown_genes{end+1} = '9262';\n",
    "                        case 3135, knockdown_genes{end+1} = '9263';\n",
    "                        case 3136, knockdown_genes{end+1} = '92745';\n",
    "                        case 3137, knockdown_genes{end+1} = '9276';\n",
    "                        case 3138, knockdown_genes{end+1} = '92912';\n",
    "                        case 3139, knockdown_genes{end+1} = '92935';\n",
    "                        case 3140, knockdown_genes{end+1} = '9296';\n",
    "                        case 3141, knockdown_genes{end+1} = '92979';\n",
    "                        case 3142, knockdown_genes{end+1} = '93';\n",
    "                        case 3143, knockdown_genes{end+1} = '93010';\n",
    "                        case 3144, knockdown_genes{end+1} = '93034';\n",
    "                        case 3145, knockdown_genes{end+1} = '93100';\n",
    "                        case 3146, knockdown_genes{end+1} = '93183';\n",
    "                        case 3147, knockdown_genes{end+1} = '9320';\n",
    "                        case 3148, knockdown_genes{end+1} = '9329';\n",
    "                        case 3149, knockdown_genes{end+1} = '9331';\n",
    "                        case 3150, knockdown_genes{end+1} = '9333';\n",
    "                        case 3151, knockdown_genes{end+1} = '9334';\n",
    "                        case 3152, knockdown_genes{end+1} = '9344';\n",
    "                        case 3153, knockdown_genes{end+1} = '9348';\n",
    "                        case 3154, knockdown_genes{end+1} = '93492';\n",
    "                        case 3155, knockdown_genes{end+1} = '93517';\n",
    "                        case 3156, knockdown_genes{end+1} = '9354';\n",
    "                        case 3157, knockdown_genes{end+1} = '9356';\n",
    "                        case 3158, knockdown_genes{end+1} = '9360';\n",
    "                        case 3159, knockdown_genes{end+1} = '9365';\n",
    "                        case 3160, knockdown_genes{end+1} = '93650';\n",
    "                        case 3161, knockdown_genes{end+1} = '9374';\n",
    "                        case 3162, knockdown_genes{end+1} = '9376';\n",
    "                        case 3163, knockdown_genes{end+1} = '9377';\n",
    "                        case 3164, knockdown_genes{end+1} = '9380';\n",
    "                        case 3165, knockdown_genes{end+1} = '9388';\n",
    "                        case 3166, knockdown_genes{end+1} = '9390';\n",
    "                        case 3167, knockdown_genes{end+1} = '9394';\n",
    "                        case 3168, knockdown_genes{end+1} = '9397';\n",
    "                        case 3169, knockdown_genes{end+1} = '93979';\n",
    "                        case 3170, knockdown_genes{end+1} = '94';\n",
    "                        case 3171, knockdown_genes{end+1} = '94005';\n",
    "                        case 3172, knockdown_genes{end+1} = '94033';\n",
    "                        case 3173, knockdown_genes{end+1} = '9415';\n",
    "                        case 3174, knockdown_genes{end+1} = '9420';\n",
    "                        case 3175, knockdown_genes{end+1} = '9426';\n",
    "                        case 3176, knockdown_genes{end+1} = '9429';\n",
    "                        case 3177, knockdown_genes{end+1} = '9435';\n",
    "                        case 3178, knockdown_genes{end+1} = '9446';\n",
    "                        case 3179, knockdown_genes{end+1} = '9451';\n",
    "                        case 3180, knockdown_genes{end+1} = '9453';\n",
    "                        case 3181, knockdown_genes{end+1} = '9468';\n",
    "                        case 3182, knockdown_genes{end+1} = '9469';\n",
    "                        case 3183, knockdown_genes{end+1} = '9475';\n",
    "                        case 3184, knockdown_genes{end+1} = '9476';\n",
    "                        case 3185, knockdown_genes{end+1} = '948';\n",
    "                        case 3186, knockdown_genes{end+1} = '9481';\n",
    "                        case 3187, knockdown_genes{end+1} = '9486';\n",
    "                        case 3188, knockdown_genes{end+1} = '9487';\n",
    "                        case 3189, knockdown_genes{end+1} = '9488';\n",
    "                        case 3190, knockdown_genes{end+1} = '9489';\n",
    "                        case 3191, knockdown_genes{end+1} = '949';\n",
    "                        case 3192, knockdown_genes{end+1} = '9497';\n",
    "                        case 3193, knockdown_genes{end+1} = '9498';\n",
    "                        case 3194, knockdown_genes{end+1} = '95';\n",
    "                        case 3195, knockdown_genes{end+1} = '9514';\n",
    "                        case 3196, knockdown_genes{end+1} = '9517';\n",
    "                        case 3197, knockdown_genes{end+1} = '952';\n",
    "                        case 3198, knockdown_genes{end+1} = '9524';\n",
    "                        case 3199, knockdown_genes{end+1} = '953';\n",
    "                        case 3200, knockdown_genes{end+1} = '9533';\n",
    "                        case 3201, knockdown_genes{end+1} = '9536';\n",
    "                        case 3202, knockdown_genes{end+1} = '954';\n",
    "                        case 3203, knockdown_genes{end+1} = '955';\n",
    "                        case 3204, knockdown_genes{end+1} = '9550';\n",
    "                        case 3205, knockdown_genes{end+1} = '9551';\n",
    "                        case 3206, knockdown_genes{end+1} = '956';\n",
    "                        case 3207, knockdown_genes{end+1} = '9562';\n",
    "                        case 3208, knockdown_genes{end+1} = '9563';\n",
    "                        case 3209, knockdown_genes{end+1} = '957';\n",
    "                        case 3210, knockdown_genes{end+1} = '9572';\n",
    "                        case 3211, knockdown_genes{end+1} = '9575';\n",
    "                        case 3212, knockdown_genes{end+1} = '9578';\n",
    "                        case 3213, knockdown_genes{end+1} = '9583';\n",
    "                        case 3214, knockdown_genes{end+1} = '9588';\n",
    "                        case 3215, knockdown_genes{end+1} = '9601';\n",
    "                        case 3216, knockdown_genes{end+1} = '9604';\n",
    "                        case 3217, knockdown_genes{end+1} = '9611';\n",
    "                        case 3218, knockdown_genes{end+1} = '9612';\n",
    "                        case 3219, knockdown_genes{end+1} = '9615';\n",
    "                        case 3220, knockdown_genes{end+1} = '9625';\n",
    "                        case 3221, knockdown_genes{end+1} = '9641';\n",
    "                        case 3222, knockdown_genes{end+1} = '9647';\n",
    "                        case 3223, knockdown_genes{end+1} = '9651';\n",
    "                        case 3224, knockdown_genes{end+1} = '9653';\n",
    "                        case 3225, knockdown_genes{end+1} = '9663';\n",
    "                        case 3226, knockdown_genes{end+1} = '9666';\n",
    "                        case 3227, knockdown_genes{end+1} = '96764';\n",
    "                        case 3228, knockdown_genes{end+1} = '9677';\n",
    "                        case 3229, knockdown_genes{end+1} = '9690';\n",
    "                        case 3230, knockdown_genes{end+1} = '97';\n",
    "                        case 3231, knockdown_genes{end+1} = '9736';\n",
    "                        case 3232, knockdown_genes{end+1} = '9739';\n",
    "                        case 3233, knockdown_genes{end+1} = '9748';\n",
    "                        case 3234, knockdown_genes{end+1} = '9757';\n",
    "                        case 3235, knockdown_genes{end+1} = '978';\n",
    "                        case 3236, knockdown_genes{end+1} = '9781';\n",
    "                        case 3237, knockdown_genes{end+1} = '9789';\n",
    "                        case 3238, knockdown_genes{end+1} = '9791';\n",
    "                        case 3239, knockdown_genes{end+1} = '98';\n",
    "                        case 3240, knockdown_genes{end+1} = '9807';\n",
    "                        case 3241, knockdown_genes{end+1} = '9810';\n",
    "                        case 3242, knockdown_genes{end+1} = '9829';\n",
    "                        case 3243, knockdown_genes{end+1} = '983';\n",
    "                        case 3244, knockdown_genes{end+1} = '9833';\n",
    "                        case 3245, knockdown_genes{end+1} = '984';\n",
    "                        case 3246, knockdown_genes{end+1} = '9843';\n",
    "                        case 3247, knockdown_genes{end+1} = '9867';\n",
    "                        case 3248, knockdown_genes{end+1} = '9869';\n",
    "                        case 3249, knockdown_genes{end+1} = '9874';\n",
    "                        case 3250, knockdown_genes{end+1} = '9890';\n",
    "                        case 3251, knockdown_genes{end+1} = '9896';\n",
    "                        case 3252, knockdown_genes{end+1} = '9917';\n",
    "                        case 3253, knockdown_genes{end+1} = '9924';\n",
    "                        case 3254, knockdown_genes{end+1} = '993';\n",
    "                        case 3255, knockdown_genes{end+1} = '994';\n",
    "                        case 3256, knockdown_genes{end+1} = '9942';\n",
    "                        case 3257, knockdown_genes{end+1} = '9943';\n",
    "                        case 3258, knockdown_genes{end+1} = '9945';\n",
    "                        case 3259, knockdown_genes{end+1} = '995';\n",
    "                        case 3260, knockdown_genes{end+1} = '9951';\n",
    "                        case 3261, knockdown_genes{end+1} = '9953';\n",
    "                        case 3262, knockdown_genes{end+1} = '9955';\n",
    "                        case 3263, knockdown_genes{end+1} = '9956';\n",
    "                        case 3264, knockdown_genes{end+1} = '9957';\n",
    "                        case 3265, knockdown_genes{end+1} = '9958';\n",
    "                        case 3266, knockdown_genes{end+1} = '9960';\n",
    "                        case 3267, knockdown_genes{end+1} = '9962';\n",
    "                        case 3268, knockdown_genes{end+1} = '9963';\n",
    "                        case 3269, knockdown_genes{end+1} = '997';\n",
    "                        case 3270, knockdown_genes{end+1} = '9990';\n",
    "                    end\n",
    "\n",
    "                    % knockdown fraction\n",
    "                    value = raw(row,2);\n",
    "                    value = value{1};\n",
    "                    if ~isnan(value)\n",
    "                        if isfloat(value)\n",
    "                            if (value >= 0) && (value <= 1)\n",
    "                                knockdown_fractions(end+1) = value;\n",
    "                            else\n",
    "                                error('ERROR - Gene Knockdown - Knockdown fractions must either be >= 0 and <= 1')\n",
    "                            end\n",
    "                        else\n",
    "                            error('ERROR - Gene Knockdown - Knockdown fractions must either be >= 0 and <= 1')\n",
    "                        end\n",
    "                    else\n",
    "                        error('ERROR - Gene Knockdown - Must enter knockdown fraction for selected gene')\n",
    "                    end\n",
    "\n",
    "                else\n",
    "                    error('ERROR - Gene Knockdown - Select values must either be X or blank')\n",
    "                end\n",
    "            else\n",
    "                error('ERROR - Gene Knockdown - Select values must either be X or blank')\n",
    "            end\n",
    "        end\n",
    "    end\n",
    "end"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Load Model"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%file load_model.m\n",
    "\n",
    "function [model,concentration_mets,concentration_values,constraints_fraction_id] = load_model(constraints_proliferation,constraints_proliferation_type,constraints_proliferation_value,constraints_value_type,constraints_value_value,constraints_value_reaction,constraints_value_count,constraints_fraction_type,constraints_fraction_value,constraints_fraction_reaction,constraints_fraction_count,constraints_maxmulti_weight,constraints_maxmulti_reaction,constraints_maxmulti_count,objective_function_direction,objective_function_weight,objective_function_reaction)\n",
    "\n",
    "    % load model file\n",
    "    load('../data/recon/recon3d_qflux.mat')\n",
    "    \n",
    "    % initialize concentrations\n",
    "    concentration_mets = {};\n",
    "    concentration_values = [];\n",
    "\n",
    "    % compartment ph's\n",
    "    concentration_mets = [concentration_mets,'h[c]','h[e]','h[g]','h[l]','h[m]','h[n]','h[r]','h[x]','h[i]'];\n",
    "    concentration_values = [concentration_values,7.2,7.4,6.35,5.5,8,7.2,7.2,7,7];\n",
    "    \n",
    "    % implement model.c_fraction and model.c_obj to keep track of reactions that will have to maximize to add constraints\n",
    "    model.c_fraction = zeros(size(model.c));\n",
    "    model.c_maxmulti = zeros(size(model.c));\n",
    "    model.c_obj = zeros(size(model.c));\n",
    "\n",
    "    % implement custom value constraints\n",
    "    for i = 1:constraints_value_count\n",
    "\n",
    "        % if custom reaction\n",
    "        if (contains(constraints_value_reaction{i},'-->')) || (contains(constraints_value_reaction{i},'<=>'))\n",
    "\n",
    "            % initialize new reaction\n",
    "            model.rxns{end+1} = sprintf('custom_value_%d',i);\n",
    "            model.S(:,end+1) = zeros(length(model.mets),1);\n",
    "            model.c(end+1) = 0;\n",
    "            model.c_fraction(end+1) = 0;\n",
    "            model.c_maxmulti(end+1) = 0;\n",
    "            model.c_obj(end+1) = 0;\n",
    "            model.rxnGeneMat(end+1,:) = zeros(1,length(model.genes));\n",
    "            model.rules{end+1} = '';\n",
    "            model.grRules{end+1} = '';\n",
    "            model.subSystems{end+1} = 'Custom Constraint';\n",
    "            model.rxnNames{end+1} = sprintf('Custom Value Constraint %d',i);\n",
    "            model.rxnKEGGID{end+1} = '';\n",
    "            model.rxnKeggOrthology{end+1} = '';\n",
    "            model.rxnConfidenceScores(end+1) = 0;\n",
    "            model.rxnReferences{end+1} = '';\n",
    "            model.rxnECNumbers{end+1} = '';\n",
    "            model.rxnNotes{end+1} = '';\n",
    "            model.rxnCOG{end+1} = '';\n",
    "            model.rxnReconMap{end+1} = '';\n",
    "\n",
    "            % irreversible reaction\n",
    "            if (contains(constraints_value_reaction{i},'-->'))\n",
    "\n",
    "                % parse reactants\n",
    "                metstring = strsplit(strtrim(constraints_value_reaction{i}(1:(strfind(constraints_value_reaction{i},'-->')-1))),' + ','CollapseDelimiters',true);\n",
    "                for j = 1:length(metstring)\n",
    "                    if ~strcmp(metstring{j},'')\n",
    "                        metstring_ind = strsplit(metstring{j},' ');\n",
    "\n",
    "                        % metabolite\n",
    "                        if ~any(strcmp(model.mets,metstring_ind{2}))\n",
    "                            error('ERROR - Constraints - Custom - Unrecognized metabolite %s',metstring_ind{2})\n",
    "                        end\n",
    "\n",
    "                        % stoichiometry\n",
    "                        [num, status] = str2num(metstring_ind{1});\n",
    "                        if status == 1\n",
    "                            model.S(strcmp(model.mets,metstring_ind{2}),end) = -num;\n",
    "                        else\n",
    "                            error('ERROR - Constraints - Custom - Unrecognized stoichiometry %s',metstring_ind{1})\n",
    "                        end   \n",
    "                    end\n",
    "                end\n",
    "\n",
    "                % parse products\n",
    "                metstring = strsplit(strtrim(constraints_value_reaction{i}((strfind(constraints_value_reaction{i},'-->')+3):end)),' + ','CollapseDelimiters',true);\n",
    "                for j = 1:length(metstring)\n",
    "                    if ~strcmp(metstring{j},'')\n",
    "                        metstring_ind = strsplit(metstring{j},' ');\n",
    "\n",
    "                        % metabolite\n",
    "                        if ~any(strcmp(model.mets,metstring_ind{2}))\n",
    "                            error('ERROR - Constraints - Custom - Unrecognized metabolite %s',metstring_ind{2})\n",
    "                        end\n",
    "\n",
    "                        % stoichiometry\n",
    "                        [num, status] = str2num(metstring_ind{1});\n",
    "                        if status == 1\n",
    "                            model.S(strcmp(model.mets,metstring_ind{2}),end) = num;\n",
    "                        else\n",
    "                            error('ERROR - Constraints - Custom - Unrecognized stoichiometry %s',metstring_ind{1})\n",
    "                        end         \n",
    "                    end\n",
    "                end\n",
    "\n",
    "                % check constraint value(s)\n",
    "                for j = 1:length(constraints_value_value{i})\n",
    "                    if constraints_value_value{i}(j) < 0\n",
    "                        error('ERROR - Constraints - Custom - For irreversible reaction, constraint value must be >= 0')\n",
    "                    end\n",
    "                end\n",
    "\n",
    "                % equal to\n",
    "                if strcmp(constraints_value_type{i},'equal')\n",
    "                    model.lb(end+1) = constraints_value_value{i}(1);\n",
    "                    model.ub(end+1) = constraints_value_value{i}(1);\n",
    "\n",
    "                % less than\n",
    "                elseif strcmp(constraints_value_type{i},'less')\n",
    "                    model.lb(end+1) = 0;\n",
    "                    model.ub(end+1) = constraints_value_value{i}(1);\n",
    "\n",
    "                % greater than\n",
    "                elseif strcmp(constraints_value_type{i},'greater')\n",
    "                    model.lb(end+1) = constraints_value_value{i}(1);\n",
    "                    model.ub(end+1) = max(model.ub);\n",
    "\n",
    "                % range\n",
    "                elseif strcmp(constraints_value_type{i},'range')\n",
    "                    if constraints_value_value{i}(1) <= constraints_value_value{i}(2)\n",
    "                        model.lb(end+1) = constraints_value_value{i}(1);\n",
    "                        model.ub(end+1) = constraints_value_value{i}(2);\n",
    "                    else\n",
    "                        error('ERROR - Constraints - Custom - For Range, min value must be <= max value');\n",
    "                    end\n",
    "                end\n",
    "                \n",
    "                % initialize formula\n",
    "                model.rxnFormulas{end+1} = '';\n",
    "\n",
    "                % substrates\n",
    "                for j = find(model.S(:,end)<0)'\n",
    "                    if isempty(model.rxnFormulas{end})\n",
    "                        model.rxnFormulas{end} = sprintf('%d %s',-full(model.S(j,end)),model.mets{j});\n",
    "                    else\n",
    "                        model.rxnFormulas{end} = [model.rxnFormulas{end},' + ',sprintf('%d %s',-full(model.S(j,end)),model.mets{j})];\n",
    "                    end\n",
    "                end\n",
    "\n",
    "                % reaction arrow\n",
    "                if (model.lb(end) < 0) && (model.ub(end) > 0)\n",
    "                    model.rxnFormulas{end} = [model.rxnFormulas{end},' <=> '];\n",
    "                elseif (model.lb(end) < 0)\n",
    "                    model.rxnFormulas{end} = [model.rxnFormulas{end},' <-- '];\n",
    "                elseif (model.ub(end) > 0)\n",
    "                    model.rxnFormulas{end} = [model.rxnFormulas{end},' --> '];\n",
    "                else\n",
    "                    model.rxnFormulas{end} = [model.rxnFormulas{end},' X--X '];\n",
    "                end\n",
    "\n",
    "                % products\n",
    "                numproducts = 0;\n",
    "                for j = find(model.S(:,end)>0)'\n",
    "                    if numproducts == 0\n",
    "                        model.rxnFormulas{end} = [model.rxnFormulas{end},sprintf('%d %s',full(model.S(j,end)),model.mets{j})];\n",
    "                    else\n",
    "                        model.rxnFormulas{end} = [model.rxnFormulas{end},' + ',sprintf('%d %s',full(model.S(j,end)),model.mets{j})];\n",
    "                    end\n",
    "                    numproducts = numproducts + 1;\n",
    "                end\n",
    "\n",
    "            % reversible reaction\n",
    "            else\n",
    "\n",
    "                % parse reactants\n",
    "                metstring = strsplit(strtrim(constraints_value_reaction{i}(1:(strfind(constraints_value_reaction{i},'<=>')-1))),' + ','CollapseDelimiters',true);\n",
    "                for j = 1:length(metstring)\n",
    "                    if ~strcmp(metstring{j},'')\n",
    "                        metstring_ind = strsplit(metstring{j},' ');\n",
    "\n",
    "                        % metabolite\n",
    "                        if ~any(strcmp(model.mets,metstring_ind{2}))\n",
    "                            error('ERROR - Constraints - Custom - Unrecognized metabolite %s',metstring_ind{2})\n",
    "                        end\n",
    "\n",
    "                        % stoichiometry\n",
    "                        [num, status] = str2num(metstring_ind{1});\n",
    "                        if status == 1\n",
    "                            model.S(strcmp(model.mets,metstring_ind{2}),end) = -num;\n",
    "                        else\n",
    "                            error('ERROR - Constraints - Custom - Unrecognized stoichiometry %s',metstring_ind{1})\n",
    "                        end   \n",
    "                    end\n",
    "                end\n",
    "\n",
    "                % parse products\n",
    "                metstring = strsplit(strtrim(constraints_value_reaction{i}((strfind(constraints_value_reaction{i},'<=>')+3):end)),' + ','CollapseDelimiters',true);\n",
    "                for j = 1:length(metstring)\n",
    "                    if ~strcmp(metstring{j},'')\n",
    "                        metstring_ind = strsplit(metstring{j},' ');\n",
    "\n",
    "                        % metabolite\n",
    "                        if ~any(strcmp(model.mets,metstring_ind{2}))\n",
    "                            error('ERROR - Constraints - Custom - Unrecognized metabolite %s',metstring_ind{2})\n",
    "                        end\n",
    "\n",
    "                        % stoichiometry\n",
    "                        [num, status] = str2num(metstring_ind{1});\n",
    "                        if status == 1\n",
    "                            model.S(strcmp(model.mets,metstring_ind{2}),end) = num;\n",
    "                        else\n",
    "                            error('ERROR - Constraints - Custom - Unrecognized stoichiometry %s',metstring_ind{1})\n",
    "                        end       \n",
    "                    end\n",
    "                end\n",
    "\n",
    "                % equal to\n",
    "                if strcmp(constraints_value_type{i},'equal')\n",
    "                    model.lb(end+1) = constraints_value_value{i}(1);\n",
    "                    model.ub(end+1) = constraints_value_value{i}(1);\n",
    "\n",
    "                % less than\n",
    "                elseif strcmp(constraints_value_type{i},'less')\n",
    "                    model.lb(end+1) = min(model.lb);\n",
    "                    model.ub(end+1) = constraints_value_value{i}(1);\n",
    "\n",
    "                % greater than\n",
    "                elseif strcmp(constraints_value_type{i},'greater')\n",
    "                    model.lb(end+1) = constraints_value_value{i}(1);\n",
    "                    model.ub(end+1) = max(model.ub);\n",
    "\n",
    "                % range\n",
    "                elseif strcmp(constraints_value_type{i},'range')\n",
    "                    if constraints_value_value{i}(1) <= constraints_value_value{i}(2)\n",
    "                        model.lb(end+1) = constraints_value_value{i}(1);\n",
    "                        model.ub(end+1) = constraints_value_value{i}(2);\n",
    "                    else\n",
    "                        error('ERROR - Constraints - Custom - For Range, min value must be <= max value');\n",
    "                    end\n",
    "                end\n",
    "                \n",
    "                % initialize formula\n",
    "                model.rxnFormulas{end+1} = '';\n",
    "\n",
    "                % substrates\n",
    "                for j = find(model.S(:,end)<0)'\n",
    "                    if isempty(model.rxnFormulas{end})\n",
    "                        model.rxnFormulas{end} = sprintf('%d %s',-full(model.S(j,end)),model.mets{j});\n",
    "                    else\n",
    "                        model.rxnFormulas{end} = [model.rxnFormulas{end},' + ',sprintf('%d %s',-full(model.S(j,end)),model.mets{j})];\n",
    "                    end\n",
    "                end\n",
    "\n",
    "                % reaction arrow\n",
    "                if (model.lb(end) < 0) && (model.ub(end) > 0)\n",
    "                    model.rxnFormulas{end} = [model.rxnFormulas{end},' <=> '];\n",
    "                elseif (model.lb(end) < 0)\n",
    "                    model.rxnFormulas{end} = [model.rxnFormulas{end},' <-- '];\n",
    "                elseif (model.ub(end) > 0)\n",
    "                    model.rxnFormulas{end} = [model.rxnFormulas{end},' --> '];\n",
    "                else\n",
    "                    model.rxnFormulas{end} = [model.rxnFormulas{end},' X--X '];\n",
    "                end\n",
    "\n",
    "                % products\n",
    "                numproducts = 0;\n",
    "                for j = find(model.S(:,end)>0)'\n",
    "                    if numproducts == 0\n",
    "                        model.rxnFormulas{end} = [model.rxnFormulas{end},sprintf('%d %s',full(model.S(j,end)),model.mets{j})];\n",
    "                    else\n",
    "                        model.rxnFormulas{end} = [model.rxnFormulas{end},' + ',sprintf('%d %s',full(model.S(j,end)),model.mets{j})];\n",
    "                    end\n",
    "                    numproducts = numproducts + 1;\n",
    "                end\n",
    "            end\n",
    "\n",
    "        % if not custom reaction\n",
    "        elseif any(strcmp(model.rxns,constraints_value_reaction{i}))\n",
    "\n",
    "            % check if biomass_reaction\n",
    "            if (strcmp(constraints_value_reaction{i},'biomass_reaction')) && (constraints_proliferation == 1)\n",
    "                error('ERROR - Constraints - Custom - If already implementing proliferation constraint, cannot create additional constraints on biomass reaction');\n",
    "            end\n",
    "\n",
    "            % equal to\n",
    "            if strcmp(constraints_value_type{i},'equal')\n",
    "                model.lb(find(strcmp(model.rxns,constraints_value_reaction{i}))) = constraints_value_value{i}(1);\n",
    "                model.ub(find(strcmp(model.rxns,constraints_value_reaction{i}))) = constraints_value_value{i}(1);\n",
    "\n",
    "            % less than\n",
    "            elseif strcmp(constraints_value_type{i},'less')\n",
    "                model.ub(find(strcmp(model.rxns,constraints_value_reaction{i}))) = constraints_value_value{i}(1);\n",
    "\n",
    "            % greater than\n",
    "            elseif strcmp(constraints_value_type{i},'greater')\n",
    "                model.lb(find(strcmp(model.rxns,constraints_value_reaction{i}))) = constraints_value_value{i}(1);\n",
    "\n",
    "            % range\n",
    "            elseif strcmp(constraints_value_type{i},'range')\n",
    "                if constraints_value_value{i}(1) <= constraints_value_value{i}(2)\n",
    "                    model.lb(find(strcmp(model.rxns,constraints_value_reaction{i}))) = constraints_value_value{i}(1);\n",
    "                    model.ub(find(strcmp(model.rxns,constraints_value_reaction{i}))) = constraints_value_value{i}(2);\n",
    "                else\n",
    "                    error('ERROR - Constraints - Custom - For Range, min value must be <= max value');\n",
    "                end\n",
    "            end\n",
    "\n",
    "        else\n",
    "            error('ERROR - Constraints - Custom - Either invalid reaction formula or non-matching reaction name');\n",
    "        end\n",
    "    end\n",
    "    \n",
    "    % record rxn id's for custom fraction of maximum constraints\n",
    "    constraints_fraction_id = [];\n",
    "    \n",
    "    % implement custom fraction of maximum constraints\n",
    "    for i = 1:constraints_fraction_count\n",
    "        \n",
    "        % if custom reaction\n",
    "        if (contains(constraints_fraction_reaction{i},'-->')) || (contains(constraints_fraction_reaction{i},'<=>'))\n",
    "\n",
    "            % irreversible reaction\n",
    "            if (contains(constraints_fraction_reaction{i},'-->'))\n",
    "\n",
    "                % determine if all() reaction\n",
    "                if strcmp(constraints_fraction_reaction{i}(1:4),'all(') && strcmp(constraints_fraction_reaction{i}(end),')')\n",
    "\n",
    "                    % reaction formula without all()\n",
    "                    rxn_no_all = constraints_fraction_reaction{i}(5:end-1);\n",
    "\n",
    "                    % save reactant and product list\n",
    "                    reactants = {};\n",
    "                    reactant_stoich = [];\n",
    "                    reactant_compartments = {};\n",
    "                    products = {};\n",
    "                    product_stoich = [];\n",
    "                    product_compartments = {};\n",
    "\n",
    "                    % parse reactants\n",
    "                    metstring = strsplit(strtrim(rxn_no_all(1:(strfind(rxn_no_all,'-->')-1))),' + ','CollapseDelimiters',true);\n",
    "                    for j = 1:length(metstring)\n",
    "                        if ~strcmp(metstring{j},'')\n",
    "                            metstring_ind = strsplit(metstring{j},' ');\n",
    "\n",
    "                            % get matching metabolite and compartments\n",
    "                            found_compartments = {};\n",
    "                            for k = 1:length(model.mets)\n",
    "                                if strcmp(model.mets{k}(1:end-3),metstring_ind{2}(1:end-2))\n",
    "                                    found_compartments{end+1} = model.mets{k}(end-1);\n",
    "                                end\n",
    "                            end\n",
    "\n",
    "                            % if found compartments\n",
    "                            if length(found_compartments) > 0\n",
    "\n",
    "                                % add metabolite\n",
    "                                reactants{end+1} = metstring_ind{2}(1:end-2);\n",
    "                                reactant_compartments{end+1} = found_compartments;\n",
    "\n",
    "                                % add stoichiometry\n",
    "                                [num, status] = str2num(metstring_ind{1});\n",
    "                                if status == 1\n",
    "                                    reactant_stoich(end+1) = num;\n",
    "                                else\n",
    "                                    error('ERROR - Constraints - Custom - Unrecognized stoichiometry %s',metstring_ind{1})\n",
    "                                end\n",
    "\n",
    "                            % if no found compartments\n",
    "                            else\n",
    "                                error('ERROR - Constraints - Custom - Could not find any compartments for metabolite %s',metstring_ind{1}(1:end-2))\n",
    "                            end \n",
    "                        end\n",
    "                    end\n",
    "\n",
    "                    % parse products\n",
    "                    metstring = strsplit(strtrim(rxn_no_all((strfind(rxn_no_all,'-->')+3):end)),' + ','CollapseDelimiters',true);\n",
    "                    for j = 1:length(metstring)\n",
    "                        if ~strcmp(metstring{j},'')\n",
    "                            metstring_ind = strsplit(metstring{j},' ');\n",
    "\n",
    "                            % get matching metabolite and compartments\n",
    "                            found_compartments = {};\n",
    "                            for k = 1:length(model.mets)\n",
    "                                if strcmp(model.mets{k}(1:end-3),metstring_ind{2}(1:end-2))\n",
    "                                    found_compartments{end+1} = model.mets{k}(end-1);\n",
    "                                end\n",
    "                            end\n",
    "\n",
    "                            % if found compartments\n",
    "                            if length(found_compartments) > 0\n",
    "\n",
    "                                % add metabolite\n",
    "                                products{end+1} = metstring_ind{2}(1:end-2);\n",
    "                                product_compartments{end+1} = found_compartments;\n",
    "\n",
    "                                % add stoichiometry\n",
    "                                [num, status] = str2num(metstring_ind{1});\n",
    "                                if status == 1\n",
    "                                    product_stoich(end+1) = num;\n",
    "                                else\n",
    "                                    error('ERROR - Constraints - Custom - Unrecognized stoichiometry %s',metstring_ind{1})\n",
    "                                end\n",
    "\n",
    "                            % if no found compartments\n",
    "                            else\n",
    "                                error('ERROR - Constraints - Custom - Could not find any compartments for metabolite %s',metstring_ind{1}(1:end-2))\n",
    "                            end \n",
    "                        end\n",
    "                    end\n",
    "\n",
    "                    % find compartments for all reactants and products\n",
    "                    all_compartments = {'c','e','g','l','m','n','r','x','i'};\n",
    "                    for j = 1:length(reactant_compartments)\n",
    "                        all_compartments = intersect(all_compartments,reactant_compartments{j});\n",
    "                    end\n",
    "                    for j = 1:length(product_compartments)\n",
    "                        all_compartments = intersect(all_compartments,product_compartments{j});\n",
    "                    end\n",
    "\n",
    "                    % if at least one compartment\n",
    "                    if length(all_compartments) > 0\n",
    "\n",
    "                        % create new reaction for all compartments\n",
    "                        for j = 1:length(all_compartments)\n",
    "                            model.rxns{end+1} = sprintf('custom_fraction_%d_%s',i,all_compartments{j});\n",
    "                            constraints_fraction_id(end+1) = length(model.rxns);\n",
    "                            model.S(:,end+1) = zeros(length(model.mets),1);\n",
    "                            model.lb(end+1) = 0;\n",
    "                            model.ub(end+1) = max(model.ub);\n",
    "                            model.c(end+1) = 0;\n",
    "                            model.c_obj(end+1) = 0;\n",
    "                            model.c_fraction(end+1) = i;\n",
    "                            model.c_maxmulti(end+1) = 0;\n",
    "                            model.rxnGeneMat(end+1,:) = zeros(1,length(model.genes));\n",
    "                            model.rules{end+1} = '';\n",
    "                            model.grRules{end+1} = '';\n",
    "                            model.subSystems{end+1} = 'Custom Constraint'\n",
    "                            model.rxnNames{end+1} = sprintf('Custom Fraction Constraint %d - Compartment %s',i,all_compartments{j});\n",
    "                            model.rxnKEGGID{end+1} = '';\n",
    "                            model.rxnKeggOrthology{end+1} = '';\n",
    "                            model.rxnConfidenceScores(end+1) = 0;\n",
    "                            model.rxnReferences{end+1} = '';\n",
    "                            model.rxnECNumbers{end+1} = '';\n",
    "                            model.rxnNotes{end+1} = '';\n",
    "                            model.rxnCOG{end+1} = '';\n",
    "                            model.rxnReconMap{end+1} = '';\n",
    "\n",
    "                            % reactants\n",
    "                            for k = 1:length(reactants)\n",
    "                                metstring = sprintf('%s[%s]',reactants{k},all_compartments{j});\n",
    "                                model.S(strcmp(model.mets,metstring),end) = -reactant_stoich(k);\n",
    "                            end\n",
    "\n",
    "                            % products\n",
    "                            for k = 1:length(products)\n",
    "                                metstring = sprintf('%s[%s]',products{k},all_compartments{j});\n",
    "                                model.S(strcmp(model.mets,metstring),end) = product_stoich(k);\n",
    "                            end\n",
    "                            \n",
    "                            % initialize formula\n",
    "                            model.rxnFormulas{end+1} = '';\n",
    "\n",
    "                            % substrates\n",
    "                            for k = find(model.S(:,end)<0)'\n",
    "                                if isempty(model.rxnFormulas{end})\n",
    "                                    model.rxnFormulas{end} = sprintf('%d %s',-full(model.S(k,end)),model.mets{k});\n",
    "                                else\n",
    "                                    model.rxnFormulas{end} = [model.rxnFormulas{end},' + ',sprintf('%d %s',-full(model.S(k,end)),model.mets{k})];\n",
    "                                end\n",
    "                            end\n",
    "\n",
    "                            % reaction arrow\n",
    "                            if (model.lb(end) < 0) && (model.ub(end) > 0)\n",
    "                                model.rxnFormulas{end} = [model.rxnFormulas{end},' <=> '];\n",
    "                            elseif (model.lb(end) < 0)\n",
    "                                model.rxnFormulas{end} = [model.rxnFormulas{end},' <-- '];\n",
    "                            elseif (model.ub(end) > 0)\n",
    "                                model.rxnFormulas{end} = [model.rxnFormulas{end},' --> '];\n",
    "                            else\n",
    "                                model.rxnFormulas{end} = [model.rxnFormulas{end},' X--X '];\n",
    "                            end\n",
    "\n",
    "                            % products\n",
    "                            numproducts = 0;\n",
    "                            for k = find(model.S(:,end)>0)'\n",
    "                                if numproducts == 0\n",
    "                                    model.rxnFormulas{end} = [model.rxnFormulas{end},sprintf('%d %s',full(model.S(k,end)),model.mets{k})];\n",
    "                                else\n",
    "                                    model.rxnFormulas{end} = [model.rxnFormulas{end},' + ',sprintf('%d %s',full(model.S(k,end)),model.mets{k})];\n",
    "                                end\n",
    "                                numproducts = numproducts + 1;\n",
    "                            end\n",
    "                        end    \n",
    "\n",
    "                    % if no compartments\n",
    "                    else\n",
    "                       error('ERROR - Constraints - Custom - Could not find any compartments that contain all reaction metabolites') \n",
    "                    end\n",
    "\n",
    "                % if not all() reaction\n",
    "                else\n",
    "\n",
    "                    % create new reaction\n",
    "                    model.rxns{end+1} = sprintf('custom_fraction_%d',i);\n",
    "                    constraints_fraction_id(end+1) = length(model.rxns);\n",
    "                    model.S(:,end+1) = zeros(length(model.mets),1);\n",
    "                    model.lb(end+1) = 0;\n",
    "                    model.ub(end+1) = max(model.ub);\n",
    "                    model.c(end+1) = 0;\n",
    "                    model.c_obj(end+1) = 0;\n",
    "                    model.c_fraction(end+1) = i;\n",
    "                    model.c_maxmulti(end+1) = 0;\n",
    "                    model.rxnGeneMat(end+1,:) = zeros(1,length(model.genes));\n",
    "                    model.rules{end+1} = '';\n",
    "                    model.grRules{end+1} = '';\n",
    "                    model.subSystems{end+1} = 'Custom Constraint';\n",
    "                    model.rxnNames{end+1} = sprintf('Custom Fraction Constraint %d',i);\n",
    "                    model.rxnKEGGID{end+1} = '';\n",
    "                    model.rxnKeggOrthology{end+1} = '';\n",
    "                    model.rxnConfidenceScores(end+1) = 0;\n",
    "                    model.rxnReferences{end+1} = '';\n",
    "                    model.rxnECNumbers{end+1} = '';\n",
    "                    model.rxnNotes{end+1} = '';\n",
    "                    model.rxnCOG{end+1} = '';\n",
    "                    model.rxnReconMap{end+1} = '';\n",
    "\n",
    "                    % parse reactants\n",
    "                    metstring = strsplit(strtrim(constraints_fraction_reaction{i}(1:(strfind(constraints_fraction_reaction{i},'-->')-1))),' + ','CollapseDelimiters',true);\n",
    "                    for j = 1:length(metstring)\n",
    "                        if ~strcmp(metstring{j},'')\n",
    "                            metstring_ind = strsplit(metstring{j},' ');\n",
    "\n",
    "                            % metabolite\n",
    "                            if ~any(strcmp(model.mets,metstring_ind{2}))\n",
    "                                error('ERROR - Constraints - Custom - Unrecognized metabolite %s',metstring_ind{2})\n",
    "                            end\n",
    "\n",
    "                            % stoichiometry\n",
    "                            [num, status] = str2num(metstring_ind{1});\n",
    "                            if status == 1\n",
    "                                model.S(strcmp(model.mets,metstring_ind{2}),end) = -num;\n",
    "                            else\n",
    "                                error('ERROR - Constraints - Custom - Unrecognized stoichiometry %s',metstring_ind{1})\n",
    "                            end  \n",
    "                        end\n",
    "                    end\n",
    "\n",
    "                    % parse products\n",
    "                    metstring = strsplit(strtrim(constraints_fraction_reaction{i}((strfind(constraints_fraction_reaction{i},'-->')+3):end)),' + ','CollapseDelimiters',true);\n",
    "                    for j = 1:length(metstring)\n",
    "                        if ~strcmp(metstring{j},'')\n",
    "                            metstring_ind = strsplit(metstring{j},' ');\n",
    "\n",
    "                            % metabolite\n",
    "                            if ~any(strcmp(model.mets,metstring_ind{2}))\n",
    "                                error('ERROR - Constraints - Custom - Unrecognized metabolite %s',metstring_ind{2})\n",
    "                            end\n",
    "\n",
    "                            % stoichiometry\n",
    "                            [num, status] = str2num(metstring_ind{1});\n",
    "                            if status == 1\n",
    "                                model.S(strcmp(model.mets,metstring_ind{2}),end) = num;\n",
    "                            else\n",
    "                                error('ERROR - Constraints - Custom - Unrecognized stoichiometry %s',metstring_ind{1})\n",
    "                            end   \n",
    "                        end\n",
    "                    end\n",
    "                    \n",
    "                    % initialize formula\n",
    "                    model.rxnFormulas{end+1} = '';\n",
    "\n",
    "                    % substrates\n",
    "                    for k = find(model.S(:,end)<0)'\n",
    "                        if isempty(model.rxnFormulas{end})\n",
    "                            model.rxnFormulas{end} = sprintf('%d %s',-full(model.S(k,end)),model.mets{k});\n",
    "                        else\n",
    "                            model.rxnFormulas{end} = [model.rxnFormulas{end},' + ',sprintf('%d %s',-full(model.S(k,end)),model.mets{k})];\n",
    "                        end\n",
    "                    end\n",
    "\n",
    "                    % reaction arrow\n",
    "                    if (model.lb(end) < 0) && (model.ub(end) > 0)\n",
    "                        model.rxnFormulas{end} = [model.rxnFormulas{end},' <=> '];\n",
    "                    elseif (model.lb(end) < 0)\n",
    "                        model.rxnFormulas{end} = [model.rxnFormulas{end},' <-- '];\n",
    "                    elseif (model.ub(end) > 0)\n",
    "                        model.rxnFormulas{end} = [model.rxnFormulas{end},' --> '];\n",
    "                    else\n",
    "                        model.rxnFormulas{end} = [model.rxnFormulas{end},' X--X '];\n",
    "                    end\n",
    "\n",
    "                    % products\n",
    "                    numproducts = 0;\n",
    "                    for k = find(model.S(:,end)>0)'\n",
    "                        if numproducts == 0\n",
    "                            model.rxnFormulas{end} = [model.rxnFormulas{end},sprintf('%d %s',full(model.S(k,end)),model.mets{k})];\n",
    "                        else\n",
    "                            model.rxnFormulas{end} = [model.rxnFormulas{end},' + ',sprintf('%d %s',full(model.S(k,end)),model.mets{k})];\n",
    "                        end\n",
    "                        numproducts = numproducts + 1;\n",
    "                    end\n",
    "                end\n",
    "\n",
    "            % reversible reaction\n",
    "            else\n",
    "                % determine if all() reaction\n",
    "                if strcmp(constraints_fraction_reaction{i}(1:4),'all(') && strcmp(constraints_fraction_reaction{i}(end),')')\n",
    "\n",
    "                    % reaction formula without all()\n",
    "                    rxn_no_all = constraints_fraction_reaction{i}(5:end-1);\n",
    "\n",
    "                    % save reactant and product list\n",
    "                    reactants = {};\n",
    "                    reactant_stoich = [];\n",
    "                    reactant_compartments = {};\n",
    "                    products = {};\n",
    "                    product_stoich = [];\n",
    "                    product_compartments = {};\n",
    "\n",
    "                    % parse reactants\n",
    "                    metstring = strsplit(strtrim(rxn_no_all(1:(strfind(rxn_no_all,'<=>')-1))),' + ','CollapseDelimiters',true);\n",
    "                    for j = 1:length(metstring)\n",
    "                        if ~strcmp(metstring{j},'')\n",
    "                            metstring_ind = strsplit(metstring{j},' ');\n",
    "\n",
    "                            % get matching metabolite and compartments\n",
    "                            found_compartments = {};\n",
    "                            for k = 1:length(model.mets)\n",
    "                                if strcmp(model.mets{k}(1:end-3),metstring_ind{2}(1:end-2))\n",
    "                                    found_compartments{end+1} = model.mets{k}(end-1);\n",
    "                                end\n",
    "                            end\n",
    "\n",
    "                            % if found compartments\n",
    "                            if length(found_compartments) > 0\n",
    "\n",
    "                                % add metabolite\n",
    "                                reactants{end+1} = metstring_ind{2}(1:end-2);\n",
    "                                reactant_compartments{end+1} = found_compartments;\n",
    "\n",
    "                                % add stoichiometry\n",
    "                                [num, status] = str2num(metstring_ind{1});\n",
    "                                if status == 1\n",
    "                                    reactant_stoich(end+1) = num;\n",
    "                                else\n",
    "                                    error('ERROR - Constraints - Custom - Unrecognized stoichiometry %s',metstring_ind{1})\n",
    "                                end\n",
    "\n",
    "                            % if no found compartments\n",
    "                            else\n",
    "                                error('ERROR - Constraints - Custom - Could not find any compartments for metabolite %s',metstring_ind{1}(1:end-2))\n",
    "                            end \n",
    "                        end\n",
    "                    end\n",
    "\n",
    "                    % parse products\n",
    "                    metstring = strsplit(strtrim(rxn_no_all((strfind(rxn_no_all,'<=>')+3):end)),' + ','CollapseDelimiters',true);\n",
    "                    for j = 1:length(metstring)\n",
    "                        if ~strcmp(metstring{j},'')\n",
    "                            metstring_ind = strsplit(metstring{j},' ');\n",
    "\n",
    "                            % get matching metabolite and compartments\n",
    "                            found_compartments = {};\n",
    "                            for k = 1:length(model.mets)\n",
    "                                if strcmp(model.mets{k}(1:end-3),metstring_ind{2}(1:end-2))\n",
    "                                    found_compartments{end+1} = model.mets{k}(end-1);\n",
    "                                end\n",
    "                            end\n",
    "\n",
    "                            % if found compartments\n",
    "                            if length(found_compartments) > 0\n",
    "\n",
    "                                % add metabolite\n",
    "                                products{end+1} = metstring_ind{2}(1:end-2);\n",
    "                                product_compartments{end+1} = found_compartments;\n",
    "\n",
    "                                % add stoichiometry\n",
    "                                [num, status] = str2num(metstring_ind{1});\n",
    "                                if status == 1\n",
    "                                    product_stoich(end+1) = num;\n",
    "                                else\n",
    "                                    error('ERROR - Constraints - Custom - Unrecognized stoichiometry %s',metstring_ind{1})\n",
    "                                end\n",
    "\n",
    "                            % if no found compartments\n",
    "                            else\n",
    "                                error('ERROR - Constraints - Custom - Could not find any compartments for metabolite %s',metstring_ind{1}(1:end-2))\n",
    "                            end \n",
    "                        end\n",
    "                    end\n",
    "\n",
    "                    % find compartments for all reactants and products\n",
    "                    all_compartments = {'c','e','g','l','m','n','r','x','i'};\n",
    "                    for j = 1:length(reactant_compartments)\n",
    "                        all_compartments = intersect(all_compartments,reactant_compartments{j});\n",
    "                    end\n",
    "                    for j = 1:length(product_compartments)\n",
    "                        all_compartments = intersect(all_compartments,product_compartments{j});\n",
    "                    end\n",
    "\n",
    "                    % if at least one compartment\n",
    "                    if length(all_compartments) > 0\n",
    "\n",
    "                        % create new reaction for all compartments\n",
    "                        for j = 1:length(all_compartments)\n",
    "                            model.rxns{end+1} = sprintf('custom_fraction_%d_%s',i,all_compartments{j});\n",
    "                            constraints_fraction_id(end+1) = length(model.rxns);\n",
    "                            model.S(:,end+1) = zeros(length(model.mets),1);\n",
    "                            model.lb(end+1) = min(model.lb);\n",
    "                            model.ub(end+1) = max(model.ub);\n",
    "                            model.c(end+1) = 0;\n",
    "                            model.c_obj(end+1) = 0;\n",
    "                            model.c_fraction(end+1) = i;\n",
    "                            model.c_maxmulti(end+1) = 0;\n",
    "                            model.rxnGeneMat(end+1,:) = zeros(1,length(model.genes));\n",
    "                            model.rules{end+1} = '';\n",
    "                            model.grRules{end+1} = '';\n",
    "                            model.subSystems{end+1} = 'Custom Constraint'\n",
    "                            model.rxnNames{end+1} = sprintf('Custom Fraction Constraint %d - Compartment %s',i,all_compartments{j});\n",
    "                            model.rxnKEGGID{end+1} = '';\n",
    "                            model.rxnKeggOrthology{end+1} = '';\n",
    "                            model.rxnConfidenceScores(end+1) = 0;\n",
    "                            model.rxnReferences{end+1} = '';\n",
    "                            model.rxnECNumbers{end+1} = '';\n",
    "                            model.rxnNotes{end+1} = '';\n",
    "                            model.rxnCOG{end+1} = '';\n",
    "                            model.rxnReconMap{end+1} = '';\n",
    "\n",
    "                            % reactants\n",
    "                            for k = 1:length(reactants)\n",
    "                                metstring = sprintf('%s[%s]',reactants{k},all_compartments{j});\n",
    "                                model.S(strcmp(model.mets,metstring),end) = -reactant_stoich(k);\n",
    "                            end\n",
    "\n",
    "                            % products\n",
    "                            for k = 1:length(products)\n",
    "                                metstring = sprintf('%s[%s]',products{k},all_compartments{j});\n",
    "                                model.S(strcmp(model.mets,metstring),end) = product_stoich(k);\n",
    "                            end\n",
    "                            \n",
    "                            % initialize formula\n",
    "                            model.rxnFormulas{end+1} = '';\n",
    "\n",
    "                            % substrates\n",
    "                            for k = find(model.S(:,end)<0)'\n",
    "                                if isempty(model.rxnFormulas{end})\n",
    "                                    model.rxnFormulas{end} = sprintf('%d %s',-full(model.S(k,end)),model.mets{k});\n",
    "                                else\n",
    "                                    model.rxnFormulas{end} = [model.rxnFormulas{end},' + ',sprintf('%d %s',-full(model.S(k,end)),model.mets{k})];\n",
    "                                end\n",
    "                            end\n",
    "\n",
    "                            % reaction arrow\n",
    "                            if (model.lb(end) < 0) && (model.ub(end) > 0)\n",
    "                                model.rxnFormulas{end} = [model.rxnFormulas{end},' <=> '];\n",
    "                            elseif (model.lb(end) < 0)\n",
    "                                model.rxnFormulas{end} = [model.rxnFormulas{end},' <-- '];\n",
    "                            elseif (model.ub(end) > 0)\n",
    "                                model.rxnFormulas{end} = [model.rxnFormulas{end},' --> '];\n",
    "                            else\n",
    "                                model.rxnFormulas{end} = [model.rxnFormulas{end},' X--X '];\n",
    "                            end\n",
    "\n",
    "                            % products\n",
    "                            numproducts = 0;\n",
    "                            for k = find(model.S(:,end)>0)'\n",
    "                                if numproducts == 0\n",
    "                                    model.rxnFormulas{end} = [model.rxnFormulas{end},sprintf('%d %s',full(model.S(k,end)),model.mets{k})];\n",
    "                                else\n",
    "                                    model.rxnFormulas{end} = [model.rxnFormulas{end},' + ',sprintf('%d %s',full(model.S(k,end)),model.mets{k})];\n",
    "                                end\n",
    "                                numproducts = numproducts + 1;\n",
    "                            end\n",
    "                        end    \n",
    "\n",
    "                    % if no compartments\n",
    "                    else\n",
    "                       error('ERROR - Constraints - Custom - Could not find any compartments that contain all reaction metabolites') \n",
    "                    end\n",
    "\n",
    "                % if not all() reaction\n",
    "                else\n",
    "\n",
    "                    % create new reaction\n",
    "                    model.rxns{end+1} = sprintf('custom_fraction_%d',i);\n",
    "                    constraints_fraction_id(end+1) = length(model.rxns);\n",
    "                    model.S(:,end+1) = zeros(length(model.mets),1);\n",
    "                    model.lb(end+1) = min(model.lb);\n",
    "                    model.ub(end+1) = max(model.ub);\n",
    "                    model.c(end+1) = 0;\n",
    "                    model.c_obj(end+1) = 0;\n",
    "                    model.c_fraction(end+1) = i;\n",
    "\t\t\t\t\tmodel.c_maxmulti(end+1) = 0;\n",
    "                    model.rxnGeneMat(end+1,:) = zeros(1,length(model.genes));\n",
    "                    model.rules{end+1} = '';\n",
    "                    model.grRules{end+1} = '';\n",
    "                    model.subSystems{end+1} = 'Custom Constraint';\n",
    "                    model.rxnNames{end+1} = sprintf('Custom Fraction Constraint %d',i);\n",
    "                    model.rxnKEGGID{end+1} = '';\n",
    "                    model.rxnKeggOrthology{end+1} = '';\n",
    "                    model.rxnConfidenceScores(end+1) = 0;\n",
    "                    model.rxnReferences{end+1} = '';\n",
    "                    model.rxnECNumbers{end+1} = '';\n",
    "                    model.rxnNotes{end+1} = '';\n",
    "                    model.rxnCOG{end+1} = '';\n",
    "                    model.rxnReconMap{end+1} = '';\n",
    "\n",
    "                    % parse reactants\n",
    "                    metstring = strsplit(strtrim(constraints_fraction_reaction{i}(1:(strfind(constraints_fraction_reaction{i},'<=>')-1))),' + ','CollapseDelimiters',true);\n",
    "                    for j = 1:length(metstring)\n",
    "                        if ~strcmp(metstring{j},'')\n",
    "                            metstring_ind = strsplit(metstring{j},' ');\n",
    "\n",
    "                            % metabolite\n",
    "                            if ~any(strcmp(model.mets,metstring_ind{2}))\n",
    "                                error('ERROR - Constraints - Custom - Unrecognized metabolite %s',metstring_ind{2})\n",
    "                            end\n",
    "\n",
    "                            % stoichiometry\n",
    "                            [num, status] = str2num(metstring_ind{1});\n",
    "                            if status == 1\n",
    "                                model.S(strcmp(model.mets,metstring_ind{2}),end) = -num;\n",
    "                            else\n",
    "                                error('ERROR - Constraints - Custom - Unrecognized stoichiometry %s',metstring_ind{1})\n",
    "                            end  \n",
    "                        end\n",
    "                    end\n",
    "\n",
    "                    % parse products\n",
    "                    metstring = strsplit(strtrim(constraints_fraction_reaction{i}((strfind(constraints_fraction_reaction{i},'<=>')+3):end)),' + ','CollapseDelimiters',true);\n",
    "                    for j = 1:length(metstring)\n",
    "                        if ~strcmp(metstring{j},'')\n",
    "                            metstring_ind = strsplit(metstring{j},' ');\n",
    "\n",
    "                            % metabolite\n",
    "                            if ~any(strcmp(model.mets,metstring_ind{2}))\n",
    "                                error('ERROR - Constraints - Custom - Unrecognized metabolite %s',metstring_ind{2})\n",
    "                            end\n",
    "\n",
    "                            % stoichiometry\n",
    "                            [num, status] = str2num(metstring_ind{1});\n",
    "                            if status == 1\n",
    "                                model.S(strcmp(model.mets,metstring_ind{2}),end) = num;\n",
    "                            else\n",
    "                                error('ERROR - Constraints - Custom - Unrecognized stoichiometry %s',metstring_ind{1})\n",
    "                            end   \n",
    "                        end\n",
    "                    end\n",
    "                    \n",
    "                    % initialize formula\n",
    "                    model.rxnFormulas{end+1} = '';\n",
    "\n",
    "                    % substrates\n",
    "                    for k = find(model.S(:,end)<0)'\n",
    "                        if isempty(model.rxnFormulas{end})\n",
    "                            model.rxnFormulas{end} = sprintf('%d %s',-full(model.S(k,end)),model.mets{k});\n",
    "                        else\n",
    "                            model.rxnFormulas{end} = [model.rxnFormulas{end},' + ',sprintf('%d %s',-full(model.S(k,end)),model.mets{k})];\n",
    "                        end\n",
    "                    end\n",
    "\n",
    "                    % reaction arrow\n",
    "                    if (model.lb(end) < 0) && (model.ub(end) > 0)\n",
    "                        model.rxnFormulas{end} = [model.rxnFormulas{end},' <=> '];\n",
    "                    elseif (model.lb(end) < 0)\n",
    "                        model.rxnFormulas{end} = [model.rxnFormulas{end},' <-- '];\n",
    "                    elseif (model.ub(end) > 0)\n",
    "                        model.rxnFormulas{end} = [model.rxnFormulas{end},' --> '];\n",
    "                    else\n",
    "                        model.rxnFormulas{end} = [model.rxnFormulas{end},' X--X '];\n",
    "                    end\n",
    "\n",
    "                    % products\n",
    "                    numproducts = 0;\n",
    "                    for k = find(model.S(:,end)>0)'\n",
    "                        if numproducts == 0\n",
    "                            model.rxnFormulas{end} = [model.rxnFormulas{end},sprintf('%d %s',full(model.S(k,end)),model.mets{k})];\n",
    "                        else\n",
    "                            model.rxnFormulas{end} = [model.rxnFormulas{end},' + ',sprintf('%d %s',full(model.S(k,end)),model.mets{k})];\n",
    "                        end\n",
    "                        numproducts = numproducts + 1;\n",
    "                    end\n",
    "                end\n",
    "            end\n",
    "\n",
    "        % if not custom reaction\n",
    "        elseif any(strcmp(model.rxns,constraints_fraction_reaction{i}))\n",
    "            constraints_fraction_id(end+1) = find(strcmp(model.rxns,constraints_fraction_reaction{i}));\n",
    "            \n",
    "            % check if biomass_reaction\n",
    "            if (strcmp(constraints_fraction_reaction{i},'biomass_reaction')) && (constraints_proliferation == 1)\n",
    "                error('ERROR - Constraints - Custom - If already implementing proliferation constraint, cannot create additional constraints on biomass reaction');\n",
    "            end\n",
    "            \n",
    "            % change model.c_fraction\n",
    "            model.c_fraction(strcmp(model.rxns,constraints_fraction_reaction{i})) = i;\n",
    "\n",
    "        else\n",
    "            error('ERROR - Constraints - Custom - Either invalid reaction formula or non-matching reaction name');\n",
    "        end\n",
    " \n",
    "    end\n",
    "\n",
    "    % implement custom max multi constraints\n",
    "    for i = 1:constraints_maxmulti_count\n",
    "        \n",
    "        % if custom reaction\n",
    "        if (contains(constraints_maxmulti_reaction{i},'-->')) || (contains(constraints_maxmulti_reaction{i},'<=>'))\n",
    "\n",
    "            % irreversible reaction\n",
    "            if (contains(constraints_maxmulti_reaction{i},'-->'))\n",
    "\n",
    "                % determine if all() reaction\n",
    "                if strcmp(constraints_maxmulti_reaction{i}(1:4),'all(') && strcmp(constraints_maxmulti_reaction{i}(end),')')\n",
    "\n",
    "                    % reaction formula without all()\n",
    "                    rxn_no_all = constraints_maxmulti_reaction{i}(5:end-1);\n",
    "\n",
    "                    % save reactant and product list\n",
    "                    reactants = {};\n",
    "                    reactant_stoich = [];\n",
    "                    reactant_compartments = {};\n",
    "                    products = {};\n",
    "                    product_stoich = [];\n",
    "                    product_compartments = {};\n",
    "\n",
    "                    % parse reactants\n",
    "                    metstring = strsplit(strtrim(rxn_no_all(1:(strfind(rxn_no_all,'-->')-1))),' + ','CollapseDelimiters',true);\n",
    "                    for j = 1:length(metstring)\n",
    "                        if ~strcmp(metstring{j},'')\n",
    "                            metstring_ind = strsplit(metstring{j},' ');\n",
    "\n",
    "                            % get matching metabolite and compartments\n",
    "                            found_compartments = {};\n",
    "                            for k = 1:length(model.mets)\n",
    "                                if strcmp(model.mets{k}(1:end-3),metstring_ind{2}(1:end-2))\n",
    "                                    found_compartments{end+1} = model.mets{k}(end-1);\n",
    "                                end\n",
    "                            end\n",
    "\n",
    "                            % if found compartments\n",
    "                            if length(found_compartments) > 0\n",
    "\n",
    "                                % add metabolite\n",
    "                                reactants{end+1} = metstring_ind{2}(1:end-2);\n",
    "                                reactant_compartments{end+1} = found_compartments;\n",
    "\n",
    "                                % add stoichiometry\n",
    "                                [num, status] = str2num(metstring_ind{1});\n",
    "                                if status == 1\n",
    "                                    reactant_stoich(end+1) = num;\n",
    "                                else\n",
    "                                    error('ERROR - Constraints - Custom - Unrecognized stoichiometry %s',metstring_ind{1})\n",
    "                                end\n",
    "\n",
    "                            % if no found compartments\n",
    "                            else\n",
    "                                error('ERROR - Constraints - Custom - Could not find any compartments for metabolite %s',metstring_ind{1}(1:end-2))\n",
    "                            end \n",
    "                        end\n",
    "                    end\n",
    "\n",
    "                    % parse products\n",
    "                    metstring = strsplit(strtrim(rxn_no_all((strfind(rxn_no_all,'-->')+3):end)),' + ','CollapseDelimiters',true);\n",
    "                    for j = 1:length(metstring)\n",
    "                        if ~strcmp(metstring{j},'')\n",
    "                            metstring_ind = strsplit(metstring{j},' ');\n",
    "\n",
    "                            % get matching metabolite and compartments\n",
    "                            found_compartments = {};\n",
    "                            for k = 1:length(model.mets)\n",
    "                                if strcmp(model.mets{k}(1:end-3),metstring_ind{2}(1:end-2))\n",
    "                                    found_compartments{end+1} = model.mets{k}(end-1);\n",
    "                                end\n",
    "                            end\n",
    "\n",
    "                            % if found compartments\n",
    "                            if length(found_compartments) > 0\n",
    "\n",
    "                                % add metabolite\n",
    "                                products{end+1} = metstring_ind{2}(1:end-2);\n",
    "                                product_compartments{end+1} = found_compartments;\n",
    "\n",
    "                                % add stoichiometry\n",
    "                                [num, status] = str2num(metstring_ind{1});\n",
    "                                if status == 1\n",
    "                                    product_stoich(end+1) = num;\n",
    "                                else\n",
    "                                    error('ERROR - Constraints - Custom - Unrecognized stoichiometry %s',metstring_ind{1})\n",
    "                                end\n",
    "\n",
    "                            % if no found compartments\n",
    "                            else\n",
    "                                error('ERROR - Constraints - Custom - Could not find any compartments for metabolite %s',metstring_ind{1}(1:end-2))\n",
    "                            end \n",
    "                        end\n",
    "                    end\n",
    "\n",
    "                    % find compartments for all reactants and products\n",
    "                    all_compartments = {'c','e','g','l','m','n','r','x','i'};\n",
    "                    for j = 1:length(reactant_compartments)\n",
    "                        all_compartments = intersect(all_compartments,reactant_compartments{j});\n",
    "                    end\n",
    "                    for j = 1:length(product_compartments)\n",
    "                        all_compartments = intersect(all_compartments,product_compartments{j});\n",
    "                    end\n",
    "\n",
    "                    % if at least one compartment\n",
    "                    if length(all_compartments) > 0\n",
    "\n",
    "                        % create new reaction for all compartments\n",
    "                        for j = 1:length(all_compartments)\n",
    "                            model.rxns{end+1} = sprintf('custom_maxmulti_%d_%s',i,all_compartments{j});\n",
    "                            model.S(:,end+1) = zeros(length(model.mets),1);\n",
    "                            model.lb(end+1) = 0;\n",
    "                            model.ub(end+1) = max(model.ub);\n",
    "                            model.c(end+1) = 0;\n",
    "                            model.c_obj(end+1) = 0;\n",
    "                            model.c_fraction(end+1) = 0;\n",
    "                            model.c_maxmulti(end+1) = constraints_maxmulti_weight(i);\n",
    "                            model.rxnGeneMat(end+1,:) = zeros(1,length(model.genes));\n",
    "                            model.rules{end+1} = '';\n",
    "                            model.grRules{end+1} = '';\n",
    "                            model.subSystems{end+1} = 'Custom Constraint'\n",
    "                            model.rxnNames{end+1} = sprintf('Custom Max Multi Constraint %d - Compartment %s',i,all_compartments{j});\n",
    "                            model.rxnKEGGID{end+1} = '';\n",
    "                            model.rxnKeggOrthology{end+1} = '';\n",
    "                            model.rxnConfidenceScores(end+1) = 0;\n",
    "                            model.rxnReferences{end+1} = '';\n",
    "                            model.rxnECNumbers{end+1} = '';\n",
    "                            model.rxnNotes{end+1} = '';\n",
    "                            model.rxnCOG{end+1} = '';\n",
    "                            model.rxnReconMap{end+1} = '';\n",
    "\n",
    "                            % reactants\n",
    "                            for k = 1:length(reactants)\n",
    "                                metstring = sprintf('%s[%s]',reactants{k},all_compartments{j});\n",
    "                                model.S(strcmp(model.mets,metstring),end) = -reactant_stoich(k);\n",
    "                            end\n",
    "\n",
    "                            % products\n",
    "                            for k = 1:length(products)\n",
    "                                metstring = sprintf('%s[%s]',products{k},all_compartments{j});\n",
    "                                model.S(strcmp(model.mets,metstring),end) = product_stoich(k);\n",
    "                            end\n",
    "                            \n",
    "                            % initialize formula\n",
    "                            model.rxnFormulas{end+1} = '';\n",
    "\n",
    "                            % substrates\n",
    "                            for k = find(model.S(:,end)<0)'\n",
    "                                if isempty(model.rxnFormulas{end})\n",
    "                                    model.rxnFormulas{end} = sprintf('%d %s',-full(model.S(k,end)),model.mets{k});\n",
    "                                else\n",
    "                                    model.rxnFormulas{end} = [model.rxnFormulas{end},' + ',sprintf('%d %s',-full(model.S(k,end)),model.mets{k})];\n",
    "                                end\n",
    "                            end\n",
    "\n",
    "                            % reaction arrow\n",
    "                            if (model.lb(end) < 0) && (model.ub(end) > 0)\n",
    "                                model.rxnFormulas{end} = [model.rxnFormulas{end},' <=> '];\n",
    "                            elseif (model.lb(end) < 0)\n",
    "                                model.rxnFormulas{end} = [model.rxnFormulas{end},' <-- '];\n",
    "                            elseif (model.ub(end) > 0)\n",
    "                                model.rxnFormulas{end} = [model.rxnFormulas{end},' --> '];\n",
    "                            else\n",
    "                                model.rxnFormulas{end} = [model.rxnFormulas{end},' X--X '];\n",
    "                            end\n",
    "\n",
    "                            % products\n",
    "                            numproducts = 0;\n",
    "                            for k = find(model.S(:,end)>0)'\n",
    "                                if numproducts == 0\n",
    "                                    model.rxnFormulas{end} = [model.rxnFormulas{end},sprintf('%d %s',full(model.S(k,end)),model.mets{k})];\n",
    "                                else\n",
    "                                    model.rxnFormulas{end} = [model.rxnFormulas{end},' + ',sprintf('%d %s',full(model.S(k,end)),model.mets{k})];\n",
    "                                end\n",
    "                                numproducts = numproducts + 1;\n",
    "                            end\n",
    "                        end    \n",
    "\n",
    "                    % if no compartments\n",
    "                    else\n",
    "                       error('ERROR - Constraints - Custom - Could not find any compartments that contain all reaction metabolites') \n",
    "                    end\n",
    "\n",
    "                % if not all() reaction\n",
    "                else\n",
    "\n",
    "                    % create new reaction\n",
    "                    model.rxns{end+1} = sprintf('custom_maxmulti_%d',i);\n",
    "                    model.S(:,end+1) = zeros(length(model.mets),1);\n",
    "                    model.lb(end+1) = 0;\n",
    "                    model.ub(end+1) = max(model.ub);\n",
    "                    model.c(end+1) = 0;\n",
    "                    model.c_obj(end+1) = 0;\n",
    "                    model.c_fraction(end+1) = 0;\n",
    "                    model.c_maxmulti(end+1) = constraints_maxmulti_weight(i);\n",
    "                    model.rxnGeneMat(end+1,:) = zeros(1,length(model.genes));\n",
    "                    model.rules{end+1} = '';\n",
    "                    model.grRules{end+1} = '';\n",
    "                    model.subSystems{end+1} = 'Custom Constraint';\n",
    "                    model.rxnNames{end+1} = sprintf('Custom Max Multi Constraint %d',i);\n",
    "                    model.rxnKEGGID{end+1} = '';\n",
    "                    model.rxnKeggOrthology{end+1} = '';\n",
    "                    model.rxnConfidenceScores(end+1) = 0;\n",
    "                    model.rxnReferences{end+1} = '';\n",
    "                    model.rxnECNumbers{end+1} = '';\n",
    "                    model.rxnNotes{end+1} = '';\n",
    "                    model.rxnCOG{end+1} = '';\n",
    "                    model.rxnReconMap{end+1} = '';\n",
    "\n",
    "                    % parse reactants\n",
    "                    metstring = strsplit(strtrim(constraints_maxmulti_reaction{i}(1:(strfind(constraints_maxmulti_reaction{i},'-->')-1))),' + ','CollapseDelimiters',true);\n",
    "                    for j = 1:length(metstring)\n",
    "                        if ~strcmp(metstring{j},'')\n",
    "                            metstring_ind = strsplit(metstring{j},' ');\n",
    "\n",
    "                            % metabolite\n",
    "                            if ~any(strcmp(model.mets,metstring_ind{2}))\n",
    "                                error('ERROR - Constraints - Custom - Unrecognized metabolite %s',metstring_ind{2})\n",
    "                            end\n",
    "\n",
    "                            % stoichiometry\n",
    "                            [num, status] = str2num(metstring_ind{1});\n",
    "                            if status == 1\n",
    "                                model.S(strcmp(model.mets,metstring_ind{2}),end) = -num;\n",
    "                            else\n",
    "                                error('ERROR - Constraints - Custom - Unrecognized stoichiometry %s',metstring_ind{1})\n",
    "                            end  \n",
    "                        end\n",
    "                    end\n",
    "\n",
    "                    % parse products\n",
    "                    metstring = strsplit(strtrim(constraints_maxmulti_reaction{i}((strfind(constraints_maxmulti_reaction{i},'-->')+3):end)),' + ','CollapseDelimiters',true);\n",
    "                    for j = 1:length(metstring)\n",
    "                        if ~strcmp(metstring{j},'')\n",
    "                            metstring_ind = strsplit(metstring{j},' ');\n",
    "\n",
    "                            % metabolite\n",
    "                            if ~any(strcmp(model.mets,metstring_ind{2}))\n",
    "                                error('ERROR - Constraints - Custom - Unrecognized metabolite %s',metstring_ind{2})\n",
    "                            end\n",
    "\n",
    "                            % stoichiometry\n",
    "                            [num, status] = str2num(metstring_ind{1});\n",
    "                            if status == 1\n",
    "                                model.S(strcmp(model.mets,metstring_ind{2}),end) = num;\n",
    "                            else\n",
    "                                error('ERROR - Constraints - Custom - Unrecognized stoichiometry %s',metstring_ind{1})\n",
    "                            end   \n",
    "                        end\n",
    "                    end\n",
    "                    \n",
    "                    % initialize formula\n",
    "                    model.rxnFormulas{end+1} = '';\n",
    "\n",
    "                    % substrates\n",
    "                    for k = find(model.S(:,end)<0)'\n",
    "                        if isempty(model.rxnFormulas{end})\n",
    "                            model.rxnFormulas{end} = sprintf('%d %s',-full(model.S(k,end)),model.mets{k});\n",
    "                        else\n",
    "                            model.rxnFormulas{end} = [model.rxnFormulas{end},' + ',sprintf('%d %s',-full(model.S(k,end)),model.mets{k})];\n",
    "                        end\n",
    "                    end\n",
    "\n",
    "                    % reaction arrow\n",
    "                    if (model.lb(end) < 0) && (model.ub(end) > 0)\n",
    "                        model.rxnFormulas{end} = [model.rxnFormulas{end},' <=> '];\n",
    "                    elseif (model.lb(end) < 0)\n",
    "                        model.rxnFormulas{end} = [model.rxnFormulas{end},' <-- '];\n",
    "                    elseif (model.ub(end) > 0)\n",
    "                        model.rxnFormulas{end} = [model.rxnFormulas{end},' --> '];\n",
    "                    else\n",
    "                        model.rxnFormulas{end} = [model.rxnFormulas{end},' X--X '];\n",
    "                    end\n",
    "\n",
    "                    % products\n",
    "                    numproducts = 0;\n",
    "                    for k = find(model.S(:,end)>0)'\n",
    "                        if numproducts == 0\n",
    "                            model.rxnFormulas{end} = [model.rxnFormulas{end},sprintf('%d %s',full(model.S(k,end)),model.mets{k})];\n",
    "                        else\n",
    "                            model.rxnFormulas{end} = [model.rxnFormulas{end},' + ',sprintf('%d %s',full(model.S(k,end)),model.mets{k})];\n",
    "                        end\n",
    "                        numproducts = numproducts + 1;\n",
    "                    end\n",
    "                end\n",
    "\n",
    "            % reversible reaction\n",
    "            else\n",
    "                % determine if all() reaction\n",
    "                if strcmp(constraints_maxmulti_reaction{i}(1:4),'all(') && strcmp(constraints_maxmulti_reaction{i}(end),')')\n",
    "\n",
    "                    % reaction formula without all()\n",
    "                    rxn_no_all = constraints_maxmulti_reaction{i}(5:end-1);\n",
    "\n",
    "                    % save reactant and product list\n",
    "                    reactants = {};\n",
    "                    reactant_stoich = [];\n",
    "                    reactant_compartments = {};\n",
    "                    products = {};\n",
    "                    product_stoich = [];\n",
    "                    product_compartments = {};\n",
    "\n",
    "                    % parse reactants\n",
    "                    metstring = strsplit(strtrim(rxn_no_all(1:(strfind(rxn_no_all,'<=>')-1))),' + ','CollapseDelimiters',true);\n",
    "                    for j = 1:length(metstring)\n",
    "                        if ~strcmp(metstring{j},'')\n",
    "                            metstring_ind = strsplit(metstring{j},' ');\n",
    "\n",
    "                            % get matching metabolite and compartments\n",
    "                            found_compartments = {};\n",
    "                            for k = 1:length(model.mets)\n",
    "                                if strcmp(model.mets{k}(1:end-3),metstring_ind{2}(1:end-2))\n",
    "                                    found_compartments{end+1} = model.mets{k}(end-1);\n",
    "                                end\n",
    "                            end\n",
    "\n",
    "                            % if found compartments\n",
    "                            if length(found_compartments) > 0\n",
    "\n",
    "                                % add metabolite\n",
    "                                reactants{end+1} = metstring_ind{2}(1:end-2);\n",
    "                                reactant_compartments{end+1} = found_compartments;\n",
    "\n",
    "                                % add stoichiometry\n",
    "                                [num, status] = str2num(metstring_ind{1});\n",
    "                                if status == 1\n",
    "                                    reactant_stoich(end+1) = num;\n",
    "                                else\n",
    "                                    error('ERROR - Constraints - Custom - Unrecognized stoichiometry %s',metstring_ind{1})\n",
    "                                end\n",
    "\n",
    "                            % if no found compartments\n",
    "                            else\n",
    "                                error('ERROR - Constraints - Custom - Could not find any compartments for metabolite %s',metstring_ind{1}(1:end-2))\n",
    "                            end \n",
    "                        end\n",
    "                    end\n",
    "\n",
    "                    % parse products\n",
    "                    metstring = strsplit(strtrim(rxn_no_all((strfind(rxn_no_all,'<=>')+3):end)),' + ','CollapseDelimiters',true);\n",
    "                    for j = 1:length(metstring)\n",
    "                        if ~strcmp(metstring{j},'')\n",
    "                            metstring_ind = strsplit(metstring{j},' ');\n",
    "\n",
    "                            % get matching metabolite and compartments\n",
    "                            found_compartments = {};\n",
    "                            for k = 1:length(model.mets)\n",
    "                                if strcmp(model.mets{k}(1:end-3),metstring_ind{2}(1:end-2))\n",
    "                                    found_compartments{end+1} = model.mets{k}(end-1);\n",
    "                                end\n",
    "                            end\n",
    "\n",
    "                            % if found compartments\n",
    "                            if length(found_compartments) > 0\n",
    "\n",
    "                                % add metabolite\n",
    "                                products{end+1} = metstring_ind{2}(1:end-2);\n",
    "                                product_compartments{end+1} = found_compartments;\n",
    "\n",
    "                                % add stoichiometry\n",
    "                                [num, status] = str2num(metstring_ind{1});\n",
    "                                if status == 1\n",
    "                                    product_stoich(end+1) = num;\n",
    "                                else\n",
    "                                    error('ERROR - Constraints - Custom - Unrecognized stoichiometry %s',metstring_ind{1})\n",
    "                                end\n",
    "\n",
    "                            % if no found compartments\n",
    "                            else\n",
    "                                error('ERROR - Constraints - Custom - Could not find any compartments for metabolite %s',metstring_ind{1}(1:end-2))\n",
    "                            end \n",
    "                        end\n",
    "                    end\n",
    "\n",
    "                    % find compartments for all reactants and products\n",
    "                    all_compartments = {'c','e','g','l','m','n','r','x','i'};\n",
    "                    for j = 1:length(reactant_compartments)\n",
    "                        all_compartments = intersect(all_compartments,reactant_compartments{j});\n",
    "                    end\n",
    "                    for j = 1:length(product_compartments)\n",
    "                        all_compartments = intersect(all_compartments,product_compartments{j});\n",
    "                    end\n",
    "\n",
    "                    % if at least one compartment\n",
    "                    if length(all_compartments) > 0\n",
    "\n",
    "                        % create new reaction for all compartments\n",
    "                        for j = 1:length(all_compartments)\n",
    "                            model.rxns{end+1} = sprintf('custom_maxmulti_%d_%s',i,all_compartments{j});\n",
    "                            model.S(:,end+1) = zeros(length(model.mets),1);\n",
    "                            model.lb(end+1) = min(model.lb);\n",
    "                            model.ub(end+1) = max(model.ub);\n",
    "                            model.c(end+1) = 0;\n",
    "                            model.c_obj(end+1) = 0;\n",
    "                            model.c_fraction(end+1) = 0;\n",
    "                            model.c_maxmulti(end+1) = constraints_maxmulti_weight(i);\n",
    "                            model.rxnGeneMat(end+1,:) = zeros(1,length(model.genes));\n",
    "                            model.rules{end+1} = '';\n",
    "                            model.grRules{end+1} = '';\n",
    "                            model.subSystems{end+1} = 'Custom Constraint'\n",
    "                            model.rxnNames{end+1} = sprintf('Custom Max Multi Constraint %d - Compartment %s',i,all_compartments{j});\n",
    "                            model.rxnKEGGID{end+1} = '';\n",
    "                            model.rxnKeggOrthology{end+1} = '';\n",
    "                            model.rxnConfidenceScores(end+1) = 0;\n",
    "                            model.rxnReferences{end+1} = '';\n",
    "                            model.rxnECNumbers{end+1} = '';\n",
    "                            model.rxnNotes{end+1} = '';\n",
    "                            model.rxnCOG{end+1} = '';\n",
    "                            model.rxnReconMap{end+1} = '';\n",
    "\n",
    "                            % reactants\n",
    "                            for k = 1:length(reactants)\n",
    "                                metstring = sprintf('%s[%s]',reactants{k},all_compartments{j});\n",
    "                                model.S(strcmp(model.mets,metstring),end) = -reactant_stoich(k);\n",
    "                            end\n",
    "\n",
    "                            % products\n",
    "                            for k = 1:length(products)\n",
    "                                metstring = sprintf('%s[%s]',products{k},all_compartments{j});\n",
    "                                model.S(strcmp(model.mets,metstring),end) = product_stoich(k);\n",
    "                            end\n",
    "                            \n",
    "                            % initialize formula\n",
    "                            model.rxnFormulas{end+1} = '';\n",
    "\n",
    "                            % substrates\n",
    "                            for k = find(model.S(:,end)<0)'\n",
    "                                if isempty(model.rxnFormulas{end})\n",
    "                                    model.rxnFormulas{end} = sprintf('%d %s',-full(model.S(k,end)),model.mets{k});\n",
    "                                else\n",
    "                                    model.rxnFormulas{end} = [model.rxnFormulas{end},' + ',sprintf('%d %s',-full(model.S(k,end)),model.mets{k})];\n",
    "                                end\n",
    "                            end\n",
    "\n",
    "                            % reaction arrow\n",
    "                            if (model.lb(end) < 0) && (model.ub(end) > 0)\n",
    "                                model.rxnFormulas{end} = [model.rxnFormulas{end},' <=> '];\n",
    "                            elseif (model.lb(end) < 0)\n",
    "                                model.rxnFormulas{end} = [model.rxnFormulas{end},' <-- '];\n",
    "                            elseif (model.ub(end) > 0)\n",
    "                                model.rxnFormulas{end} = [model.rxnFormulas{end},' --> '];\n",
    "                            else\n",
    "                                model.rxnFormulas{end} = [model.rxnFormulas{end},' X--X '];\n",
    "                            end\n",
    "\n",
    "                            % products\n",
    "                            numproducts = 0;\n",
    "                            for k = find(model.S(:,end)>0)'\n",
    "                                if numproducts == 0\n",
    "                                    model.rxnFormulas{end} = [model.rxnFormulas{end},sprintf('%d %s',full(model.S(k,end)),model.mets{k})];\n",
    "                                else\n",
    "                                    model.rxnFormulas{end} = [model.rxnFormulas{end},' + ',sprintf('%d %s',full(model.S(k,end)),model.mets{k})];\n",
    "                                end\n",
    "                                numproducts = numproducts + 1;\n",
    "                            end\n",
    "                        end    \n",
    "\n",
    "                    % if no compartments\n",
    "                    else\n",
    "                       error('ERROR - Constraints - Custom - Could not find any compartments that contain all reaction metabolites') \n",
    "                    end\n",
    "\n",
    "                % if not all() reaction\n",
    "                else\n",
    "\n",
    "                    % create new reaction\n",
    "                    model.rxns{end+1} = sprintf('custom_maxmulti_%d',i);\n",
    "                    model.S(:,end+1) = zeros(length(model.mets),1);\n",
    "                    model.lb(end+1) = min(model.lb);\n",
    "                    model.ub(end+1) = max(model.ub);\n",
    "                    model.c(end+1) = 0;\n",
    "                    model.c_obj(end+1) = 0;\n",
    "                    model.c_fraction(end+1) = 0;\n",
    "                    model.c_maxmulti(end+1) = constraints_maxmulti_weight(i);\n",
    "                    model.rxnGeneMat(end+1,:) = zeros(1,length(model.genes));\n",
    "                    model.rules{end+1} = '';\n",
    "                    model.grRules{end+1} = '';\n",
    "                    model.subSystems{end+1} = 'Custom Constraint';\n",
    "                    model.rxnNames{end+1} = sprintf('Custom Max Multi Constraint %d',i);\n",
    "                    model.rxnKEGGID{end+1} = '';\n",
    "                    model.rxnKeggOrthology{end+1} = '';\n",
    "                    model.rxnConfidenceScores(end+1) = 0;\n",
    "                    model.rxnReferences{end+1} = '';\n",
    "                    model.rxnECNumbers{end+1} = '';\n",
    "                    model.rxnNotes{end+1} = '';\n",
    "                    model.rxnCOG{end+1} = '';\n",
    "                    model.rxnReconMap{end+1} = '';\n",
    "\n",
    "                    % parse reactants\n",
    "                    metstring = strsplit(strtrim(constraints_maxmulti_reaction{i}(1:(strfind(constraints_maxmulti_reaction{i},'<=>')-1))),' + ','CollapseDelimiters',true);\n",
    "                    for j = 1:length(metstring)\n",
    "                        if ~strcmp(metstring{j},'')\n",
    "                            metstring_ind = strsplit(metstring{j},' ');\n",
    "\n",
    "                            % metabolite\n",
    "                            if ~any(strcmp(model.mets,metstring_ind{2}))\n",
    "                                error('ERROR - Constraints - Custom - Unrecognized metabolite %s',metstring_ind{2})\n",
    "                            end\n",
    "\n",
    "                            % stoichiometry\n",
    "                            [num, status] = str2num(metstring_ind{1});\n",
    "                            if status == 1\n",
    "                                model.S(strcmp(model.mets,metstring_ind{2}),end) = -num;\n",
    "                            else\n",
    "                                error('ERROR - Constraints - Custom - Unrecognized stoichiometry %s',metstring_ind{1})\n",
    "                            end  \n",
    "                        end\n",
    "                    end\n",
    "\n",
    "                    % parse products\n",
    "                    metstring = strsplit(strtrim(constraints_maxmulti_reaction{i}((strfind(constraints_maxmulti_reaction{i},'<=>')+3):end)),' + ','CollapseDelimiters',true);\n",
    "                    for j = 1:length(metstring)\n",
    "                        if ~strcmp(metstring{j},'')\n",
    "                            metstring_ind = strsplit(metstring{j},' ');\n",
    "\n",
    "                            % metabolite\n",
    "                            if ~any(strcmp(model.mets,metstring_ind{2}))\n",
    "                                error('ERROR - Constraints - Custom - Unrecognized metabolite %s',metstring_ind{2})\n",
    "                            end\n",
    "\n",
    "                            % stoichiometry\n",
    "                            [num, status] = str2num(metstring_ind{1});\n",
    "                            if status == 1\n",
    "                                model.S(strcmp(model.mets,metstring_ind{2}),end) = num;\n",
    "                            else\n",
    "                                error('ERROR - Constraints - Custom - Unrecognized stoichiometry %s',metstring_ind{1})\n",
    "                            end   \n",
    "                        end\n",
    "                    end\n",
    "                    \n",
    "                    % initialize formula\n",
    "                    model.rxnFormulas{end+1} = '';\n",
    "\n",
    "                    % substrates\n",
    "                    for k = find(model.S(:,end)<0)'\n",
    "                        if isempty(model.rxnFormulas{end})\n",
    "                            model.rxnFormulas{end} = sprintf('%d %s',-full(model.S(k,end)),model.mets{k});\n",
    "                        else\n",
    "                            model.rxnFormulas{end} = [model.rxnFormulas{end},' + ',sprintf('%d %s',-full(model.S(k,end)),model.mets{k})];\n",
    "                        end\n",
    "                    end\n",
    "\n",
    "                    % reaction arrow\n",
    "                    if (model.lb(end) < 0) && (model.ub(end) > 0)\n",
    "                        model.rxnFormulas{end} = [model.rxnFormulas{end},' <=> '];\n",
    "                    elseif (model.lb(end) < 0)\n",
    "                        model.rxnFormulas{end} = [model.rxnFormulas{end},' <-- '];\n",
    "                    elseif (model.ub(end) > 0)\n",
    "                        model.rxnFormulas{end} = [model.rxnFormulas{end},' --> '];\n",
    "                    else\n",
    "                        model.rxnFormulas{end} = [model.rxnFormulas{end},' X--X '];\n",
    "                    end\n",
    "\n",
    "                    % products\n",
    "                    numproducts = 0;\n",
    "                    for k = find(model.S(:,end)>0)'\n",
    "                        if numproducts == 0\n",
    "                            model.rxnFormulas{end} = [model.rxnFormulas{end},sprintf('%d %s',full(model.S(k,end)),model.mets{k})];\n",
    "                        else\n",
    "                            model.rxnFormulas{end} = [model.rxnFormulas{end},' + ',sprintf('%d %s',full(model.S(k,end)),model.mets{k})];\n",
    "                        end\n",
    "                        numproducts = numproducts + 1;\n",
    "                    end\n",
    "                end\n",
    "            end\n",
    "\n",
    "        % if not custom reaction\n",
    "        elseif any(strcmp(model.rxns,constraints_maxmulti_reaction{i}))\n",
    "            \n",
    "            % change model.c_maxmulti\n",
    "            model.c_maxmulti(strcmp(model.rxns,constraints_maxmulti_reaction{i})) = constraints_maxmulti_weight(i);\n",
    "\n",
    "        else\n",
    "            error('ERROR - Constraints - Custom - Either invalid reaction formula or non-matching reaction name');\n",
    "        end\n",
    " \n",
    "    end\n",
    "\n",
    "    % implement objective functions\n",
    "    for i = 1:length(objective_function_reaction)\n",
    "\n",
    "        % if custom reaction\n",
    "        if (contains(objective_function_reaction{i},'-->')) || (contains(objective_function_reaction{i},'<=>'))\n",
    "\n",
    "            % irreversible reaction\n",
    "            if (contains(objective_function_reaction{i},'-->'))\n",
    "\n",
    "                % determine if all() reaction\n",
    "                if strcmp(objective_function_reaction{i}(1:4),'all(') && strcmp(objective_function_reaction{i}(end),')')\n",
    "\n",
    "                    % reaction formula without all()\n",
    "                    rxn_no_all = objective_function_reaction{i}(5:end-1);\n",
    "\n",
    "                    % save reactant and product list\n",
    "                    reactants = {};\n",
    "                    reactant_stoich = [];\n",
    "                    reactant_compartments = {};\n",
    "                    products = {};\n",
    "                    product_stoich = [];\n",
    "                    product_compartments = {};\n",
    "\n",
    "                    % parse reactants\n",
    "                    metstring = strsplit(strtrim(rxn_no_all(1:(strfind(rxn_no_all,'-->')-1))),' + ','CollapseDelimiters',true);\n",
    "                    for j = 1:length(metstring)\n",
    "                        if ~strcmp(metstring{j},'')\n",
    "                            metstring_ind = strsplit(metstring{j},' ');\n",
    "\n",
    "                            % get matching metabolite and compartments\n",
    "                            found_compartments = {};\n",
    "                            for k = 1:length(model.mets)\n",
    "                                if strcmp(model.mets{k}(1:end-3),metstring_ind{2}(1:end-2))\n",
    "                                    found_compartments{end+1} = model.mets{k}(end-1);\n",
    "                                end\n",
    "                            end\n",
    "\n",
    "                            % if found compartments\n",
    "                            if length(found_compartments) > 0\n",
    "\n",
    "                                % add metabolite\n",
    "                                reactants{end+1} = metstring_ind{2}(1:end-2);\n",
    "                                reactant_compartments{end+1} = found_compartments;\n",
    "\n",
    "                                % add stoichiometry\n",
    "                                [num, status] = str2num(metstring_ind{1});\n",
    "                                if status == 1\n",
    "                                    reactant_stoich(end+1) = num;\n",
    "                                else\n",
    "                                    error('ERROR - Objective Function - Unrecognized stoichiometry %s',metstring_ind{1})\n",
    "                                end\n",
    "\n",
    "                            % if no found compartments\n",
    "                            else\n",
    "                                error('ERROR - Objective Function - Could not find any compartments for metabolite %s',metstring_ind{1}(1:end-2))\n",
    "                            end \n",
    "                        end\n",
    "                    end\n",
    "\n",
    "                    % parse products\n",
    "                    metstring = strsplit(strtrim(rxn_no_all((strfind(rxn_no_all,'-->')+3):end)),' + ','CollapseDelimiters',true);\n",
    "                    for j = 1:length(metstring)\n",
    "                        if ~strcmp(metstring{j},'')\n",
    "                            metstring_ind = strsplit(metstring{j},' ');\n",
    "\n",
    "                            % get matching metabolite and compartments\n",
    "                            found_compartments = {};\n",
    "                            for k = 1:length(model.mets)\n",
    "                                if strcmp(model.mets{k}(1:end-3),metstring_ind{2}(1:end-2))\n",
    "                                    found_compartments{end+1} = model.mets{k}(end-1);\n",
    "                                end\n",
    "                            end\n",
    "\n",
    "                            % if found compartments\n",
    "                            if length(found_compartments) > 0\n",
    "\n",
    "                                % add metabolite\n",
    "                                products{end+1} = metstring_ind{2}(1:end-2);\n",
    "                                product_compartments{end+1} = found_compartments;\n",
    "\n",
    "                                % add stoichiometry\n",
    "                                [num, status] = str2num(metstring_ind{1});\n",
    "                                if status == 1\n",
    "                                    product_stoich(end+1) = num;\n",
    "                                else\n",
    "                                    error('ERROR - Objective Function - Unrecognized stoichiometry %s',metstring_ind{1})\n",
    "                                end\n",
    "\n",
    "                            % if no found compartments\n",
    "                            else\n",
    "                                error('ERROR - Objective Function - Could not find any compartments for metabolite %s',metstring_ind{1}(1:end-2))\n",
    "                            end \n",
    "                        end\n",
    "                    end\n",
    "\n",
    "                    % find compartments for all reactants and products\n",
    "                    all_compartments = {'c','e','g','l','m','n','r','x','i'};\n",
    "                    for j = 1:length(reactant_compartments)\n",
    "                        all_compartments = intersect(all_compartments,reactant_compartments{j});\n",
    "                    end\n",
    "                    for j = 1:length(product_compartments)\n",
    "                        all_compartments = intersect(all_compartments,product_compartments{j});\n",
    "                    end\n",
    "\n",
    "                    % if at least one compartment\n",
    "                    if length(all_compartments) > 0\n",
    "\n",
    "                        % create new reaction for all compartments\n",
    "                        for j = 1:length(all_compartments)\n",
    "                            model.rxns{end+1} = sprintf('custom_objective_%d_%s',i,all_compartments{j});\n",
    "                            model.S(:,end+1) = zeros(length(model.mets),1);\n",
    "                            model.lb(end+1) = 0;\n",
    "                            model.ub(end+1) = max(model.ub);\n",
    "                            if strcmp(objective_function_direction{i},'MAX')\n",
    "                                model.c(end+1) = objective_function_weight{i};\n",
    "                                model.c_obj(end+1) = objective_function_weight{i};\n",
    "                            elseif strcmp(objective_function_direction{i},'MIN')\n",
    "                                model.c(end+1) = -objective_function_weight{i};\n",
    "                                model.c_obj(end+1) = -objective_function_weight{i};\n",
    "                            end\n",
    "                            model.c_fraction(end+1) = 0;\n",
    "                            model.c_maxmulti(end+1) = 0;\n",
    "                            model.rxnGeneMat(end+1,:) = zeros(1,length(model.genes));\n",
    "                            model.rules{end+1} = '';\n",
    "                            model.grRules{end+1} = '';\n",
    "                            model.subSystems{end+1} = 'Custom Objective'\n",
    "                            model.rxnNames{end+1} = sprintf('Custom Objective Function %d - Compartment %s',i,all_compartments{j});\n",
    "                            model.rxnKEGGID{end+1} = '';\n",
    "                            model.rxnKeggOrthology{end+1} = '';\n",
    "                            model.rxnConfidenceScores(end+1) = 0;\n",
    "                            model.rxnReferences{end+1} = '';\n",
    "                            model.rxnECNumbers{end+1} = '';\n",
    "                            model.rxnNotes{end+1} = '';\n",
    "                            model.rxnCOG{end+1} = '';\n",
    "                            model.rxnReconMap{end+1} = '';\n",
    "\n",
    "                            % reactants\n",
    "                            for k = 1:length(reactants)\n",
    "                                metstring = sprintf('%s[%s]',reactants{k},all_compartments{j});\n",
    "                                model.S(strcmp(model.mets,metstring),end) = -reactant_stoich(k);\n",
    "                            end\n",
    "\n",
    "                            % products\n",
    "                            for k = 1:length(products)\n",
    "                                metstring = sprintf('%s[%s]',products{k},all_compartments{j});\n",
    "                                model.S(strcmp(model.mets,metstring),end) = product_stoich(k);\n",
    "                            end\n",
    "                            \n",
    "                            % initialize formula\n",
    "                            model.rxnFormulas{end+1} = '';\n",
    "\n",
    "                            % substrates\n",
    "                            for k = find(model.S(:,end)<0)'\n",
    "                                if isempty(model.rxnFormulas{end})\n",
    "                                    model.rxnFormulas{end} = sprintf('%d %s',-full(model.S(k,end)),model.mets{k});\n",
    "                                else\n",
    "                                    model.rxnFormulas{end} = [model.rxnFormulas{end},' + ',sprintf('%d %s',-full(model.S(k,end)),model.mets{k})];\n",
    "                                end\n",
    "                            end\n",
    "\n",
    "                            % reaction arrow\n",
    "                            if (model.lb(end) < 0) && (model.ub(end) > 0)\n",
    "                                model.rxnFormulas{end} = [model.rxnFormulas{end},' <=> '];\n",
    "                            elseif (model.lb(end) < 0)\n",
    "                                model.rxnFormulas{end} = [model.rxnFormulas{end},' <-- '];\n",
    "                            elseif (model.ub(end) > 0)\n",
    "                                model.rxnFormulas{end} = [model.rxnFormulas{end},' --> '];\n",
    "                            else\n",
    "                                model.rxnFormulas{end} = [model.rxnFormulas{end},' X--X '];\n",
    "                            end\n",
    "\n",
    "                            % products\n",
    "                            numproducts = 0;\n",
    "                            for k = find(model.S(:,end)>0)'\n",
    "                                if numproducts == 0\n",
    "                                    model.rxnFormulas{end} = [model.rxnFormulas{end},sprintf('%d %s',full(model.S(k,end)),model.mets{k})];\n",
    "                                else\n",
    "                                    model.rxnFormulas{end} = [model.rxnFormulas{end},' + ',sprintf('%d %s',full(model.S(k,end)),model.mets{k})];\n",
    "                                end\n",
    "                                numproducts = numproducts + 1;\n",
    "                            end\n",
    "                        end    \n",
    "\n",
    "                    % if no compartments\n",
    "                    else\n",
    "                       error('ERROR - Objective Function - Could not find any compartments that contain all reaction metabolites') \n",
    "                    end\n",
    "\n",
    "                % if not all() reaction\n",
    "                else\n",
    "\n",
    "                    % create new reaction\n",
    "                    model.rxns{end+1} = sprintf('custom_objective_%d',i);\n",
    "                    model.S(:,end+1) = zeros(length(model.mets),1);\n",
    "                    model.lb(end+1) = 0;\n",
    "                    model.ub(end+1) = max(model.ub);\n",
    "                    if strcmp(objective_function_direction{i},'MAX')\n",
    "                        model.c(end+1) = objective_function_weight{i};\n",
    "                        model.c_obj(end+1) = objective_function_weight{i};\n",
    "                    elseif strcmp(objective_function_direction{i},'MIN')\n",
    "                        model.c(end+1) = -objective_function_weight{i};\n",
    "                        model.c_obj(end+1) = -objective_function_weight{i};\n",
    "                    end\n",
    "                    model.c_fraction(end+1) = 0;\n",
    "                    model.c_maxmulti(end+1) = 0;\n",
    "                    model.rxnGeneMat(end+1,:) = zeros(1,length(model.genes));\n",
    "                    model.rules{end+1} = '';\n",
    "                    model.grRules{end+1} = '';\n",
    "                    model.subSystems{end+1} = 'Custom Objective';\n",
    "                    model.rxnNames{end+1} = sprintf('Custom Objective Function %d',i);\n",
    "                    model.rxnKEGGID{end+1} = '';\n",
    "                    model.rxnKeggOrthology{end+1} = '';\n",
    "                    model.rxnConfidenceScores(end+1) = 0;\n",
    "                    model.rxnReferences{end+1} = '';\n",
    "                    model.rxnECNumbers{end+1} = '';\n",
    "                    model.rxnNotes{end+1} = '';\n",
    "                    model.rxnCOG{end+1} = '';\n",
    "                    model.rxnReconMap{end+1} = '';\n",
    "\n",
    "                    % parse reactants\n",
    "                    metstring = strsplit(strtrim(objective_function_reaction{i}(1:(strfind(objective_function_reaction{i},'-->')-1))),' + ','CollapseDelimiters',true);\n",
    "                    for j = 1:length(metstring)\n",
    "                        if ~strcmp(metstring{j},'')\n",
    "                            metstring_ind = strsplit(metstring{j},' ');\n",
    "\n",
    "                            % metabolite\n",
    "                            if ~any(strcmp(model.mets,metstring_ind{2}))\n",
    "                                error('ERROR - Objective Function - Unrecognized metabolite %s',metstring_ind{2})\n",
    "                            end\n",
    "\n",
    "                            % stoichiometry\n",
    "                            [num, status] = str2num(metstring_ind{1});\n",
    "                            if status == 1\n",
    "                                model.S(strcmp(model.mets,metstring_ind{2}),end) = -num;\n",
    "                            else\n",
    "                                error('ERROR - Objective Function - Unrecognized stoichiometry %s',metstring_ind{1})\n",
    "                            end  \n",
    "                        end\n",
    "                    end\n",
    "\n",
    "                    % parse products\n",
    "                    metstring = strsplit(strtrim(objective_function_reaction{i}((strfind(objective_function_reaction{i},'-->')+3):end)),' + ','CollapseDelimiters',true);\n",
    "                    for j = 1:length(metstring)\n",
    "                        if ~strcmp(metstring{j},'')\n",
    "                            metstring_ind = strsplit(metstring{j},' ');\n",
    "\n",
    "                            % metabolite\n",
    "                            if ~any(strcmp(model.mets,metstring_ind{2}))\n",
    "                                error('ERROR - Objective Function - Unrecognized metabolite %s',metstring_ind{2})\n",
    "                            end\n",
    "\n",
    "                            % stoichiometry\n",
    "                            [num, status] = str2num(metstring_ind{1});\n",
    "                            if status == 1\n",
    "                                model.S(strcmp(model.mets,metstring_ind{2}),end) = num;\n",
    "                            else\n",
    "                                error('ERROR - Objective Function - Unrecognized stoichiometry %s',metstring_ind{1})\n",
    "                            end   \n",
    "                        end\n",
    "                    end\n",
    "                    \n",
    "                    % initialize formula\n",
    "                    model.rxnFormulas{end+1} = '';\n",
    "\n",
    "                    % substrates\n",
    "                    for k = find(model.S(:,end)<0)'\n",
    "                        if isempty(model.rxnFormulas{end})\n",
    "                            model.rxnFormulas{end} = sprintf('%d %s',-full(model.S(k,end)),model.mets{k});\n",
    "                        else\n",
    "                            model.rxnFormulas{end} = [model.rxnFormulas{end},' + ',sprintf('%d %s',-full(model.S(k,end)),model.mets{k})];\n",
    "                        end\n",
    "                    end\n",
    "\n",
    "                    % reaction arrow\n",
    "                    if (model.lb(end) < 0) && (model.ub(end) > 0)\n",
    "                        model.rxnFormulas{end} = [model.rxnFormulas{end},' <=> '];\n",
    "                    elseif (model.lb(end) < 0)\n",
    "                        model.rxnFormulas{end} = [model.rxnFormulas{end},' <-- '];\n",
    "                    elseif (model.ub(end) > 0)\n",
    "                        model.rxnFormulas{end} = [model.rxnFormulas{end},' --> '];\n",
    "                    else\n",
    "                        model.rxnFormulas{end} = [model.rxnFormulas{end},' X--X '];\n",
    "                    end\n",
    "\n",
    "                    % products\n",
    "                    numproducts = 0;\n",
    "                    for k = find(model.S(:,end)>0)'\n",
    "                        if numproducts == 0\n",
    "                            model.rxnFormulas{end} = [model.rxnFormulas{end},sprintf('%d %s',full(model.S(k,end)),model.mets{k})];\n",
    "                        else\n",
    "                            model.rxnFormulas{end} = [model.rxnFormulas{end},' + ',sprintf('%d %s',full(model.S(k,end)),model.mets{k})];\n",
    "                        end\n",
    "                        numproducts = numproducts + 1;\n",
    "                    end\n",
    "                end\n",
    "\n",
    "            % reversible reaction\n",
    "            else\n",
    "                % determine if all() reaction\n",
    "                if strcmp(objective_function_reaction{i}(1:4),'all(') && strcmp(objective_function_reaction{i}(end),')')\n",
    "\n",
    "                    % reaction formula without all()\n",
    "                    rxn_no_all = objective_function_reaction{i}(5:end-1);\n",
    "\n",
    "                    % save reactant and product list\n",
    "                    reactants = {};\n",
    "                    reactant_stoich = [];\n",
    "                    reactant_compartments = {};\n",
    "                    products = {};\n",
    "                    product_stoich = [];\n",
    "                    product_compartments = {};\n",
    "\n",
    "                    % parse reactants\n",
    "                    metstring = strsplit(strtrim(rxn_no_all(1:(strfind(rxn_no_all,'<=>')-1))),' + ','CollapseDelimiters',true);\n",
    "                    for j = 1:length(metstring)\n",
    "                        if ~strcmp(metstring{j},'')\n",
    "                            metstring_ind = strsplit(metstring{j},' ');\n",
    "\n",
    "                            % get matching metabolite and compartments\n",
    "                            found_compartments = {};\n",
    "                            for k = 1:length(model.mets)\n",
    "                                if strcmp(model.mets{k}(1:end-3),metstring_ind{2}(1:end-2))\n",
    "                                    found_compartments{end+1} = model.mets{k}(end-1);\n",
    "                                end\n",
    "                            end\n",
    "\n",
    "                            % if found compartments\n",
    "                            if length(found_compartments) > 0\n",
    "\n",
    "                                % add metabolite\n",
    "                                reactants{end+1} = metstring_ind{2}(1:end-2);\n",
    "                                reactant_compartments{end+1} = found_compartments;\n",
    "\n",
    "                                % add stoichiometry\n",
    "                                [num, status] = str2num(metstring_ind{1});\n",
    "                                if status == 1\n",
    "                                    reactant_stoich(end+1) = num;\n",
    "                                else\n",
    "                                    error('ERROR - Objective Function - Unrecognized stoichiometry %s',metstring_ind{1})\n",
    "                                end\n",
    "\n",
    "                            % if no found compartments\n",
    "                            else\n",
    "                                error('ERROR - Objective Function - Could not find any compartments for metabolite %s',metstring_ind{1}(1:end-2))\n",
    "                            end \n",
    "                        end\n",
    "                    end\n",
    "\n",
    "                    % parse products\n",
    "                    metstring = strsplit(strtrim(rxn_no_all((strfind(rxn_no_all,'<=>')+3):end)),' + ','CollapseDelimiters',true);\n",
    "                    for j = 1:length(metstring)\n",
    "                        if ~strcmp(metstring{j},'')\n",
    "                            metstring_ind = strsplit(metstring{j},' ');\n",
    "\n",
    "                            % get matching metabolite and compartments\n",
    "                            found_compartments = {};\n",
    "                            for k = 1:length(model.mets)\n",
    "                                if strcmp(model.mets{k}(1:end-3),metstring_ind{2}(1:end-2))\n",
    "                                    found_compartments{end+1} = model.mets{k}(end-1);\n",
    "                                end\n",
    "                            end\n",
    "\n",
    "                            % if found compartments\n",
    "                            if length(found_compartments) > 0\n",
    "\n",
    "                                % add metabolite\n",
    "                                products{end+1} = metstring_ind{2}(1:end-2);\n",
    "                                product_compartments{end+1} = found_compartments;\n",
    "\n",
    "                                % add stoichiometry\n",
    "                                [num, status] = str2num(metstring_ind{1});\n",
    "                                if status == 1\n",
    "                                    product_stoich(end+1) = num;\n",
    "                                else\n",
    "                                    error('ERROR - Objective Function - Unrecognized stoichiometry %s',metstring_ind{1})\n",
    "                                end\n",
    "\n",
    "                            % if no found compartments\n",
    "                            else\n",
    "                                error('ERROR - Objective Function - Could not find any compartments for metabolite %s',metstring_ind{1}(1:end-2))\n",
    "                            end \n",
    "                        end\n",
    "                    end\n",
    "\n",
    "                    % find compartments for all reactants and products\n",
    "                    all_compartments = {'c','e','g','l','m','n','r','x','i'};\n",
    "                    for j = 1:length(reactant_compartments)\n",
    "                        all_compartments = intersect(all_compartments,reactant_compartments{j});\n",
    "                    end\n",
    "                    for j = 1:length(product_compartments)\n",
    "                        all_compartments = intersect(all_compartments,product_compartments{j});\n",
    "                    end\n",
    "\n",
    "                    % if at least one compartment\n",
    "                    if length(all_compartments) > 0\n",
    "\n",
    "                        % create new reaction for all compartments\n",
    "                        for j = 1:length(all_compartments)\n",
    "                            model.rxns{end+1} = sprintf('custom_objective_%d_%s',i,all_compartments{j});\n",
    "                            model.S(:,end+1) = zeros(length(model.mets),1);\n",
    "                            model.lb(end+1) = min(model.lb);\n",
    "                            model.ub(end+1) = max(model.ub);\n",
    "                            if strcmp(objective_function_direction{i},'MAX')\n",
    "                                model.c(end+1) = objective_function_weight{i};\n",
    "                                model.c_obj(end+1) = objective_function_weight{i};\n",
    "                            elseif strcmp(objective_function_direction{i},'MIN')\n",
    "                                model.c(end+1) = -objective_function_weight{i};\n",
    "                                model.c_obj(end+1) = -objective_function_weight{i};\n",
    "                            end\n",
    "                            model.c_fraction(end+1) = 0;\n",
    "                            model.c_maxmulti(end+1) = 0;\n",
    "                            model.rxnGeneMat(end+1,:) = zeros(1,length(model.genes));\n",
    "                            model.rules{end+1} = '';\n",
    "                            model.grRules{end+1} = '';\n",
    "                            model.subSystems{end+1} = 'Custom Objective';\n",
    "                            model.rxnNames{end+1} = sprintf('Custom Objective Function %d - Compartment %s',i,all_compartments{j});\n",
    "                            model.rxnKEGGID{end+1} = '';\n",
    "                            model.rxnKeggOrthology{end+1} = '';\n",
    "                            model.rxnConfidenceScores(end+1) = 0;\n",
    "                            model.rxnReferences{end+1} = '';\n",
    "                            model.rxnECNumbers{end+1} = '';\n",
    "                            model.rxnNotes{end+1} = '';\n",
    "                            model.rxnCOG{end+1} = '';\n",
    "                            model.rxnReconMap{end+1} = '';\n",
    "\n",
    "                            % reactants\n",
    "                            for k = 1:length(reactants)\n",
    "                                metstring = sprintf('%s[%s]',reactants{k},all_compartments{j});\n",
    "                                model.S(strcmp(model.mets,metstring),end) = -reactant_stoich(k);\n",
    "                            end\n",
    "\n",
    "                            % products\n",
    "                            for k = 1:length(products)\n",
    "                                metstring = sprintf('%s[%s]',products{k},all_compartments{j});\n",
    "                                model.S(strcmp(model.mets,metstring),end) = product_stoich(k);\n",
    "                            end\n",
    "                            \n",
    "                            % initialize formula\n",
    "                            model.rxnFormulas{end+1} = '';\n",
    "\n",
    "                            % substrates\n",
    "                            for k = find(model.S(:,end)<0)'\n",
    "                                if isempty(model.rxnFormulas{end})\n",
    "                                    model.rxnFormulas{end} = sprintf('%d %s',-full(model.S(k,end)),model.mets{k});\n",
    "                                else\n",
    "                                    model.rxnFormulas{end} = [model.rxnFormulas{end},' + ',sprintf('%d %s',-full(model.S(k,end)),model.mets{k})];\n",
    "                                end\n",
    "                            end\n",
    "\n",
    "                            % reaction arrow\n",
    "                            if (model.lb(end) < 0) && (model.ub(end) > 0)\n",
    "                                model.rxnFormulas{end} = [model.rxnFormulas{end},' <=> '];\n",
    "                            elseif (model.lb(end) < 0)\n",
    "                                model.rxnFormulas{end} = [model.rxnFormulas{end},' <-- '];\n",
    "                            elseif (model.ub(end) > 0)\n",
    "                                model.rxnFormulas{end} = [model.rxnFormulas{end},' --> '];\n",
    "                            else\n",
    "                                model.rxnFormulas{end} = [model.rxnFormulas{end},' X--X '];\n",
    "                            end\n",
    "\n",
    "                            % products\n",
    "                            numproducts = 0;\n",
    "                            for k = find(model.S(:,end)>0)'\n",
    "                                if numproducts == 0\n",
    "                                    model.rxnFormulas{end} = [model.rxnFormulas{end},sprintf('%d %s',full(model.S(k,end)),model.mets{k})];\n",
    "                                else\n",
    "                                    model.rxnFormulas{end} = [model.rxnFormulas{end},' + ',sprintf('%d %s',full(model.S(k,end)),model.mets{k})];\n",
    "                                end\n",
    "                                numproducts = numproducts + 1;\n",
    "                            end\n",
    "                        end    \n",
    "\n",
    "                    % if no compartments\n",
    "                    else\n",
    "                       error('ERROR - Objective Function - Could not find any compartments that contain all reaction metabolites') \n",
    "                    end\n",
    "\n",
    "                % if not all() reaction\n",
    "                else\n",
    "\n",
    "                    % create new reaction\n",
    "                    model.rxns{end+1} = sprintf('custom_objective_%d',i);\n",
    "                    model.S(:,end+1) = zeros(length(model.mets),1);\n",
    "                    model.lb(end+1) = min(model.lb);\n",
    "                    model.ub(end+1) = max(model.ub);\n",
    "                    if strcmp(objective_function_direction{i},'MAX')\n",
    "                        model.c(end+1) = objective_function_weight{i};\n",
    "                        model.c_obj(end+1) = objective_function_weight{i};\n",
    "                    elseif strcmp(objective_function_direction{i},'MIN')\n",
    "                        model.c(end+1) = -objective_function_weight{i};\n",
    "                        model.c_obj(end+1) = -objective_function_weight{i};\n",
    "                    end\n",
    "                    model.c_fraction(end+1) = 0;\n",
    "                        model.c_maxmulti(end+1) = 0;\n",
    "                    model.rxnGeneMat(end+1,:) = zeros(1,length(model.genes));\n",
    "                    model.rules{end+1} = '';\n",
    "                    model.grRules{end+1} = '';\n",
    "                    model.subSystems{end+1} = 'Custom Objective';\n",
    "                    model.rxnNames{end+1} = 'Custom Objective Function';\n",
    "                    model.rxnKEGGID{end+1} = '';\n",
    "                    model.rxnKeggOrthology{end+1} = '';\n",
    "                    model.rxnConfidenceScores(end+1) = 0;\n",
    "                    model.rxnReferences{end+1} = '';\n",
    "                    model.rxnECNumbers{end+1} = '';\n",
    "                    model.rxnNotes{end+1} = '';\n",
    "                    model.rxnCOG{end+1} = '';\n",
    "                    model.rxnReconMap{end+1} = '';\n",
    "\n",
    "                    % parse reactants\n",
    "                    metstring = strsplit(strtrim(objective_function_reaction{i}(1:(strfind(objective_function_reaction{i},'<=>')-1))),' + ','CollapseDelimiters',true);\n",
    "                    for j = 1:length(metstring)\n",
    "                        if ~strcmp(metstring{j},'')\n",
    "                            metstring_ind = strsplit(metstring{j},' ');\n",
    "\n",
    "                            % metabolite\n",
    "                            if ~any(strcmp(model.mets,metstring_ind{2}))\n",
    "                                error('ERROR - Objective Function - Unrecognized metabolite %s',metstring_ind{2})\n",
    "                            end\n",
    "\n",
    "                            % stoichiometry\n",
    "                            [num, status] = str2num(metstring_ind{1});\n",
    "                            if status == 1\n",
    "                                model.S(strcmp(model.mets,metstring_ind{2}),end) = -num;\n",
    "                            else\n",
    "                                error('ERROR - Objective Function - Unrecognized stoichiometry %s',metstring_ind{1})\n",
    "                            end  \n",
    "                        end\n",
    "                    end\n",
    "\n",
    "                    % parse products\n",
    "                    metstring = strsplit(strtrim(objective_function_reaction{i}((strfind(objective_function_reaction{i},'<=>')+3):end)),' + ','CollapseDelimiters',true);\n",
    "                    for j = 1:length(metstring)\n",
    "                        if ~strcmp(metstring{j},'')\n",
    "                            metstring_ind = strsplit(metstring{j},' ');\n",
    "\n",
    "                            % metabolite\n",
    "                            if ~any(strcmp(model.mets,metstring_ind{2}))\n",
    "                                error('ERROR - Objective Function - Unrecognized metabolite %s',metstring_ind{2})\n",
    "                            end\n",
    "\n",
    "                            % stoichiometry\n",
    "                            [num, status] = str2num(metstring_ind{1});\n",
    "                            if status == 1\n",
    "                                model.S(strcmp(model.mets,metstring_ind{2}),end) = num;\n",
    "                            else\n",
    "                                error('ERROR - Objective Function - Unrecognized stoichiometry %s',metstring_ind{1})\n",
    "                            end   \n",
    "                        end\n",
    "                    end\n",
    "                    \n",
    "                    % initialize formula\n",
    "                    model.rxnFormulas{end+1} = '';\n",
    "\n",
    "                    % substrates\n",
    "                    for k = find(model.S(:,end)<0)'\n",
    "                        if isempty(model.rxnFormulas{end})\n",
    "                            model.rxnFormulas{end} = sprintf('%d %s',-full(model.S(k,end)),model.mets{k});\n",
    "                        else\n",
    "                            model.rxnFormulas{end} = [model.rxnFormulas{end},' + ',sprintf('%d %s',-full(model.S(k,end)),model.mets{k})];\n",
    "                        end\n",
    "                    end\n",
    "\n",
    "                    % reaction arrow\n",
    "                    if (model.lb(end) < 0) && (model.ub(end) > 0)\n",
    "                        model.rxnFormulas{end} = [model.rxnFormulas{end},' <=> '];\n",
    "                    elseif (model.lb(end) < 0)\n",
    "                        model.rxnFormulas{end} = [model.rxnFormulas{end},' <-- '];\n",
    "                    elseif (model.ub(end) > 0)\n",
    "                        model.rxnFormulas{end} = [model.rxnFormulas{end},' --> '];\n",
    "                    else\n",
    "                        model.rxnFormulas{end} = [model.rxnFormulas{end},' X--X '];\n",
    "                    end\n",
    "\n",
    "                    % products\n",
    "                    numproducts = 0;\n",
    "                    for k = find(model.S(:,end)>0)'\n",
    "                        if numproducts == 0\n",
    "                            model.rxnFormulas{end} = [model.rxnFormulas{end},sprintf('%d %s',full(model.S(k,end)),model.mets{k})];\n",
    "                        else\n",
    "                            model.rxnFormulas{end} = [model.rxnFormulas{end},' + ',sprintf('%d %s',full(model.S(k,end)),model.mets{k})];\n",
    "                        end\n",
    "                        numproducts = numproducts + 1;\n",
    "                    end\n",
    "                end\n",
    "            end\n",
    "\n",
    "        % if not custom reaction\n",
    "        elseif any(strcmp(model.rxns,objective_function_reaction{i}))\n",
    "\n",
    "            if strcmp(objective_function_direction{i},'MAX')\n",
    "                model.c(strcmp(model.rxns,objective_function_reaction{i})) = objective_function_weight{i};\n",
    "                model.c_obj(strcmp(model.rxns,objective_function_reaction{i})) = objective_function_weight{i};\n",
    "            elseif strcmp(objective_function_direction{i},'MIN')\n",
    "                model.c(strcmp(model.rxns,objective_function_reaction{i})) = -objective_function_weight{i};\n",
    "                model.c_obj(strcmp(model.rxns,objective_function_reaction{i})) = -objective_function_weight{i};\n",
    "            end\n",
    "\n",
    "        else\n",
    "            error('ERROR - Objective Function - Either invalid reaction formula or non-matching reaction name');\n",
    "        end\n",
    "\n",
    "    end\n",
    "end"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Load Media"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%file load_media.m\n",
    "\n",
    "function [model,concentration_mets,concentration_values,exchange_rxn] = load_media(model,media_choice,media_constraint_metabolite,media_constraint_uptake,concentration_mets,concentration_values,module_5fu,module_cis,module_cpa,module_dox)\n",
    "\n",
    "    % get list of available media files\n",
    "    media_dir = dir('../data/media/output/*.csv');\n",
    "    available_media = {};\n",
    "    for i = 1:length(media_dir)\n",
    "        available_media{end+1} = media_dir(i).name;\n",
    "    end\n",
    "\n",
    "    % check if media files exists\n",
    "    for i = 1:length(media_choice)\n",
    "        if ~any(strcmp(available_media,sprintf('%s.csv',media_choice{i})))\n",
    "            error('ERROR - Can''t find media file %s.csv',media_choice{i})\n",
    "        end\n",
    "    end\n",
    "\n",
    "    % initialize exchange reactions and flux values\n",
    "    exchange_rxn = {};\n",
    "    exchange_value = [];\n",
    "\n",
    "    % iterate over media files\n",
    "    for i = 1:length(media_choice)\n",
    "\n",
    "        % load media file\n",
    "        f = fopen(sprintf('../data/media/output/%s.csv',media_choice{i}),'r');\n",
    "        data = textscan(f,'%s %s %f %f','Delimiter',',','headerLines',1);\n",
    "        fclose(f);\n",
    "\n",
    "        % add to exchange reactions\n",
    "        for j = 1:length(data{2})\n",
    "            if ~any(strcmp(exchange_rxn,data{2}{j}))\n",
    "                exchange_rxn{end+1} = data{2}{j};\n",
    "            end\n",
    "        end\n",
    "        \n",
    "        % add to concentration list\n",
    "        for j = 1:length(data{1})\n",
    "            if ~any(strcmp(concentration_mets,data{1}{j}))\n",
    "                concentration_mets{end+1} = data{1}{j};\n",
    "                concentration_values(end+1) = data{3}(j);\n",
    "            else\n",
    "                concentration_values(strcmp(concentration_mets,data{1}{j})) = concentration_values(strcmp(concentration_mets,data{1}{j})) + data{3}(j);\n",
    "            end\n",
    "        end\n",
    "        \n",
    "    end\n",
    "\n",
    "    % add exchange reactions for oxygen, carbon dioxide, water, protons, hydroxyl\n",
    "    exchange_rxn = [exchange_rxn,{'EX_o2[e]','EX_co2[e]','EX_h2o[e]','EX_h[e]','EX_oh1[e]'}];\n",
    "    \n",
    "    % add exchange reactions for drug modules\n",
    "    if module_5fu == 1\n",
    "        exchange_rxn = [exchange_rxn,{'EX_5fu[e]'}];\n",
    "    end\n",
    "    if module_cis == 1\n",
    "        exchange_rxn = [exchange_rxn,{'EX_cis[e]'}];\n",
    "    end\n",
    "    if module_cpa == 1\n",
    "        exchange_rxn = [exchange_rxn,{'EX_cpa[e]'}];\n",
    "    end\n",
    "    if module_dox == 1\n",
    "        exchange_rxn = [exchange_rxn,{'EX_doxQ[e]'}];\n",
    "    end\n",
    "    \n",
    "    % parse specific media constraints\n",
    "    media_constraint_exchange = {};\n",
    "    for i = 1:length(media_constraint_metabolite)\n",
    "        if ~any(strcmp(model.mets,sprintf('%s[e]',media_constraint_metabolite{i})))\n",
    "            error('ERROR - Media - Specific Media Constraints - Extracellular metabolite %s not found',media_constraint_metabolite{i})\n",
    "        elseif ~any(strcmp(model.rxns,sprintf('EX_%s[e]',media_constraint_metabolite{i})))\n",
    "            error('ERROR - Media - Specific Media Constraints - Exchange reaction for metabolite %s not found',media_constraint_metabolite{i})\n",
    "        elseif ~any(strcmp(exchange_rxn,sprintf('EX_%s[e]',media_constraint_metabolite{i})))\n",
    "            error('ERROR - Media - Specific Media Constraints - Metabolite %s not found in selected medium makeup',media_constraint_metabolite{i})\n",
    "        else\n",
    "            media_constraint_exchange{end+1} = sprintf('EX_%s[e]',media_constraint_metabolite{i});\n",
    "        end\n",
    "    end\n",
    "    \n",
    "    % restrict input exchange reactions to those in media\n",
    "    for i = 1:length(model.rxns)\n",
    "        if length(model.rxns{i}) >= 3\n",
    "            if strcmp(model.rxns{i}(1:3),'EX_')\n",
    "                if ~any(strcmp(exchange_rxn,model.rxns{i}))\n",
    "                    model.lb(i) = 0;\n",
    "                else\n",
    "                    if any(strcmp(media_constraint_exchange,model.rxns{i}))\n",
    "                        model.lb(i) = -media_constraint_uptake(strcmp(media_constraint_exchange,model.rxns{i}));\n",
    "                    end\n",
    "                end\n",
    "            end\n",
    "        end\n",
    "    end\n",
    "end"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Load G and concentration values"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%file load_deltag.m\n",
    "\n",
    "function [deltag_lower,deltag_upper] = load_deltag(model,recalculate_thermodynamics,default_concentration_lb,default_concentration_ub,concentration_ranges_all,concentration_ranges_sample_folder,concentration_ranges_sample,concentration_mets,concentration_values,samples_all,samples_all_source)\n",
    "\n",
    "    % if recalculating thermodynamics\n",
    "    if recalculate_thermodynamics == 1\n",
    "    \n",
    "        % load datasets for all samples\n",
    "        conc_all_met = {};\n",
    "        conc_all_lower = [];\n",
    "        conc_all_upper = [];\n",
    "        for i = 1:length(concentration_ranges_all)\n",
    "        \n",
    "            % load data\n",
    "            f = fopen(sprintf('../data/concentration/_ALL_/%s.csv',concentration_ranges_all{i}),'r');\n",
    "            data = textscan(f,'%s %f %f','Delimiter',',','headerLines',1);\n",
    "            fclose(f);\n",
    "            \n",
    "            % add metabolite id, lower and upper bound\n",
    "            for j = 1:length(data{1})\n",
    "                \n",
    "                % if metabolite already added\n",
    "                if any(strcmp(conc_all_met,data{1}{j}))\n",
    "                \n",
    "                    % change lower bound if lower than original\n",
    "                    if data{2}(j) < conc_all_lower(strcmp(conc_all_met,data{1}{j}))\n",
    "                        conc_all_lower(strcmp(conc_all_met,data{1}{j})) = data{2}(j);\n",
    "                    end\n",
    "                    \n",
    "                    % change upper bound if greater than original\n",
    "                    if data{3}(j) > conc_all_upper(strcmp(conc_all_met,data{1}{j}))\n",
    "                        conc_all_upper(strcmp(conc_all_met,data{1}{j})) = data{3}(j);\n",
    "                    end  \n",
    "                \n",
    "                % if metabolite not already added\n",
    "                else\n",
    "                \n",
    "                    % add info\n",
    "                    conc_all_met{end+1} = data{1}{j};\n",
    "                    conc_all_lower(end+1) = data{2}(j);\n",
    "                    conc_all_upper(end+1) = data{3}(j);        \n",
    "                end\n",
    "            end\n",
    "        end\n",
    "        \n",
    "        % load datasets for individual samples\n",
    "        conc_individual_sample = {};\n",
    "        conc_individual_dataset = {};\n",
    "        conc_individual_met = {};\n",
    "        conc_individual_lower = {};\n",
    "        conc_individual_upper = {};\n",
    "        for i = 1:length(concentration_ranges_sample)\n",
    "        \n",
    "            % add \"all\" concentrations to individual sample list\n",
    "            conc_individual_sample{end+1} = concentration_ranges_sample{i};\n",
    "            conc_individual_dataset{end+1} = concentration_ranges_sample_folder{i};\n",
    "            conc_individual_met{end+1} = conc_all_met;\n",
    "            conc_individual_lower{end+1} = conc_all_lower;\n",
    "            conc_individual_upper{end+1} = conc_all_upper;\n",
    "                \n",
    "            % load data\n",
    "            f = fopen(sprintf('../data/concentration/%s/%s.csv',concentration_ranges_sample_folder{i},concentration_ranges_sample{i}),'r');\n",
    "            data = textscan(f,'%s %f %f','Delimiter',',','headerLines',1);\n",
    "            fclose(f);\n",
    "                \n",
    "            % add metabolite id, lower and upper bound\n",
    "            for j = 1:length(data{1})\n",
    "\n",
    "                % if metabolite already added\n",
    "                if any(strcmp(conc_individual_met{end},data{1}{j}))\n",
    "\n",
    "                    % change lower bound if lower than original\n",
    "                    if data{2}(j) < conc_individual_lower{end}(strcmp(conc_individual_met{end},data{1}{j}))\n",
    "                        conc_individual_lower{end}(strcmp(conc_individual_met{end},data{1}{j})) = data{2}(j);\n",
    "                    end\n",
    "\n",
    "                    % change upper bound if greater than original\n",
    "                    if data{3}(j) > conc_individual_upper{end}(strcmp(conc_individual_met{end},data{1}{j}))\n",
    "                        conc_individual_upper{end}(strcmp(conc_individual_met{end},data{1}{j})) = data{3}(j);\n",
    "                    end    \n",
    "\n",
    "                % if metabolite not already added\n",
    "                else\n",
    "\n",
    "                    % add info\n",
    "                    conc_individual_met{end}{end+1} = data{1}{j};\n",
    "                    conc_individual_lower{end}(end+1) = data{2}(j);\n",
    "                    conc_individual_upper{end}(end+1) = data{3}(j);\n",
    "                end\n",
    "            end\n",
    "        end\n",
    "        \n",
    "        % load deltaG file\n",
    "        f = fopen('../data/deltag/deltag.csv','r');\n",
    "        data = textscan(f,'%s %f %f %f %f %f','Delimiter',',','headerLines',1);\n",
    "        fclose(f);\n",
    "        \n",
    "        % calculate deltag ranges based on concentrations\n",
    "        deltag_lower = -Inf*ones(length(model.rxns),length(samples_all));\n",
    "        deltag_upper = Inf*ones(length(model.rxns),length(samples_all));\n",
    "        \n",
    "        % implement concentrations for all samples\n",
    "        for i = 1:length(model.rxns)\n",
    "            \n",
    "            % if deltag data available\n",
    "            if any(strcmp(data{1},model.rxns{i}))\n",
    "            \n",
    "                % get concentrations of reactants and products\n",
    "                reactant_power = [];\n",
    "                reactant_lower = [];\n",
    "                reactant_upper = [];\n",
    "                product_power = [];\n",
    "                product_lower = [];\n",
    "                product_upper = [];\n",
    "\n",
    "                % iterate over reactants\n",
    "                for j = find(model.S(:,i) < 0)'\n",
    "                    reactant_power = full(model.S(j,i));\n",
    "\n",
    "                    % if metabolite found in concentration data\n",
    "                    if any(strcmp(conc_all_met,model.mets{j}))\n",
    "                        reactant_lower(end+1) = conc_all_lower(strcmp(conc_all_met,model.mets{j}));\n",
    "                        reactant_upper(end+1) = conc_all_upper(strcmp(conc_all_met,model.mets{j}));\n",
    "\n",
    "                    % if metabolite not found in concentration data\n",
    "                    else\n",
    "                        reactant_lower(end+1) = default_concentration_lb;\n",
    "                        reactant_upper(end+1) = default_concentration_ub;\n",
    "                    end\n",
    "                end\n",
    "\n",
    "                % iterate over products\n",
    "                for j = find(model.S(:,i) > 0)'\n",
    "                    product_power = full(model.S(j,i));\n",
    "\n",
    "                    % if metabolite found in concentration data\n",
    "                    if any(strcmp(conc_all_met,model.mets{j}))\n",
    "                        reactant_lower(end+1) = conc_all_lower(strcmp(conc_all_met,model.mets{j}));\n",
    "                        reactant_upper(end+1) = conc_all_upper(strcmp(conc_all_met,model.mets{j}));\n",
    "\n",
    "                    % if metabolite not found in concentration data\n",
    "                    else\n",
    "                        reactant_lower(end+1) = default_concentration_lb;\n",
    "                        reactant_upper(end+1) = default_concentration_ub;\n",
    "                    end\n",
    "                end\n",
    "\n",
    "                % calculate deltag bounds\n",
    "                Q_max = prod(product_upper.^product_power)/prod(reactant_lower.^reactant_power);\n",
    "                Q_min = prod(product_lower.^product_power)/prod(reactant_upper.^reactant_power);\n",
    "                deltag_lower(i,:) = data{4}(strcmp(data{1},model.rxns{i})) - 8.3144598/1000*310.15*log(Q_max);\n",
    "                deltag_upper(i,:) = data{4}(strcmp(data{1},model.rxns{i})) - 8.3144598/1000*310.15*log(Q_min);\n",
    "            end\n",
    "        end\n",
    "        \n",
    "        % implement concentrations for individual samples\n",
    "        for a = 1:length(conc_individual_sample)\n",
    "        \n",
    "            % implement concentrations for all samples\n",
    "            for i = 1:length(model.rxns)\n",
    "\n",
    "                % if deltag data available\n",
    "                if any(strcmp(data{1},model.rxns{i}))\n",
    "\n",
    "                    % get concentrations of reactants and products\n",
    "                    reactant_power = [];\n",
    "                    reactant_lower = [];\n",
    "                    reactant_upper = [];\n",
    "                    product_power = [];\n",
    "                    product_lower = [];\n",
    "                    product_upper = [];\n",
    "\n",
    "                    % iterate over reactants\n",
    "                    for j = find(model.S(:,i) < 0)'\n",
    "                        reactant_power = full(model.S(j,i));\n",
    "\n",
    "                        % if metabolite found in concentration data\n",
    "                        if any(strcmp(conc_individual_met{a},model.mets{j}))\n",
    "                            reactant_lower(end+1) = conc_individual_lower{a}(strcmp(conc_individual_met{a},model.mets{j}));\n",
    "                            reactant_upper(end+1) = conc_individual_upper{a}(strcmp(conc_individual_met{a},model.mets{j}));\n",
    "\n",
    "                        % if metabolite not found in concentration data\n",
    "                        else\n",
    "                            reactant_lower(end+1) = default_concentration_lb;\n",
    "                            reactant_upper(end+1) = default_concentration_ub;\n",
    "                        end\n",
    "                    end\n",
    "\n",
    "                    % iterate over products\n",
    "                    for j = find(model.S(:,i) > 0)'\n",
    "                        product_power = full(model.S(j,i));\n",
    "\n",
    "                        % if metabolite found in concentration data\n",
    "                        if any(strcmp(conc_individual_met{a},model.mets{j}))\n",
    "                            reactant_lower(end+1) = conc_individual_lower{a}(strcmp(conc_individual_met{a},model.mets{j}));\n",
    "                            reactant_upper(end+1) = conc_individual_upper{a}(strcmp(conc_individual_met{a},model.mets{j}));\n",
    "\n",
    "                        % if metabolite not found in concentration data\n",
    "                        else\n",
    "                            reactant_lower(end+1) = default_concentration_lb;\n",
    "                            reactant_upper(end+1) = default_concentration_ub;\n",
    "                        end\n",
    "                    end\n",
    "\n",
    "                    % calculate deltag bounds\n",
    "                    Q_max = prod(product_upper.^product_power)/prod(reactant_lower.^reactant_power);\n",
    "                    Q_min = prod(product_lower.^product_power)/prod(reactant_upper.^reactant_power);\n",
    "                    deltag_lower(i,strcmp(samples_all,conc_individual_sample{a})) = data{4}(strcmp(data{1},model.rxns{i})) - 8.3144598/1000*310.15*log(Q_max)\n",
    "                    deltag_upper(i,strcmp(samples_all,conc_individual_sample{a})) = data{4}(strcmp(data{1},model.rxns{i})) - 8.3144598/1000*310.15*log(Q_min)\n",
    "                end\n",
    "            end\n",
    "        end\n",
    "        \n",
    "    % if not recalculating thermodynamics\n",
    "    else\n",
    "        \n",
    "        % load deltaG file\n",
    "        f = fopen('../data/deltag/deltag.csv','r');\n",
    "        data = textscan(f,'%s %f %f %f %f %f','Delimiter',',','headerLines',1);\n",
    "        fclose(f);\n",
    "        \n",
    "        % use provided minimum and maximum transformed deltag values\n",
    "        deltag_lower = -Inf*ones(length(model.rxns),length(samples_all));\n",
    "        deltag_upper = Inf*ones(length(model.rxns),length(samples_all));\n",
    "        for i = 1:length(model.rxns)\n",
    "            if any(strcmp(data{1},model.rxns{i}))\n",
    "                deltag_lower(i,:) = data{5}(strcmp(data{1},model.rxns{i}));\n",
    "                deltag_upper(i,:) = data{6}(strcmp(data{1},model.rxns{i}));\n",
    "            end\n",
    "        end\n",
    "    end\n",
    "end"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Load/calculate Vmax for each sample"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%file calculate_vmax.m\n",
    "\n",
    "function [vmax_forward,vmax_reverse] = calculate_vmax(model,samples_all,samples_all_source,samples_all_source_protein,use_mutation,knockdown_genes,knockdown_fractions,gene_knockdown_screen)\n",
    "\n",
    "    % initialize vmax\n",
    "    vmax_forward = nan(length(model.rxns),length(samples_all));\n",
    "    vmax_reverse = nan(length(model.rxns),length(samples_all));\n",
    "    \n",
    "    % if gene knockdown, parse GPR rules and reaction kcat's\n",
    "    if (length(knockdown_genes) > 0)\n",
    " \n",
    "        % parse GPR rules\n",
    "        parsedGPR = GPRparser(model);\n",
    "                    \n",
    "        % load forward reaction kcat's\n",
    "        f = fopen('../data/kcat/kcat_forward.csv','r');\n",
    "        kcat_forward_original = textscan(f,'%s %f %s','Delimiter',',','headerLines',1);\n",
    "        fclose(f);\n",
    "\n",
    "        % load reverse reaction kcat's\n",
    "        f = fopen('../data/kcat/kcat_reverse.csv','r');\n",
    "        kcat_reverse_original = textscan(f,'%s %f %s','Delimiter',',','headerLines',1);\n",
    "        fclose(f);\n",
    "    end\n",
    "    \n",
    "    % load list of TCGA patients with IDH1 mutations\n",
    "    idh1_patients_H133Q = importdata('../data/idh1/H133Q.txt')';\n",
    "    idh1_patients_A134D = importdata('../data/idh1/A134D.txt')';\n",
    "    idh1_patients_R100Q = importdata('../data/idh1/R100Q.txt')';\n",
    "    idh1_patients_R132H = importdata('../data/idh1/R132H.txt')';\n",
    "    idh1_patients_R132C = importdata('../data/idh1/R132C.txt')';\n",
    "    idh1_patients_R132G = importdata('../data/idh1/R132G.txt')';\n",
    "    idh1_patients_R132W = importdata('../data/idh1/R132W.txt')';\n",
    "    idh1_patients_R132A = importdata('../data/idh1/R132A.txt')';\n",
    "    idh1_patients_R132Q = importdata('../data/idh1/R132Q.txt')';\n",
    "    idh1_patients_R132K = importdata('../data/idh1/R132K.txt')';\n",
    "    idh1_patients_R132N = importdata('../data/idh1/R132N.txt')';\n",
    "    idh1_patients_all = [idh1_patients_H133Q,idh1_patients_A134D,idh1_patients_R100Q,idh1_patients_R132H,idh1_patients_R132C,idh1_patients_R132G,idh1_patients_R132W,idh1_patients_R132A,idh1_patients_R132Q,idh1_patients_R132K,idh1_patients_R132N];\n",
    "\n",
    "    % IDH1 kcat, forward reaction\n",
    "    idh1_normal_H133Q = 45*60*60;\n",
    "    idh1_normal_R132A = 10.4*60*60;\n",
    "    idh1_normal_R132G = 9.3*60*60;\n",
    "    idh1_normal_R132Q = 9.2*60*60;\n",
    "    idh1_normal_R132K = 7.2*60*60;\n",
    "    idh1_normal_R100Q = 5.6*60*60;\n",
    "    idh1_normal_R132C = 4.4*60*60;\n",
    "    idh1_normal_R132H = 2.4*60*60;\n",
    "    idh1_normal_A134D = 2.3*60*60;\n",
    "    idh1_normal_R132W = 1.21*60*60;\n",
    "    idh1_normal_R132N = 0.047*60*60;\n",
    "\n",
    "    % IDH1 kcat, neomorphic reaction\n",
    "    idh1_neomorphic_H133Q = 0*60*60;\n",
    "    idh1_neomorphic_A134D = 0*60*60;\n",
    "    idh1_neomorphic_R100Q = 0.34*60*60;\n",
    "    idh1_neomorphic_R132A = 0.37*60*60;\n",
    "    idh1_neomorphic_R132W = 0.54*60*60;\n",
    "    idh1_neomorphic_R132K = 0.57*60*60;\n",
    "    idh1_neomorphic_R132N = 0.79*60*60;\n",
    "    idh1_neomorphic_R132G = 1.59*60*60;\n",
    "    idh1_neomorphic_R132C = 1.60*60*60;\n",
    "    idh1_neomorphic_R132H = 4.2*60*60;\n",
    "    idh1_neomorphic_R132Q = 4.7*60*60;\n",
    "\n",
    "    % iterate over samples\n",
    "    for i = 1:length(samples_all)\n",
    "\n",
    "        % load vmax data\n",
    "        if (use_mutation == 1) || ( ( (exist(sprintf('../data/mutation/%s/%s.csv',samples_all_source{i},samples_all{i}),'file')) || (any(strcmp(idh1_patients_all,samples_all{i}))) ) && (use_mutation == 2))\n",
    "            f = fopen(sprintf('../data/vmax/%s/yes_mutation/%s.csv',samples_all_source_protein{i},samples_all{i}),'r');\n",
    "        else\n",
    "            f = fopen(sprintf('../data/vmax/%s/no_mutation/%s.csv',samples_all_source_protein{i},samples_all{i}),'r');\n",
    "        end\n",
    "        vmax = textscan(f,'%s %f %f','Delimiter',',','HeaderLines',1);\n",
    "        vmax{1} = strrep(vmax{1},'\"','');\n",
    "        fclose(f);\n",
    "        \n",
    "        % put in vmax values\n",
    "        vmax_forward(1:length(vmax{1}),i) = vmax{2};\n",
    "        vmax_reverse(1:length(vmax{1}),i) = vmax{3};\n",
    " \n",
    "        % if gene knockdown, change vmax values for reactions containing that gene\n",
    "        if (length(knockdown_genes) > 0)\n",
    "             \n",
    "            % load protein expression\n",
    "            f = fopen(sprintf('../data/protein/output/%s/%s.csv',samples_all_source_protein{i},samples_all{i}),'r');\n",
    "            protein = textscan(f,'%s %f','Delimiter',',');\n",
    "            protein{1} = strrep(protein{1},'\"','');\n",
    "            fclose(f);\n",
    "            \n",
    "            % if using mutation data\n",
    "            if ( (exist(sprintf('../data/mutation/%s/%s.csv',samples_all_source{i},samples_all{i}),'file')) && (use_mutation == 1) ) || ( (exist(sprintf('../data/mutation/%s/%s.csv',samples_all_source{i},samples_all{i}),'file')) && (use_mutation == 2) )\n",
    "\n",
    "                % load mutation data\n",
    "                f = fopen(sprintf('../data/mutation/%s/%s.csv',samples_all_source{i},samples_all{i}),'r');\n",
    "                mutation = textscan(f,'%s %f','Delimiter',',','headerLines',1);\n",
    "                fclose(f);\n",
    "\n",
    "                % multiply protein expression data by envision scores\n",
    "                for j = 1:length(mutation{1})\n",
    "                    if any(strcmp(protein{1},mutation{1}{j}))\n",
    "                        protein{2}(find(strcmp(protein{1},mutation{1}{j}))) = protein{2}(find(strcmp(protein{1},mutation{1}{j}))) * mutation{2}(j);\n",
    "                    end\n",
    "                end\n",
    "            end\n",
    "            \n",
    "            % gene knockdowns\n",
    "            for j = 1:length(knockdown_genes)\n",
    "                if any(strcmp(protein{1},model.geneSymbols{find(strcmp(model.genes,knockdown_genes{j}))}))\n",
    "                    protein{2}(find(strcmp(protein{1},model.geneSymbols{find(strcmp(model.genes,knockdown_genes{j}))}))) = protein{2}(find(strcmp(protein{1},model.geneSymbols{find(strcmp(model.genes,knockdown_genes{j}))}))) * knockdown_fractions(j);\n",
    "                end\n",
    "            end\n",
    "            \n",
    "            % convert ppm to mmol/gDW\n",
    "            m = 50; % average mass of protein, in kDa\n",
    "            fdw = 0.5; % fraction of dry weight that is protein\n",
    "            Na = 6.0221409*10^23; % Avogadro constant\n",
    "            gDa = 1.66054*10^-24; % number of grams per dalton\n",
    "\n",
    "            protein{2} = protein{2}/1000000; % protein/million --> protein/protein\n",
    "            protein{2} = protein{2} / Na; % protein/protein --> mol/protein\n",
    "            protein{2} = protein{2} * 1000; % mol/protein --> mmol/protein\n",
    "            protein{2} = protein{2} / (m*1000); % mmol/protein --> mmol/Daltons_protein\n",
    "            protein{2} = protein{2} / gDa; % mmol/Daltons --> mmol/grams_protein\n",
    "            protein{2} = protein{2} * fdw; % mmol/grams_protein --> mmol/gDW\n",
    "            \n",
    "            % get reactions that gene(s) is(are) involved in\n",
    "            reaction_list = [];\n",
    "            for j = 1:length(knockdown_genes)\n",
    "                reaction_list = [reaction_list,find(full(model.rxnGeneMat(:,strcmp(model.genes,knockdown_genes{j}))))'];\n",
    "            end\n",
    "            reaction_list = unique(reaction_list);\n",
    "            \n",
    "            % get protein expression for each reaction with grRule and EC number containing gene(s) being knockdown\n",
    "            protein_rxn = nan(1,length(model.rxns));\n",
    "            for j = reaction_list\n",
    "                if (~isempty(model.rules{j})) && (~isempty(model.rxnECNumbers{j}))\n",
    "\n",
    "                    % get genes associated with reaction\n",
    "                    genes = find(model.rxnGeneMat(j,:));\n",
    "                    allfound = 1;\n",
    "                    for k = 1:length(genes)\n",
    "                        if ~any(strcmp(protein{1},model.geneSymbols{genes(k)}));\n",
    "                            allfound = 0;\n",
    "                            break\n",
    "                        end\n",
    "                    end\n",
    "                    if allfound == 1\n",
    "\n",
    "                        % create model structure for selectGeneFromGPR\n",
    "                        modelRxn = {};\n",
    "                        modelRxn.rxns = {model.rxns{j}};\n",
    "                        modelRxn.genes = {};\n",
    "                        for k = 1:length(genes)\n",
    "                            modelRxn.genes{end+1} = model.genes{genes(k)};\n",
    "                        end\n",
    "                        modelRxn.genes = modelRxn.genes';\n",
    "\n",
    "                        % calculate reaction expression\n",
    "                        values = [];\n",
    "                        for k = 1:length(genes)\n",
    "                            values(end+1) = protein{2}(strcmp(protein{1},model.geneSymbols{genes(k)}));\n",
    "                        end\n",
    "                        % updated version of selectGeneFromGPR\n",
    "                        protein_rxn(j) = selectGeneFromGPR(modelRxn,modelRxn.genes,values',{parsedGPR{j}});\n",
    "\n",
    "                    else\n",
    "                       protein_rxn(j) = NaN; \n",
    "                    end\n",
    "\n",
    "                else\n",
    "                    protein_rxn(j) = NaN;\n",
    "                end\n",
    "            end\n",
    "            \n",
    "            % load kcat values\n",
    "            kcat_forward = kcat_forward_original;\n",
    "            kcat_reverse = kcat_reverse_original;\n",
    "            \n",
    "            % alter kcat reactions for IDH1 normal reactions\n",
    "            if (use_mutation == 1) or (use_mutation == 2)\n",
    "                if any(strcmp(idh1_patients_R132N,samples_all{i}))\n",
    "                    kcat_forward{2}(strcmp(kcat_forward{1},'ICDHy')) = idh1_normal_R132N;\n",
    "                    kcat_forward{2}(strcmp(kcat_forward{1},'ICDHyp')) = idh1_normal_R132N;\n",
    "                    kcat_forward{2}(strcmp(kcat_forward{1},'IDH1_R132')) = idh1_neomorphic_R132N;\n",
    "                elseif any(strcmp(idh1_patients_R132W,samples_all{i}))\n",
    "                    kcat_forward{2}(strcmp(kcat_forward{1},'ICDHy')) = idh1_normal_R132W;\n",
    "                    kcat_forward{2}(strcmp(kcat_forward{1},'ICDHyp')) = idh1_normal_R132W;\n",
    "                    kcat_forward{2}(strcmp(kcat_forward{1},'IDH1_R132')) = idh1_neomorphic_R132W;\n",
    "                elseif any(strcmp(idh1_patients_A134D,samples_all{i}))\n",
    "                    kcat_forward{2}(strcmp(kcat_forward{1},'ICDHy')) = idh1_normal_A134D;\n",
    "                    kcat_forward{2}(strcmp(kcat_forward{1},'ICDHyp')) = idh1_normal_A134D;\n",
    "                    kcat_forward{2}(strcmp(kcat_forward{1},'IDH1_R132')) = idh1_neomorphic_A134D;\n",
    "                elseif any(strcmp(idh1_patients_R132H,samples_all{i}))\n",
    "                    kcat_forward{2}(strcmp(kcat_forward{1},'ICDHy')) = idh1_normal_R132H;\n",
    "                    kcat_forward{2}(strcmp(kcat_forward{1},'ICDHyp')) = idh1_normal_R132H;\n",
    "                    kcat_forward{2}(strcmp(kcat_forward{1},'IDH1_R132')) = idh1_neomorphic_R132H;\n",
    "                elseif any(strcmp(idh1_patients_R132C,samples_all{i}))\n",
    "                    kcat_forward{2}(strcmp(kcat_forward{1},'ICDHy')) = idh1_normal_R132C;\n",
    "                    kcat_forward{2}(strcmp(kcat_forward{1},'ICDHyp')) = idh1_normal_R132C;\n",
    "                    kcat_forward{2}(strcmp(kcat_forward{1},'IDH1_R132')) = idh1_neomorphic_R132C;\n",
    "                elseif any(strcmp(idh1_patients_R100Q,samples_all{i}))\n",
    "                    kcat_forward{2}(strcmp(kcat_forward{1},'ICDHy')) = idh1_normal_R100Q;\n",
    "                    kcat_forward{2}(strcmp(kcat_forward{1},'ICDHyp')) = idh1_normal_R100Q;\n",
    "                    kcat_forward{2}(strcmp(kcat_forward{1},'IDH1_R132')) = idh1_neomorphic_R100Q;\n",
    "                elseif any(strcmp(idh1_patients_R132K,samples_all{i}))\n",
    "                    kcat_forward{2}(strcmp(kcat_forward{1},'ICDHy')) = idh1_normal_R132K;\n",
    "                    kcat_forward{2}(strcmp(kcat_forward{1},'ICDHyp')) = idh1_normal_R132K;\n",
    "                    kcat_forward{2}(strcmp(kcat_forward{1},'IDH1_R132')) = idh1_neomorphic_R132K;\n",
    "                elseif any(strcmp(idh1_patients_R132Q,samples_all{i}))\n",
    "                    kcat_forward{2}(strcmp(kcat_forward{1},'ICDHy')) = idh1_normal_R132Q;\n",
    "                    kcat_forward{2}(strcmp(kcat_forward{1},'ICDHyp')) = idh1_normal_R132Q;\n",
    "                    kcat_forward{2}(strcmp(kcat_forward{1},'IDH1_R132')) = idh1_neomorphic_R132Q;\n",
    "                elseif any(strcmp(idh1_patients_R132G,samples_all{i}))\n",
    "                    kcat_forward{2}(strcmp(kcat_forward{1},'ICDHy')) = idh1_normal_R132G;\n",
    "                    kcat_forward{2}(strcmp(kcat_forward{1},'ICDHyp')) = idh1_normal_R132G;\n",
    "                    kcat_forward{2}(strcmp(kcat_forward{1},'IDH1_R132')) = idh1_neomorphic_R132G;\n",
    "                elseif any(strcmp(idh1_patients_R132A,samples_all{i}))\n",
    "                    kcat_forward{2}(strcmp(kcat_forward{1},'ICDHy')) = idh1_normal_R132A;\n",
    "                    kcat_forward{2}(strcmp(kcat_forward{1},'ICDHyp')) = idh1_normal_R132A;\n",
    "                    kcat_forward{2}(strcmp(kcat_forward{1},'IDH1_R132')) = idh1_neomorphic_R132A;\n",
    "                elseif any(strcmp(idh1_patients_H133Q,samples_all{i}))\n",
    "                    kcat_forward{2}(strcmp(kcat_forward{1},'ICDHy')) = idh1_normal_H133Q;\n",
    "                    kcat_forward{2}(strcmp(kcat_forward{1},'ICDHyp')) = idh1_normal_H133Q;\n",
    "                    kcat_forward{2}(strcmp(kcat_forward{1},'IDH1_R132')) = idh1_neomorphic_H133Q;\n",
    "                end\n",
    "            end\n",
    "            \n",
    "            % change vmax for reactions containing gene(s) being knockdown\n",
    "            for j = reaction_list\n",
    "                if ~strcmp(model.subSystems{j},'Custom Constraint') && ~strcmp(model.subSystems{j},'Custom Objective');\n",
    "\n",
    "                    % forward vmax\n",
    "                    if model.ub(j) > 0\n",
    "                        if (~(isnan(protein_rxn(j)))) && (~(isnan(kcat_forward{2}(strcmp(kcat_forward{1},model.rxns{j})))))\n",
    "                            vmax_forward(j,i) = protein_rxn(j) * kcat_forward{2}(strcmp(kcat_forward{1},model.rxns{j}));\n",
    "                        end\n",
    "                    end\n",
    "\n",
    "                    % reverse vmax\n",
    "                    if model.lb(j) < 0\n",
    "                        if (~(isnan(protein_rxn(j)))) && (~(isnan(kcat_reverse{2}(strcmp(kcat_reverse{1},model.rxns{j})))))\n",
    "                            vmax_reverse(j,i) = protein_rxn(j) * kcat_reverse{2}(strcmp(kcat_reverse{1},model.rxns{j}));\n",
    "                        end\n",
    "                    end\n",
    "                end\n",
    "            end\n",
    "        end\n",
    "    end\n",
    "end"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Run FBA for each sample"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%file run_fba.m\n",
    "\n",
    "\n",
    "function [] = run_fba(output_folder,input_file,model,type_of_analysis,samples_all,samples_all_source,samples_all_source_protein,vmax_forward,vmax_reverse,objective_function_direction,objective_function_weight,objective_function_reaction,exchange_rxn,deltag_lower,deltag_upper,constraints_fraction_count,constraints_fraction_type,constraints_fraction_reaction,constraints_fraction_value,constraints_fraction_id,constraints_maxmulti_fraction,constraints_maxmulti_count,pfba_fraction,fva_all_fraction,fva_select_fraction,fva_select_groups,fva_select_reactions,sampler_fraction,sampler_samples,sampler_steps,media_sensitivity_critical_fraction,media_sensitivity_critical_metabolite1,media_sensitivity_spectrum_fraction,media_sensitivity_spectrum_metabolite1,media_sensitivity_spectrum_metabolite2,media_sensitivity_spectrum_points,parallel_cores,knockdown_genes,knockdown_fractions,catalase_version,module_5fu,module_cis,module_cpa,module_dox);\n",
    "\n",
    "    % initialize results\n",
    "    temp = strsplit(input_file,'.xlsx');\n",
    "    temp = strsplit(temp{1},'/');\n",
    "    input_filename = temp{end};\n",
    "    mkdir(sprintf('results/%s/%s',output_folder,input_filename))\n",
    "    if type_of_analysis == 2\n",
    "        mkdir(sprintf('results/%s/%s/pfba',output_folder,input_filename))\n",
    "    elseif (type_of_analysis == 3) || (type_of_analysis == 4)\n",
    "        mkdir(sprintf('results/%s/%s/fva',output_folder,input_filename))\n",
    "    elseif type_of_analysis == 5\n",
    "        mkdir(sprintf('results/%s/%s/sampling',output_folder,input_filename))\n",
    "    elseif type_of_analysis == 7\n",
    "        mkdir(sprintf('results/%s/%s/media_sensitivity',output_folder,input_filename))\n",
    "    end\n",
    "    \n",
    "    % create objective value file\n",
    "    f_obj = fopen(sprintf('results/%s/%s/objval.tsv',output_folder,input_filename),'w');\n",
    "    fprintf(f_obj,'SAMPLE\\tSOURCE\\tOBJVAL\\n');\n",
    "        \n",
    "    % create critical media sensitivity file\n",
    "    if type_of_analysis == 6\n",
    "        f_media = fopen(sprintf('results/%s/%s/media_sensitivity_critical.tsv',output_folder,input_filename),'w');\n",
    "        fprintf(f_media,'SAMPLE\\tSOURCE\\tCRITICAL UPTAKE [mmol/gDW/hr]\\n');\n",
    "    end\n",
    "\n",
    "    % save initial model object\n",
    "    modelOriginal = model;\n",
    "    \n",
    "    % iterate over each sample\n",
    "    for i = 1:length(samples_all)\n",
    "        \n",
    "        % load original model\n",
    "        model = modelOriginal;\n",
    "\n",
    "        % implement vmax values\n",
    "        for j = 1:length(model.rxns)\n",
    "\n",
    "            % forward value\n",
    "            if (model.ub(j) == max(model.ub)) && (~(isnan(vmax_forward(j,i))))\n",
    "                model.ub(j) = vmax_forward(j,i);\n",
    "            end\n",
    "\n",
    "            % reverse value\n",
    "            if (model.lb(j) == min(model.lb)) && (~(isnan(vmax_reverse(j,i))))\n",
    "                model.lb(j) = -vmax_reverse(j,i);\n",
    "            end   \n",
    "        end\n",
    "\n",
    "        % implement deltag constraints\n",
    "        for j = 1:length(model.rxns)\n",
    "\n",
    "            % if deltag range only positive, restrict forward direction\n",
    "            if deltag_lower(j,i) > 0\n",
    "                model.ub(j) = 0;\n",
    "            end\n",
    "\n",
    "            % if deltag range only negative, restrict reverse direction (if reaction is reversible)\n",
    "            if deltag_upper(j,i) < 0\n",
    "                model.lb(j) = 0;\n",
    "            end\n",
    "        end\n",
    "        \n",
    "        % implement catalase\n",
    "        catalase_reactions = {'r0010','CATm','CATp','CATr'};\n",
    "        if catalase_version == 0\n",
    "            for j = 1:length(catalase_reactions)\n",
    "                model.lb(strcmp(model.rxns,strcat(catalase_reactions{j},'_nadph'))) = 0;\n",
    "                model.ub(strcmp(model.rxns,strcat(catalase_reactions{j},'_nadph'))) = 0;\n",
    "            end\n",
    "        else\n",
    "            for j = 1:length(catalase_reactions)\n",
    "                model.lb(strcmp(model.rxns,catalase_reactions{j})) = 0;\n",
    "                model.ub(strcmp(model.rxns,catalase_reactions{j})) = 0;\n",
    "            end\n",
    "        end\n",
    "        \n",
    "        % implement 5-fluorouracil module\n",
    "        if module_5fu == 0\n",
    "            for j = 1:length(model.rxns)\n",
    "                if strcmp(model.subSystems{j},'MODULE 5-FLUOROURACIL');\n",
    "                    model.lb(j) = 0;\n",
    "                    model.ub(j) = 0;\n",
    "                end\n",
    "            end\n",
    "        end\n",
    "        \n",
    "        % implement cisplatin module\n",
    "        if module_cis == 0\n",
    "            for j = 1:length(model.rxns)\n",
    "                if strcmp(model.subSystems{j},'MODULE CISPLATIN');\n",
    "                    model.lb(j) = 0;\n",
    "                    model.ub(j) = 0;\n",
    "                end\n",
    "            end\n",
    "        end\n",
    "        \n",
    "        % implement cyclophosphamide module\n",
    "        if module_cpa == 0\n",
    "            for j = 1:length(model.rxns)\n",
    "                if strcmp(model.subSystems{j},'MODULE CYCLOPHOSPHAMIDE');\n",
    "                    model.lb(j) = 0;\n",
    "                    model.ub(j) = 0;\n",
    "                end\n",
    "            end\n",
    "        end\n",
    "        \n",
    "        % implement doxorubicin module\n",
    "        if module_dox == 0\n",
    "            for j = 1:length(model.rxns)\n",
    "                if strcmp(model.subSystems{j},'MODULE DOXORUBICIN');\n",
    "                    model.lb(j) = 0;\n",
    "                    model.ub(j) = 0;\n",
    "                end\n",
    "            end\n",
    "        end\n",
    "        \n",
    "        % setup LP problem\n",
    "        model.A = model.S;\n",
    "        model.rhs = model.b;\n",
    "        model.obj = model.c;\n",
    "        model.sense = repmat('=',1,length(model.mets));\n",
    "        model.vtype = repmat('C',1,length(model.rxns));\n",
    "        model.varnames = model.rxns;\n",
    "        model.modelsense = lower(objective_function_direction);\n",
    "\n",
    "        % reset objective value\n",
    "        model.c = zeros(size(model.c));\n",
    "        model.obj = zeros(size(model.obj));\n",
    "\n",
    "        % find fraction of maximum constraints\n",
    "        max_constraint_values = [];\n",
    "        for j = 1:constraints_fraction_count\n",
    "\n",
    "            % implement objective value\n",
    "            model.c(model.c_fraction == j) = 1;\n",
    "            model.obj(model.c_fraction == j) = 1;\n",
    "            model.modelsense = 'max';\n",
    "\n",
    "            % get objective value\n",
    "            params.outputflag = 0;\n",
    "            result = gurobi(model,params);\n",
    "            max_constraint_values(end+1) = result.objval;\n",
    "\n",
    "            % reset objective value\n",
    "            model.c = zeros(size(model.c));\n",
    "            model.obj = zeros(size(model.obj));\n",
    "        end\n",
    "\n",
    "        % implement fraction of maximum constraints\n",
    "        for j = 1:constraints_fraction_count\n",
    "\n",
    "            % equal to\n",
    "            if strcmp(constraints_fraction_type{j},'equal')\n",
    "                model.lb(constraints_fraction_id(j)) = constraints_fraction_value{j}(1) * max_constraint_values(j);\n",
    "                model.ub(constraints_fraction_id(j)) = constraints_fraction_value{j}(1) * max_constraint_values(j);\n",
    "\n",
    "            % less than\n",
    "            elseif strcmp(constraints_fraction_type{j},'less')\n",
    "                model.ub(constraints_fraction_id(j)) = constraints_fraction_value{j}(1) * max_constraint_values(j);\n",
    "\n",
    "            % greater than\n",
    "            elseif strcmp(constraints_fraction_type{j},'greater')\n",
    "                model.lb(constraints_fraction_id(j)) = constraints_fraction_value{j}(1) * max_constraint_values(j);\n",
    "\n",
    "            % range\n",
    "            elseif strcmp(constraints_fraction_type{j},'range')\n",
    "                if constraints_fraction_value{j}(1) <= constraints_fraction_value{j}(2)\n",
    "                    model.lb(constraints_fraction_id(j)) = constraints_fraction_value{j}(1) * max_constraint_values(j);\n",
    "                    model.ub(constraints_fraction_id(j)) = constraints_fraction_value{j}(2) * max_constraint_values(j);\n",
    "                else\n",
    "                    error('ERROR - Constraints - Custom - For Range, min value must be <= max value');\n",
    "                end\n",
    "            end      \n",
    "        end  \n",
    "\n",
    "\t\t% find max multi constraints\n",
    "\t\tif constraints_maxmulti_count > 0\n",
    "\t\t\n",
    "\t\t\t% implement objective value\n",
    "\t\t\tmodel.c = model.c_maxmulti;\n",
    "\t\t\tmodel.obj = model.c_maxmulti;\n",
    "\t\t\tmodel.modelsense = 'max';\n",
    "\n",
    "            % get objective value\n",
    "            params.outputflag = 0;\n",
    "            result = gurobi(model,params);\n",
    "            maxmulti_value = result.objval;\n",
    "\n",
    "            % reset objective value\n",
    "            model.c = zeros(size(model.c));\n",
    "            model.obj = zeros(size(model.obj));\n",
    "        end\n",
    "\n",
    "        % implement objective function\n",
    "        model.c = model.c_obj;\n",
    "        model.obj = model.c_obj;\n",
    "        model.modelsense = 'max';\n",
    "\t\t\n",
    "\t\t% implement max multi constraint\n",
    "\t\tif constraints_maxmulti_count > 0\n",
    "\t\t\tmodel.S(end+1,:) = model.c_maxmulti;\n",
    "\t\t\tmodel.b(end+1) = constraints_maxmulti_fraction*maxmulti_value;\n",
    "\t\t\tmodel.csense(end+1) = 'E';\n",
    "\t\t\tmodel.mets{end+1} = 'maxmulti';\n",
    "\t\t\tmodel.metCharges(end+1) = 0;\n",
    "\t\t\tmodel.metFormulas{end+1} = '';\n",
    "\t\t\tmodel.metSmiles{end+1} = '';\n",
    "\t\t\tmodel.metNames{end+1} = 'Max Multi Constraint';\n",
    "\t\t\tmodel.metHMDBID{end+1} = '';\n",
    "\t\t\tmodel.metInChIString{end+1} = '';\n",
    "\t\t\tmodel.metKEGGID{end+1} = '';\n",
    "\t\t\tmodel.metPubChemID{end+1} = '';\n",
    "\t\t\tmodel.metCHEBIID{end+1} = '';\n",
    "\t\t\tmodel.metPdMap{end+1} = '';\n",
    "\t\t\tmodel.metReconMap{end+1} = '';\n",
    "\t\tend\n",
    "\t\t\n",
    "\t\t% create parameters\n",
    "\t\tmodel.A = model.S;\n",
    "\t\tmodel.rhs = model.b;\n",
    "\t\tmodel.sense = repmat('=',1,length(model.csense));\n",
    "\t\tfor j = 1:length(model.csense)\n",
    "\t\t\tif model.csense(i) == 'G'\n",
    "\t\t\t\tmodel.sense(i) = '>';\n",
    "\t\t\telseif model.csense(i) == 'L'\n",
    "\t\t\t\tmodel.sense(i) = '<';\n",
    "\t\t\tend\n",
    "\t\tend\n",
    "\t\t\n",
    "        % objective function value\n",
    "        params.outputflag = 0;\n",
    "        result = gurobi(model,params);\n",
    "        fprintf(f_obj,'%s\\t%s\\t%0.9f\\n',samples_all{i},samples_all_source_protein{i},result.objval);\n",
    "\n",
    "        % parsimonious fba\n",
    "        if type_of_analysis == 2\n",
    "\n",
    "            % ensure only one maximization objective\n",
    "            if length(objective_function_reaction) ~= 1\n",
    "                error('ERROR - General - Type of Analysis - Can only perform pFBA with maximizing one objective function reaction');\n",
    "            elseif ~strcmp(objective_function_direction{1},'MAX')\n",
    "                error('ERROR - General - Type of Analysis - Can only perform pFBA with maximizing one objective function reaction');\n",
    "            end\n",
    "\n",
    "            % create irreversible model\n",
    "            model_irrev = model;\n",
    "            for j = 1:length(model_irrev.rxns)\n",
    "\n",
    "                % if both directions, split into two separate reactions\n",
    "                if (model_irrev.lb(j) < 0) && (model_irrev.ub(j) > 0)\n",
    "\n",
    "                    % create new reaction\n",
    "                    model_irrev.rxns{end+1} = sprintf('%s_REV',model_irrev.rxns{j});\n",
    "                    model_irrev.S(:,end+1) = -model_irrev.S(:,j);\n",
    "                    model_irrev.lb(end+1) = 0;\n",
    "                    model_irrev.ub(end+1) = -model_irrev.lb(j);\n",
    "                    model_irrev.c(end+1) = 0;\n",
    "                    model_irrev.rxnGeneMat(end+1,:) = model_irrev.rxnGeneMat(j,:);\n",
    "                    model_irrev.rules{end+1} = model_irrev.rules{j};\n",
    "                    model_irrev.grRules{end+1} = model_irrev.grRules{j};\n",
    "                    model_irrev.subSystems{end+1} = model_irrev.subSystems{j};\n",
    "                    model_irrev.rxnNames{end+1} = model_irrev.rxnNames{j};\n",
    "                    model_irrev.rxnKEGGID{end+1} = model_irrev.rxnKEGGID{j};\n",
    "                    model_irrev.rxnKeggOrthology{end+1} = model_irrev.rxnKeggOrthology{j};\n",
    "                    model_irrev.rxnConfidenceScores(end+1) = model_irrev.rxnConfidenceScores(j);\n",
    "                    model_irrev.rxnReferences{end+1} = model_irrev.rxnReferences{j};\n",
    "                    model_irrev.rxnECNumbers{end+1} = model_irrev.rxnECNumbers{j};\n",
    "                    model_irrev.rxnNotes{end+1} = model_irrev.rxnNotes{j};\n",
    "                    model_irrev.rxnCOG{end+1} = model_irrev.rxnCOG{j};\n",
    "                    model_irrev.rxnReconMap{end+1} = model_irrev.rxnReconMap{j};\n",
    "\n",
    "                    % make original reaction irreversible\n",
    "                    model_irrev.lb(j) = 0;                \n",
    "\n",
    "                % if just reverse direction, reverse to forward direction\n",
    "                elseif (model_irrev.lb(j) < 0)\n",
    "\n",
    "                    % reverse reaction\n",
    "                    model_irrev.rxns{j} = sprintf('%s_REV',model_irrev.rxns{j});\n",
    "                    model_irrev.S(:,j) = -model_irrev.S(:,j);\n",
    "                    model_irrev.ub(j) = -model_irrev.lb(j);\n",
    "                    model_irrev.lb(j) = 0;\n",
    "\n",
    "                end\n",
    "            end\n",
    "\n",
    "            % implement objective value constraint\n",
    "            model_irrev.S(end+1,:) = zeros(1,length(model_irrev.rxns));\n",
    "            model_irrev.S(end,find(model_irrev.c)) = 1;\n",
    "            model_irrev.b(end+1) = pfba_fraction * result.objval;\n",
    "\n",
    "            % implement minimization of total flux\n",
    "            original_model_irrev_c = model_irrev.c;\n",
    "            model_irrev.c = zeros(size(model_irrev.c));\n",
    "            model_irrev.c(~original_model_irrev_c) = 1;\n",
    "\n",
    "            % setup LP problem\n",
    "            model_irrev.A = model_irrev.S;\n",
    "            model_irrev.rhs = model_irrev.b;\n",
    "            model_irrev.obj = model_irrev.c;\n",
    "            model_irrev.sense = repmat('=',1,length(model_irrev.mets)+1);\n",
    "            model_irrev.vtype = repmat('C',1,length(model_irrev.rxns));\n",
    "            model_irrev.varnames = model_irrev.rxns;\n",
    "            model_irrev.modelsense = 'min';\n",
    "\n",
    "            % pFBA fluxes\n",
    "            params.outputflag = 0;\n",
    "            result = gurobi(model_irrev,params);\n",
    "\n",
    "            % collect flux values\n",
    "            flux_value = [];\n",
    "            flux_lb = [];\n",
    "            flux_ub = [];\n",
    "            for j = 1:length(model.rxns)\n",
    "                if (any(strcmp(model_irrev.rxns,model.rxns{j}))) && (any(strcmp(model_irrev.rxns,sprintf('%s_REV',model.rxns{j}))))\n",
    "                    if (result.x(strcmp(model_irrev.rxns,model.rxns{j})) > 0) && (result.x(strcmp(model_irrev.rxns,sprintf('%s_REV',model.rxns{j}))) > 0)\n",
    "                        error(sprintf('pFBA - both directions of %s have a nonzero flux',model.rxns{j}));\n",
    "                    elseif result.x(strcmp(model_irrev.rxns,sprintf('%s_REV',model.rxns{j}))) > 0\n",
    "                        flux_value(end+1) = -result.x(strcmp(model_irrev.rxns,sprintf('%s_REV',model.rxns{j})));\n",
    "                    else\n",
    "                        flux_value(end+1) = result.x(strcmp(model_irrev.rxns,model.rxns{j}));\n",
    "                    end\n",
    "                    flux_lb(end+1) = -model_irrev.ub(strcmp(model_irrev.rxns,sprintf('%s_REV',model.rxns{j})));\n",
    "                    flux_ub(end+1) = model_irrev.ub(strcmp(model_irrev.rxns,model.rxns{j}));\n",
    "                elseif any(strcmp(model_irrev.rxns,sprintf('%s_REV',model.rxns{j})))\n",
    "                    flux_value(end+1) = -result.x(strcmp(model_irrev.rxns,sprintf('%s_REV',model.rxns{j})));\n",
    "                    flux_lb(end+1) = -model_irrev.ub(strcmp(model_irrev.rxns,sprintf('%s_REV',model.rxns{j})));\n",
    "                    flux_ub(end+1) = 0;\n",
    "                else\n",
    "                    flux_value(end+1) = result.x(strcmp(model_irrev.rxns,model.rxns{j}));\n",
    "                    flux_lb(end+1) = 0;\n",
    "                    flux_ub(end+1) = model_irrev.ub(strcmp(model_irrev.rxns,model.rxns{j}));\n",
    "                end\n",
    "            end\n",
    "\n",
    "            % output results\n",
    "            f_pfba = fopen(sprintf('results/%s/%s/pfba/%s.tsv',output_folder,input_filename,samples_all{i}),'w');\n",
    "            fprintf(f_pfba,'REACTION\\tFLUX [mmol/gDW/hr]\\tLB [mmol/gDW/hr]\\tUB [mmol/gDW/hr]\\tNAME\\tFORMULA\\tSUBSYSTEM\\tGR RULE\\tEC NUMBER\\tKEGG\\tNOTES\\tREFERENCES\\n');\n",
    "            for j = 1:length(model.rxns)\n",
    "                fprintf(f_pfba,'%s\\t%0.9f\\t%0.9f\\t%0.9f\\t%s\\t%s\\t%s\\t%s\\t%s\\t%s\\t%s\\t%s\\n',model.rxns{j},flux_value(j),flux_lb(j),flux_ub(j),model.rxnNames{j},model.rxnFormulas{j},model.subSystems{j},model.grRules{j},model.rxnECNumbers{j},model.rxnKEGGID{j},model.rxnNotes{j},model.rxnReferences{j});\n",
    "            end\n",
    "            fclose(f_pfba);\n",
    "\n",
    "        % flux variability analysis - all reactions\n",
    "        elseif type_of_analysis == 3\n",
    "\n",
    "            % ensure only one maximization objective\n",
    "            if length(objective_function_reaction) ~= 1\n",
    "                error('ERROR - General - Type of Analysis - Can only perform FVA with maximizing one objective function reaction');\n",
    "            elseif ~strcmp(objective_function_direction{1},'MAX')\n",
    "                error('ERROR - General - Type of Analysis - Can only perform FVA with maximizing one objective function reaction');\n",
    "            end\n",
    "            \n",
    "            % implement objective value constraint\n",
    "            model.S(end+1,:) = zeros(1,length(model.rxns));\n",
    "            model.S(end,find(model.c)) = 1;\n",
    "            model.b(end+1) = fva_all_fraction * result.objval;\n",
    "\n",
    "            % clear model.c\n",
    "            model.c = zeros(size(model.c));\n",
    "\n",
    "            % calculate FVA fluxes\n",
    "            flux_min = [];\n",
    "            flux_max = [];\n",
    "            for j = 1:length(model.rxns)\n",
    "\n",
    "                % set model.c\n",
    "                model.c(j) = 1;\n",
    "\n",
    "                % min value\n",
    "                model.A = model.S;\n",
    "                model.rhs = model.b;\n",
    "                model.obj = model.c;\n",
    "                model.sense = [repmat('=',1,length(model.mets)),'>'];\n",
    "                model.vtype = repmat('C',1,length(model.rxns));\n",
    "                model.varnames = model.rxns;\n",
    "                model.modelsense = 'min';\n",
    "\n",
    "                params.outputflag = 0;\n",
    "                result = gurobi(model,params);\n",
    "                flux_min(end+1) = result.objval;\n",
    "\n",
    "                % max value\n",
    "                model.A = model.S;\n",
    "                model.rhs = model.b;\n",
    "                model.obj = model.c;\n",
    "                model.sense = [repmat('=',1,length(model.mets)),'>'];\n",
    "                model.vtype = repmat('C',1,length(model.rxns));\n",
    "                model.varnames = model.rxns;\n",
    "                model.modelsense = 'max';\n",
    "\n",
    "                params.outputflag = 0;\n",
    "                result = gurobi(model,params);\n",
    "                flux_max(end+1) = result.objval;\n",
    "\n",
    "                % clear model.c\n",
    "                model.c = zeros(size(model.c));\n",
    "            end\n",
    "\n",
    "            % output results\n",
    "            f_fva = fopen(sprintf('results/%s/%s/fva/%s.tsv',output_folder,input_filename,samples_all{i}),'w');\n",
    "            fprintf(f_fva,'REACTION\\tMIN [mmol/gDW/hr]\\tMAX [mmol/gDW/hr]\\tLB [mmol/gDW/hr]\\tUB [mmol/gDW/hr]\\tNAME\\tFORMULA\\tSUBSYSTEM\\tGR RULE\\tEC NUMBER\\tKEGG\\tNOTES\\tREFERENCES\\n');\n",
    "            for j = 1:length(model.rxns)\n",
    "                fprintf(f_fva,'%s\\t%0.9f\\t%0.9f\\t%0.9f\\t%0.9f\\t%s\\t%s\\t%s\\t%s\\t%s\\t%s\\t%s\\t%s\\n',model.rxns{j},flux_min(j),flux_max(j),model.lb(j),model.ub(j),model.rxnNames{j},model.rxnFormulas{j},model.subSystems{j},model.grRules{j},model.rxnECNumbers{j},model.rxnKEGGID{j},model.rxnNotes{j},model.rxnReferences{j});\n",
    "            end\n",
    "            fclose(f_fva);\n",
    "\n",
    "        % flux variability analysis - select reactions\n",
    "        elseif type_of_analysis == 4\n",
    "\n",
    "            % ensure only one maximization objective\n",
    "            if length(objective_function_reaction) ~= 1\n",
    "                error('ERROR - General - Type of Analysis - Can only perform FVA with maximizing one objective function reaction');\n",
    "            elseif ~strcmp(objective_function_direction{1},'MAX')\n",
    "                error('ERROR - General - Type of Analysis - Can only perform FVA with maximizing one objective function reaction');\n",
    "            end\n",
    "\n",
    "            % reaction list\n",
    "            rxnlist = [];\n",
    "            for j = 1:length(fva_select_reactions)\n",
    "                if any(strcmp(model.rxns,fva_select_reactions{j}))\n",
    "                    rxnlist(end+1) = find(strcmp(model.rxns,fva_select_reactions{j}));\n",
    "                else\n",
    "                    error('ERROR - General - Options- FVA - Select reactions - Reaction \"%s\" not found in model',fva_select_reactions{j});\n",
    "                end\n",
    "            end\n",
    "\n",
    "            % add objective reactions\n",
    "            for j = 1:length(model.rxns)\n",
    "                if model.c(j) && ~any(j == rxnlist)\n",
    "                    rxnlist(end+1) = j;\n",
    "                end\n",
    "            end\n",
    "\n",
    "            % reaction groups\n",
    "            for j = 1:length(fva_select_groups)\n",
    "                switch fva_select_groups(j)\n",
    "\n",
    "                    % all nadp+ --> nadph reactions\n",
    "                    case 1\n",
    "\n",
    "                        % find nadp\n",
    "                        nadp = [];\n",
    "                        for k = 1:length(model.mets)\n",
    "                            if length(model.mets{k}) == 7\n",
    "                                if strcmp(model.mets{k}(1:5),'nadp[')\n",
    "                                    nadp(end+1) = k;\n",
    "                                end\n",
    "                            end\n",
    "                        end\n",
    "\n",
    "                        % find nadph\n",
    "                        nadph = [];\n",
    "                        for k = 1:length(model.mets)\n",
    "                            if length(model.mets{k}) == 8\n",
    "                                if strcmp(model.mets{k}(1:6),'nadph[')\n",
    "                                    nadph(end+1) = k;\n",
    "                                end\n",
    "                            end\n",
    "                        end\n",
    "\n",
    "                        % find reactions\n",
    "                        for k = 1:length(model.rxns)\n",
    "\n",
    "                            % check forward direction\n",
    "                            if model.ub(k) > 0\n",
    "\n",
    "                                % find reactant\n",
    "                                reactant_found = false;\n",
    "                                for m = find(full(model.S(:,k)) < 0)'\n",
    "                                    if any(m == nadp)\n",
    "                                        reactant_found = true;\n",
    "                                        break;\n",
    "                                    end\n",
    "                                end\n",
    "\n",
    "                                % find product\n",
    "                                product_found = false;\n",
    "                                for m = find(full(model.S(:,k)) > 0)'\n",
    "                                    if any(m == nadph)\n",
    "                                        product_found = true;\n",
    "                                        break;\n",
    "                                    end\n",
    "                                end\n",
    "\n",
    "                                % add\n",
    "                                if reactant_found && product_found && ~any(k == rxnlist)\n",
    "                                    rxnlist(end+1) = k;\n",
    "                                end\n",
    "                            end\n",
    "\n",
    "                            % check reverse direction\n",
    "                            if model.ub(k) < 0\n",
    "\n",
    "                                % find reactant\n",
    "                                reactant_found = false;\n",
    "                                for m = find(full(model.S(:,k)) > 0)'\n",
    "                                    if any(m == nadp)\n",
    "                                        reactant_found = true;\n",
    "                                        break;\n",
    "                                    end\n",
    "                                end\n",
    "\n",
    "                                % find product\n",
    "                                product_found = false;\n",
    "                                for m = find(full(model.S(:,k)) < 0)'\n",
    "                                    if any(m == nadph)\n",
    "                                        product_found = true;\n",
    "                                        break;\n",
    "                                    end\n",
    "                                end\n",
    "\n",
    "                                % add\n",
    "                                if reactant_found && product_found && ~any(k == rxnlist)\n",
    "                                    rxnlist(end+1) = k;\n",
    "                                end\n",
    "                            end\n",
    "                        end\n",
    "\n",
    "\n",
    "                    % all nadph --> nadp+ reactions\n",
    "                    case 2\n",
    "\n",
    "                        % find nadp\n",
    "                        nadp = [];\n",
    "                        for k = 1:length(model.mets)\n",
    "                            if length(model.mets{k}) == 7\n",
    "                                if strcmp(model.mets{k}(1:5),'nadp[')\n",
    "                                    nadp(end+1) = k;\n",
    "                                end\n",
    "                            end\n",
    "                        end\n",
    "\n",
    "                        % find nadph\n",
    "                        nadph = [];\n",
    "                        for k = 1:length(model.mets)\n",
    "                            if length(model.mets{k}) == 8\n",
    "                                if strcmp(model.mets{k}(1:6),'nadph[')\n",
    "                                    nadph(end+1) = k;\n",
    "                                end\n",
    "                            end\n",
    "                        end\n",
    "\n",
    "                        % find reactions\n",
    "                        for k = 1:length(model.rxns)\n",
    "\n",
    "                            % check forward direction\n",
    "                            if model.ub(k) > 0\n",
    "\n",
    "                                % find reactant\n",
    "                                reactant_found = false;\n",
    "                                for m = find(full(model.S(:,k)) < 0)'\n",
    "                                    if any(m == nadph)\n",
    "                                        reactant_found = true;\n",
    "                                        break;\n",
    "                                    end\n",
    "                                end\n",
    "\n",
    "                                % find product\n",
    "                                product_found = false;\n",
    "                                for m = find(full(model.S(:,k)) > 0)'\n",
    "                                    if any(m == nadp)\n",
    "                                        product_found = true;\n",
    "                                        break;\n",
    "                                    end\n",
    "                                end\n",
    "\n",
    "                                % add\n",
    "                                if reactant_found && product_found && ~any(k == rxnlist)\n",
    "                                    rxnlist(end+1) = k;\n",
    "                                end\n",
    "                            end\n",
    "\n",
    "                            % check reverse direction\n",
    "                            if model.ub(k) < 0\n",
    "\n",
    "                                % find reactant\n",
    "                                reactant_found = false;\n",
    "                                for m = find(full(model.S(:,k)) > 0)'\n",
    "                                    if any(m == nadph)\n",
    "                                        reactant_found = true;\n",
    "                                        break;\n",
    "                                    end\n",
    "                                end\n",
    "\n",
    "                                % find product\n",
    "                                product_found = false;\n",
    "                                for m = find(full(model.S(:,k)) < 0)'\n",
    "                                    if any(m == nadp)\n",
    "                                        product_found = true;\n",
    "                                        break;\n",
    "                                    end\n",
    "                                end\n",
    "\n",
    "                                % add\n",
    "                                if reactant_found && product_found && ~any(k == rxnlist)\n",
    "                                    rxnlist(end+1) = k;\n",
    "                                end\n",
    "                            end\n",
    "                        end\n",
    "\n",
    "                    % all nad+ --> nadh reactions\n",
    "                    case 3\n",
    "\n",
    "                        % find nad\n",
    "                        nad = [];\n",
    "                        for k = 1:length(model.mets)\n",
    "                            if length(model.mets{k}) == 6\n",
    "                                if strcmp(model.mets{k}(1:4),'nad[')\n",
    "                                    nad(end+1) = k;\n",
    "                                end\n",
    "                            end\n",
    "                        end\n",
    "\n",
    "                        % find nadh\n",
    "                        nadh = [];\n",
    "                        for k = 1:length(model.mets)\n",
    "                            if length(model.mets{k}) == 7\n",
    "                                if strcmp(model.mets{k}(1:5),'nadh[')\n",
    "                                    nadh(end+1) = k;\n",
    "                                end\n",
    "                            end\n",
    "                        end\n",
    "\n",
    "                        % find reactions\n",
    "                        for k = 1:length(model.rxns)\n",
    "\n",
    "                            % check forward direction\n",
    "                            if model.ub(k) > 0\n",
    "\n",
    "                                % find reactant\n",
    "                                reactant_found = false;\n",
    "                                for m = find(full(model.S(:,k)) < 0)'\n",
    "                                    if any(m == nad)\n",
    "                                        reactant_found = true;\n",
    "                                        break;\n",
    "                                    end\n",
    "                                end\n",
    "\n",
    "                                % find product\n",
    "                                product_found = false;\n",
    "                                for m = find(full(model.S(:,k)) > 0)'\n",
    "                                    if any(m == nadh)\n",
    "                                        product_found = true;\n",
    "                                        break;\n",
    "                                    end\n",
    "                                end\n",
    "\n",
    "                                % add\n",
    "                                if reactant_found && product_found && ~any(k == rxnlist)\n",
    "                                    rxnlist(end+1) = k;\n",
    "                                end\n",
    "                            end\n",
    "\n",
    "                            % check reverse direction\n",
    "                            if model.ub(k) < 0\n",
    "\n",
    "                                % find reactant\n",
    "                                reactant_found = false;\n",
    "                                for m = find(full(model.S(:,k)) > 0)'\n",
    "                                    if any(m == nad)\n",
    "                                        reactant_found = true;\n",
    "                                        break;\n",
    "                                    end\n",
    "                                end\n",
    "\n",
    "                                % find product\n",
    "                                product_found = false;\n",
    "                                for m = find(full(model.S(:,k)) < 0)'\n",
    "                                    if any(m == nadh)\n",
    "                                        product_found = true;\n",
    "                                        break;\n",
    "                                    end\n",
    "                                end\n",
    "\n",
    "                                % add\n",
    "                                if reactant_found && product_found && ~any(k == rxnlist)\n",
    "                                    rxnlist(end+1) = k;\n",
    "                                end\n",
    "                            end\n",
    "                        end\n",
    "\n",
    "                    % all nadh --> nad+ reactions\n",
    "                    case 4\n",
    "\n",
    "                        % find nad\n",
    "                        nad = [];\n",
    "                        for k = 1:length(model.mets)\n",
    "                            if length(model.mets{k}) == 6\n",
    "                                if strcmp(model.mets{k}(1:4),'nad[')\n",
    "                                    nad(end+1) = k;\n",
    "                                end\n",
    "                            end\n",
    "                        end\n",
    "\n",
    "                        % find nadh\n",
    "                        nadh = [];\n",
    "                        for k = 1:length(model.mets)\n",
    "                            if length(model.mets{k}) == 7\n",
    "                                if strcmp(model.mets{k}(1:5),'nadh[')\n",
    "                                    nadh(end+1) = k;\n",
    "                                end\n",
    "                            end\n",
    "                        end\n",
    "\n",
    "                        % find reactions\n",
    "                        for k = 1:length(model.rxns)\n",
    "\n",
    "                            % check forward direction\n",
    "                            if model.ub(k) > 0\n",
    "\n",
    "                                % find reactant\n",
    "                                reactant_found = false;\n",
    "                                for m = find(full(model.S(:,k)) < 0)'\n",
    "                                    if any(m == nadh)\n",
    "                                        reactant_found = true;\n",
    "                                        break;\n",
    "                                    end\n",
    "                                end\n",
    "\n",
    "                                % find product\n",
    "                                product_found = false;\n",
    "                                for m = find(full(model.S(:,k)) > 0)'\n",
    "                                    if any(m == nad)\n",
    "                                        product_found = true;\n",
    "                                        break;\n",
    "                                    end\n",
    "                                end\n",
    "\n",
    "                                % add\n",
    "                                if reactant_found && product_found && ~any(k == rxnlist)\n",
    "                                    rxnlist(end+1) = k;\n",
    "                                end\n",
    "                            end\n",
    "\n",
    "                            % check reverse direction\n",
    "                            if model.ub(k) < 0\n",
    "\n",
    "                                % find reactant\n",
    "                                reactant_found = false;\n",
    "                                for m = find(full(model.S(:,k)) > 0)'\n",
    "                                    if any(m == nadh)\n",
    "                                        reactant_found = true;\n",
    "                                        break;\n",
    "                                    end\n",
    "                                end\n",
    "\n",
    "                                % find product\n",
    "                                product_found = false;\n",
    "                                for m = find(full(model.S(:,k)) < 0)'\n",
    "                                    if any(m == nad)\n",
    "                                        product_found = true;\n",
    "                                        break;\n",
    "                                    end\n",
    "                                end\n",
    "\n",
    "                                % add\n",
    "                                if reactant_found && product_found && ~any(k == rxnlist)\n",
    "                                    rxnlist(end+1) = k;\n",
    "                                end\n",
    "                            end\n",
    "                        end\n",
    "\n",
    "\n",
    "                    % all exchange reactions\n",
    "                    case 5\n",
    "\n",
    "                        % find reactions\n",
    "                        for k = 1:length(model.rxns)\n",
    "                            if any(strcmp(exchange_rxn,model.rxns{k})) && ~any(k == rxnlist)\n",
    "                                rxnlist(end+1) = k;\n",
    "                            end\n",
    "                        end \n",
    "                end\n",
    "            end\n",
    "\n",
    "            % implement objective value constraint\n",
    "            model.S(end+1,:) = zeros(1,length(model.rxns));\n",
    "            model.S(end,find(model.c)) = 1;\n",
    "            model.b(end+1) = fva_select_fraction * result.objval;\n",
    "\n",
    "            % clear model.c\n",
    "            model.c = zeros(size(model.c));\n",
    "\n",
    "            % calculate FVA fluxes\n",
    "            flux_min = nan(length(model.rxns),1);\n",
    "            flux_max = nan(length(model.rxns),1);\n",
    "            for j = rxnlist\n",
    "\n",
    "                % set model.c\n",
    "                model.c(j) = 1;\n",
    "\n",
    "                % min value\n",
    "                model.A = model.S;\n",
    "                model.rhs = model.b;\n",
    "                model.obj = model.c;\n",
    "                model.sense = [repmat('=',1,length(model.mets)),'>'];\n",
    "                model.vtype = repmat('C',1,length(model.rxns));\n",
    "                model.varnames = model.rxns;\n",
    "                model.modelsense = 'min';\n",
    "\n",
    "                params.outputflag = 0;\n",
    "                result = gurobi(model,params);\n",
    "                try\n",
    "                    flux_min(j) = result.objval;\n",
    "                catch\n",
    "                    flux_min(j) = nan;\n",
    "                end\n",
    "\n",
    "                % max value\n",
    "                model.A = model.S;\n",
    "                model.rhs = model.b;\n",
    "                model.obj = model.c;\n",
    "                model.sense = [repmat('=',1,length(model.mets)),'>'];\n",
    "                model.vtype = repmat('C',1,length(model.rxns));\n",
    "                model.varnames = model.rxns;\n",
    "                model.modelsense = 'max';\n",
    "\n",
    "                params.outputflag = 0;\n",
    "                result = gurobi(model,params);\n",
    "                try\n",
    "                    flux_max(j) = result.objval;\n",
    "                catch\n",
    "                    flux_max(j) = nan;\n",
    "                end\n",
    "\n",
    "                % clear model.c\n",
    "                model.c = zeros(size(model.c));\n",
    "            end\n",
    "\n",
    "            % output results\n",
    "            f_fva = fopen(sprintf('results/%s/%s/fva/%s.tsv',output_folder,input_filename,samples_all{i}),'w');\n",
    "            fprintf(f_fva,'REACTION\\tMIN [mmol/gDW/hr]\\tMAX [mmol/gDW/hr]\\tLB [mmol/gDW/hr]\\tUB [mmol/gDW/hr]\\tNAME\\tFORMULA\\tSUBSYSTEM\\tGR RULE\\tEC NUMBER\\tKEGG\\tNOTES\\tREFERENCES\\n');\n",
    "            for j = rxnlist\n",
    "                fprintf(f_fva,'%s\\t%0.9f\\t%0.9f\\t%0.9f\\t%0.9f\\t%s\\t%s\\t%s\\t%s\\t%s\\t%s\\t%s\\t%s\\n',model.rxns{j},flux_min(j),flux_max(j),model.lb(j),model.ub(j),model.rxnNames{j},model.rxnFormulas{j},model.subSystems{j},model.grRules{j},model.rxnECNumbers{j},model.rxnKEGGID{j},model.rxnNotes{j},model.rxnReferences{j});\n",
    "            end\n",
    "            fclose(f_fva);\n",
    "\n",
    "        % uniform sampling\n",
    "        elseif type_of_analysis == 5\n",
    "\n",
    "            % not yet implemented\n",
    "            error('Uniform sampling not yet implemented!')\n",
    "\n",
    "        % media sensitivity - critical value\n",
    "        elseif type_of_analysis == 6\n",
    "\n",
    "            % ensure sum(model.c) = 1\n",
    "            if sum(model.c) ~= 1\n",
    "                error('ERROR - General - Media Sensitivity - Critical Value - Can only perform media sensitivity with one objective function reaction')\n",
    "            end\n",
    "            \n",
    "            % ensure only one maximization objective\n",
    "            if length(objective_function_reaction) ~= 1\n",
    "                error('ERROR - General - Type of Analysis - Can only perform media sensitivity with maximizing one objective function reaction');\n",
    "            elseif ~strcmp(objective_function_direction{1},'MAX')\n",
    "                error('ERROR - General - Type of Analysis - Can only perform media sensitivity with maximizing one objective function reaction');\n",
    "            end\n",
    "\n",
    "            % check metabolite 1 name\n",
    "            if ~any(strcmp(model.mets,sprintf('%s[e]',media_sensitivity_critical_metabolite1)))\n",
    "                error('ERROR - General - Media Sensitivity Options - Critical Value - Metabolite - Metabolite 1 not found in extracellular compartment')\n",
    "\n",
    "            % check if exchange reaction exists\n",
    "            elseif ~any(strcmp(model.rxns,sprintf('EX_%s[e]',media_sensitivity_critical_metabolite1)))\n",
    "                error('ERROR - General - Media Sensitivity Options - Critical Value - Metabolite - Metabolite 1 not involved in exchange reaction')\n",
    "\n",
    "            % ensure that exchange reaction is in exchange_rxn\n",
    "            elseif ~any(strcmp(exchange_rxn,sprintf('EX_%s[e]',media_sensitivity_critical_metabolite1)))\n",
    "                error('ERROR - General - Media Sensitivity Options - Critical Value - Metabolite - Exchange reaction for metabolite 1 not found in media composition')\n",
    "            end\n",
    "\n",
    "            % find minimum exchange value while maintaining objective value\n",
    "            original_objective = find(model.c);\n",
    "            original_objective_value = result.objval;\n",
    "            model.lb(original_objective) = media_sensitivity_critical_fraction*original_objective_value;\n",
    "            model.c = zeros(size(model.c));\n",
    "            model.c(strcmp(model.rxns,sprintf('EX_%s[e]',media_sensitivity_critical_metabolite1))) = 1;\n",
    "\n",
    "            model.A = model.S;\n",
    "            model.rhs = model.b;\n",
    "            model.obj = model.c;\n",
    "            model.sense = repmat('=',1,length(model.mets));\n",
    "            model.vtype = repmat('C',1,length(model.rxns));\n",
    "            model.varnames = model.rxns;\n",
    "            model.modelsense = 'max';\n",
    "\n",
    "            params.outputflag = 0;\n",
    "            result = gurobi(model,params);\n",
    "            critical_val = result.objval;\n",
    "\n",
    "            % if critical value is negative\n",
    "            if critical_val < 0\n",
    "\n",
    "                % save negative of value\n",
    "                fprintf(f_media,'%s\\t%s\\t%f\\n',samples_all{i},samples_all_source{i},-critical_val);\n",
    "\n",
    "            % if critical value is positive or zero\n",
    "            else\n",
    "\n",
    "                % save nan\n",
    "                fprintf(f_media,'%s\\t%s\\tnan\\n',samples_all{i},samples_all_source{i});\n",
    "            end\n",
    "\n",
    "        % media sensitivity - spectrum\n",
    "        elseif type_of_analysis == 7\n",
    "\n",
    "            % ensure sum(model.c) = 1\n",
    "            if sum(model.c) ~= 1\n",
    "                error('ERROR - General - Media Sensitivity - Can only perform media sensitivity with one objective function reaction')\n",
    "\n",
    "            % ensure only one maximization objective\n",
    "            if length(objective_function_reaction) ~= 1\n",
    "                error('ERROR - General - Type of Analysis - Can only perform media sensitivity with maximizing one objective function reaction');\n",
    "            elseif ~strcmp(objective_function_direction{1},'MAX')\n",
    "                error('ERROR - General - Type of Analysis - Can only perform media sensitivity with maximizing one objective function reaction');\n",
    "            end\n",
    "\n",
    "            % check metabolite 1 name\n",
    "            if ~any(strcmp(model.mets,sprintf('%s[e]',media_sensitivity_spectrum_metabolite1)))\n",
    "                error('ERROR - General - Media Sensitivity Options - Metabolite - Metabolite 1 not found in extracellular compartment')\n",
    "\n",
    "            % check if exchange reaction exists\n",
    "            elseif ~any(strcmp(model.rxns,sprintf('EX_%s[e]',media_sensitivity_spectrum_metabolite1)))\n",
    "                error('ERROR - General - Media Sensitivity Options - Metabolite - Metabolite 1 not involved in exchange reaction')\n",
    "\n",
    "            % ensure that exchange reaction is in exchange_rxn\n",
    "            elseif ~any(strcmp(exchange_rxn,sprintf('EX_%s[e]',media_sensitivity_spectrum_metabolite1)))\n",
    "                error('ERROR - General - Media Sensitivity Options - Metabolite - Exchange reaction for metabolite 1 not found in media composition')\n",
    "            end  \n",
    "\n",
    "            if ~isnan(media_sensitivity_spectrum_metabolite2)\n",
    "\n",
    "                % check metabolite 2 name\n",
    "                if ~any(strcmp(model.mets,sprintf('%s[e]',media_sensitivity_spectrum_metabolite2)))\n",
    "                    error('ERROR - General - Media Sensitivity Options - Metabolite - Metabolite 2 not found in extracellular compartment')\n",
    "\n",
    "                % check if exchange reaction exists\n",
    "                elseif ~any(strcmp(model.rxns,sprintf('EX_%s[e]',media_sensitivity_spectrum_metabolite2)))\n",
    "                    error('ERROR - General - Media Sensitivity Options - Metabolite - Metabolite 2 not involved in exchange reaction')\n",
    "\n",
    "                % ensure that exchange reaction is in exchange_rxn\n",
    "                elseif ~any(strcmp(exchange_rxn,sprintf('EX_%s[e]',media_sensitivity_spectrum_metabolite2)))\n",
    "                    error('ERROR - General - Media Sensitivity Options - Metabolite - Exchange reaction for metabolite 2 not found in media composition')\n",
    "                end\n",
    "            end\n",
    "\n",
    "            % if 1 metabolite\n",
    "            if isnan(media_sensitivity_spectrum_metabolite2)\n",
    "\n",
    "                % find minimum exchange value while maintaining objective value\n",
    "                original_objective = find(model.c);\n",
    "                original_objective_value = result.objval;\n",
    "                original_objective_lb = model.lb(original_objective);\n",
    "                model.lb(original_objective) = media_sensitivity_spectrum_fraction*original_objective_value;\n",
    "                model.c = zeros(size(model.c));\n",
    "                model.c(strcmp(model.rxns,sprintf('EX_%s[e]',media_sensitivity_spectrum_metabolite1))) = 1;\n",
    "\n",
    "                model.A = model.S;\n",
    "                model.rhs = model.b;\n",
    "                model.obj = model.c;\n",
    "                model.sense = repmat('=',1,length(model.mets));\n",
    "                model.vtype = repmat('C',1,length(model.rxns));\n",
    "                model.varnames = model.rxns;\n",
    "                model.modelsense = 'max';\n",
    "\n",
    "                params.outputflag = 0;\n",
    "                result = gurobi(model,params);\n",
    "                start_val = result.objval;\n",
    "\n",
    "                % if critical value is negative\n",
    "                if start_val < 0\n",
    "\n",
    "                    % return original objective function\n",
    "                    model.c = zeros(size(model.c));\n",
    "                    model.c(original_objective) = 1;\n",
    "                    model.lb(original_objective) = original_objective_lb;\n",
    "\n",
    "                    % initiailize output file\n",
    "                    f_media = fopen(sprintf('results/%s/%s/media_sensitivity/%s.tsv',output_folder,input_filename,samples_all{i}),'w');\n",
    "                    fprintf(f_media,'%s UPTAKE LB [mmol/gDW/hr]\\tOBJVAL\\n',sprintf('EX_%s[e]',media_sensitivity_spectrum_metabolite1));\n",
    "\n",
    "                    % progressively increase flux lb\n",
    "                    uptake = [];\n",
    "                    objval = [];\n",
    "                    for j = 1:media_sensitivity_spectrum_points\n",
    "\n",
    "                        % set lower bound\n",
    "                        model.lb(strcmp(model.rxns,sprintf('EX_%s[e]',media_sensitivity_spectrum_metabolite1))) = start_val*(media_sensitivity_points-j)/(media_sensitivity_points-1);\n",
    "\n",
    "                        % find objective value\n",
    "                        model.A = model.S;\n",
    "                        model.rhs = model.b;\n",
    "                        model.obj = model.c;\n",
    "                        model.sense = repmat('=',1,length(model.mets));\n",
    "                        model.vtype = repmat('C',1,length(model.rxns));\n",
    "                        model.varnames = model.rxns;\n",
    "                        model.modelsense = 'max';\n",
    "\n",
    "                        params.outputflag = 0;\n",
    "                        result = gurobi(model,params);\n",
    "\n",
    "                        % output objective value\n",
    "                        fprintf(f_media,'%f\\t%f\\n',model.lb(strcmp(model.rxns,sprintf('EX_%s[e]',media_sensitivity_spectrum_metabolite1))),result.objval);\n",
    "                        uptake(end+1) = -model.lb(strcmp(model.rxns,sprintf('EX_%s[e]',media_sensitivity_spectrum_metabolite1)));\n",
    "                        objval(end+1) = result.objval;\n",
    "                    end\n",
    "\n",
    "                    % close output file\n",
    "                    fclose(f_media);\n",
    "\n",
    "                    % create image\n",
    "                    figure; hold on;\n",
    "                    plot(uptake,objval,'k-');\n",
    "                    plot([0,max(uptake)],[max(objval),max(objval)],'k--');\n",
    "                    hold off;\n",
    "                    xlabel(sprintf('%s Uptake Rate [mmol/gDW/hr]',media_sensitivity_spectrum_metabolite1),'Interpreter','none');\n",
    "                    ylabel('Objective Value');\n",
    "                    title(sprintf('Critical %s Uptake Rate = %g mmol/gDW/hr',media_sensitivity_spectrum_metabolite1,max(uptake)),'Interpreter','none');\n",
    "                    saveas(gcf,sprintf('results/%s/%s/media_sensitivity/%s.png',output_folder,input_filename,samples_all{i}));        \n",
    "\n",
    "                % if critical value is positive or zero\n",
    "                else\n",
    "\n",
    "                    % create output file\n",
    "                    f_media = fopen(sprintf('results/%s/%s/media_sensitivity/%s.tsv',output_folder,input_filename,samples_all{i}),'w');\n",
    "                    fprintf(f_media,'%s UPTAKE LB [mmol/gDW/hr]\\tOBJVAL\\n',sprintf('EX_%s[e]',media_sensitivity_spectrum_metabolite1));    \n",
    "                    fprintf(f_media,'nan\\t%f\\n',original_objective_value);\n",
    "                    fclose(f_media);          \n",
    "                end\n",
    "\n",
    "            % if 2 metabolites\n",
    "            else\n",
    "\n",
    "                % find minimum exchanges value while maintaining objective value\n",
    "                original_objective = find(model.c);\n",
    "                original_objective_value = result.objval;\n",
    "                original_objective_lb = model.lb(original_objective);\n",
    "                model.lb(original_objective) = original_objective_value;\n",
    "                model.c = zeros(size(model.c));\n",
    "\n",
    "                % % 1-1 ratio\n",
    "                model.c(strcmp(model.rxns,sprintf('EX_%s[e]',media_sensitivity_spectrum_metabolite1))) = 1;\n",
    "                model.c(strcmp(model.rxns,sprintf('EX_%s[e]',media_sensitivity_spectrum_metabolite2))) = 1;\n",
    "\n",
    "                model.A = model.S;\n",
    "                model.rhs = model.b;\n",
    "                model.obj = model.c;\n",
    "                model.sense = repmat('=',1,length(model.mets));\n",
    "                model.vtype = repmat('C',1,length(model.rxns));\n",
    "                model.varnames = model.rxns;\n",
    "                model.modelsense = 'max';\n",
    "\n",
    "                params.outputflag = 0;\n",
    "                result = gurobi(model,params);\n",
    "                start_val1 = result.x(strcmp(model.rxns,sprintf('EX_%s[e]',media_sensitivity_spectrum_metabolite1)));\n",
    "                start_val2 = result.x(strcmp(model.rxns,sprintf('EX_%s[e]',media_sensitivity_spectrum_metabolite2)));\n",
    "\n",
    "                % % 10-1 ratio\n",
    "                model.c(strcmp(model.rxns,sprintf('EX_%s[e]',media_sensitivity_spectrum_metabolite1))) = 10;\n",
    "                model.c(strcmp(model.rxns,sprintf('EX_%s[e]',media_sensitivity_spectrum_metabolite2))) = 1;\n",
    "\n",
    "                model.A = model.S;\n",
    "                model.rhs = model.b;\n",
    "                model.obj = model.c;\n",
    "                model.sense = repmat('=',1,length(model.mets));\n",
    "                model.vtype = repmat('C',1,length(model.rxns));\n",
    "                model.varnames = model.rxns;\n",
    "                model.modelsense = 'max';\n",
    "\n",
    "                params.outputflag = 0;\n",
    "                result = gurobi(model,params);\n",
    "                if result.x(strcmp(model.rxns,sprintf('EX_%s[e]',media_sensitivity_spectrum_metabolite1))) < start_val1\n",
    "                    start_val1 = result.x(strcmp(model.rxns,sprintf('EX_%s[e]',media_sensitivity_spectrum_metabolite1)));\n",
    "                end\n",
    "                if result.x(strcmp(model.rxns,sprintf('EX_%s[e]',media_sensitivity_spectrum_metabolite2))) < start_val2\n",
    "                    start_val2 = result.x(strcmp(model.rxns,sprintf('EX_%s[e]',media_sensitivity_spectrum_metabolite2)));\n",
    "                end\n",
    "\n",
    "                % % 1-10 ratio\n",
    "                model.c(strcmp(model.rxns,sprintf('EX_%s[e]',media_sensitivity_spectrum_metabolite1))) = 1;\n",
    "                model.c(strcmp(model.rxns,sprintf('EX_%s[e]',media_sensitivity_spectrum_metabolite2))) = 10;\n",
    "\n",
    "                model.A = model.S;\n",
    "                model.rhs = model.b;\n",
    "                model.obj = model.c;\n",
    "                model.sense = repmat('=',1,length(model.mets));\n",
    "                model.vtype = repmat('C',1,length(model.rxns));\n",
    "                model.varnames = model.rxns;\n",
    "                model.modelsense = 'max';\n",
    "\n",
    "                params.outputflag = 0;\n",
    "                result = gurobi(model,params);\n",
    "                if result.x(strcmp(model.rxns,sprintf('EX_%s[e]',media_sensitivity_spectrum_metabolite1))) < start_val1\n",
    "                    start_val1 = result.x(strcmp(model.rxns,sprintf('EX_%s[e]',media_sensitivity_spectrum_metabolite1)));\n",
    "                end\n",
    "                if result.x(strcmp(model.rxns,sprintf('EX_%s[e]',media_sensitivity_spectrum_metabolite2))) < start_val2\n",
    "                    start_val2 = result.x(strcmp(model.rxns,sprintf('EX_%s[e]',media_sensitivity_spectrum_metabolite2)));\n",
    "                end\n",
    "\n",
    "                % if both critical values are negative\n",
    "                if (start_val1 < 0) && (start_val2 < 0)\n",
    "\n",
    "                    % result original objective function\n",
    "                    model.c = zeros(size(model.c));\n",
    "                    model.c(original_objective) = 1;\n",
    "                    model.lb(original_objective) = original_objective_lb;\n",
    "\n",
    "                    % initiailize output file\n",
    "                    f_media = fopen(sprintf('results/%s/%s/media_sensitivity/%s.tsv',output_folder,input_filename,samples_all{i}),'w');\n",
    "                    fprintf(f_media,'%s UPTAKE LB [mmol/gDW/hr]\\t%s UPTAKE LB [mmol/gDW/hr]\\tOBJVAL\\n',sprintf('EX_%s[e]',media_sensitivity_spectrum_metabolite1),sprintf('EX_%s[e]',media_sensitivity_spectrum_metabolite2));\n",
    "\n",
    "                    % progressively increase flux lb\n",
    "                    uptake1 = [];\n",
    "                    uptake2 = [];\n",
    "                    objval = [];\n",
    "                    for j = 1:media_sensitivity_points\n",
    "                        for k = 1:media_sensitivity_points\n",
    "\n",
    "                            % set lower bound\n",
    "                            model.lb(strcmp(model.rxns,sprintf('EX_%s[e]',media_sensitivity_spectrum_metabolite1))) = start_val1*(media_sensitivity_points-j)/(media_sensitivity_points-1);\n",
    "                            model.lb(strcmp(model.rxns,sprintf('EX_%s[e]',media_sensitivity_spectrum_metabolite2))) = start_val2*(media_sensitivity_points-k)/(media_sensitivity_points-1);\n",
    "\n",
    "                            % find objective value\n",
    "                            model.A = model.S;\n",
    "                            model.rhs = model.b;\n",
    "                            model.obj = model.c;\n",
    "                            model.sense = repmat('=',1,length(model.mets));\n",
    "                            model.vtype = repmat('C',1,length(model.rxns));\n",
    "                            model.varnames = model.rxns;\n",
    "                            model.modelsense = 'max';\n",
    "\n",
    "                            params.outputflag = 0;\n",
    "                            result = gurobi(model,params);\n",
    "\n",
    "                            % output objective value\n",
    "                            fprintf(f_media,'%f\\t%f\\t%f\\n',model.lb(strcmp(model.rxns,sprintf('EX_%s[e]',media_sensitivity_spectrum_metabolite1))),model.lb(strcmp(model.rxns,sprintf('EX_%s[e]',media_sensitivity_spectrum_metabolite2))),result.objval);\n",
    "                            uptake1(end+1) = -model.lb(strcmp(model.rxns,sprintf('EX_%s[e]',media_sensitivity_spectrum_metabolite1)));\n",
    "                            uptake2(end+1) = -model.lb(strcmp(model.rxns,sprintf('EX_%s[e]',media_sensitivity_spectrum_metabolite2)));\n",
    "                            objval(end+1) = result.objval;\n",
    "                        end\n",
    "                    end            \n",
    "\n",
    "                    % close output file\n",
    "                    fclose(f_media);\n",
    "\n",
    "                    % create image\n",
    "                    Z = reshape(objval,media_sensitivity_points,media_sensitivity_points);\n",
    "                    uptake2 = uptake2;\n",
    "                    figure;\n",
    "                    imagesc(uptake1,uptake2,Z);\n",
    "                    xlabel(sprintf('%s Uptake Rate [mmol/gDW/hr]',media_sensitivity_spectrum_metabolite1),'Interpreter','none');\n",
    "                    ylabel(sprintf('%s Uptake Rate [mmol/gDW/hr]',media_sensitivity_spectrum_metabolite2),'Interpreter','none');\n",
    "                    c = colorbar;\n",
    "                    c.Label.String = 'Objective Value';\n",
    "                    c.Label.FontSize = 12;\n",
    "                    set(gca,'YDir','normal');\n",
    "                    saveas(gcf,sprintf('results/%s/%s/media_sensitivity/%s.png',output_folder,input_filename,samples_all{i}));\n",
    "\n",
    "                % if one critical value is not negative\n",
    "                else\n",
    "\n",
    "                    % create output file\n",
    "                    f_media = fopen(sprintf('results/%s/%s/media_sensitivity/%s.tsv',output_folder,input_filename,samples_all{i}),'w');\n",
    "                    fprintf(f_media,'%s UPTAKE LB [mmol/gDW/hr]\\t%s UPTAKE LB [mmol/gDW/hr]\\tOBJVAL\\n',sprintf('EX_%s[e]',media_sensitivity_spectrum_metabolite1),sprintf('EX_%s[e]',media_sensitivity_spectrum_metabolite2));\n",
    "                    fprintf(f_media,'nan\\tnan\\t%f\\n',original_objective_value);\n",
    "                    fclose(f_media);\n",
    "                end\n",
    "            end\n",
    "        end\n",
    "    end\n",
    "        \n",
    "    % close objective value file\n",
    "    %fclose(f_obj);\n",
    "    \n",
    "    % close critical media sensitivity file\n",
    "    if type_of_analysis == 6\n",
    "        fclose(f_media);\n",
    "    end\n",
    "\n",
    "end"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Master Cell"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": false
   },
   "outputs": [],
   "source": [
    "tic\n",
    "\n",
    "% if gene knockdown screen\n",
    "if gene_knockdown_screen\n",
    "\n",
    "    % parse info from first input file\n",
    "    input_file = input_files(1);\n",
    "\n",
    "    % parse input file - general\n",
    "    [type_of_analysis,pfba_fraction,fva_all_fraction,fva_select_fraction,fva_select_groups,fva_select_reactions,sampler_fraction,sampler_samples,sampler_steps,media_sensitivity_critical_fraction,media_sensitivity_critical_metabolite1,media_sensitivity_spectrum_fraction,media_sensitivity_spectrum_metabolite1,media_sensitivity_spectrum_metabolite2,media_sensitivity_spectrum_points] = parse_general(input_file{1});\n",
    "\n",
    "    % parse input file - constraints - proliferation\n",
    "    [constraints_proliferation,constraints_proliferation_type,constraints_proliferation_value] = parse_constraints_proliferation(input_file{1});\n",
    "\n",
    "    % parse input file - constraints - value\n",
    "    [constraints_value_type,constraints_value_value,constraints_value_reaction,constraints_value_count] = parse_constraints_value(input_file{1});\n",
    "\n",
    "    % parse input file - constraints - fraction\n",
    "    [constraints_fraction_type,constraints_fraction_value,constraints_fraction_reaction,constraints_fraction_count] = parse_constraints_fraction(input_file{1});\n",
    "    \n",
    "    % parse input file - constraints - max multi\n",
    "    [constraints_maxmulti_weight,constraints_maxmulti_reaction,constraints_maxmulti_fraction,constraints_maxmulti_count] = parse_constraints_maxmulti(input_file{1});\n",
    "    \n",
    "    % parse input file - objective function\n",
    "    [objective_function_direction,objective_function_weight,objective_function_reaction] = parse_objective(input_file{1});\n",
    "\n",
    "    % parse input file - drugs\n",
    "    [required_drugs] = parse_drugs(input_file{1});\n",
    "\n",
    "    % parse input file - mutations\n",
    "    [use_mutation] = parse_mutations(input_file{1});\n",
    "\n",
    "    % parse input file - samples - TCGA\n",
    "    [samples_all,samples_all_source,samples_all_source_protein] = parse_samples_tcga(input_file{1},constraints_proliferation,required_drugs,use_mutation);\n",
    "    \n",
    "    % parse input file - samples - CCLE\n",
    "    [samples_all,samples_all_source,samples_all_source_protein] = parse_samples_ccle(input_file{1},constraints_proliferation,required_drugs,use_mutation,samples_all,samples_all_source,samples_all_source_protein);\n",
    "    \n",
    "    % parse input file - samples - GTEx\n",
    "    [samples_all,samples_all_source,samples_all_source_protein] = parse_samples_gtex(input_file{1},constraints_proliferation,required_drugs,use_mutation,samples_all,samples_all_source,samples_all_source_protein);\n",
    "    \n",
    "    % parse input file - samples - other\n",
    "    [samples_all,samples_all_source,samples_all_source_protein] = parse_samples_other(input_file{1},constraints_proliferation,required_drugs,use_mutation,samples_all,samples_all_source,samples_all_source_protein);\n",
    "    \n",
    "    % make sure at least one sample chosen\n",
    "    if isempty(samples_all)\n",
    "        error('ERROR - Samples - Must choose at least one sample')\n",
    "    end\n",
    "\n",
    "    % parse input file - media\n",
    "    [media_choice,media_constraint_metabolite,media_constraint_uptake] = parse_media(input_file{1});\n",
    "\n",
    "    % parse input file - concentrations\n",
    "    [recalculate_thermodynamics,default_concentration_lb,default_concentration_ub,concentration_ranges_all,concentration_ranges_sample_folder,concentration_ranges_sample] = parse_concentrations(input_file{1});\n",
    "    \n",
    "    % parse input file - modules\n",
    "    [catalase_version, module_5fu, module_cis, module_cpa, module_dox] = parse_modules(input_file{1});\n",
    "    \n",
    "    % iterate over input files\n",
    "    %parpool(parallel_cores);\n",
    "    %parfor a = 1:length(input_files)\n",
    "    for a = 1:length(input_files)\n",
    "        input_file = input_files(a);\n",
    "    \n",
    "        % parse input file - gene knockdown\n",
    "        [knockdown_genes,knockdown_fractions] = parse_knockdown(input_file{1});\n",
    "        \n",
    "        % load model\n",
    "        [model,concentration_mets,concentration_values,constraints_fraction_id] = load_model(constraints_proliferation,constraints_proliferation_type,constraints_proliferation_value,constraints_value_type,constraints_value_value,constraints_value_reaction,constraints_value_count,constraints_fraction_type,constraints_fraction_value,constraints_fraction_reaction,constraints_fraction_count,constraints_maxmulti_weight,constraints_maxmulti_reaction,constraints_maxmulti_count,objective_function_direction,objective_function_weight,objective_function_reaction);\n",
    "        \n",
    "        % load media\n",
    "        [model,concentration_mets,concentration_values,exchange_rxn] = load_media(model,media_choice,media_constraint_metabolite,media_constraint_uptake,concentration_mets,concentration_values,module_5fu,module_cis,module_cpa,module_dox);\n",
    "\n",
    "        % load deltag values\n",
    "        [deltag_lower,deltag_upper] = load_deltag(model,recalculate_thermodynamics,default_concentration_lb,default_concentration_ub,concentration_ranges_all,concentration_ranges_sample_folder,concentration_ranges_sample,concentration_mets,concentration_values,samples_all,samples_all_source);\n",
    "\n",
    "        % calculate Vmax for each sample\n",
    "        [vmax_forward,vmax_reverse] = calculate_vmax(model,samples_all,samples_all_source,samples_all_source_protein,use_mutation,knockdown_genes,knockdown_fractions,gene_knockdown_screen);\n",
    "\n",
    "        % run FBA for each sample\n",
    "        run_fba(output_folder,input_file{1},model,type_of_analysis,samples_all,samples_all_source,samples_all_source_protein,vmax_forward,vmax_reverse,objective_function_direction,objective_function_weight,objective_function_reaction,exchange_rxn,deltag_lower,deltag_upper,constraints_fraction_count,constraints_fraction_type,constraints_fraction_reaction,constraints_fraction_value,constraints_fraction_id,constraints_maxmulti_fraction,constraints_maxmulti_count,pfba_fraction,fva_all_fraction,fva_select_fraction,fva_select_groups,fva_select_reactions,sampler_fraction,sampler_samples,sampler_steps,media_sensitivity_critical_fraction,media_sensitivity_critical_metabolite1,media_sensitivity_spectrum_fraction,media_sensitivity_spectrum_metabolite1,media_sensitivity_spectrum_metabolite2,media_sensitivity_spectrum_points,parallel_cores,knockdown_genes,knockdown_fractions,catalase_version,module_5fu,module_cis,module_cpa,module_dox);\n",
    "    end\n",
    "\n",
    "% if objective function screen\n",
    "elseif objective_function_screen\n",
    "\n",
    "    % parse info from first input file\n",
    "    input_file = input_files(1);\n",
    "\n",
    "    % parse input file - general\n",
    "    [type_of_analysis,pfba_fraction,fva_all_fraction,fva_select_fraction,fva_select_groups,fva_select_reactions,sampler_fraction,sampler_samples,sampler_steps,media_sensitivity_critical_fraction,media_sensitivity_critical_metabolite1,media_sensitivity_spectrum_fraction,media_sensitivity_spectrum_metabolite1,media_sensitivity_spectrum_metabolite2,media_sensitivity_spectrum_points] = parse_general(input_file{1});\n",
    "\n",
    "    % parse input file - constraints - proliferation\n",
    "    [constraints_proliferation,constraints_proliferation_type,constraints_proliferation_value] = parse_constraints_proliferation(input_file{1});\n",
    "\n",
    "    % parse input file - constraints - value\n",
    "    [constraints_value_type,constraints_value_value,constraints_value_reaction,constraints_value_count] = parse_constraints_value(input_file{1});\n",
    "\n",
    "    % parse input file - constraints - fraction\n",
    "    [constraints_fraction_type,constraints_fraction_value,constraints_fraction_reaction,constraints_fraction_count] = parse_constraints_fraction(input_file{1});\n",
    "    \n",
    "    % parse input file - constraints - max multi\n",
    "    [constraints_maxmulti_weight,constraints_maxmulti_reaction,constraints_maxmulti_fraction,constraints_maxmulti_count] = parse_constraints_maxmulti(input_file{1});\n",
    "\n",
    "    % parse input file - drugs\n",
    "    [required_drugs] = parse_drugs(input_file{1});\n",
    "\n",
    "    % parse input file - mutations\n",
    "    [use_mutation] = parse_mutations(input_file{1});\n",
    "\n",
    "    % parse input file - samples - TCGA\n",
    "    [samples_all,samples_all_source,samples_all_source_protein] = parse_samples_tcga(input_file{1},constraints_proliferation,required_drugs,use_mutation);\n",
    "    \n",
    "    % parse input file - samples - CCLE\n",
    "    [samples_all,samples_all_source,samples_all_source_protein] = parse_samples_ccle(input_file{1},constraints_proliferation,required_drugs,use_mutation,samples_all,samples_all_source,samples_all_source_protein);\n",
    "    \n",
    "    % parse input file - samples - GTEx\n",
    "    [samples_all,samples_all_source,samples_all_source_protein] = parse_samples_gtex(input_file{1},constraints_proliferation,required_drugs,use_mutation,samples_all,samples_all_source,samples_all_source_protein);\n",
    "    \n",
    "    % parse input file - samples - other\n",
    "    [samples_all,samples_all_source,samples_all_source_protein] = parse_samples_other(input_file{1},constraints_proliferation,required_drugs,use_mutation,samples_all,samples_all_source,samples_all_source_protein);\n",
    "    \n",
    "    % make sure at least one sample chosen\n",
    "    if isempty(samples_all)\n",
    "        error('ERROR - Samples - Must choose at least one sample')\n",
    "    end\n",
    "\n",
    "    % parse input file - media\n",
    "    [media_choice,media_constraint_metabolite,media_constraint_uptake] = parse_media(input_file{1});\n",
    "\n",
    "    % parse input file - concentrations\n",
    "    [recalculate_thermodynamics,default_concentration_lb,default_concentration_ub,concentration_ranges_all,concentration_ranges_sample_folder,concentration_ranges_sample] = parse_concentrations(input_file{1});\n",
    "    \n",
    "    % parse input file - modules\n",
    "    [catalase_version, module_5fu, module_cis, module_cpa, module_dox] = parse_modules(input_file{1});\n",
    "    \n",
    "    % parse input file - gene knockdown\n",
    "    [knockdown_genes,knockdown_fractions] = parse_knockdown(input_file{1});\n",
    "    \n",
    "    % iterate over input files\n",
    "    %parpool(parallel_cores);\n",
    "    %parfor a = 1:length(input_files)\n",
    "    for a = 1:length(input_files)\n",
    "        input_file = input_files(a);\n",
    "    \n",
    "        % parse input file - objective function\n",
    "        [objective_function_direction,objective_function_weight,objective_function_reaction] = parse_objective(input_file{1});\n",
    "        \n",
    "        % load model\n",
    "        [model,concentration_mets,concentration_values,constraints_fraction_id] = load_model(constraints_proliferation,constraints_proliferation_type,constraints_proliferation_value,constraints_value_type,constraints_value_value,constraints_value_reaction,constraints_value_count,constraints_fraction_type,constraints_fraction_value,constraints_fraction_reaction,constraints_fraction_count,constraints_maxmulti_weight,constraints_maxmulti_reaction,constraints_maxmulti_count,objective_function_direction,objective_function_weight,objective_function_reaction);\n",
    "        \n",
    "        % load media\n",
    "        [model,concentration_mets,concentration_values,exchange_rxn] = load_media(model,media_choice,media_constraint_metabolite,media_constraint_uptake,concentration_mets,concentration_values,module_5fu,module_cis,module_cpa,module_dox);\n",
    "\n",
    "        % load deltag values\n",
    "        [deltag_lower,deltag_upper] = load_deltag(model,recalculate_thermodynamics,default_concentration_lb,default_concentration_ub,concentration_ranges_all,concentration_ranges_sample_folder,concentration_ranges_sample,concentration_mets,concentration_values,samples_all,samples_all_source);\n",
    "\n",
    "        % calculate Vmax for each sample\n",
    "        [vmax_forward,vmax_reverse] = calculate_vmax(model,samples_all,samples_all_source,samples_all_source_protein,use_mutation,knockdown_genes,knockdown_fractions,gene_knockdown_screen);\n",
    "\n",
    "        % run FBA for each sample\n",
    "        run_fba(output_folder,input_file{1},model,type_of_analysis,samples_all,samples_all_source,samples_all_source_protein,vmax_forward,vmax_reverse,objective_function_direction,objective_function_weight,objective_function_reaction,exchange_rxn,deltag_lower,deltag_upper,constraints_fraction_count,constraints_fraction_type,constraints_fraction_reaction,constraints_fraction_value,constraints_fraction_id,constraints_maxmulti_fraction,constraints_maxmulti_count,pfba_fraction,fva_all_fraction,fva_select_fraction,fva_select_groups,fva_select_reactions,sampler_fraction,sampler_samples,sampler_steps,media_sensitivity_critical_fraction,media_sensitivity_critical_metabolite1,media_sensitivity_spectrum_fraction,media_sensitivity_spectrum_metabolite1,media_sensitivity_spectrum_metabolite2,media_sensitivity_spectrum_points,parallel_cores,knockdown_genes,knockdown_fractions,catalase_version,module_5fu,module_cis,module_cpa,module_dox);\n",
    "    end\n",
    "\n",
    "% if not either\n",
    "else\n",
    "        \n",
    "    % iterate over input files\n",
    "    %parpool(parallel_cores);\n",
    "    %parfor a = 1:length(input_files)\n",
    "    for a = 1:length(input_files)\n",
    "        input_file = input_files(a);\n",
    "\n",
    "        % parse input file - general\n",
    "        [type_of_analysis,pfba_fraction,fva_all_fraction,fva_select_fraction,fva_select_groups,fva_select_reactions,sampler_fraction,sampler_samples,sampler_steps,media_sensitivity_critical_fraction,media_sensitivity_critical_metabolite1,media_sensitivity_spectrum_fraction,media_sensitivity_spectrum_metabolite1,media_sensitivity_spectrum_metabolite2,media_sensitivity_spectrum_points] = parse_general(input_file{1});\n",
    "\n",
    "        % parse input file - constraints - proliferation\n",
    "        [constraints_proliferation,constraints_proliferation_type,constraints_proliferation_value] = parse_constraints_proliferation(input_file{1});\n",
    "\n",
    "        % parse input file - constraints - value\n",
    "        [constraints_value_type,constraints_value_value,constraints_value_reaction,constraints_value_count] = parse_constraints_value(input_file{1});\n",
    "\n",
    "        % parse input file - constraints - fraction\n",
    "        [constraints_fraction_type,constraints_fraction_value,constraints_fraction_reaction,constraints_fraction_count] = parse_constraints_fraction(input_file{1});\n",
    "        \n",
    "        % parse input file - constraints - max multi\n",
    "        [constraints_maxmulti_weight,constraints_maxmulti_reaction,constraints_maxmulti_fraction,constraints_maxmulti_count] = parse_constraints_maxmulti(input_file{1});\n",
    "\n",
    "        % parse input file - objective function\n",
    "        [objective_function_direction,objective_function_weight,objective_function_reaction] = parse_objective(input_file{1});\n",
    "        \n",
    "        % parse input file - drugs\n",
    "        [required_drugs] = parse_drugs(input_file{1});\n",
    "\n",
    "        % parse input file - mutations\n",
    "        [use_mutation] = parse_mutations(input_file{1});\n",
    "\n",
    "        % parse input file - samples - TCGA\n",
    "        [samples_all,samples_all_source,samples_all_source_protein] = parse_samples_tcga(input_file{1},constraints_proliferation,required_drugs,use_mutation);\n",
    "        \n",
    "        % parse input file - samples - CCLE\n",
    "        [samples_all,samples_all_source,samples_all_source_protein] = parse_samples_ccle(input_file{1},constraints_proliferation,required_drugs,use_mutation,samples_all,samples_all_source,samples_all_source_protein);\n",
    "        \n",
    "        % parse input file - samples - GTEx\n",
    "        [samples_all,samples_all_source,samples_all_source_protein] = parse_samples_gtex(input_file{1},constraints_proliferation,required_drugs,use_mutation,samples_all,samples_all_source,samples_all_source_protein);\n",
    "         \n",
    "        % parse input file - samples - other\n",
    "        [samples_all,samples_all_source,samples_all_source_protein] = parse_samples_other(input_file{1},constraints_proliferation,required_drugs,use_mutation,samples_all,samples_all_source,samples_all_source_protein);\n",
    "    \n",
    "        % make sure at least one sample chosen\n",
    "        if isempty(samples_all)\n",
    "            error('ERROR - Samples - Must choose at least one sample')\n",
    "        end\n",
    "\n",
    "        % parse input file - media\n",
    "        [media_choice,media_constraint_metabolite,media_constraint_uptake] = parse_media(input_file{1});\n",
    "\n",
    "        % parse input file - concentrations\n",
    "        [recalculate_thermodynamics,default_concentration_lb,default_concentration_ub,concentration_ranges_all,concentration_ranges_sample_folder,concentration_ranges_sample] = parse_concentrations(input_file{1});\n",
    "\n",
    "        % parse input file - modules\n",
    "        [catalase_version, module_5fu, module_cis, module_cpa, module_dox] = parse_modules(input_file{1});\n",
    "\n",
    "        % parse input file - gene knockdown\n",
    "        [knockdown_genes,knockdown_fractions] = parse_knockdown(input_file{1});\n",
    "\n",
    "        % load model\n",
    "        [model,concentration_mets,concentration_values,constraints_fraction_id] = load_model(constraints_proliferation,constraints_proliferation_type,constraints_proliferation_value,constraints_value_type,constraints_value_value,constraints_value_reaction,constraints_value_count,constraints_fraction_type,constraints_fraction_value,constraints_fraction_reaction,constraints_fraction_count,constraints_maxmulti_weight,constraints_maxmulti_reaction,constraints_maxmulti_count,objective_function_direction,objective_function_weight,objective_function_reaction);\n",
    "        \n",
    "        % load media\n",
    "        [model,concentration_mets,concentration_values,exchange_rxn] = load_media(model,media_choice,media_constraint_metabolite,media_constraint_uptake,concentration_mets,concentration_values,module_5fu,module_cis,module_cpa,module_dox);\n",
    "\n",
    "        % load deltag values\n",
    "        [deltag_lower,deltag_upper] = load_deltag(model,recalculate_thermodynamics,default_concentration_lb,default_concentration_ub,concentration_ranges_all,concentration_ranges_sample_folder,concentration_ranges_sample,concentration_mets,concentration_values,samples_all,samples_all_source);\n",
    "\n",
    "        % calculate Vmax for each sample\n",
    "        [vmax_forward,vmax_reverse] = calculate_vmax(model,samples_all,samples_all_source,samples_all_source_protein,use_mutation,knockdown_genes,knockdown_fractions,gene_knockdown_screen);\n",
    "\n",
    "        % run FBA for each sample\n",
    "        run_fba(output_folder,input_file{1},model,type_of_analysis,samples_all,samples_all_source,samples_all_source_protein,vmax_forward,vmax_reverse,objective_function_direction,objective_function_weight,objective_function_reaction,exchange_rxn,deltag_lower,deltag_upper,constraints_fraction_count,constraints_fraction_type,constraints_fraction_reaction,constraints_fraction_value,constraints_fraction_id,constraints_maxmulti_fraction,constraints_maxmulti_count,pfba_fraction,fva_all_fraction,fva_select_fraction,fva_select_groups,fva_select_reactions,sampler_fraction,sampler_samples,sampler_steps,media_sensitivity_critical_fraction,media_sensitivity_critical_metabolite1,media_sensitivity_spectrum_fraction,media_sensitivity_spectrum_metabolite1,media_sensitivity_spectrum_metabolite2,media_sensitivity_spectrum_points,parallel_cores,knockdown_genes,knockdown_fractions,catalase_version,module_5fu,module_cis,module_cpa,module_dox);\n",
    "    end    \n",
    "end\n",
    "\n",
    "toc"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Matlab",
   "language": "matlab",
   "name": "matlab"
  },
  "language_info": {
   "codemirror_mode": "octave",
   "file_extension": ".m",
   "help_links": [
    {
     "text": "MetaKernel Magics",
     "url": "https://metakernel.readthedocs.io/en/latest/source/README.html"
    }
   ],
   "mimetype": "text/x-octave",
   "name": "matlab",
   "version": "0.16.7"
  },
  "toc": {
   "nav_menu": {},
   "number_sections": true,
   "sideBar": true,
   "skip_h1_title": true,
   "title_cell": "Table of Contents",
   "title_sidebar": "Contents",
   "toc_cell": true,
   "toc_position": {},
   "toc_section_display": true,
   "toc_window_display": false
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
