{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Data Pipeline - Recon3D metabolic network"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Input Data"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "'Recon3D_301.mat' and 'Recon3DModel_301.mat' - Go to https://vmh.uni.lu/#downloadview, choose 'Recon3D_3.01', unzip zip file."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Code"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Initialize Recon3D model"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Load Recon3D model"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "load('input_data/Recon3D_301.mat');\n",
    "load('input_data/Recon3DModel_301.mat');\n",
    "model = Recon3D;"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Fix reaction subsystems"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "% model.subSystems\n",
    "for i = 1:length(model.rxns)\n",
    "    model.subSystems{i} = model.subSystems{i}{1};\n",
    "end"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Clear objective values"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "model.c = zeros(size(model.c));"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Clear dm/dt"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "model.b = zeros(size(model.b));"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Remove erroneous energy-generating cycles"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Implement bounds from Recon3DModel"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "for i = 1:length(model.rxns)\n",
    "    if any(strcmp(Recon3DModel.rxns,model.rxns{i}))\n",
    "        model.lb(i) = Recon3DModel.lb(strcmp(Recon3DModel.rxns,model.rxns{i}));\n",
    "        model.ub(i) = Recon3DModel.ub(strcmp(Recon3DModel.rxns,model.rxns{i}));\n",
    "    else\n",
    "        model.lb(i) = 0;\n",
    "        model.ub(i) = 0;\n",
    "    end\n",
    "end"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Add back bounds from Recon3D if doesn't cause erroneous energy-generating cycles"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Erroneous energy-generating cycles to test (from Fritzemeier et al):<br>\n",
    "atp[c] + h2o[c] --> adp[c] + h[c] + pi[c]<br>\n",
    "ctp[c] + h2o[c] --> cdp[c] + h[c] + pi[c]<br>\n",
    "gtp[c] + h2o[c] --> gdp[c] + h[c] + pi[c]<br>\n",
    "utp[c] + h2o[c] --> udp[c] + h[c] + pi[c]<br>\n",
    "itp[c] + h2o[c] --> idp[c] + h[c] + pi[c]<br>\n",
    "nadh[c] --> h[c] + nad[c]<br>\n",
    "nadph[c] --> h[c] + nadp[c]<br>\n",
    "fadh2[c] --> 2 h[c] + fad[c]<br>\n",
    "q10h2[c] --> 2 h[c] + q10[c]<br>\n",
    "h2o[c] + accoa[c] --> h[c] + ac[c] + coa[c]<br>\n",
    "glu_L[c] + h2o[c] --> akg[c] + nh4[c] + h[c]<br>\n",
    "h[i] --> h[c]<br>"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Restrict exchange fluxes to 0, all other fluxes to +/- 1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "for i = 1:length(model.rxns)\n",
    "    if (model.lb(i) ~= 0) || (model.ub(i) ~= 0)\n",
    "        if strcmp(model.subSystems{i},'Exchange/demand reaction')\n",
    "            if length(model.rxns{i}) >= 3\n",
    "                if strcmp(model.rxns{i}(1:3),'EX_')\n",
    "                    model.lb(i) = 0;\n",
    "                    model.ub(i) = 0;\n",
    "                else\n",
    "                    model.lb(i) = 0;\n",
    "                    model.ub(i) = 0;\n",
    "                end\n",
    "            else\n",
    "                model.lb(i) = 0;\n",
    "                model.ub(i) = 0;\n",
    "            end     \n",
    "        else\n",
    "            if model.lb(i) < 0\n",
    "                model.lb(i) = -1;\n",
    "            else\n",
    "                model.lb(i) = 0;\n",
    "            end\n",
    "            if model.ub(i) > 0\n",
    "                model.ub(i) = 1;\n",
    "            else\n",
    "                model.ub(i) = 0;\n",
    "            end\n",
    "        end\n",
    "    end\n",
    "end"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Add reactions sequentually, keep if still no erroneous energy-generating cycles"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "for i = 1:length(model.rxns)\n",
    "    \n",
    "    % if not exchange reaction\n",
    "    if ~strcmp(model.subSystems{i},'Exchange/demand reaction')\n",
    "    \n",
    "        % forward direction\n",
    "        if (model.ub(i) == 0) && (Recon3D.ub(i) > 0)\n",
    "            \n",
    "            % set new upper bound\n",
    "            model.ub(i) = 1;\n",
    "            model_keep = model;\n",
    "            \n",
    "            % test dissipation reactions\n",
    "            objectives = [];\n",
    "\n",
    "            % atp\n",
    "            model.rxns{end+1} = 'energy_dissipation_atp';\n",
    "            model.S(:,end+1) = zeros(length(model.mets),1);\n",
    "            model.S(strcmp(model.mets,'atp[c]'),end) = -1;\n",
    "            model.S(strcmp(model.mets,'h2o[c]'),end) = -1;\n",
    "            model.S(strcmp(model.mets,'adp[c]'),end) = 1;\n",
    "            model.S(strcmp(model.mets,'h[c]'),end) = 1;\n",
    "            model.S(strcmp(model.mets,'pi[c]'),end) = 1;\n",
    "            model.lb(end+1) = 0;\n",
    "            model.ub(end+1) = 9999999;\n",
    "            model.c(end+1) = 1;\n",
    "            model.rxnGeneMat(end+1,:) = zeros(1,length(model.genes));\n",
    "            model.rules{end+1} = '';\n",
    "            model.grRules{end+1} = '';\n",
    "            model.subSystems{end+1} = 'Energy dissipation reaction';\n",
    "            model.rxnNames{end+1} = 'Energy dissipation - ATP';\n",
    "            model.rxnKEGGID{end+1} = '';\n",
    "            model.rxnKeggOrthology{end+1} = '';\n",
    "            model.rxnConfidenceScores(end+1) = 0';\n",
    "            model.rxnReferences{end+1} = '';\n",
    "            model.rxnECNumbers{end+1} = '';\n",
    "            model.rxnNotes{end+1} = '';\n",
    "            model.rxnCOG{end+1} = '';\n",
    "            model.rxnReconMap{end+1} = '';\n",
    "\n",
    "            model.A = model.S;\n",
    "            model.rhs = model.b;\n",
    "            model.obj = model.c;\n",
    "            model.sense = repmat('=',1,length(model.mets));\n",
    "            model.vtype = repmat('C',1,length(model.rxns));\n",
    "            model.varnames = model.rxns;\n",
    "            model.modelsense = 'max';\n",
    "\n",
    "            params.outputflag = 0;\n",
    "            result = gurobi(model,params);\n",
    "            objectives(end+1) = result.objval;\n",
    "\n",
    "            model.c(end) = 0;\n",
    "            model.lb(end) = 0;\n",
    "            model.ub(end) = 0;\n",
    "\n",
    "            % ctp\n",
    "            model.rxns{end+1} = 'energy_dissipation_ctp';\n",
    "            model.S(:,end+1) = zeros(length(model.mets),1);\n",
    "            model.S(strcmp(model.mets,'ctp[c]'),end) = -1;\n",
    "            model.S(strcmp(model.mets,'h2o[c]'),end) = -1;\n",
    "            model.S(strcmp(model.mets,'cdp[c]'),end) = 1;\n",
    "            model.S(strcmp(model.mets,'h[c]'),end) = 1;\n",
    "            model.S(strcmp(model.mets,'pi[c]'),end) = 1;\n",
    "            model.lb(end+1) = 0;\n",
    "            model.ub(end+1) = 9999999;\n",
    "            model.c(end+1) = 1;\n",
    "            model.rxnGeneMat(end+1,:) = zeros(1,length(model.genes));\n",
    "            model.rules{end+1} = '';\n",
    "            model.grRules{end+1} = '';\n",
    "            model.subSystems{end+1} = 'Energy dissipation reaction';\n",
    "            model.rxnNames{end+1} = 'Energy dissipation - CTP';\n",
    "            model.rxnKEGGID{end+1} = '';\n",
    "            model.rxnKeggOrthology{end+1} = '';\n",
    "            model.rxnConfidenceScores(end+1) = 0';\n",
    "            model.rxnReferences{end+1} = '';\n",
    "            model.rxnECNumbers{end+1} = '';\n",
    "            model.rxnNotes{end+1} = '';\n",
    "            model.rxnCOG{end+1} = '';\n",
    "            model.rxnReconMap{end+1} = '';\n",
    "\n",
    "            model.A = model.S;\n",
    "            model.rhs = model.b;\n",
    "            model.obj = model.c;\n",
    "            model.sense = repmat('=',1,length(model.mets));\n",
    "            model.vtype = repmat('C',1,length(model.rxns));\n",
    "            model.varnames = model.rxns;\n",
    "            model.modelsense = 'max';\n",
    "\n",
    "            params.outputflag = 0;\n",
    "            result = gurobi(model,params);\n",
    "            objectives(end+1) = result.objval;\n",
    "\n",
    "            model.c(end) = 0;\n",
    "            model.lb(end) = 0;\n",
    "            model.ub(end) = 0;\n",
    "\n",
    "            % gtp\n",
    "            model.rxns{end+1} = 'energy_dissipation_gtp';\n",
    "            model.S(:,end+1) = zeros(length(model.mets),1);\n",
    "            model.S(strcmp(model.mets,'gtp[c]'),end) = -1;\n",
    "            model.S(strcmp(model.mets,'h2o[c]'),end) = -1;\n",
    "            model.S(strcmp(model.mets,'gdp[c]'),end) = 1;\n",
    "            model.S(strcmp(model.mets,'h[c]'),end) = 1;\n",
    "            model.S(strcmp(model.mets,'pi[c]'),end) = 1;\n",
    "            model.lb(end+1) = 0;\n",
    "            model.ub(end+1) = 9999999;\n",
    "            model.c(end+1) = 1;\n",
    "            model.rxnGeneMat(end+1,:) = zeros(1,length(model.genes));\n",
    "            model.rules{end+1} = '';\n",
    "            model.grRules{end+1} = '';\n",
    "            model.subSystems{end+1} = 'Energy dissipation reaction';\n",
    "            model.rxnNames{end+1} = 'Energy dissipation - GTP';\n",
    "            model.rxnKEGGID{end+1} = '';\n",
    "            model.rxnKeggOrthology{end+1} = '';\n",
    "            model.rxnConfidenceScores(end+1) = 0';\n",
    "            model.rxnReferences{end+1} = '';\n",
    "            model.rxnECNumbers{end+1} = '';\n",
    "            model.rxnNotes{end+1} = '';\n",
    "            model.rxnCOG{end+1} = '';\n",
    "            model.rxnReconMap{end+1} = '';\n",
    "\n",
    "            model.A = model.S;\n",
    "            model.rhs = model.b;\n",
    "            model.obj = model.c;\n",
    "            model.sense = repmat('=',1,length(model.mets));\n",
    "            model.vtype = repmat('C',1,length(model.rxns));\n",
    "            model.varnames = model.rxns;\n",
    "            model.modelsense = 'max';\n",
    "\n",
    "            params.outputflag = 0;\n",
    "            result = gurobi(model,params);\n",
    "            objectives(end+1) = result.objval;\n",
    "\n",
    "            model.c(end) = 0;\n",
    "            model.lb(end) = 0;\n",
    "            model.ub(end) = 0;\n",
    "\n",
    "            % utp\n",
    "            model.rxns{end+1} = 'energy_dissipation_utp';\n",
    "            model.S(:,end+1) = zeros(length(model.mets),1);\n",
    "            model.S(strcmp(model.mets,'utp[c]'),end) = -1;\n",
    "            model.S(strcmp(model.mets,'h2o[c]'),end) = -1;\n",
    "            model.S(strcmp(model.mets,'udp[c]'),end) = 1;\n",
    "            model.S(strcmp(model.mets,'h[c]'),end) = 1;\n",
    "            model.S(strcmp(model.mets,'pi[c]'),end) = 1;\n",
    "            model.lb(end+1) = 0;\n",
    "            model.ub(end+1) = 9999999;\n",
    "            model.c(end+1) = 1;\n",
    "            model.rxnGeneMat(end+1,:) = zeros(1,length(model.genes));\n",
    "            model.rules{end+1} = '';\n",
    "            model.grRules{end+1} = '';\n",
    "            model.subSystems{end+1} = 'Energy dissipation reaction';\n",
    "            model.rxnNames{end+1} = 'Energy dissipation - UTP';\n",
    "            model.rxnKEGGID{end+1} = '';\n",
    "            model.rxnKeggOrthology{end+1} = '';\n",
    "            model.rxnConfidenceScores(end+1) = 0';\n",
    "            model.rxnReferences{end+1} = '';\n",
    "            model.rxnECNumbers{end+1} = '';\n",
    "            model.rxnNotes{end+1} = '';\n",
    "            model.rxnCOG{end+1} = '';\n",
    "            model.rxnReconMap{end+1} = '';\n",
    "\n",
    "            model.A = model.S;\n",
    "            model.rhs = model.b;\n",
    "            model.obj = model.c;\n",
    "            model.sense = repmat('=',1,length(model.mets));\n",
    "            model.vtype = repmat('C',1,length(model.rxns));\n",
    "            model.varnames = model.rxns;\n",
    "            model.modelsense = 'max';\n",
    "\n",
    "            params.outputflag = 0;\n",
    "            result = gurobi(model,params);\n",
    "            objectives(end+1) = result.objval;\n",
    "\n",
    "            model.c(end) = 0;\n",
    "            model.lb(end) = 0;\n",
    "            model.ub(end) = 0;\n",
    "\n",
    "            % itp\n",
    "            model.rxns{end+1} = 'energy_dissipation_itp';\n",
    "            model.S(:,end+1) = zeros(length(model.mets),1);\n",
    "            model.S(strcmp(model.mets,'itp[c]'),end) = -1;\n",
    "            model.S(strcmp(model.mets,'h2o[c]'),end) = -1;\n",
    "            model.S(strcmp(model.mets,'idp[c]'),end) = 1;\n",
    "            model.S(strcmp(model.mets,'h[c]'),end) = 1;\n",
    "            model.S(strcmp(model.mets,'pi[c]'),end) = 1;\n",
    "            model.lb(end+1) = 0;\n",
    "            model.ub(end+1) = 9999999;\n",
    "            model.c(end+1) = 1;\n",
    "            model.rxnGeneMat(end+1,:) = zeros(1,length(model.genes));\n",
    "            model.rules{end+1} = '';\n",
    "            model.grRules{end+1} = '';\n",
    "            model.subSystems{end+1} = 'Energy dissipation reaction';\n",
    "            model.rxnNames{end+1} = 'Energy dissipation - ITP';\n",
    "            model.rxnKEGGID{end+1} = '';\n",
    "            model.rxnKeggOrthology{end+1} = '';\n",
    "            model.rxnConfidenceScores(end+1) = 0';\n",
    "            model.rxnReferences{end+1} = '';\n",
    "            model.rxnECNumbers{end+1} = '';\n",
    "            model.rxnNotes{end+1} = '';\n",
    "            model.rxnCOG{end+1} = '';\n",
    "            model.rxnReconMap{end+1} = '';\n",
    "\n",
    "            model.A = model.S;\n",
    "            model.rhs = model.b;\n",
    "            model.obj = model.c;\n",
    "            model.sense = repmat('=',1,length(model.mets));\n",
    "            model.vtype = repmat('C',1,length(model.rxns));\n",
    "            model.varnames = model.rxns;\n",
    "            model.modelsense = 'max';\n",
    "\n",
    "            params.outputflag = 0;\n",
    "            result = gurobi(model,params);\n",
    "            objectives(end+1) = result.objval;\n",
    "\n",
    "            model.c(end) = 0;\n",
    "            model.lb(end) = 0;\n",
    "            model.ub(end) = 0;\n",
    "\n",
    "            % nadh\n",
    "            model.rxns{end+1} = 'energy_dissipation_nadh';\n",
    "            model.S(:,end+1) = zeros(length(model.mets),1);\n",
    "            model.S(strcmp(model.mets,'nadh[c]'),end) = -1;\n",
    "            model.S(strcmp(model.mets,'h[c]'),end) = 1;\n",
    "            model.S(strcmp(model.mets,'nad[c]'),end) = 1;\n",
    "            model.lb(end+1) = 0;\n",
    "            model.ub(end+1) = 9999999;\n",
    "            model.c(end+1) = 1;\n",
    "            model.rxnGeneMat(end+1,:) = zeros(1,length(model.genes));\n",
    "            model.rules{end+1} = '';\n",
    "            model.grRules{end+1} = '';\n",
    "            model.subSystems{end+1} = 'Energy dissipation reaction';\n",
    "            model.rxnNames{end+1} = 'Energy dissipation - NADH';\n",
    "            model.rxnKEGGID{end+1} = '';\n",
    "            model.rxnKeggOrthology{end+1} = '';\n",
    "            model.rxnConfidenceScores(end+1) = 0';\n",
    "            model.rxnReferences{end+1} = '';\n",
    "            model.rxnECNumbers{end+1} = '';\n",
    "            model.rxnNotes{end+1} = '';\n",
    "            model.rxnCOG{end+1} = '';\n",
    "            model.rxnReconMap{end+1} = '';\n",
    "\n",
    "            model.A = model.S;\n",
    "            model.rhs = model.b;\n",
    "            model.obj = model.c;\n",
    "            model.sense = repmat('=',1,length(model.mets));\n",
    "            model.vtype = repmat('C',1,length(model.rxns));\n",
    "            model.varnames = model.rxns;\n",
    "            model.modelsense = 'max';\n",
    "\n",
    "            params.outputflag = 0;\n",
    "            result = gurobi(model,params);\n",
    "            objectives(end+1) = result.objval;\n",
    "\n",
    "            model.c(end) = 0;\n",
    "            model.lb(end) = 0;\n",
    "            model.ub(end) = 0;\n",
    "\n",
    "            % nadph\n",
    "            model.rxns{end+1} = 'energy_dissipation_nadph';\n",
    "            model.S(:,end+1) = zeros(length(model.mets),1);\n",
    "            model.S(strcmp(model.mets,'nadph[c]'),end) = -1;\n",
    "            model.S(strcmp(model.mets,'h[c]'),end) = 1;\n",
    "            model.S(strcmp(model.mets,'nadp[c]'),end) = 1;\n",
    "            model.lb(end+1) = 0;\n",
    "            model.ub(end+1) = 9999999;\n",
    "            model.c(end+1) = 1;\n",
    "            model.rxnGeneMat(end+1,:) = zeros(1,length(model.genes));\n",
    "            model.rules{end+1} = '';\n",
    "            model.grRules{end+1} = '';\n",
    "            model.subSystems{end+1} = 'Energy dissipation reaction';\n",
    "            model.rxnNames{end+1} = 'Energy dissipation - NADPH';\n",
    "            model.rxnKEGGID{end+1} = '';\n",
    "            model.rxnKeggOrthology{end+1} = '';\n",
    "            model.rxnConfidenceScores(end+1) = 0';\n",
    "            model.rxnReferences{end+1} = '';\n",
    "            model.rxnECNumbers{end+1} = '';\n",
    "            model.rxnNotes{end+1} = '';\n",
    "            model.rxnCOG{end+1} = '';\n",
    "            model.rxnReconMap{end+1} = '';\n",
    "\n",
    "            model.A = model.S;\n",
    "            model.rhs = model.b;\n",
    "            model.obj = model.c;\n",
    "            model.sense = repmat('=',1,length(model.mets));\n",
    "            model.vtype = repmat('C',1,length(model.rxns));\n",
    "            model.varnames = model.rxns;\n",
    "            model.modelsense = 'max';\n",
    "\n",
    "            params.outputflag = 0;\n",
    "            result = gurobi(model,params);\n",
    "            objectives(end+1) = result.objval;\n",
    "\n",
    "            model.c(end) = 0;\n",
    "            model.lb(end) = 0;\n",
    "            model.ub(end) = 0;\n",
    "\n",
    "            % fadh2\n",
    "            model.rxns{end+1} = 'energy_dissipation_fadh2';\n",
    "            model.S(:,end+1) = zeros(length(model.mets),1);\n",
    "            model.S(strcmp(model.mets,'fadh2[c]'),end) = -1;\n",
    "            model.S(strcmp(model.mets,'h[c]'),end) = 2;\n",
    "            model.S(strcmp(model.mets,'fad[c]'),end) = 1;\n",
    "            model.lb(end+1) = 0;\n",
    "            model.ub(end+1) = 9999999;\n",
    "            model.c(end+1) = 1;\n",
    "            model.rxnGeneMat(end+1,:) = zeros(1,length(model.genes));\n",
    "            model.rules{end+1} = '';\n",
    "            model.grRules{end+1} = '';\n",
    "            model.subSystems{end+1} = 'Energy dissipation reaction';\n",
    "            model.rxnNames{end+1} = 'Energy dissipation - FADH2';\n",
    "            model.rxnKEGGID{end+1} = '';\n",
    "            model.rxnKeggOrthology{end+1} = '';\n",
    "            model.rxnConfidenceScores(end+1) = 0';\n",
    "            model.rxnReferences{end+1} = '';\n",
    "            model.rxnECNumbers{end+1} = '';\n",
    "            model.rxnNotes{end+1} = '';\n",
    "            model.rxnCOG{end+1} = '';\n",
    "            model.rxnReconMap{end+1} = '';\n",
    "\n",
    "            model.A = model.S;\n",
    "            model.rhs = model.b;\n",
    "            model.obj = model.c;\n",
    "            model.sense = repmat('=',1,length(model.mets));\n",
    "            model.vtype = repmat('C',1,length(model.rxns));\n",
    "            model.varnames = model.rxns;\n",
    "            model.modelsense = 'max';\n",
    "\n",
    "            params.outputflag = 0;\n",
    "            result = gurobi(model,params);\n",
    "            objectives(end+1) = result.objval;\n",
    "\n",
    "            model.c(end) = 0;\n",
    "            model.lb(end) = 0;\n",
    "            model.ub(end) = 0;\n",
    "\n",
    "            % q10h2\n",
    "            model.rxns{end+1} = 'energy_dissipation_q10h2';\n",
    "            model.S(:,end+1) = zeros(length(model.mets),1);\n",
    "            model.S(strcmp(model.mets,'q10h2[c]'),end) = -1;\n",
    "            model.S(strcmp(model.mets,'h[c]'),end) = 2;\n",
    "            model.S(strcmp(model.mets,'q10[c]'),end) = 1;\n",
    "            model.lb(end+1) = 0;\n",
    "            model.ub(end+1) = 9999999;\n",
    "            model.c(end+1) = 1;\n",
    "            model.rxnGeneMat(end+1,:) = zeros(1,length(model.genes));\n",
    "            model.rules{end+1} = '';\n",
    "            model.grRules{end+1} = '';\n",
    "            model.subSystems{end+1} = 'Energy dissipation reaction';\n",
    "            model.rxnNames{end+1} = 'Energy dissipation - Q10H2';\n",
    "            model.rxnKEGGID{end+1} = '';\n",
    "            model.rxnKeggOrthology{end+1} = '';\n",
    "            model.rxnConfidenceScores(end+1) = 0';\n",
    "            model.rxnReferences{end+1} = '';\n",
    "            model.rxnECNumbers{end+1} = '';\n",
    "            model.rxnNotes{end+1} = '';\n",
    "            model.rxnCOG{end+1} = '';\n",
    "            model.rxnReconMap{end+1} = '';\n",
    "\n",
    "            model.A = model.S;\n",
    "            model.rhs = model.b;\n",
    "            model.obj = model.c;\n",
    "            model.sense = repmat('=',1,length(model.mets));\n",
    "            model.vtype = repmat('C',1,length(model.rxns));\n",
    "            model.varnames = model.rxns;\n",
    "            model.modelsense = 'max';\n",
    "\n",
    "            params.outputflag = 0;\n",
    "            result = gurobi(model,params);\n",
    "            objectives(end+1) = result.objval;\n",
    "\n",
    "            model.c(end) = 0;\n",
    "            model.lb(end) = 0;\n",
    "            model.ub(end) = 0;\n",
    "\n",
    "            % accoa\n",
    "            model.rxns{end+1} = 'energy_dissipation_accoa';\n",
    "            model.S(:,end+1) = zeros(length(model.mets),1);\n",
    "            model.S(strcmp(model.mets,'h2o[c]'),end) = -1;\n",
    "            model.S(strcmp(model.mets,'accoa[c]'),end) = -1;\n",
    "            model.S(strcmp(model.mets,'h[c]'),end) = 1;\n",
    "            model.S(strcmp(model.mets,'ac[c]'),end) = 1;\n",
    "            model.S(strcmp(model.mets,'coa[c]'),end) = 1;\n",
    "            model.lb(end+1) = 0;\n",
    "            model.ub(end+1) = 9999999;\n",
    "            model.c(end+1) = 1;\n",
    "            model.rxnGeneMat(end+1,:) = zeros(1,length(model.genes));\n",
    "            model.rules{end+1} = '';\n",
    "            model.grRules{end+1} = '';\n",
    "            model.subSystems{end+1} = 'Energy dissipation reaction';\n",
    "            model.rxnNames{end+1} = 'Energy dissipation - ACCOA';\n",
    "            model.rxnKEGGID{end+1} = '';\n",
    "            model.rxnKeggOrthology{end+1} = '';\n",
    "            model.rxnConfidenceScores(end+1) = 0';\n",
    "            model.rxnReferences{end+1} = '';\n",
    "            model.rxnECNumbers{end+1} = '';\n",
    "            model.rxnNotes{end+1} = '';\n",
    "            model.rxnCOG{end+1} = '';\n",
    "            model.rxnReconMap{end+1} = '';\n",
    "\n",
    "            model.A = model.S;\n",
    "            model.rhs = model.b;\n",
    "            model.obj = model.c;\n",
    "            model.sense = repmat('=',1,length(model.mets));\n",
    "            model.vtype = repmat('C',1,length(model.rxns));\n",
    "            model.varnames = model.rxns;\n",
    "            model.modelsense = 'max';\n",
    "\n",
    "            params.outputflag = 0;\n",
    "            result = gurobi(model,params);\n",
    "            objectives(end+1) = result.objval;\n",
    "\n",
    "            model.c(end) = 0;\n",
    "            model.lb(end) = 0;\n",
    "            model.ub(end) = 0;\n",
    "\n",
    "            % glu\n",
    "            model.rxns{end+1} = 'energy_dissipation_glu';\n",
    "            model.S(:,end+1) = zeros(length(model.mets),1);\n",
    "            model.S(strcmp(model.mets,'glu_L[c]'),end) = -1;\n",
    "            model.S(strcmp(model.mets,'h2o[c]'),end) = -1;\n",
    "            model.S(strcmp(model.mets,'akg[c]'),end) = 1;\n",
    "            model.S(strcmp(model.mets,'nh4[c]'),end) = 1;\n",
    "            model.S(strcmp(model.mets,'h[c]'),end) = 2;\n",
    "            model.lb(end+1) = 0;\n",
    "            model.ub(end+1) = 9999999;\n",
    "            model.c(end+1) = 1;\n",
    "            model.rxnGeneMat(end+1,:) = zeros(1,length(model.genes));\n",
    "            model.rules{end+1} = '';\n",
    "            model.grRules{end+1} = '';\n",
    "            model.subSystems{end+1} = 'Energy dissipation reaction';\n",
    "            model.rxnNames{end+1} = 'Energy dissipation - GLU';\n",
    "            model.rxnKEGGID{end+1} = '';\n",
    "            model.rxnKeggOrthology{end+1} = '';\n",
    "            model.rxnConfidenceScores(end+1) = 0';\n",
    "            model.rxnReferences{end+1} = '';\n",
    "            model.rxnECNumbers{end+1} = '';\n",
    "            model.rxnNotes{end+1} = '';\n",
    "            model.rxnCOG{end+1} = '';\n",
    "            model.rxnReconMap{end+1} = '';\n",
    "\n",
    "            model.A = model.S;\n",
    "            model.rhs = model.b;\n",
    "            model.obj = model.c;\n",
    "            model.sense = repmat('=',1,length(model.mets));\n",
    "            model.vtype = repmat('C',1,length(model.rxns));\n",
    "            model.varnames = model.rxns;\n",
    "            model.modelsense = 'max';\n",
    "\n",
    "            params.outputflag = 0;\n",
    "            result = gurobi(model,params);\n",
    "            objectives(end+1) = result.objval;\n",
    "\n",
    "            model.c(end) = 0;\n",
    "            model.lb(end) = 0;\n",
    "            model.ub(end) = 0;\n",
    "\n",
    "            % proton\n",
    "            model.rxns{end+1} = 'energy_dissipation_proton';\n",
    "            model.S(:,end+1) = zeros(length(model.mets),1);\n",
    "            model.S(strcmp(model.mets,'h[i]'),end) = -1;\n",
    "            model.S(strcmp(model.mets,'h[c]'),end) = 1;\n",
    "            model.lb(end+1) = 0;\n",
    "            model.ub(end+1) = 9999999;\n",
    "            model.c(end+1) = 1;\n",
    "            model.rxnGeneMat(end+1,:) = zeros(1,length(model.genes));\n",
    "            model.rules{end+1} = '';\n",
    "            model.grRules{end+1} = '';\n",
    "            model.subSystems{end+1} = 'Energy dissipation reaction';\n",
    "            model.rxnNames{end+1} = 'Energy dissipation - PROTON';\n",
    "            model.rxnKEGGID{end+1} = '';\n",
    "            model.rxnKeggOrthology{end+1} = '';\n",
    "            model.rxnConfidenceScores(end+1) = 0';\n",
    "            model.rxnReferences{end+1} = '';\n",
    "            model.rxnECNumbers{end+1} = '';\n",
    "            model.rxnNotes{end+1} = '';\n",
    "            model.rxnCOG{end+1} = '';\n",
    "            model.rxnReconMap{end+1} = '';\n",
    "\n",
    "            model.A = model.S;\n",
    "            model.rhs = model.b;\n",
    "            model.obj = model.c;\n",
    "            model.sense = repmat('=',1,length(model.mets));\n",
    "            model.vtype = repmat('C',1,length(model.rxns));\n",
    "            model.varnames = model.rxns;\n",
    "            model.modelsense = 'max';\n",
    "\n",
    "            params.outputflag = 0;\n",
    "            result = gurobi(model,params);\n",
    "            objectives(end+1) = result.objval;\n",
    "\n",
    "            model.c(end) = 0;\n",
    "            model.lb(end) = 0;\n",
    "            model.ub(end) = 0;\n",
    "            \n",
    "            % keep only if all objective values are still 0\n",
    "            if sum(objectives == 0) ~= length(objectives)\n",
    "                model_keep.ub(i) = 0;\n",
    "            end\n",
    "            model = model_keep;\n",
    "            \n",
    "        end\n",
    "\n",
    "        % reverse direction\n",
    "        if (model.lb(i) == 0) && (Recon3D.lb(i) < 0)\n",
    "            \n",
    "            % set new upper bound\n",
    "            model.lb(i) = -1;\n",
    "            model_keep = model;\n",
    "            \n",
    "            % test dissipation reactions\n",
    "            objectives = [];\n",
    "\n",
    "            % atp\n",
    "            model.rxns{end+1} = 'energy_dissipation_atp';\n",
    "            model.S(:,end+1) = zeros(length(model.mets),1);\n",
    "            model.S(strcmp(model.mets,'atp[c]'),end) = -1;\n",
    "            model.S(strcmp(model.mets,'h2o[c]'),end) = -1;\n",
    "            model.S(strcmp(model.mets,'adp[c]'),end) = 1;\n",
    "            model.S(strcmp(model.mets,'h[c]'),end) = 1;\n",
    "            model.S(strcmp(model.mets,'pi[c]'),end) = 1;\n",
    "            model.lb(end+1) = 0;\n",
    "            model.ub(end+1) = 9999999;\n",
    "            model.c(end+1) = 1;\n",
    "            model.rxnGeneMat(end+1,:) = zeros(1,length(model.genes));\n",
    "            model.rules{end+1} = '';\n",
    "            model.grRules{end+1} = '';\n",
    "            model.subSystems{end+1} = 'Energy dissipation reaction';\n",
    "            model.rxnNames{end+1} = 'Energy dissipation - ATP';\n",
    "            model.rxnKEGGID{end+1} = '';\n",
    "            model.rxnKeggOrthology{end+1} = '';\n",
    "            model.rxnConfidenceScores(end+1) = 0';\n",
    "            model.rxnReferences{end+1} = '';\n",
    "            model.rxnECNumbers{end+1} = '';\n",
    "            model.rxnNotes{end+1} = '';\n",
    "            model.rxnCOG{end+1} = '';\n",
    "            model.rxnReconMap{end+1} = '';\n",
    "\n",
    "            model.A = model.S;\n",
    "            model.rhs = model.b;\n",
    "            model.obj = model.c;\n",
    "            model.sense = repmat('=',1,length(model.mets));\n",
    "            model.vtype = repmat('C',1,length(model.rxns));\n",
    "            model.varnames = model.rxns;\n",
    "            model.modelsense = 'max';\n",
    "\n",
    "            params.outputflag = 0;\n",
    "            result = gurobi(model,params);\n",
    "            objectives(end+1) = result.objval;\n",
    "\n",
    "            model.c(end) = 0;\n",
    "            model.lb(end) = 0;\n",
    "            model.ub(end) = 0;\n",
    "\n",
    "            % ctp\n",
    "            model.rxns{end+1} = 'energy_dissipation_ctp';\n",
    "            model.S(:,end+1) = zeros(length(model.mets),1);\n",
    "            model.S(strcmp(model.mets,'ctp[c]'),end) = -1;\n",
    "            model.S(strcmp(model.mets,'h2o[c]'),end) = -1;\n",
    "            model.S(strcmp(model.mets,'cdp[c]'),end) = 1;\n",
    "            model.S(strcmp(model.mets,'h[c]'),end) = 1;\n",
    "            model.S(strcmp(model.mets,'pi[c]'),end) = 1;\n",
    "            model.lb(end+1) = 0;\n",
    "            model.ub(end+1) = 9999999;\n",
    "            model.c(end+1) = 1;\n",
    "            model.rxnGeneMat(end+1,:) = zeros(1,length(model.genes));\n",
    "            model.rules{end+1} = '';\n",
    "            model.grRules{end+1} = '';\n",
    "            model.subSystems{end+1} = 'Energy dissipation reaction';\n",
    "            model.rxnNames{end+1} = 'Energy dissipation - CTP';\n",
    "            model.rxnKEGGID{end+1} = '';\n",
    "            model.rxnKeggOrthology{end+1} = '';\n",
    "            model.rxnConfidenceScores(end+1) = 0';\n",
    "            model.rxnReferences{end+1} = '';\n",
    "            model.rxnECNumbers{end+1} = '';\n",
    "            model.rxnNotes{end+1} = '';\n",
    "            model.rxnCOG{end+1} = '';\n",
    "            model.rxnReconMap{end+1} = '';\n",
    "\n",
    "            model.A = model.S;\n",
    "            model.rhs = model.b;\n",
    "            model.obj = model.c;\n",
    "            model.sense = repmat('=',1,length(model.mets));\n",
    "            model.vtype = repmat('C',1,length(model.rxns));\n",
    "            model.varnames = model.rxns;\n",
    "            model.modelsense = 'max';\n",
    "\n",
    "            params.outputflag = 0;\n",
    "            result = gurobi(model,params);\n",
    "            objectives(end+1) = result.objval;\n",
    "\n",
    "            model.c(end) = 0;\n",
    "            model.lb(end) = 0;\n",
    "            model.ub(end) = 0;\n",
    "\n",
    "            % gtp\n",
    "            model.rxns{end+1} = 'energy_dissipation_gtp';\n",
    "            model.S(:,end+1) = zeros(length(model.mets),1);\n",
    "            model.S(strcmp(model.mets,'gtp[c]'),end) = -1;\n",
    "            model.S(strcmp(model.mets,'h2o[c]'),end) = -1;\n",
    "            model.S(strcmp(model.mets,'gdp[c]'),end) = 1;\n",
    "            model.S(strcmp(model.mets,'h[c]'),end) = 1;\n",
    "            model.S(strcmp(model.mets,'pi[c]'),end) = 1;\n",
    "            model.lb(end+1) = 0;\n",
    "            model.ub(end+1) = 9999999;\n",
    "            model.c(end+1) = 1;\n",
    "            model.rxnGeneMat(end+1,:) = zeros(1,length(model.genes));\n",
    "            model.rules{end+1} = '';\n",
    "            model.grRules{end+1} = '';\n",
    "            model.subSystems{end+1} = 'Energy dissipation reaction';\n",
    "            model.rxnNames{end+1} = 'Energy dissipation - GTP';\n",
    "            model.rxnKEGGID{end+1} = '';\n",
    "            model.rxnKeggOrthology{end+1} = '';\n",
    "            model.rxnConfidenceScores(end+1) = 0';\n",
    "            model.rxnReferences{end+1} = '';\n",
    "            model.rxnECNumbers{end+1} = '';\n",
    "            model.rxnNotes{end+1} = '';\n",
    "            model.rxnCOG{end+1} = '';\n",
    "            model.rxnReconMap{end+1} = '';\n",
    "\n",
    "            model.A = model.S;\n",
    "            model.rhs = model.b;\n",
    "            model.obj = model.c;\n",
    "            model.sense = repmat('=',1,length(model.mets));\n",
    "            model.vtype = repmat('C',1,length(model.rxns));\n",
    "            model.varnames = model.rxns;\n",
    "            model.modelsense = 'max';\n",
    "\n",
    "            params.outputflag = 0;\n",
    "            result = gurobi(model,params);\n",
    "            objectives(end+1) = result.objval;\n",
    "\n",
    "            model.c(end) = 0;\n",
    "            model.lb(end) = 0;\n",
    "            model.ub(end) = 0;\n",
    "\n",
    "            % utp\n",
    "            model.rxns{end+1} = 'energy_dissipation_utp';\n",
    "            model.S(:,end+1) = zeros(length(model.mets),1);\n",
    "            model.S(strcmp(model.mets,'utp[c]'),end) = -1;\n",
    "            model.S(strcmp(model.mets,'h2o[c]'),end) = -1;\n",
    "            model.S(strcmp(model.mets,'udp[c]'),end) = 1;\n",
    "            model.S(strcmp(model.mets,'h[c]'),end) = 1;\n",
    "            model.S(strcmp(model.mets,'pi[c]'),end) = 1;\n",
    "            model.lb(end+1) = 0;\n",
    "            model.ub(end+1) = 9999999;\n",
    "            model.c(end+1) = 1;\n",
    "            model.rxnGeneMat(end+1,:) = zeros(1,length(model.genes));\n",
    "            model.rules{end+1} = '';\n",
    "            model.grRules{end+1} = '';\n",
    "            model.subSystems{end+1} = 'Energy dissipation reaction';\n",
    "            model.rxnNames{end+1} = 'Energy dissipation - UTP';\n",
    "            model.rxnKEGGID{end+1} = '';\n",
    "            model.rxnKeggOrthology{end+1} = '';\n",
    "            model.rxnConfidenceScores(end+1) = 0';\n",
    "            model.rxnReferences{end+1} = '';\n",
    "            model.rxnECNumbers{end+1} = '';\n",
    "            model.rxnNotes{end+1} = '';\n",
    "            model.rxnCOG{end+1} = '';\n",
    "            model.rxnReconMap{end+1} = '';\n",
    "\n",
    "            model.A = model.S;\n",
    "            model.rhs = model.b;\n",
    "            model.obj = model.c;\n",
    "            model.sense = repmat('=',1,length(model.mets));\n",
    "            model.vtype = repmat('C',1,length(model.rxns));\n",
    "            model.varnames = model.rxns;\n",
    "            model.modelsense = 'max';\n",
    "\n",
    "            params.outputflag = 0;\n",
    "            result = gurobi(model,params);\n",
    "            objectives(end+1) = result.objval;\n",
    "\n",
    "            model.c(end) = 0;\n",
    "            model.lb(end) = 0;\n",
    "            model.ub(end) = 0;\n",
    "\n",
    "            % itp\n",
    "            model.rxns{end+1} = 'energy_dissipation_itp';\n",
    "            model.S(:,end+1) = zeros(length(model.mets),1);\n",
    "            model.S(strcmp(model.mets,'itp[c]'),end) = -1;\n",
    "            model.S(strcmp(model.mets,'h2o[c]'),end) = -1;\n",
    "            model.S(strcmp(model.mets,'idp[c]'),end) = 1;\n",
    "            model.S(strcmp(model.mets,'h[c]'),end) = 1;\n",
    "            model.S(strcmp(model.mets,'pi[c]'),end) = 1;\n",
    "            model.lb(end+1) = 0;\n",
    "            model.ub(end+1) = 9999999;\n",
    "            model.c(end+1) = 1;\n",
    "            model.rxnGeneMat(end+1,:) = zeros(1,length(model.genes));\n",
    "            model.rules{end+1} = '';\n",
    "            model.grRules{end+1} = '';\n",
    "            model.subSystems{end+1} = 'Energy dissipation reaction';\n",
    "            model.rxnNames{end+1} = 'Energy dissipation - ITP';\n",
    "            model.rxnKEGGID{end+1} = '';\n",
    "            model.rxnKeggOrthology{end+1} = '';\n",
    "            model.rxnConfidenceScores(end+1) = 0';\n",
    "            model.rxnReferences{end+1} = '';\n",
    "            model.rxnECNumbers{end+1} = '';\n",
    "            model.rxnNotes{end+1} = '';\n",
    "            model.rxnCOG{end+1} = '';\n",
    "            model.rxnReconMap{end+1} = '';\n",
    "\n",
    "            model.A = model.S;\n",
    "            model.rhs = model.b;\n",
    "            model.obj = model.c;\n",
    "            model.sense = repmat('=',1,length(model.mets));\n",
    "            model.vtype = repmat('C',1,length(model.rxns));\n",
    "            model.varnames = model.rxns;\n",
    "            model.modelsense = 'max';\n",
    "\n",
    "            params.outputflag = 0;\n",
    "            result = gurobi(model,params);\n",
    "            objectives(end+1) = result.objval;\n",
    "\n",
    "            model.c(end) = 0;\n",
    "            model.lb(end) = 0;\n",
    "            model.ub(end) = 0;\n",
    "\n",
    "            % nadh\n",
    "            model.rxns{end+1} = 'energy_dissipation_nadh';\n",
    "            model.S(:,end+1) = zeros(length(model.mets),1);\n",
    "            model.S(strcmp(model.mets,'nadh[c]'),end) = -1;\n",
    "            model.S(strcmp(model.mets,'h[c]'),end) = 1;\n",
    "            model.S(strcmp(model.mets,'nad[c]'),end) = 1;\n",
    "            model.lb(end+1) = 0;\n",
    "            model.ub(end+1) = 9999999;\n",
    "            model.c(end+1) = 1;\n",
    "            model.rxnGeneMat(end+1,:) = zeros(1,length(model.genes));\n",
    "            model.rules{end+1} = '';\n",
    "            model.grRules{end+1} = '';\n",
    "            model.subSystems{end+1} = 'Energy dissipation reaction';\n",
    "            model.rxnNames{end+1} = 'Energy dissipation - NADH';\n",
    "            model.rxnKEGGID{end+1} = '';\n",
    "            model.rxnKeggOrthology{end+1} = '';\n",
    "            model.rxnConfidenceScores(end+1) = 0';\n",
    "            model.rxnReferences{end+1} = '';\n",
    "            model.rxnECNumbers{end+1} = '';\n",
    "            model.rxnNotes{end+1} = '';\n",
    "            model.rxnCOG{end+1} = '';\n",
    "            model.rxnReconMap{end+1} = '';\n",
    "\n",
    "            model.A = model.S;\n",
    "            model.rhs = model.b;\n",
    "            model.obj = model.c;\n",
    "            model.sense = repmat('=',1,length(model.mets));\n",
    "            model.vtype = repmat('C',1,length(model.rxns));\n",
    "            model.varnames = model.rxns;\n",
    "            model.modelsense = 'max';\n",
    "\n",
    "            params.outputflag = 0;\n",
    "            result = gurobi(model,params);\n",
    "            objectives(end+1) = result.objval;\n",
    "\n",
    "            model.c(end) = 0;\n",
    "            model.lb(end) = 0;\n",
    "            model.ub(end) = 0;\n",
    "\n",
    "            % nadph\n",
    "            model.rxns{end+1} = 'energy_dissipation_nadph';\n",
    "            model.S(:,end+1) = zeros(length(model.mets),1);\n",
    "            model.S(strcmp(model.mets,'nadph[c]'),end) = -1;\n",
    "            model.S(strcmp(model.mets,'h[c]'),end) = 1;\n",
    "            model.S(strcmp(model.mets,'nadp[c]'),end) = 1;\n",
    "            model.lb(end+1) = 0;\n",
    "            model.ub(end+1) = 9999999;\n",
    "            model.c(end+1) = 1;\n",
    "            model.rxnGeneMat(end+1,:) = zeros(1,length(model.genes));\n",
    "            model.rules{end+1} = '';\n",
    "            model.grRules{end+1} = '';\n",
    "            model.subSystems{end+1} = 'Energy dissipation reaction';\n",
    "            model.rxnNames{end+1} = 'Energy dissipation - NADPH';\n",
    "            model.rxnKEGGID{end+1} = '';\n",
    "            model.rxnKeggOrthology{end+1} = '';\n",
    "            model.rxnConfidenceScores(end+1) = 0';\n",
    "            model.rxnReferences{end+1} = '';\n",
    "            model.rxnECNumbers{end+1} = '';\n",
    "            model.rxnNotes{end+1} = '';\n",
    "            model.rxnCOG{end+1} = '';\n",
    "            model.rxnReconMap{end+1} = '';\n",
    "\n",
    "            model.A = model.S;\n",
    "            model.rhs = model.b;\n",
    "            model.obj = model.c;\n",
    "            model.sense = repmat('=',1,length(model.mets));\n",
    "            model.vtype = repmat('C',1,length(model.rxns));\n",
    "            model.varnames = model.rxns;\n",
    "            model.modelsense = 'max';\n",
    "\n",
    "            params.outputflag = 0;\n",
    "            result = gurobi(model,params);\n",
    "            objectives(end+1) = result.objval;\n",
    "\n",
    "            model.c(end) = 0;\n",
    "            model.lb(end) = 0;\n",
    "            model.ub(end) = 0;\n",
    "\n",
    "            % fadh2\n",
    "            model.rxns{end+1} = 'energy_dissipation_fadh2';\n",
    "            model.S(:,end+1) = zeros(length(model.mets),1);\n",
    "            model.S(strcmp(model.mets,'fadh2[c]'),end) = -1;\n",
    "            model.S(strcmp(model.mets,'h[c]'),end) = 2;\n",
    "            model.S(strcmp(model.mets,'fad[c]'),end) = 1;\n",
    "            model.lb(end+1) = 0;\n",
    "            model.ub(end+1) = 9999999;\n",
    "            model.c(end+1) = 1;\n",
    "            model.rxnGeneMat(end+1,:) = zeros(1,length(model.genes));\n",
    "            model.rules{end+1} = '';\n",
    "            model.grRules{end+1} = '';\n",
    "            model.subSystems{end+1} = 'Energy dissipation reaction';\n",
    "            model.rxnNames{end+1} = 'Energy dissipation - FADH2';\n",
    "            model.rxnKEGGID{end+1} = '';\n",
    "            model.rxnKeggOrthology{end+1} = '';\n",
    "            model.rxnConfidenceScores(end+1) = 0';\n",
    "            model.rxnReferences{end+1} = '';\n",
    "            model.rxnECNumbers{end+1} = '';\n",
    "            model.rxnNotes{end+1} = '';\n",
    "            model.rxnCOG{end+1} = '';\n",
    "            model.rxnReconMap{end+1} = '';\n",
    "\n",
    "            model.A = model.S;\n",
    "            model.rhs = model.b;\n",
    "            model.obj = model.c;\n",
    "            model.sense = repmat('=',1,length(model.mets));\n",
    "            model.vtype = repmat('C',1,length(model.rxns));\n",
    "            model.varnames = model.rxns;\n",
    "            model.modelsense = 'max';\n",
    "\n",
    "            params.outputflag = 0;\n",
    "            result = gurobi(model,params);\n",
    "            objectives(end+1) = result.objval;\n",
    "\n",
    "            model.c(end) = 0;\n",
    "            model.lb(end) = 0;\n",
    "            model.ub(end) = 0;\n",
    "\n",
    "            % q10h2\n",
    "            model.rxns{end+1} = 'energy_dissipation_q10h2';\n",
    "            model.S(:,end+1) = zeros(length(model.mets),1);\n",
    "            model.S(strcmp(model.mets,'q10h2[c]'),end) = -1;\n",
    "            model.S(strcmp(model.mets,'h[c]'),end) = 2;\n",
    "            model.S(strcmp(model.mets,'q10[c]'),end) = 1;\n",
    "            model.lb(end+1) = 0;\n",
    "            model.ub(end+1) = 9999999;\n",
    "            model.c(end+1) = 1;\n",
    "            model.rxnGeneMat(end+1,:) = zeros(1,length(model.genes));\n",
    "            model.rules{end+1} = '';\n",
    "            model.grRules{end+1} = '';\n",
    "            model.subSystems{end+1} = 'Energy dissipation reaction';\n",
    "            model.rxnNames{end+1} = 'Energy dissipation - Q10H2';\n",
    "            model.rxnKEGGID{end+1} = '';\n",
    "            model.rxnKeggOrthology{end+1} = '';\n",
    "            model.rxnConfidenceScores(end+1) = 0';\n",
    "            model.rxnReferences{end+1} = '';\n",
    "            model.rxnECNumbers{end+1} = '';\n",
    "            model.rxnNotes{end+1} = '';\n",
    "            model.rxnCOG{end+1} = '';\n",
    "            model.rxnReconMap{end+1} = '';\n",
    "\n",
    "            model.A = model.S;\n",
    "            model.rhs = model.b;\n",
    "            model.obj = model.c;\n",
    "            model.sense = repmat('=',1,length(model.mets));\n",
    "            model.vtype = repmat('C',1,length(model.rxns));\n",
    "            model.varnames = model.rxns;\n",
    "            model.modelsense = 'max';\n",
    "\n",
    "            params.outputflag = 0;\n",
    "            result = gurobi(model,params);\n",
    "            objectives(end+1) = result.objval;\n",
    "\n",
    "            model.c(end) = 0;\n",
    "            model.lb(end) = 0;\n",
    "            model.ub(end) = 0;\n",
    "\n",
    "            % accoa\n",
    "            model.rxns{end+1} = 'energy_dissipation_accoa';\n",
    "            model.S(:,end+1) = zeros(length(model.mets),1);\n",
    "            model.S(strcmp(model.mets,'h2o[c]'),end) = -1;\n",
    "            model.S(strcmp(model.mets,'accoa[c]'),end) = -1;\n",
    "            model.S(strcmp(model.mets,'h[c]'),end) = 1;\n",
    "            model.S(strcmp(model.mets,'ac[c]'),end) = 1;\n",
    "            model.S(strcmp(model.mets,'coa[c]'),end) = 1;\n",
    "            model.lb(end+1) = 0;\n",
    "            model.ub(end+1) = 9999999;\n",
    "            model.c(end+1) = 1;\n",
    "            model.rxnGeneMat(end+1,:) = zeros(1,length(model.genes));\n",
    "            model.rules{end+1} = '';\n",
    "            model.grRules{end+1} = '';\n",
    "            model.subSystems{end+1} = 'Energy dissipation reaction';\n",
    "            model.rxnNames{end+1} = 'Energy dissipation - ACCOA';\n",
    "            model.rxnKEGGID{end+1} = '';\n",
    "            model.rxnKeggOrthology{end+1} = '';\n",
    "            model.rxnConfidenceScores(end+1) = 0';\n",
    "            model.rxnReferences{end+1} = '';\n",
    "            model.rxnECNumbers{end+1} = '';\n",
    "            model.rxnNotes{end+1} = '';\n",
    "            model.rxnCOG{end+1} = '';\n",
    "            model.rxnReconMap{end+1} = '';\n",
    "\n",
    "            model.A = model.S;\n",
    "            model.rhs = model.b;\n",
    "            model.obj = model.c;\n",
    "            model.sense = repmat('=',1,length(model.mets));\n",
    "            model.vtype = repmat('C',1,length(model.rxns));\n",
    "            model.varnames = model.rxns;\n",
    "            model.modelsense = 'max';\n",
    "\n",
    "            params.outputflag = 0;\n",
    "            result = gurobi(model,params);\n",
    "            objectives(end+1) = result.objval;\n",
    "\n",
    "            model.c(end) = 0;\n",
    "            model.lb(end) = 0;\n",
    "            model.ub(end) = 0;\n",
    "\n",
    "            % glu\n",
    "            model.rxns{end+1} = 'energy_dissipation_glu';\n",
    "            model.S(:,end+1) = zeros(length(model.mets),1);\n",
    "            model.S(strcmp(model.mets,'glu_L[c]'),end) = -1;\n",
    "            model.S(strcmp(model.mets,'h2o[c]'),end) = -1;\n",
    "            model.S(strcmp(model.mets,'akg[c]'),end) = 1;\n",
    "            model.S(strcmp(model.mets,'nh4[c]'),end) = 1;\n",
    "            model.S(strcmp(model.mets,'h[c]'),end) = 2;\n",
    "            model.lb(end+1) = 0;\n",
    "            model.ub(end+1) = 9999999;\n",
    "            model.c(end+1) = 1;\n",
    "            model.rxnGeneMat(end+1,:) = zeros(1,length(model.genes));\n",
    "            model.rules{end+1} = '';\n",
    "            model.grRules{end+1} = '';\n",
    "            model.subSystems{end+1} = 'Energy dissipation reaction';\n",
    "            model.rxnNames{end+1} = 'Energy dissipation - GLU';\n",
    "            model.rxnKEGGID{end+1} = '';\n",
    "            model.rxnKeggOrthology{end+1} = '';\n",
    "            model.rxnConfidenceScores(end+1) = 0';\n",
    "            model.rxnReferences{end+1} = '';\n",
    "            model.rxnECNumbers{end+1} = '';\n",
    "            model.rxnNotes{end+1} = '';\n",
    "            model.rxnCOG{end+1} = '';\n",
    "            model.rxnReconMap{end+1} = '';\n",
    "\n",
    "            model.A = model.S;\n",
    "            model.rhs = model.b;\n",
    "            model.obj = model.c;\n",
    "            model.sense = repmat('=',1,length(model.mets));\n",
    "            model.vtype = repmat('C',1,length(model.rxns));\n",
    "            model.varnames = model.rxns;\n",
    "            model.modelsense = 'max';\n",
    "\n",
    "            params.outputflag = 0;\n",
    "            result = gurobi(model,params);\n",
    "            objectives(end+1) = result.objval;\n",
    "\n",
    "            model.c(end) = 0;\n",
    "            model.lb(end) = 0;\n",
    "            model.ub(end) = 0;\n",
    "\n",
    "            % proton\n",
    "            model.rxns{end+1} = 'energy_dissipation_proton';\n",
    "            model.S(:,end+1) = zeros(length(model.mets),1);\n",
    "            model.S(strcmp(model.mets,'h[i]'),end) = -1;\n",
    "            model.S(strcmp(model.mets,'h[c]'),end) = 1;\n",
    "            model.lb(end+1) = 0;\n",
    "            model.ub(end+1) = 9999999;\n",
    "            model.c(end+1) = 1;\n",
    "            model.rxnGeneMat(end+1,:) = zeros(1,length(model.genes));\n",
    "            model.rules{end+1} = '';\n",
    "            model.grRules{end+1} = '';\n",
    "            model.subSystems{end+1} = 'Energy dissipation reaction';\n",
    "            model.rxnNames{end+1} = 'Energy dissipation - PROTON';\n",
    "            model.rxnKEGGID{end+1} = '';\n",
    "            model.rxnKeggOrthology{end+1} = '';\n",
    "            model.rxnConfidenceScores(end+1) = 0';\n",
    "            model.rxnReferences{end+1} = '';\n",
    "            model.rxnECNumbers{end+1} = '';\n",
    "            model.rxnNotes{end+1} = '';\n",
    "            model.rxnCOG{end+1} = '';\n",
    "            model.rxnReconMap{end+1} = '';\n",
    "\n",
    "            model.A = model.S;\n",
    "            model.rhs = model.b;\n",
    "            model.obj = model.c;\n",
    "            model.sense = repmat('=',1,length(model.mets));\n",
    "            model.vtype = repmat('C',1,length(model.rxns));\n",
    "            model.varnames = model.rxns;\n",
    "            model.modelsense = 'max';\n",
    "\n",
    "            params.outputflag = 0;\n",
    "            result = gurobi(model,params);\n",
    "            objectives(end+1) = result.objval;\n",
    "\n",
    "            model.c(end) = 0;\n",
    "            model.lb(end) = 0;\n",
    "            model.ub(end) = 0;\n",
    "            \n",
    "            % keep only if all objective values are still 0\n",
    "            if sum(objectives == 0) ~= length(objectives)\n",
    "                model_keep.lb(i) = 0;\n",
    "            end\n",
    "            model = model_keep;\n",
    "\n",
    "        end\n",
    "    end\n",
    "end"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Reset flux ranges"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "% reset flux ranges\n",
    "for i = 1:length(model.rxns)\n",
    "    if strcmp(model.rxns{i},'biomass_reaction')\n",
    "        model.lb(i) = 0;\n",
    "        model.ub(i) = 9999999;\n",
    "    elseif length(model.rxns{i}) >= 3\n",
    "        if strcmp(model.rxns{i}(1:3),'EX_')\n",
    "            model.lb(i) = -9999999;\n",
    "            model.ub(i) = 9999999;\n",
    "        else\n",
    "            if model.lb(i) < 0\n",
    "                model.lb(i) = -9999999;\n",
    "            end\n",
    "            if model.ub(i) > 0\n",
    "                model.ub(i) = 9999999;\n",
    "            end\n",
    "        end\n",
    "    else\n",
    "        if model.lb(i) < 0\n",
    "            model.lb(i) = -9999999;\n",
    "        end\n",
    "        if model.ub(i) > 0\n",
    "            model.ub(i) = 9999999;\n",
    "        end\n",
    "    end\n",
    "end"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Add/change redox-related reaction information in Recon3D"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Split reaction FTHFDH into two separate cytosolic and mitochondrial reactions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "% fix cytosolic reaction\n",
    "i = find(strcmp(model.rxns,'FTHFDH'));\n",
    "model.rxnGeneMat(i,strcmp(model.genes,'160428.1')) = 0;\n",
    "model.rules{i} = '(x(615))';\n",
    "model.grRules{i} = '(10840.1)';\n",
    "\n",
    "% create new mitochondrial reaction\n",
    "model.rxns{end+1} = 'FTHFDHm';\n",
    "model.S(:,end+1) = zeros(length(model.mets),1);\n",
    "model.S(strcmp(model.mets,'10fthf[m]'),end) = -1;\n",
    "model.S(strcmp(model.mets,'h2o[m]'),end) = -1;\n",
    "model.S(strcmp(model.mets,'nadp[m]'),end) = -1;\n",
    "model.S(strcmp(model.mets,'h[m]'),end) = 1;\n",
    "model.S(strcmp(model.mets,'nadph[m]'),end) = 1;\n",
    "model.S(strcmp(model.mets,'co2[m]'),end) = 1;\n",
    "model.S(strcmp(model.mets,'thf[m]'),end) = 1;\n",
    "model.lb(end+1) = model.lb(i);\n",
    "model.ub(end+1) = model.ub(i);\n",
    "model.c(end+1) = model.c(i);\n",
    "model.rxnGeneMat(end+1,:) = zeros(1,length(model.genes));\n",
    "model.rxnGeneMat(end,strcmp(model.genes,'160428.1')) = 1;\n",
    "model.rules{end+1} = '(x(626))';\n",
    "model.grRules{end+1} = '(160428.1)';\n",
    "model.subSystems{end+1} = model.subSystems{i};\n",
    "model.rxnNames{end+1} = model.rxnNames{i};\n",
    "model.rxnKEGGID{end+1} = model.rxnKEGGID{i};\n",
    "model.rxnKeggOrthology{end+1} = model.rxnKeggOrthology{i};\n",
    "model.rxnConfidenceScores(end+1) = model.rxnConfidenceScores(i);\n",
    "model.rxnReferences{end+1} = model.rxnReferences{i};\n",
    "model.rxnECNumbers{end+1} = model.rxnECNumbers{i};\n",
    "model.rxnNotes{end+1} = model.rxnNotes{i};\n",
    "model.rxnCOG{end+1} = model.rxnCOG{i};\n",
    "model.rxnReconMap{end+1} = model.rxnReconMap{i};"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Remove gene MTHFR from reaction MTHFD"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "i = find(strcmp(model.rxns,'MTHFD'));\n",
    "model.rxnGeneMat(i,strcmp(model.genes,'4524.1')) = 0;\n",
    "model.rules{i} = '(x(628))';\n",
    "model.grRules{i} = '(4522.1)';"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Add reactions catalyzed by NOX isoforms"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "% NOX1/2/3/5, transfer from cytoplasm to extracellular\n",
    "model.rxns{end+1} = 'NOXce';\n",
    "model.S(:,end+1) = zeros(length(model.mets),1);\n",
    "model.S(strcmp(model.mets,'nadph[c]'),end) = -1;\n",
    "model.S(strcmp(model.mets,'o2[c]'),end) = -2;\n",
    "model.S(strcmp(model.mets,'o2s[e]'),end) = 2;\n",
    "model.S(strcmp(model.mets,'nadp[c]'),end) = 1;\n",
    "model.S(strcmp(model.mets,'h[c]'),end) = 1;\n",
    "model.lb(end+1) = 0;\n",
    "model.ub(end+1) = max(model.ub);\n",
    "model.c(end+1) = 0;\n",
    "model.rxnGeneMat(end+1,:) = zeros(1,length(model.genes));\n",
    "model.rxnGeneMat(:,(end+1):(end+4)) = zeros(length(model.rxns),4);\n",
    "model.rxnGeneMat(end,(end-3):end) = 1;\n",
    "model.rules{end+1} = '(x(2249)) | (x(2250)) | (x(2251)) | (x(2252))';\n",
    "model.genes{end+1} = '27035.1';\n",
    "model.genes{end+1} = '1536.1';\n",
    "model.genes{end+1} = '50508.1';\n",
    "model.genes{end+1} = '79400.1';\n",
    "model.grRules{end+1} = '(27035.1) or (1536.1) or (50508.1) or (79400.1)';\n",
    "model.subSystems{end+1} = 'ROS detoxification';\n",
    "model.rxnNames{end+1} = 'NADPH oxidase, cytoplasm to extracellular';\n",
    "model.rxnKEGGID{end+1} = '';\n",
    "model.rxnKeggOrthology{end+1} = '';\n",
    "model.rxnConfidenceScores(end+1) = 0;\n",
    "model.rxnReferences{end+1} = '';\n",
    "model.rxnECNumbers{end+1} = '1.6.3.1';\n",
    "model.rxnNotes{end+1} = '';\n",
    "model.rxnCOG{end+1} = '';\n",
    "model.rxnReconMap{end+1} = '';\n",
    "\n",
    "% NOX4, transfer from nucleus to cytoplasm\n",
    "model.mets{end+1} = 'nadph[n]';\n",
    "model.S(end+1,:) = zeros(1,length(model.rxns));\n",
    "model.metNames{end+1} = model.metNames{strcmp(model.mets,'nadph[c]')};\n",
    "model.metFormulas{end+1} = model.metFormulas{strcmp(model.mets,'nadph[c]')};\n",
    "model.metCharges(end+1) = model.metCharges(strcmp(model.mets,'nadph[c]'));\n",
    "model.metCHEBIID{end+1} = model.metCHEBIID{strcmp(model.mets,'nadph[c]')};\n",
    "model.metKEGGID{end+1} = model.metKEGGID{strcmp(model.mets,'nadph[c]')};\n",
    "model.metPubChemID{end+1} = model.metPubChemID{strcmp(model.mets,'nadph[c]')};\n",
    "model.metInChIString{end+1} = model.metInChIString{strcmp(model.mets,'nadph[c]')};\n",
    "model.metHMDBID{end+1} = model.metHMDBID{strcmp(model.mets,'nadph[c]')};\n",
    "model.metSmiles{end+1} = model.metSmiles{strcmp(model.mets,'nadph[c]')};\n",
    "model.metPdMap{end+1} = model.metPdMap{strcmp(model.mets,'nadph[c]')};\n",
    "model.metReconMap{end+1} = model.metReconMap{strcmp(model.mets,'nadph[c]')};\n",
    "model.b(end+1) = model.b(strcmp(model.mets,'nadph[c]'));\n",
    "model.csense(end+1) = 'E';\n",
    "\n",
    "model.mets{end+1} = 'nadp[n]';\n",
    "model.S(end+1,:) = zeros(1,length(model.rxns));\n",
    "model.metNames{end+1} = model.metNames{strcmp(model.mets,'nadp[c]')};\n",
    "model.metFormulas{end+1} = model.metFormulas{strcmp(model.mets,'nadp[c]')};\n",
    "model.metCharges(end+1) = model.metCharges(strcmp(model.mets,'nadp[c]'));\n",
    "model.metCHEBIID{end+1} = model.metCHEBIID{strcmp(model.mets,'nadp[c]')};\n",
    "model.metKEGGID{end+1} = model.metKEGGID{strcmp(model.mets,'nadp[c]')};\n",
    "model.metPubChemID{end+1} = model.metPubChemID{strcmp(model.mets,'nadp[c]')};\n",
    "model.metInChIString{end+1} = model.metInChIString{strcmp(model.mets,'nadp[c]')};\n",
    "model.metHMDBID{end+1} = model.metHMDBID{strcmp(model.mets,'nadp[c]')};\n",
    "model.metSmiles{end+1} = model.metSmiles{strcmp(model.mets,'nadp[c]')};\n",
    "model.metPdMap{end+1} = model.metPdMap{strcmp(model.mets,'nadph[c]')};\n",
    "model.metReconMap{end+1} = model.metReconMap{strcmp(model.mets,'nadp[c]')};\n",
    "model.b(end+1) = model.b(strcmp(model.mets,'nadp[c]'));\n",
    "model.csense(end+1) = 'E';\n",
    "\n",
    "model.rxns{end+1} = 'NOXnc';\n",
    "model.S(:,end+1) = zeros(length(model.mets),1);\n",
    "model.S(strcmp(model.mets,'nadph[n]'),end) = -1;\n",
    "model.S(strcmp(model.mets,'o2[n]'),end) = -2;\n",
    "model.S(strcmp(model.mets,'o2s[c]'),end) = 2;\n",
    "model.S(strcmp(model.mets,'nadp[n]'),end) = 1;\n",
    "model.S(strcmp(model.mets,'h[n]'),end) = 1;\n",
    "model.lb(end+1) = 0;\n",
    "model.ub(end+1) = max(model.ub);\n",
    "model.c(end+1) = 0;\n",
    "model.rxnGeneMat(end+1,:) = zeros(1,length(model.genes));\n",
    "model.rxnGeneMat(:,end+1) = zeros(length(model.rxns),1);\n",
    "model.rxnGeneMat(end,end) = 1;\n",
    "model.rules{end+1} = '(x(2253))';\n",
    "model.genes{end+1} = '50507.1';\n",
    "model.grRules{end+1} = '(50507.1)';\n",
    "model.subSystems{end+1} = 'ROS detoxification';\n",
    "model.rxnNames{end+1} = 'NADPH oxidase, nucleus to cytoplasm';\n",
    "model.rxnKEGGID{end+1} = '';\n",
    "model.rxnKeggOrthology{end+1} = '';\n",
    "model.rxnConfidenceScores(end+1) = 0;\n",
    "model.rxnReferences{end+1} = '';\n",
    "model.rxnECNumbers{end+1} = '1.6.3.1';\n",
    "model.rxnNotes{end+1} = '';\n",
    "model.rxnCOG{end+1} = '';\n",
    "model.rxnReconMap{end+1} = '';\n",
    "\n",
    "% NOX4, transfer from endoplasmic reticulum to cytoplasm\n",
    "model.rxns{end+1} = 'NOXrc';\n",
    "model.S(:,end+1) = zeros(length(model.mets),1);\n",
    "model.S(strcmp(model.mets,'nadph[r]'),end) = -1;\n",
    "model.S(strcmp(model.mets,'o2[r]'),end) = -2;\n",
    "model.S(strcmp(model.mets,'o2s[c]'),end) = 2;\n",
    "model.S(strcmp(model.mets,'nadp[r]'),end) = 1;\n",
    "model.S(strcmp(model.mets,'h[r]'),end) = 1;\n",
    "model.lb(end+1) = 0;\n",
    "model.ub(end+1) = max(model.ub);\n",
    "model.c(end+1) = 0;\n",
    "model.rxnGeneMat(end+1,:) = zeros(1,length(model.genes));\n",
    "model.rxnGeneMat(end,end) = 1;\n",
    "model.rules{end+1} = '(x(2253))';\n",
    "model.grRules{end+1} = '(50507.1)';\n",
    "model.subSystems{end+1} = 'ROS detoxification';\n",
    "model.rxnNames{end+1} = 'NADPH oxidase, endoplasmic reticulum to cytoplasm';\n",
    "model.rxnKEGGID{end+1} = '';\n",
    "model.rxnKeggOrthology{end+1} = '';\n",
    "model.rxnConfidenceScores(end+1) = 0;\n",
    "model.rxnReferences{end+1} = '';\n",
    "model.rxnECNumbers{end+1} = '1.6.3.1';\n",
    "model.rxnNotes{end+1} = '';\n",
    "model.rxnCOG{end+1} = '';\n",
    "model.rxnReconMap{end+1} = '';"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Fix aspects of model variable in order to run successfully"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Remove EC numbers for transport reactions (many incorrect)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "for i = 1:length(model.rxns)\n",
    "    if any(strcmp({'Transport, endoplasmic reticular','Transport, extracellular','Transport, golgi apparatus','Transport, lysosomal','Transport, mitochondrial','Transport, nuclear','Transport, peroxisomal'},model.subSystems{i}))\n",
    "        model.rxnECNumbers{i} = '';\n",
    "    end\n",
    "end"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Remove gene '0' and any associations with Recon3D reactions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "model.rules{strcmp(model.rxns,'HDCA24Gtr')} = '';\n",
    "model.grRules{strcmp(model.rxns,'HDCA24Gtr')} = '';\n",
    "\n",
    "i = find(strcmp(model.genes,'0'));\n",
    "for j = (i+1):(length(model.genes))\n",
    "    for k = find(model.rxnGeneMat(:,j))'\n",
    "        model.rules{k} = strrep(model.rules{k},sprintf('x(%d)',j),sprintf('x(%d)',j-1));\n",
    "    end\n",
    "end\n",
    "model.genes(i) = [];\n",
    "model.rxnGeneMat(:,i) = [];"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Remove gene '100507855' and any associations with Recon3D reactions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "model.rules{strcmp(model.rxns,'r0345')} = 'x(161) | x(1877) | x(1876) | x(165) | x(160)';\n",
    "model.grRules{strcmp(model.rxns,'r0345')} = '204.1 or 203.1 or 122481.1 or 26289.1 or 205.1';\n",
    "\n",
    "i = find(strcmp(model.genes,'100507855.1'));\n",
    "for j = (i+1):(length(model.genes))\n",
    "    for k = find(model.rxnGeneMat(:,j))'\n",
    "        model.rules{k} = strrep(model.rules{k},sprintf('x(%d)',j),sprintf('x(%d)',j-1));\n",
    "    end\n",
    "end\n",
    "model.genes(i) = [];\n",
    "model.rxnGeneMat(:,i) = [];"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Remove gene '8041' and any associations with Recon3D reactions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "model.rules{strcmp(model.rxns,'HMR_4790')} = '';\n",
    "model.grRules{strcmp(model.rxns,'HMR_4790')} = '';\n",
    "\n",
    "i = find(strcmp(model.genes,'8041.1'));\n",
    "for j = (i+1):(length(model.genes))\n",
    "    for k = find(model.rxnGeneMat(:,j))'\n",
    "        model.rules{k} = strrep(model.rules{k},sprintf('x(%d)',j),sprintf('x(%d)',j-1));\n",
    "    end\n",
    "end\n",
    "model.genes(i) = [];\n",
    "model.rxnGeneMat(:,i) = [];"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Rewrite 'model.rules' to fix any original errors"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "for i = 1:length(model.rxns)\n",
    "    genes = {};\n",
    "    geneids = {};\n",
    "    isoforms = [];\n",
    "    lengths = [];\n",
    "    for j = find(model.rxnGeneMat(i,:))\n",
    "        genes{end+1} = model.genes{j};\n",
    "        geneids{end+1} = num2str(j);\n",
    "        s = strsplit(model.genes{j},'.');\n",
    "        isoforms(end+1) = str2num(s{2});\n",
    "        lengths(end+1) = length(model.genes{j});\n",
    "    end\n",
    "    [~,index] = sort(lengths,'descend');\n",
    "    genes = genes(index);\n",
    "    geneids = geneids(index);\n",
    "    isoforms = isoforms(index);\n",
    "    \n",
    "    [~,index] = sort(isoforms,'descend');\n",
    "    genes = genes(index);\n",
    "    geneids = geneids(index);\n",
    "    \n",
    "    model.rules{i} = model.grRules{i};\n",
    "    for j = 1:length(genes)\n",
    "        model.rules{i} = strrep(model.rules{i},genes{j},sprintf('x(%s)',geneids{j}));\n",
    "    end\n",
    "    model.rules{i} = strrep(strrep(model.rules{i},'or','|'),'and','&');\n",
    "end"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Remove gene isoform information - genes with different isoforms are combined into one gene ID"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "% new list of genes without isoforms\n",
    "genes = {};\n",
    "for i = 1:length(model.genes)\n",
    "    s = strsplit(model.genes{i},'.');\n",
    "    if ~any(strcmp(genes,s{1}))\n",
    "        genes{end+1} = s{1};\n",
    "    end\n",
    "end\n",
    "genes = sort(genes);\n",
    "\n",
    "% rewrite model.rules and model.rxnGeneMat for new gene list without isoforms\n",
    "rxnGeneMat = zeros(length(model.rxns),length(genes));\n",
    "for i = 1:length(model.rxns)\n",
    "    for j = find(model.rxnGeneMat(i,:))\n",
    "        s = strsplit(model.genes{j},'.');\n",
    "        model.rules{i} = strrep(model.rules{i},sprintf('x(%d)',j),sprintf('x(%d)',find(strcmp(genes,s{1}))));\n",
    "        rxnGeneMat(i,find(strcmp(genes,s{1}))) = 1;\n",
    "    end\n",
    "end\n",
    "model.rxnGeneMat = rxnGeneMat;\n",
    "model.genes = genes;\n",
    "\n",
    "for i = 1:length(model.rxns)\n",
    "    \n",
    "    % if has rule\n",
    "    if ~isempty(model.rules{i})\n",
    "        rule = model.rules{i};\n",
    "        \n",
    "        % insert symbolic variables\n",
    "        s = {};\n",
    "        for j = find(model.rxnGeneMat(i,:))\n",
    "            s{end+1} = sprintf('x%d',j);\n",
    "            rule = strrep(rule,sprintf('x(%d)',j),sprintf('x%d',j));\n",
    "        end\n",
    "        \n",
    "        % create symbolic variables\n",
    "        for j = 1:length(s)\n",
    "            syms (s{j});\n",
    "        end\n",
    "        \n",
    "        % simplify grRule\n",
    "        rule = char(simplify(sym(rule)));\n",
    "        \n",
    "        % insert gene variables\n",
    "        for j = sort(find(model.rxnGeneMat(i,:)),'descend')\n",
    "            rule = strrep(rule,sprintf('x%d',j),sprintf('x(%d)',j));\n",
    "        end\n",
    "        \n",
    "        % replace rules\n",
    "        model.rules{i} = rule;\n",
    "        \n",
    "        % replace grRules\n",
    "        for j = sort(find(model.rxnGeneMat(i,:)),'descend')\n",
    "            rule = strrep(strrep(strrep(rule,sprintf('x(%d)',j),sprintf('(%s)',model.genes{j})),'|','or'),'&','and');\n",
    "        end\n",
    "        model.grRules{i} = rule;\n",
    "        \n",
    "    end\n",
    "end\n",
    "\n",
    "% rewrite model.rxnGeneMat with simplified gene rules\n",
    "model.rxnGeneMat = zeros(length(model.rxns),length(genes));\n",
    "for i = 1:length(model.rxns)\n",
    "    for j = 1:length(model.genes)\n",
    "        if contains(model.rules{i},sprintf('x(%d)',j))\n",
    "            model.rxnGeneMat(i,j) = 1;\n",
    "        end\n",
    "    end   \n",
    "end\n",
    "\n",
    "% remove genes no longer in GPR rules\n",
    "gene_length_original = length(model.genes);\n",
    "genes_to_remove = sort(find(sum(model.rxnGeneMat)==0),'descend');\n",
    "for i = genes_to_remove\n",
    "    model.genes(i) = [];\n",
    "    model.rxnGeneMat(:,i) = [];    \n",
    "    for j = (i+1):gene_length_original\n",
    "        for k = 1:length(model.rxns)\n",
    "            model.rules{k} = strrep(model.rules{k},sprintf('x(%d)',j),sprintf('x(%d)',j-1));\n",
    "        end\n",
    "    end\n",
    "end"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Fix metabolite and reaction ID's"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Fix metabolite CHEBI ID's"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "for i = 1:length(model.mets)\n",
    "    if length(model.metCHEBIID{i}) > 6\n",
    "        if strcmpi(model.metCHEBIID{i}(1:6),'CHEBI:')\n",
    "            model.metCHEBIID{i} = model.metCHEBIID{i}(7:end);\n",
    "        end\n",
    "    end \n",
    "end"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Fix reaction EC numbers"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "for i = 1:length(model.rxns)\n",
    "    \n",
    "    % if has ec number\n",
    "    if ~isempty(model.rxnECNumbers{i})\n",
    "        \n",
    "        % if TCDB number, remove\n",
    "        if strcmp(model.rxnECNumbers{i}(1:5),'TCDB:')\n",
    "            model.rxnECNumbers{i} = '';\n",
    "            \n",
    "        % if spontaneous, remove\n",
    "        elseif any(strcmp({'NonEnzymatic','Spontaneous','spontaneous'},model.rxnECNumbers{i}))\n",
    "            model.rxnECNumbers{i} = '';\n",
    "        \n",
    "        else\n",
    "            \n",
    "            % split on \",\" and \";\"\n",
    "            ecs = strsplit(model.rxnECNumbers{i},{',',';','or',' , ',' ; ',' or ',' '});\n",
    "            \n",
    "            % iterate over individual ec's\n",
    "            to_remove = [];\n",
    "            for j = 1:length(ecs)\n",
    "                \n",
    "                % if empty, remove\n",
    "                if isempty(ecs{j})\n",
    "                    to_remove(end+1) = j;\n",
    "                \n",
    "                % if starts with \"EC:\", remove \"EC:\"\n",
    "                elseif strcmp(ecs{j}(1:3),'EC:')\n",
    "                    ecs{j} = ecs{j}(4:end);\n",
    "                end\n",
    "                \n",
    "            end\n",
    "            \n",
    "            % remove any empty ec's\n",
    "            for j = sort(to_remove,'descend')\n",
    "                ecs(j) = [];\n",
    "            end\n",
    "            \n",
    "            % see if can remove any general ec's that contain more specific ec\n",
    "            % eg. can remove 1.2.3.- if also contains 1.2.3.4\n",
    "            % also removes duplicate ec's\n",
    "            dash1_index = []; dash1_val1 = {};\n",
    "            dash2_index = []; dash2_val1 = {}; dash2_val2 = {};\n",
    "            dash3_index = []; dash3_val1 = {}; dash3_val2 = {}; dash3_val3 = {};\n",
    "            nodash_index = []; nodash_val1 = {}; nodash_val2 = {}; nodash_val3 = {}; nodash_val4 = {};\n",
    "            for j = 1:length(ecs)\n",
    "                split = strsplit(ecs{j},'.');\n",
    "                if strcmp(split{2},'-')\n",
    "                    dash1_index(end+1) = j;\n",
    "                    dash1_val1{end+1} = split{1};\n",
    "                elseif strcmp(split{3},'-')\n",
    "                    dash2_index(end+1) = j;\n",
    "                    dash2_val1{end+1} = split{1};\n",
    "                    dash2_val2{end+1} = split{2};\n",
    "                elseif strcmp(split{4},'-')\n",
    "                    dash3_index(end+1) = j;\n",
    "                    dash3_val1{end+1} = split{1};\n",
    "                    dash3_val2{end+1} = split{2};\n",
    "                    dash3_val3{end+1} = split{3};\n",
    "                else\n",
    "                    nodash_index(end+1) = j;\n",
    "                    nodash_val1{end+1} = split{1};\n",
    "                    nodash_val2{end+1} = split{2};\n",
    "                    nodash_val3{end+1} = split{3};\n",
    "                    nodash_val4{end+1} = split{4};\n",
    "                end\n",
    "            end\n",
    "            \n",
    "            to_remove = [];\n",
    "            \n",
    "            for j = 1:length(dash1_index)\n",
    "                for k = (j+1):length(dash1_index)\n",
    "                    if strcmp(dash1_val1{j},dash1_val1{k})\n",
    "                        if ~any(to_remove == dash1_index(j))\n",
    "                            to_remove(end+1) = dash1_index(j);\n",
    "                        end\n",
    "                    end\n",
    "                end\n",
    "                for k = 1:length(dash2_index)\n",
    "                    if strcmp(dash1_val1{j},dash2_val1{k})\n",
    "                        if ~any(to_remove == dash1_index(j))\n",
    "                            to_remove(end+1) = dash1_index(j);\n",
    "                        end\n",
    "                    end\n",
    "                end\n",
    "                for k = 1:length(dash3_index)\n",
    "                    if strcmp(dash1_val1{j},dash3_val1{k})\n",
    "                        if ~any(to_remove == dash1_index(j))\n",
    "                            to_remove(end+1) = dash1_index(j);\n",
    "                        end\n",
    "                    end\n",
    "                end\n",
    "                for k = 1:length(nodash_index)\n",
    "                    if strcmp(dash1_val1{j},nodash_val1{k})\n",
    "                        if ~any(to_remove == dash1_index(j))\n",
    "                            to_remove(end+1) = dash1_index(j);\n",
    "                        end\n",
    "                    end\n",
    "                end\n",
    "            end\n",
    "            \n",
    "            for j = 1:length(dash2_index)\n",
    "                for k = (j+1):length(dash2_index)\n",
    "                    if strcmp(dash2_val1{j},dash2_val1{k}) && strcmp(dash2_val2{j},dash2_val2{k})\n",
    "                        if ~any(to_remove == dash2_index(j))\n",
    "                            to_remove(end+1) = dash2_index(j);\n",
    "                        end\n",
    "                    end\n",
    "                end\n",
    "                for k = 1:length(dash3_index)\n",
    "                    if strcmp(dash2_val1{j},dash3_val1{k}) && strcmp(dash2_val2{j},dash3_val2{k})\n",
    "                        if ~any(to_remove == dash2_index(j))\n",
    "                            to_remove(end+1) = dash2_index(j);\n",
    "                        end\n",
    "                    end\n",
    "                end\n",
    "                for k = 1:length(nodash_index)\n",
    "                    if strcmp(dash2_val1{j},nodash_val1{k}) && strcmp(dash2_val2{j},nodash_val2{k})\n",
    "                        if ~any(to_remove == dash2_index(j))\n",
    "                            to_remove(end+1) = dash2_index(j);\n",
    "                        end\n",
    "                    end\n",
    "                end\n",
    "            end\n",
    "            \n",
    "            for j = 1:length(dash3_index)\n",
    "                for k = (j+1):length(dash3_index)\n",
    "                    if strcmp(dash3_val1{j},dash3_val1{k}) && strcmp(dash3_val2{j},dash3_val2{k}) && strcmp(dash3_val3{j},dash3_val3{k})\n",
    "                        if ~any(to_remove == dash3_index(j))\n",
    "                            to_remove(end+1) = dash3_index(j);\n",
    "                        end\n",
    "                    end\n",
    "                end\n",
    "                for k = 1:length(nodash_index)\n",
    "                    if strcmp(dash3_val1{j},nodash_val1{k}) && strcmp(dash3_val2{j},nodash_val2{k}) && strcmp(dash3_val3{j},nodash_val3{k})\n",
    "                        if ~any(to_remove == dash3_index(j))\n",
    "                            to_remove(end+1) = dash3_index(j);\n",
    "                        end\n",
    "                    end\n",
    "                end\n",
    "            end\n",
    "            \n",
    "            for j = 1:length(nodash_index)\n",
    "                for k = (j+1):length(nodash_index)\n",
    "                    if strcmp(nodash_val1{j},nodash_val1{k}) && strcmp(nodash_val2{j},nodash_val2{k}) && strcmp(nodash_val3{j},nodash_val3{k}) && strcmp(nodash_val4{j},nodash_val4{k})\n",
    "                        if ~any(to_remove == nodash_index(j))\n",
    "                            to_remove(end+1) = nodash_index(j);\n",
    "                        end\n",
    "                    end\n",
    "                end\n",
    "            end\n",
    "        \n",
    "            for j = sort(to_remove,'descend')\n",
    "                ecs(j) = [];\n",
    "            end\n",
    "\n",
    "            % create new ec number string\n",
    "            model.rxnECNumbers{i} = strjoin(ecs,' | ');\n",
    "            \n",
    "        end\n",
    "    end\n",
    "end"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Obtain gene symbols for each gene ID"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "% initialize list of gene symbols\n",
    "model.geneSymbols = {};\n",
    "\n",
    "% get conversions\n",
    "% split into groups of 300 genes\n",
    "for i = 1:floor(length(genes)/300)\n",
    "    conversion = webread(sprintf('https://biodbnet-abcc.ncifcrf.gov/webServices/rest.php/biodbnetRestApi.json?method=db2db&input=geneid&inputValues=%s&outputs=genesymbol&taxonId=9606&format=row',strjoin(model.genes((300*(i-1)+1):(300*i)),',')));\n",
    "    for j = 1:length(conversion)\n",
    "        model.geneSymbols{end+1} = conversion(j).GeneSymbol;\n",
    "    end\n",
    "    pause(10)\n",
    "end\n",
    "conversion = webread(sprintf('https://biodbnet-abcc.ncifcrf.gov/webServices/rest.php/biodbnetRestApi.json?method=db2db&input=geneid&inputValues=%s&outputs=genesymbol&taxonId=9606&format=row',strjoin(model.genes((300*floor(length(genes)/300)+1):end),',')));\n",
    "for j = 1:length(conversion)\n",
    "    model.geneSymbols{end+1} = conversion(j).GeneSymbol;\n",
    "end\n",
    "\n",
    "model.geneSymbols = model.geneSymbols';"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Remove EC numbers for transport reactions"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Many EC numbers for transport reactions are incorrect"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "for i = 1:length(model.rxns)\n",
    "    if any(strcmp({'Transport, endoplasmic reticular','Transport, extracellular','Transport, golgi apparatus','Transport, lysosomal','Transport, mitochondrial','Transport, nuclear','Transport, peroxisomal'},model.subSystems{i}))\n",
    "        model.rxnECNumbers{i} = '';\n",
    "    end\n",
    "end"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Add KEGG ID's for reactions"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Load reactions json file"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "data = fileread('../deltag/input_data/reactions.json');\n",
    "data = jsondecode(data);"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Add available KEGG ID's"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "for i = 1:length(data.results)\n",
    "    temp = data.results(i);\n",
    "    rxn = temp.abbreviation;\n",
    "    keggid = temp.keggId;\n",
    "    if ~strcmp(keggid,'') && length(keggid)>0 && sum(strcmp(model.rxns,rxn))==1  \n",
    "        model.rxnKEGGID{strcmp(model.rxns,rxn)} = keggid;\n",
    "    end\n",
    "end"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Remove reactions for accurate results"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Removing the following subsystems accurately prevents de novo production of essential amino acids: 'Protein assembly', 'Protein degradation', 'Protein modification', 'Protein formation'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "subsystems_to_remove = {'Protein assembly','Protein degradation','Protein modification','Protein formation'};\n",
    "for i = 1:length(model.rxns)\n",
    "    if any(strcmp(subsystems_to_remove,model.subSystems{i}))\n",
    "        model.lb(i) = 0;\n",
    "        model.ub(i) = 0;\n",
    "    end\n",
    "end"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Remove bile acid synthesis"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "for i = 1:length(model.rxns)\n",
    "    if strcmp(model.subSystems{i},'Bile acid synthesis')\n",
    "        model.lb(i) = 0;\n",
    "        model.ub(i) = 0;\n",
    "    end\n",
    "end"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Remove NQO1 futile cycle reactions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "for i = 1:length(model.rxns)\n",
    "    if any(strcmp({'NADPQNOXR','NADQNOXR','HMR_9538','HMR_9722'},model.rxns{i}))\n",
    "        model.lb(i) = 0;\n",
    "        model.ub(i) = 0;\n",
    "    end\n",
    "end"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Remove oxalosuccinic acid-producing IDH1 reactions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "for i = 1:length(model.rxns)\n",
    "    if any(strcmp({'r0423','r0424','r0425','r0422'},model.rxns{i}))\n",
    "        model.lb(i) = 0;\n",
    "        model.ub(i) = 0;\n",
    "    end\n",
    "end"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Remove duplicate reactions"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Some reactions exist where stoichiometric vector is an exact multiple of that of other reaction"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "reactions_to_remove = {'HMR_7257','PEHSFABPe','G6PDH2c','GNDc','PGLc','RPEc'};\n",
    "for i = 1:length(model.rxns)\n",
    "    if any(strcmp(reactions_to_remove,model.rxns{i}))\n",
    "        model.lb(i) = 0;\n",
    "        model.ub(i) = 0;\n",
    "    end\n",
    "end"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Cleanup"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Write reaction formulas"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "% reaction formulas\n",
    "formulas = {};\n",
    "for i = 1:length(model.rxns)\n",
    "    \n",
    "    % initialize formula\n",
    "    formulas{end+1} = '';\n",
    "    \n",
    "    % substrates\n",
    "    for j = find(model.S(:,i)<0)'\n",
    "        if isempty(formulas{end})\n",
    "            formulas{end} = sprintf('%d %s',-full(model.S(j,i)),model.mets{j});\n",
    "        else\n",
    "            formulas{end} = [formulas{end},' + ',sprintf('%d %s',-full(model.S(j,i)),model.mets{j})];\n",
    "        end\n",
    "    end\n",
    "    \n",
    "    % reaction arrow\n",
    "    if (model.lb(i) < 0) && (model.ub(i) > 0)\n",
    "        formulas{end} = [formulas{end},' <=> '];\n",
    "    elseif (model.lb(i) < 0)\n",
    "        formulas{end} = [formulas{end},' <-- '];\n",
    "    elseif (model.ub(i) > 0)\n",
    "        formulas{end} = [formulas{end},' --> '];\n",
    "    else\n",
    "        formulas{end} = [formulas{end},' X--X '];\n",
    "    end\n",
    "    \n",
    "    % products\n",
    "    numproducts = 0;\n",
    "    for j = find(model.S(:,i)>0)'\n",
    "        if numproducts == 0\n",
    "            formulas{end} = [formulas{end},sprintf('%d %s',full(model.S(j,i)),model.mets{j})];\n",
    "        else\n",
    "            formulas{end} = [formulas{end},' + ',sprintf('%d %s',full(model.S(j,i)),model.mets{j})];\n",
    "        end\n",
    "        numproducts = numproducts + 1;\n",
    "    end\n",
    "    \n",
    "end\n",
    "\n",
    "% add to model variable\n",
    "model.rxnFormulas = formulas;"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Remove quotations from reaction references"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "% remove quotations from reaction references\n",
    "for i = 1:length(model.rxns)\n",
    "    model.rxnReferences{i} = strrep(strrep(model.rxnReferences{i},\"'\",\"\"),'\"','');\n",
    "end"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Metabolite names"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "model.metNames{6588} = model.metNames{6588}(3:end);\n",
    "model.metNames{7590} = model.metNames{7590}(3:end);\n",
    "model.metNames{7591} = model.metNames{7591}(3:end);"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### IDH1 new reactions"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### IDH1 neomorphic reactions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "model.rxns{end+1} = 'IDH1_R132';\n",
    "model.S(:,end+1) = zeros(length(model.mets),1);\n",
    "model.S(strcmp(model.mets,'akg[c]'),end) = -1;\n",
    "model.S(strcmp(model.mets,'nadph[c]'),end) = -1;\n",
    "model.S(strcmp(model.mets,'M00653[c]'),end) = 1;\n",
    "model.S(strcmp(model.mets,'nadp[c]'),end) = 1;\n",
    "model.lb(end+1) = 0;\n",
    "model.ub(end+1) = max(model.ub);\n",
    "model.c(end+1) = 0;\n",
    "model.rxnGeneMat(end+1,:) = zeros(1,length(model.genes));\n",
    "model.rxnGeneMat(end,strcmp(model.geneSymbols,'IDH1')) = 1;\n",
    "model.rules{end+1} = 'x(1301)';\n",
    "model.grRules{end+1} = '(3417)';\n",
    "model.subSystems{end+1} = 'IDH1 neomorphic';\n",
    "model.rxnNames{end+1} = 'IDH1 neomorphic reaction';\n",
    "model.rxnKEGGID{end+1} = '';\n",
    "model.rxnKeggOrthology{end+1} = '';\n",
    "model.rxnConfidenceScores(end+1) = 0;\n",
    "model.rxnReferences{end+1} = '';\n",
    "model.rxnECNumbers{end+1} = '';\n",
    "model.rxnNotes{end+1} = '';\n",
    "model.rxnCOG{end+1} = '';\n",
    "model.rxnReconMap{end+1} = '';\n",
    "model.rxnFormulas{end+1} =  '1 akg[c] + 1 nadph[c] --> 1 M00653[c] + 1 nadp[c]';"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### 2HG c->e reaction"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "model.rxns{end+1} = '2HG_c_e';\n",
    "model.S(:,end+1) = zeros(length(model.mets),1);\n",
    "model.S(strcmp(model.mets,'M00653[c]'),end) = -1;\n",
    "model.S(strcmp(model.mets,'M00653[e]'),end) = 1;\n",
    "model.lb(end+1) = 0;\n",
    "model.ub(end+1) = max(model.ub);\n",
    "model.c(end+1) = 0;\n",
    "model.rxnGeneMat(end+1,:) = zeros(1,length(model.genes));\n",
    "model.rules{end+1} = '';\n",
    "model.grRules{end+1} = '';\n",
    "model.subSystems{end+1} = 'IDH1 R132';\n",
    "model.rxnNames{end+1} = '2HG transport, cytosol to extracellular';\n",
    "model.rxnKEGGID{end+1} = '';\n",
    "model.rxnKeggOrthology{end+1} = '';\n",
    "model.rxnConfidenceScores(end+1) = 0;\n",
    "model.rxnReferences{end+1} = '';\n",
    "model.rxnECNumbers{end+1} = '';\n",
    "model.rxnNotes{end+1} = '';\n",
    "model.rxnCOG{end+1} = '';\n",
    "model.rxnReconMap{end+1} = '';\n",
    "model.rxnFormulas{end+1} =  '1 M00653[c] --> 1 M00653[e]';"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### 2HG e-> reaction"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "model.rxns{end+1} = '2HG_e_null';\n",
    "model.S(:,end+1) = zeros(length(model.mets),1);\n",
    "model.S(strcmp(model.mets,'M00653[e]'),end) = -1;\n",
    "model.lb(end+1) = 0;\n",
    "model.ub(end+1) = max(model.ub);\n",
    "model.c(end+1) = 0;\n",
    "model.rxnGeneMat(end+1,:) = zeros(1,length(model.genes));\n",
    "model.rules{end+1} = '';\n",
    "model.grRules{end+1} = '';\n",
    "model.subSystems{end+1} = 'IDH1 R132';\n",
    "model.rxnNames{end+1} = '2HG transport, extracellular to null';\n",
    "model.rxnKEGGID{end+1} = '';\n",
    "model.rxnKeggOrthology{end+1} = '';\n",
    "model.rxnConfidenceScores(end+1) = 0;\n",
    "model.rxnReferences{end+1} = '';\n",
    "model.rxnECNumbers{end+1} = '';\n",
    "model.rxnNotes{end+1} = '';\n",
    "model.rxnCOG{end+1} = '';\n",
    "model.rxnReconMap{end+1} = '';\n",
    "model.rxnFormulas{end+1} =  '1 M00653[e] --> ';"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Drug modules"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### 5-fluorouracil"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Metabolites"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "% 5fu[e]\n",
    "model.S(end+1,:) = zeros(1,length(model.rxns));\n",
    "model.mets{end+1} = '5fu[e]';\n",
    "model.b(end+1) = 0;\n",
    "model.csense(end+1) = 'E';\n",
    "model.metCharges(end+1) = 0;\n",
    "model.metFormulas{end+1} = 'C4H3FN2O2';\n",
    "model.metSmiles{end+1} = 'C1=C(C(=O)NC(=O)N1)F';\n",
    "model.metNames{end+1} = '5-fluorouracil';\n",
    "model.metHMDBID{end+1} = 'HMDB0014684';\n",
    "model.metInChIString{end+1} = 'InChI=1S/C4H3FN2O2/c5-2-1-6-4(9)7-3(2)8/h1H,(H2,6,7,8,9)';\n",
    "model.metKEGGID{end+1} = '';\n",
    "model.metPubChemID{end+1} = '3385';\n",
    "model.metCHEBIID{end+1} = '46345';\n",
    "model.metPdMap{end+1} = '';\n",
    "model.metReconMap{end+1} = '';\n",
    "\n",
    "% 5fu[c]\n",
    "model.S(end+1,:) = zeros(1,length(model.rxns));\n",
    "model.mets{end+1} = '5fu[c]';\n",
    "model.b(end+1) = 0;\n",
    "model.csense(end+1) = 'E';\n",
    "model.metCharges(end+1) = 0;\n",
    "model.metFormulas{end+1} = 'C4H3FN2O2';\n",
    "model.metSmiles{end+1} = 'C1=C(C(=O)NC(=O)N1)F';\n",
    "model.metNames{end+1} = '5-Fluorouracil';\n",
    "model.metHMDBID{end+1} = 'HMDB0014684';\n",
    "model.metInChIString{end+1} = 'InChI=1S/C4H3FN2O2/c5-2-1-6-4(9)7-3(2)8/h1H,(H2,6,7,8,9)';\n",
    "model.metKEGGID{end+1} = '';\n",
    "model.metPubChemID{end+1} = '3385';\n",
    "model.metCHEBIID{end+1} = '46345';\n",
    "model.metPdMap{end+1} = '';\n",
    "model.metReconMap{end+1} = '';\n",
    "\n",
    "% fdurd[c]\n",
    "model.S(end+1,:) = zeros(1,length(model.rxns));\n",
    "model.mets{end+1} = 'fdurd[c]';\n",
    "model.b(end+1) = 0;\n",
    "model.csense(end+1) = 'E';\n",
    "model.metCharges(end+1) = 0;\n",
    "model.metFormulas{end+1} = 'C9H11FN2O5';\n",
    "model.metSmiles{end+1} = 'C1C(C(OC1N2C=C(C(=O)NC2=O)F)CO)O';\n",
    "model.metNames{end+1} = 'Fluorodeoxyuridine';\n",
    "model.metHMDBID{end+1} = 'HMDB0014467';\n",
    "model.metInChIString{end+1} = 'InChI=1S/C9H11FN2O5/c10-4-2-12(9(16)11-8(4)15)7-1-5(14)6(3-13)17-7/h2,5-7,13-14H,1,3H2,(H,11,15,16)/t5-,6+,7+/m0/s1';\n",
    "model.metKEGGID{end+1} = '';\n",
    "model.metPubChemID{end+1} = '5790';\n",
    "model.metCHEBIID{end+1} = '60761';\n",
    "model.metPdMap{end+1} = '';\n",
    "model.metReconMap{end+1} = '';\n",
    "\n",
    "% fdump[c]\n",
    "model.S(end+1,:) = zeros(1,length(model.rxns));\n",
    "model.mets{end+1} = 'fdump[c]';\n",
    "model.b(end+1) = 0;\n",
    "model.csense(end+1) = 'E';\n",
    "model.metCharges(end+1) = 0;\n",
    "model.metFormulas{end+1} = 'C9H10FN2O8P';\n",
    "model.metSmiles{end+1} = 'C1C(C(OC1(N2C=CC(=O)NC2=O)F)COP(=O)(O)O)O';\n",
    "model.metNames{end+1} = 'Fluorodeoxyuridine monophosphate';\n",
    "model.metHMDBID{end+1} = '';\n",
    "model.metInChIString{end+1} = 'InChI=1S/C9H12FN2O8P/c10-9(12-2-1-7(14)11-8(12)15)3-5(13)6(20-9)4-19-21(16,17)18/h1-2,5-6,13H,3-4H2,(H,11,14,15)(H2,16,17,18)/t5-,6+,9-/m0/s1';\n",
    "model.metKEGGID{end+1} = '';\n",
    "model.metPubChemID{end+1} = '66697563';\n",
    "model.metCHEBIID{end+1} = '';\n",
    "model.metPdMap{end+1} = '';\n",
    "model.metReconMap{end+1} = '';\n",
    "\n",
    "% fdump[e]\n",
    "model.S(end+1,:) = zeros(1,length(model.rxns));\n",
    "model.mets{end+1} = 'fdump[e]';\n",
    "model.b(end+1) = 0;\n",
    "model.csense(end+1) = 'E';\n",
    "model.metCharges(end+1) = 0;\n",
    "model.metFormulas{end+1} = 'C9H10FN2O8P';\n",
    "model.metSmiles{end+1} = 'C1C(C(OC1(N2C=CC(=O)NC2=O)F)COP(=O)(O)O)O';\n",
    "model.metNames{end+1} = 'Fluorodeoxyuridine monophosphate';\n",
    "model.metHMDBID{end+1} = '';\n",
    "model.metInChIString{end+1} = 'InChI=1S/C9H12FN2O8P/c10-9(12-2-1-7(14)11-8(12)15)3-5(13)6(20-9)4-19-21(16,17)18/h1-2,5-6,13H,3-4H2,(H,11,14,15)(H2,16,17,18)/t5-,6+,9-/m0/s1';\n",
    "model.metKEGGID{end+1} = '';\n",
    "model.metPubChemID{end+1} = '66697563';\n",
    "model.metCHEBIID{end+1} = '';\n",
    "model.metPdMap{end+1} = '';\n",
    "model.metReconMap{end+1} = '';\n",
    "\n",
    "% fdudp[c]\n",
    "model.S(end+1,:) = zeros(1,length(model.rxns));\n",
    "model.mets{end+1} = 'fdudp[c]';\n",
    "model.b(end+1) = 0;\n",
    "model.csense(end+1) = 'E';\n",
    "model.metCharges(end+1) = 0;\n",
    "model.metFormulas{end+1} = 'C9H10FN2O11P2';\n",
    "model.metSmiles{end+1} = 'C1C(C(OC1(N2C=CC(=O)NC2=O)F)COP(=O)(O)OP(=O)(O)O)O';\n",
    "model.metNames{end+1} = 'Fluorodeoxyuridine diphosphate';\n",
    "model.metHMDBID{end+1} = '';\n",
    "model.metInChIString{end+1} = 'InChI=1S/C9H13FN2O11P2/c10-9(12-2-1-7(14)11-8(12)15)3-5(13)6(22-9)4-21-25(19,20)23-24(16,17)18/h1-2,5-6,13H,3-4H2,(H,19,20)(H,11,14,15)(H2,16,17,18)/t5-,6+,9-/m0/s1';\n",
    "model.metKEGGID{end+1} = '';\n",
    "model.metPubChemID{end+1} = '87203011';\n",
    "model.metCHEBIID{end+1} = '';\n",
    "model.metPdMap{end+1} = '';\n",
    "model.metReconMap{end+1} = '';\n",
    "\n",
    "% furd[c]\n",
    "model.S(end+1,:) = zeros(1,length(model.rxns));\n",
    "model.mets{end+1} = 'furd[c]';\n",
    "model.b(end+1) = 0;\n",
    "model.csense(end+1) = 'E';\n",
    "model.metCharges(end+1) = 0;\n",
    "model.metFormulas{end+1} = 'C9H11FN2O6';\n",
    "model.metSmiles{end+1} = 'C1=C(C(=O)NC(=O)N1C2C(C(C(O2)CO)O)O)F';\n",
    "model.metNames{end+1} = 'Fluorouridine';\n",
    "model.metHMDBID{end+1} = 'HMDB0060396';\n",
    "model.metInChIString{end+1} = 'InChI=1S/C9H11FN2O6/c10-3-1-12(9(17)11-7(3)16)8-6(15)5(14)4(2-13)18-8/h1,4-6,8,13-15H,2H2,(H,11,16,17)/t4-,5-,6-,8-/m1/s1';\n",
    "model.metKEGGID{end+1} = '';\n",
    "model.metPubChemID{end+1} = '9427';\n",
    "model.metCHEBIID{end+1} = '40154';\n",
    "model.metPdMap{end+1} = '';\n",
    "model.metReconMap{end+1} = '';\n",
    "\n",
    "% fump[c]\n",
    "model.S(end+1,:) = zeros(1,length(model.rxns));\n",
    "model.mets{end+1} = 'fump[c]';\n",
    "model.b(end+1) = 0;\n",
    "model.csense(end+1) = 'E';\n",
    "model.metCharges(end+1) = 0;\n",
    "model.metFormulas{end+1} = 'C9H10FN2O9P';\n",
    "model.metSmiles{end+1} = 'C1=CN(C(=O)NC1=O)C2(C(C(C(O2)COP(=O)(O)O)O)O)F';\n",
    "model.metNames{end+1} = 'Fluorouridine monophosphate';\n",
    "model.metHMDBID{end+1} = '';\n",
    "model.metInChIString{end+1} = 'InChI=1S/C9H12FN2O9P/c10-9(12-2-1-5(13)11-8(12)16)7(15)6(14)4(21-9)3-20-22(17,18)19/h1-2,4,6-7,14-15H,3H2,(H,11,13,16)(H2,17,18,19)/t4-,6-,7-,9+/m1/s1';\n",
    "model.metKEGGID{end+1} = '';\n",
    "model.metPubChemID{end+1} = '67305234';\n",
    "model.metCHEBIID{end+1} = '';\n",
    "model.metPdMap{end+1} = '';\n",
    "model.metReconMap{end+1} = '';\n",
    "\n",
    "% fudp[c]\n",
    "model.S(end+1,:) = zeros(1,length(model.rxns));\n",
    "model.mets{end+1} = 'fudp[c]';\n",
    "model.b(end+1) = 0;\n",
    "model.csense(end+1) = 'E';\n",
    "model.metCharges(end+1) = 0;\n",
    "model.metFormulas{end+1} = 'C9H10FN2O12P2';\n",
    "model.metSmiles{end+1} = 'C1=CN(C(=O)NC1=O)C2(C(C(C(O2)COP(=O)(O)OP(=O)(O)O)O)O)F';\n",
    "model.metNames{end+1} = 'Fluorouridine diphosphate';\n",
    "model.metHMDBID{end+1} = '';\n",
    "model.metInChIString{end+1} = 'InChI=1S/C9H13FN2O12P2/c10-9(12-2-1-5(13)11-8(12)16)7(15)6(14)4(23-9)3-22-26(20,21)24-25(17,18)19/h1-2,4,6-7,14-15H,3H2,(H,20,21)(H,11,13,16)(H2,17,18,19)/t4-,6-,7-,9+/m1/s1';\n",
    "model.metKEGGID{end+1} = '';\n",
    "model.metPubChemID{end+1} = '87438244';\n",
    "model.metCHEBIID{end+1} = '';\n",
    "model.metPdMap{end+1} = '';\n",
    "model.metReconMap{end+1} = '';\n",
    "\n",
    "% dhfu[c]\n",
    "model.S(end+1,:) = zeros(1,length(model.rxns));\n",
    "model.mets{end+1} = 'dhfu[c]';\n",
    "model.b(end+1) = 0;\n",
    "model.csense(end+1) = 'E';\n",
    "model.metCharges(end+1) = 0;\n",
    "model.metFormulas{end+1} = 'C4H5FN2O2';\n",
    "model.metSmiles{end+1} = 'C1C(C(=O)NC(=O)N1)F';\n",
    "model.metNames{end+1} = '5,6-Dihydro-5-fluorouracil';\n",
    "model.metHMDBID{end+1} = 'HMDB0060402';\n",
    "model.metInChIString{end+1} = 'InChI=1S/C4H5FN2O2/c5-2-1-6-4(9)7-3(2)8/h2H,1H2,(H2,6,7,8,9)';\n",
    "model.metKEGGID{end+1} = '';\n",
    "model.metPubChemID{end+1} = '121997';\n",
    "model.metCHEBIID{end+1} = '80624';\n",
    "model.metPdMap{end+1} = '';\n",
    "model.metReconMap{end+1} = '';\n",
    "\n",
    "% fupa[c]\n",
    "model.S(end+1,:) = zeros(1,length(model.rxns));\n",
    "model.mets{end+1} = 'fupa[c]';\n",
    "model.b(end+1) = 0;\n",
    "model.csense(end+1) = 'E';\n",
    "model.metCharges(end+1) = 0;\n",
    "model.metFormulas{end+1} = 'C4H6FN2O3';\n",
    "model.metSmiles{end+1} = 'C(C(C(=O)O)F)NC(=O)N';\n",
    "model.metNames{end+1} = 'Alpha-fluoro-ureidopropionic acid';\n",
    "model.metHMDBID{end+1} = 'HMDB0060435';\n",
    "model.metInChIString{end+1} = 'InChI=1S/C4H7FN2O3/c5-2(3(8)9)1-7-4(6)10/h2H,1H2,(H,8,9)(H3,6,7,10)';\n",
    "model.metKEGGID{end+1} = '';\n",
    "model.metPubChemID{end+1} = '151244';\n",
    "model.metCHEBIID{end+1} = '80625';\n",
    "model.metPdMap{end+1} = '';\n",
    "model.metReconMap{end+1} = '';\n",
    "model.rxnCOG{end+1} = '';\n",
    "model.rxnKeggOrthology{end+1} = '';\n",
    "model.rxnReconMap{end+1} = '';\n",
    "\n",
    "% fbal[c]\n",
    "model.S(end+1,:) = zeros(1,length(model.rxns));\n",
    "model.mets{end+1} = 'fbal[c]';\n",
    "model.b(end+1) = 0;\n",
    "model.csense(end+1) = 'E';\n",
    "model.metCharges(end+1) = 0;\n",
    "model.metFormulas{end+1} = 'C3H6FNO2';\n",
    "model.metSmiles{end+1} = 'C(CNF)C(=O)O';\n",
    "model.metNames{end+1} = 'Fluoro-beta-alanine';\n",
    "model.metHMDBID{end+1} = '';\n",
    "model.metInChIString{end+1} = 'InChI=1S/C3H6FNO2/c4-5-2-1-3(6)7/h5H,1-2H2,(H,6,7)';\n",
    "model.metKEGGID{end+1} = '';\n",
    "model.metPubChemID{end+1} = '22665129';\n",
    "model.metCHEBIID{end+1} = '';\n",
    "model.metPdMap{end+1} = '';\n",
    "model.metReconMap{end+1} = '';\n",
    "\n",
    "% fbal[e]\n",
    "model.S(end+1,:) = zeros(1,length(model.rxns));\n",
    "model.mets{end+1} = 'fbal[e]';\n",
    "model.b(end+1) = 0;\n",
    "model.csense(end+1) = 'E';\n",
    "model.metCharges(end+1) = 0;\n",
    "model.metFormulas{end+1} = 'C3H6FNO2';\n",
    "model.metSmiles{end+1} = 'C(CNF)C(=O)O';\n",
    "model.metNames{end+1} = 'Fluoro-beta-alanine';\n",
    "model.metHMDBID{end+1} = '';\n",
    "model.metInChIString{end+1} = 'InChI=1S/C3H6FNO2/c4-5-2-1-3(6)7/h5H,1-2H2,(H,6,7)';\n",
    "model.metKEGGID{end+1} = '';\n",
    "model.metPubChemID{end+1} = '22665129';\n",
    "model.metCHEBIID{end+1} = '';\n",
    "model.metPdMap{end+1} = '';\n",
    "model.metReconMap{end+1} = '';"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Reactions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "% 5fu[e] <=>\n",
    "model.S(:,end+1) = zeros(length(model.mets),1);\n",
    "model.S(strcmp(model.mets,'5fu[e]'),end) = -1;\n",
    "model.rxns{end+1} = 'EX_5fu[e]';\n",
    "model.lb(end+1) = min(model.lb);\n",
    "model.ub(end+1) = max(model.ub);\n",
    "model.c(end+1) = 0;\n",
    "model.rules{end+1} = '';\n",
    "model.grRules{end+1} = '';\n",
    "model.rxnGeneMat(end+1,:) = zeros(1,length(model.genes));\n",
    "model.rxnConfidenceScores(end+1) = 0;\n",
    "model.rxnNames{end+1} = 'Exchange of 5-Fluorouracil';\n",
    "model.rxnNotes{end+1} = '';\n",
    "model.rxnECNumbers{end+1} = '';\n",
    "model.rxnReferences{end+1} = '';\n",
    "model.rxnKEGGID{end+1} = '';\n",
    "model.subSystems{end+1} = 'MODULE 5-FLUOROURACIL';\n",
    "model.rxnCOG{end+1} = '';\n",
    "model.rxnKeggOrthology{end+1} = '';\n",
    "model.rxnReconMap{end+1} = '';\n",
    "model.rxnFormulas{end+1} = '1 5fu[e] <=> ';\n",
    "\n",
    "% 5fu[e] --> 5fu[c]\n",
    "model.S(:,end+1) = zeros(length(model.mets),1);\n",
    "model.S(strcmp(model.mets,'5fu[e]'),end) = -1;\n",
    "model.S(strcmp(model.mets,'5fu[c]'),end) = 1;\n",
    "model.rxns{end+1} = '5fu_transport_e_c';\n",
    "model.lb(end+1) = 0;\n",
    "model.ub(end+1) = max(model.ub);\n",
    "model.c(end+1) = 0;\n",
    "model.rules{end+1} = 'x(666) | x(155)';\n",
    "model.grRules{end+1} = '(2030) or (10864)';\n",
    "model.rxnGeneMat(end+1,:) = zeros(1,length(model.genes));\n",
    "model.rxnGeneMat(end,666) = 1;\n",
    "model.rxnGeneMat(end,155) = 1;\n",
    "model.rxnConfidenceScores(end+1) = 0;\n",
    "model.rxnNames{end+1} = 'Transport of 5-Fluorouracil from extracellular to cytosol';\n",
    "model.rxnNotes{end+1} = '';\n",
    "model.rxnECNumbers{end+1} = 'X.X.X.X';\n",
    "model.rxnReferences{end+1} = '';\n",
    "model.rxnKEGGID{end+1} = '';\n",
    "model.subSystems{end+1} = 'MODULE 5-FLUOROURACIL';\n",
    "model.rxnCOG{end+1} = '';\n",
    "model.rxnKeggOrthology{end+1} = '';\n",
    "model.rxnReconMap{end+1} = '';\n",
    "model.rxnFormulas{end+1} = '1 5fu[e] --> 1 5fu[c]';\n",
    "\n",
    "% 5fu[c] --> 5fu[e]\n",
    "model.S(:,end+1) = zeros(length(model.mets),1);\n",
    "model.S(strcmp(model.mets,'5fu[c]'),end) = -1;\n",
    "model.S(strcmp(model.mets,'atp[c]'),end) = -1;\n",
    "model.S(strcmp(model.mets,'h2o[c]'),end) = -1;\n",
    "model.S(strcmp(model.mets,'5fu[e]'),end) = 1;\n",
    "model.S(strcmp(model.mets,'adp[c]'),end) = 1;\n",
    "model.S(strcmp(model.mets,'pi[c]'),end) = 1;\n",
    "model.rxns{end+1} = '5fu_transport_c_e';\n",
    "model.lb(end+1) = 0;\n",
    "model.ub(end+1) = max(model.ub);\n",
    "model.c(end+1) = 0;\n",
    "model.rules{end+1} = 'x(2996) | x(62) | x(20) | x(2943) | x(3174)';\n",
    "model.grRules{end+1} = '(8714) or (10257) or (10057) or (85320) or (9429)';\n",
    "model.rxnGeneMat(end+1,:) = zeros(1,length(model.genes));\n",
    "model.rxnGeneMat(end,2996) = 1;\n",
    "model.rxnGeneMat(end,62) = 1;\n",
    "model.rxnGeneMat(end,20) = 1;\n",
    "model.rxnGeneMat(end,2943) = 1;\n",
    "model.rxnGeneMat(end,3174) = 1;\n",
    "model.rxnConfidenceScores(end+1) = 0;\n",
    "model.rxnNames{end+1} = 'Transport of 5-Fluorouracil from cytosol to extracellular';\n",
    "model.rxnNotes{end+1} = '';\n",
    "model.rxnECNumbers{end+1} = 'X.X.X.X';\n",
    "model.rxnReferences{end+1} = '';\n",
    "model.rxnKEGGID{end+1} = '';\n",
    "model.subSystems{end+1} = 'MODULE 5-FLUOROURACIL';\n",
    "model.rxnCOG{end+1} = '';\n",
    "model.rxnKeggOrthology{end+1} = '';\n",
    "model.rxnReconMap{end+1} = '';\n",
    "model.rxnFormulas{end+1} = '1 5fu[c] + 1 atp[c] + 1 h2o[c] --> 1 5fu[e] + 1 adp[c] + 1 pi[c]';\n",
    "\n",
    "% 5fu[c] <=> fdurd[c]\n",
    "model.S(:,end+1) = zeros(length(model.mets),1);\n",
    "model.S(strcmp(model.mets,'5fu[c]'),end) = -1;\n",
    "model.S(strcmp(model.mets,'2dr1p[c]'),end) = -1;\n",
    "model.S(strcmp(model.mets,'fdurd[c]'),end) = 1;\n",
    "model.S(strcmp(model.mets,'pi[c]'),end) = 1;\n",
    "model.rxns{end+1} = '5fu_fdurd';\n",
    "model.lb(end+1) = min(model.lb);\n",
    "model.ub(end+1) = max(model.ub);\n",
    "model.c(end+1) = 0;\n",
    "model.rules{end+1} = 'x(630)';\n",
    "model.grRules{end+1} = '(1890)';\n",
    "model.rxnGeneMat(end+1,:) = zeros(1,length(model.genes));\n",
    "model.rxnGeneMat(end,630) = 1;\n",
    "model.rxnConfidenceScores(end+1) = 0;\n",
    "model.rxnNames{end+1} = '5-Fluorouracil to Fluorodeoxyuridine';\n",
    "model.rxnNotes{end+1} = '';\n",
    "model.rxnECNumbers{end+1} = '2.4.2.2 | 2.4.2.4';\n",
    "model.rxnReferences{end+1} = '';\n",
    "model.rxnKEGGID{end+1} = '';\n",
    "model.subSystems{end+1} = 'MODULE 5-FLUOROURACIL';\n",
    "model.rxnCOG{end+1} = '';\n",
    "model.rxnKeggOrthology{end+1} = '';\n",
    "model.rxnReconMap{end+1} = '';\n",
    "model.rxnFormulas{end+1} = '1 5fu[c] + 1 2dr1p[c] <=> 1 fdurd[c] + 1 pi[c]';\n",
    "\n",
    "% fdurd[c] --> fdump[c]\n",
    "model.S(:,end+1) = zeros(length(model.mets),1);\n",
    "model.S(strcmp(model.mets,'fdurd[c]'),end) = -1;\n",
    "model.S(strcmp(model.mets,'atp[c]'),end) = -1;\n",
    "model.S(strcmp(model.mets,'fdump[c]'),end) = 1;\n",
    "model.S(strcmp(model.mets,'adp[c]'),end) = 1;\n",
    "model.S(strcmp(model.mets,'h[c]'),end) = 1;\n",
    "model.rxns{end+1} = 'fdurd_fdump';\n",
    "model.lb(end+1) = 0;\n",
    "model.ub(end+1) = max(model.ub);\n",
    "model.c(end+1) = 0;\n",
    "model.rules{end+1} = 'x(2601)';\n",
    "model.grRules{end+1} = '(7083)';\n",
    "model.rxnGeneMat(end+1,:) = zeros(1,length(model.genes));\n",
    "model.rxnGeneMat(end,2601) = 1;\n",
    "model.rxnConfidenceScores(end+1) = 0;\n",
    "model.rxnNames{end+1} = 'Fluorodeoxyuridine to Fluorodeoxyuridine monophosphate';\n",
    "model.rxnNotes{end+1} = '';\n",
    "model.rxnECNumbers{end+1} = '2.7.1.21 | 2.7.1.145';\n",
    "model.rxnReferences{end+1} = '';\n",
    "model.rxnKEGGID{end+1} = '';\n",
    "model.subSystems{end+1} = 'MODULE 5-FLUOROURACIL';\n",
    "model.rxnCOG{end+1} = '';\n",
    "model.rxnKeggOrthology{end+1} = '';\n",
    "model.rxnReconMap{end+1} = '';\n",
    "model.rxnFormulas{end+1} = '1 fdurd[c] + 1 atp[c] --> 1 fdump[c] + 1 adp[c] + 1 h[c]';\n",
    "\n",
    "% fdudp[c] <=> fdump[c]\n",
    "model.S(:,end+1) = zeros(length(model.mets),1);\n",
    "model.S(strcmp(model.mets,'fdump[c]'),end) = 1;\n",
    "model.S(strcmp(model.mets,'atp[c]'),end) = 1;\n",
    "model.S(strcmp(model.mets,'fdudp[c]'),end) = -1;\n",
    "model.S(strcmp(model.mets,'adp[c]'),end) = -1;\n",
    "model.rxns{end+1} = 'fdudp_fdump';\n",
    "model.lb(end+1) = min(model.lb);\n",
    "model.ub(end+1) = max(model.ub);\n",
    "model.c(end+1) = 0;\n",
    "model.rules{end+1} = '';\n",
    "model.grRules{end+1} = '';\n",
    "model.rxnGeneMat(end+1,:) = zeros(1,length(model.genes));\n",
    "model.rxnConfidenceScores(end+1) = 0;\n",
    "model.rxnNames{end+1} = 'Fluorodeoxyuridine diphosphate to Fluorodeoxyuridine monophosphate';\n",
    "model.rxnNotes{end+1} = '';\n",
    "model.rxnECNumbers{end+1} = '2.7.4.4 | 2.7.4.9';\n",
    "model.rxnReferences{end+1} = '';\n",
    "model.rxnKEGGID{end+1} = '';\n",
    "model.subSystems{end+1} = 'MODULE 5-FLUOROURACIL';\n",
    "model.rxnCOG{end+1} = '';\n",
    "model.rxnKeggOrthology{end+1} = '';\n",
    "model.rxnReconMap{end+1} = '';\n",
    "model.rxnFormulas{end+1} = '1 fdudp[c] + 1 adp[c] <=> 1 fdump[c] + 1 atp[c]';\n",
    "\n",
    "% 5fu[c] <=> furd[c]\n",
    "model.S(:,end+1) = zeros(length(model.mets),1);\n",
    "model.S(strcmp(model.mets,'5fu[c]'),end) = -1;\n",
    "model.S(strcmp(model.mets,'r1p[c]'),end) = -1;\n",
    "model.S(strcmp(model.mets,'furd[c]'),end) = 1;\n",
    "model.S(strcmp(model.mets,'pi[c]'),end) = 1;\n",
    "model.rxns{end+1} = '5fu_furd';\n",
    "model.lb(end+1) = min(model.lb);\n",
    "model.ub(end+1) = max(model.ub);\n",
    "model.c(end+1) = 0;\n",
    "model.rules{end+1} = 'x(2661) | x(478)';\n",
    "model.grRules{end+1} = '(7378) or (151531)';\n",
    "model.rxnGeneMat(end+1,:) = zeros(1,length(model.genes));\n",
    "model.rxnGeneMat(end,2661) = 1;\n",
    "model.rxnGeneMat(end,478) = 1;\n",
    "model.rxnConfidenceScores(end+1) = 0;\n",
    "model.rxnNames{end+1} = '5-Fluorouracil to Fluorouridine';\n",
    "model.rxnNotes{end+1} = '';\n",
    "model.rxnECNumbers{end+1} = '2.4.2.2 | 2.4.2.3';\n",
    "model.rxnReferences{end+1} = '';\n",
    "model.rxnKEGGID{end+1} = '';\n",
    "model.subSystems{end+1} = 'MODULE 5-FLUOROURACIL';\n",
    "model.rxnCOG{end+1} = '';\n",
    "model.rxnKeggOrthology{end+1} = '';\n",
    "model.rxnReconMap{end+1} = '';\n",
    "model.rxnFormulas{end+1} = '1 5fu[c] + 1 r1p[c] <=> 1 furd[c] + 1 pi[c]';\n",
    "\n",
    "% furd[c] --> fump[c]\n",
    "model.S(:,end+1) = zeros(length(model.mets),1);\n",
    "model.S(strcmp(model.mets,'furd[c]'),end) = -1;\n",
    "model.S(strcmp(model.mets,'atp[c]'),end) = -1;\n",
    "model.S(strcmp(model.mets,'fump[c]'),end) = 1;\n",
    "model.S(strcmp(model.mets,'adp[c]'),end) = 1;\n",
    "model.S(strcmp(model.mets,'h[c]'),end) = 1;\n",
    "model.rxns{end+1} = 'furd_fump';\n",
    "model.lb(end+1) = 0;\n",
    "model.ub(end+1) = max(model.ub);\n",
    "model.c(end+1) = 0;\n",
    "model.rules{end+1} = 'x(2839) | x(2658) | x(1987)';\n",
    "model.grRules{end+1} = '(83549) or (7371) or (54963)';\n",
    "model.rxnGeneMat(end+1,:) = zeros(1,length(model.genes));\n",
    "model.rxnGeneMat(end,2839) = 1;\n",
    "model.rxnGeneMat(end,2658) = 1;\n",
    "model.rxnGeneMat(end,1987) = 1;\n",
    "model.rxnConfidenceScores(end+1) = 0;\n",
    "model.rxnNames{end+1} = 'Fluorouridine to Fluorouridine monophosphate';\n",
    "model.rxnNotes{end+1} = '';\n",
    "model.rxnECNumbers{end+1} = '2.7.1.48 | 2.7.1.213';\n",
    "model.rxnReferences{end+1} = '';\n",
    "model.rxnKEGGID{end+1} = '';\n",
    "model.subSystems{end+1} = 'MODULE 5-FLUOROURACIL';\n",
    "model.rxnCOG{end+1} = '';\n",
    "model.rxnKeggOrthology{end+1} = '';\n",
    "model.rxnReconMap{end+1} = '';\n",
    "model.rxnFormulas{end+1} = '1 furd[c] + 1 atp[c] --> 1 fump[c] + 1 adp[c] + 1 h[c]';\n",
    "\n",
    "% 5fu[c] <=> fump[c]\n",
    "model.S(:,end+1) = zeros(length(model.mets),1);\n",
    "model.S(strcmp(model.mets,'5fu[c]'),end) = -1;\n",
    "model.S(strcmp(model.mets,'prpp[c]'),end) = -1;\n",
    "model.S(strcmp(model.mets,'fump[c]'),end) = 1;\n",
    "model.S(strcmp(model.mets,'ppi[c]'),end) = 1;\n",
    "model.rxns{end+1} = '5fu_fump';\n",
    "model.lb(end+1) = min(model.lb);\n",
    "model.ub(end+1) = max(model.ub);\n",
    "model.c(end+1) = 0;\n",
    "model.rules{end+1} = 'x(1960) | x(2659)';\n",
    "model.grRules{end+1} = '(5471) or (7372)';\n",
    "model.rxnGeneMat(end+1,:) = zeros(1,length(model.genes));\n",
    "model.rxnGeneMat(end,1960) = 1;\n",
    "model.rxnGeneMat(end,2659) = 1;\n",
    "model.rxnConfidenceScores(end+1) = 0;\n",
    "model.rxnNames{end+1} = '5-Fluorouracil to Fluorouridine monophosphate';\n",
    "model.rxnNotes{end+1} = '';\n",
    "model.rxnECNumbers{end+1} = '2.4.2.9';\n",
    "model.rxnReferences{end+1} = '';\n",
    "model.rxnKEGGID{end+1} = '';\n",
    "model.subSystems{end+1} = 'MODULE 5-FLUOROURACIL';\n",
    "model.rxnCOG{end+1} = '';\n",
    "model.rxnKeggOrthology{end+1} = '';\n",
    "model.rxnReconMap{end+1} = '';\n",
    "model.rxnFormulas{end+1} = '1 5fu[c] + 1 prpp[c] <=> 1 fump[c] + 1 ppi[c]';\n",
    "\n",
    "% fump[c] <=> fudp[c]\n",
    "model.S(:,end+1) = zeros(length(model.mets),1);\n",
    "model.S(strcmp(model.mets,'fump[c]'),end) = -1;\n",
    "model.S(strcmp(model.mets,'atp[c]'),end) = -1;\n",
    "model.S(strcmp(model.mets,'fudp[c]'),end) = 1;\n",
    "model.S(strcmp(model.mets,'adp[c]'),end) = 1;\n",
    "model.rxns{end+1} = 'fump_fudp';\n",
    "model.lb(end+1) = min(model.lb);\n",
    "model.ub(end+1) = max(model.ub);\n",
    "model.c(end+1) = 0;\n",
    "model.rules{end+1} = '';\n",
    "model.grRules{end+1} = '';\n",
    "model.rxnGeneMat(end+1,:) = zeros(1,length(model.genes));\n",
    "model.rxnConfidenceScores(end+1) = 0;\n",
    "model.rxnNames{end+1} = 'Fluorouridine monophosphate to Fluorouridine diphosphate';\n",
    "model.rxnNotes{end+1} = '';\n",
    "model.rxnECNumbers{end+1} = '2.7.4.4 | 2.7.4.14 | 2.7.4.22';\n",
    "model.rxnReferences{end+1} = '';\n",
    "model.rxnKEGGID{end+1} = '';\n",
    "model.subSystems{end+1} = 'MODULE 5-FLUOROURACIL';\n",
    "model.rxnCOG{end+1} = '';\n",
    "model.rxnKeggOrthology{end+1} = '';\n",
    "model.rxnReconMap{end+1} = '';\n",
    "model.rxnFormulas{end+1} = '1 fump[c] + 1 atp[c] <=> 1 fudp[c] + 1 adp[c]';\n",
    "\n",
    "% fudp[c] --> fdudp[c]\n",
    "model.S(:,end+1) = zeros(length(model.mets),1);\n",
    "model.S(strcmp(model.mets,'fudp[c]'),end) = -1;\n",
    "model.S(strcmp(model.mets,'trdrd[c]'),end) = -1;\n",
    "model.S(strcmp(model.mets,'fdudp[c]'),end) = 1;\n",
    "model.S(strcmp(model.mets,'trdox[c]'),end) = 1;\n",
    "model.S(strcmp(model.mets,'h2o[c]'),end) = 1;\n",
    "model.rxns{end+1} = 'fudp_fdudp';\n",
    "model.lb(end+1) = 0;\n",
    "model.ub(end+1) = max(model.ub);\n",
    "model.c(end+1) = 0;\n",
    "model.rules{end+1} = 'x(2348) & (x(2349) | x(1635))';\n",
    "model.grRules{end+1} = '(6240) and ((6241) or (50484))';\n",
    "model.rxnGeneMat(end+1,:) = zeros(1,length(model.genes));\n",
    "model.rxnGeneMat(end,2348) = 1;\n",
    "model.rxnGeneMat(end,2349) = 1;\n",
    "model.rxnGeneMat(end,1635) = 1;\n",
    "model.rxnConfidenceScores(end+1) = 0;\n",
    "model.rxnNames{end+1} = 'Fluorouridine diphosphate to Fluorodeoxyuridine diphosphate';\n",
    "model.rxnNotes{end+1} = '';\n",
    "model.rxnECNumbers{end+1} = '1.17.4.1';\n",
    "model.rxnReferences{end+1} = '';\n",
    "model.rxnKEGGID{end+1} = '';\n",
    "model.subSystems{end+1} = 'MODULE 5-FLUOROURACIL';\n",
    "model.rxnCOG{end+1} = '';\n",
    "model.rxnKeggOrthology{end+1} = '';\n",
    "model.rxnReconMap{end+1} = '';\n",
    "model.rxnFormulas{end+1} = '1 fudp[c] + 1 trdrd[c] --> 1 fdudp[c] + 1 trdox[c] + 1 h2o[c]';\n",
    "\n",
    "% 5fu[c] --> dhfu[c]\n",
    "model.S(:,end+1) = zeros(length(model.mets),1);\n",
    "model.S(strcmp(model.mets,'5fu[c]'),end) = -1;\n",
    "model.S(strcmp(model.mets,'nadph[c]'),end) = -1;\n",
    "model.S(strcmp(model.mets,'h[c]'),end) = -1;\n",
    "model.S(strcmp(model.mets,'dhfu[c]'),end) = 1;\n",
    "model.S(strcmp(model.mets,'nadp[c]'),end) = 1;\n",
    "model.rxns{end+1} = '5fu_dhfu';\n",
    "model.lb(end+1) = 0;\n",
    "model.ub(end+1) = max(model.ub);\n",
    "model.c(end+1) = 0;\n",
    "model.rules{end+1} = 'x(610)';\n",
    "model.grRules{end+1} = '(1806)';\n",
    "model.rxnGeneMat(end+1,:) = zeros(1,length(model.genes));\n",
    "model.rxnGeneMat(end,610) = 1;\n",
    "model.rxnConfidenceScores(end+1) = 0;\n",
    "model.rxnNames{end+1} = '5-Fluorouracil to 5,6-Dihydro-5-fluorouracil';\n",
    "model.rxnNotes{end+1} = '';\n",
    "model.rxnECNumbers{end+1} = '1.3.1.2';\n",
    "model.rxnReferences{end+1} = '';\n",
    "model.rxnKEGGID{end+1} = '';\n",
    "model.subSystems{end+1} = 'MODULE 5-FLUOROURACIL';\n",
    "model.rxnCOG{end+1} = '';\n",
    "model.rxnKeggOrthology{end+1} = '';\n",
    "model.rxnReconMap{end+1} = '';\n",
    "model.rxnFormulas{end+1} = '1 5fu[c] + 1 nadph[c] + 1 h[c] --> 1 dhfu[c] + 1 nadp[c]';\n",
    "\n",
    "% dhfu[c] --> fupa[c]\n",
    "model.S(:,end+1) = zeros(length(model.mets),1);\n",
    "model.S(strcmp(model.mets,'dhfu[c]'),end) = -1;\n",
    "model.S(strcmp(model.mets,'h2o[c]'),end) = -1;\n",
    "model.S(strcmp(model.mets,'fupa[c]'),end) = 1;\n",
    "model.S(strcmp(model.mets,'h[c]'),end) = 1;\n",
    "model.rxns{end+1} = 'dhfu_fupa';\n",
    "model.lb(end+1) = 0;\n",
    "model.ub(end+1) = max(model.ub);\n",
    "model.c(end+1) = 0;\n",
    "model.rules{end+1} = 'x(611) | x(612) | x(613)';\n",
    "model.grRules{end+1} = '(1807) or (1808) or (1809)';\n",
    "model.rxnGeneMat(end+1,:) = zeros(1,length(model.genes));\n",
    "model.rxnGeneMat(end,611) = 1;\n",
    "model.rxnGeneMat(end,612) = 1;\n",
    "model.rxnGeneMat(end,613) = 1;\n",
    "model.rxnConfidenceScores(end+1) = 0;\n",
    "model.rxnNames{end+1} = '5,6-Dihydro-5-fluorouracil to alpha-fluoro-ureidopropionic acid';\n",
    "model.rxnNotes{end+1} = '';\n",
    "model.rxnECNumbers{end+1} = '3.5.2.2';\n",
    "model.rxnReferences{end+1} = '';\n",
    "model.rxnKEGGID{end+1} = '';\n",
    "model.subSystems{end+1} = 'MODULE 5-FLUOROURACIL';\n",
    "model.rxnCOG{end+1} = '';\n",
    "model.rxnKeggOrthology{end+1} = '';\n",
    "model.rxnReconMap{end+1} = '';\n",
    "model.rxnFormulas{end+1} = '1 dhfu[c] + 1 h2o[c] --> 1 fupa[c] + 1 h[c]';\n",
    "\n",
    "% fupa[c] --> fbal[c]\n",
    "model.S(:,end+1) = zeros(length(model.mets),1);\n",
    "model.S(strcmp(model.mets,'fupa[c]'),end) = -1;\n",
    "model.S(strcmp(model.mets,'h[c]'),end) = -2;\n",
    "model.S(strcmp(model.mets,'h2o[c]'),end) = -1;\n",
    "model.S(strcmp(model.mets,'fbal[c]'),end) = 1;\n",
    "model.S(strcmp(model.mets,'co2[c]'),end) = 1;\n",
    "model.S(strcmp(model.mets,'nh4[c]'),end) = 1;\n",
    "model.rxns{end+1} = 'fupa_fbal';\n",
    "model.lb(end+1) = 0;\n",
    "model.ub(end+1) = max(model.ub);\n",
    "model.c(end+1) = 0;\n",
    "model.rules{end+1} = 'x(1793)';\n",
    "model.grRules{end+1} = '(51733)';\n",
    "model.rxnGeneMat(end+1,:) = zeros(1,length(model.genes));\n",
    "model.rxnGeneMat(end,1793) = 1;\n",
    "model.rxnConfidenceScores(end+1) = 0;\n",
    "model.rxnNames{end+1} = 'Alpha-fluoro-ureidopropionic acid to fluoro-beta-alanine';\n",
    "model.rxnNotes{end+1} = '';\n",
    "model.rxnECNumbers{end+1} = '3.5.1.6';\n",
    "model.rxnReferences{end+1} = '';\n",
    "model.rxnKEGGID{end+1} = '';\n",
    "model.subSystems{end+1} = 'MODULE 5-FLUOROURACIL';\n",
    "model.rxnCOG{end+1} = '';\n",
    "model.rxnKeggOrthology{end+1} = '';\n",
    "model.rxnReconMap{end+1} = '';\n",
    "model.rxnFormulas{end+1} = '1 fupa[c] + 2 h[c] + 1 h2o[c] --> 1 fbal[c] + 1 co2[c] + 1 nh4[c]';\n",
    "\n",
    "% fbal[c] <=> fbal[e]\n",
    "model.S(:,end+1) = zeros(length(model.mets),1);\n",
    "model.S(strcmp(model.mets,'fbal[e]'),end) = 1;\n",
    "model.S(strcmp(model.mets,'fbal[c]'),end) = -1;\n",
    "model.rxns{end+1} = 'fbal_transport_c_e';\n",
    "model.lb(end+1) = min(model.lb);\n",
    "model.ub(end+1) = max(model.ub);\n",
    "model.c(end+1) = 0;\n",
    "model.rules{end+1} = '';\n",
    "model.grRules{end+1} = '';\n",
    "model.rxnGeneMat(end+1,:) = zeros(1,length(model.genes));\n",
    "model.rxnConfidenceScores(end+1) = 0;\n",
    "model.rxnNames{end+1} = 'Transport of Fluoro-beta-alanine from cytosol to extracellular';\n",
    "model.rxnNotes{end+1} = '';\n",
    "model.rxnECNumbers{end+1} = '';\n",
    "model.rxnReferences{end+1} = '';\n",
    "model.rxnKEGGID{end+1} = '';\n",
    "model.subSystems{end+1} = 'MODULE 5-FLUOROURACIL';\n",
    "model.rxnCOG{end+1} = '';\n",
    "model.rxnKeggOrthology{end+1} = '';\n",
    "model.rxnReconMap{end+1} = '';\n",
    "model.rxnFormulas{end+1} = '1 fbal[c] <=> 1 fbal[e]';\n",
    "\n",
    "% fbal[e] <=>\n",
    "model.S(:,end+1) = zeros(length(model.mets),1);\n",
    "model.S(strcmp(model.mets,'fbal[e]'),end) = -1;\n",
    "model.rxns{end+1} = 'EX_fbal[e]';\n",
    "model.lb(end+1) = min(model.lb);\n",
    "model.ub(end+1) = max(model.ub);\n",
    "model.c(end+1) = 0;\n",
    "model.rules{end+1} = '';\n",
    "model.grRules{end+1} = '';\n",
    "model.rxnGeneMat(end+1,:) = zeros(1,length(model.genes));\n",
    "model.rxnConfidenceScores(end+1) = 0;\n",
    "model.rxnNames{end+1} = 'Exchange of Fluoro-beta-alanine';\n",
    "model.rxnNotes{end+1} = '';\n",
    "model.rxnECNumbers{end+1} = '';\n",
    "model.rxnReferences{end+1} = '';\n",
    "model.rxnKEGGID{end+1} = '';\n",
    "model.subSystems{end+1} = 'MODULE 5-FLUOROURACIL';\n",
    "model.rxnCOG{end+1} = '';\n",
    "model.rxnKeggOrthology{end+1} = '';\n",
    "model.rxnReconMap{end+1} = '';\n",
    "model.rxnFormulas{end+1} = '1 fbal[e] <=> ';\n",
    "\n",
    "% artificial fdump[c] sink\n",
    "model.S(:,end+1) = zeros(length(model.mets),1);\n",
    "model.S(strcmp(model.mets,'fdump[c]'),end) = -1;\n",
    "model.rxns{end+1} = 'SINK_fdump[c]';\n",
    "model.lb(end+1) = 0;\n",
    "model.ub(end+1) = max(model.ub);\n",
    "model.c(end+1) = 0;\n",
    "model.rules{end+1} = '';\n",
    "model.grRules{end+1} = '';\n",
    "model.rxnGeneMat(end+1,:) = zeros(1,length(model.genes));\n",
    "model.rxnConfidenceScores(end+1) = 0;\n",
    "model.rxnNames{end+1} = 'Artificial sink on Fluorodeoxyuridine monophosphate';\n",
    "model.rxnNotes{end+1} = '';\n",
    "model.rxnECNumbers{end+1} = '';\n",
    "model.rxnReferences{end+1} = '';\n",
    "model.rxnKEGGID{end+1} = '';\n",
    "model.subSystems{end+1} = 'MODULE 5-FLUOROURACIL';\n",
    "model.rxnCOG{end+1} = '';\n",
    "model.rxnKeggOrthology{end+1} = '';\n",
    "model.rxnReconMap{end+1} = '';\n",
    "model.rxnFormulas{end+1} = '1 fdump[c] --> ';"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Doxorubicin"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Metabolites"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 151,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n"
     ]
    }
   ],
   "source": [
    "% doxQ[e]\n",
    "model.S(end+1,:) = zeros(1,length(model.rxns));\n",
    "model.mets{end+1} = 'doxQ[e]';\n",
    "model.b(end+1) = 0;\n",
    "model.csense(end+1) = 'E';\n",
    "model.metCharges(end+1) = 0;\n",
    "model.metFormulas{end+1} = 'C27H29NO11';\n",
    "model.metSmiles{end+1} = 'CC1C(C(CC(O1)OC2CC(CC3=C2C(=C4C(=C3O)C(=O)C5=C(C4=O)C(=CC=C5)OC)O)(C(=O)CO)O)N)O';\n",
    "model.metNames{end+1} = 'Doxorubicin';\n",
    "model.metHMDBID{end+1} = 'HMDB0015132';\n",
    "model.metInChIString{end+1} = 'InChI=1S/C27H29NO11/c1-10-22(31)13(28)6-17(38-10)39-15-8-27(36,16(30)9-29)7-12-19(15)26(35)21-20(24(12)33)23(32)11-4-3-5-14(37-2)18(11)25(21)34/h3-5,10,13,15,17,22,29,31,33,35-36H,6-9,28H2,1-2H3/t10-,13-,15-,17-,22+,27-/m0/s1';\n",
    "model.metKEGGID{end+1} = '';\n",
    "model.metPubChemID{end+1} = '31703';\n",
    "model.metCHEBIID{end+1} = '28748';\n",
    "model.metPdMap{end+1} = '';\n",
    "model.metReconMap{end+1} = '';\n",
    "\n",
    "% doxQ[c]\n",
    "model.S(end+1,:) = zeros(1,length(model.rxns));\n",
    "model.mets{end+1} = 'doxQ[c]';\n",
    "model.b(end+1) = 0;\n",
    "model.csense(end+1) = 'E';\n",
    "model.metCharges(end+1) = 0;\n",
    "model.metFormulas{end+1} = 'C27H29NO11';\n",
    "model.metSmiles{end+1} = 'CC1C(C(CC(O1)OC2CC(CC3=C2C(=C4C(=C3O)C(=O)C5=C(C4=O)C(=CC=C5)OC)O)(C(=O)CO)O)N)O';\n",
    "model.metNames{end+1} = 'Doxorubicin';\n",
    "model.metHMDBID{end+1} = 'HMDB0015132';\n",
    "model.metInChIString{end+1} = 'InChI=1S/C27H29NO11/c1-10-22(31)13(28)6-17(38-10)39-15-8-27(36,16(30)9-29)7-12-19(15)26(35)21-20(24(12)33)23(32)11-4-3-5-14(37-2)18(11)25(21)34/h3-5,10,13,15,17,22,29,31,33,35-36H,6-9,28H2,1-2H3/t10-,13-,15-,17-,22+,27-/m0/s1';\n",
    "model.metKEGGID{end+1} = '';\n",
    "model.metPubChemID{end+1} = '31703';\n",
    "model.metCHEBIID{end+1} = '28748';\n",
    "model.metPdMap{end+1} = '';\n",
    "model.metReconMap{end+1} = '';\n",
    "\n",
    "% doxSQ[c]\n",
    "model.S(end+1,:) = zeros(1,length(model.rxns));\n",
    "model.mets{end+1} = 'doxSQ[c]';\n",
    "model.b(end+1) = 0;\n",
    "model.csense(end+1) = 'E';\n",
    "model.metCharges(end+1) = 0;\n",
    "model.metFormulas{end+1} = 'C27H30NO11';\n",
    "model.metSmiles{end+1} = 'CC1C(C(CC(O1)OC2CC(CC3=C2C(=C4C(C5=C(C=CC=C5OC)C(C4=C3O)O)O)O)(C(=O)CO)O)N)O';\n",
    "model.metNames{end+1} = 'Doxorubicin semiquinone';\n",
    "model.metHMDBID{end+1} = 'HMDB0060822';\n",
    "model.metInChIString{end+1} = 'InChI=1S/C27H33NO11/c1-10-22(31)13(28)6-17(38-10)39-15-8-27(36,16(30)9-29)7-12-19(15)26(35)21-20(24(12)33)23(32)11-4-3-5-14(37-2)18(11)25(21)34/h3-5,10,13,15,17,22-23,25,29,31-36H,6-9,28H2,1-2H3';\n",
    "model.metKEGGID{end+1} = '';\n",
    "model.metPubChemID{end+1} = '131769967';\n",
    "model.metCHEBIID{end+1} = '';\n",
    "model.metPdMap{end+1} = '';\n",
    "model.metReconMap{end+1} = '';\n",
    "\n",
    "% doxHQ[c]\n",
    "model.S(end+1,:) = zeros(1,length(model.rxns));\n",
    "model.mets{end+1} = 'doxHQ[c]';\n",
    "model.b(end+1) = 0;\n",
    "model.csense(end+1) = 'E';\n",
    "model.metCharges(end+1) = 0;\n",
    "model.metFormulas{end+1} = 'C27H31NO11';\n",
    "model.metSmiles{end+1} = 'CC1C(C(CC(O1)OC2CC(CC3=C2C(=C4C(=C3O)C(=O)C5=C(C4=O)C(=CC=C5)OC)O)(C(CO)O)O)N)O';\n",
    "model.metNames{end+1} = 'Doxorubicinol';\n",
    "model.metHMDBID{end+1} = 'HMDB0041884';\n",
    "model.metInChIString{end+1} = 'InChI=1S/C27H31NO11/c1-10-22(31)13(28)6-17(38-10)39-15-8-27(36,16(30)9-29)7-12-19(15)26(35)21-20(24(12)33)23(32)11-4-3-5-14(37-2)18(11)25(21)34/h3-5,10,13,15-17,22,29-31,33,35-36H,6-9,28H2,1-2H3/t10-,13-,15-,16-,17-,22+,27-/m0/s1';\n",
    "model.metKEGGID{end+1} = '';\n",
    "model.metPubChemID{end+1} = '83970';\n",
    "model.metCHEBIID{end+1} = '133817';\n",
    "model.metPdMap{end+1} = '';\n",
    "model.metReconMap{end+1} = '';\n",
    "\n",
    "% doxHQ[e]\n",
    "model.S(end+1,:) = zeros(1,length(model.rxns));\n",
    "model.mets{end+1} = 'doxHQ[e]';\n",
    "model.b(end+1) = 0;\n",
    "model.csense(end+1) = 'E';\n",
    "model.metCharges(end+1) = 0;\n",
    "model.metFormulas{end+1} = 'C27H31NO11';\n",
    "model.metSmiles{end+1} = 'CC1C(C(CC(O1)OC2CC(CC3=C2C(=C4C(=C3O)C(=O)C5=C(C4=O)C(=CC=C5)OC)O)(C(CO)O)O)N)O';\n",
    "model.metNames{end+1} = 'Doxorubicinol';\n",
    "model.metHMDBID{end+1} = 'HMDB0041884';\n",
    "model.metInChIString{end+1} = 'InChI=1S/C27H31NO11/c1-10-22(31)13(28)6-17(38-10)39-15-8-27(36,16(30)9-29)7-12-19(15)26(35)21-20(24(12)33)23(32)11-4-3-5-14(37-2)18(11)25(21)34/h3-5,10,13,15-17,22,29-31,33,35-36H,6-9,28H2,1-2H3/t10-,13-,15-,16-,17-,22+,27-/m0/s1';\n",
    "model.metKEGGID{end+1} = '';\n",
    "model.metPubChemID{end+1} = '83970';\n",
    "model.metCHEBIID{end+1} = '133817';\n",
    "model.metPdMap{end+1} = '';\n",
    "model.metReconMap{end+1} = '';\n",
    "\n",
    "% doxdeoxy[c]\n",
    "model.S(end+1,:) = zeros(1,length(model.rxns));\n",
    "model.mets{end+1} = 'doxdeoxy[c]';\n",
    "model.b(end+1) = 0;\n",
    "model.csense(end+1) = 'E';\n",
    "model.metCharges(end+1) = 0;\n",
    "model.metFormulas{end+1} = 'C21H18O8';\n",
    "model.metSmiles{end+1} = 'COC1=CC=CC2=C1C(=O)C3=C(C4=C(CC(CC4)(C(=O)CO)O)C(=C3C2=O)O)O';\n",
    "model.metNames{end+1} = '7-Deoxydoxorubicinone';\n",
    "model.metHMDBID{end+1} = '';\n",
    "model.metInChIString{end+1} = 'InChI=1S/C21H18O8/c1-29-12-4-2-3-10-14(12)20(27)16-15(18(10)25)19(26)11-7-21(28,13(23)8-22)6-5-9(11)17(16)24/h2-4,22,24,26,28H,5-8H2,1H3/t21-/m1/s1';\n",
    "model.metKEGGID{end+1} = '';\n",
    "model.metPubChemID{end+1} = '83958';\n",
    "model.metCHEBIID{end+1} = '';\n",
    "model.metPdMap{end+1} = '';\n",
    "model.metReconMap{end+1} = '';\n",
    "\n",
    "% doxdeoxy[e]\n",
    "model.S(end+1,:) = zeros(1,length(model.rxns));\n",
    "model.mets{end+1} = 'doxdeoxy[e]';\n",
    "model.b(end+1) = 0;\n",
    "model.csense(end+1) = 'E';\n",
    "model.metCharges(end+1) = 0;\n",
    "model.metFormulas{end+1} = 'C21H18O8';\n",
    "model.metSmiles{end+1} = 'COC1=CC=CC2=C1C(=O)C3=C(C4=C(CC(CC4)(C(=O)CO)O)C(=C3C2=O)O)O';\n",
    "model.metNames{end+1} = '7-Deoxydoxorubicinone';\n",
    "model.metHMDBID{end+1} = '';\n",
    "model.metInChIString{end+1} = 'InChI=1S/C21H18O8/c1-29-12-4-2-3-10-14(12)20(27)16-15(18(10)25)19(26)11-7-21(28,13(23)8-22)6-5-9(11)17(16)24/h2-4,22,24,26,28H,5-8H2,1H3/t21-/m1/s1';\n",
    "model.metKEGGID{end+1} = '';\n",
    "model.metPubChemID{end+1} = '83958';\n",
    "model.metCHEBIID{end+1} = '';\n",
    "model.metPdMap{end+1} = '';\n",
    "model.metReconMap{end+1} = '';\n",
    "\n",
    "% daun[c]\n",
    "model.S(end+1,:) = zeros(1,length(model.rxns));\n",
    "model.mets{end+1} = 'daun[c]';\n",
    "model.b(end+1) = 0;\n",
    "model.csense(end+1) = 'E';\n",
    "model.metCharges(end+1) = 0;\n",
    "model.metFormulas{end+1} = 'C6H13NO3';\n",
    "model.metSmiles{end+1} = 'CC(C(C(CC=O)N)O)O';\n",
    "model.metNames{end+1} = 'Daunosamine';\n",
    "model.metHMDBID{end+1} = '';\n",
    "model.metInChIString{end+1} = 'InChI=1S/C6H13NO3/c1-4(9)6(10)5(7)2-3-8/h3-6,9-10H,2,7H2,1H3/t4-,5-,6+/m0/s1';\n",
    "model.metKEGGID{end+1} = '';\n",
    "model.metPubChemID{end+1} = '160128';\n",
    "model.metCHEBIID{end+1} = '';\n",
    "model.metPdMap{end+1} = '';\n",
    "model.metReconMap{end+1} = '';\n",
    "\n",
    "% daun[e]\n",
    "model.S(end+1,:) = zeros(1,length(model.rxns));\n",
    "model.mets{end+1} = 'daun[e]';\n",
    "model.b(end+1) = 0;\n",
    "model.csense(end+1) = 'E';\n",
    "model.metCharges(end+1) = 0;\n",
    "model.metFormulas{end+1} = 'C6H13NO3';\n",
    "model.metSmiles{end+1} = 'CC(C(C(CC=O)N)O)O';\n",
    "model.metNames{end+1} = 'Daunosamine';\n",
    "model.metHMDBID{end+1} = '';\n",
    "model.metInChIString{end+1} = 'InChI=1S/C6H13NO3/c1-4(9)6(10)5(7)2-3-8/h3-6,9-10H,2,7H2,1H3/t4-,5-,6+/m0/s1';\n",
    "model.metKEGGID{end+1} = '';\n",
    "model.metPubChemID{end+1} = '160128';\n",
    "model.metCHEBIID{end+1} = '';\n",
    "model.metPdMap{end+1} = '';\n",
    "model.metReconMap{end+1} = '';"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Reactions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 152,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n"
     ]
    }
   ],
   "source": [
    "% % % import metabolites\n",
    "\n",
    "% doxQ[e] <=>\n",
    "model.S(:,end+1) = zeros(length(model.mets),1);\n",
    "model.S(strcmp(model.mets,'doxQ[e]'),end) = -1;\n",
    "model.rxns{end+1} = 'EX_doxQ[e]';\n",
    "model.lb(end+1) = min(model.lb);\n",
    "model.ub(end+1) = max(model.ub);\n",
    "model.c(end+1) = 0;\n",
    "model.rules{end+1} = '';\n",
    "model.grRules{end+1} = '';\n",
    "model.rxnGeneMat(end+1,:) = zeros(1,length(model.genes));\n",
    "model.rxnConfidenceScores(end+1) = 0;\n",
    "model.rxnNames{end+1} = 'Exchange of Doxorubicin';\n",
    "model.rxnNotes{end+1} = '';\n",
    "model.rxnECNumbers{end+1} = '';\n",
    "model.rxnReferences{end+1} = '';\n",
    "model.rxnKEGGID{end+1} = '';\n",
    "model.subSystems{end+1} = 'MODULE DOXORUBICIN';\n",
    "model.rxnCOG{end+1} = '';\n",
    "model.rxnKeggOrthology{end+1} = '';\n",
    "model.rxnReconMap{end+1} = '';\n",
    "model.rxnFormulas{end+1} = '1 doxQ[e] <=> ';\n",
    "\n",
    "% doxHQ[e] <=>\n",
    "model.S(:,end+1) = zeros(length(model.mets),1);\n",
    "model.S(strcmp(model.mets,'doxHQ[e]'),end) = -1;\n",
    "model.rxns{end+1} = 'EX_doxHQ[e]';\n",
    "model.lb(end+1) = min(model.lb);\n",
    "model.ub(end+1) = max(model.ub);\n",
    "model.c(end+1) = 0;\n",
    "model.rules{end+1} = '';\n",
    "model.grRules{end+1} = '';\n",
    "model.rxnGeneMat(end+1,:) = zeros(1,length(model.genes));\n",
    "model.rxnConfidenceScores(end+1) = 0;\n",
    "model.rxnNames{end+1} = 'Exchange of Doxorubicinol';\n",
    "model.rxnNotes{end+1} = '';\n",
    "model.rxnECNumbers{end+1} = '';\n",
    "model.rxnReferences{end+1} = '';\n",
    "model.rxnKEGGID{end+1} = '';\n",
    "model.subSystems{end+1} = 'MODULE DOXORUBICIN';\n",
    "model.rxnCOG{end+1} = '';\n",
    "model.rxnKeggOrthology{end+1} = '';\n",
    "model.rxnReconMap{end+1} = '';\n",
    "model.rxnFormulas{end+1} = '1 doxHQ[e] <=> ';\n",
    "\n",
    "% doxdeoxy[e] <=>\n",
    "model.S(:,end+1) = zeros(length(model.mets),1);\n",
    "model.S(strcmp(model.mets,'doxdeoxy[e]'),end) = -1;\n",
    "model.rxns{end+1} = 'EX_doxdeoxy[e]';\n",
    "model.lb(end+1) = min(model.lb);\n",
    "model.ub(end+1) = max(model.ub);\n",
    "model.c(end+1) = 0;\n",
    "model.rules{end+1} = '';\n",
    "model.grRules{end+1} = '';\n",
    "model.rxnGeneMat(end+1,:) = zeros(1,length(model.genes));\n",
    "model.rxnConfidenceScores(end+1) = 0;\n",
    "model.rxnNames{end+1} = 'Exchange of 7-Deoxydoxorubicinone';\n",
    "model.rxnNotes{end+1} = '';\n",
    "model.rxnECNumbers{end+1} = '';\n",
    "model.rxnReferences{end+1} = '';\n",
    "model.rxnKEGGID{end+1} = '';\n",
    "model.subSystems{end+1} = 'MODULE DOXORUBICIN';\n",
    "model.rxnCOG{end+1} = '';\n",
    "model.rxnKeggOrthology{end+1} = '';\n",
    "model.rxnReconMap{end+1} = '';\n",
    "model.rxnFormulas{end+1} = '1 doxdeoxy[e] <=> ';\n",
    "\n",
    "% daun[e] <=>\n",
    "model.S(:,end+1) = zeros(length(model.mets),1);\n",
    "model.S(strcmp(model.mets,'daun[e]'),end) = -1;\n",
    "model.rxns{end+1} = 'EX_daun[e]';\n",
    "model.lb(end+1) = min(model.lb);\n",
    "model.ub(end+1) = max(model.ub);\n",
    "model.c(end+1) = 0;\n",
    "model.rules{end+1} = '';\n",
    "model.grRules{end+1} = '';\n",
    "model.rxnGeneMat(end+1,:) = zeros(1,length(model.genes));\n",
    "model.rxnConfidenceScores(end+1) = 0;\n",
    "model.rxnNames{end+1} = 'Exchange of Daunosamine';\n",
    "model.rxnNotes{end+1} = '';\n",
    "model.rxnECNumbers{end+1} = '';\n",
    "model.rxnReferences{end+1} = '';\n",
    "model.rxnKEGGID{end+1} = '';\n",
    "model.subSystems{end+1} = 'MODULE DOXORUBICIN';\n",
    "model.rxnCOG{end+1} = '';\n",
    "model.rxnKeggOrthology{end+1} = '';\n",
    "model.rxnReconMap{end+1} = '';\n",
    "model.rxnFormulas{end+1} = '1 daun[e] <=> ';\n",
    "\n",
    "% % % transport\n",
    "\n",
    "% doxQ[e] --> doxQ[c]\n",
    "model.S(:,end+1) = zeros(length(model.mets),1);\n",
    "model.S(strcmp(model.mets,'doxQ[e]'),end) = -1;\n",
    "model.S(strcmp(model.mets,'doxQ[c]'),end) = 1;\n",
    "model.rxns{end+1} = 'doxQ_transport_e_c';\n",
    "model.lb(end+1) = 0;\n",
    "model.ub(end+1) = max(model.ub);\n",
    "model.c(end+1) = 0;\n",
    "model.rules{end+1} = 'x(2949)';\n",
    "model.grRules{end+1} = '(85413)';\n",
    "model.rxnGeneMat(end+1,:) = zeros(1,length(model.genes));\n",
    "model.rxnGeneMat(end,2949) = 1;\n",
    "model.rxnConfidenceScores(end+1) = 0;\n",
    "model.rxnNames{end+1} = 'Transport of Doxorubicin from extracellular to cytosol';\n",
    "model.rxnNotes{end+1} = '';\n",
    "model.rxnECNumbers{end+1} = 'X.X.X.X';\n",
    "model.rxnReferences{end+1} = '';\n",
    "model.rxnKEGGID{end+1} = '';\n",
    "model.subSystems{end+1} = 'MODULE DOXORUBICIN';\n",
    "model.rxnCOG{end+1} = '';\n",
    "model.rxnKeggOrthology{end+1} = '';\n",
    "model.rxnReconMap{end+1} = '';\n",
    "model.rxnFormulas{end+1} = '1 doxQ[e] --> 1 doxQ[c]';\n",
    "\n",
    "% doxQ[c] --> doxQ[e], no ATP\n",
    "model.S(:,end+1) = zeros(length(model.mets),1);\n",
    "model.S(strcmp(model.mets,'doxQ[c]'),end) = -1;\n",
    "model.S(strcmp(model.mets,'doxQ[e]'),end) = 1;\n",
    "model.rxns{end+1} = 'doxQ_transport_c_e_noATP';\n",
    "model.lb(end+1) = 0;\n",
    "model.ub(end+1) = max(model.ub);\n",
    "model.c(end+1) = 0;\n",
    "model.rules{end+1} = 'x(167)';\n",
    "model.grRules{end+1} = '(10928)';\n",
    "model.rxnGeneMat(end+1,:) = zeros(1,length(model.genes));\n",
    "model.rxnGeneMat(end,167) = 1;\n",
    "model.rxnConfidenceScores(end+1) = 0;\n",
    "model.rxnNames{end+1} = 'Transport of Doxorubicin from cytosol to extracellular, no ATP';\n",
    "model.rxnNotes{end+1} = '';\n",
    "model.rxnECNumbers{end+1} = 'X.X.X.X';\n",
    "model.rxnReferences{end+1} = '';\n",
    "model.rxnKEGGID{end+1} = '';\n",
    "model.subSystems{end+1} = 'MODULE DOXORUBICIN';\n",
    "model.rxnCOG{end+1} = '';\n",
    "model.rxnKeggOrthology{end+1} = '';\n",
    "model.rxnReconMap{end+1} = '';\n",
    "model.rxnFormulas{end+1} = '1 doxQ[c] --> 1 doxQ[e]';\n",
    "\n",
    "% doxQ[c] --> doxQ[e], with ATP\n",
    "model.S(:,end+1) = zeros(length(model.mets),1);\n",
    "model.S(strcmp(model.mets,'doxQ[c]'),end) = -1;\n",
    "model.S(strcmp(model.mets,'atp[c]'),end) = -1;\n",
    "model.S(strcmp(model.mets,'h2o[c]'),end) = -1;\n",
    "model.S(strcmp(model.mets,'doxQ[e]'),end) = 1;\n",
    "model.S(strcmp(model.mets,'adp[c]'),end) = 1;\n",
    "model.S(strcmp(model.mets,'pi[c]'),end) = 1;\n",
    "model.rxns{end+1} = 'doxQ_transport_c_e_withATP';\n",
    "model.lb(end+1) = 0;\n",
    "model.ub(end+1) = max(model.ub);\n",
    "model.c(end+1) = 0;\n",
    "model.rules{end+1} = 'x(1823) | x(1499) | x(331) | x(62) | x(3174)';\n",
    "model.grRules{end+1} = '(5243) or (4363) or (1244) or (10257) or (9429)';\n",
    "model.rxnGeneMat(end+1,:) = zeros(1,length(model.genes));\n",
    "model.rxnGeneMat(end,1823) = 1;\n",
    "model.rxnGeneMat(end,1499) = 1;\n",
    "model.rxnGeneMat(end,331) = 1;\n",
    "model.rxnGeneMat(end,62) = 1;\n",
    "model.rxnGeneMat(end,3174) = 1;\n",
    "model.rxnConfidenceScores(end+1) = 0;\n",
    "model.rxnNames{end+1} = 'Transport of Doxorubicin from cytosol to extracellular, with ATP';\n",
    "model.rxnNotes{end+1} = '';\n",
    "model.rxnECNumbers{end+1} = 'X.X.X.X';\n",
    "model.rxnReferences{end+1} = '';\n",
    "model.rxnKEGGID{end+1} = '';\n",
    "model.subSystems{end+1} = 'MODULE DOXORUBICIN';\n",
    "model.rxnCOG{end+1} = '';\n",
    "model.rxnKeggOrthology{end+1} = '';\n",
    "model.rxnReconMap{end+1} = '';\n",
    "model.rxnFormulas{end+1} = '1 doxQ[c] + 1 atp[c] + 1 h2o[c] --> 1 doxQ[e] + 1 adp[c] + 1 pi[c]';\n",
    "\n",
    "% doxHQ[c] <=> doxHQ[e]\n",
    "model.S(:,end+1) = zeros(length(model.mets),1);\n",
    "model.S(strcmp(model.mets,'doxHQ[e]'),end) = 1;\n",
    "model.S(strcmp(model.mets,'doxHQ[c]'),end) = -1;\n",
    "model.rxns{end+1} = 'doxHQ_transport_c_e';\n",
    "model.lb(end+1) = min(model.lb);\n",
    "model.ub(end+1) = max(model.ub);\n",
    "model.c(end+1) = 0;\n",
    "model.rules{end+1} = '';\n",
    "model.grRules{end+1} = '';\n",
    "model.rxnGeneMat(end+1,:) = zeros(1,length(model.genes));\n",
    "model.rxnConfidenceScores(end+1) = 0;\n",
    "model.rxnNames{end+1} = 'Transport of Doxorubicinol from cytosol to extracellular';\n",
    "model.rxnNotes{end+1} = '';\n",
    "model.rxnECNumbers{end+1} = '';\n",
    "model.rxnReferences{end+1} = '';\n",
    "model.rxnKEGGID{end+1} = '';\n",
    "model.subSystems{end+1} = 'MODULE DOXORUBICIN';\n",
    "model.rxnCOG{end+1} = '';\n",
    "model.rxnKeggOrthology{end+1} = '';\n",
    "model.rxnReconMap{end+1} = '';\n",
    "model.rxnFormulas{end+1} = '1 doxHQ[c] <=> 1 doxHQ[e]';\n",
    "\n",
    "% doxdeoxy[c] <=> doxdeoxy[e]\n",
    "model.S(:,end+1) = zeros(length(model.mets),1);\n",
    "model.S(strcmp(model.mets,'doxdeoxy[e]'),end) = 1;\n",
    "model.S(strcmp(model.mets,'doxdeoxy[c]'),end) = -1;\n",
    "model.rxns{end+1} = 'doxdeoxy_transport_c_e';\n",
    "model.lb(end+1) = min(model.lb);\n",
    "model.ub(end+1) = max(model.ub);\n",
    "model.c(end+1) = 0;\n",
    "model.rules{end+1} = '';\n",
    "model.grRules{end+1} = '';\n",
    "model.rxnGeneMat(end+1,:) = zeros(1,length(model.genes));\n",
    "model.rxnConfidenceScores(end+1) = 0;\n",
    "model.rxnNames{end+1} = 'Transport of 7-Deoxydoxorubicinone from cytosol to extracellular';\n",
    "model.rxnNotes{end+1} = '';\n",
    "model.rxnECNumbers{end+1} = '';\n",
    "model.rxnReferences{end+1} = '';\n",
    "model.rxnKEGGID{end+1} = '';\n",
    "model.subSystems{end+1} = 'MODULE DOXORUBICIN';\n",
    "model.rxnCOG{end+1} = '';\n",
    "model.rxnKeggOrthology{end+1} = '';\n",
    "model.rxnReconMap{end+1} = '';\n",
    "model.rxnFormulas{end+1} = '1 doxdeoxy[c] <=> 1 doxdeoxy[e]';\n",
    "\n",
    "% daun[c] <=> daun[e]\n",
    "model.S(:,end+1) = zeros(length(model.mets),1);\n",
    "model.S(strcmp(model.mets,'daun[e]'),end) = 1;\n",
    "model.S(strcmp(model.mets,'daun[c]'),end) = -1;\n",
    "model.rxns{end+1} = 'daun_transport_c_e';\n",
    "model.lb(end+1) = min(model.lb);\n",
    "model.ub(end+1) = max(model.ub);\n",
    "model.c(end+1) = 0;\n",
    "model.rules{end+1} = '';\n",
    "model.grRules{end+1} = '';\n",
    "model.rxnGeneMat(end+1,:) = zeros(1,length(model.genes));\n",
    "model.rxnConfidenceScores(end+1) = 0;\n",
    "model.rxnNames{end+1} = 'Transport of Daunosamine from cytosol to extracellular';\n",
    "model.rxnNotes{end+1} = '';\n",
    "model.rxnECNumbers{end+1} = '';\n",
    "model.rxnReferences{end+1} = '';\n",
    "model.rxnKEGGID{end+1} = '';\n",
    "model.subSystems{end+1} = 'MODULE DOXORUBICIN';\n",
    "model.rxnCOG{end+1} = '';\n",
    "model.rxnKeggOrthology{end+1} = '';\n",
    "model.rxnReconMap{end+1} = '';\n",
    "model.rxnFormulas{end+1} = '1 daun[c] <=> 1 daun[e]';\n",
    "\n",
    "% % % doxQ and doxSQ\n",
    "\n",
    "% doxQ[c] --> doxSQ[c], NOS3\n",
    "model.S(:,end+1) = zeros(length(model.mets),1);\n",
    "model.S(strcmp(model.mets,'doxQ[c]'),end) = -1;\n",
    "model.S(strcmp(model.mets,'o2[c]'),end) = -2;\n",
    "model.S(strcmp(model.mets,'arg_L[c]'),end) = -1;\n",
    "model.S(strcmp(model.mets,'nadph[c]'),end) = -1.5;\n",
    "model.S(strcmp(model.mets,'h[c]'),end) = -1.5;\n",
    "model.S(strcmp(model.mets,'doxSQ[c]'),end) = 1;\n",
    "model.S(strcmp(model.mets,'h2o[c]'),end) = 2;\n",
    "model.S(strcmp(model.mets,'no[c]'),end) = 1;\n",
    "model.S(strcmp(model.mets,'citr_L[c]'),end) = 1;\n",
    "model.S(strcmp(model.mets,'nadp[c]'),end) = 1.5;\n",
    "model.rxns{end+1} = 'doxQ_doxSQ_NOS3';\n",
    "model.lb(end+1) = 0;\n",
    "model.ub(end+1) = max(model.ub);\n",
    "model.c(end+1) = 0;\n",
    "model.rules{end+1} = 'x(1598)';\n",
    "model.grRules{end+1} = '(4846)';\n",
    "model.rxnGeneMat(end+1,:) = zeros(1,length(model.genes));\n",
    "model.rxnGeneMat(end,1598) = 1;\n",
    "model.rxnConfidenceScores(end+1) = 0;\n",
    "model.rxnNames{end+1} = 'Doxorubicin to doxorubicin semiquinone, NOS3';\n",
    "model.rxnNotes{end+1} = '';\n",
    "model.rxnECNumbers{end+1} = '1.14.13.39';\n",
    "model.rxnReferences{end+1} = '';\n",
    "model.rxnKEGGID{end+1} = '';\n",
    "model.subSystems{end+1} = 'MODULE DOXORUBICIN';\n",
    "model.rxnCOG{end+1} = '';\n",
    "model.rxnKeggOrthology{end+1} = '';\n",
    "model.rxnReconMap{end+1} = '';\n",
    "model.rxnFormulas{end+1} = '1 doxQ[c] + 2 o2[c] + 1 arg_L[c] + 1.5 nadph[c] + 1.5 h[c] --> 1 doxSQ[c] + 2 h2o[c] + 1 no[c] + 1 citr_L[c] + 1.5 nadp[c]';\n",
    "\n",
    "% doxQ[c] --> doxSQ[c], POR\n",
    "model.S(:,end+1) = zeros(length(model.mets),1);\n",
    "model.S(strcmp(model.mets,'doxQ[c]'),end) = -1;\n",
    "model.S(strcmp(model.mets,'nadph[c]'),end) = -0.5;\n",
    "model.S(strcmp(model.mets,'h[c]'),end) = -0.5;\n",
    "model.S(strcmp(model.mets,'doxSQ[c]'),end) = 1;\n",
    "model.S(strcmp(model.mets,'nadp[c]'),end) = 0.5;\n",
    "model.rxns{end+1} = 'doxQ_doxSQ_POR';\n",
    "model.lb(end+1) = 0;\n",
    "model.ub(end+1) = max(model.ub);\n",
    "model.c(end+1) = 0;\n",
    "model.rules{end+1} = 'x(1933)';\n",
    "model.grRules{end+1} = '(5447)';\n",
    "model.rxnGeneMat(end+1,:) = zeros(1,length(model.genes));\n",
    "model.rxnGeneMat(end,1933) = 1;\n",
    "model.rxnConfidenceScores(end+1) = 0;\n",
    "model.rxnNames{end+1} = 'Doxorubicin to doxorubicin semiquinone, POR';\n",
    "model.rxnNotes{end+1} = '';\n",
    "model.rxnECNumbers{end+1} = '1.6.2.4';\n",
    "model.rxnReferences{end+1} = '';\n",
    "model.rxnKEGGID{end+1} = '';\n",
    "model.subSystems{end+1} = 'MODULE DOXORUBICIN';\n",
    "model.rxnCOG{end+1} = '';\n",
    "model.rxnKeggOrthology{end+1} = '';\n",
    "model.rxnReconMap{end+1} = '';\n",
    "model.rxnFormulas{end+1} = '1 doxQ[c] + 0.5 nadph[c] + 0.5 h[c] --> 1 doxSQ[c] + 0.5 nadp[c]';\n",
    "\n",
    "% doxSQ[c] --> doxQ\n",
    "model.S(:,end+1) = zeros(length(model.mets),1);\n",
    "model.S(strcmp(model.mets,'doxSQ[c]'),end) = -1;\n",
    "model.S(strcmp(model.mets,'o2[c]'),end) = -1;\n",
    "model.S(strcmp(model.mets,'doxQ[c]'),end) = 1;\n",
    "model.S(strcmp(model.mets,'o2s[c]'),end) = 1;\n",
    "model.S(strcmp(model.mets,'h[c]'),end) = 1;\n",
    "model.rxns{end+1} = 'doxSQ_doxQ';\n",
    "model.lb(end+1) = 0;\n",
    "model.ub(end+1) = max(model.ub);\n",
    "model.c(end+1) = 0;\n",
    "model.rules{end+1} = '';\n",
    "model.grRules{end+1} = '';\n",
    "model.rxnGeneMat(end+1,:) = zeros(1,length(model.genes));\n",
    "model.rxnConfidenceScores(end+1) = 0;\n",
    "model.rxnNames{end+1} = 'Doxorubicin semiquinone to doxorubicin';\n",
    "model.rxnNotes{end+1} = '';\n",
    "model.rxnECNumbers{end+1} = '';\n",
    "model.rxnReferences{end+1} = '';\n",
    "model.rxnKEGGID{end+1} = '';\n",
    "model.subSystems{end+1} = 'MODULE DOXORUBICIN';\n",
    "model.rxnCOG{end+1} = '';\n",
    "model.rxnKeggOrthology{end+1} = '';\n",
    "model.rxnReconMap{end+1} = '';\n",
    "model.rxnFormulas{end+1} = '1 doxSQ[c] + 1 o2[c] --> 1 doxQ[c] + 1 o2s[c] + 1 h[c]';\n",
    "\n",
    "% % % doxQ metabolism\n",
    "\n",
    "% doxQ[c] --> doxHQ[c]\n",
    "model.S(:,end+1) = zeros(length(model.mets),1);\n",
    "model.S(strcmp(model.mets,'doxQ[c]'),end) = -1;\n",
    "model.S(strcmp(model.mets,'nadph[c]'),end) = -1;\n",
    "model.S(strcmp(model.mets,'h[c]'),end) = -1;\n",
    "model.S(strcmp(model.mets,'doxHQ[c]'),end) = 1;\n",
    "model.S(strcmp(model.mets,'nadp[c]'),end) = 1;\n",
    "model.rxns{end+1} = 'doxQ_doxHQ';\n",
    "model.lb(end+1) = 0;\n",
    "model.ub(end+1) = max(model.ub);\n",
    "model.c(end+1) = 0;\n",
    "model.rules{end+1} = 'x(2999) | x(3004) | x(74) | x(2978)';\n",
    "model.grRules{end+1} = '(873) or (874) or (10327) or (8644)';\n",
    "model.rxnGeneMat(end+1,:) = zeros(1,length(model.genes));\n",
    "model.rxnGeneMat(end,2999) = 1;\n",
    "model.rxnGeneMat(end,3004) = 1;\n",
    "model.rxnGeneMat(end,74) = 1;\n",
    "model.rxnGeneMat(end,2978) = 1;\n",
    "model.rxnConfidenceScores(end+1) = 0;\n",
    "model.rxnNames{end+1} = 'Doxorubicin to doxorubicinol';\n",
    "model.rxnNotes{end+1} = '';\n",
    "model.rxnECNumbers{end+1} = '1.1.1.2 | 1.1.1.184';\n",
    "model.rxnReferences{end+1} = '';\n",
    "model.rxnKEGGID{end+1} = '';\n",
    "model.subSystems{end+1} = 'MODULE DOXORUBICIN';\n",
    "model.rxnCOG{end+1} = '';\n",
    "model.rxnKeggOrthology{end+1} = '';\n",
    "model.rxnReconMap{end+1} = '';\n",
    "model.rxnFormulas{end+1} = '1 doxQ[c] + 1 nadph[c] + 1 h[c] --> 1 doxHQ[c] + 1 nadp[c]';\n",
    "\n",
    "% doxQ[c] --> doxdeoxy[c]\n",
    "model.S(:,end+1) = zeros(length(model.mets),1);\n",
    "model.S(strcmp(model.mets,'doxQ[c]'),end) = -1;\n",
    "model.S(strcmp(model.mets,'nadph[c]'),end) = -0.5;\n",
    "model.S(strcmp(model.mets,'h[c]'),end) = -1.5;\n",
    "model.S(strcmp(model.mets,'doxdeoxy[c]'),end) = 1;\n",
    "model.S(strcmp(model.mets,'daun[c]'),end) = 1;\n",
    "model.S(strcmp(model.mets,'nadp[c]'),end) = 0.5;\n",
    "model.rxns{end+1} = 'doxQ_doxdeoxy';\n",
    "model.lb(end+1) = 0;\n",
    "model.ub(end+1) = max(model.ub);\n",
    "model.c(end+1) = 0;\n",
    "model.rules{end+1} = 'x(1933)';\n",
    "model.grRules{end+1} = '(5447)';\n",
    "model.rxnGeneMat(end+1,:) = zeros(1,length(model.genes));\n",
    "model.rxnGeneMat(end,1933) = 1;\n",
    "model.rxnConfidenceScores(end+1) = 0;\n",
    "model.rxnNames{end+1} = 'Doxorubicin to 7-deoxydoxorubicinone';\n",
    "model.rxnNotes{end+1} = '';\n",
    "model.rxnECNumbers{end+1} = '1.6.2.4';\n",
    "model.rxnReferences{end+1} = '';\n",
    "model.rxnKEGGID{end+1} = '';\n",
    "model.subSystems{end+1} = 'MODULE DOXORUBICIN';\n",
    "model.rxnCOG{end+1} = '';\n",
    "model.rxnKeggOrthology{end+1} = '';\n",
    "model.rxnReconMap{end+1} = '';\n",
    "model.rxnFormulas{end+1} = '1 doxQ[c] + 0.5 nadph[c] + 1.5 h[c] --> 1 doxHQ[c] + 1 daun[c] + 0.5 nadp[c]';"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Cisplatin"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Metabolites"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 153,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n"
     ]
    }
   ],
   "source": [
    "% cis[e]\n",
    "model.S(end+1,:) = zeros(1,length(model.rxns));\n",
    "model.mets{end+1} = 'cis[e]';\n",
    "model.b(end+1) = 0;\n",
    "model.csense(end+1) = 'E';\n",
    "model.metCharges(end+1) = 0;\n",
    "model.metFormulas{end+1} = 'Cl2H6N2Pt';\n",
    "model.metSmiles{end+1} = 'N.N.Cl[Pt]Cl';\n",
    "model.metNames{end+1} = 'Cisplatin';\n",
    "model.metHMDBID{end+1} = '';\n",
    "model.metInChIString{end+1} = 'InChI=1S/2ClH.2H3N.Pt/h2*1H;2*1H3;/q;;;;+2/p-2';\n",
    "model.metKEGGID{end+1} = '';\n",
    "model.metPubChemID{end+1} = '5702198';\n",
    "model.metCHEBIID{end+1} = '';\n",
    "model.metPdMap{end+1} = '';\n",
    "model.metReconMap{end+1} = '';\n",
    "\n",
    "% cis[c]\n",
    "model.S(end+1,:) = zeros(1,length(model.rxns));\n",
    "model.mets{end+1} = 'cis[c]';\n",
    "model.b(end+1) = 0;\n",
    "model.csense(end+1) = 'E';\n",
    "model.metCharges(end+1) = 0;\n",
    "model.metFormulas{end+1} = 'Cl2H6N2Pt';\n",
    "model.metSmiles{end+1} = 'N.N.Cl[Pt]Cl';\n",
    "model.metNames{end+1} = 'Cisplatin';\n",
    "model.metHMDBID{end+1} = '';\n",
    "model.metInChIString{end+1} = 'InChI=1S/2ClH.2H3N.Pt/h2*1H;2*1H3;/q;;;;+2/p-2';\n",
    "model.metKEGGID{end+1} = '';\n",
    "model.metPubChemID{end+1} = '5702198';\n",
    "model.metCHEBIID{end+1} = '';\n",
    "model.metPdMap{end+1} = '';\n",
    "model.metReconMap{end+1} = '';\n",
    "\n",
    "% cis[m]\n",
    "model.S(end+1,:) = zeros(1,length(model.rxns));\n",
    "model.mets{end+1} = 'cis[m]';\n",
    "model.b(end+1) = 0;\n",
    "model.csense(end+1) = 'E';\n",
    "model.metCharges(end+1) = 0;\n",
    "model.metFormulas{end+1} = 'Cl2H6N2Pt';\n",
    "model.metSmiles{end+1} = 'N.N.Cl[Pt]Cl';\n",
    "model.metNames{end+1} = 'Cisplatin';m\n",
    "model.metHMDBID{end+1} = '';\n",
    "model.metInChIString{end+1} = 'InChI=1S/2ClH.2H3N.Pt/h2*1H;2*1H3;/q;;;;+2/p-2';\n",
    "model.metKEGGID{end+1} = '';\n",
    "model.metPubChemID{end+1} = '5702198';\n",
    "model.metCHEBIID{end+1} = '';\n",
    "model.metPdMap{end+1} = '';\n",
    "model.metReconMap{end+1} = '';\n",
    "\n",
    "% cis-mtDNA[m]\n",
    "model.S(end+1,:) = zeros(1,length(model.rxns));\n",
    "model.mets{end+1} = 'cis-mtDNA[m]';\n",
    "model.b(end+1) = 0;\n",
    "model.csense(end+1) = 'E';\n",
    "model.metCharges(end+1) = 0;\n",
    "model.metFormulas{end+1} = '';\n",
    "model.metSmiles{end+1} = '';\n",
    "model.metNames{end+1} = 'Cisplatin-mitochondrial DNA';\n",
    "model.metHMDBID{end+1} = '';\n",
    "model.metInChIString{end+1} = '';\n",
    "model.metKEGGID{end+1} = '';\n",
    "model.metPubChemID{end+1} = '';\n",
    "model.metCHEBIID{end+1} = '';\n",
    "model.metPdMap{end+1} = '';\n",
    "model.metReconMap{end+1} = '';\n",
    "\n",
    "% cisGSH[e]\n",
    "model.S(end+1,:) = zeros(1,length(model.rxns));\n",
    "model.mets{end+1} = 'cisGSH[e]';\n",
    "model.b(end+1) = 0;\n",
    "model.csense(end+1) = 'E';\n",
    "model.metCharges(end+1) = 0;\n",
    "model.metFormulas{end+1} = 'C10Cl2H21N5O6PtS';\n",
    "model.metSmiles{end+1} = '';\n",
    "model.metNames{end+1} = 'Cisplatin-GSH';\n",
    "model.metHMDBID{end+1} = '';\n",
    "model.metInChIString{end+1} = '';\n",
    "model.metKEGGID{end+1} = '';\n",
    "model.metPubChemID{end+1} = '';\n",
    "model.metCHEBIID{end+1} = '';\n",
    "model.metPdMap{end+1} = '';\n",
    "model.metReconMap{end+1} = '';\n",
    "\n",
    "% cisGSH[c]\n",
    "model.S(end+1,:) = zeros(1,length(model.rxns));\n",
    "model.mets{end+1} = 'cisGSH[c]';\n",
    "model.b(end+1) = 0;\n",
    "model.csense(end+1) = 'E';\n",
    "model.metCharges(end+1) = 0;\n",
    "model.metFormulas{end+1} = 'C10Cl2H21N5O6PtS';\n",
    "model.metSmiles{end+1} = '';\n",
    "model.metNames{end+1} = 'Cisplatin-GSH';\n",
    "model.metHMDBID{end+1} = '';\n",
    "model.metInChIString{end+1} = '';\n",
    "model.metKEGGID{end+1} = '';\n",
    "model.metPubChemID{end+1} = '';\n",
    "model.metCHEBIID{end+1} = '';\n",
    "model.metPdMap{end+1} = '';\n",
    "model.metReconMap{end+1} = '';\n",
    "\n",
    "% cisMT[e]\n",
    "model.S(end+1,:) = zeros(1,length(model.rxns));\n",
    "model.mets{end+1} = 'cisMT[e]';\n",
    "model.b(end+1) = 0;\n",
    "model.csense(end+1) = 'E';\n",
    "model.metCharges(end+1) = 0;\n",
    "model.metFormulas{end+1} = '';\n",
    "model.metSmiles{end+1} = '';\n",
    "model.metNames{end+1} = 'Cisplatin-metallothionein';\n",
    "model.metHMDBID{end+1} = '';\n",
    "model.metInChIString{end+1} = '';\n",
    "model.metKEGGID{end+1} = '';\n",
    "model.metPubChemID{end+1} = '';\n",
    "model.metCHEBIID{end+1} = '';\n",
    "model.metPdMap{end+1} = '';\n",
    "model.metReconMap{end+1} = '';\n",
    "\n",
    "% cisMT[c]\n",
    "model.S(end+1,:) = zeros(1,length(model.rxns));\n",
    "model.mets{end+1} = 'cisMT[c]';\n",
    "model.b(end+1) = 0;\n",
    "model.csense(end+1) = 'E';\n",
    "model.metCharges(end+1) = 0;\n",
    "model.metFormulas{end+1} = '';\n",
    "model.metSmiles{end+1} = '';\n",
    "model.metNames{end+1} = 'Cisplatin-metallothionein';\n",
    "model.metHMDBID{end+1} = '';\n",
    "model.metInChIString{end+1} = '';\n",
    "model.metKEGGID{end+1} = '';\n",
    "model.metPubChemID{end+1} = '';\n",
    "model.metCHEBIID{end+1} = '';\n",
    "model.metPdMap{end+1} = '';\n",
    "model.metReconMap{end+1} = '';"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Reactions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 154,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n"
     ]
    }
   ],
   "source": [
    "% % % import metabolites\n",
    "\n",
    "% cis[e] <=>\n",
    "model.S(:,end+1) = zeros(length(model.mets),1);\n",
    "model.S(strcmp(model.mets,'cis[e]'),end) = -1;\n",
    "model.rxns{end+1} = 'EX_cis[e]';\n",
    "model.lb(end+1) = min(model.lb);\n",
    "model.ub(end+1) = max(model.ub);\n",
    "model.c(end+1) = 0;\n",
    "model.rules{end+1} = '';\n",
    "model.grRules{end+1} = '';\n",
    "model.rxnGeneMat(end+1,:) = zeros(1,length(model.genes));\n",
    "model.rxnConfidenceScores(end+1) = 0;\n",
    "model.rxnNames{end+1} = 'Exchange of Cisplatin';\n",
    "model.rxnNotes{end+1} = '';\n",
    "model.rxnECNumbers{end+1} = '';\n",
    "model.rxnReferences{end+1} = '';\n",
    "model.rxnKEGGID{end+1} = '';\n",
    "model.subSystems{end+1} = 'MODULE CISPLATIN';\n",
    "model.rxnCOG{end+1} = '';\n",
    "model.rxnKeggOrthology{end+1} = '';\n",
    "model.rxnReconMap{end+1} = '';\n",
    "model.rxnFormulas{end+1} = '1 cis[e] <=> ';\n",
    "\n",
    "% cisGSH[e] <=>\n",
    "model.S(:,end+1) = zeros(length(model.mets),1);\n",
    "model.S(strcmp(model.mets,'cisGSH[e]'),end) = -1;\n",
    "model.rxns{end+1} = 'EX_cisGSH[e]';\n",
    "model.lb(end+1) = min(model.lb);\n",
    "model.ub(end+1) = max(model.ub);\n",
    "model.c(end+1) = 0;\n",
    "model.rules{end+1} = '';\n",
    "model.grRules{end+1} = '';\n",
    "model.rxnGeneMat(end+1,:) = zeros(1,length(model.genes));\n",
    "model.rxnConfidenceScores(end+1) = 0;\n",
    "model.rxnNames{end+1} = 'Exchange of Cisplatin-GSH';\n",
    "model.rxnNotes{end+1} = '';\n",
    "model.rxnECNumbers{end+1} = '';\n",
    "model.rxnReferences{end+1} = '';\n",
    "model.rxnKEGGID{end+1} = '';\n",
    "model.subSystems{end+1} = 'MODULE CISPLATIN';\n",
    "model.rxnCOG{end+1} = '';\n",
    "model.rxnKeggOrthology{end+1} = '';\n",
    "model.rxnReconMap{end+1} = '';\n",
    "model.rxnFormulas{end+1} = '1 cisGSH[e] <=> ';\n",
    "\n",
    "% cisMT[e] <=>\n",
    "model.S(:,end+1) = zeros(length(model.mets),1);\n",
    "model.S(strcmp(model.mets,'cisMT[e]'),end) = -1;\n",
    "model.rxns{end+1} = 'EX_cisMT[e]';\n",
    "model.lb(end+1) = min(model.lb);\n",
    "model.ub(end+1) = max(model.ub);\n",
    "model.c(end+1) = 0;\n",
    "model.rules{end+1} = '';\n",
    "model.grRules{end+1} = '';\n",
    "model.rxnGeneMat(end+1,:) = zeros(1,length(model.genes));\n",
    "model.rxnConfidenceScores(end+1) = 0;\n",
    "model.rxnNames{end+1} = 'Exchange of Cisplatin-metallothionein';\n",
    "model.rxnNotes{end+1} = '';\n",
    "model.rxnECNumbers{end+1} = '';\n",
    "model.rxnReferences{end+1} = '';\n",
    "model.rxnKEGGID{end+1} = '';\n",
    "model.subSystems{end+1} = 'MODULE CISPLATIN';\n",
    "model.rxnCOG{end+1} = '';\n",
    "model.rxnKeggOrthology{end+1} = '';\n",
    "model.rxnReconMap{end+1} = '';\n",
    "model.rxnFormulas{end+1} = '1 cisMT[e] <=> ';\n",
    "\n",
    "% % % transport\n",
    "\n",
    "% cis[e] --> cis[c]\n",
    "model.S(:,end+1) = zeros(length(model.mets),1);\n",
    "model.S(strcmp(model.mets,'cis[e]'),end) = -1;\n",
    "model.S(strcmp(model.mets,'cis[c]'),end) = 1;\n",
    "model.rxns{end+1} = 'cis_transport_e_c';\n",
    "model.lb(end+1) = 0;\n",
    "model.ub(end+1) = max(model.ub);\n",
    "model.c(end+1) = 0;\n",
    "model.rules{end+1} = 'x(375)';\n",
    "model.grRules{end+1} = '(1317)';\n",
    "model.rxnGeneMat(end+1,:) = zeros(1,length(model.genes));\n",
    "model.rxnGeneMat(end,375) = 1;\n",
    "model.rxnConfidenceScores(end+1) = 0;\n",
    "model.rxnNames{end+1} = 'Transport of Cisplatin from extracellular to cytosol';\n",
    "model.rxnNotes{end+1} = '';\n",
    "model.rxnECNumbers{end+1} = 'X.X.X.X';\n",
    "model.rxnReferences{end+1} = '';\n",
    "model.rxnKEGGID{end+1} = '';\n",
    "model.subSystems{end+1} = 'MODULE CISPLATIN';\n",
    "model.rxnCOG{end+1} = '';\n",
    "model.rxnKeggOrthology{end+1} = '';\n",
    "model.rxnReconMap{end+1} = '';\n",
    "model.rxnFormulas{end+1} = '1 cis[e] --> 1 cis[c]';\n",
    "\n",
    "% cis[c] --> cis[e]\n",
    "model.S(:,end+1) = zeros(length(model.mets),1);\n",
    "model.S(strcmp(model.mets,'cis[c]'),end) = -1;\n",
    "model.S(strcmp(model.mets,'atp[c]'),end) = -1;\n",
    "model.S(strcmp(model.mets,'h2o[c]'),end) = -1;\n",
    "model.S(strcmp(model.mets,'cis[e]'),end) = 1;\n",
    "model.S(strcmp(model.mets,'adp[c]'),end) = 1;\n",
    "model.S(strcmp(model.mets,'pi[c]'),end) = 1;\n",
    "model.rxns{end+1} = 'cis_transport_c_e';\n",
    "model.lb(end+1) = 0;\n",
    "model.ub(end+1) = max(model.ub);\n",
    "model.c(end+1) = 0;\n",
    "model.rules{end+1} = 'x(3174) | x(3269) | x(3270)';\n",
    "model.grRules{end+1} = '(9429) or (538) or (540)';\n",
    "model.rxnGeneMat(end+1,:) = zeros(1,length(model.genes));\n",
    "model.rxnGeneMat(end,3174) = 1;\n",
    "model.rxnGeneMat(:,(end+1):(end+2)) = zeros(length(model.rxns),2);\n",
    "model.rxnGeneMat(end,3269) = 1;\n",
    "model.rxnGeneMat(end,3270) = 1;\n",
    "model.genes{end+1} = '538';\n",
    "model.genes{end+1} = '540';\n",
    "model.geneSymbols{end+1} = 'ATP7A';\n",
    "model.geneSymbols{end+1} = 'ATP7B';\n",
    "model.rxnConfidenceScores(end+1) = 0;\n",
    "model.rxnNames{end+1} = 'Transport of Cisplatin from cytosol to extracellular';\n",
    "model.rxnNotes{end+1} = '';\n",
    "model.rxnECNumbers{end+1} = 'X.X.X.X';\n",
    "model.rxnReferences{end+1} = '';\n",
    "model.rxnKEGGID{end+1} = '';\n",
    "model.subSystems{end+1} = 'MODULE CISPLATIN';\n",
    "model.rxnCOG{end+1} = '';\n",
    "model.rxnKeggOrthology{end+1} = '';\n",
    "model.rxnReconMap{end+1} = '';\n",
    "model.rxnFormulas{end+1} = '1 cis[c] + 1 atp[c] + 1 h2o[c] --> 1 cis[e] + 1 adp[c] + 1 pi[c]';\n",
    "\n",
    "% cisGSH[c] --> cisGSH[e]\n",
    "model.S(:,end+1) = zeros(length(model.mets),1);\n",
    "model.S(strcmp(model.mets,'cisGSH[c]'),end) = -1;\n",
    "model.S(strcmp(model.mets,'atp[c]'),end) = -1;\n",
    "model.S(strcmp(model.mets,'h2o[c]'),end) = -1;\n",
    "model.S(strcmp(model.mets,'cisGSH[e]'),end) = 1;\n",
    "model.S(strcmp(model.mets,'adp[c]'),end) = 1;\n",
    "model.S(strcmp(model.mets,'pi[c]'),end) = 1;\n",
    "model.rxns{end+1} = 'cisGSH_transport_c_e';\n",
    "model.lb(end+1) = 0;\n",
    "model.ub(end+1) = max(model.ub);\n",
    "model.c(end+1) = 0;\n",
    "model.rules{end+1} = 'x(331)';\n",
    "model.grRules{end+1} = '(1244)';\n",
    "model.rxnGeneMat(end+1,:) = zeros(1,length(model.genes));\n",
    "model.rxnGeneMat(end,331) = 1;\n",
    "model.rxnConfidenceScores(end+1) = 0;\n",
    "model.rxnNames{end+1} = 'Transport of Cisplatin-GSH from cytosol to extracellular';\n",
    "model.rxnNotes{end+1} = '';\n",
    "model.rxnECNumbers{end+1} = 'X.X.X.X';\n",
    "model.rxnReferences{end+1} = '';\n",
    "model.rxnKEGGID{end+1} = '';\n",
    "model.subSystems{end+1} = 'MODULE CISPLATIN';\n",
    "model.rxnCOG{end+1} = '';\n",
    "model.rxnKeggOrthology{end+1} = '';\n",
    "model.rxnReconMap{end+1} = '';\n",
    "model.rxnFormulas{end+1} = '1 cisGSH[c] + 1 atp[c] + 1 h2o[c] --> 1 cisGSH[e] + 1 adp[c] + 1 pi[c]';\n",
    "\n",
    "% cisMT[c] <=> cisMT[e]\n",
    "model.S(:,end+1) = zeros(length(model.mets),1);\n",
    "model.S(strcmp(model.mets,'cisMT[e]'),end) = 1;\n",
    "model.S(strcmp(model.mets,'cisMT[c]'),end) = -1;\n",
    "model.rxns{end+1} = 'cisMT_transport_c_e';\n",
    "model.lb(end+1) = min(model.lb);\n",
    "model.ub(end+1) = max(model.ub);\n",
    "model.c(end+1) = 0;\n",
    "model.rules{end+1} = '';\n",
    "model.grRules{end+1} = '';\n",
    "model.rxnGeneMat(end+1,:) = zeros(1,length(model.genes));\n",
    "model.rxnConfidenceScores(end+1) = 0;\n",
    "model.rxnNames{end+1} = 'Transport of Cisplatin-metallothionein from cytosol to extracellular';\n",
    "model.rxnNotes{end+1} = '';\n",
    "model.rxnECNumbers{end+1} = '';\n",
    "model.rxnReferences{end+1} = '';\n",
    "model.rxnKEGGID{end+1} = '';\n",
    "model.subSystems{end+1} = 'MODULE CISPLATIN';\n",
    "model.rxnCOG{end+1} = '';\n",
    "model.rxnKeggOrthology{end+1} = '';\n",
    "model.rxnReconMap{end+1} = '';\n",
    "model.rxnFormulas{end+1} = '1 cisMT[c] <=> 1 cisMT[e]';\n",
    "\n",
    "% cis[c] --> cis[m]\n",
    "model.S(:,end+1) = zeros(length(model.mets),1);\n",
    "model.S(strcmp(model.mets,'cis[c]'),end) = -1;\n",
    "model.S(strcmp(model.mets,'cis[m]'),end) = 1;\n",
    "model.rxns{end+1} = 'cis_transport_c_m';\n",
    "model.lb(end+1) = 0;\n",
    "model.ub(end+1) = max(model.ub);\n",
    "model.c(end+1) = 0;\n",
    "model.rules{end+1} = 'x(3271)';\n",
    "model.grRules{end+1} = '(10063)';\n",
    "model.rxnGeneMat(end+1,:) = zeros(1,length(model.genes));\n",
    "model.rxnGeneMat(:,end+1) = zeros(length(model.rxns),1);\n",
    "model.rxnGeneMat(end,3271) = 1;\n",
    "model.genes{end+1} = '10063';\n",
    "model.geneSymbols{end+1} = 'COX17';\n",
    "model.rxnConfidenceScores(end+1) = 0;\n",
    "model.rxnNames{end+1} = 'Transport of Cisplatin from cytosol to mitochondrion';\n",
    "model.rxnNotes{end+1} = '';\n",
    "model.rxnECNumbers{end+1} = '1.9.3.1';\n",
    "model.rxnReferences{end+1} = '';\n",
    "model.rxnKEGGID{end+1} = '';\n",
    "model.subSystems{end+1} = 'MODULE CISPLATIN';\n",
    "model.rxnCOG{end+1} = '';\n",
    "model.rxnKeggOrthology{end+1} = '';\n",
    "model.rxnReconMap{end+1} = '';\n",
    "model.rxnFormulas{end+1} = '1 cis[c] --> 1 cis[m]';\n",
    "\n",
    "% % % cisplatin mechanism\n",
    "\n",
    "% cis[m] --> cis-mtDNA[m]\n",
    "model.S(:,end+1) = zeros(length(model.mets),1);\n",
    "model.S(strcmp(model.mets,'cis[m]'),end) = -1;\n",
    "model.S(strcmp(model.mets,'o2[m]'),end) = -0.5;\n",
    "model.S(strcmp(model.mets,'cis-mtDNA[m]'),end) = 1;\n",
    "model.S(strcmp(model.mets,'o2s[m]'),end) = 0.5;\n",
    "model.rxns{end+1} = 'cis_cis-mtDNA';\n",
    "model.lb(end+1) = 0;\n",
    "model.ub(end+1) = max(model.ub);\n",
    "model.c(end+1) = 0;\n",
    "model.rules{end+1} = 'x(1519) | x(1520) | x(1521) | x(1522) | x(1523) | x(1524) | x(1525) | x(1516)';\n",
    "model.grRules{end+1} = '(4535) or (4536) or (4537) or (4538) or (4539) or (4540) or (4541) or (4519)';\n",
    "model.rxnGeneMat(end+1,:) = zeros(1,length(model.genes));\n",
    "model.rxnGeneMat(end,1519) = 1;\n",
    "model.rxnGeneMat(end,1520) = 1;\n",
    "model.rxnGeneMat(end,1521) = 1;\n",
    "model.rxnGeneMat(end,1522) = 1;\n",
    "model.rxnGeneMat(end,1523) = 1;\n",
    "model.rxnGeneMat(end,1524) = 1;\n",
    "model.rxnGeneMat(end,1525) = 1;\n",
    "model.rxnGeneMat(end,1516) = 1;\n",
    "model.rxnConfidenceScores(end+1) = 0;\n",
    "model.rxnNames{end+1} = 'Cisplatin to cisplatin-mtDNA';\n",
    "model.rxnNotes{end+1} = '';\n",
    "model.rxnECNumbers{end+1} = '1.9.3.1';\n",
    "model.rxnReferences{end+1} = '';\n",
    "model.rxnKEGGID{end+1} = '';\n",
    "model.subSystems{end+1} = 'MODULE CISPLATIN';\n",
    "model.rxnCOG{end+1} = '';\n",
    "model.rxnKeggOrthology{end+1} = '';\n",
    "model.rxnReconMap{end+1} = '';\n",
    "model.rxnFormulas{end+1} = '1 cis[m] + 0.5 o2[m] --> 1 cis-mtDNA[m] + 0.5 o2s[m]';\n",
    "\n",
    "% cisplatin clearance\n",
    "\n",
    "% cis[c] --> cisGSH[c]\n",
    "model.S(:,end+1) = zeros(length(model.mets),1);\n",
    "model.S(strcmp(model.mets,'cis[c]]'),end) = -1;\n",
    "model.S(strcmp(model.mets,'gthrd[c]'),end) = -1;\n",
    "model.S(strcmp(model.mets,'cisGSH[c]'),end) = 1;\n",
    "model.S(strcmp(model.mets,'h[c]'),end) = 1;\n",
    "model.rxns{end+1} = 'cis_cisGSH';\n",
    "model.lb(end+1) = 0;\n",
    "model.ub(end+1) = max(model.ub);\n",
    "model.c(end+1) = 0;\n",
    "model.rules{end+1} = 'x(1179) | x(3176) | x(1184) | x(1185)';\n",
    "model.grRules{end+1} = '(2944) or (9446) or (2950) or (2952)';\n",
    "model.rxnGeneMat(end+1,:) = zeros(1,length(model.genes));\n",
    "model.rxnGeneMat(end,1179) = 1;\n",
    "model.rxnGeneMat(end,3176) = 1;\n",
    "model.rxnGeneMat(end,1184) = 1;\n",
    "model.rxnGeneMat(end,1185) = 1;\n",
    "model.rxnConfidenceScores(end+1) = 0;\n",
    "model.rxnNames{end+1} = 'Cisplatin to cisplatin-GSH';\n",
    "model.rxnNotes{end+1} = '';\n",
    "model.rxnECNumbers{end+1} = '2.5.1.18';\n",
    "model.rxnReferences{end+1} = '';\n",
    "model.rxnKEGGID{end+1} = '';\n",
    "model.subSystems{end+1} = 'MODULE CISPLATIN';\n",
    "model.rxnCOG{end+1} = '';\n",
    "model.rxnKeggOrthology{end+1} = '';\n",
    "model.rxnReconMap{end+1} = '';\n",
    "model.rxnFormulas{end+1} = '1 cis[c] + 1 gthrd[c] --> 1 cisGSH[c] + 1 h[c]';\n",
    "\n",
    "% cis[c] --> cisMT[c]\n",
    "model.S(:,end+1) = zeros(length(model.mets),1);\n",
    "model.S(strcmp(model.mets,'cis[c]]'),end) = -2;\n",
    "model.S(strcmp(model.mets,'gthrd[c]'),end) = -2;\n",
    "model.S(strcmp(model.mets,'cisMT[c]'),end) = 1;\n",
    "model.S(strcmp(model.mets,'gthox[c]'),end) = 1;\n",
    "model.S(strcmp(model.mets,'h[c]'),end) = 2;\n",
    "model.rxns{end+1} = 'cis_cisMT';\n",
    "model.lb(end+1) = 0;\n",
    "model.ub(end+1) = max(model.ub);\n",
    "model.c(end+1) = 0;\n",
    "model.rules{end+1} = 'x(3272) | x(3273) | x(3274) | x(3275)';\n",
    "model.grRules{end+1} = '1';\n",
    "model.rxnGeneMat(end+1,:) = zeros(1,length(model.genes));\n",
    "model.rxnGeneMat(:,(end+1):(end+4)) = zeros(length(model.rxns),4);\n",
    "model.rxnGeneMat(end,3272) = 1;\n",
    "model.rxnGeneMat(end,3273) = 1;\n",
    "model.rxnGeneMat(end,3274) = 1;\n",
    "model.rxnGeneMat(end,3275) = 1;\n",
    "model.genes{end+1} = '4489';\n",
    "model.genes{end+1} = '4502';\n",
    "model.genes{end+1} = '4504';\n",
    "model.genes{end+1} = '84560';\n",
    "model.geneSymbols{end+1} = 'MT1A';\n",
    "model.geneSymbols{end+1} = 'MT2A';\n",
    "model.geneSymbols{end+1} = 'MT3';\n",
    "model.geneSymbols{end+1} = 'MT4';\n",
    "model.rxnConfidenceScores(end+1) = 0;\n",
    "model.rxnNames{end+1} = 'Cisplatin to cisplatin-metallothionein';\n",
    "model.rxnNotes{end+1} = '';\n",
    "model.rxnECNumbers{end+1} = '1.8.5.1';\n",
    "model.rxnReferences{end+1} = '';\n",
    "model.rxnKEGGID{end+1} = '';\n",
    "model.subSystems{end+1} = 'MODULE CISPLATIN';\n",
    "model.rxnCOG{end+1} = '';\n",
    "model.rxnKeggOrthology{end+1} = '';\n",
    "model.rxnReconMap{end+1} = '';\n",
    "model.rxnFormulas{end+1} = '2 cis[c] + 2 gthrd[c] --> 1 cisMT[c] + 1 gthox[c] + 2 h[c]';\n",
    "\n",
    "% % % sinks\n",
    "\n",
    "% artificial cis-mtDNA[m] sink\n",
    "model.S(:,end+1) = zeros(length(model.mets),1);\n",
    "model.S(strcmp(model.mets,'cis-mtDNA[m]'),end) = -1;\n",
    "model.rxns{end+1} = 'SINK_cis-mtDNA[m]';\n",
    "model.lb(end+1) = 0;\n",
    "model.ub(end+1) = max(model.ub);\n",
    "model.c(end+1) = 0;\n",
    "model.rules{end+1} = '';\n",
    "model.grRules{end+1} = '';\n",
    "model.rxnGeneMat(end+1,:) = zeros(1,length(model.genes));\n",
    "model.rxnConfidenceScores(end+1) = 0;\n",
    "model.rxnNames{end+1} = 'Artificial sink on Cisplatin-mitochondrial DNA';\n",
    "model.rxnNotes{end+1} = '';\n",
    "model.rxnECNumbers{end+1} = '';\n",
    "model.rxnReferences{end+1} = '';\n",
    "model.rxnKEGGID{end+1} = '';\n",
    "model.subSystems{end+1} = 'MODULE CISPLATIN';\n",
    "model.rxnCOG{end+1} = '';\n",
    "model.rxnKeggOrthology{end+1} = '';\n",
    "model.rxnReconMap{end+1} = '';\n",
    "model.rxnFormulas{end+1} = '1 cis-mtDNA[m] --> ';"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Cyclophosphamide"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Metabolites"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 155,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n"
     ]
    }
   ],
   "source": [
    "% cpa[e]\n",
    "model.S(end+1,:) = zeros(1,length(model.rxns));\n",
    "model.mets{end+1} = 'cpa[e]';\n",
    "model.b(end+1) = 0;\n",
    "model.csense(end+1) = 'E';\n",
    "model.metCharges(end+1) = 0;\n",
    "model.metFormulas{end+1} = 'C7H15Cl2N2O2P';\n",
    "model.metSmiles{end+1} = 'C1CNP(=O)(OC1)N(CCCl)CCCl';\n",
    "model.metNames{end+1} = 'Cyclophosphamide';\n",
    "model.metHMDBID{end+1} = 'HMDB0014672';\n",
    "model.metInChIString{end+1} = 'InChI=1S/C7H15Cl2N2O2P/c8-2-5-11(6-3-9)14(12)10-4-1-7-13-14/h1-7H2,(H,10,12)';\n",
    "model.metKEGGID{end+1} = '';\n",
    "model.metPubChemID{end+1} = '2907';\n",
    "model.metCHEBIID{end+1} = '4027';\n",
    "model.metPdMap{end+1} = '';\n",
    "model.metReconMap{end+1} = '';\n",
    "\n",
    "% cpa[c]\n",
    "model.S(end+1,:) = zeros(1,length(model.rxns));\n",
    "model.mets{end+1} = 'cpa[c]';\n",
    "model.b(end+1) = 0;\n",
    "model.csense(end+1) = 'E';\n",
    "model.metCharges(end+1) = 0;\n",
    "model.metFormulas{end+1} = 'C7H15Cl2N2O2P';\n",
    "model.metSmiles{end+1} = 'C1CNP(=O)(OC1)N(CCCl)CCCl';\n",
    "model.metNames{end+1} = 'Cyclophosphamide';\n",
    "model.metHMDBID{end+1} = 'HMDB0014672';\n",
    "model.metInChIString{end+1} = 'InChI=1S/C7H15Cl2N2O2P/c8-2-5-11(6-3-9)14(12)10-4-1-7-13-14/h1-7H2,(H,10,12)';\n",
    "model.metKEGGID{end+1} = '';\n",
    "model.metPubChemID{end+1} = '2907';\n",
    "model.metCHEBIID{end+1} = '4027';\n",
    "model.metPdMap{end+1} = '';\n",
    "model.metReconMap{end+1} = '';\n",
    "\n",
    "% dce-cpa[e]\n",
    "model.S(end+1,:) = zeros(1,length(model.rxns));\n",
    "model.mets{end+1} = 'dce-cpa[e]';\n",
    "model.b(end+1) = 0;\n",
    "model.csense(end+1) = 'E';\n",
    "model.metCharges(end+1) = 0;\n",
    "model.metFormulas{end+1} = 'C5H12ClN2O2P';\n",
    "model.metSmiles{end+1} = 'C1CNP(=O)(OC1)NCCCl';\n",
    "model.metNames{end+1} = 'Dechloroethylcyclophosphamide';\n",
    "model.metHMDBID{end+1} = 'HMDB0013858';\n",
    "model.metInChIString{end+1} = 'InChI=1S/C5H12ClN2O2P/c6-2-4-8-11(9)7-3-1-5-10-11/h1-5H2,(H2,7,8,9)';\n",
    "model.metKEGGID{end+1} = '';\n",
    "model.metPubChemID{end+1} = '114861';\n",
    "model.metCHEBIID{end+1} = '80558';\n",
    "model.metPdMap{end+1} = '';\n",
    "model.metReconMap{end+1} = '';\n",
    "\n",
    "% dce-cpa[c]\n",
    "model.S(end+1,:) = zeros(1,length(model.rxns));\n",
    "model.mets{end+1} = 'dce-cpa[c]';\n",
    "model.b(end+1) = 0;\n",
    "model.csense(end+1) = 'E';\n",
    "model.metCharges(end+1) = 0;\n",
    "model.metFormulas{end+1} = 'C5H12ClN2O2P';\n",
    "model.metSmiles{end+1} = 'C1CNP(=O)(OC1)NCCCl';\n",
    "model.metNames{end+1} = 'Dechloroethylcyclophosphamide';\n",
    "model.metHMDBID{end+1} = 'HMDB0013858';\n",
    "model.metInChIString{end+1} = 'InChI=1S/C5H12ClN2O2P/c6-2-4-8-11(9)7-3-1-5-10-11/h1-5H2,(H2,7,8,9)';\n",
    "model.metKEGGID{end+1} = '';\n",
    "model.metPubChemID{end+1} = '114861';\n",
    "model.metCHEBIID{end+1} = '80558';\n",
    "model.metPdMap{end+1} = '';\n",
    "model.metReconMap{end+1} = '';\n",
    "\n",
    "% caa[e]\n",
    "model.S(end+1,:) = zeros(1,length(model.rxns));\n",
    "model.mets{end+1} = 'caa[e]';\n",
    "model.b(end+1) = 0;\n",
    "model.csense(end+1) = 'E';\n",
    "model.metCharges(end+1) = 0;\n",
    "model.metFormulas{end+1} = 'C2H3ClO';\n",
    "model.metSmiles{end+1} = 'C(C=O)Cl';\n",
    "model.metNames{end+1} = 'Chloroacetaldehyde';\n",
    "model.metHMDBID{end+1} = 'HMDB0013860';\n",
    "model.metInChIString{end+1} = 'InChI=1S/C2H3ClO/c3-1-2-4/h2H,1H2';\n",
    "model.metKEGGID{end+1} = '';\n",
    "model.metPubChemID{end+1} = '33';\n",
    "model.metCHEBIID{end+1} = '27871';\n",
    "model.metPdMap{end+1} = '';\n",
    "model.metReconMap{end+1} = '';\n",
    "\n",
    "% caa[c]\n",
    "model.S(end+1,:) = zeros(1,length(model.rxns));\n",
    "model.mets{end+1} = 'caa[c]';\n",
    "model.b(end+1) = 0;\n",
    "model.csense(end+1) = 'E';\n",
    "model.metCharges(end+1) = 0;\n",
    "model.metFormulas{end+1} = 'C2H3ClO';\n",
    "model.metSmiles{end+1} = 'C(C=O)Cl';\n",
    "model.metNames{end+1} = 'Chloroacetaldehyde';\n",
    "model.metHMDBID{end+1} = 'HMDB0013860';\n",
    "model.metInChIString{end+1} = 'InChI=1S/C2H3ClO/c3-1-2-4/h2H,1H2';\n",
    "model.metKEGGID{end+1} = '';\n",
    "model.metPubChemID{end+1} = '33';\n",
    "model.metCHEBIID{end+1} = '27871';\n",
    "model.metPdMap{end+1} = '';\n",
    "model.metReconMap{end+1} = '';\n",
    "\n",
    "% 4oh-cpa[c]\n",
    "model.S(end+1,:) = zeros(1,length(model.rxns));\n",
    "model.mets{end+1} = '4oh-cpa[c]';\n",
    "model.b(end+1) = 0;\n",
    "model.csense(end+1) = 'E';\n",
    "model.metCharges(end+1) = 0;\n",
    "model.metFormulas{end+1} = 'C7H15Cl2N2O3P';\n",
    "model.metSmiles{end+1} = 'C1COP(=O)(NC1O)N(CCCl)CCCl';\n",
    "model.metNames{end+1} = '4-Hydroxycyclophosphamide';\n",
    "model.metHMDBID{end+1} = 'HMDB0013856';\n",
    "model.metInChIString{end+1} = 'InChI=1S/C7H15Cl2N2O3P/c8-2-4-11(5-3-9)15(13)10-7(12)1-6-14-15/h7,12H,1-6H2,(H,10,13)';\n",
    "model.metKEGGID{end+1} = '';\n",
    "model.metPubChemID{end+1} = '99735';\n",
    "model.metCHEBIID{end+1} = '1864';\n",
    "model.metPdMap{end+1} = '';\n",
    "model.metReconMap{end+1} = '';\n",
    "\n",
    "% 4k-cpa[e]\n",
    "model.S(end+1,:) = zeros(1,length(model.rxns));\n",
    "model.mets{end+1} = '4k-cpa[e]';\n",
    "model.b(end+1) = 0;\n",
    "model.csense(end+1) = 'E';\n",
    "model.metCharges(end+1) = 0;\n",
    "model.metFormulas{end+1} = 'C7H13Cl2N2O3P';\n",
    "model.metSmiles{end+1} = 'C1COP(=O)(NC1=O)N(CCCl)CCCl';\n",
    "model.metNames{end+1} = '4-Ketocyclophosphamide';\n",
    "model.metHMDBID{end+1} = 'HMDB0060686';\n",
    "model.metInChIString{end+1} = 'InChI=1S/C7H13Cl2N2O3P/c8-2-4-11(5-3-9)15(13)10-7(12)1-6-14-15/h1-6H2,(H,10,12,13)';\n",
    "model.metKEGGID{end+1} = '';\n",
    "model.metPubChemID{end+1} = '33676';\n",
    "model.metCHEBIID{end+1} = '1887';\n",
    "model.metPdMap{end+1} = '';\n",
    "model.metReconMap{end+1} = '';\n",
    "\n",
    "% 4k-cpa[c]\n",
    "model.S(end+1,:) = zeros(1,length(model.rxns));\n",
    "model.mets{end+1} = '4k-cpa[c]';\n",
    "model.b(end+1) = 0;\n",
    "model.csense(end+1) = 'E';\n",
    "model.metCharges(end+1) = 0;\n",
    "model.metFormulas{end+1} = 'C7H13Cl2N2O3P';\n",
    "model.metSmiles{end+1} = 'C1COP(=O)(NC1=O)N(CCCl)CCCl';\n",
    "model.metNames{end+1} = '4-Ketocyclophosphamide';\n",
    "model.metHMDBID{end+1} = 'HMDB0060686';\n",
    "model.metInChIString{end+1} = 'InChI=1S/C7H13Cl2N2O3P/c8-2-4-11(5-3-9)15(13)10-7(12)1-6-14-15/h1-6H2,(H,10,12,13)';\n",
    "model.metKEGGID{end+1} = '';\n",
    "model.metPubChemID{end+1} = '33676';\n",
    "model.metCHEBIID{end+1} = '1887';\n",
    "model.metPdMap{end+1} = '';\n",
    "model.metReconMap{end+1} = '';\n",
    "\n",
    "% aldph[c]\n",
    "model.S(end+1,:) = zeros(1,length(model.rxns));\n",
    "model.mets{end+1} = 'aldph[c]';\n",
    "model.b(end+1) = 0;\n",
    "model.csense(end+1) = 'E';\n",
    "model.metCharges(end+1) = 0;\n",
    "model.metFormulas{end+1} = 'C7H15Cl2N2O3P';\n",
    "model.metSmiles{end+1} = 'C(COP(=O)(N)N(CCCl)CCCl)C=O';\n",
    "model.metNames{end+1} = 'Aldophosphamide';\n",
    "model.metHMDBID{end+1} = 'HMDB0060433';\n",
    "model.metInChIString{end+1} = 'InChI=1S/C7H15Cl2N2O3P/c8-2-4-11(5-3-9)15(10,13)14-7-1-6-12/h6H,1-5,7H2,(H2,10,13)';\n",
    "model.metKEGGID{end+1} = '';\n",
    "model.metPubChemID{end+1} = '107744';\n",
    "model.metCHEBIID{end+1} = '2560';\n",
    "model.metPdMap{end+1} = '';\n",
    "model.metReconMap{end+1} = '';\n",
    "\n",
    "% alcph[e]\n",
    "model.S(end+1,:) = zeros(1,length(model.rxns));\n",
    "model.mets{end+1} = 'alcph[e]';\n",
    "model.b(end+1) = 0;\n",
    "model.csense(end+1) = 'E';\n",
    "model.metCharges(end+1) = 0;\n",
    "model.metFormulas{end+1} = 'C7H17Cl2N2O3P';\n",
    "model.metSmiles{end+1} = 'C(CO)COP(=O)(N)N(CCCl)CCCl';\n",
    "model.metNames{end+1} = 'Alcophosphamide';\n",
    "model.metHMDBID{end+1} = 'HMDB0060432';\n",
    "model.metInChIString{end+1} = 'InChI=1S/C7H17Cl2N2O3P/c8-2-4-11(5-3-9)15(10,13)14-7-1-6-12/h12H,1-7H2,(H2,10,13)';\n",
    "model.metKEGGID{end+1} = '';\n",
    "model.metPubChemID{end+1} = '98612';\n",
    "model.metCHEBIID{end+1} = '80559';\n",
    "model.metPdMap{end+1} = '';\n",
    "model.metReconMap{end+1} = '';\n",
    "\n",
    "% alcph[c]\n",
    "model.S(end+1,:) = zeros(1,length(model.rxns));\n",
    "model.mets{end+1} = 'alcph[c]';\n",
    "model.b(end+1) = 0;\n",
    "model.csense(end+1) = 'E';\n",
    "model.metCharges(end+1) = 0;\n",
    "model.metFormulas{end+1} = 'C7H17Cl2N2O3P';\n",
    "model.metSmiles{end+1} = 'C(CO)COP(=O)(N)N(CCCl)CCCl';\n",
    "model.metNames{end+1} = 'Alcophosphamide';\n",
    "model.metHMDBID{end+1} = 'HMDB0060432';\n",
    "model.metInChIString{end+1} = 'InChI=1S/C7H17Cl2N2O3P/c8-2-4-11(5-3-9)15(10,13)14-7-1-6-12/h12H,1-7H2,(H2,10,13)';\n",
    "model.metKEGGID{end+1} = '';\n",
    "model.metPubChemID{end+1} = '98612';\n",
    "model.metCHEBIID{end+1} = '80559';\n",
    "model.metPdMap{end+1} = '';\n",
    "model.metReconMap{end+1} = '';\n",
    "\n",
    "% cph[e]\n",
    "model.S(end+1,:) = zeros(1,length(model.rxns));\n",
    "model.mets{end+1} = 'cph[e]';\n",
    "model.b(end+1) = 0;\n",
    "model.csense(end+1) = 'E';\n",
    "model.metCharges(end+1) = 0;\n",
    "model.metFormulas{end+1} = 'C7H15Cl2N2O4P';\n",
    "model.metSmiles{end+1} = 'C(COP(=O)(N)N(CCCl)CCCl)C(=O)O';\n",
    "model.metNames{end+1} = 'Carboxyphosphamide';\n",
    "model.metHMDBID{end+1} = 'HMDB0060449';\n",
    "model.metInChIString{end+1} = 'InChI=1S/C7H15Cl2N2O4P/c8-2-4-11(5-3-9)16(10,14)15-6-1-7(12)13/h1-6H2,(H2,10,14)(H,12,13)';\n",
    "model.metKEGGID{end+1} = '';\n",
    "model.metPubChemID{end+1} = '31515';\n",
    "model.metCHEBIID{end+1} = '3410';\n",
    "model.metPdMap{end+1} = '';\n",
    "model.metReconMap{end+1} = '';\n",
    "\n",
    "% cph[c]\n",
    "model.S(end+1,:) = zeros(1,length(model.rxns));\n",
    "model.mets{end+1} = 'cph[c]';\n",
    "model.b(end+1) = 0;\n",
    "model.csense(end+1) = 'E';\n",
    "model.metCharges(end+1) = 0;\n",
    "model.metFormulas{end+1} = 'C7H15Cl2N2O4P';\n",
    "model.metSmiles{end+1} = 'C(COP(=O)(N)N(CCCl)CCCl)C(=O)O';\n",
    "model.metNames{end+1} = 'Carboxyphosphamide';\n",
    "model.metHMDBID{end+1} = 'HMDB0060449';\n",
    "model.metInChIString{end+1} = 'InChI=1S/C7H15Cl2N2O4P/c8-2-4-11(5-3-9)16(10,14)15-6-1-7(12)13/h1-6H2,(H2,10,14)(H,12,13)';\n",
    "model.metKEGGID{end+1} = '';\n",
    "model.metPubChemID{end+1} = '31515';\n",
    "model.metCHEBIID{end+1} = '3410';\n",
    "model.metPdMap{end+1} = '';\n",
    "model.metReconMap{end+1} = '';\n",
    "\n",
    "% phmust[c]\n",
    "model.S(end+1,:) = zeros(1,length(model.rxns));\n",
    "model.mets{end+1} = 'phmust[c]';\n",
    "model.b(end+1) = 0;\n",
    "model.csense(end+1) = 'E';\n",
    "model.metCharges(end+1) = 0;\n",
    "model.metFormulas{end+1} = 'C4H11Cl2N2O2P';\n",
    "model.metSmiles{end+1} = 'C(CCl)N(CCCl)P(=O)(N)O';\n",
    "model.metNames{end+1} = 'Phosphoramide mustard';\n",
    "model.metHMDBID{end+1} = 'HMDB0060689';\n",
    "model.metInChIString{end+1} = 'InChI=1S/C4H11Cl2N2O2P/c5-1-3-8(4-2-6)11(7,9)10/h1-4H2,(H3,7,9,10)';\n",
    "model.metKEGGID{end+1} = '';\n",
    "model.metPubChemID{end+1} = '96356';\n",
    "model.metCHEBIID{end+1} = '8163';\n",
    "model.metPdMap{end+1} = '';\n",
    "model.metReconMap{end+1} = '';\n",
    "\n",
    "% acr[c]\n",
    "model.S(end+1,:) = zeros(1,length(model.rxns));\n",
    "model.mets{end+1} = 'acr[c]';\n",
    "model.b(end+1) = 0;\n",
    "model.csense(end+1) = 'E';\n",
    "model.metCharges(end+1) = 0;\n",
    "model.metFormulas{end+1} = 'C3H4O';\n",
    "model.metSmiles{end+1} = 'C=CC=O';\n",
    "model.metNames{end+1} = 'Acrolein';\n",
    "model.metHMDBID{end+1} = 'HMDB0041822';\n",
    "model.metInChIString{end+1} = 'InChI=1S/C3H4O/c1-2-3-4/h2-3H,1H2';\n",
    "model.metKEGGID{end+1} = '';\n",
    "model.metPubChemID{end+1} = '7847';\n",
    "model.metCHEBIID{end+1} = '15368';\n",
    "model.metPdMap{end+1} = '';\n",
    "model.metReconMap{end+1} = '';\n",
    "\n",
    "% phmust-GSH[e]\n",
    "model.S(end+1,:) = zeros(1,length(model.rxns));\n",
    "model.mets{end+1} = 'phmust-GSH[e]';\n",
    "model.b(end+1) = 0;\n",
    "model.csense(end+1) = 'E';\n",
    "model.metCharges(end+1) = 0;\n",
    "model.metFormulas{end+1} = 'C14H26Cl2N5O8PS';\n",
    "model.metSmiles{end+1} = '';\n",
    "model.metNames{end+1} = 'Phosphoramide mustard-GSH';\n",
    "model.metHMDBID{end+1} = '';\n",
    "model.metInChIString{end+1} = '';\n",
    "model.metKEGGID{end+1} = '';\n",
    "model.metPubChemID{end+1} = '';\n",
    "model.metCHEBIID{end+1} = '';\n",
    "model.metPdMap{end+1} = '';\n",
    "model.metReconMap{end+1} = '';\n",
    "\n",
    "% phmust-GSH[c]\n",
    "model.S(end+1,:) = zeros(1,length(model.rxns));\n",
    "model.mets{end+1} = 'phmust-GSH[c]';\n",
    "model.b(end+1) = 0;\n",
    "model.csense(end+1) = 'E';\n",
    "model.metCharges(end+1) = 0;\n",
    "model.metFormulas{end+1} = 'C14H26Cl2N5O8PS';\n",
    "model.metSmiles{end+1} = '';\n",
    "model.metNames{end+1} = 'Phosphoramide mustard-GSH';\n",
    "model.metHMDBID{end+1} = '';\n",
    "model.metInChIString{end+1} = '';\n",
    "model.metKEGGID{end+1} = '';\n",
    "model.metPubChemID{end+1} = '';\n",
    "model.metCHEBIID{end+1} = '';\n",
    "model.metPdMap{end+1} = '';\n",
    "model.metReconMap{end+1} = '';\n",
    "\n",
    "% acr-GSH[e]\n",
    "model.S(end+1,:) = zeros(1,length(model.rxns));\n",
    "model.mets{end+1} = 'acr-GSH[e]';\n",
    "model.b(end+1) = 0;\n",
    "model.csense(end+1) = 'E';\n",
    "model.metCharges(end+1) = 0;\n",
    "model.metFormulas{end+1} = 'C13H19N3O7S';\n",
    "model.metSmiles{end+1} = '';\n",
    "model.metNames{end+1} = 'Acrolein-GSH';\n",
    "model.metHMDBID{end+1} = '';\n",
    "model.metInChIString{end+1} = '';\n",
    "model.metKEGGID{end+1} = '';\n",
    "model.metPubChemID{end+1} = '';\n",
    "model.metCHEBIID{end+1} = '';\n",
    "model.metPdMap{end+1} = '';\n",
    "model.metReconMap{end+1} = '';\n",
    "\n",
    "% acr-GSH[c]\n",
    "model.S(end+1,:) = zeros(1,length(model.rxns));\n",
    "model.mets{end+1} = 'acr-GSH[c]';\n",
    "model.b(end+1) = 0;\n",
    "model.csense(end+1) = 'E';\n",
    "model.metCharges(end+1) = 0;\n",
    "model.metFormulas{end+1} = 'C13H19N3O7S';\n",
    "model.metSmiles{end+1} = '';\n",
    "model.metNames{end+1} = 'Acrolein-GSH';\n",
    "model.metHMDBID{end+1} = '';\n",
    "model.metInChIString{end+1} = '';\n",
    "model.metKEGGID{end+1} = '';\n",
    "model.metPubChemID{end+1} = '';\n",
    "model.metCHEBIID{end+1} = '';\n",
    "model.metPdMap{end+1} = '';\n",
    "model.metReconMap{end+1} = '';"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Reactions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 156,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n"
     ]
    }
   ],
   "source": [
    "% % % import metabolites\n",
    "\n",
    "% cpa[e] <=>\n",
    "model.S(:,end+1) = zeros(length(model.mets),1);\n",
    "model.S(strcmp(model.mets,'cpa[e]'),end) = -1;\n",
    "model.rxns{end+1} = 'EX_cpa[e]';\n",
    "model.lb(end+1) = min(model.lb);\n",
    "model.ub(end+1) = max(model.ub);\n",
    "model.c(end+1) = 0;\n",
    "model.rules{end+1} = '';\n",
    "model.grRules{end+1} = '';\n",
    "model.rxnGeneMat(end+1,:) = zeros(1,length(model.genes));\n",
    "model.rxnConfidenceScores(end+1) = 0;\n",
    "model.rxnNames{end+1} = 'Exchange of Cyclophosphamide';\n",
    "model.rxnNotes{end+1} = '';\n",
    "model.rxnECNumbers{end+1} = '';\n",
    "model.rxnReferences{end+1} = '';\n",
    "model.rxnKEGGID{end+1} = '';\n",
    "model.subSystems{end+1} = 'MODULE CYCLOPHOSPHAMIDE';\n",
    "model.rxnCOG{end+1} = '';\n",
    "model.rxnKeggOrthology{end+1} = '';\n",
    "model.rxnReconMap{end+1} = '';\n",
    "model.rxnFormulas{end+1} = '1 cpa[e] <=> ';\n",
    "\n",
    "% dce-cpa[e] <=>\n",
    "model.S(:,end+1) = zeros(length(model.mets),1);\n",
    "model.S(strcmp(model.mets,'dce-cpa[e]'),end) = -1;\n",
    "model.rxns{end+1} = 'EX_dce-cpa[e]';\n",
    "model.lb(end+1) = min(model.lb);\n",
    "model.ub(end+1) = max(model.ub);\n",
    "model.c(end+1) = 0;\n",
    "model.rules{end+1} = '';\n",
    "model.grRules{end+1} = '';\n",
    "model.rxnGeneMat(end+1,:) = zeros(1,length(model.genes));\n",
    "model.rxnConfidenceScores(end+1) = 0;\n",
    "model.rxnNames{end+1} = 'Exchange of Dechloroethylcyclophosphamide';\n",
    "model.rxnNotes{end+1} = '';\n",
    "model.rxnECNumbers{end+1} = '';\n",
    "model.rxnReferences{end+1} = '';\n",
    "model.rxnKEGGID{end+1} = '';\n",
    "model.subSystems{end+1} = 'MODULE CYCLOPHOSPHAMIDE';\n",
    "model.rxnCOG{end+1} = '';\n",
    "model.rxnKeggOrthology{end+1} = '';\n",
    "model.rxnReconMap{end+1} = '';\n",
    "model.rxnFormulas{end+1} = '1 dce-cpa[e] <=> ';\n",
    "\n",
    "% caa[e] <=>\n",
    "model.S(:,end+1) = zeros(length(model.mets),1);\n",
    "model.S(strcmp(model.mets,'caa[e]'),end) = -1;\n",
    "model.rxns{end+1} = 'EX_caa[e]';\n",
    "model.lb(end+1) = min(model.lb);\n",
    "model.ub(end+1) = max(model.ub);\n",
    "model.c(end+1) = 0;\n",
    "model.rules{end+1} = '';\n",
    "model.grRules{end+1} = '';\n",
    "model.rxnGeneMat(end+1,:) = zeros(1,length(model.genes));\n",
    "model.rxnConfidenceScores(end+1) = 0;\n",
    "model.rxnNames{end+1} = 'Exchange of Chloroacetaldehyde';\n",
    "model.rxnNotes{end+1} = '';\n",
    "model.rxnECNumbers{end+1} = '';\n",
    "model.rxnReferences{end+1} = '';\n",
    "model.rxnKEGGID{end+1} = '';\n",
    "model.subSystems{end+1} = 'MODULE CYCLOPHOSPHAMIDE';\n",
    "model.rxnCOG{end+1} = '';\n",
    "model.rxnKeggOrthology{end+1} = '';\n",
    "model.rxnReconMap{end+1} = '';\n",
    "model.rxnFormulas{end+1} = '1 caa[e] <=> ';\n",
    "\n",
    "% 4k-cpa[e] <=>\n",
    "model.S(:,end+1) = zeros(length(model.mets),1);\n",
    "model.S(strcmp(model.mets,'4k-cpa[e]'),end) = -1;\n",
    "model.rxns{end+1} = 'EX_4k-cpa[e]';\n",
    "model.lb(end+1) = min(model.lb);\n",
    "model.ub(end+1) = max(model.ub);\n",
    "model.c(end+1) = 0;\n",
    "model.rules{end+1} = '';\n",
    "model.grRules{end+1} = '';\n",
    "model.rxnGeneMat(end+1,:) = zeros(1,length(model.genes));\n",
    "model.rxnConfidenceScores(end+1) = 0;\n",
    "model.rxnNames{end+1} = 'Exchange of 4-Ketocyclophosphamide';\n",
    "model.rxnNotes{end+1} = '';\n",
    "model.rxnECNumbers{end+1} = '';\n",
    "model.rxnReferences{end+1} = '';\n",
    "model.rxnKEGGID{end+1} = '';\n",
    "model.subSystems{end+1} = 'MODULE CYCLOPHOSPHAMIDE';\n",
    "model.rxnCOG{end+1} = '';\n",
    "model.rxnKeggOrthology{end+1} = '';\n",
    "model.rxnReconMap{end+1} = '';\n",
    "model.rxnFormulas{end+1} = '1 4k-cpa[e] <=> ';\n",
    "\n",
    "% alcph[e] <=>\n",
    "model.S(:,end+1) = zeros(length(model.mets),1);\n",
    "model.S(strcmp(model.mets,'alcph[e]'),end) = -1;\n",
    "model.rxns{end+1} = 'EX_alcph[e]';\n",
    "model.lb(end+1) = min(model.lb);\n",
    "model.ub(end+1) = max(model.ub);\n",
    "model.c(end+1) = 0;\n",
    "model.rules{end+1} = '';\n",
    "model.grRules{end+1} = '';\n",
    "model.rxnGeneMat(end+1,:) = zeros(1,length(model.genes));\n",
    "model.rxnConfidenceScores(end+1) = 0;\n",
    "model.rxnNames{end+1} = 'Exchange of Alcophosphamide';\n",
    "model.rxnNotes{end+1} = '';\n",
    "model.rxnECNumbers{end+1} = '';\n",
    "model.rxnReferences{end+1} = '';\n",
    "model.rxnKEGGID{end+1} = '';\n",
    "model.subSystems{end+1} = 'MODULE CYCLOPHOSPHAMIDE';\n",
    "model.rxnCOG{end+1} = '';\n",
    "model.rxnKeggOrthology{end+1} = '';\n",
    "model.rxnReconMap{end+1} = '';\n",
    "model.rxnFormulas{end+1} = '1 alcph[e] <=> ';\n",
    "\n",
    "% cph[e] <=>\n",
    "model.S(:,end+1) = zeros(length(model.mets),1);\n",
    "model.S(strcmp(model.mets,'cph[e]'),end) = -1;\n",
    "model.rxns{end+1} = 'EX_cph[e]';\n",
    "model.lb(end+1) = min(model.lb);\n",
    "model.ub(end+1) = max(model.ub);\n",
    "model.c(end+1) = 0;\n",
    "model.rules{end+1} = '';\n",
    "model.grRules{end+1} = '';\n",
    "model.rxnGeneMat(end+1,:) = zeros(1,length(model.genes));\n",
    "model.rxnConfidenceScores(end+1) = 0;\n",
    "model.rxnNames{end+1} = 'Exchange of Carboxyphosphamide';\n",
    "model.rxnNotes{end+1} = '';\n",
    "model.rxnECNumbers{end+1} = '';\n",
    "model.rxnReferences{end+1} = '';\n",
    "model.rxnKEGGID{end+1} = '';\n",
    "model.subSystems{end+1} = 'MODULE CYCLOPHOSPHAMIDE';\n",
    "model.rxnCOG{end+1} = '';\n",
    "model.rxnKeggOrthology{end+1} = '';\n",
    "model.rxnReconMap{end+1} = '';\n",
    "model.rxnFormulas{end+1} = '1 cph[e] <=> ';\n",
    "\n",
    "% phmust-GSH[e] <=>\n",
    "model.S(:,end+1) = zeros(length(model.mets),1);\n",
    "model.S(strcmp(model.mets,'phmust-GSH[e]'),end) = -1;\n",
    "model.rxns{end+1} = 'EX_phmust-GSH[e]';\n",
    "model.lb(end+1) = min(model.lb);\n",
    "model.ub(end+1) = max(model.ub);\n",
    "model.c(end+1) = 0;\n",
    "model.rules{end+1} = '';\n",
    "model.grRules{end+1} = '';\n",
    "model.rxnGeneMat(end+1,:) = zeros(1,length(model.genes));\n",
    "model.rxnConfidenceScores(end+1) = 0;\n",
    "model.rxnNames{end+1} = 'Exchange of Phosphoramide mustard-GSH';\n",
    "model.rxnNotes{end+1} = '';\n",
    "model.rxnECNumbers{end+1} = '';\n",
    "model.rxnReferences{end+1} = '';\n",
    "model.rxnKEGGID{end+1} = '';\n",
    "model.subSystems{end+1} = 'MODULE CYCLOPHOSPHAMIDE';\n",
    "model.rxnCOG{end+1} = '';\n",
    "model.rxnKeggOrthology{end+1} = '';\n",
    "model.rxnReconMap{end+1} = '';\n",
    "model.rxnFormulas{end+1} = '1 phmust-GSH[e] <=> ';\n",
    "\n",
    "% acr-GSH[e] <=>\n",
    "model.S(:,end+1) = zeros(length(model.mets),1);\n",
    "model.S(strcmp(model.mets,'acr-GSH[e]'),end) = -1;\n",
    "model.rxns{end+1} = 'EX_acr-GSH[e]';\n",
    "model.lb(end+1) = min(model.lb);\n",
    "model.ub(end+1) = max(model.ub);\n",
    "model.c(end+1) = 0;\n",
    "model.rules{end+1} = '';\n",
    "model.grRules{end+1} = '';\n",
    "model.rxnGeneMat(end+1,:) = zeros(1,length(model.genes));\n",
    "model.rxnConfidenceScores(end+1) = 0;\n",
    "model.rxnNames{end+1} = 'Exchange of Acrolein-GSH';\n",
    "model.rxnNotes{end+1} = '';\n",
    "model.rxnECNumbers{end+1} = '';\n",
    "model.rxnReferences{end+1} = '';\n",
    "model.rxnKEGGID{end+1} = '';\n",
    "model.subSystems{end+1} = 'MODULE CYCLOPHOSPHAMIDE';\n",
    "model.rxnCOG{end+1} = '';\n",
    "model.rxnKeggOrthology{end+1} = '';\n",
    "model.rxnReconMap{end+1} = '';\n",
    "model.rxnFormulas{end+1} = '1 acr-GSH[e] <=> ';\n",
    "\n",
    "% % % transport\n",
    "\n",
    "% cpa[e] --> cpa[c]\n",
    "model.S(:,end+1) = zeros(length(model.mets),1);\n",
    "model.S(strcmp(model.mets,'cpa[e]'),end) = -1;\n",
    "model.S(strcmp(model.mets,'cpa[c]'),end) = 1;\n",
    "model.rxns{end+1} = 'cpa_transport_e_c';\n",
    "model.lb(end+1) = 0;\n",
    "model.ub(end+1) = max(model.ub);\n",
    "model.c(end+1) = 0;\n",
    "model.rules{end+1} = '';\n",
    "model.grRules{end+1} = '';\n",
    "model.rxnGeneMat(end+1,:) = zeros(1,length(model.genes));\n",
    "model.rxnConfidenceScores(end+1) = 0;\n",
    "model.rxnNames{end+1} = 'Transport of Cyclophosphamide from extracellular to cytosol';\n",
    "model.rxnNotes{end+1} = '';\n",
    "model.rxnECNumbers{end+1} = '';\n",
    "model.rxnReferences{end+1} = '';\n",
    "model.rxnKEGGID{end+1} = '';\n",
    "model.subSystems{end+1} = 'MODULE CYCLOPHOSPHAMIDE';\n",
    "model.rxnCOG{end+1} = '';\n",
    "model.rxnKeggOrthology{end+1} = '';\n",
    "model.rxnReconMap{end+1} = '';\n",
    "model.rxnFormulas{end+1} = '1 cpa[e] --> 1 cpa[c]';\n",
    "\n",
    "% cpa[c] --> cpa[e]\n",
    "model.S(:,end+1) = zeros(length(model.mets),1);\n",
    "model.S(strcmp(model.mets,'cpa[c]'),end) = -1;\n",
    "model.S(strcmp(model.mets,'atp[c]'),end) = -1;\n",
    "model.S(strcmp(model.mets,'h2o[c]'),end) = -1;\n",
    "model.S(strcmp(model.mets,'cpa[e]'),end) = 1;\n",
    "model.S(strcmp(model.mets,'adp[c]'),end) = 1;\n",
    "model.S(strcmp(model.mets,'pi[c]'),end) = 1;\n",
    "model.rxns{end+1} = 'cpa_transport_c_e';\n",
    "model.lb(end+1) = 0;\n",
    "model.ub(end+1) = max(model.ub);\n",
    "model.c(end+1) = 0;\n",
    "model.rules{end+1} = 'x(1823) | x(3174)';\n",
    "model.grRules{end+1} = '(5243) or (9429)';\n",
    "model.rxnGeneMat(end+1,:) = zeros(1,length(model.genes));\n",
    "model.rxnGeneMat(end,1823) = 1;\n",
    "model.rxnGeneMat(end,3174) = 1;\n",
    "model.rxnConfidenceScores(end+1) = 0;\n",
    "model.rxnNames{end+1} = 'Transport of Cyclophosphamide from cytosol to extracellular';\n",
    "model.rxnNotes{end+1} = '';\n",
    "model.rxnECNumbers{end+1} = 'X.X.X.X';\n",
    "model.rxnReferences{end+1} = '';\n",
    "model.rxnKEGGID{end+1} = '';\n",
    "model.subSystems{end+1} = 'MODULE CYCLOPHOSPHAMIDE';\n",
    "model.rxnCOG{end+1} = '';\n",
    "model.rxnKeggOrthology{end+1} = '';\n",
    "model.rxnReconMap{end+1} = '';\n",
    "model.rxnFormulas{end+1} = '1 cpa[c] + 1 atp[c] + 1 h2o[c] --> 1 cpa[e] + 1 adp[c] + 1 pi[c]';\n",
    "\n",
    "% dce-cpa[c] <=> dce-cpa[e]\n",
    "model.S(:,end+1) = zeros(length(model.mets),1);\n",
    "model.S(strcmp(model.mets,'dce-cpa[e]'),end) = 1;\n",
    "model.S(strcmp(model.mets,'dce-cpa[c]'),end) = -1;\n",
    "model.rxns{end+1} = 'dce-cpa_transport_c_e';\n",
    "model.lb(end+1) = min(model.lb);\n",
    "model.ub(end+1) = max(model.ub);\n",
    "model.c(end+1) = 0;\n",
    "model.rules{end+1} = '';\n",
    "model.grRules{end+1} = '';\n",
    "model.rxnGeneMat(end+1,:) = zeros(1,length(model.genes));\n",
    "model.rxnConfidenceScores(end+1) = 0;\n",
    "model.rxnNames{end+1} = 'Transport of Dechloroethylcyclophosphamide from cytosol to extracellular';\n",
    "model.rxnNotes{end+1} = '';\n",
    "model.rxnECNumbers{end+1} = '';\n",
    "model.rxnReferences{end+1} = '';\n",
    "model.rxnKEGGID{end+1} = '';\n",
    "model.subSystems{end+1} = 'MODULE CYCLOPHOSPHAMIDE';\n",
    "model.rxnCOG{end+1} = '';\n",
    "model.rxnKeggOrthology{end+1} = '';\n",
    "model.rxnReconMap{end+1} = '';\n",
    "model.rxnFormulas{end+1} = '1 dce-cpa[c] <=> 1 dce-cpa[e]';\n",
    "\n",
    "% caa[c] <=> caa[e]\n",
    "model.S(:,end+1) = zeros(length(model.mets),1);\n",
    "model.S(strcmp(model.mets,'caa[e]'),end) = 1;\n",
    "model.S(strcmp(model.mets,'caa[c]'),end) = -1;\n",
    "model.rxns{end+1} = 'caa_transport_c_e';\n",
    "model.lb(end+1) = min(model.lb);\n",
    "model.ub(end+1) = max(model.ub);\n",
    "model.c(end+1) = 0;\n",
    "model.rules{end+1} = '';\n",
    "model.grRules{end+1} = '';\n",
    "model.rxnGeneMat(end+1,:) = zeros(1,length(model.genes));\n",
    "model.rxnConfidenceScores(end+1) = 0;\n",
    "model.rxnNames{end+1} = 'Transport of Chloroacetaldehyde from cytosol to extracellular';\n",
    "model.rxnNotes{end+1} = '';\n",
    "model.rxnECNumbers{end+1} = '';\n",
    "model.rxnReferences{end+1} = '';\n",
    "model.rxnKEGGID{end+1} = '';\n",
    "model.subSystems{end+1} = 'MODULE CYCLOPHOSPHAMIDE';\n",
    "model.rxnCOG{end+1} = '';\n",
    "model.rxnKeggOrthology{end+1} = '';\n",
    "model.rxnReconMap{end+1} = '';\n",
    "model.rxnFormulas{end+1} = '1 caa[c] <=> 1 caa[e]';\n",
    "\n",
    "% 4k-cpa[c] <=> 4k-cpa[e]\n",
    "model.S(:,end+1) = zeros(length(model.mets),1);\n",
    "model.S(strcmp(model.mets,'4k-cpa[e]'),end) = 1;\n",
    "model.S(strcmp(model.mets,'4k-cpa[c]'),end) = -1;\n",
    "model.rxns{end+1} = '4k-cpa_transport_c_e';\n",
    "model.lb(end+1) = min(model.lb);\n",
    "model.ub(end+1) = max(model.ub);\n",
    "model.c(end+1) = 0;\n",
    "model.rules{end+1} = '';\n",
    "model.grRules{end+1} = '';\n",
    "model.rxnGeneMat(end+1,:) = zeros(1,length(model.genes));\n",
    "model.rxnConfidenceScores(end+1) = 0;\n",
    "model.rxnNames{end+1} = 'Transport of 4-Ketocyclophosphamide from cytosol to extracellular';\n",
    "model.rxnNotes{end+1} = '';\n",
    "model.rxnECNumbers{end+1} = '';\n",
    "model.rxnReferences{end+1} = '';\n",
    "model.rxnKEGGID{end+1} = '';\n",
    "model.subSystems{end+1} = 'MODULE CYCLOPHOSPHAMIDE';\n",
    "model.rxnCOG{end+1} = '';\n",
    "model.rxnKeggOrthology{end+1} = '';\n",
    "model.rxnReconMap{end+1} = '';\n",
    "model.rxnFormulas{end+1} = '1 4k-cpa[c] <=> 1 4k-cpa[e]';\n",
    "\n",
    "% alcph[c] <=> alcph[e]\n",
    "model.S(:,end+1) = zeros(length(model.mets),1);\n",
    "model.S(strcmp(model.mets,'alcph[e]'),end) = 1;\n",
    "model.S(strcmp(model.mets,'alcph[c]'),end) = -1;\n",
    "model.rxns{end+1} = 'alcph_transport_c_e';\n",
    "model.lb(end+1) = min(model.lb);\n",
    "model.ub(end+1) = max(model.ub);\n",
    "model.c(end+1) = 0;\n",
    "model.rules{end+1} = '';\n",
    "model.grRules{end+1} = '';\n",
    "model.rxnGeneMat(end+1,:) = zeros(1,length(model.genes));\n",
    "model.rxnConfidenceScores(end+1) = 0;\n",
    "model.rxnNames{end+1} = 'Transport of Alcophosphamide from cytosol to extracellular';\n",
    "model.rxnNotes{end+1} = '';\n",
    "model.rxnECNumbers{end+1} = '';\n",
    "model.rxnReferences{end+1} = '';\n",
    "model.rxnKEGGID{end+1} = '';\n",
    "model.subSystems{end+1} = 'MODULE CYCLOPHOSPHAMIDE';\n",
    "model.rxnCOG{end+1} = '';\n",
    "model.rxnKeggOrthology{end+1} = '';\n",
    "model.rxnReconMap{end+1} = '';\n",
    "model.rxnFormulas{end+1} = '1 alcph[c] <=> 1 alcph[e]';\n",
    "\n",
    "% cph[c] <=> cph[e]\n",
    "model.S(:,end+1) = zeros(length(model.mets),1);\n",
    "model.S(strcmp(model.mets,'cph[e]'),end) = 1;\n",
    "model.S(strcmp(model.mets,'cph[c]'),end) = -1;\n",
    "model.rxns{end+1} = 'cph_transport_c_e';\n",
    "model.lb(end+1) = min(model.lb);\n",
    "model.ub(end+1) = max(model.ub);\n",
    "model.c(end+1) = 0;\n",
    "model.rules{end+1} = '';\n",
    "model.grRules{end+1} = '';\n",
    "model.rxnGeneMat(end+1,:) = zeros(1,length(model.genes));\n",
    "model.rxnConfidenceScores(end+1) = 0;\n",
    "model.rxnNames{end+1} = 'Transport of Carboxyphosphamide from cytosol to extracellular';\n",
    "model.rxnNotes{end+1} = '';\n",
    "model.rxnECNumbers{end+1} = '';\n",
    "model.rxnReferences{end+1} = '';\n",
    "model.rxnKEGGID{end+1} = '';\n",
    "model.subSystems{end+1} = 'MODULE CYCLOPHOSPHAMIDE';\n",
    "model.rxnCOG{end+1} = '';\n",
    "model.rxnKeggOrthology{end+1} = '';\n",
    "model.rxnReconMap{end+1} = '';\n",
    "model.rxnFormulas{end+1} = '1 cph[c] <=> 1 cph[e]';\n",
    "\n",
    "% phmust-GSH[c] --> phmust-GSH[e]\n",
    "model.S(:,end+1) = zeros(length(model.mets),1);\n",
    "model.S(strcmp(model.mets,'phmust-GSH[c]'),end) = -1;\n",
    "model.S(strcmp(model.mets,'atp[c]'),end) = -1;\n",
    "model.S(strcmp(model.mets,'h2o[c]'),end) = -1;\n",
    "model.S(strcmp(model.mets,'phmust-GSH[e]'),end) = 1;\n",
    "model.S(strcmp(model.mets,'adp[c]'),end) = 1;\n",
    "model.S(strcmp(model.mets,'pi[c]'),end) = 1;\n",
    "model.rxns{end+1} = 'phmust-GSH_transport_c_e';\n",
    "model.lb(end+1) = 0;\n",
    "model.ub(end+1) = max(model.ub);\n",
    "model.c(end+1) = 0;\n",
    "model.rules{end+1} = 'x(331)';\n",
    "model.grRules{end+1} = '(1244)';\n",
    "model.rxnGeneMat(end+1,:) = zeros(1,length(model.genes));\n",
    "model.rxnGeneMat(end,331) = 1;\n",
    "model.rxnConfidenceScores(end+1) = 0;\n",
    "model.rxnNames{end+1} = 'Transport of Phosphoramide mustard-GSH from cytosol to extracellular';\n",
    "model.rxnNotes{end+1} = '';\n",
    "model.rxnECNumbers{end+1} = 'X.X.X.X';\n",
    "model.rxnReferences{end+1} = '';\n",
    "model.rxnKEGGID{end+1} = '';\n",
    "model.subSystems{end+1} = 'MODULE CYCLOPHOSPHAMIDE';\n",
    "model.rxnCOG{end+1} = '';\n",
    "model.rxnKeggOrthology{end+1} = '';\n",
    "model.rxnReconMap{end+1} = '';\n",
    "model.rxnFormulas{end+1} = '1 phmust-GSH[c] + 1 atp[c] + 1 h2o[c] --> 1 phmust-GSH[e] + 1 adp[c] + 1 pi[c]';\n",
    "\n",
    "% acr-GSH[c] --> acr-GSH[e]\n",
    "model.S(:,end+1) = zeros(length(model.mets),1);\n",
    "model.S(strcmp(model.mets,'acr-GSH[c]'),end) = -1;\n",
    "model.S(strcmp(model.mets,'atp[c]'),end) = -1;\n",
    "model.S(strcmp(model.mets,'h2o[c]'),end) = -1;\n",
    "model.S(strcmp(model.mets,'acr-GSH[e]'),end) = 1;\n",
    "model.S(strcmp(model.mets,'adp[c]'),end) = 1;\n",
    "model.S(strcmp(model.mets,'pi[c]'),end) = 1;\n",
    "model.rxns{end+1} = 'acr-GSH_transport_c_e';\n",
    "model.lb(end+1) = 0;\n",
    "model.ub(end+1) = max(model.ub);\n",
    "model.c(end+1) = 0;\n",
    "model.rules{end+1} = 'x(331)';\n",
    "model.grRules{end+1} = '(1244)';\n",
    "model.rxnGeneMat(end+1,:) = zeros(1,length(model.genes));\n",
    "model.rxnGeneMat(end,331) = 1;\n",
    "model.rxnConfidenceScores(end+1) = 0;\n",
    "model.rxnNames{end+1} = 'Transport of Acrolein-GSH from cytosol to extracellular';\n",
    "model.rxnNotes{end+1} = '';\n",
    "model.rxnECNumbers{end+1} = 'X.X.X.X';\n",
    "model.rxnReferences{end+1} = '';\n",
    "model.rxnKEGGID{end+1} = '';\n",
    "model.subSystems{end+1} = 'MODULE CYCLOPHOSPHAMIDE';\n",
    "model.rxnCOG{end+1} = '';\n",
    "model.rxnKeggOrthology{end+1} = '';\n",
    "model.rxnReconMap{end+1} = '';\n",
    "model.rxnFormulas{end+1} = '1 acr-GSH[c] + 1 atp[c] + 1 h2o[c] --> 1 acr-GSH[e] + 1 adp[c] + 1 pi[c]';\n",
    "\n",
    "% % % cpa mechanism\n",
    "\n",
    "% cpa[c] --> 4oh-cpa[c]\n",
    "model.S(:,end+1) = zeros(length(model.mets),1);\n",
    "model.S(strcmp(model.mets,'cpa[c]'),end) = -1;\n",
    "model.S(strcmp(model.mets,'o2[c]'),end) = -1;\n",
    "model.S(strcmp(model.mets,'nadph[c]'),end) = -1;\n",
    "model.S(strcmp(model.mets,'h[c]'),end) = -1;\n",
    "model.S(strcmp(model.mets,'4oh-cpa[c]'),end) = 1;\n",
    "model.S(strcmp(model.mets,'h2o[c]'),end) = 1;\n",
    "model.S(strcmp(model.mets,'nadp[c]'),end) = 1;\n",
    "model.rxns{end+1} = 'cpa_4oh-cpa';\n",
    "model.lb(end+1) = 0;\n",
    "model.ub(end+1) = max(model.ub);\n",
    "model.c(end+1) = 0;\n",
    "model.rules{end+1} = 'x(501) | x(504) | x(502) | x(514)';\n",
    "model.grRules{end+1} = '(1555) or (1559) or (1557) or (1576)';\n",
    "model.rxnGeneMat(end+1,:) = zeros(1,length(model.genes));\n",
    "model.rxnGeneMat(end,501) = 1;\n",
    "model.rxnGeneMat(end,504) = 1;\n",
    "model.rxnGeneMat(end,502) = 1;\n",
    "model.rxnGeneMat(end,514) = 1;\n",
    "model.rxnConfidenceScores(end+1) = 0;\n",
    "model.rxnNames{end+1} = 'Cyclophosphamide to 4-hydroxycyclophosphamide';\n",
    "model.rxnNotes{end+1} = '';\n",
    "model.rxnECNumbers{end+1} = '1.14.14.1';\n",
    "model.rxnReferences{end+1} = '';\n",
    "model.rxnKEGGID{end+1} = '';\n",
    "model.subSystems{end+1} = 'MODULE CYCLOPHOSPHAMIDE';\n",
    "model.rxnCOG{end+1} = '';\n",
    "model.rxnKeggOrthology{end+1} = '';\n",
    "model.rxnReconMap{end+1} = '';\n",
    "model.rxnFormulas{end+1} = '1 cpa[c] + 1 o2[c] + 1 nadph[c] + 1 h[c] --> 1 4oh-cpa[c] + 1 h2o[c] + 1 nadp[c]';\n",
    "\n",
    "% 4oh-cpa[e] <=> aldph[c]\n",
    "model.S(:,end+1) = zeros(length(model.mets),1);\n",
    "model.S(strcmp(model.mets,'4oh-cpa[c]'),end) = -1;\n",
    "model.S(strcmp(model.mets,'aldph[c]'),end) = 1;\n",
    "model.rxns{end+1} = '4oh-cpa_aldph';\n",
    "model.lb(end+1) = min(model.lb);\n",
    "model.ub(end+1) = max(model.ub);\n",
    "model.c(end+1) = 0;\n",
    "model.rules{end+1} = '';\n",
    "model.grRules{end+1} = '';\n",
    "model.rxnGeneMat(end+1,:) = zeros(1,length(model.genes));\n",
    "model.rxnConfidenceScores(end+1) = 0;\n",
    "model.rxnNames{end+1} = '4-hydroxycyclophosphamide to aldophosphamide';\n",
    "model.rxnNotes{end+1} = '';\n",
    "model.rxnECNumbers{end+1} = '';\n",
    "model.rxnReferences{end+1} = '';\n",
    "model.rxnKEGGID{end+1} = '';\n",
    "model.subSystems{end+1} = 'MODULE CYCLOPHOSPHAMIDE';\n",
    "model.rxnCOG{end+1} = '';\n",
    "model.rxnKeggOrthology{end+1} = '';\n",
    "model.rxnReconMap{end+1} = '';\n",
    "model.rxnFormulas{end+1} = '1 4oh-cpa[e] <=> 1 aldph[c]';\n",
    "\n",
    "% aldph[c] --> phmust[c]\n",
    "model.S(:,end+1) = zeros(length(model.mets),1);\n",
    "model.S(strcmp(model.mets,'aldph[c]'),end) = -1;\n",
    "model.S(strcmp(model.mets,'phmust[c]'),end) = 1;\n",
    "model.S(strcmp(model.mets,'acr[c]'),end) = 1;\n",
    "model.rxns{end+1} = 'aldph_phmust';\n",
    "model.lb(end+1) = 0;\n",
    "model.ub(end+1) = max(model.ub);\n",
    "model.c(end+1) = 0;\n",
    "model.rules{end+1} = '';\n",
    "model.grRules{end+1} = '';\n",
    "model.rxnGeneMat(end+1,:) = zeros(1,length(model.genes));\n",
    "model.rxnConfidenceScores(end+1) = 0;\n",
    "model.rxnNames{end+1} = 'Aldophosphamide to phosphoramide mustard';\n",
    "model.rxnNotes{end+1} = '';\n",
    "model.rxnECNumbers{end+1} = '';\n",
    "model.rxnReferences{end+1} = '';\n",
    "model.rxnKEGGID{end+1} = '';\n",
    "model.subSystems{end+1} = 'MODULE CYCLOPHOSPHAMIDE';\n",
    "model.rxnCOG{end+1} = '';\n",
    "model.rxnKeggOrthology{end+1} = '';\n",
    "model.rxnReconMap{end+1} = '';\n",
    "model.rxnFormulas{end+1} = '1 aldph[c] --> 1 phmust[c] + 1 acr[c]';\n",
    "\n",
    "% % % cpa metabolism\n",
    "\n",
    "% cpa[c] --> dce-cpa[c]\n",
    "model.S(:,end+1) = zeros(length(model.mets),1);\n",
    "model.S(strcmp(model.mets,'cpa[c]'),end) = -1;\n",
    "model.S(strcmp(model.mets,'o2[c]'),end) = -1;\n",
    "model.S(strcmp(model.mets,'nadph[c]'),end) = -1;\n",
    "model.S(strcmp(model.mets,'h[c]'),end) = -1;\n",
    "model.S(strcmp(model.mets,'dce-cpa[c]'),end) = 1;\n",
    "model.S(strcmp(model.mets,'caa[c]'),end) = 1;\n",
    "model.S(strcmp(model.mets,'h2o[c]'),end) = 1;\n",
    "model.S(strcmp(model.mets,'nadp[c]'),end) = 1;\n",
    "model.rxns{end+1} = 'cpa_dce-cpa';\n",
    "model.lb(end+1) = 0;\n",
    "model.ub(end+1) = max(model.ub);\n",
    "model.c(end+1) = 0;\n",
    "model.rules{end+1} = 'x(514) | x(515)';\n",
    "model.grRules{end+1} = '(1576) or (1577)';\n",
    "model.rxnGeneMat(end+1,:) = zeros(1,length(model.genes));\n",
    "model.rxnGeneMat(end,514) = 1;\n",
    "model.rxnGeneMat(end,515) = 1;\n",
    "model.rxnConfidenceScores(end+1) = 0;\n",
    "model.rxnNames{end+1} = 'Cyclophosphamide to dechloroethylcyclophosphamide';\n",
    "model.rxnNotes{end+1} = '';\n",
    "model.rxnECNumbers{end+1} = '1.14.14.1';\n",
    "model.rxnReferences{end+1} = '';\n",
    "model.rxnKEGGID{end+1} = '';\n",
    "model.subSystems{end+1} = 'MODULE CYCLOPHOSPHAMIDE';\n",
    "model.rxnCOG{end+1} = '';\n",
    "model.rxnKeggOrthology{end+1} = '';\n",
    "model.rxnReconMap{end+1} = '';\n",
    "model.rxnFormulas{end+1} = '1 cpa[c] + 1 o2[c] + 1 nadph[c] + 1 h[c] --> 1 dce-cpa[c] + 1 caa[c] + 1 h2o[c] + 1 nadp[c]';\n",
    "\n",
    "% 4oh-cpa[c] --> 4k-cpa[c]\n",
    "model.S(:,end+1) = zeros(length(model.mets),1);\n",
    "model.S(strcmp(model.mets,'4oh-cpa[c]'),end) = -1;\n",
    "model.S(strcmp(model.mets,'nad[c]'),end) = -1;\n",
    "model.S(strcmp(model.mets,'4k-cpa[c]'),end) = 1;\n",
    "model.S(strcmp(model.mets,'nadh[c]'),end) = 1;\n",
    "model.S(strcmp(model.mets,'h[c]'),end) = 1;\n",
    "model.rxns{end+1} = '4oh-cpa_4k-cpa';\n",
    "model.lb(end+1) = 0;\n",
    "model.ub(end+1) = max(model.ub);\n",
    "model.c(end+1) = 0;\n",
    "model.rules{end+1} = 'x(330) | x(339) | x(344) | x(351) | x(355) | x(363) | x(370)';\n",
    "model.grRules{end+1} = '(124) or (125) or (126) or (127) or (128) or (130) or (131)';\n",
    "model.rxnGeneMat(end+1,:) = zeros(1,length(model.genes));\n",
    "model.rxnGeneMat(end,330) = 1;\n",
    "model.rxnGeneMat(end,339) = 1;\n",
    "model.rxnGeneMat(end,344) = 1;\n",
    "model.rxnGeneMat(end,351) = 1;\n",
    "model.rxnGeneMat(end,355) = 1;\n",
    "model.rxnGeneMat(end,363) = 1;\n",
    "model.rxnGeneMat(end,370) = 1;\n",
    "model.rxnConfidenceScores(end+1) = 0;\n",
    "model.rxnNames{end+1} = '4-Hydroxycyclophosphamide to 4-ketocyclophosphamide';\n",
    "model.rxnNotes{end+1} = '';\n",
    "model.rxnECNumbers{end+1} = '1.1.1.1';\n",
    "model.rxnReferences{end+1} = '';\n",
    "model.rxnKEGGID{end+1} = '';\n",
    "model.subSystems{end+1} = 'MODULE CYCLOPHOSPHAMIDE';\n",
    "model.rxnCOG{end+1} = '';\n",
    "model.rxnKeggOrthology{end+1} = '';\n",
    "model.rxnReconMap{end+1} = '';\n",
    "model.rxnFormulas{end+1} = '1 4oh-cpa[c] + 1 nad[c] --> 1 4k-cpa[c] + 1 nadh[c] + 1 h[c]';\n",
    "\n",
    "% aldph[c] --> alcph[c]\n",
    "model.S(:,end+1) = zeros(length(model.mets),1);\n",
    "model.S(strcmp(model.mets,'aldph[c]'),end) = -1;\n",
    "model.S(strcmp(model.mets,'nadph[c]'),end) = -1;\n",
    "model.S(strcmp(model.mets,'h[c]'),end) = -1;\n",
    "model.S(strcmp(model.mets,'alcph[c]'),end) = 1;\n",
    "model.S(strcmp(model.mets,'nadp[c]'),end) = 1;\n",
    "model.rxns{end+1} = 'aldph_alcph';\n",
    "model.lb(end+1) = 0;\n",
    "model.ub(end+1) = max(model.ub);\n",
    "model.c(end+1) = 0;\n",
    "model.rules{end+1} = 'x(74) | x(814) | x(2190)';\n",
    "model.grRules{end+1} = '(10327) or (231) or (57016)';\n",
    "model.rxnGeneMat(end+1,:) = zeros(1,length(model.genes));\n",
    "model.rxnGeneMat(end,74) = 1;\n",
    "model.rxnGeneMat(end,814) = 1;\n",
    "model.rxnGeneMat(end,2190) = 1;\n",
    "model.rxnConfidenceScores(end+1) = 0;\n",
    "model.rxnNames{end+1} = 'Aldophosphamide to alcophosphamide';\n",
    "model.rxnNotes{end+1} = '';\n",
    "model.rxnECNumbers{end+1} = '1.1.1.21';\n",
    "model.rxnReferences{end+1} = '';\n",
    "model.rxnKEGGID{end+1} = '';\n",
    "model.subSystems{end+1} = 'MODULE CYCLOPHOSPHAMIDE';\n",
    "model.rxnCOG{end+1} = '';\n",
    "model.rxnKeggOrthology{end+1} = '';\n",
    "model.rxnReconMap{end+1} = '';\n",
    "model.rxnFormulas{end+1} = '1 aldph[c] + 1 nadph[c] + 1 h[c] --> 1 alcph[c] + 1 nadp[c]';\n",
    "\n",
    "% aldph[c] --> cph[c]\n",
    "model.S(:,end+1) = zeros(length(model.mets),1);\n",
    "model.S(strcmp(model.mets,'aldph[c]'),end) = -1;\n",
    "model.S(strcmp(model.mets,'nad[c]'),end) = -1;\n",
    "model.S(strcmp(model.mets,'h2o[c]'),end) = -1;\n",
    "model.S(strcmp(model.mets,'cph[c]'),end) = 1;\n",
    "model.S(strcmp(model.mets,'nadh[c]'),end) = 1;\n",
    "model.S(strcmp(model.mets,'h[c]'),end) = 1;\n",
    "model.rxns{end+1} = 'aldph_cph';\n",
    "model.lb(end+1) = 0;\n",
    "model.ub(end+1) = max(model.ub);\n",
    "model.c(end+1) = 0;\n",
    "model.rules{end+1} = 'x(716) | x(728) | x(2709)';\n",
    "model.grRules{end+1} = '(216) or (218) or (7915)';\n",
    "model.rxnGeneMat(end+1,:) = zeros(1,length(model.genes));\n",
    "model.rxnGeneMat(end,716) = 1;\n",
    "model.rxnGeneMat(end,728) = 1;\n",
    "model.rxnGeneMat(end,2709) = 1;\n",
    "model.rxnConfidenceScores(end+1) = 0;\n",
    "model.rxnNames{end+1} = 'Aldophosphamide to carboxyphosphamide';\n",
    "model.rxnNotes{end+1} = '';\n",
    "model.rxnECNumbers{end+1} = '1.2.1.3';\n",
    "model.rxnReferences{end+1} = '';\n",
    "model.rxnKEGGID{end+1} = '';\n",
    "model.subSystems{end+1} = 'MODULE CYCLOPHOSPHAMIDE';\n",
    "model.rxnCOG{end+1} = '';\n",
    "model.rxnKeggOrthology{end+1} = '';\n",
    "model.rxnReconMap{end+1} = '';\n",
    "model.rxnFormulas{end+1} = '1 aldph[c] + 1 nad[c] + 1 h2o[c] --> 1 cph[c] + 1 nadh[c] + 1 h[c]';\n",
    "\n",
    "% phmust[c] --> phmust-GSH[c]\n",
    "model.S(:,end+1) = zeros(length(model.mets),1);\n",
    "model.S(strcmp(model.mets,'phmust[c]'),end) = -1;\n",
    "model.S(strcmp(model.mets,'gthrd[c]'),end) = -1;\n",
    "model.S(strcmp(model.mets,'phmust-GSH[c]'),end) = 1;\n",
    "model.S(strcmp(model.mets,'h[c]'),end) = 1;\n",
    "model.rxns{end+1} = 'phmust_phmust-GSH';\n",
    "model.lb(end+1) = 0;\n",
    "model.ub(end+1) = max(model.ub);\n",
    "model.c(end+1) = 0;\n",
    "model.rules{end+1} = 'x(1175)';\n",
    "model.grRules{end+1} = '(2938)';\n",
    "model.rxnGeneMat(end+1,:) = zeros(1,length(model.genes));\n",
    "model.rxnGeneMat(end,1175) = 1;\n",
    "model.rxnConfidenceScores(end+1) = 0;\n",
    "model.rxnNames{end+1} = 'Phosphoramide mustard to phosphoramide mustard-GSH';\n",
    "model.rxnNotes{end+1} = '';\n",
    "model.rxnECNumbers{end+1} = '2.5.1.18';\n",
    "model.rxnReferences{end+1} = '';\n",
    "model.rxnKEGGID{end+1} = '';\n",
    "model.subSystems{end+1} = 'MODULE CYCLOPHOSPHAMIDE';\n",
    "model.rxnCOG{end+1} = '';\n",
    "model.rxnKeggOrthology{end+1} = '';\n",
    "model.rxnReconMap{end+1} = '';\n",
    "model.rxnFormulas{end+1} = '1 phmust[c] + 1 gthrd[c] --> 1 phmust-GSH + 1 h[c]';\n",
    "\n",
    "% acr[c] --> acr-GSH[c]\n",
    "model.S(:,end+1) = zeros(length(model.mets),1);\n",
    "model.S(strcmp(model.mets,'acr[c]'),end) = -1;\n",
    "model.S(strcmp(model.mets,'gthrd[c]'),end) = -1;\n",
    "model.S(strcmp(model.mets,'acr-GSH[c]'),end) = 1;\n",
    "model.S(strcmp(model.mets,'h[c]'),end) = 1;\n",
    "model.rxns{end+1} = 'acr_acr-GSH';\n",
    "model.lb(end+1) = 0;\n",
    "model.ub(end+1) = max(model.ub);\n",
    "model.c(end+1) = 0;\n",
    "model.rules{end+1} = 'x(1175)';\n",
    "model.grRules{end+1} = '(2938)';\n",
    "model.rxnGeneMat(end+1,:) = zeros(1,length(model.genes));\n",
    "model.rxnGeneMat(end,1175) = 1;\n",
    "model.rxnConfidenceScores(end+1) = 0;\n",
    "model.rxnNames{end+1} = 'Acrolein to acrolein-GSH';\n",
    "model.rxnNotes{end+1} = '';\n",
    "model.rxnECNumbers{end+1} = '2.5.1.18';\n",
    "model.rxnReferences{end+1} = '';\n",
    "model.rxnKEGGID{end+1} = '';\n",
    "model.subSystems{end+1} = 'MODULE CYCLOPHOSPHAMIDE';\n",
    "model.rxnCOG{end+1} = '';\n",
    "model.rxnKeggOrthology{end+1} = '';\n",
    "model.rxnReconMap{end+1} = '';\n",
    "model.rxnFormulas{end+1} = '1 acr[c] + 1 gthrd[c] --> 1 acr-GSH + 1 h[c]';"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Redox cont."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Remove wrong genes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n"
     ]
    }
   ],
   "source": [
    "% PRDX1 and PRDX2\n",
    "i = find(strcmp(model.rxns,'GTHP'));\n",
    "model.rxnGeneMat(i,strcmp(model.genes,'5052')) = 0;\n",
    "model.rxnGeneMat(i,strcmp(model.genes,'7001')) = 0;\n",
    "model.rules{i} = 'x(1147) | x(1148) | x(1150)';\n",
    "model.grRules{i} =  '(2876) or (2877) or (2879)';\n",
    "\n",
    "% PRDX3\n",
    "i = find(strcmp(model.rxns,'GTHPm'));\n",
    "model.rxnGeneMat(i,strcmp(model.genes,'10935')) = 0;\n",
    "model.rules{i} = 'x(1147) | x(1150)';\n",
    "model.grRules{i} =  '(2876) or (2879)';\n",
    "\n",
    "% PRDX6\n",
    "i = find(strcmp(model.rxns,'HMR_0960'));\n",
    "model.rxnGeneMat(i,strcmp(model.genes,'9588')) = 0;\n",
    "model.rules{i} = 'x(954) | x(1147) | x(1148) | x(1149) | x(1150) | x(1151) | x(1152) | x(1619)';\n",
    "model.grRules{i} =  '(257202) or (2876) or (2877) or (2878) or (2879) or (2880) or (2882) or (493869)';\n",
    "\n",
    "i = find(strcmp(model.rxns,'HMR_0963'));\n",
    "model.rxnGeneMat(i,strcmp(model.genes,'9588')) = 0;\n",
    "model.rules{i} = 'x(954) | x(1147) | x(1148) | x(1149) | x(1150) | x(1151) | x(1152) | x(1619)';\n",
    "model.grRules{i} =  '(257202) or (2876) or (2877) or (2878) or (2879) or (2880) or (2882) or (493869)';\n",
    "\n",
    "i = find(strcmp(model.rxns,'HMR_0988'));\n",
    "model.rxnGeneMat(i,strcmp(model.genes,'9588')) = 0;\n",
    "model.rules{i} = 'x(954) | x(1147) | x(1148) | x(1149) | x(1150) | x(1151) | x(1152) | x(1619)';\n",
    "model.grRules{i} =  '(257202) or (2876) or (2877) or (2878) or (2879) or (2880) or (2882) or (493869)';\n",
    "\n",
    "i = find(strcmp(model.rxns,'HMR_2441'));\n",
    "model.rxnGeneMat(i,strcmp(model.genes,'9588')) = 0;\n",
    "model.rules{i} = 'x(954) | x(1147) | x(1148) | x(1149) | x(1150) | x(1151) | x(1152) | x(1619)';\n",
    "model.grRules{i} =  '(257202) or (2876) or (2877) or (2878) or (2879) or (2880) or (2882) or (493869)';\n",
    "\n",
    "i = find(strcmp(model.rxns,'HMR_1048'));\n",
    "model.rxnGeneMat(i,strcmp(model.genes,'9588')) = 0;\n",
    "model.rules{i} = 'x(954) | x(1147) | x(1148) | x(1149) | x(1150) | x(1151) | x(1152) | x(1619)';\n",
    "model.grRules{i} =  '(257202) or (2876) or (2877) or (2878) or (2879) or (2880) or (2882) or (493869)';\n",
    "\n",
    "i = find(strcmp(model.rxns,'HMR_1049'));\n",
    "model.rxnGeneMat(i,strcmp(model.genes,'9588')) = 0;\n",
    "model.rules{i} = 'x(954) | x(1147) | x(1148) | x(1149) | x(1150) | x(1151) | x(1152) | x(1619)';\n",
    "model.grRules{i} =  '(257202) or (2876) or (2877) or (2878) or (2879) or (2880) or (2882) or (493869)';"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Metabolites"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n"
     ]
    }
   ],
   "source": [
    "% Pr-SH[c]\n",
    "model.S(end+1,:) = zeros(1,length(model.rxns));\n",
    "model.mets{end+1} = 'Pr-SH[c]';\n",
    "model.b(end+1) = 0;\n",
    "model.csense(end+1) = 'E';\n",
    "model.metCharges(end+1) = 0;\n",
    "model.metFormulas{end+1} = 'X';\n",
    "model.metSmiles{end+1} = '';\n",
    "model.metNames{end+1} = 'Protein, reduced';\n",
    "model.metHMDBID{end+1} = '';\n",
    "model.metInChIString{end+1} = '';\n",
    "model.metKEGGID{end+1} = '';\n",
    "model.metPubChemID{end+1} = '';\n",
    "model.metCHEBIID{end+1} = '';\n",
    "model.metPdMap{end+1} = '';\n",
    "model.metReconMap{end+1} = '';\n",
    "\n",
    "% Pr-SSG[c]\n",
    "model.S(end+1,:) = zeros(1,length(model.rxns));\n",
    "model.mets{end+1} = 'Pr-SSG[c]';\n",
    "model.b(end+1) = 0;\n",
    "model.csense(end+1) = 'E';\n",
    "model.metCharges(end+1) = 0;\n",
    "model.metFormulas{end+1} = 'X';\n",
    "model.metSmiles{end+1} = '';\n",
    "model.metNames{end+1} = 'Protein, glutathionylated';\n",
    "model.metHMDBID{end+1} = '';\n",
    "model.metInChIString{end+1} = '';\n",
    "model.metKEGGID{end+1} = '';\n",
    "model.metPubChemID{end+1} = '';\n",
    "model.metCHEBIID{end+1} = '';\n",
    "model.metPdMap{end+1} = '';\n",
    "model.metReconMap{end+1} = '';\n",
    "\n",
    "% Pr-SS[c]\n",
    "model.S(end+1,:) = zeros(1,length(model.rxns));\n",
    "model.mets{end+1} = 'Pr-SS[c]';\n",
    "model.b(end+1) = 0;\n",
    "model.csense(end+1) = 'E';\n",
    "model.metCharges(end+1) = 0;\n",
    "model.metFormulas{end+1} = 'X';\n",
    "model.metSmiles{end+1} = '';\n",
    "model.metNames{end+1} = 'Protein, oxidized';\n",
    "model.metHMDBID{end+1} = '';\n",
    "model.metInChIString{end+1} = '';\n",
    "model.metKEGGID{end+1} = '';\n",
    "model.metPubChemID{end+1} = '';\n",
    "model.metCHEBIID{end+1} = '';\n",
    "model.metPdMap{end+1} = '';\n",
    "model.metReconMap{end+1} = '';\n",
    "\n",
    "% Pr-SH[m]\n",
    "model.S(end+1,:) = zeros(1,length(model.rxns));\n",
    "model.mets{end+1} = 'Pr-SH[m]';\n",
    "model.b(end+1) = 0;\n",
    "model.csense(end+1) = 'E';\n",
    "model.metCharges(end+1) = 0;\n",
    "model.metFormulas{end+1} = 'X';\n",
    "model.metSmiles{end+1} = '';\n",
    "model.metNames{end+1} = 'Protein, reduced';\n",
    "model.metHMDBID{end+1} = '';\n",
    "model.metInChIString{end+1} = '';\n",
    "model.metKEGGID{end+1} = '';\n",
    "model.metPubChemID{end+1} = '';\n",
    "model.metCHEBIID{end+1} = '';\n",
    "model.metPdMap{end+1} = '';\n",
    "model.metReconMap{end+1} = '';\n",
    "\n",
    "% Pr-SSG[m]\n",
    "model.S(end+1,:) = zeros(1,length(model.rxns));\n",
    "model.mets{end+1} = 'Pr-SSG[c]';\n",
    "model.b(end+1) = 0;\n",
    "model.csense(end+1) = 'E';\n",
    "model.metCharges(end+1) = 0;\n",
    "model.metFormulas{end+1} = 'X';\n",
    "model.metSmiles{end+1} = '';\n",
    "model.metNames{end+1} = 'Protein, glutathionylated';\n",
    "model.metHMDBID{end+1} = '';\n",
    "model.metInChIString{end+1} = '';\n",
    "model.metKEGGID{end+1} = '';\n",
    "model.metPubChemID{end+1} = '';\n",
    "model.metCHEBIID{end+1} = '';\n",
    "model.metPdMap{end+1} = '';\n",
    "model.metReconMap{end+1} = '';\n",
    "\n",
    "% Pr-SS[m]\n",
    "model.S(end+1,:) = zeros(1,length(model.rxns));\n",
    "model.mets{end+1} = 'Pr-SS[m]';\n",
    "model.b(end+1) = 0;\n",
    "model.csense(end+1) = 'E';\n",
    "model.metCharges(end+1) = 0;\n",
    "model.metFormulas{end+1} = 'X';\n",
    "model.metSmiles{end+1} = '';\n",
    "model.metNames{end+1} = 'Protein, oxidized';\n",
    "model.metHMDBID{end+1} = '';\n",
    "model.metInChIString{end+1} = '';\n",
    "model.metKEGGID{end+1} = '';\n",
    "model.metPubChemID{end+1} = '';\n",
    "model.metCHEBIID{end+1} = '';\n",
    "model.metPdMap{end+1} = '';\n",
    "model.metReconMap{end+1} = '';"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Reactions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n"
     ]
    }
   ],
   "source": [
    "% 1 h2o2[c] + 1 trdrd[c] --> 2 h2o[c] + 1 trdox[c]\n",
    "model.S(:,end+1) = zeros(length(model.mets),1);\n",
    "model.S(strcmp(model.mets,'h2o2[c]'),end) = -1;\n",
    "model.S(strcmp(model.mets,'trdrd[c]'),end) = -1;\n",
    "model.S(strcmp(model.mets,'h2o[c]'),end) = 2;\n",
    "model.S(strcmp(model.mets,'trdox[c]'),end) = 1;\n",
    "model.rxns{end+1} = 'Prx_c';\n",
    "model.lb(end+1) = 0;\n",
    "model.ub(end+1) = max(model.ub);\n",
    "model.c(end+1) = 0;\n",
    "model.rules{end+1} = 'x(1646) | x(2590)';\n",
    "model.grRules{end+1} =  '(5052) or (7001)';\n",
    "model.rxnGeneMat(end+1,:) = zeros(1,length(model.genes));\n",
    "model.rxnGeneMat(end,strcmp(model.genes,'5052')) = 1;\n",
    "model.rxnGeneMat(end,strcmp(model.genes,'7001')) = 1;\n",
    "model.rxnConfidenceScores(end+1) = 0;\n",
    "model.rxnNames{end+1} = 'Peroxiredoxin, cytoplasmic';\n",
    "model.rxnNotes{end+1} = '';\n",
    "model.rxnECNumbers{end+1} = '1.11.1.15';\n",
    "model.rxnReferences{end+1} = '';\n",
    "model.rxnKEGGID{end+1} = '';\n",
    "model.subSystems{end+1} = 'ROS detoxification';\n",
    "model.rxnCOG{end+1} = '';\n",
    "model.rxnKeggOrthology{end+1} = '';\n",
    "model.rxnReconMap{end+1} = '';\n",
    "model.rxnFormulas{end+1} = '1 h2o2[c] + 1 trdrd[c] --> 2 h2o[c] + 1 trdox[c]';\n",
    "\n",
    "% 1 h2o2[m] + 1 trdrd[m] --> 2 h2o[m] + 1 trdox[m]\n",
    "model.S(:,end+1) = zeros(length(model.mets),1);\n",
    "model.S(strcmp(model.mets,'h2o2[m]'),end) = -1;\n",
    "model.S(strcmp(model.mets,'trdrd[m]'),end) = -1;\n",
    "model.S(strcmp(model.mets,'h2o[m]'),end) = 2;\n",
    "model.S(strcmp(model.mets,'trdox[m]'),end) = 1;\n",
    "model.rxns{end+1} = 'Prx_m';\n",
    "model.lb(end+1) = 0;\n",
    "model.ub(end+1) = max(model.ub);\n",
    "model.c(end+1) = 0;\n",
    "model.rules{end+1} = 'x(168)';\n",
    "model.grRules{end+1} =  '(10935)';\n",
    "model.rxnGeneMat(end+1,:) = zeros(1,length(model.genes));\n",
    "model.rxnGeneMat(end,strcmp(model.genes,'10935')) = 1;\n",
    "model.rxnConfidenceScores(end+1) = 0;\n",
    "model.rxnNames{end+1} = 'Peroxiredoxin, mitochondrial';\n",
    "model.rxnNotes{end+1} = '';\n",
    "model.rxnECNumbers{end+1} = '1.11.1.15';\n",
    "model.rxnReferences{end+1} = '';\n",
    "model.rxnKEGGID{end+1} = '';\n",
    "model.subSystems{end+1} = 'ROS detoxification';\n",
    "model.rxnCOG{end+1} = '';\n",
    "model.rxnKeggOrthology{end+1} = '';\n",
    "model.rxnReconMap{end+1} = '';\n",
    "model.rxnFormulas{end+1} = '1 h2o2[m] + 1 trdrd[m] --> 2 h2o[m] + 1 trdox[m]';\n",
    "\n",
    "% 1 h2o2[c] + 1 Pr-SH[c] + 1 gthrd[c] --> 2 h2o[c] + 1 Pr-SSG[c]\n",
    "model.S(:,end+1) = zeros(length(model.mets),1);\n",
    "model.S(strcmp(model.mets,'h2o2[c]'),end) = -1;\n",
    "model.S(strcmp(model.mets,'Pr-SH[c]'),end) = -1;\n",
    "model.S(strcmp(model.mets,'gthrd[c]'),end) = -1;\n",
    "model.S(strcmp(model.mets,'h2o[c]'),end) = 2;\n",
    "model.S(strcmp(model.mets,'Pr-SSG[c]'),end) = 1;\n",
    "model.rxns{end+1} = 'Protein_thiol_SSG';\n",
    "model.lb(end+1) = 0;\n",
    "model.ub(end+1) = max(model.ub);\n",
    "model.c(end+1) = 0;\n",
    "model.rules{end+1} = '';\n",
    "model.grRules{end+1} =  '';\n",
    "model.rxnGeneMat(end+1,:) = zeros(1,length(model.genes));\n",
    "model.rxnConfidenceScores(end+1) = 0;\n",
    "model.rxnNames{end+1} = 'Protein thiol oxidation, SSG';\n",
    "model.rxnNotes{end+1} = '';\n",
    "model.rxnECNumbers{end+1} = '';\n",
    "model.rxnReferences{end+1} = '';\n",
    "model.rxnKEGGID{end+1} = '';\n",
    "model.subSystems{end+1} = 'ROS detoxification';\n",
    "model.rxnCOG{end+1} = '';\n",
    "model.rxnKeggOrthology{end+1} = '';\n",
    "model.rxnReconMap{end+1} = '';\n",
    "model.rxnFormulas{end+1} = '1 h2o2[c] + 1 Pr-SH[c] + 1 gthrd[c] --> 2 h2o[c] + 1 Pr-SSG[c]';\n",
    "\n",
    "% 1 Pr-SSG[c] + 1 gthrd[c] --> 1 Pr-SH[c] + 1 gthox[c]\n",
    "model.S(:,end+1) = zeros(length(model.mets),1);\n",
    "model.S(strcmp(model.mets,'Pr-SSG[c]'),end) = -1;\n",
    "model.S(strcmp(model.mets,'gthrd[c]'),end) = -1;\n",
    "model.S(strcmp(model.mets,'Pr-SH[c]'),end) = 1;\n",
    "model.S(strcmp(model.mets,'gthox[c]'),end) = 1;\n",
    "model.rxns{end+1} = 'Grx_c';\n",
    "model.lb(end+1) = 0;\n",
    "model.ub(end+1) = max(model.ub);\n",
    "model.c(end+1) = 0;\n",
    "model.rules{end+1} = 'x(1096)';\n",
    "model.grRules{end+1} =  '(2745)';\n",
    "model.rxnGeneMat(end+1,:) = zeros(1,length(model.genes));\n",
    "model.rxnGeneMat(end,strcmp(model.genes,'2745')) = 1;\n",
    "model.rxnConfidenceScores(end+1) = 0;\n",
    "model.rxnNames{end+1} = 'Glutaredoxin, mitochondrial';\n",
    "model.rxnNotes{end+1} = '';\n",
    "model.rxnECNumbers{end+1} = '1.20.4.1';\n",
    "model.rxnReferences{end+1} = '';\n",
    "model.rxnKEGGID{end+1} = '';\n",
    "model.subSystems{end+1} = 'ROS detoxification';\n",
    "model.rxnCOG{end+1} = '';\n",
    "model.rxnKeggOrthology{end+1} = '';\n",
    "model.rxnReconMap{end+1} = '';\n",
    "model.rxnFormulas{end+1} = '1 Pr-SSG[c] + 1 gthrd[c] --> 1 Pr-SH[c] + 1 gthox[c]';\n",
    "\n",
    "% 1 h2o2[c] + 1 Pr-SH[c] --> 2 h2o[c] + 1 Pr-SS[c]\n",
    "model.S(:,end+1) = zeros(length(model.mets),1);\n",
    "model.S(strcmp(model.mets,'h2o2[c]'),end) = -1;\n",
    "model.S(strcmp(model.mets,'Pr-SH[c]'),end) = -1;\n",
    "model.S(strcmp(model.mets,'h2o[c]'),end) = 2;\n",
    "model.S(strcmp(model.mets,'Pr-SS[c]'),end) = 1;\n",
    "model.rxns{end+1} = 'Protein_thiol_SS';\n",
    "model.lb(end+1) = 0;\n",
    "model.ub(end+1) = max(model.ub);\n",
    "model.c(end+1) = 0;\n",
    "model.rules{end+1} = '';\n",
    "model.grRules{end+1} =  '';\n",
    "model.rxnGeneMat(end+1,:) = zeros(1,length(model.genes));\n",
    "model.rxnConfidenceScores(end+1) = 0;\n",
    "model.rxnNames{end+1} = 'Protein thiol oxidation, SS';\n",
    "model.rxnNotes{end+1} = '';\n",
    "model.rxnECNumbers{end+1} = '';\n",
    "model.rxnReferences{end+1} = '';\n",
    "model.rxnKEGGID{end+1} = '';\n",
    "model.subSystems{end+1} = 'ROS detoxification';\n",
    "model.rxnCOG{end+1} = '';\n",
    "model.rxnKeggOrthology{end+1} = '';\n",
    "model.rxnReconMap{end+1} = '';\n",
    "model.rxnFormulas{end+1} = '1 h2o2[c] + 1 Pr-SH[c] --> 2 h2o[c] + 1 Pr-SS[c]';\n",
    "\n",
    "% 1 Pr-SS[c] + 1 trdrd[c] --> 1 Pr-SH[c] + 1 trdox[c]\n",
    "model.S(:,end+1) = zeros(length(model.mets),1);\n",
    "model.S(strcmp(model.mets,'Pr-SS[c]'),end) = -1;\n",
    "model.S(strcmp(model.mets,'trdrd[c]'),end) = -1;\n",
    "model.S(strcmp(model.mets,'Pr-SH[c]'),end) = 1;\n",
    "model.S(strcmp(model.mets,'trdox[c]'),end) = 1;\n",
    "model.rxns{end+1} = 'Protein_thiol_reduction_c';\n",
    "model.lb(end+1) = 0;\n",
    "model.ub(end+1) = max(model.ub);\n",
    "model.c(end+1) = 0;\n",
    "model.rules{end+1} = '';\n",
    "model.grRules{end+1} =  '';\n",
    "model.rxnGeneMat(end+1,:) = zeros(1,length(model.genes));\n",
    "model.rxnConfidenceScores(end+1) = 0;\n",
    "model.rxnNames{end+1} = 'Protein thiol reduction by thioredoxin';\n",
    "model.rxnNotes{end+1} = '';\n",
    "model.rxnECNumbers{end+1} = '';\n",
    "model.rxnReferences{end+1} = '';\n",
    "model.rxnKEGGID{end+1} = '';\n",
    "model.subSystems{end+1} = 'ROS detoxification';\n",
    "model.rxnCOG{end+1} = '';\n",
    "model.rxnKeggOrthology{end+1} = '';\n",
    "model.rxnReconMap{end+1} = '';\n",
    "model.rxnFormulas{end+1} = '1 Pr-SS[c] + 1 trdrd[c] --> 1 Pr-SH[c] + 1 trdox[c]';\n",
    "\n",
    "% 1 h2o2[m] + 1 Pr-SH[m] + 1 gthrd[m] --> 2 h2o[m] + 1 Pr-SSG[m]\n",
    "model.S(:,end+1) = zeros(length(model.mets),1);\n",
    "model.S(strcmp(model.mets,'h2o2[m]'),end) = -1;\n",
    "model.S(strcmp(model.mets,'Pr-SH[m]'),end) = -1;\n",
    "model.S(strcmp(model.mets,'gthrd[m]'),end) = -1;\n",
    "model.S(strcmp(model.mets,'h2o[m]'),end) = 2;\n",
    "model.S(strcmp(model.mets,'Pr-SSG[m]'),end) = 1;\n",
    "model.rxns{end+1} = 'Protein_thiol_SSG_m';\n",
    "model.lb(end+1) = 0;\n",
    "model.ub(end+1) = max(model.ub);\n",
    "model.c(end+1) = 0;\n",
    "model.rules{end+1} = '';\n",
    "model.grRules{end+1} =  '';\n",
    "model.rxnGeneMat(end+1,:) = zeros(1,length(model.genes));\n",
    "model.rxnConfidenceScores(end+1) = 0;\n",
    "model.rxnNames{end+1} = 'Protein thiol oxidation, SSG';\n",
    "model.rxnNotes{end+1} = '';\n",
    "model.rxnECNumbers{end+1} = '';\n",
    "model.rxnReferences{end+1} = '';\n",
    "model.rxnKEGGID{end+1} = '';\n",
    "model.subSystems{end+1} = 'ROS detoxification';\n",
    "model.rxnCOG{end+1} = '';\n",
    "model.rxnKeggOrthology{end+1} = '';\n",
    "model.rxnReconMap{end+1} = '';\n",
    "model.rxnFormulas{end+1} = '1 h2o2[m] + 1 Pr-SH[m] + 1 gthrd[m] --> 2 h2o[m] + 1 Pr-SSG[m]';\n",
    "\n",
    "% 1 Pr-SSG[m] + 1 gthrd[m] --> 1 Pr-SH[m] + 1 gthox[m]\n",
    "model.S(:,end+1) = zeros(length(model.mets),1);\n",
    "model.S(strcmp(model.mets,'Pr-SSG[m]'),end) = -1;\n",
    "model.S(strcmp(model.mets,'gthrd[m]'),end) = -1;\n",
    "model.S(strcmp(model.mets,'Pr-SH[m]'),end) = 1;\n",
    "model.S(strcmp(model.mets,'gthox[m]'),end) = 1;\n",
    "model.rxns{end+1} = 'Grx_m';\n",
    "model.lb(end+1) = 0;\n",
    "model.ub(end+1) = max(model.ub);\n",
    "model.c(end+1) = 0;\n",
    "model.rules{end+1} = 'x(1670)';\n",
    "model.grRules{end+1} =  '(51022)';\n",
    "model.rxnGeneMat(end+1,:) = zeros(1,length(model.genes));\n",
    "model.rxnGeneMat(end,strcmp(model.genes,'51022')) = 1;\n",
    "model.rxnConfidenceScores(end+1) = 0;\n",
    "model.rxnNames{end+1} = 'Glutaredoxin, mitochondrial';\n",
    "model.rxnNotes{end+1} = '';\n",
    "model.rxnECNumbers{end+1} = '1.20.4.1';\n",
    "model.rxnReferences{end+1} = '';\n",
    "model.rxnKEGGID{end+1} = '';\n",
    "model.subSystems{end+1} = 'ROS detoxification';\n",
    "model.rxnCOG{end+1} = '';\n",
    "model.rxnKeggOrthology{end+1} = '';\n",
    "model.rxnReconMap{end+1} = '';\n",
    "model.rxnFormulas{end+1} = '1 Pr-SSG[m] + 1 gthrd[m] --> 1 Pr-SH[m] + 1 gthox[m]';\n",
    "\n",
    "% 1 h2o2[m] + 1 Pr-SH[m] --> 2 h2o[m] + 1 Pr-SS[m]\n",
    "model.S(:,end+1) = zeros(length(model.mets),1);\n",
    "model.S(strcmp(model.mets,'h2o2[m]'),end) = -1;\n",
    "model.S(strcmp(model.mets,'Pr-SH[m]'),end) = -1;\n",
    "model.S(strcmp(model.mets,'h2o[m]'),end) = 2;\n",
    "model.S(strcmp(model.mets,'Pr-SS[m]'),end) = 1;\n",
    "model.rxns{end+1} = 'Protein_thiol_SS_m';\n",
    "model.lb(end+1) = 0;\n",
    "model.ub(end+1) = max(model.ub);\n",
    "model.c(end+1) = 0;\n",
    "model.rules{end+1} = '';\n",
    "model.grRules{end+1} =  '';\n",
    "model.rxnGeneMat(end+1,:) = zeros(1,length(model.genes));\n",
    "model.rxnConfidenceScores(end+1) = 0;\n",
    "model.rxnNames{end+1} = 'Protein thiol oxidation, SS';\n",
    "model.rxnNotes{end+1} = '';\n",
    "model.rxnECNumbers{end+1} = '';\n",
    "model.rxnReferences{end+1} = '';\n",
    "model.rxnKEGGID{end+1} = '';\n",
    "model.subSystems{end+1} = 'ROS detoxification';\n",
    "model.rxnCOG{end+1} = '';\n",
    "model.rxnKeggOrthology{end+1} = '';\n",
    "model.rxnReconMap{end+1} = '';\n",
    "model.rxnFormulas{end+1} = '1 h2o2[m] + 1 Pr-SH[m] --> 2 h2o[m] + 1 Pr-SS[m]';\n",
    "\n",
    "% 1 Pr-SS[m] + 1 trdrd[m] --> 1 Pr-SH[m] + 1 trdox[m]\n",
    "model.S(:,end+1) = zeros(length(model.mets),1);\n",
    "model.S(strcmp(model.mets,'Pr-SS[m]'),end) = -1;\n",
    "model.S(strcmp(model.mets,'trdrd[m]'),end) = -1;\n",
    "model.S(strcmp(model.mets,'Pr-SH[m]'),end) = 1;\n",
    "model.S(strcmp(model.mets,'trdox[m]'),end) = 1;\n",
    "model.rxns{end+1} = 'Protein_thiol_reduction_m';\n",
    "model.lb(end+1) = 0;\n",
    "model.ub(end+1) = max(model.ub);\n",
    "model.c(end+1) = 0;\n",
    "model.rules{end+1} = '';\n",
    "model.grRules{end+1} =  '';\n",
    "model.rxnGeneMat(end+1,:) = zeros(1,length(model.genes));\n",
    "model.rxnConfidenceScores(end+1) = 0;\n",
    "model.rxnNames{end+1} = 'Protein thiol reduction by thioredoxin, mitochondrial';\n",
    "model.rxnNotes{end+1} = '';\n",
    "model.rxnECNumbers{end+1} = '';\n",
    "model.rxnReferences{end+1} = '';\n",
    "model.rxnKEGGID{end+1} = '';\n",
    "model.subSystems{end+1} = 'ROS detoxification';\n",
    "model.rxnCOG{end+1} = '';\n",
    "model.rxnKeggOrthology{end+1} = '';\n",
    "model.rxnReconMap{end+1} = '';\n",
    "model.rxnFormulas{end+1} = '1 Pr-SS[m] + 1 trdrd[m] --> 1 Pr-SH[m] + 1 trdox[m]';"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Catalase reactions involving NADPH"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n"
     ]
    }
   ],
   "source": [
    "% fix CATr genes and EC number\n",
    "model.rules{strcmp(model.rxns,'CATr')} = model.rules{strcmp(model.rxns,'r0010')};\n",
    "model.grRules{strcmp(model.rxns,'CATr')} = model.grRules{strcmp(model.rxns,'r0010')};\n",
    "model.rxnGeneMat(strcmp(model.rxns,'CATr'),:) = model.rxnGeneMat(strcmp(model.rxns,'r0010'),:);\n",
    "model.rxnECNumbers{strcmp(model.rxns,'CATr')} = model.rxnECNumbers{strcmp(model.rxns,'r0010')};"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n"
     ]
    }
   ],
   "source": [
    "reaction_names = {'r0010','CATm','CATp','CATr'};\n",
    "compartment_names = {'[c]','[m]','[x]','[r]'};\n",
    "for i = 1:length(reaction_names)\n",
    "    model.S(:,end+1) = model.S(:,strcmp(model.rxns,reaction_names{i}));\n",
    "    model.S(strcmp(model.mets,strcat('nadph',compartment_names{i})),end) = -0.1282;\n",
    "    model.S(strcmp(model.mets,strcat('nadp',compartment_names{i})),end) = 0.1282;\n",
    "    model.S(strcmp(model.mets,strcat('h',compartment_names{i})),end) = 0.1282;\n",
    "    model.rxns{end+1} = strcat(reaction_names{i},'_nadph');\n",
    "    model.lb(end+1) = model.lb(strcmp(model.rxns,reaction_names{i}));\n",
    "    model.ub(end+1) = model.ub(strcmp(model.rxns,reaction_names{i}));\n",
    "    model.c(end+1) = model.c(strcmp(model.rxns,reaction_names{i}));\n",
    "    model.rules{end+1} = model.rules{strcmp(model.rxns,reaction_names{i})};\n",
    "    model.grRules{end+1} =  model.grRules{strcmp(model.rxns,reaction_names{i})};\n",
    "    model.rxnGeneMat(end+1,:) = model.rxnGeneMat(strcmp(model.rxns,reaction_names{i}),:);\n",
    "    model.rxnConfidenceScores(end+1) = model.rxnConfidenceScores(strcmp(model.rxns,reaction_names{i}));\n",
    "    model.rxnNames{end+1} = model.rxnNames{strcmp(model.rxns,reaction_names{i})};\n",
    "    model.rxnNotes{end+1} = model.rxnNotes{strcmp(model.rxns,reaction_names{i})};\n",
    "    model.rxnECNumbers{end+1} = model.rxnECNumbers{strcmp(model.rxns,reaction_names{i})};\n",
    "    model.rxnReferences{end+1} = model.rxnReferences{strcmp(model.rxns,reaction_names{i})};\n",
    "    model.rxnKEGGID{end+1} = model.rxnKEGGID{strcmp(model.rxns,reaction_names{i})};\n",
    "    model.subSystems{end+1} = model.subSystems{strcmp(model.rxns,reaction_names{i})};\n",
    "    model.rxnCOG{end+1} = model.rxnCOG{strcmp(model.rxns,reaction_names{i})};\n",
    "    model.rxnKeggOrthology{end+1} = model.rxnKeggOrthology{strcmp(model.rxns,reaction_names{i})};\n",
    "    model.rxnReconMap{end+1} = model.rxnReconMap{strcmp(model.rxns,reaction_names{i})};\n",
    "    model.rxnFormulas{end+1} = strcat(model.rxnFormulas{strcmp(model.rxns,reaction_names{i})}(1:9),' + 0.1282 nadph',compartment_names{i},model.rxnFormulas{strcmp(model.rxns,reaction_names{i})}(10:end),' + 0.1282 nadp',compartment_names{i},' + 0.1282 h',compartment_names{i});\n",
    "end"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Save model file "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 64,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n"
     ]
    }
   ],
   "source": [
    "#load('recon3d_qflux.mat','model');"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 74,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n"
     ]
    }
   ],
   "source": [
    "save('recon3d_qflux.mat','model');"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Create information files"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Create reaction information file"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 75,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n"
     ]
    }
   ],
   "source": [
    "f = fopen('reactions.tsv','w');\n",
    "fprintf(f,'RXNID\\tNAME\\tFORMULA\\tSUBSYSTEM\\tGR RULE\\tEC NUMBER\\tKEGG\\tNOTES\\tREFERENCES\\n');\n",
    "for i = 1:length(model.rxns)\n",
    "    fprintf(f,'%s\\t%s\\t%s\\t%s\\t%s\\t%s\\t%s\\t%s\\t%s\\n',model.rxns{i},model.rxnNames{i},model.rxnFormulas{i},model.subSystems{i},model.grRules{i},model.rxnECNumbers{i},model.rxnKEGGID{i},model.rxnNotes{i},model.rxnReferences{i});\n",
    "end\n",
    "fclose(f);"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Create metabolite information file"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 76,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n"
     ]
    }
   ],
   "source": [
    "f = fopen('metabolites.tsv','w');\n",
    "fprintf(f,'METID\\tNAME\\tFORMULA\\tCHARGE\\tCHEBI\\tKEGG\\tPUBCHEM\\tHMDB\\tINCHI STRING\\tSMILE\\n');\n",
    "for i = 1:length(model.mets)\n",
    "    fprintf(f,'%s\\t%s\\t%s\\t%d\\t%s\\t%s\\t%s\\t%s\\t%s\\t%s\\n',model.mets{i},model.metNames{i},model.metFormulas{i},model.metCharges(i),model.metCHEBIID{i},model.metKEGGID{i},model.metPubChemID{i},model.metHMDBID{i},model.metInChIString{i},model.metSmiles{i});\n",
    "end\n",
    "fclose(f);"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Create gene information file"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 77,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n"
     ]
    }
   ],
   "source": [
    "f = fopen('genes.tsv','w');\n",
    "fprintf(f,'GENEID\\tSYMBOL\\n');\n",
    "for i = 1:length(model.genes)\n",
    "    fprintf(f,'%s\\t%s\\n',model.genes{i},model.geneSymbols{i});\n",
    "end\n",
    "fclose(f);"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Matlab",
   "language": "matlab",
   "name": "matlab"
  },
  "language_info": {
   "codemirror_mode": "octave",
   "file_extension": ".m",
   "help_links": [
    {
     "text": "MetaKernel Magics",
     "url": "https://metakernel.readthedocs.io/en/latest/source/README.html"
    }
   ],
   "mimetype": "text/x-octave",
   "name": "matlab",
   "version": "0.16.7"
  },
  "toc": {
   "nav_menu": {},
   "number_sections": true,
   "sideBar": true,
   "skip_h1_title": true,
   "title_cell": "Table of Contents",
   "title_sidebar": "Contents",
   "toc_cell": true,
   "toc_position": {},
   "toc_section_display": true,
   "toc_window_display": false
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
